---
title: "GLM and Permutations for Contrasts 0,1"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: hide
  html_notebook:
    code_folding: hide
---

```{r, warning=FALSE, message=FALSE, comment=FALSE, fig.width=18, fig.height=5}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)

require(cybrBSA)


#TESTING
#glm_formula <- "Allele ~ Bulk * Parent + Rep"

#ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)))

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))

```

## Functions for GLM all and Permutation all

```{r}
glm_cb2_all <- function(..., W, formula, numgroups = FALSE, outputlength = 8) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  output <- summary(glm_fit)$coefficients[c(((length(summary(glm_fit)$coefficients)*0)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.25)),
                                            ((length(summary(glm_fit)$coefficients)*0.5)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.75)))]

  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
} 

```

```{r}
################################################################################
# No replicates, fix Parent
cybrPermute_cb2_all <- function(dataset, 
                                R = 0, perp = 1, outlength = 4, 
                                glmform = "Allele ~ Bulk * Parent", 
                                inputlabels = c("Intercept", "Bulk", "Parent", "Interaction")){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled aD")
  print(paste("Your labels are:", inputlabels))

    if(R > 0){ #INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Rep, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == 1) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest    
    }else{ #DO NOT INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == perp) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    }
  
    #PERMUTE TWICE
  if(R > 0){
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Rep, 
             Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep, 
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, 
             Rep, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep,
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }else{
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #RUN THE GLM
    if(R > 0) {
      shuffletoglm2 %>% na.omit() %>%
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }else{
      shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }
    
    
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}


```

# Load In GLM Prep Data

```{r, eval = FALSE}
#CuSO4 CSS1
HNGLVDRXY_prep <- readRDS("StandardGLM/HNGLVDRXY_CSS1_glm_prep.rds") %>% 
  mutate(Bulk = gsub("Unselected", "aD", Bulk)) %>% mutate_if(is.character,as.factor)

HKTFTDRX2_prep <- readRDS("StandardGLM/HKTFTDRX2_CSS1_glm_prep.rds")%>% mutate_if(is.character,as.factor)

#CuSO4 CSS8
HVYTYDRX2_prep <- readRDS("StandardGLM/HVYTYDRX2_CuSO4_CSS8_glm_prep.rds")%>% mutate_if(is.character,as.factor) %>% select(-Dose)

HTG3TDMXY_prep <- readRDS("StandardGLM/HTG3TDMXY_CSS8_glm_prep.rds") %>% 
  mutate(DoseCol = Bulk) %>%
  select(-Bulk) %>% 
  select(DoseCol = DoseCol, Bulk = B, Allele = Allele, POS = POS, Parent = Parent, SmoothCount = SmoothCount, CHROM = CHROM, Rep = Rep, Pool = Pool) %>% na.omit() %>% 
  filter(CHROM != "VIII", CHROM != "M") %>% mutate_if(is.character,as.factor)

```


### Define contrasts for each

```{r, eval = FALSE}
contrasts(HNGLVDRXY_prep$Allele) <- matrix(c(1, 0), ncol = 1)
contrasts(HNGLVDRXY_prep$Bulk) <- matrix(c(1, 0), ncol = 1)
contrasts(HNGLVDRXY_prep$Parent) <- matrix(c(1, 0), ncol = 1)

contrasts(HKTFTDRX2_prep$Allele) <- matrix(c(1, 0), ncol = 1)
contrasts(HKTFTDRX2_prep$Bulk) <- matrix(c(1, 0), ncol = 1)
contrasts(HKTFTDRX2_prep$Parent) <- matrix(c(1, 0), ncol = 1)

contrasts(HVYTYDRX2_prep$Allele) <- matrix(c(1, 0), ncol = 1)
contrasts(HVYTYDRX2_prep$Bulk) <- matrix(c(1, 0), ncol = 1)
contrasts(HVYTYDRX2_prep$Parent) <- matrix(c(1, 0), ncol = 1)

contrasts(HTG3TDMXY_prep$Allele) <- matrix(c(1, 0), ncol = 1)
contrasts(HTG3TDMXY_prep$Bulk) <- matrix(c(1, 0), ncol = 1)
contrasts(HTG3TDMXY_prep$Parent) <- matrix(c(1, 0), ncol = 1)

```


# Run Permutations

### HNGLVDRXY_CSS1 (CuSO4 CSS I A)
```{r, eval = FALSE}

HNGLVDRXY_parentonly <- cybrPermute_cb2_all(HNGLVDRXY_prep)


HNGLVDRXY_parentonly %>% group_by(Factor) %>%
  summarize(quant095 = quantile(abs(Summary), 0.95, na.rm = TRUE),
             quant099 = quantile(abs(Summary), 0.99, na.rm = TRUE),
             quant0995 = quantile(abs(Summary), 0.995, na.rm = TRUE)) -> HNGLVDRXY_allpQs

saveRDS(HNGLVDRXY_allpQs, file = "StandardGLM/Ec01_HNGLVDRXY_allpQs.rds")

```


### CuSO4 2024, no replicates

```{r, eval = FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}

HTG3TDMXY_CSS8_parentonly <- cybrPermute_cb2_all(HTG3TDMXY_prep)

HTG3TDMXY_CSS8_parentonly %>% group_by(Factor) %>%
 summarize(quant095 = quantile(abs(Summary), 0.95, na.rm = TRUE),
            quant099 = quantile(abs(Summary), 0.99, na.rm = TRUE),
            quant0995 = quantile(abs(Summary), 0.995, na.rm = TRUE)) -> HTG3TDMXY_allpQs

saveRDS(HTG3TDMXY_allpQs, file = "StandardGLM/Ec01_HTG3TDMXY_allpQs.rds")
```

### HKTFTDRX2 (CuSO4)

Redoing this 1/8/24

```{r, eval = FALSE}
unique(HKTFTDRX2_prep$Allele)

HKTFTDRX2_parentonly <- cybrPermute_cb2_all(HKTFTDRX2_prep, perp = 2)


HKTFTDRX2_parentonly %>% group_by(Factor) %>%
  summarize(quant095 = quantile(abs(Summary), 0.95, na.rm = TRUE),
             quant099 = quantile(abs(Summary), 0.99, na.rm = TRUE),
             quant0995 = quantile(abs(Summary), 0.995, na.rm = TRUE)) -> HKTFTDRX2_allpQs

saveRDS(HKTFTDRX2_allpQs, file = "StandardGLM/Ec01_HKTFTDRX2_allpQs.rds")

```

### CSS8
REDO THIS FOR CSS8: 1/9/24 - running

```{r, eval = FALSE}

HVYTYDRX2_cu_prep_parentonly <- cybrPermute_cb2_all(HVYTYDRX2_prep, perp = 2)

HVYTYDRX2_cu_prep_parentonly %>% na.omit()

HVYTYDRX2_cu_prep_parentonly %>% group_by(Factor) %>%
  summarize(quant095 = quantile(abs(Summary), 0.95, na.rm = TRUE),
             quant099 = quantile(abs(Summary), 0.99, na.rm = TRUE),
             quant0995 = quantile(abs(Summary), 0.995, na.rm = TRUE)) -> HVYTYDRX2_allpQs

saveRDS(HVYTYDRX2_allpQs, file = "StandardGLM/Ec01_HVYTYDRX2_allpQs.rds")

# HVYTYDRX2_prep[,c("CHROM", "Allele", "Bulk", "Parent", "SmoothCount")]
# HTG3TDMXY_prep[,c("CHROM", "Allele", "Bulk", "Parent", "SmoothCount")]
# str(HVYTYDRX2_prep)
# str(HTG3TDMXY_prep)
```

```{r, eval = FALSE}

HTG3TDMXY_parentonly_rep <- cybrPermute_cb2_all(HTG3TDMXY_prep, perp = 2, R = 2,
                                                glmform = "Allele ~ Bulk*Parent + Rep",
                                                inputlabels = c("cept", "Bulk", "Parent", "Rep", "Interaction"))

HTG3TDMXY_parentonly_rep %>% group_by(Factor) %>%
  summarize(quant095 = quantile(abs(Summary), 0.95, na.rm = TRUE),
             quant099 = quantile(abs(Summary), 0.99, na.rm = TRUE),
             quant0995 = quantile(abs(Summary), 0.995, na.rm = TRUE)) -> HTG3TDMXY_allpQs_rep

saveRDS(HTG3TDMXY_allpQs_rep, file = "StandardGLM/Ec01_HTG3TDMXY_allpQs_rep.rds")


```

# Run the GLMs on these same samples with the contrasts at 1 and 0

### HNV

```{r, eval = FALSE}
inputlabels <- c("cept", "Bulk", "Parent", "Interaction")
HVYTYDRX2_prep %>% na.omit() %>% 
      group_by(Pool, CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(
                                    Allele = Allele, Bulk = Bulk, Parent = Parent,#Rep = Rep,
                                    W = SmoothCount,
                                    formula = "Allele ~ Bulk * Parent",
                                    outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> HVYTYDRX2_glm_Ec

saveRDS(HVYTYDRX2_glm_Ec, file = "HVYTYDRX2_glm_Ec01.rds")
```
### HNG
```{r, eval = FALSE}
inputlabels <- c("cept", "Bulk", "Parent", "Interaction")
HNGLVDRXY_prep %>% na.omit() %>% 
      group_by(Pool, CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(
                                    Allele = Allele, Bulk = Bulk, Parent = Parent, #Rep = Rep,
                                    W = SmoothCount,
                                    formula = "Allele ~ Bulk * Parent",
                                    outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> HNGLVDRXY_glm_Ec

saveRDS(HNGLVDRXY_glm_Ec, file = "HNGLVDRXY_glm_Ec01.rds")

```

### HKT

```{r, eval = FALSE}
inputlabels <- c("cept", "Bulk", "Parent", "Interaction")
HKTFTDRX2_prep %>% na.omit() %>% 
      group_by(Pool, CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(
                                    Allele = Allele, Bulk = Bulk, Parent = Parent,#Rep = Rep,
                                    W = SmoothCount,
                                    formula = "Allele ~ Bulk * Parent",
                                    outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> HKTFTDRX2_glm_Ec

saveRDS(HKTFTDRX2_glm_Ec, file = "HKTFTDRX2_glm_Ec01.rds")
```

### HTG
```{r, eval = FALSE}
inputlabels <- c("cept", "Bulk", "Parent", "Interaction")
HTG3TDMXY_prep %>% na.omit() %>% 
      group_by(Pool, CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(
                                    Allele = Allele, Bulk = Bulk, Parent = Parent,#Rep = Rep,
                                    W = SmoothCount,
                                    formula = "Allele ~ Bulk * Parent",
                                    outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> HTG3TDMXY_glm_Ec

saveRDS(HTG3TDMXY_glm_Ec, file = "HTG3TDMXY_glm_Ec01.rds")
```

Replicates

```{r, eval = FALSE}
inputlabels <- c("cept", "Bulk", "Parent", "Interaction", "Rep1", "Rep2")

HTG3TDMXY_prep %>% na.omit() %>%
      group_by(Pool, CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk, 
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = "Allele ~ Bulk * Parent + Rep",
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> HTG3TDMXY_glm_Ec_rep

saveRDS(HTG3TDMXY_glm_Ec_rep, file = "HTG3TDMXY_glm_Ec01_rep.rds")
```

# Plotting to Compare

## Load pre-run permutations

```{r}
HNGLVDRXY_allpQs <- readRDS("StandardGLM/Ec01_HNGLVDRXY_allpQs.rds")
HTG3TDMXY_allpQs <- readRDS("StandardGLM/Ec01_HTG3TDMXY_allpQs.rds")
HKTFTDRX2_allpQs <- readRDS("StandardGLM/Ec01_HKTFTDRX2_allpQs.rds")
HVYTYDRX2_allpQs <- readRDS("StandardGLM/Ec01_HVYTYDRX2_allpQs.rds")
HTG3TDMXY_allpQs_rep <- readRDS("StandardGLM/Ec01_HTG3TDMXY_allpQs_rep.rds")

HVYTYDRX2_glm_Ec <- readRDS("HVYTYDRX2_glm_Ec01.rds")
HNGLVDRXY_glm_Ec <- readRDS("HNGLVDRXY_glm_Ec01.rds")
HKTFTDRX2_glm_Ec <- readRDS("HKTFTDRX2_glm_Ec01.rds")
HTG3TDMXY_glm_Ec <- readRDS("HTG3TDMXY_glm_Ec01.rds")
HTG3TDMXY_glm_Ec_rep <- readRDS("HTG3TDMXY_glm_Ec01_rep.rds")

```

Effects vs Z-scores for rep vs no rep
```{r, fig.width = 12, fig.height = 5, eval = FALSE}

HTG3TDMXY_glm_Ec_rep %>% merge(HTG3TDMXY_allpQs_rep) %>% 
  mutate(intpeaks = POS*(abs(Summary) > quant095)*(Factor == "Interaction")) %>%
  mutate(bulkpeaks = POS*(abs(Summary) > quant095)*(Factor == "Bulk")) %>%
  filter(d == "Z") %>%

  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  #geom_vline(aes(xintercept = bulkpeaks), color = "lightpink", alpha = 0.2) +
  geom_vline(aes(xintercept = intpeaks), color = "lightgreen", alpha = 0.8) +
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  geom_line() + facet_grid(d~CHROM, space = "free", scales = "free") + ggtitle("Z-score w Replicates")

################################################################################
HTG3TDMXY_glm_Ec %>% merge(HTG3TDMXY_allpQs) %>% 
  mutate(intpeaks = POS*(abs(Summary) > quant095)*(Factor == "Interaction")) %>%
  mutate(bulkpeaks = POS*(abs(Summary) > quant095)*(Factor == "Bulk")) %>%
  filter(d == "Z") %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  #geom_vline(aes(xintercept = bulkpeaks), color = "lightpink", alpha = 0.2) +
  geom_vline(aes(xintercept = intpeaks), color = "lightgreen", alpha = 0.8) +
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  geom_line() + facet_grid(d~CHROM, space = "free", scales = "free") + ggtitle("Z-score No replicates")

```

### Effects

```{r, fig.width = 12, fig.height = 5, eval = FALSE}
HNGLVDRXY_glm_Ec %>% merge(HNGLVDRXY_allpQs) %>% filter(CHROM != "I") %>%
  mutate(intpeaks = POS*(abs(Summary) > quant095)*(Factor == "Interaction")) %>%
  mutate(bulkpeaks = POS*(abs(Summary) > quant095)*(Factor == "Bulk")) %>%
  filter(d == "Effect") %>%

  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_vline(aes(xintercept = bulkpeaks), color = "lightpink", alpha = 0.2) +
  geom_vline(aes(xintercept = intpeaks), color = "lightblue", alpha = 0.8) +
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  
  geom_line() + facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("HNGLVDRXY Effect")

################################################################################
HKTFTDRX2_glm_Ec %>% merge(HKTFTDRX2_allpQs) %>% filter(CHROM != "I") %>%
  mutate(intpeaks = POS*(abs(Summary) > quant095)*(Factor == "Interaction")) %>%
  mutate(bulkpeaks = POS*(abs(Summary) > quant095)*(Factor == "Bulk")) %>%
  filter(d == "Effect") %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_vline(aes(xintercept = bulkpeaks), color = "lightpink", alpha = 0.2) +
  geom_vline(aes(xintercept = intpeaks), color = "lightblue", alpha = 0.8) +
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  geom_line() + facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("HKTFTDRX2 Effect ")

################################################################################
HVYTYDRX2_glm_Ec %>% merge(HVYTYDRX2_allpQs) %>% filter(CHROM != "VIII") %>%
  mutate(intpeaks = POS*(abs(Summary) > quant095)*(Factor == "Interaction")) %>%
  mutate(bulkpeaks = POS*(abs(Summary) > quant095)*(Factor == "Bulk")) %>%
  filter(d == "Effect") %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_vline(aes(xintercept = bulkpeaks), color = "lightpink", alpha = 0.2) +
  geom_vline(aes(xintercept = intpeaks), color = "lightblue", alpha = 0.8) +
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  geom_line() + facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("HVYTYDRX2 Effect")

```

### Z-scores

```{r, fig.width = 12, fig.height = 5, eval = FALSE}
HNGLVDRXY_glm_Ec %>% merge(HNGLVDRXY_allpQs) %>% filter(CHROM != "I") %>%
  mutate(intpeaks = POS*(abs(Summary) > quant095)*(Factor == "Interaction")) %>%
  mutate(bulkpeaks = POS*(abs(Summary) > quant095)*(Factor == "Bulk")) %>%
  filter(d == "Z") %>%

  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_vline(aes(xintercept = bulkpeaks), color = "lightpink", alpha = 0.2) +
  geom_vline(aes(xintercept = intpeaks), color = "lightgreen", alpha = 0.8) +
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  
  geom_line() + facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("HNGLVDRXY Z")

HKTFTDRX2_glm_Ec %>% merge(HKTFTDRX2_allpQs) %>% filter(CHROM != "I") %>%
  mutate(intpeaks = POS*(abs(Summary) > quant095)*(Factor == "Interaction")) %>%
  mutate(bulkpeaks = POS*(abs(Summary) > quant095)*(Factor == "Bulk")) %>%
  filter(d == "Z") %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_vline(aes(xintercept = bulkpeaks), color = "lightpink", alpha = 0.2) +
  geom_vline(aes(xintercept = intpeaks), color = "lightgreen", alpha = 0.8) +
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  geom_line() + facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("HKTFTDRX2 Z ")

HVYTYDRX2_glm_Ec %>% merge(HVYTYDRX2_allpQs) %>% filter(CHROM != "VIII") %>%
  mutate(intpeaks = POS*(abs(Summary) > quant095)*(Factor == "Interaction")) %>%
  mutate(bulkpeaks = POS*(abs(Summary) > quant095)*(Factor == "Bulk")) %>%
  filter(d == "Z") %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_vline(aes(xintercept = bulkpeaks), color = "lightpink", alpha = 0.2) +
  geom_vline(aes(xintercept = intpeaks), color = "lightgreen", alpha = 0.8) +
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  geom_line() + facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("HVYTYDRX2 Z")

```

### All at once?

```{r, fig.width = 12, fig.height = 5}
rbind(HTG3TDMXY_glm_Ec[HTG3TDMXY_glm_Ec$CHROM != "VIII",], 
      #HTG3TDMXY_glm_Ec_rep[HTG3TDMXY_glm_Ec_rep$CHROM != "VIII",],
      HNGLVDRXY_glm_Ec[HNGLVDRXY_glm_Ec$CHROM != "I",],
      HKTFTDRX2_glm_Ec[HKTFTDRX2_glm_Ec$CHROM != "I",],
      HVYTYDRX2_glm_Ec[HVYTYDRX2_glm_Ec$CHROM != "VIII",]) -> CuSO4_contrastglm

CuSO4_contrastglm %>% filter(d == "Effect", Factor != "cept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + facet_grid(Pool ~ CHROM, space = "free", scales = "free") +
  ggtitle("Effects, 0/1")

CuSO4_contrastglm %>% filter(d == "Z", Factor != "cept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + facet_grid(Pool ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, 0/1")

```

All at once WITH cutoffs?

```{r, fig.width = 12, fig.height = 5}
rbind(#merge(HTG3TDMXY_glm_Ec[HTG3TDMXY_glm_Ec$CHROM != "VIII",], HTG3TDMXY_allpQs), 
      merge(HTG3TDMXY_glm_Ec_rep[HTG3TDMXY_glm_Ec_rep$CHROM != "VIII",], HTG3TDMXY_allpQs_rep), 
      merge(HNGLVDRXY_glm_Ec[HNGLVDRXY_glm_Ec$CHROM != "I",], HNGLVDRXY_allpQs),
      merge(HKTFTDRX2_glm_Ec[HKTFTDRX2_glm_Ec$CHROM != "I",], HKTFTDRX2_allpQs),
      merge(HVYTYDRX2_glm_Ec[HVYTYDRX2_glm_Ec$CHROM != "VIII",], HVYTYDRX2_allpQs)) -> CuSO4_contrastglm_cutoffs

IDkey <- data.frame(Pool = c("HTG3TDMXY", "HNGLVDRXY", "HKTFTDRX2", "HVYTYDRX2"),
                    CSS = c("VIII", "I", "I", "VIII"),
                    Label = c("CSS8 H", "CSS1 H", "CSS1 L", "CSS8 L"),
                    Selection = c("High", "High", "Low", "Low"))

CuSO4_contrastglm_cutoffs %>% merge(IDkey) %>% filter(Selection == "Low") -> CuSO4_contrastglm_cutoffs

# CuSO4_contrastglm_cutoffs %>% filter(d == "Effect", Factor != "cept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
#   geom_line() + facet_grid(Pool ~ CHROM, space = "free", scales = "free") +
#   ggtitle("Effects, 0/1, Interaction Peaks")

# CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept") %>%
#   mutate(peak = POS * as.numeric(Factor == "Interaction" & abs(Summary) > quant095)) %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
#   geom_vline(aes(xintercept = peak), color = "lightgreen", alpha = 0.01) + 
#   geom_hline(aes(yintercept = quant095, color = Factor)) +
#   geom_line() + facet_grid(Label ~ CHROM, space = "free", scales = "free") +
#   ggtitle("Z scores, 0/1, Interaction Peaks")
# 
# CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept") %>%
#   mutate(peak = POS * as.numeric(Factor == "Bulk" & abs(Summary) > quant095)) %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
#   geom_vline(aes(xintercept = peak), color = "salmon", alpha = 0.01) + 
#   geom_hline(aes(yintercept = quant095, color = Factor)) +
#   geom_line() + facet_grid(Label ~ CHROM, space = "free", scales = "free") +
#   ggtitle("Z scores, 0/1, Bulk Peaks")
# 
# CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept") %>%
#   mutate(peak = POS * as.numeric(Factor == "Parent" & abs(Summary) > quant095)) %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
#   geom_vline(aes(xintercept = peak), color = "lightblue", alpha = 0.01) + 
#   geom_hline(aes(yintercept = quant095, color = Factor)) +
#   geom_line() + facet_grid(Label ~ CHROM, space = "free", scales = "free") +
#   ggtitle("Z scores, 0/1, Parent Peaks")

```

# Peak Calling

```{r, fig.width = 12, fig.height = 5}

CuSO4_contrastglm_cutoffs %>% group_by(Pool, Label, Factor, CHROM, d, CSS) %>% arrange(POS) %>%
  na.omit() %>%
  mutate(is_above_threshold = abs(Summary) > quant095) %>%
  
  mutate(segment_change = is_above_threshold != lag(is_above_threshold, default = FALSE)) %>%
  mutate(segment_id = cumsum(segment_change)) %>%
  # Group by chromosome and segment_id to count the points in each segment
  group_by(Pool, Label, Factor, CHROM, d, CSS, segment_id) %>%
  mutate(points_in_segment = n()) %>%
  # Now filter to keep only those segments that are above the threshold and have 5 or more points
  filter(is_above_threshold, points_in_segment >= 100) %>%
 # mutate(change = is_above_threshold != lag(is_above_threshold, default = first(is_above_threshold))) %>%
  #mutate(group_id = cumsum(change)) %>%
  #group_by(Pool, Factor, CHROM, d, group_id) %>%
 # filter(is_above_threshold) %>%
  summarize(max_value = max(abs(Summary), na.rm = TRUE),
            ppeakPOS = POS[which.max(abs(Summary))]) %>%
  ungroup() -> maybepeaks

CuSO4_contrastglm_cutoffs %>% group_by(Pool, Label, Factor, CHROM, d, CSS) %>% arrange(POS) %>%
  na.omit() %>%
  mutate(is_above_threshold = abs(Summary) > quant095) %>%
  filter(is_above_threshold == TRUE) %>%
  summarize(max_value = max(abs(Summary), na.rm = TRUE),
            ppeakPOS = POS[which.max(abs(Summary))]) -> peaks

maybepeaks %>% filter(Factor == "Interaction", d == "Z")
#left_join(CuSO4_contrastglm_cutoffs, maybepeaks) -> CuSO4_contrastglm_peaks

CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept", Factor != "Parent") %>% ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_vline(data = maybepeaks[maybepeaks$Factor == "Interaction" & maybepeaks$d == "Z",], aes(xintercept = ppeakPOS), color = "purple", linewidth = 2, alpha = 0.3) +
  geom_hline(aes(yintercept = quant095)) +
  geom_point(size = 0.2, alpha = 0.5) + facet_grid(Label ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, 0/1, Interaction Peaks") +
  scale_color_manual(values = c("black", "purple3", "gray"))

CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept", Factor != "Parent") %>% ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_vline(data = maybepeaks[maybepeaks$Factor == "Bulk" & maybepeaks$d == "Z",], aes(xintercept = ppeakPOS), color = "black", linewidth = 2, alpha = 0.3) +
  geom_hline(aes(yintercept = quant095)) +
  geom_point(size = 0.2, alpha = 0.5) + facet_grid(Label ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, 0/1, Bulk Peaks") +
  scale_color_manual(values = c("black", "purple3", "gray"))

# CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept") %>% ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
#   geom_vline(aes(xintercept = ppeakPOS*as.numeric(Factor == "Bulk" & d == "Z")), color = "black", linewidth = 2, alpha = 0.3) +
#   geom_line() + facet_grid(Label ~ CHROM, space = "free", scales = "free") +
#   ggtitle("Z scores, 0/1, Bulk Peaks")

# CuSO4_contrastglm_cutoffs %>% filter(Factor == "Interaction"& d == "Z") %>% ggplot(aes(x = POS, y = abs(Summary))) + geom_point(data = maybepeaks, aes(x = ppeakPOS, y = max_value, color = Factor)) +
#   geom_hline(data = HTG3TDMXY_allpQs_rep, aes(yintercept = quant095, color = Factor)) + facet_grid( ~ CHROM, space = "free", scales = "free")
```

```{r, fig.width = 12, fig.height = 5}

# CuSO4_contrastglm_cutoffs %>% group_by(CSS, CHROM, POS) %>% 
#   mutate(peak = as.numeric(abs(Summary) > quant095)) %>%
#   summarize(doublepeak = sum(peak) > 1) %>%
#   filter(doublepeak == TRUE) -> doublepeaks
#   
#  CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk") %>% 
#   ggplot(aes(x = POS, y = abs(Summary), color = Label)) + 
#   geom_hline(aes(yintercept = quant095, color = Label)) +
#    geom_vline(data = doublepeaks, aes(xintercept = POS, color = CSS), alpha = 0.01) +
#   geom_point(size = 0.2, alpha = 0.1) + 
#   facet_grid(CSS~ CHROM, space = "free", scales = "free") +
#   ggtitle("Z scores, 0/1, Bulk Peaks") +
#   scale_color_manual(values = c("navy", "navy", "darkorange", "darkorange", "navy", "orange"))

# CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk", Label %in% c("CSS1 A", "CSS8 B")) %>% 
#   ggplot(aes(x = POS, y = abs(Summary), color = Label)) + 
#   #geom_vline(data = maybepeaks[maybepeaks$Factor == "Bulk" & maybepeaks$d == "Z",], aes(xintercept = ppeakPOS), linewidth = 2, alpha = 0.3) +
#   geom_hline(aes(yintercept = quant095, color = Label)) +
#   #geom_line() + 
#   geom_point(size = 0.2, alpha = 0.1) + 
#   facet_grid(~ CHROM, space = "free", scales = "free") +
#   ggtitle("High Selection") +
#   scale_color_manual(values = c("navy",  "darkorange", "darkorange"))
# 
# CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk", Label %in% c("CSS1 B", "CSS8 A")) %>% 
#   ggplot(aes(x = POS, y = abs(Summary), color = Label)) + 
#   #geom_vline(data = maybepeaks[maybepeaks$Factor == "Bulk" & maybepeaks$d == "Z",], aes(xintercept = ppeakPOS), linewidth = 2, alpha = 0.3) +
#   geom_hline(aes(yintercept = quant095, color = Label)) +
#   #geom_line() + 
#   geom_point(size = 0.2, alpha = 0.1) + 
#   facet_grid(~ CHROM, space = "free", scales = "free") +
#   ggtitle("Low Selection") +
#   scale_color_manual(values = c("navy",  "darkorange", "darkorange"))

#All Additive Effects AND Agreement Peaks

CuSO4_contrastglm_cutoffs %>% group_by(CHROM, POS, d, Factor) %>% 
  mutate(peak = as.numeric(abs(Summary) > quant095)) %>%
  summarize(doublepeak = sum(peak) > 3) %>%
  filter(doublepeak == TRUE) -> quadpeaks

CuSO4_contrastglm_cutoffs %>% filter(CHROM %in% c("I", "VIII")) %>% group_by(CHROM, POS, d, Factor) %>% 
  mutate(peak = as.numeric(abs(Summary) > quant095)) %>%
  summarize(doublepeak = sum(peak) > 1) %>%
  filter(doublepeak == TRUE) -> quadpeaks_CSS

quadpeaks <- rbind(quadpeaks_CSS, quadpeaks)

CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_vline(data = quadpeaks[quadpeaks$Factor == "Bulk" & quadpeaks$d == "Z",], aes(xintercept = POS), linewidth = 2, alpha = 0.3, color = "turquoise") +
  geom_hline(aes(yintercept = quant095)) +
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid( ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  ylab("abs Z Score") +
  theme(legend.position = "none")

#################################################################################

```

```{r, fig.width = 12, fig.height = 5}

#CHANGE THESE TO PUT BACK THE HIGH BULKS
maybepeaks %>% filter(Label != "Css8 H", Label != "CSS1 H") -> maybepeaks

################################################################################

#Comparison of Selection
CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Selection)) + 
  geom_hline(aes(yintercept = quant095, color = Selection)) +
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid(CSS~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  scale_color_manual(values = c("darkturquoise",  "orange")) +
  ylab("abs Z Score") +
  theme(legend.position = "none")

#Comparison of CSS
CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + 
  geom_hline(aes(yintercept = quant095, color = CSS)) +
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid( ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  scale_color_manual(values = c("navy",  "darkorange")) +
  ylab("abs Z Score") +
  theme(legend.position = "none")

#Comparison of CSS but separated out
CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + 
  geom_hline(aes(yintercept = quant095, color = CSS)) +
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid(CSS ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  scale_color_manual(values = c("navy",  "darkorange")) +
  ylab("abs Z Score") +
  theme(legend.position = "none")

#Comparison of CSS but then colored by Selection
CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Selection)) + 
  geom_hline(aes(yintercept = quant095, color = Selection)) +
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid(CSS ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  scale_color_manual(values = c("black",  "darkturquoise")) +
  ylab("abs Z Score") +
  theme(legend.position = "none")

#All Additive Effects
CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) +
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid( ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  ylab("abs Z Score") +
  theme(legend.position = "none")

#ALL EFFECTS
CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid( ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  scale_color_manual(values = c("#000000",  "purple", "#80808020")) +
  ylab("abs Z Score") +
  theme(legend.position = "none")


#INTERACTION EFFECTS
CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept", Factor != "Parent") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  geom_vline(data = maybepeaks[maybepeaks$Factor == "Interaction" & maybepeaks$d == "Z",], aes(xintercept = ppeakPOS), color = "purple", linewidth = 2, alpha = 0.3) +

  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid(CSS ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  scale_color_manual(values = c("#00000020",  "purple")) +
  ylab("abs Z Score") +
  theme(legend.position = "none")

#INTERACTION EFFECTS
# CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept", Factor != "Parent") %>% 
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
#   geom_hline(aes(yintercept = quant095, color = Factor)) +
#   geom_vline(data = maybepeaks[maybepeaks$Factor == "Interaction" & maybepeaks$d == "Z",], aes(xintercept = ppeakPOS, color = CSS), linewidth = 2, alpha = 0.3) +
# 
#   geom_point(size = 0.4, alpha = 0.1) + 
#   facet_grid(CSS ~ CHROM, space = "free", scales = "free") +
#   ggtitle("Z scores, Selection")  +
#   scale_color_manual(values = c("gray",  "#24588A", "purple", "#ED7B01")) +
#   ylab("abs Z Score") +
#   theme(legend.position = "none")

CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept", Factor != "Parent") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  # geom_vline(data = maybepeaks[maybepeaks$Factor == "Interaction" & maybepeaks$d == "Z",], 
  #            aes(xintercept = ppeakPOS, color = CSS), linewidth = 2, alpha = 0.3) +
  geom_vline(data = peaks[peaks$d == "Z" & peaks$Factor == "Interaction",], aes(xintercept = ppeakPOS, color = CSS), size = 2, alpha = 0.4) +
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid(CSS ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  scale_color_manual(values = c("black",  "#24588A", "purple", "#ED7B01")) +
  ylab("abs Z Score") +
  theme(legend.position = "none")

#INT ONLY
CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept", Factor == "Interaction") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  geom_hline(aes(yintercept = quant095, color = Factor)) +
  # geom_vline(data = maybepeaks[maybepeaks$Factor == "Interaction" & maybepeaks$d == "Z",], 
  #            aes(xintercept = ppeakPOS, color = CSS), linewidth = 2, alpha = 0.3) +
  geom_vline(data = peaks[peaks$d == "Z" & peaks$Factor == "Interaction",], aes(xintercept = ppeakPOS, color = CSS), size = 2, alpha = 0.4) +
  geom_point(size = 1, alpha = 0.1) + 
  facet_grid(CSS ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  scale_color_manual(values = c("#24588A", "purple", "#ED7B01")) +
  ylab("abs Z Score") +
  theme(legend.position = "none")

#INT ONLY
CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor != "cept", Factor == "Interaction") %>% 
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) +
  # geom_vline(data = maybepeaks[maybepeaks$Factor == "Interaction" & maybepeaks$d == "Z",], 
  #            aes(xintercept = ppeakPOS, color = CSS), linewidth = 2, alpha = 0.3) +
  geom_vline(data = peaks[peaks$d == "Z" & peaks$Factor == "Interaction",], aes(xintercept = ppeakPOS, color = CSS), size = 2, alpha = 0.4) +
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid(CSS ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  scale_color_manual(values = c("purple", "#ED7B01")) +
  ylab("abs Z Score") +
  theme(legend.position = "none")


#Additive wtih genes

#genes <- c("CRS5", "SOD1", "CUP1-1", "CUP1-2", "MET17", "MUP1", "PCL1", "OYE3", "FRE1", "FRE7", "CTR1", "IRC7")
genes <- c("CUP1-1", "CUP1-2", "FRE1")

sgd_orfs <- read.csv("SGD_ORFs.csv")
sgd_orfs %>% filter(Gene.symbol %in% genes) %>%
  transmute(Gene = Gene.symbol,
            POS = Gene.chromosomeLocation.start,
            CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_copper

sgd_copper %>% arrange(CHROM)

CuSO4_contrastglm_cutoffs %>% filter(d == "Z", Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_vline(data = quadpeaks[quadpeaks$Factor == "Bulk" & quadpeaks$d == "Z",], aes(xintercept = POS), linewidth = 2, alpha = 0.3, color = "gray90") +
  geom_hline(aes(yintercept = quant095)) +
  geom_vline(data = sgd_copper, aes(xintercept = POS), size = 2, alpha = 0.4, color = "red") + 
  geom_point(size = 0.4, alpha = 0.1) + 
  facet_grid( ~ CHROM, space = "free", scales = "free") +
  ggtitle("Z scores, Selection")  +
  ylab("abs Z Score") +
  theme(legend.position = "none")

```

### Circos Plots????

```{r}
cybr_circos <- function(d1, d8, peaklist1 = NULL, peaklist8 = NULL, maxy = NULL, color1 = "#24588A", color8 = "#ED7B01", opacity = "50"){
  
  color1fade_50 <- paste(color1, "50", sep = "")
  color8fade_50 <- paste(color8, "50", sep = "")
  
  color1fade <- paste(color1, opacity, sep = "")
  color8fade <- paste(color8, opacity, sep = "")

  if(is.null(d8)){
    d8 <- d1
    include8 <- FALSE
  }else{
    include8 <- TRUE
  }
    #SET UP DATA
  df8 <- data.frame(sectors = as.character(d8$CHROM),
                 x = d8$POS,
                 y = abs(d8$summary),
                  label = d8$label)

  df1 <- data.frame(sectors = as.character(d1$CHROM),
                 x = d1$POS,
                 y = abs(d1$summary),
                  label = d1$label)

  #Take just the interactions of each
  df8int <- df8 %>% filter(label == "Interaction")
  df1int <- df1 %>% filter(label == "Interaction")

    #REORDER THE CHROMOSOMES
  df8int$sectors <- factor(df8int$sectors, levels = as.character(as.roman(1:16)))
  df1int$sectors <- factor(df1int$sectors, levels = as.character(as.roman(1:16)))
  
  ##############################################################################
  dfall <- rbind(df1int, df8int) %>% na.omit()
  circos.par("track.height" = 0.3, start.degree = 90, cell.padding = c(0,0))
  circos.initialize(sectors = dfall$sectors, x = dfall$x)

  if(is.null(maxy)){
    #I think this makes the sizes?
    circos.track(ylim = c(max(c(dfall$y)), 0), dfall$sectors, y = dfall$y, 
        panel.fun = function(x, y) {
            circos.text(CELL_META$xcenter, 
                0 - mm_y(5), 
                CELL_META$sector.index,
                niceFacing = FALSE)
            circos.axis(labels.cex = 0.1)
    })
  }else{
    circos.track(ylim = c(maxy, 0), dfall$sectors, y = dfall$y, 
      panel.fun = function(x, y) {
          circos.text(CELL_META$xcenter, 
              0 - mm_y(5), 
              CELL_META$sector.index,
              niceFacing = FALSE)
          circos.axis(labels.cex = 0.1)
      })
  }
  #Makes the chromosome overlap parts
  #CHROMOSOME I
  draw.sector(84, #RIGHT, originally 83.5
              90, #LEFT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color1fade_50, border = FALSE)
  if(include8 == TRUE){
    #CHROMOSOME VIII
    draw.sector(289.5, #LEFT
              305.4, #RIGHT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color8fade_50, border = FALSE)
  }
  
  
  #Makes the lines
  circos.trackPoints(df8int$sectors, df8int$x, abs(df8int$y), col = color8, pch = 16, cex = 0.1)
  circos.trackPoints(df1int$sectors, df1int$x, abs(df1int$y), col = color1, pch = 16, cex = 0.1)
  
  if(is.null(peaklist8) == FALSE){
    if(length(peaklist8$POS) >= 1){
      for(i in 1:length(peaklist8$POS)){
      circos.link(peaklist8$CHROM[i], 
                  peaklist8$POS[i], 
                  #Add 8 after
                  "VIII", c(0, max(dfall$x[dfall$sectors == "VIII"])),  
                  
                  col = color8fade, 
                  h.ratio = 0.3, 
                  border = color8fade, 
                  lwd = 1)
      }
    }else{
      print("No interactions on Chr 8")
    }
    
  }

  if(is.null(peaklist1) == FALSE){
    if(length(peaklist1$POS) >= 1){
      for(i in 1:length(peaklist1$POS)){
        circos.link("I",c(0, max(dfall$x[dfall$sectors == "I"])),  
                    #add 1 first
                    peaklist1$CHROM[i], 
                    peaklist1$POS[i], 
                    col = color1fade, 
                    h.ratio = 0.3, 
                    border = color1fade, 
                    lwd = 1)
      }
    }else{
      print("No interactions on Chr 1")
    }
  }  
  
  circos.clear()
}
```


```{r fig.height=6, fig.width=6, message=FALSE, warning=FALSE}
CuSO4_contrastglm_cutoffs %>% filter(Factor == "Interaction", Label == "CSS1 L", d == "Z") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = factor(CHROM), POS = POS, summary = abs(Summary), label = Factor) -> CuSO4_CSS1_circ

CuSO4_contrastglm_cutoffs %>% filter(Factor == "Interaction", Label == "CSS8 L", d == "Z") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = factor(CHROM), POS = POS, summary = abs(Summary), label = Factor) -> CuSO4_CSS8_circ

maybepeaks %>% filter(Factor == "Interaction", Label == "CSS1 L", d == "Z") %>% transmute(CHROM = factor(CHROM), POS = ppeakPOS) -> CSS1_peaks
maybepeaks %>% filter(Factor == "Interaction", Label == "CSS8 L", d == "Z") %>% transmute(CHROM = factor(CHROM), POS = ppeakPOS) -> CSS8_peaks

################################################################################ CuSO4 CSS VIII
# HVYTYDRX2_CuSO4_CSS8_contrastGLM %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
#  mutate(peakbroad = abs(Effect) >= quant095) %>% na.omit() %>%
# transmute(CHROM = CHROM, POS = POS, Effect = Effect, label = Factor) -> HVYTYDRX2_CuSO4_CSS8_circ
# 
# HVYTYDRX2_CuSO4_CSS8_contrastGLM %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
#   filter(abs(Effect) >= quant095) %>% select(CHROM, POS) %>% mutate(label = "Interaction", Effect = 1) -> HVYTYDRX2_CuSO4_CSS8_circpeaks

cybr_circos(d1 = rbind(CuSO4_CSS1_circ, ChromosomeScale2), 
            d8 = rbind(CuSO4_CSS8_circ, ChromosomeScale2),
            peaklist1 = CSS1_peaks, 
            peaklist8 = CSS8_peaks)

```

