---
title: "AllSequencingRuns Figures"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
require(ggplot2)

require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)

require(cybrBSA)


#TESTING
#glm_formula <- "Allele ~ Bulk * Parent + Rep"

#ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)))

ggplot2::theme_set(theme_light() + theme(axis.text.x = element_blank(), legend.position = "bottom") + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "gray70"),
        panel.border = element_rect(colour = "gray70", fill=NA, linewidth=0.4))+
    theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
    theme(text = element_text(size = 10)))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))

```



## Load in Data

### Load Genes of Interest for CuSO4

```{r}
sgd_orfs <- read.csv("SGD_ORFs.csv")

genes <- c("CRS5", "SOD1", "CUP1-1", "CUP1-2", "MET17", "MUP1", "PCL1", "OYE3", "FRE1", "FRE7", "CTR1", "IRC7")
sgd_orfs %>% filter(Gene.symbol %in% genes) -> sgd_copper

genes_eh <- c("VMA4", "MTL1", "PTR2", "LEU9", "IRC15")
sgd_orfs %>% filter(Gene.symbol %in% genes_eh) -> sgd_copper_eh
```

### Cataloging samples

Which samples are which?

```{r, include = FALSE}
MQCRuns <- read.csv("C:\\Users\\cassa\\OneDrive\\Documents\\SiegalLab\\Sequencing_Tuboweb\\AllMultiQCRuns.csv")

# MQCRuns %>% select(-DoseUnits, -SurvivalPercentage, -Dose, -Barcodes, -PercUndetermined) %>% 
#   na.omit() %>% 
#   filter(grepl("undetermined", Library) == FALSE) %>%
#   pivot_wider(names_from = Bulk, values_from = Parent) %>% select(Pool, ShortName, MQCDate, Dilute, CuSO4, Fluconazole, Cycloheximide, H2O2, Zeocin) %>% distinct() %>% arrange(Pool)

MQCRuns %>% select(-DoseUnits, -SurvivalPercentage, -Dose, -Barcodes, -PercUndetermined) %>% 
  na.omit() %>% 
  filter(grepl("undetermined", Library) == FALSE) %>%
  pivot_wider(names_from = Parent, values_from = Bulk) %>% select(Pool, ShortName, MQCDate, Wine1, Oak1, Wine8, Oak8) %>% distinct() %>% arrange(Pool) -> mqcbyparent

table(mqcbyparent$Pool, mqcbyparent$Wine1)
table(mqcbyparent$Pool, mqcbyparent$Wine8)


#MQCRuns %>% filter(Pool == "HKTMWDRX2")
```


### GLM Outputs

```{r}
#files <- as.data.frame(x = dir("StandardGLM"))
#files %>% filter(grepl("glm_all.rds",`dir("StandardGLM")`)) %>% transmute(files = `dir("StandardGLM")`) -> all_files

#Copper
HNGLVDRXY_CuSO4_CSS1_glm_all <- readRDS("StandardGLM/HNGLVDRXY_CuSO4_CSS1_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))	 #Cu1	
HKTFTDRX2_CuSO4_CSS1_glm_all <- readRDS("StandardGLM/HKTFTDRX2_CuSO4_CSS1_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu1			
xHNGLVDRXY_CuSO4_CSS1_glm_all <- readRDS("StandardGLM/xHNGLVDRXY_CuSO4_CSS1_glm_all.rds")%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))	#Cu1
HVYTYDRX2_CuSO4_CSS8_glm_all <- readRDS("StandardGLM/HVYTYDRX2_CuSO4_CSS8_glm_all.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu8

HTG3TDMXY_CuSO4_CSS8_glm_noreps_all <- readRDS(file = "StandardGLM/HTG3TDMXY_CuSO4_CSS8_glm_noreps.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu8

HTG3TDMXY_CuSO4_CSS8_glm <- readRDS(file = "StandardGLM/HTG3TDMXY_CuSO4_CSS8_glm.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu8

## Bulk Only
HTG3TDMXY_CSS8_glm_BULK <- readRDS(file = "StandardGLM/HTG3TDMXY_CSS8_glm_BULK.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu8
HVYTYDRX2_cu_CSS8_glm_BULK <- readRDS(file = "StandardGLM/HVYTYDRX2_cu_CSS8_glm_BULK.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu8
HKTFTDRX2_CSS1_glm_BULK <- readRDS(file = "StandardGLM/HKTFTDRX2_CSS1_glm_BULK.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu1
HNGLVDRXY_CSS1_glm_BULK <- readRDS(file = "StandardGLM/HNGLVDRXY_CSS1_glm_BULK.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu1


#Fluc
HGVMVDRX2_CSS1_glm_alld <- readRDS("StandardGLM/HGVMVDRX2_CSS1_glm_alld.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))	        #fluc1 duplicating reps
HVYTYDRX2_Fluc_CSS8_glm_all <- readRDS("StandardGLM/HVYTYDRX2_Fluc_CSS8_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #fluc8		
HGVMVDRX2norep_CSS1_glm_all <- readRDS("StandardGLM/HGVMVDRX2norep_CSS1_glm_all.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #fluc1 without replicates


#H2O2
HJ5HKDRX3_CSS1_glm_all <- readRDS("StandardGLM/HJ5HKDRX3_CSS1_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))			#H2O21
HJ5HKDRX3_CSS8_glm_all <- readRDS("StandardGLM/HJ5HKDRX3_CSS8_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))			#H2O28
HJ5HKDRX3_CSS8_glm_all_noreps <- readRDS("StandardGLM/HJ5HKDRX3_CSS8_glm_all_noreps.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))

#Zeocin + Cycloheximide 
HKTMWDRX2_Cyc_CSS1_glm_all_noreps <- readRDS("StandardGLM/HKTMWDRX2_Cyc_CSS1_glm_all_noreps.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))
HKTMWDRX2_Zeo_CSS1_glm_all_noreps <- readRDS("StandardGLM/HKTMWDRX2_Zeo_CSS1_glm_all_noreps.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))


```

### GLM Cutoffs

```{r}
#CuSO4 CSS1
HNGLVDRXY_CSS1_chrparentQs <- readRDS("StandardGLM/HNGLVDRXY_CSS1_chrparentQs.rds") %>% mutate(ds = "HNGLVDRXY", CSS = "I")
HNGLVDRXY_CSS1_parQs <- readRDS("StandardGLM/HNGLVDRXY_CSS1_parQs.rds") %>% mutate(ds = "HNGLVDRXY", CSS = "I")

HKTFTDRX2_CSS1_chrparentQs <- readRDS("StandardGLM/HKTFTDRX2_CSS1_chrparentQs.rds") %>% mutate(ds = "HKTFTDRX2", CSS = "I")
HKTFTDRX2_CSS1_parQs <- readRDS("StandardGLM/HKTFTDRX2_CSS1_parQs.rds") %>% mutate(ds = "HKTFTDRX2", CSS = "I")

#CuSO4 CSS8 - pending run
HVYTYDRX2_cu_CSS8_chrparentQs <- readRDS("StandardGLM/HVYTYDRX2_cu_CSS8_chrparentQs.rds") %>% mutate(ds = "HVYTYDRX2", CSS = "VIII")
HVYTYDRX2_cu_CSS8_parQs <- readRDS("StandardGLM/HVYTYDRX2_cu_CSS8_parQs.rds") %>% mutate(ds = "HVYTYDRX2", CSS = "VIII")

#New 2024
HTG3TDMXY_CSS8_glm_chrparentQs <- readRDS(file = "StandardGLM/HTG3TDMXY_CSS8_glm_chrparentQs.rds") %>% mutate(ds = "HTG3TDMXY", CSS = "VIII")
HTG3TDMXY_CSS8_glm_parQs <- readRDS(file = "StandardGLM/HTG3TDMXY_CSS8_glm_parQs.rds")  %>% mutate(ds = "HTG3TDMXY", CSS = "VIII")
HTG3TDMXY_CSS8_chrparentR6_Qs <- readRDS(file = "StandardGLM/HTG3TDMXY_CSS8_chrparentR6_Qs.rds")  %>% mutate(ds = "HTG3TDMXY", CSS = "VIII")
HTG3TDMXY_CSS8_parentonlyR6_parQs <- readRDS(file = "StandardGLM/HTG3TDMXY_CSS8_parentonlyR6_parQs.rds")  %>% mutate(ds = "HTG3TDMXY", CSS = "VIII")

#Fluconazole
HVYTYDRX2_F_CSS8_chrparentQs <- readRDS("StandardGLM/HVYTYDRX2_F_CSS8_chrparentQs.rds")
HVYTYDRX2_F_CSS8_parQs <- readRDS("StandardGLM/HVYTYDRX2_F_CSS8_parQs.rds")

HGVMVDRX2_F_CSS1d_chrparentQs <- readRDS("StandardGLM/HGVMVDRX2_F_CSS1d_chrparentQs.rds") #duplicate reps
HGVMVDRX2_F_CSS1d_parQs <- readRDS("StandardGLM/HGVMVDRX2_F_CSS1d_parQs.rds")

HGVMVDRX2_F_CSS1_chrparentQs<- readRDS("StandardGLM/HGVMVDRX2_F_CSS1_chrparentQs.rds") #no replicates
HGVMVDRX2_F_CSS1_parQs<- readRDS("StandardGLM/HGVMVDRX2_F_CSS1_parQs.rds")

#H2O2
HJ5HKDRX3_H_CSS8_chrparentQs <- readRDS("StandardGLM/HJ5HKDRX3_H_CSS8_chrparentQs.rds")
HJ5HKDRX3_H_CSS8_parQs <- readRDS("StandardGLM/HJ5HKDRX3_H_CSS8_parQs.rds")

HJ5HKDRX3norep_CSS8_chrparentQs <- readRDS("StandardGLM/HJ5HKDRX3norep_CSS8_chrparentQs.rds")
HJ5HKDRX3norep_CSS8_parQs <- readRDS("StandardGLM/HJ5HKDRX3norep_CSS8_parQs.rds")

#Cycloheximide
HKTMWDRX2_Cyc_norep_CSS8_chrparentQs<- readRDS("StandardGLM/HKTMWDRX2_Cyc_norep_CSS8_chrparentQs.rds")
HKTMWDRX2_Cyc_norep_CSS8_parQs<- readRDS("StandardGLM/HKTMWDRX2_Cyc_norep_CSS8_parQs.rds")

#Zeocin
HKTMWDRX2_Zeo_norep_CSS8_chrparentQs<- readRDS("StandardGLM/HKTMWDRX2_Zeo_norep_CSS8_chrparentQs.rds")
HKTMWDRX2_Zeo_norep_CSS8_parQs<- readRDS("StandardGLM/HKTMWDRX2_Zeo_norep_CSS8_parQs.rds")

```


# Plotting by Experiment

This includes all experiments and FDR of 0.05, 0.01, and 0.005. Each cutoff has only been run from one permutation, so multiple would ideally be done on the HPC.

## Plot CuSO4 Data

Individual Dataset HNGLVDRXY

```{r, fig.height = 4, fig.width = 9}

# xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_chrparentQs) %>% filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#     geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI Chromosome-Specific Cutoffs")

xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI Factor-Specific Cutoffs")
```

Individual Set HKTFTDRX2_CSS1_parQs

```{r, fig.height = 4, fig.width = 9}
# HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_chrparentQs) %>% filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#     geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI Chromosome-Specific Cutoffs")

HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI Factor-Specific Cutoffs")
```


```{r, fig.height = 4, fig.width = 9}
# HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_chrparentQs) %>% filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#     geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8 Chromosome-Specific Cutoffs")

HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8 Factor-Specific Cutoffs")
```



```{r, fig.height = 5, fig.width = 9}
rbind(data.frame(HKTFTDRX2_CuSO4_CSS1_glm_all, ds = "HKTFTDRX2", CSS = "I"),
      data.frame(HVYTYDRX2_CuSO4_CSS8_glm_all, ds = "HVYTYDRX2", CSS = "VIII"),
      data.frame(HTG3TDMXY_CuSO4_CSS8_glm, ds = "HTG3TDMXY", CSS = "VIII"), #2024 ADDITION
      #data.frame(HNGLVDRXY_CSS8_glm_all, ds = "HNGLVDRXY", CSS = "I"),
      data.frame(xHNGLVDRXY_CuSO4_CSS1_glm_all, ds = "HNGLVDRXYx", CSS = "I")) -> CuSO4_Data

CuSO4_Data %>% filter(Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(ds~CHROM, scales = "free", space = "free") + ggtitle("CuSO4 Bulk Effects")

CuSO4_Data %>% filter(Factor == "Parent") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(ds~CHROM, scales = "free", space = "free") + ggtitle("CuSO4 Parent Effects")

CuSO4_Data %>% filter(Factor == "Interaction") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(ds~CHROM, scales = "free", space = "free") + ggtitle("CuSO4 Interaction Effects")

```

### New 2024 Data (does not include permutations yet)

```{r, fig.height = 4, fig.width = 9, eval = FALSE}
# HTG3TDMXY_CuSO4_CSS8_glm_noreps %>% filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   ylim(0, 20) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8 2024")
```
```{r, fig.height = 4, fig.width = 9}
HTG3TDMXY_CuSO4_CSS8_glm %>% merge(HTG3TDMXY_CSS8_parentonlyR6_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8 With Replicates")

HTG3TDMXY_CuSO4_CSS8_glm_noreps_all %>% merge(HTG3TDMXY_CSS8_glm_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8 No Replicates")
```

### Bulk Only Effects

```{r, fig.height = 4, fig.width = 9}
HTG3TDMXY_CSS8_glm_BULK %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CSS8 b")

#This one's blank?
HVYTYDRX2_cu_CSS8_glm_BULK %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CSS8 a")

HKTFTDRX2_CSS1_glm_BULK %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CSS1 b")

HNGLVDRXY_CSS1_glm_BULK %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CSS1 a")

rbind(data.frame(HTG3TDMXY_CSS8_glm_BULK, ds = "CSS8b", CSS = "VIII"),
      data.frame(HVYTYDRX2_cu_CSS8_glm_BULK, ds = "CSS8a", CSS = "VIII"),
      data.frame(HKTFTDRX2_CSS1_glm_BULK, ds = "CSS1b", CSS = "I"),
      data.frame(HNGLVDRXY_CSS1_glm_BULK, ds = "CSS1a", CSS = "I")) %>%
  filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + 
  facet_grid(ds~CHROM, space = "free", scales = "free") +
  scale_color_manual(values = c("navy", "darkorange"))
  
```


## Plot Fluconazole Data

### CSS1 (no replicates)

Because the replicates were not balanced in this experiment, and did not correspond with specific days, I ran a model without a replicate effect which should just average the replicates together at each locus. The permutations for this were then done the same way.

The permutations are currently broken so all of this will just not have cutoffs

```{r, fig.height = 4, fig.width = 9}

# HGVMVDRX2norep_CSS1_glm_all %>% #CHANGE DATASET
#   merge(HGVMVDRX2_F_CSS1_chrparentQs) %>% #CHANGE CUTOFF FILE
#   
#   filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + 
#   
#   ggtitle("Fluc CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HGVMVDRX2norep_CSS1_glm_all %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```



### CSS1 (duplicated reps)

Because the replicates in this experiment are unbalanced, I duplicated the bulks which did not have a replicate to match the other two. This *should* result in a replicate effect of 0 for those samples, since they are fixed effects, and then allow the remaining replicates to capture the effects of difference by day. This is not verified yet, and I will be comparing this to the data without having any replicates (using the original CSSI script).

```{r, fig.height = 4, fig.width = 9}

################################################################################
# HGVMVDRX2_CSS1_glm_alld %>% #CHANGE DATASET
#   merge(HGVMVDRX2_F_CSS1d_chrparentQs) %>% #CHANGE CUTOFF FILE
#   
#   filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + 
#   
#   ggtitle("Fluc CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HGVMVDRX2_CSS1_glm_alld %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1d_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```

### How similar are the replicates?

We can use the two replicate effects here, which result from the random assignment of a, b, or c, to see how consistent our cutoffs will be. Each cutoff should be run >100 times on the HPC to get a true mean for the FDR, but to get each of these loaded quickly, that step has not been completed as of 12/19/23.

```{r, fig.height = 4, fig.width = 9}
# HGVMVDRX2_CSS1_glm_alld %>% #CHANGE DATASET
#   merge(HGVMVDRX2_F_CSS1d_chrparentQs) %>% #CHANGE CUTOFF FILE
#   
#   filter(grepl("Rep", Factor)) %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + 
#   
#   ggtitle("Fluc CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

HGVMVDRX2_CSS1_glm_alld %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1d_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(grepl("Rep", Factor)) %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```


### How similar are the duplicated vs non-rep'd results?

```{r, fig.height = 4, fig.width = 9, eval = FALSE}
rbind(data.frame(HGVMVDRX2_CSS1_glm_alld, dataset = "Dup"),
      data.frame(HGVMVDRX2norep_CSS1_glm_all, dataset = "NoRep")) %>%
  filter(grepl("Rep", Factor) == FALSE) %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = dataset)) + geom_line() + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  ggtitle("Fluc CSS I Comparison") #CHANGE TITLE
```

### CSS8
```{r, fig.height = 4, fig.width = 9}
################################################################################
# HVYTYDRX2_Fluc_CSS8_glm_all %>% #CHANGE DATASET
#   merge(HVYTYDRX2_F_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
#   
#   filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + 
#   
#   ggtitle("Fluc CSS VIII Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HVYTYDRX2_Fluc_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HVYTYDRX2_F_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS VIII Factor-Specific Cutoffs") #CHANGE TITLE
```

```{r, fig.height = 4, fig.width = 9, eval = FALSE}

rbind(data.frame(HGVMVDRX2_CSS1_glm_alld, Pool = "HGVMVDRX2", CSS = "I"), 
      data.frame(HVYTYDRX2_Fluc_CSS8_glm_all, Pool = "HVYTYDRX2", CSS = "VIII")) %>%
  filter(Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(Pool~CHROM, scales = "free", space = "free") + ggtitle("Fluconazole Bulk Effects")
  
 
rbind(data.frame(HGVMVDRX2_CSS1_glm_alld, Pool = "HGVMVDRX2", CSS = "I"), 
      data.frame(HVYTYDRX2_Fluc_CSS8_glm_all, Pool = "HVYTYDRX2", CSS = "VIII")) %>%
  filter(Factor == "Interaction") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(Pool~CHROM, scales = "free", space = "free") + ggtitle("Fluconazole Interaction Effects")
  

```

## Plot H2O2 Data

### Standard Analysis with Replicates
CSS 8

```{r, fig.height = 4, fig.width = 9}

################################################################################
# HJ5HKDRX3_CSS8_glm_all %>% #CHANGE DATASET
#   merge(HJ5HKDRX3_H_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
#   
#   filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + 
#   
#   ggtitle("H2O2 CSS VIII Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HJ5HKDRX3_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HJ5HKDRX3_H_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII Factor-Specific Cutoffs") #CHANGE TITLE
```

```{r, fig.height = 4, fig.width = 9}
rbind(data.frame(HJ5HKDRX3_CSS1_glm_all, Pool = "HJ5HKDRX3", CSS = "I"), 
      data.frame(HJ5HKDRX3_CSS8_glm_all, Pool = "HJ5HKDRX3", CSS = "VIII")) %>%
  filter(Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(CSS~CHROM, scales = "free", space = "free") + ggtitle("Fluconazole Bulk Effects")
  
```


### Averaged values without replicates

```{r, fig.height = 4, fig.width = 9}

################################################################################
# HJ5HKDRX3_CSS8_glm_all_noreps %>% #CHANGE DATASET
#   merge(HJ5HKDRX3norep_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
#   
#   filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + 
#   
#   ggtitle("H2O2 CSS VIII Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HJ5HKDRX3_CSS8_glm_all_noreps %>%  #CHANGE DATASET
  merge(HJ5HKDRX3norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII Factor-Specific Cutoffs") #CHANGE TITLE
```

## Plot Zeocin Data

### Averaged values without replicates

```{r, fig.height = 5, fig.width = 9}

################################################################################
# HKTMWDRX2_Zeo_CSS1_glm_all_noreps %>% #CHANGE DATASET
#   merge(HKTMWDRX2_Zeo_norep_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
#   
#   filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + 
#   
#   ggtitle("Zeocin CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HKTMWDRX2_Zeo_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Zeo_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Zeocin CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```

## Plot Cycloheximide Data

### Averaged values without replicates

```{r, fig.height = 5, fig.width = 9}

################################################################################
# HKTMWDRX2_Cyc_CSS1_glm_all_noreps %>% #CHANGE DATASET
#   merge(HKTMWDRX2_Cyc_norep_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
#   
#   filter(Factor != "Intercept") %>%
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
#   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
#   geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
#   facet_grid(~CHROM, space = "free", scales = "free") + 
#   
#   ggtitle("Cycloheximide CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HKTMWDRX2_Cyc_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Cyc_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Cycloheximide CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```


# Plotting only whole-genome FDR 0.05 cutoffs

```{r, fig.height = 5, fig.width = 12}
################################################################################ CuSO4 CSS I
xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    #geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  #geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())

################################################################################ CuSO4 CSS I 
HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI ")+
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())

################################################################################

HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8") + 
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) 

################################################################################ Fluconazole CSS I (w/o Rep)
HGVMVDRX2norep_CSS1_glm_all %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) 

HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8 Factor-Specific Cutoffs")

################################################################################ Fluconazole CSS I
HGVMVDRX2_CSS1_glm_alld %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1d_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) 

################################################################################ Fluconazole CSS VIII
HVYTYDRX2_Fluc_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HVYTYDRX2_F_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS VIII ")+
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())  

################################################################################ H2O2 CSS VIII
HJ5HKDRX3_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HJ5HKDRX3_H_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII ")+
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())  

################################################################################ H2O2 CSS I
HJ5HKDRX3_CSS8_glm_all_noreps %>%  #CHANGE DATASET
  merge(HJ5HKDRX3norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) 

################################################################################ Zeocin CSS I
HKTMWDRX2_Zeo_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Zeo_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Zeocin CSS I ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) 

################################################################################ Cyclohex CSS I
HKTMWDRX2_Cyc_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Cyc_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Cycloheximide CSS I ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) 

################################################################################ CuSO4 CSS VIII 2024 with replicates
HTG3TDMXY_CuSO4_CSS8_glm %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_parentonlyR6_parQs)%>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("CuSO4 CSS VIII ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) 

################################################################################ CuSO4 CSS VIII 2024 without replicates
HTG3TDMXY_CuSO4_CSS8_glm_noreps_all %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_glm_parQs)%>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("CuSO4 CSS VIII w/o Replicates") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) 
```
# Where are the SGD Genes?

```{r, fig.width=12, fig.height=6}
sgd_copper %>% mutate(POS = Gene.chromosomeLocation.start,
                      CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_copper

sgd_copper_eh %>% mutate(POS = Gene.chromosomeLocation.start,
                      CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_copper_eh

HTG3TDMXY_CuSO4_CSS8_glm_noreps_all %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_glm_parQs) -> HTG_noreps #CHANGE CUTOFF FILE
  
xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>%
  rbind(HTG_noreps) %>%
  filter(Factor == "Bulk") %>%
    mutate(peak = abs(Summary) > quant095) %>%

  ggplot(aes(x = POS, y = abs(Summary))) +   
  geom_vline(aes(xintercept = POS*peak), alpha = 0.1, color = "gray90") +
  geom_line() + 
  geom_hline(aes(yintercept = quant095), linetype = 1) + 
  
  geom_vline(data = sgd_copper, aes(xintercept = POS, color = paste(CHROM, Gene.symbol)), size = 2, alpha = 0.4) + 
  
  geom_vline(data = sgd_copper_eh, aes(xintercept = POS, linetype = paste(CHROM, Gene.symbol)), color = "firebrick", size = 1.5) + 
  ylim(0,20) +
  facet_grid(CSS~CHROM, space = "free", scales = "free") + 
  
  ggtitle("CuSO4 Additive") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank(), legend.position = "bottom") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "gray50"),
        panel.border = element_rect(colour = "gray50", fill=NA, linewidth=1))

xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>%
  rbind(HTG_noreps) %>%
  filter(Factor == "Interaction") %>%
  mutate(peak = abs(Summary) > quant095) %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_vline(aes(xintercept = POS*peak), alpha = 0.1, color = "gray85") +
  geom_line() + 
  geom_hline(aes(yintercept = quant095), linetype = 1) + 
  geom_vline(data = sgd_copper, aes(xintercept = POS, color = paste(CHROM, Gene.symbol)), size = 2, alpha = 0.4) + 
  
  geom_vline(data = sgd_copper_eh, aes(xintercept = POS, linetype = paste(CHROM, Gene.symbol)), color = "firebrick", size = 1.5) + 

  facet_grid(CSS~CHROM, space = "free", scales = "free") + 
  ylim(0,20) +
  
  ggtitle("CuSO4 Interaction Effects") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank(), legend.position = "bottom") + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "gray50"),
        panel.border = element_rect(colour = "gray50", fill=NA, linewidth=1))


```

# Calling Peaks using 5% FDR Cutoffs

```{r, fig.height = 5, fig.width = 12}

################################################################################ CuSO4 CSS I
xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())+
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ CuSO4 CSS I 
HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% 
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI ")+
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())+
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ CuSO4 CSS VIII

HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8") + 
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006))

################################################################################ Fluconazole CSS I (w/o Rep)
HGVMVDRX2norep_CSS1_glm_all %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006)) 


################################################################################ Fluconazole CSS I
HGVMVDRX2_CSS1_glm_alld %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1d_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ Fluconazole CSS VIII
HVYTYDRX2_Fluc_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HVYTYDRX2_F_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS VIII ")+
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())  +
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ H2O2 CSS VIII
HJ5HKDRX3_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HJ5HKDRX3_H_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII ")+
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())  +
  scale_alpha_discrete(range = c(0, 0.006))

################################################################################ H2O2 CSS I
HJ5HKDRX3_CSS8_glm_all_noreps %>%  #CHANGE DATASET
  merge(HJ5HKDRX3norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ Zeocin CSS I
HKTMWDRX2_Zeo_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Zeo_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%

  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
  
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Zeocin CSS I ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006))
################################################################################ Cyclohex CSS I
HKTMWDRX2_Cyc_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Cyc_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
  
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Cycloheximide CSS I ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006)) 


################################################################################ CuSO4 CSS VIII 2024 with replicates
HTG3TDMXY_CuSO4_CSS8_glm %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_parentonlyR6_parQs)%>% #CHANGE CUTOFF FILE
  
filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
  
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("CuSO4 CSS 8 Replicates") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ CuSO4 CSS VIII 2024 without replicates
HTG3TDMXY_CuSO4_CSS8_glm_noreps_all %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_glm_parQs)%>% #CHANGE CUTOFF FILE
  
filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_point(size = 0.2, alpha = 0.5) + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
  
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  scale_color_manual(values = c("black",  "violet", "gray", "gray90", "gray90")) +
  ggtitle("CuSO4 CSS 8 No Replicates") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006))



```


### Plotting Multiple Experiments for Bulk Effects

CuSO4 Merge

```{r, fig.height = 5, fig.width = 12}

xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit()  -> da

HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% 
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit()  -> db

HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit()  -> d8

HTG3TDMXY_CuSO4_CSS8_glm_noreps_all %>% merge(HTG3TDMXY_CSS8_parentonlyR6_parQs) %>% 
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit()  -> d824r

################################################################################

rbind(da, db, d8, d824r) %>%
  filter(Factor == "Bulk") %>%

  ggplot(aes(x = POS, y = abs(Summary), color = ds)) + geom_point(size = 0.4, alpha = 0.5) + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = ds), size = 2) +
    geom_hline(aes(yintercept = quant095, color = ds), linetype = "dashed") + 

  facet_grid(CSS~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 Bulk Effects") + 
  ylab("abs(Zscore)") +
  scale_color_manual(values = c("gray30", "#24588A", "firebrick", "#ED7B01")) +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.004))

rbind(da, db, d8, d824r) %>%
  filter(Factor == "Interaction") %>%

  ggplot(aes(x = POS, y = abs(Summary), color = ds)) + geom_point(size = 0.4, alpha = 0.5) + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = ds), size = 2) +
    geom_hline(aes(yintercept = quant095, color = ds), linetype = "dashed") + 

  facet_grid(CSS~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 Interaction Effects") + 
  ylab("abs(Zscore)") +
  scale_color_manual(values = c("gray30", "#24588A", "firebrick", "#ED7B01")) +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.004))

################################################################################
#Do bulk stuff separated by each run

rbind(da, db, d8, d824r) %>%
  filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_point(size = 0.4, alpha = 0.2) + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = CSS), size = 2) +
  geom_hline(aes(yintercept = quant095, color = CSS)) + 

  facet_grid(ds~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 Bulk Effects") + 
  ylab("abs(Zscore)") +
  scale_color_manual(values = c("#24588A", "#ED7B01")) +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.004))

rbind(da, db, d8, d824r) %>%
  filter(Factor == "Interaction") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_point(size = 0.4, alpha = 0.2) + 
  
  geom_vline(aes(xintercept = POS*peakbroad, color = CSS), size = 2, alpha = 0.01) +
  geom_hline(aes(yintercept = quant095, color = CSS)) + 

  facet_grid(ds~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 Interaction Effects") + 
  ylab("abs(Zscore)") +
  scale_color_manual(values = c("#24588A", "#ED7B01")) +
  theme(axis.text.x = element_blank()) 
```
Fluconazole Merge

```{r, fig.height = 5, fig.width = 12}
################################################################################ Fluconazole CSS I (w/o Rep)
HGVMVDRX2norep_CSS1_glm_all %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>% mutate(ds = "HGVMVDRX2norep_CSS1_glm_all",
                                                                        CSS = "I") -> fa

################################################################################ Fluconazole CSS I
HGVMVDRX2_CSS1_glm_alld %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1d_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>% mutate(ds = "HGVMVDRX2_CSS1_glm_alld",
                                                                        CSS = "I") -> fb
################################################################################ Fluconazole CSS VIII
HVYTYDRX2_Fluc_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HVYTYDRX2_F_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>% mutate(ds = "HVYTYDRX2_Fluc_CSS8_glm_all",
                                                                        CSS = "VIII") -> fc

 rbind(fa, fb, fc) %>%
  filter(Factor == "Bulk") %>%

  ggplot(aes(x = POS, y = abs(Summary), color = ds)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = ds), size = 2) +
  geom_hline(aes(yintercept = quant095, color = ds), linetype = "dashed") + 

  facet_grid(CSS~CHROM, space = "free", scales = "free") + ggtitle("Fluconazole Bulk Effects") + 
  ylab("abs(Zscore)") +
  scale_color_manual(values = c("gray30", "#24588A", "#ED7B01")) +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.004))

```

# CuSO4 Plotting for TAGC - Interaction Peaks

```{r, fig.height = 4, fig.width = 12}
################################################################################ CuSO4 CSS I
xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
    filter(grepl("Rep", Factor) == FALSE) %>%

  mutate(ofinterest = as.numeric(Factor == "Interaction")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  
  geom_vline(aes(xintercept = POS*peakbroad), color = "#CBC3E3", alpha = 0.5) +
    
  geom_point(alpha = 0.2, size = 0.2) + 
  
  #geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  scale_color_manual(values = c("black",  "purple", "gray", "gray90", "gray90")) +

  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())+
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ CuSO4 CSS I 
HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% 
  filter(Factor != "Intercept") %>%  filter(grepl("Rep", Factor) == FALSE) %>%

mutate(ofinterest = as.numeric(Factor == "Interaction")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  
  geom_vline(aes(xintercept = POS*peakbroad), color = "#CBC3E3", alpha = 0.5) +
    
  geom_point(alpha = 0.2, size = 0.2) + 
  
  #geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  scale_color_manual(values = c("black",  "purple", "gray", "gray90", "gray90")) +

  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI ")+
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())+
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ CuSO4 CSS VIII

HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor != "Intercept") %>%
    filter(grepl("Rep", Factor) == FALSE) %>%

mutate(ofinterest = as.numeric(Factor == "Interaction")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  
  geom_vline(aes(xintercept = POS*peakbroad), color = "#CBC3E3", alpha = 0.5) +
    
  geom_point(alpha = 0.2, size = 0.2) + 
  
  #geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  scale_color_manual(values = c("black",  "purple", "gray", "gray90", "gray90")) +

  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8") + 
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006))

################################################################################ CuSO4 CSS VIII 24

HTG3TDMXY_CuSO4_CSS8_glm %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_parentonlyR6_parQs)%>% #CHANGE CUTOFF FILE
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(1:16)))) %>%
  
  filter(Factor != "Intercept") %>%
  filter(grepl("Rep", Factor) == FALSE) %>%
  mutate(ofinterest = as.numeric(Factor == "Interaction")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  
  geom_vline(aes(xintercept = POS*peakbroad), color = "#CBC3E3", alpha = 0.5) +
  
  geom_point(alpha = 0.2, size = 0.2) + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
    scale_color_manual(values = c("black",  "purple", "gray", "gray90", "gray90")) +
  scale_linetype_discrete(c(0, 1)) +
  ggtitle("CuSO4 CSS 8 Replicates") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())
```

```{r, fig.height = 4, fig.width = 12}
################################################################################ CuSO4 CSS I
xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
    filter(grepl("Rep", Factor) == FALSE) %>%

  mutate(ofinterest = as.numeric(Factor == "Bulk")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  
  geom_vline(aes(xintercept = POS*peakbroad), color = "gray", alpha = 0.5) +
    
  geom_point(alpha = 0.2, size = 0.2) + 
  
  #geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  scale_color_manual(values = c("black",  "purple", "gray", "gray90", "gray90")) +

  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI ") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())+
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ CuSO4 CSS I 
HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% 
  filter(Factor != "Intercept") %>%  filter(grepl("Rep", Factor) == FALSE) %>%

mutate(ofinterest = as.numeric(Factor == "Bulk")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  
  geom_vline(aes(xintercept = POS*peakbroad), color = "gray", alpha = 0.5) +
    
  geom_point(alpha = 0.2, size = 0.2) + 
  
  #geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  scale_color_manual(values = c("black",  "purple", "gray", "gray90", "gray90")) +

  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI ")+
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())+
  scale_alpha_discrete(range = c(0, 0.006)) 

################################################################################ CuSO4 CSS VIII

HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor != "Intercept") %>%
    filter(grepl("Rep", Factor) == FALSE) %>%

mutate(ofinterest = as.numeric(Factor == "Bulk")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  
  geom_vline(aes(xintercept = POS*peakbroad), color = "gray", alpha = 0.5) +
    
  geom_point(alpha = 0.2, size = 0.2) + 
  
  #geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  scale_color_manual(values = c("black",  "purple", "gray", "gray90", "gray90")) +

  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSS8") + 
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006))

################################################################################ CuSO4 CSS VIII 24

HTG3TDMXY_CuSO4_CSS8_glm %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_parentonlyR6_parQs)%>% #CHANGE CUTOFF FILE
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(1:16)))) %>%
  
  filter(Factor != "Intercept") %>%
  filter(grepl("Rep", Factor) == FALSE) %>%
  mutate(ofinterest = as.numeric(Factor == "Bulk")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + 
  
  geom_vline(aes(xintercept = POS*peakbroad), color = "gray", alpha = 0.5) +
  
  geom_point(alpha = 0.2, size = 0.2) + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
    scale_color_manual(values = c("black",  "purple", "gray", "gray90", "gray90")) +
  scale_linetype_discrete(c(0, 1)) +
  ggtitle("CuSO4 CSS 8 Replicates") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())
```

### Compare peaks in common

```{r}
xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% mutate(ofinterest = as.numeric(Factor == "Bulk")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  select(CHROM, POS, peakbroad) %>%
  filter(peakbroad == TRUE) -> HNG_p_bulk

HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% mutate(ofinterest = as.numeric(Factor == "Bulk")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  select(CHROM, POS, peakbroad) %>%
  filter(peakbroad == TRUE) -> HKT_p_bulk

merge(HNG_p_bulk, HKT_p_bulk) -> CuSO4_1_BulkPeaks

xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% mutate(ofinterest = as.numeric(Factor == "Interaction")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  select(CHROM, POS, peakbroad) %>%
  filter(peakbroad == TRUE) -> HNG_p_int

HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% mutate(ofinterest = as.numeric(Factor == "Interaction")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  select(CHROM, POS, peakbroad) %>%
  filter(peakbroad == TRUE) -> HKT_p_int

merge(HNG_p_int, HKT_p_int) -> CuSO4_1_IntPeaks
```

```{r}
HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% mutate(ofinterest = as.numeric(Factor == "Bulk")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  select(CHROM, POS, peakbroad) %>%
  filter(peakbroad == TRUE) -> HVY_p_bulk

HTG3TDMXY_CuSO4_CSS8_glm %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_parentonlyR6_parQs)%>% mutate(ofinterest = as.numeric(Factor == "Bulk")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  select(CHROM, POS, peakbroad) %>%
  filter(peakbroad == TRUE) -> HTG_p_bulk

merge(HVY_p_bulk, HTG_p_bulk) -> CuSO4_8_BulkPeaks

HVYTYDRX2_CuSO4_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HVYTYDRX2_cu_CSS8_parQs) %>% mutate(ofinterest = as.numeric(Factor == "Interaction")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  select(CHROM, POS, peakbroad) %>%
  filter(peakbroad == TRUE) -> HVY_p_int

HTG3TDMXY_CuSO4_CSS8_glm %>% merge(HTG3TDMXY_CSS8_parentonlyR6_parQs) %>% mutate(ofinterest = as.numeric(Factor == "Interaction")) %>%
  mutate(peakbroad = ofinterest*abs(Summary) >= quant095) %>% na.omit() %>%
  select(CHROM, POS, peakbroad) %>%
  filter(peakbroad == TRUE) -> HTG_p_int

merge(HVY_p_int, HTG_p_int) -> CuSO4_8_IntPeaks

```

### And now adding to the plots

```{r, fig.width=12, fig.height=5}
rbind(HTG3TDMXY_CuSO4_CSS8_glm, HVYTYDRX2_CuSO4_CSS8_glm_all) %>%
  filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_vline(data = CuSO4_8_BulkPeaks, aes(xintercept = POS), color = "gray", size = 2, alpha = 0.5) +
  geom_vline(data = rbind(sgd_copper, sgd_copper_eh), aes(xintercept = POS, color = paste(CHROM, Gene.symbol, sep = "  -  ")), size = 2, linetype = "dashed") +
  geom_point(alpha = 0.6, size = 0.4) + 
    facet_grid(~CHROM, space = "free", scales = "free") + 
  scale_linetype_discrete(c(0, 1)) +
  ggtitle("CuSO4 CSS 8 Additive Peaks") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())

rbind(HTG3TDMXY_CuSO4_CSS8_glm, HVYTYDRX2_CuSO4_CSS8_glm_all) %>%
  filter(Factor == "Interaction") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_vline(data = CuSO4_8_IntPeaks, aes(xintercept = POS), color = "#CBC3E3", size = 2) +
    geom_vline(data = rbind(sgd_copper, sgd_copper_eh), aes(xintercept = POS, color = paste(CHROM, Gene.symbol, sep = "  -  ")), size = 2, linetype = "dashed") +
  geom_point(alpha = 0.6, size = 0.4, color = "purple") + 
  ylim(0, 20)+
    facet_grid(~CHROM, space = "free", scales = "free") + 
  scale_linetype_discrete(c(0, 1)) +
  ggtitle("CuSO4 CSS 8 Interaction Peaks") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())
```

Plotting interaction vs parent effects to see how well they match

```{r, fig.height = 4, fig.width = 12, eval = FALSE}
HTG3TDMXY_CuSO4_CSS8_glm %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_parentonlyR6_parQs)%>% #CHANGE CUTOFF FILE
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(1:16)))) %>%
  
filter(Factor != "Intercept", Factor != "Bulk") %>%
  filter(grepl("Rep", Factor) == FALSE) %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_point(alpha = 0.2, size = 0.2) + 
  
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
    scale_color_manual(values = c("purple", "gray", "gray90", "gray90")) +

  ggtitle("CuSO4 CSS 8 Replicates, Parent vs Interaction") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank(), legend.position = "none") +
  scale_alpha_discrete(range = c(0, 0.006)) 

HTG3TDMXY_CuSO4_CSS8_glm %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_parentonlyR6_parQs)%>% #CHANGE CUTOFF FILE
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(1:16)))) %>%
  select(Factor, CHROM, POS, Summary) %>%
  pivot_wider(names_from = Factor, values_from = Summary) %>%
 # na.omit() %>%

  ggplot(aes(x = POS, y = abs(Parent) - abs(Interaction))) + geom_point(alpha = 0.2, size = 0.2) +
  facet_grid(~CHROM, space = "free", scales = "free") +
  ggtitle("CuSO4 CSS 8 Replicates, Parent MINUS Interaction") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())

HTG3TDMXY_CuSO4_CSS8_glm %>%  #CHANGE DATASET
  merge(HTG3TDMXY_CSS8_parentonlyR6_parQs)%>% #CHANGE CUTOFF FILE
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(1:16)))) %>%
  select(Factor, CHROM, POS, Summary) %>% 
  pivot_wider(names_from = Factor, values_from = Summary) %>%
  ggplot(aes(x = POS, y = log2(abs(Parent)/abs(Interaction)))) + geom_point(alpha = 0.2, size = 0.2) + 
  geom_hline(yintercept = 0) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  ggtitle("CuSO4 CSS 8 Replicates, log2(Parent/Interaction)") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) 
```


# Plotting Interactions based on 5% FDR

### Function for circos plots

```{r}
# d1 = rbind(xHNGLVDRXY_CuSO4_CSS1_circ, ChromosomeScale2)
# d8 = rbind(HVYTYDRX2_CuSO4_CSS8_circ, ChromosomeScale2)
# peaklist8 = HVYTYDRX2_CuSO4_CSS8_circpeaks

cybr_circos <- function(d1, d8, peaklist1 = NULL, peaklist8 = NULL, maxy = NULL, color1 = "#24588A", color8 = "#ED7B01", opacity = "50"){
  
  color1fade_50 <- paste(color1, "50", sep = "")
  color8fade_50 <- paste(color8, "50", sep = "")
  
  color1fade <- paste(color1, opacity, sep = "")
  color8fade <- paste(color8, opacity, sep = "")

  if(is.null(d8)){
    d8 <- d1
    include8 <- FALSE
  }else{
    include8 <- TRUE
  }
    #SET UP DATA
  df8 <- data.frame(sectors = as.character(d8$CHROM),
                 x = d8$POS,
                 y = abs(d8$summary),
                  label = d8$label)

  df1 <- data.frame(sectors = as.character(d1$CHROM),
                 x = d1$POS,
                 y = abs(d1$summary),
                  label = d1$label)

  #Take just the interactions of each
  df8int <- df8 %>% filter(label == "Interaction")
  df1int <- df1 %>% filter(label == "Interaction")

    #REORDER THE CHROMOSOMES
  df8int$sectors <- factor(df8int$sectors, levels = as.character(as.roman(1:16)))
  df1int$sectors <- factor(df1int$sectors, levels = as.character(as.roman(1:16)))
  
  ##############################################################################
  dfall <- rbind(df1int, df8int) %>% na.omit()
  circos.par("track.height" = 0.3, start.degree = 90, cell.padding = c(0,0))
  circos.initialize(sectors = dfall$sectors, x = dfall$x)

  if(is.null(maxy)){
    #I think this makes the sizes?
    circos.track(ylim = c(max(c(dfall$y)), 0), dfall$sectors, y = dfall$y, 
        panel.fun = function(x, y) {
            circos.text(CELL_META$xcenter, 
                0 - mm_y(5), 
                CELL_META$sector.index,
                niceFacing = FALSE)
            circos.axis(labels.cex = 0.1)
    })
  }else{
    circos.track(ylim = c(maxy, 0), dfall$sectors, y = dfall$y, 
      panel.fun = function(x, y) {
          circos.text(CELL_META$xcenter, 
              0 - mm_y(5), 
              CELL_META$sector.index,
              niceFacing = FALSE)
          circos.axis(labels.cex = 0.1)
      })
  }
  #Makes the chromosome overlap parts
  #CHROMOSOME I
  draw.sector(83.5, #RIGHT
              90, #LEFT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color1fade_50, border = FALSE)
  if(include8 == TRUE){
    #CHROMOSOME VIII
    draw.sector(289.5, #LEFT
              305.4, #RIGHT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color8fade_50, border = FALSE)
  }
  
  
  #Makes the lines
  circos.trackPoints(df8int$sectors, df8int$x, abs(df8int$y), col = color8, pch = 16, cex = 0.1)
  circos.trackPoints(df1int$sectors, df1int$x, abs(df1int$y), col = color1, pch = 16, cex = 0.1)
  
  if(is.null(peaklist8) == FALSE){
    if(length(peaklist8$POS) >= 1){
      for(i in 1:length(peaklist8$POS)){
      circos.link(peaklist8$CHROM[i], 
                  peaklist8$POS[i], 
                  #Add 8 after
                  "VIII", c(0, max(dfall$x[dfall$sectors == "VIII"])),  
                  
                  col = color8fade, 
                  h.ratio = 0.3, 
                  border = color8fade, 
                  lwd = 1)
      }
    }else{
      print("No interactions on Chr 8")
    }
    
  }

  if(is.null(peaklist1) == FALSE){
    if(length(peaklist1$POS) >= 1){
      for(i in 1:length(peaklist1$POS)){
        circos.link("I",c(0, max(dfall$x[dfall$sectors == "I"])),  
                    #add 1 first
                    peaklist1$CHROM[i], 
                    peaklist1$POS[i], 
                    col = color1fade, 
                    h.ratio = 0.3, 
                    border = color1fade, 
                    lwd = 1)
      }
    }else{
      print("No interactions on Chr 1")
    }
  }  
  
  circos.clear()
}
```


## CuSO4 Interactions

### First CSSI Experiment (without replicates)

```{r fig.height=8, fig.width=, message=FALSE, warning=FALSE}
################################################################################ CuSO4 CSS I
xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% filter(Factor == "Interaction") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = CHROM, POS = POS, summary = Summary, label = Factor) -> xHNGLVDRXY_CuSO4_CSS1_circ

xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% filter(Factor == "Interaction") %>%
  filter(abs(Summary) >= quant095) %>% select(CHROM, POS) %>% mutate(label = "Interaction", summary = 1) -> xHNGLVDRXY_CuSO4_CSS1_circpeaks

################################################################################ CuSO4 CSS VIII
HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = CHROM, POS = POS, summary = Summary, label = Factor) -> HVYTYDRX2_CuSO4_CSS8_circ

HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
  filter(abs(Summary) >= quant095) %>% select(CHROM, POS) %>% mutate(label = "Interaction", summary = 1) -> HVYTYDRX2_CuSO4_CSS8_circpeaks


cybr_circos(d1 = rbind(xHNGLVDRXY_CuSO4_CSS1_circ, ChromosomeScale2), 
            d8 = rbind(HVYTYDRX2_CuSO4_CSS8_circ, ChromosomeScale2),
            peaklist1 = xHNGLVDRXY_CuSO4_CSS1_circpeaks, 
            peaklist8 = HVYTYDRX2_CuSO4_CSS8_circpeaks)

```
### Second CSSI Experiment (with replicates)

```{r, fig.height= 8, fig.width= 8}
################################################################################ CuSO4 CSS I
HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% filter(Factor == "Interaction") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = CHROM, POS = POS, summary = Summary, label = Factor) -> xHNGLVDRXY_CuSO4_CSS1_circ

HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% filter(Factor == "Interaction") %>%
  filter(abs(Summary) >= quant095) %>% select(CHROM, POS) %>% mutate(label = "Interaction", summary = 1) -> xHNGLVDRXY_CuSO4_CSS1_circpeaks

################################################################################ CuSO4 CSS VIII
HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = CHROM, POS = POS, summary = Summary, label = Factor) -> HVYTYDRX2_CuSO4_CSS8_circ

HVYTYDRX2_CuSO4_CSS8_glm_all %>% merge(HVYTYDRX2_cu_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
  filter(abs(Summary) >= quant095) %>% select(CHROM, POS) %>% mutate(label = "Interaction", summary = 1) -> HVYTYDRX2_CuSO4_CSS8_circpeaks


cybr_circos(d1 = rbind(xHNGLVDRXY_CuSO4_CSS1_circ, ChromosomeScale2), 
            d8 = rbind(HVYTYDRX2_CuSO4_CSS8_circ, ChromosomeScale2),
            peaklist1 = xHNGLVDRXY_CuSO4_CSS1_circpeaks, 
            peaklist8 = HVYTYDRX2_CuSO4_CSS8_circpeaks)

```

## Fluconazole Interactions w/o Replicates

Allele ~ Bulk*Parent used for CSS I

```{r, fig.height= 8, fig.width= 8}

################################################################################ Fluconazole CSS I (w/o Rep)
HGVMVDRX2norep_CSS1_glm_all %>% merge(HGVMVDRX2_F_CSS1_parQs) %>% filter(Factor == "Interaction") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = CHROM, POS = POS, summary = Summary, label = Factor) -> HGVMVDRX2norep_CSS1_circ

HGVMVDRX2norep_CSS1_glm_all %>% merge(HGVMVDRX2_F_CSS1_parQs) %>% filter(Factor == "Interaction") %>%
  filter(abs(Summary) >= quant095) %>% select(CHROM, POS) %>% mutate(label = "Interaction", summary = 1) -> HGVMVDRX2norep_CSS1_circpeaks

################################################################################ Fluconazole CSS VIII
HVYTYDRX2_Fluc_CSS8_glm_all %>% merge(HVYTYDRX2_F_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = CHROM, POS = POS, summary = Summary, label = Factor) -> HVYTYDRX2_Fluc_CSS8_circ

HVYTYDRX2_Fluc_CSS8_glm_all %>% merge(HVYTYDRX2_F_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
  filter(abs(Summary) >= quant095) %>% select(CHROM, POS) %>% mutate(label = "Interaction", summary = 1) -> HVYTYDRX2_Fluc_CSS8_circpeaks


cybr_circos(d1 = rbind(HGVMVDRX2norep_CSS1_circ, ChromosomeScale2), 
            d8 = rbind(HVYTYDRX2_Fluc_CSS8_circ, ChromosomeScale2),
            peaklist1 = HGVMVDRX2norep_CSS1_circpeaks, 
            peaklist8 = HVYTYDRX2_Fluc_CSS8_circpeaks)

```

## Fluconazole Interactions with doubled replicates

Allele ~ Bulk*Parent+Rep used for both

```{r, fig.height= 8, fig.width= 8}

################################################################################ Fluconazole CSS I (w/o Rep)
HGVMVDRX2_CSS1_glm_alld %>% merge(HGVMVDRX2_F_CSS1d_parQs) %>% filter(Factor == "Interaction") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = CHROM, POS = POS, summary = Summary, label = Factor) -> HGVMVDRX2norep_CSS1_circ

HGVMVDRX2_CSS1_glm_alld %>% merge(HGVMVDRX2_F_CSS1d_parQs) %>% filter(Factor == "Interaction") %>%
  filter(abs(Summary) >= quant095) %>% select(CHROM, POS) %>% mutate(label = "Interaction", summary = 1) -> HGVMVDRX2norep_CSS1_circpeaks

################################################################################ Fluconazole CSS VIII
HVYTYDRX2_Fluc_CSS8_glm_all %>% merge(HVYTYDRX2_F_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
transmute(CHROM = CHROM, POS = POS, summary = Summary, label = Factor) -> HVYTYDRX2_Fluc_CSS8_circ

HVYTYDRX2_Fluc_CSS8_glm_all %>% merge(HVYTYDRX2_F_CSS8_parQs) %>% filter(Factor == "Interaction") %>%
  filter(abs(Summary) >= quant095) %>% select(CHROM, POS) %>% mutate(label = "Interaction", summary = 1) -> HVYTYDRX2_Fluc_CSS8_circpeaks


cybr_circos(d1 = rbind(HGVMVDRX2norep_CSS1_circ, ChromosomeScale2), 
            d8 = rbind(HVYTYDRX2_Fluc_CSS8_circ, ChromosomeScale2),
            peaklist1 = HGVMVDRX2norep_CSS1_circpeaks, 
            peaklist8 = HVYTYDRX2_Fluc_CSS8_circpeaks)
```

## Hydrogen Peroxide, Zeocin, and Cycloheximide Interactions

Not doing circos plots because we can see where the interactions are from here and it isn't _really_ worth comparing for these experiments

```{r, fig.height = 4, fig.width = 12}

################################################################################ H2O2 CSS VIII
HJ5HKDRX3_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HJ5HKDRX3_H_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor == "Interaction") %>%
 mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII ")+
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank())  +
  scale_alpha_discrete(range = c(0, 0.006)) + scale_color_manual(values = "#ED7B01")

################################################################################ H2O2 CSS VIII w/o reps
HJ5HKDRX3_CSS8_glm_all_noreps %>%  #CHANGE DATASET
  merge(HJ5HKDRX3norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor == "Interaction") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
    geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 

  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII w/0 replicates") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006)) + scale_color_manual(values = "#ED7B01")

################################################################################ Zeocin CSS I
HKTMWDRX2_Zeo_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Zeo_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor == "Interaction") %>%

  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
  
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Zeocin CSS I w/o replicates") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006)) + scale_color_manual(values = "#24588A")

################################################################################ Cyclohex CSS I
HKTMWDRX2_Cyc_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Cyc_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor == "Interaction") %>%
  mutate(peakbroad = abs(Summary) >= quant095) %>% na.omit() %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  
  geom_vline(aes(xintercept = POS, alpha = peakbroad, color = Factor), size = 2) +
  
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  #   geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  # geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Cycloheximide CSS I w/o replicates") +
  xlab("abs(Zscore)") +
  theme(axis.text.x = element_blank()) +
  scale_alpha_discrete(range = c(0, 0.006)) + scale_color_manual(values = "#24588A")
```

