---
title: "2023 Comparing Coverage and SNPs"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: hide
  html_notebook:
    code_folding: hide
---

```{r, warning=FALSE, message=FALSE, comment=FALSE, fig.width=18, fig.height=5}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)

require(cybrBSA)


#TESTING
#glm_formula <- "Allele ~ Bulk * Parent + Rep"

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV"),
                     POS = c(375215, 116167,389572, 198671, 46271),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO"),
                     cat = c("Floc", "Fixed", "Floc", "Fixed", "Fixed"))

```

# Functions

```{r}
glm_cb <- function(..., W, formula) {
  data <- list(...)
  
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  
  glm_formula <- as.formula(formula)
  
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #CHANGE THIS TO ADJUST TYPE OF REGRESSION
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  #Output: Effect, Standard Error, Z-score 
  #(ignores p-value since it can be calculated from Z score; multiply by 1 if you would like to retain it)
  return(summary(glm_fit)$coefficients[1:((length(summary(glm_fit)$coefficients)*0.75))])
  
}

glm_cb2_short <- function(..., W, formula, numgroups = FALSE, outputlength = 4, return = c("Z")) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  #MAYBEWORKS
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  #Return specific ones like z-score only
  if(return %in% "Z"){
    output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients)*0.5)+1):((length(summary(glm_fit)$coefficients)*0.75))]
  }
  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
}

```

Note: if looking for smoothed vs unsmoothed data, the `Figures_2023_BiologyofGenomes.rmd` file has these at the beginning.

# Loading In Data

GQ Cutoff is 98 for all samples, so find out how many of these are under 98 and in which samples that tends to occur

```{r}
setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)
parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) %>% filter(Unique == 1) -> pSNPs
pSNPs$CHROM <- factor(pSNPs$CHROM, levels = as.character(as.roman(1:17)))

```

## Table Processing
```{r}
# MQCRuns <- read.csv("C:\\Users\\cassa\\OneDrive\\Documents\\SiegalLab\\Sequencing_Tuboweb\\AllMultiQCRuns.csv")
# 
# MQCRuns %>% select(Pool, ShortName, VCF_Table) %>% distinct() -> RawFiles
# 
# for(i in 1:length(RawFiles$VCF_Table)){
#   cybrInputGATKTable(RawFiles$VCF_Table[i]) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
#   select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) %>% mutate(Pool = RawFiles$Pool[i])-> rawdata
# 
#   if(i > 1){
#     alldata <- rbind(rawdata, alldata)
#   }else{
#     alldata <- rawdata
#   }
# 
# }
# 
# alldata %>% distinct() %>%
#   mutate(Dataset = gsub(".*_n01_", "", Dataset)) %>%
#   mutate(Dataset = gsub("\\.fastq$", "", Dataset)) -> alldata
# 
# saveRDS(alldata, file = "allseqruns_oct23.rds")

alldata <- readRDS("allseqruns_oct23.rds")

#unique(alldata$Pool)
```
For any new datasets:

```{r, eval = FALSE}
#LOAD IN ORIGINAL
alldata <- readRDS("allseqruns_oct23.rds")

#LOAD IN TABLE
MQCRuns <- read.csv("C:\\Users\\cassa\\OneDrive\\Documents\\SiegalLab\\Sequencing_Tuboweb\\AllMultiQCRuns.csv")

MQCRuns %>% select(Pool, ShortName, VCF_Table) %>% distinct() -> RawFiles

var <- 1 #CHANGE THIS FOR THE INDEX OF RawFiles YOU'RE ANALYZING
RawFiles[var] #CHECK THAT IT'S RIGHT

for(i in var){
  cybrInputGATKTable(RawFiles$VCF_Table[i]) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) %>% mutate(Pool = RawFiles$Pool[i])-> rawdata

}

rawdata %>% distinct() %>%
  mutate(Dataset = gsub(".*_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub("\\.fastq$", "", Dataset)) -> rawdata

#SMOOTHING ETC
rawdata %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Pool, Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% group_by(Pool, Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 200, FUN = median, align = "center"))) -> rawdata_smoothed #%>% na.omit()

################################################################################
# COMBINE WITH ALLDATA FILES
################################################################################

alldata <- rbind(alldata, rawdata)

alldata_called <- readRDS(file = "Oct23_alldata_called.rds")
alldata_called <- rbind(alldata_called, rawdata_called)

alldata_smoothed <- readRDS(file = "Oct23_alldata_smoothed.rds")
alldata_smoothed <- rbind(alldata_smoothed, rawdata_smoothed)

# saveRDS(alldata, file = "allseqruns_dec23.rds")
# saveRDS(alldata_called, file = "alldata_called.rds")
# saveRDS(alldata_smoothed, file = "alldata_smoothed.rds")

################################################################################
# for quick use, add in as run here:
rawdata_smoothed -> Hsomething_smoothed
Hsomething_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> Hsomething_counts

```

```{r}
# alldata %>% #head(3000) %>% 
#   merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>% 
#   group_by(Dataset, CHROM, POS) %>%
#   mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
#          Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
#   select(Pool, Dataset, POS, CHROM, Coverage, Wine, Oak) %>% 
#   pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> alldata_called
# 
# alldata_called %>% group_by(Pool, Dataset, CHROM, Allele) %>% arrange(POS) %>%
#   reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 200, FUN = median, align = "center"))) -> alldata_smoothed #%>% na.omit() 
# 
# saveRDS(alldata_called, file = "Oct23_alldata_called.rds")
# saveRDS(alldata_smoothed, file = "Oct23_alldata_smoothed.rds")

alldata_called <- readRDS(file = "Oct23_alldata_called.rds")
alldata_smoothed <- readRDS(file = "Oct23_alldata_smoothed.rds")

```

Use _e dataset to ensure that the ends of the chromosomes are not there

```{r, fig.width = 15, fig.height = 4}
alldata_smoothed %>%
  filter(POS > 30000) %>% 
  merge(ChromosomeScale[,c("CHROM", "End")]) %>% 
  filter(End > 0) %>% 
  group_by(CHROM) %>% 
  filter(POS < End) -> alldata_smoothed_e
```

## Separate each out into independent datasets to run GLM

```{r}
# HNGLVDRXY = CuSO4a (CSS1)
alldata_smoothed_e %>% filter(Pool == "HNGLVDRXY") -> HNGLVDRXY_smoothed
HNGLVDRXY_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HNGLVDRXY_counts

# HKTFTDRX2 = CuSO4b (CSS1)
alldata_smoothed_e %>% filter(Pool == "HKTFTDRX2") -> HKTFTDRX2_smoothed
HKTFTDRX2_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HKTFTDRX2_counts

# HVYTYDRX2 = CSS8 Fluc and CuSO4, also CSS1
alldata_smoothed_e %>% filter(Pool == "HVYTYDRX2") -> HVYTYDRX2_smoothed
HVYTYDRX2_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HVYTYDRX2_counts

# HGVMDRX2 = Fluconazole CSS I
alldata_smoothed_e %>% filter(Pool == "HGVMVDRX2") -> HGVMVDRX2_smoothed
HGVMVDRX2_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HGVMVDRX2_counts

# HJ5HKDRX3 = H2O2 data
alldata_smoothed_e %>% filter(Pool == "HJ5HKDRX3") -> HJ5HKDRX3_smoothed
HJ5HKDRX3_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HJ5HKDRX3_counts

# HKTMWDRX2 = Zeocin/Cycloheximide
alldata_smoothed_e %>% filter(Pool == "HKTMWDRX2") -> HKTMWDRX2_smoothed
HKTMWDRX2_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HKTMWDRX2_counts

# ADD NEW ONES DOWN HERE


```

## Get each run GLM ready

```{r}

```

# Analysis Standard

### Fluc CSS  VIII - DONE (Z score)

```{r, eval = FALSE}

# Pick out the specific groups of the dataset
HVYTYDRX2_smoothed %>%  separate(Dataset, into = c("Parent", "Rep_Bulk_Dose"), sep = "_") %>% 
  #CHANGE THE REPS ETC
  separate(Rep_Bulk_Dose, into = c("Rep", "Bulk", "Dose"), sep="(?<=[a-z,A-Z,0-9])") %>%
  #cont
  mutate(Parent = gsub("0", "O", Parent)) -> HVYTYDRX2_FlucCuSO4_81

#################################################################################
HVYTYDRX2_FlucCuSO4_81 %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '8')) %>% 
  filter(Bulk != "C") %>%
  mutate_if(is.character, as.factor) -> HVYTYDRX2_Fluc_CSS8_glm_prep

HVYTYDRX2_FlucCuSO4_81 %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '1')) %>% 
  filter(Bulk != "C") %>%
  mutate_if(is.character, as.factor) -> HVYTYDRX2_Fluc_CSS1_glm_prep

HVYTYDRX2_FlucCuSO4_81 %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '8')) %>% 
  filter(Bulk != "F") %>%
  mutate_if(is.character, as.factor) -> HVYTYDRX2_CuSO4_CSS8_glm_prep

#################################################################################
#Test once
testdata <- HVYTYDRX2_Fluc_CSS8_glm_prep %>% filter(CHROM == "I", POS == 60526) %>% arrange(Bulk, Parent, Rep)
table(testdata$Rep, testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Rep, testdata$Bulk, testdata$Parent))$Freq) == 16
formula = "Allele ~ Bulk * Parent + Rep"
glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HVYTYDRX2_Fluc_CSS8_glm_prep %>% select(-Dose) %>% na.omit() %>%
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                            numgroups = 16, outputlength = 5),
            Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> HVYTYDRX2_Fluc_CSS8_glm_all

saveRDS(HVYTYDRX2_Fluc_CSS8_glm_all, file = "StandardGLM/HVYTYDRX2_Fluc_CSS8_glm_all.rds")
```

### H2O2 CSS  VIII (DONE - effect?)

```{r, eval = FALSE}

# Pick out the specific groups of the dataset
HJ5HKDRX3_smoothed %>% separate(Dataset, into = c("Bulk", "Parent", "Rep"), sep = "_") %>% 
  mutate(Parent = gsub("0", "O", Parent)) %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '8')) %>% 
  filter(Bulk != "Z") %>%
  mutate_if(is.character, as.factor) -> HJ5HKDRX3_CSS8_glm_prep

#Test once
testdata <- HJ5HKDRX3_CSS8_glm_prep %>% na.omit() %>% filter(CHROM == "X", POS == 640028) %>% arrange(Bulk, Parent, Rep)
table(testdata$Rep, testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Rep, testdata$Bulk, testdata$Parent))$Freq) == 16
formula = "Allele ~ Bulk * Parent + Rep"
glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HJ5HKDRX3_CSS8_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                            numgroups = 16, outputlength = 5),
            Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> HJ5HKDRX3_CSS8_glm_all

HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HJ5HKDRX3", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HJ5HKDRX3_CSS8_merge_all

saveRDS(HJ5HKDRX3_CSS8_glm_all, file = "StandardGLM/HJ5HKDRX3_CSS8_glm_all.rds")
```


### H2O2 CSS  VIII (DONE - effect?)
```{r, eval = FALSE}

# Pick out the specific groups of the dataset
HJ5HKDRX3_smoothed %>% separate(Dataset, into = c("Bulk", "Parent", "Rep"), sep = "_") %>% 
  mutate(Parent = gsub("0", "O", Parent)) %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '8')) %>% 
  filter(Bulk != "Z") %>%
  mutate_if(is.character, as.factor) -> HJ5HKDRX3_CSS8_glm_prep

#Test once
testdata <- HJ5HKDRX3_CSS8_glm_prep %>% na.omit() %>% filter(CHROM == "X", POS == 640028) %>% arrange(Bulk, Parent, Rep)
table(testdata$Rep, testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Rep, testdata$Bulk, testdata$Parent))$Freq) == 16
formula = "Allele ~ Bulk * Parent + Rep"
glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HJ5HKDRX3_CSS8_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                            numgroups = 16, outputlength = 5),
            Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> HJ5HKDRX3_CSS8_glm_all

HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HJ5HKDRX3", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HJ5HKDRX3_CSS8_merge_all

saveRDS(HJ5HKDRX3_CSS8_glm_all, file = "StandardGLM/HJ5HKDRX3_CSS8_glm_all.rds")
```

### HNGLVDRXY = CuSO4a, CSS I - IP (Z score)

```{r, eval = FALSE}

# Pick out the specific groups of the dataset
HNGLVDRXY_smoothed %>% #head() %>%
    mutate(Dataset = gsub("CuSO4_CSSI_", "", Dataset)) %>%
    mutate(Dataset = gsub("Unselected", "Unselected_", Dataset)) %>%
    mutate(Dataset = gsub("Selected", "Selected_", Dataset)) %>%
  separate(Dataset, into = c("Bulk", "Parent"), sep = "_") %>% 
  mutate_if(is.character, as.factor) -> HNGLVDRXY_CSS1_glm_prep

#Test once
# testdata <- HNGLVDRXY_CSS1_glm_prep %>% na.omit() %>% filter(CHROM == "II", POS == 65975) %>% arrange(Bulk, Parent)
# table(testdata$Bulk, testdata$Parent)
# formula = "Allele ~ Bulk * Parent"
# glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HNGLVDRXY_CSS1_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             #Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 8, outputlength = 4),
            #MAKE SURE THIS IS THE SAME LENGTH AS OUTPUT LENGTH
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HNGLVDRXY_CuSO4_CSS1_glm_all

#HNGLVDRXY_CuSO4_CSS1_glm_all %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HNGLVDRXY", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HNGLVDRXY_CSS8_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HNGLVDRXY_CSS8_merge_all

saveRDS(HNGLVDRXY_CuSO4_CSS1_glm_all, file = "StandardGLM/HNGLVDRXY_CuSO4_CSS1_glm_all.rds")
```

# Duplicating reps to balance them (should have fixed effect of 0)
### HGVMDRX2 = Fluconazole CSS I (DONE - effect?)
 
```{r, eval = FALSE}
# Pick out the specific groups of the dataset
HGVMVDRX2_smoothed %>% separate(Dataset, into = c("Parent", "Bulk", "Rep"), sep = "_") %>% 
  mutate_if(is.character, as.factor) -> HGVMVDRX2_CSS1_glm_prep

#DUPLICATE MISSING REPLICATES (this experiment only)
HGVMVDRX2_CSS1_glm_prep %>% filter(Parent == "OakI", Bulk == "Dilute", Rep %in% c("B")) %>% 
  mutate(Rep = "C") %>% rbind(HGVMVDRX2_CSS1_glm_prep) -> HGVMVDRX2_CSS1_glm_prepd
HGVMVDRX2_CSS1_glm_prep %>% filter(Parent == "OakI", Bulk == "Dilute", Rep %in% c("B")) %>% 
  mutate(Rep = "D") %>% rbind(HGVMVDRX2_CSS1_glm_prepd) -> HGVMVDRX2_CSS1_glm_prepd
HGVMVDRX2_CSS1_glm_prep %>% filter(Parent == "WineI", Bulk == "Fluc", Rep %in% c("C")) %>% 
  mutate(Rep = "B") %>% rbind(HGVMVDRX2_CSS1_glm_prepd) -> HGVMVDRX2_CSS1_glm_prepd


#Test once
testdata <- HGVMVDRX2_CSS1_glm_prep %>% na.omit() %>% filter(CHROM == "I", POS == 128455) %>% arrange(Bulk, Parent, Rep)
table(testdata$Rep, testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Rep, testdata$Bulk, testdata$Parent))$Freq)
formula = "Allele ~ Bulk * Parent + Rep"
glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HGVMVDRX2_CSS1_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640021) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                            numgroups = 18, outputlength = 6),
            Factor = (c("Intercept", "Bulk", "Parent", "Rep1", "Rep2", "Interaction"))) -> HGVMVDRX2_CSS1_glm_all

HGVMVDRX2_CSS1_glm_all %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HGVMVDRX2", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HGVMVDRX2_CSS1_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HGVMVDRX2_CSS1_merge_all

saveRDS(HGVMVDRX2_CSS1_glm_all, file = "StandardGLM/HGVMVDRX2_CSS1_glm_all.rds")
```

# Averaging across replicates to fill in missing data (HJ5HKDRX3 and HKTMWDRX2)

### H2O2 
```{r, fig.width = 15, fig.height = 4}
HJ5HKDRX3_CSS8_glm_prep %>% group_by(CHROM, POS, Bulk, Parent, Allele) %>% na.omit() %>% summarize(AvgCount = mean(SmoothCount, na.rm = TRUE)) -> HJ5HKDRX3_CSS8_glm_prep_noreps

#Test once
testdata <- HJ5HKDRX3_CSS8_glm_prep_noreps %>% na.omit() %>% filter(CHROM == "I", POS == 30782) %>% arrange(Bulk, Parent)
table(testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Bulk, testdata$Parent))$Freq)
formula = "Allele ~ Bulk * Parent"
glm(formula = formula, family = "binomial", data = testdata, weights = AvgCount)

# Run full dataset
HJ5HKDRX3_CSS8_glm_prep_noreps %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             W = AvgCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 8, outputlength = 4),
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HJ5HKDRX3_CSS8_glm_all_noreps

HJ5HKDRX3_CSS8_glm_all_noreps %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HJ5HKDRX3", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HJ5HKDRX3_CSS8_merge_all

saveRDS(HJ5HKDRX3_CSS8_glm_all_noreps, file = "StandardGLM/Z_HJ5HKDRX3_CSS8_glm_all_noreps.rds")

ChromosomeScale %>% select(CHROM, POS) %>% mutate(Summary = NA, Factor = NA) -> ChromScale3

HJ5HKDRX3_CSS8_glm_all_noreps %>%  rbind(ChromScale3) %>% ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")
```

```{r, fig.width = 15, fig.height = 4}
HJ5HKDRX3_CSS1_glm_prep %>% group_by(CHROM, POS, Bulk, Parent, Allele) %>% na.omit() %>% summarize(AvgCount = mean(SmoothCount, na.rm = TRUE)) -> HJ5HKDRX3_CSS1_glm_prep_noreps

#Test once
testdata <- HJ5HKDRX3_CSS1_glm_prep_noreps %>% na.omit() %>% filter(CHROM == "I", POS == 30479) %>% arrange(Bulk, Parent)
table(testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Bulk, testdata$Parent))$Freq)
formula = "Allele ~ Bulk * Parent"
glm(formula = formula, family = "binomial", data = testdata, weights = AvgCount)

# Run full dataset
HJ5HKDRX3_CSS1_glm_prep_noreps %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640021) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             W = AvgCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 1, outputlength = 4),
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HJ5HKDRX3_CSS1_glm_all_noreps

HJ5HKDRX3_CSS1_glm_all_noreps %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HJ5HKDRX3", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HJ5HKDRX3_CSS1_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HJ5HKDRX3_CSS1_merge_all

saveRDS(HJ5HKDRX3_CSS1_glm_all_noreps, file = "StandardGLM/Z_HJ5HKDRX3_CSS1_glm_all_noreps.rds")

ChromosomeScale %>% select(CHROM, POS) %>% mutate(Summary = NA, Factor = NA) -> ChromScale3

HJ5HKDRX3_CSS1_glm_all_noreps %>%  rbind(ChromScale3) %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")
```

### Zeocin and Cycloheximide CSSI

Do the same no-rep thing for Zeocin and Cycloheximide to see if that helps

```{r, fig.width = 15, fig.height = 4}

# Zeocin
HKTMWDRX2_smoothed %>% 
  #Remove or add "Rep" and "Dose" etc
  separate(Dataset, into = c("Parent", "Bulk", "Rep"), sep="(?<=[A-Z])") %>% 
  mutate(Parent = gsub("0", "O", Parent)) %>%
  
  #Filter for Chr 1 and Bulk D or H
  filter(Bulk != "C") %>%

  #Make everything factors
  mutate_if(is.character, as.factor) -> HKTMWDRX2_Zeo_CSS1_glm_prep

# Cycloheximide
HKTMWDRX2_smoothed %>% 
  #Remove or add "Rep" and "Dose" etc
  separate(Dataset, into = c("Parent", "Bulk", "Rep"), sep="(?<=[A-Z])") %>% 
  mutate(Parent = gsub("0", "O", Parent)) %>%
  
  #Filter for Chr 1 and Bulk D or H
  filter(Bulk != "Z") %>%

  #Make everything factors
  mutate_if(is.character, as.factor) -> HKTMWDRX2_Cyc_CSS1_glm_prep
```

```{r, fig.width = 15, fig.height = 4}
HKTMWDRX2_Cyc_CSS1_glm_prep %>% group_by(CHROM, POS, Bulk, Parent, Allele) %>% na.omit() %>% summarize(AvgCount = mean(SmoothCount, na.rm = TRUE)) -> HKTMWDRX2_Cyc_CSS1_glm_prep_noreps

HKTMWDRX2_Cyc_CSS1_glm_prep_noreps %>% filter(Parent == "W", CHROM == "I")

#Test once
testdata <- HKTMWDRX2_Cyc_CSS1_glm_prep_noreps %>% na.omit() %>% filter(CHROM == "I", POS == 34991) %>% arrange(Bulk, Parent)
table(testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Bulk, testdata$Parent))$Freq)
formula = "Allele ~ Bulk * Parent"
glm(formula = formula, family = "binomial", data = testdata, weights = AvgCount)

HKTMWDRX2_Cyc_CSS1_glm_prep_noreps %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640021) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             W = AvgCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 1, outputlength = 4),
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HKTMWDRX2_Cyc_CSS1_glm_all_noreps

#PLOT TO CHECK
HKTMWDRX2_Cyc_CSS1_glm_all_noreps %>% rbind(ChromScale3) %>% filter(CHROM != "I") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_vline(data = myloci, aes(xintercept = POS, color = cat), linewidth = 1, linetype = "dashed") +
  facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("Cycloheximide CSS I (Rep Avg)")

```

```{r, fig.width = 15, fig.height = 4}
HKTMWDRX2_Zeo_CSS1_glm_prep %>% group_by(CHROM, POS, Bulk, Parent, Allele) %>% na.omit() %>% summarize(AvgCount = mean(SmoothCount, na.rm = TRUE)) -> HKTMWDRX2_Zeo_CSS1_glm_prep_noreps

#Test once
# testdata <- HKTMWDRX2_Zeo_CSS1_glm_prep_noreps %>% na.omit() %>% filter(CHROM == "I", POS == 34991) %>% arrange(Bulk, Parent)
# table(testdata$Bulk, testdata$Parent)
# sum(as.data.frame(table(testdata$Bulk, testdata$Parent))$Freq)
# formula = "Allele ~ Bulk * Parent"
# glm(formula = formula, family = "binomial", data = testdata, weights = AvgCount)

HKTMWDRX2_Zeo_CSS1_glm_prep_noreps %>% na.omit() %>% 
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             W = AvgCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 1, outputlength = 4),
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HKTMWDRX2_Zeo_CSS1_glm_all_noreps

#PLOT TO CHECK
HKTMWDRX2_Zeo_CSS1_glm_all_noreps %>% rbind(ChromScale3) %>% filter(CHROM != "I") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() +
  geom_vline(data = myloci, aes(xintercept = POS, color = cat), linewidth = 1, linetype = "dashed") +
facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("Zeocin CSS I (Rep Avg)")

```

