---
title: "AllSequencingRuns Figures"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
require(ggplot2)

require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)

require(cybrBSA)


#TESTING
#glm_formula <- "Allele ~ Bulk * Parent + Rep"

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))

```



## Load in Data

### Cataloging samples

Which samples are which?

```{r}
MQCRuns <- read.csv("C:\\Users\\cassa\\OneDrive\\Documents\\SiegalLab\\Sequencing_Tuboweb\\AllMultiQCRuns.csv")

# MQCRuns %>% select(-DoseUnits, -SurvivalPercentage, -Dose, -Barcodes, -PercUndetermined) %>% 
#   na.omit() %>% 
#   filter(grepl("undetermined", Library) == FALSE) %>%
#   pivot_wider(names_from = Bulk, values_from = Parent) %>% select(Pool, ShortName, MQCDate, Dilute, CuSO4, Fluconazole, Cycloheximide, H2O2, Zeocin) %>% distinct() %>% arrange(Pool)

MQCRuns %>% select(-DoseUnits, -SurvivalPercentage, -Dose, -Barcodes, -PercUndetermined) %>% 
  na.omit() %>% 
  filter(grepl("undetermined", Library) == FALSE) %>%
  pivot_wider(names_from = Parent, values_from = Bulk) %>% select(Pool, ShortName, MQCDate, Wine1, Oak1, Wine8, Oak8) %>% distinct() %>% arrange(Pool) -> mqcbyparent

table(mqcbyparent$Pool, mqcbyparent$Wine1)
table(mqcbyparent$Pool, mqcbyparent$Wine8)


#MQCRuns %>% filter(Pool == "HKTMWDRX2")
```


### GLM Outputs

```{r}
#files <- as.data.frame(x = dir("StandardGLM"))
#files %>% filter(grepl("glm_all.rds",`dir("StandardGLM")`)) %>% transmute(files = `dir("StandardGLM")`) -> all_files

#Copper
HNGLVDRXY_CuSO4_CSS1_glm_all <- readRDS("StandardGLM/HNGLVDRXY_CuSO4_CSS1_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))	 #Cu1	
HKTFTDRX2_CuSO4_CSS1_glm_all <- readRDS("StandardGLM/HKTFTDRX2_CuSO4_CSS1_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu1			
xHNGLVDRXY_CuSO4_CSS1_glm_all <- readRDS("StandardGLM/xHNGLVDRXY_CuSO4_CSS1_glm_all.rds")%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))	#Cu1
HVYTYDRX2_CuSO4_CSS8_glm_all <- readRDS("StandardGLM/HVYTYDRX2_CuSO4_CSS8_glm_all.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #Cu8

#Fluc
HGVMVDRX2_CSS1_glm_alld <- readRDS("StandardGLM/HGVMVDRX2_CSS1_glm_alld.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))	        #fluc1 duplicating reps
HVYTYDRX2_Fluc_CSS8_glm_all <- readRDS("StandardGLM/HVYTYDRX2_Fluc_CSS8_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #fluc8		
HGVMVDRX2norep_CSS1_glm_all <- readRDS("StandardGLM/HGVMVDRX2norep_CSS1_glm_all.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000))))) #fluc1 without replicates


#H2O2
HJ5HKDRX3_CSS1_glm_all <- readRDS("StandardGLM/HJ5HKDRX3_CSS1_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))			#H2O21
HJ5HKDRX3_CSS8_glm_all <- readRDS("StandardGLM/HJ5HKDRX3_CSS8_glm_all.rds")	%>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))			#H2O28
HJ5HKDRX3_CSS8_glm_all_noreps <- readRDS("StandardGLM/HJ5HKDRX3_CSS8_glm_all_noreps.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))

#Zeocin + Cycloheximide 
HKTMWDRX2_Cyc_CSS1_glm_all_noreps <- readRDS("StandardGLM/HKTMWDRX2_Cyc_CSS1_glm_all_noreps.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))
HKTMWDRX2_Zeo_CSS1_glm_all_noreps <- readRDS("StandardGLM/HKTMWDRX2_Zeo_CSS1_glm_all_noreps.rds") %>% 
  mutate(CHROM = factor(CHROM, levels = as.character(as.roman(c(1:16,1000)))))



```

### GLM Cutoffs

```{r}
#CuSO4
HNGLVDRXY_CSS1_chrparentQs <- readRDS("StandardGLM/HNGLVDRXY_CSS1_chrparentQs.rds") %>% mutate(ds = "HNGLVDRXY", CSS = "I")
HNGLVDRXY_CSS1_parQs <- readRDS("StandardGLM/HNGLVDRXY_CSS1_parQs.rds") %>% mutate(ds = "HNGLVDRXY", CSS = "I")

HKTFTDRX2_CSS1_chrparentQs <- readRDS("StandardGLM/HKTFTDRX2_CSS1_chrparentQs.rds") %>% mutate(ds = "HKTFTDRX2", CSS = "I")
HKTFTDRX2_CSS1_parQs <- readRDS("StandardGLM/HKTFTDRX2_CSS1_parQs.rds") %>% mutate(ds = "HKTFTDRX2", CSS = "I")

#Fluconazole
HVYTYDRX2_F_CSS8_chrparentQs <- readRDS("StandardGLM/HVYTYDRX2_F_CSS8_chrparentQs.rds")
HVYTYDRX2_F_CSS8_parQs <- readRDS("StandardGLM/HVYTYDRX2_F_CSS8_parQs.rds")

HGVMVDRX2_F_CSS1d_chrparentQs <- readRDS("StandardGLM/HGVMVDRX2_F_CSS1d_chrparentQs.rds") #duplicate reps
HGVMVDRX2_F_CSS1d_parQs <- readRDS("StandardGLM/HGVMVDRX2_F_CSS1d_parQs.rds")

HGVMVDRX2_F_CSS1_chrparentQs<- readRDS("StandardGLM/HGVMVDRX2_F_CSS1_chrparentQs.rds") #no replicates
HGVMVDRX2_F_CSS1_parQs<- readRDS("StandardGLM/HGVMVDRX2_F_CSS1_parQs.rds")

#H2O2
HJ5HKDRX3_H_CSS8_chrparentQs <- readRDS("StandardGLM/HJ5HKDRX3_H_CSS8_chrparentQs.rds")
HJ5HKDRX3_H_CSS8_parQs <- readRDS("StandardGLM/HJ5HKDRX3_H_CSS8_parQs.rds")

HJ5HKDRX3norep_CSS8_chrparentQs <- readRDS("StandardGLM/HJ5HKDRX3norep_CSS8_chrparentQs.rds")
HJ5HKDRX3norep_CSS8_parQs <- readRDS("StandardGLM/HJ5HKDRX3norep_CSS8_parQs.rds")

#Cycloheximide
HKTMWDRX2_Cyc_norep_CSS8_chrparentQs<- readRDS("StandardGLM/HKTMWDRX2_Cyc_norep_CSS8_chrparentQs.rds")
HKTMWDRX2_Cyc_norep_CSS8_parQs<- readRDS("StandardGLM/HKTMWDRX2_Cyc_norep_CSS8_parQs.rds")

#Zeocin
HKTMWDRX2_Zeo_norep_CSS8_chrparentQs<- readRDS("StandardGLM/HKTMWDRX2_Zeo_norep_CSS8_chrparentQs.rds")
HKTMWDRX2_Zeo_norep_CSS8_parQs<- readRDS("StandardGLM/HKTMWDRX2_Zeo_norep_CSS8_parQs.rds")

```


## Plot CuSO4 Data

Individual Dataset HNGLVDRXY

```{r, fig.height = 3, fig.width = 9}

xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_chrparentQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI Chromosome-Specific Cutoffs")

xHNGLVDRXY_CuSO4_CSS1_glm_all %>% merge(HNGLVDRXY_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI Factor-Specific Cutoffs")
```

Individual Set HKTFTDRX2_CSS1_parQs

```{r, fig.height = 3, fig.width = 9}
HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_chrparentQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI Chromosome-Specific Cutoffs")

HKTFTDRX2_CuSO4_CSS1_glm_all %>% merge(HKTFTDRX2_CSS1_parQs) %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("CuSO4 CSSI Factor-Specific Cutoffs")
```



```{r, fig.height = 3, fig.width = 9}
rbind(data.frame(HKTFTDRX2_CuSO4_CSS1_glm_all, ds = "HKTFTDRX2", CSS = "I"),
      data.frame(HVYTYDRX2_CuSO4_CSS8_glm_all, ds = "HVYTYDRX2", CSS = "VIII"),
      #data.frame(HNGLVDRXY_CSS8_glm_all, ds = "HNGLVDRXY", CSS = "I"),
      data.frame(xHNGLVDRXY_CuSO4_CSS1_glm_all, ds = "HNGLVDRXYx", CSS = "I")) -> CuSO4_Data

CuSO4_Data %>% filter(Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(ds~CHROM, scales = "free", space = "free") + ggtitle("CuSO4 Bulk Effects")

CuSO4_Data %>% filter(Factor == "Parent") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(ds~CHROM, scales = "free", space = "free") + ggtitle("CuSO4 Parent Effects")

CuSO4_Data %>% filter(Factor == "Interaction") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(ds~CHROM, scales = "free", space = "free") + ggtitle("CuSO4 Interaction Effects")

```

## Plot Fluconazole Data

### CSS1 (no replicates)

Because the replicates were not balanced in this experiment, and did not correspond with specific days, I ran a model without a replicate effect which should just average the replicates together at each locus. The permutations for this were then done the same way.

The permutations are currently broken so all of this will just not have cutoffs

```{r, fig.height = 3, fig.width = 9}

HGVMVDRX2norep_CSS1_glm_all %>% #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1_chrparentQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HGVMVDRX2norep_CSS1_glm_all %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```



### CSS1 (duplicated reps)

Because the replicates in this experiment are unbalanced, I duplicated the bulks which did not have a replicate to match the other two. This *should* result in a replicate effect of 0 for those samples, since they are fixed effects, and then allow the remaining replicates to capture the effects of difference by day. This is not verified yet, and I will be comparing this to the data without having any replicates (using the original CSSI script).

```{r, fig.height = 3, fig.width = 9}

################################################################################
HGVMVDRX2_CSS1_glm_alld %>% #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1d_chrparentQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HGVMVDRX2_CSS1_glm_alld %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1d_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```

### How similar are the replicates?

We can use the two replicate effects here, which result from the random assignment of a, b, or c, to see how consistent our cutoffs will be. Each cutoff should be run >100 times on the HPC to get a true mean for the FDR, but to get each of these loaded quickly, that step has not been completed as of 12/19/23.

```{r, fig.height = 3, fig.width = 9}
HGVMVDRX2_CSS1_glm_alld %>% #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1d_chrparentQs) %>% #CHANGE CUTOFF FILE
  
  filter(grepl("Rep", Factor)) %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

HGVMVDRX2_CSS1_glm_alld %>%  #CHANGE DATASET
  merge(HGVMVDRX2_F_CSS1d_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(grepl("Rep", Factor)) %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```


### How similar are the duplicated vs non-rep'd results?

```{r, fig.height = 3, fig.width = 9}
rbind(data.frame(HGVMVDRX2_CSS1_glm_alld, dataset = "Dup"),
      data.frame(HGVMVDRX2norep_CSS1_glm_all, dataset = "NoRep")) %>%
  filter(grepl("Rep", Factor) == FALSE) %>%
  
  ggplot(aes(x = POS, y = abs(Summary), color = dataset)) + geom_line() + 
  facet_grid(Factor~CHROM, space = "free", scales = "free") + 
  ggtitle("Fluc CSS I Comparison") #CHANGE TITLE
```

### CSS8
```{r, fig.height = 3, fig.width = 9}
################################################################################
HVYTYDRX2_Fluc_CSS8_glm_all %>% #CHANGE DATASET
  merge(HVYTYDRX2_F_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS VIII Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HVYTYDRX2_Fluc_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HVYTYDRX2_F_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Fluc CSS VIII Factor-Specific Cutoffs") #CHANGE TITLE
```

```{r, fig.height = 3, fig.width = 9}

rbind(data.frame(HGVMVDRX2_CSS1_glm_alld, Pool = "HGVMVDRX2", CSS = "I"), 
      data.frame(HVYTYDRX2_Fluc_CSS8_glm_all, Pool = "HVYTYDRX2", CSS = "VIII")) %>%
  filter(Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(Pool~CHROM, scales = "free", space = "free") + ggtitle("Fluconazole Bulk Effects")
  
 
rbind(data.frame(HGVMVDRX2_CSS1_glm_alld, Pool = "HGVMVDRX2", CSS = "I"), 
      data.frame(HVYTYDRX2_Fluc_CSS8_glm_all, Pool = "HVYTYDRX2", CSS = "VIII")) %>%
  filter(Factor == "Interaction") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(Pool~CHROM, scales = "free", space = "free") + ggtitle("Fluconazole Interaction Effects")
  

```

## Plot H2O2 Data

### Standard Analysis with Replicates
CSS 8

```{r, fig.height = 3, fig.width = 9}

################################################################################
HJ5HKDRX3_CSS8_glm_all %>% #CHANGE DATASET
  merge(HJ5HKDRX3_H_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HJ5HKDRX3_CSS8_glm_all %>%  #CHANGE DATASET
  merge(HJ5HKDRX3_H_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII Factor-Specific Cutoffs") #CHANGE TITLE
```

```{r, fig.height = 3, fig.width = 9}
rbind(data.frame(HJ5HKDRX3_CSS1_glm_all, Pool = "HJ5HKDRX3", CSS = "I"), 
      data.frame(HJ5HKDRX3_CSS8_glm_all, Pool = "HJ5HKDRX3", CSS = "VIII")) %>%
  filter(Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary), color = CSS)) + geom_line() + facet_grid(CSS~CHROM, scales = "free", space = "free") + ggtitle("Fluconazole Bulk Effects")
  
```


### Averaged values without replicates

```{r, fig.height = 3, fig.width = 9}

################################################################################
HJ5HKDRX3_CSS8_glm_all_noreps %>% #CHANGE DATASET
  merge(HJ5HKDRX3norep_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HJ5HKDRX3_CSS8_glm_all_noreps %>%  #CHANGE DATASET
  merge(HJ5HKDRX3norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("H2O2 CSS VIII Factor-Specific Cutoffs") #CHANGE TITLE
```

## Plot Zeocin Data

### Averaged values without replicates

```{r, fig.height = 3, fig.width = 9}

################################################################################
HKTMWDRX2_Zeo_CSS1_glm_all_noreps %>% #CHANGE DATASET
  merge(HKTMWDRX2_Zeo_norep_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Zeocin CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HKTMWDRX2_Zeo_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Zeo_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Zeocin CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```

## Plot Cycloheximide Data

### Averaged values without replicates

```{r, fig.height = 3, fig.width = 9}

################################################################################
HKTMWDRX2_Cyc_CSS1_glm_all_noreps %>% #CHANGE DATASET
  merge(HKTMWDRX2_Cyc_norep_CSS8_chrparentQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
  geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Cycloheximide CSS I Chromosome-Specific Cutoffs") #CHANGE TITLE

################################################################################
HKTMWDRX2_Cyc_CSS1_glm_all_noreps %>%  #CHANGE DATASET
  merge(HKTMWDRX2_Cyc_norep_CSS8_parQs) %>% #CHANGE CUTOFF FILE
  
  filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
  geom_hline(aes(yintercept = quant095, color = Factor), linetype = 1) + 
    geom_hline(aes(yintercept = quant099, color = Factor), linetype = 2) + 
  geom_hline(aes(yintercept = quant0995, color = Factor), linetype = 3) + 
  facet_grid(~CHROM, space = "free", scales = "free") + 
  
  ggtitle("Cycloheximide CSS I Factor-Specific Cutoffs") #CHANGE TITLE
```