---
title: "2023 Comparing Coverage and SNPs"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: hide
  html_notebook:
    code_folding: hide
---

```{r, warning=FALSE, message=FALSE, comment=FALSE, fig.width=18, fig.height=5}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
require(ggplot2)

require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)

require(cybrBSA)


#TESTING
#glm_formula <- "Allele ~ Bulk * Parent + Rep"

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))

```

# Functions

```{r}
glm_cb2_short <- function(..., W, formula, numgroups = FALSE, outputlength = 4, return = c("Z")) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  #MAYBEWORKS
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  #Return specific ones like z-score only
  if(return %in% "Z"){
    output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients)*0.5)+1):((length(summary(glm_fit)$coefficients)*0.75))]
  }
  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
}

############################ PERMUTATIONS ######################################
# Replicates, all scrambled
cybrPermute_Rep <- function(dataset){
  
  start.time <- Sys.time()
  
  dataset %>% 
  distinct() %>% ungroup() %>%
  group_by(CHROM, POS, Allele, 
           Bulk, 
           Rep, #might not have these
           Parent) %>% 
  summarize(culprits = length((SmoothCount))) %>% 
  merge(dataset) %>% 
  filter(culprits == 1) %>% 
  ungroup() %>%
  distinct() %>% #THIS IS IMPORTANT
  pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest

  #these are now all of the ones that can be used to permute
  newnewtest %>% filter(Bulk != "F", CHROM != "I", CHROM != "III", CHROM != "VIII") %>% select(CHROM, Oak, Wine) %>% 
    mutate(rownum = sample(1:nrow(.))) %>% arrange(rownum) %>% na.omit() -> NullToSample
  
  #Trying this again
  
  dataset[1:nrow(NullToSample),] %>% ungroup() %>% select(CHROM, POS, Bulk, Parent, Rep) %>% cbind(NullToSample[,c("Oak", "Wine")]) %>%
    pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") %>%
    #Original Script
    group_by(CHROM, POS) %>%
    mutate_if(is.character, as.factor) %>%
      summarize(Summary = glm_cb2_short(Allele = Allele,
                               Bulk = Bulk,
                               Parent = Parent,
                               Rep = Rep,
                               W = SmoothCount,
                               formula = "Allele ~ Bulk * Parent + Rep",
                               outputlength = 5),
              Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> glmresult

  end.time = Sys.time()
  print(end.time - start.time)
  
  return(glmresult)
  
}

# No replicates, fix CHROM, Parent
# No replicates, fix Parent

# Replicates, fix CHROM, Parent
cybrPermute_byCHRParent_Rep <- function(dataset){
   start.time <- Sys.time()
  
   print("Make sure that dilute bulk is labeled D")
   
    dataset %>% 
      distinct() %>% ungroup() %>%
      group_by(CHROM, POS, Allele, 
               Bulk, 
               Rep, #might not have these
               Parent) %>% 
      summarize(culprits = length((SmoothCount))) %>% 
      merge(dataset) %>% 
      filter(culprits == 1) %>% 
      ungroup() %>%
      distinct() %>% #THIS IS IMPORTANT
      pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    
    #these are now all of the ones that can be used to permute
    newnewtest %>% filter(Bulk == "D") %>% select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(CHROM, Parent) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine, POS = POS,
                POS2 = sample(POS)) %>%
      group_by(CHROM, Parent, POS2) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine,
                POS = POS2, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA
    
    newnewtest %>% filter(Bulk == "D") %>% select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(CHROM, Parent) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine, POS = POS,
                POS2 = sample(POS)) %>%
      group_by(CHROM, Parent, POS2) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine,
                POS = POS2, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB
    
    rbind(shuffled_DiluteA, shuffled_DiluteB) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm
    
    #Trying this again
    
    shuffletoglm %>% na.omit() %>%
      #Original Script
      group_by(CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_short(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = "Allele ~ Bulk * Parent + Rep",
                                numgroups = 16, outputlength = 5),
                Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> glmresult
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}

# Replicates, fix Parent
cybrPermute_byParent_Rep <- function(dataset){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled D")

  dataset %>% 
  distinct() %>% ungroup() %>%
  group_by(CHROM, POS, Allele, 
           Bulk, 
           Rep, #might not have these
           Parent) %>% 
  summarize(culprits = length((SmoothCount))) %>% 
  merge(dataset) %>% 
  filter(culprits == 1) %>% 
  ungroup() %>%
  distinct() %>% #THIS IS IMPORTANT
  pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
  
  #these are now all of the ones that can be used to permute
    newnewtest %>% filter(Bulk == "D",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, 
                Loc = sample(Loc)) %>%
      group_by(Parent, Loc) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine,
                Loc = Loc, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "D",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, 
                Loc = sample(Loc)) %>%
      group_by(Parent, Loc) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine,
                Loc = Loc, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #Trying this again
    
    shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_short(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = "Allele ~ Bulk * Parent + Rep",
                                numgroups = 16, outputlength = 5),
                Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> glmresult2
  
    end.time = Sys.time()
    print(end.time - start.time)
    
    return(glmresult2)
}
  
```

Note: if looking for smoothed vs unsmoothed data, the `Figures_2023_BiologyofGenomes.rmd` file has these at the beginning.

## Load in data for GLM analysis

Load all smoothed files from AllSequencingRuns_PreProcess

```{r}
HNGLVDRXY_smoothed<- readRDS(file = "StandardGLM/HNGLVDRXY_smoothed.rds")
HKTFTDRX2_smoothed<- readRDS(file = "StandardGLM/HKTFTDRX2_smoothed.rds")
HVYTYDRX2_smoothed<- readRDS(file = "StandardGLM/HVYTYDRX2_smoothed.rds")
HGVMVDRX2_smoothed<- readRDS(file = "StandardGLM/HGVMVDRX2_smoothed.rds")
HJ5HKDRX3_smoothed<- readRDS(file = "StandardGLM/HJ5HKDRX3_smoothed.rds")
HKTMWDRX2_smoothed<- readRDS(file = "StandardGLM/HKTMWDRX2_smoothed.rds")
xHNGLVDRXY_smoothed<- readRDS(file = "StandardGLM/xHNGLVDRXY_smoothed.rds")
```

# Analysis Standard

### HVYTYDRX2 into Fluc8, CuSO48, and Fluc1
```{r, eval = FALSE}

# Pick out the specific groups of the dataset
HVYTYDRX2_smoothed %>%  separate(Dataset, into = c("Parent", "Rep_Bulk_Dose"), sep = "_") %>% 
  #CHANGE THE REPS ETC
  separate(Rep_Bulk_Dose, into = c("Rep", "Bulk", "Dose"), sep = "(?<=[a-z,A-Z,0-9])") %>%
  #cont
  mutate(Parent = gsub("0", "O", Parent)) -> HVYTYDRX2_FlucCuSO4_81

#################################################################################
HVYTYDRX2_FlucCuSO4_81 %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '8')) %>% 
  filter(Bulk != "C") %>%
  mutate_if(is.character, as.factor) -> HVYTYDRX2_Fluc_CSS8_glm_prep

HVYTYDRX2_FlucCuSO4_81 %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '1')) %>% 
  filter(Bulk != "C") %>%
  mutate_if(is.character, as.factor) -> HVYTYDRX2_Fluc_CSS1_glm_prep

HVYTYDRX2_FlucCuSO4_81 %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '8')) %>% 
  filter(Bulk != "F") %>%
  mutate_if(is.character, as.factor) -> HVYTYDRX2_CuSO4_CSS8_glm_prep

#SAVE FOR PERMUTATIONS
saveRDS(HVYTYDRX2_Fluc_CSS8_glm_prep, file = "StandardGLM/HVYTYDRX2_Fluc_CSS8_glm_prep.rds")
saveRDS(HVYTYDRX2_Fluc_CSS1_glm_prep, file = "StandardGLM/HVYTYDRX2_Fluc_CSS1_glm_prep.rds")
saveRDS(HVYTYDRX2_CuSO4_CSS8_glm_prep, file = "StandardGLM/HVYTYDRX2_CuSO4_CSS8_glm_prep.rds")

```

### Fluc CSS VIII - DONE (Z score)

```{r}
#Test once
testdata <- HVYTYDRX2_Fluc_CSS8_glm_prep %>% filter(CHROM == "I", POS == 60526) %>% arrange(Bulk, Parent, Rep)
table(testdata$Rep, testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Rep, testdata$Bulk, testdata$Parent))$Freq) == 16
formula = "Allele ~ Bulk * Parent + Rep"
glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HVYTYDRX2_Fluc_CSS8_glm_prep %>% select(-Dose) %>% na.omit() %>%
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                             outputlength = 5),
            Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> HVYTYDRX2_Fluc_CSS8_glm_all

saveRDS(HVYTYDRX2_Fluc_CSS8_glm_all, file = "StandardGLM/HVYTYDRX2_Fluc_CSS8_glm_all.rds")
```


### CuSO4 CSS VIII - DONE

```{r, eval = FALSE}

#Test once
testdata <- HVYTYDRX2_CuSO4_CSS8_glm_prep %>% filter((Parent == "O8" & Dose == 6) |
                                         (Parent == "W8" & Dose == 5) | 
                                         (Parent == "O8" & is.na(Dose) == T) |
                                         (Parent == "W8" & is.na(Dose) == T) ) %>% filter(CHROM == "I", POS == 106953) %>% arrange(Bulk, Parent, Rep)
table(testdata$Rep, testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Rep, testdata$Bulk, testdata$Parent))$Freq) == 16
formula = "Allele ~ Bulk * Parent + Rep"
glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

#This isn't right
HVYTYDRX2_CuSO4_CSS8_glm_prep %>% filter((Parent == "O8" & Dose == 6) |
                                         (Parent == "W8" & Dose == 5) | 
                                         (Parent == "O8" & is.na(Dose) == T) |
                                         (Parent == "W8" & is.na(Dose) == T) ) %>%
# Run full dataset
#HVYTYDRX2_CuSO4_CSS8_glm_prep %>% 
  select(-Dose) %>% na.omit() %>%
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                            numgroups = 16, outputlength = 5),
            Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> HVYTYDRX2_CuSO4_CSS8_glm_all

saveRDS(HVYTYDRX2_CuSO4_CSS8_glm_all, file = "StandardGLM/HVYTYDRX2_CuSO4_CSS8_glm_all.rds")
```

### H2O2 CSS  VIII - DONE (Z score)

```{r, eval = FALSE}

# Pick out the specific groups of the dataset
HJ5HKDRX3_smoothed %>% separate(Dataset, into = c("Bulk", "Parent", "Rep"), sep = "_") %>% 
  mutate(Parent = gsub("0", "O", Parent)) %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '8')) %>% 
  filter(Bulk != "Z") %>%
  mutate_if(is.character, as.factor) -> HJ5HKDRX3_CSS8_glm_prep

#SAVE FOR PERMUTATIONS
saveRDS(HJ5HKDRX3_CSS8_glm_prep, file = "StandardGLM/HJ5HKDRX3_CSS8_glm_prep.rds")

#Test once
# testdata <- HJ5HKDRX3_CSS8_glm_prep %>% na.omit() %>% filter(CHROM == "X", POS == 640028) %>% arrange(Bulk, Parent, Rep)
# table(testdata$Rep, testdata$Bulk, testdata$Parent)
# sum(as.data.frame(table(testdata$Rep, testdata$Bulk, testdata$Parent))$Freq) == 16
# formula = "Allele ~ Bulk * Parent + Rep"
# glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HJ5HKDRX3_CSS8_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                            numgroups = 16, outputlength = 5),
            Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> HJ5HKDRX3_CSS8_glm_all

HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HJ5HKDRX3", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HJ5HKDRX3_CSS8_merge_all

saveRDS(HJ5HKDRX3_CSS8_glm_all, file = "StandardGLM/HJ5HKDRX3_CSS8_glm_all.rds")

#HJ5HKDRX3_CSS8_perm <- cybrPermute_Rep(dataset = HJ5HKDRX3_CSS8_glm_prep)

```


### CuSO4a, CSS I - DONE (Z score)
Redoing 11/28 to add back in Chr 8
```{r, eval = FALSE}

# Pick out the specific groups of the dataset
HNGLVDRXY_smoothed %>% #head() %>%
    mutate(Dataset = gsub("CuSO4_CSSI_", "", Dataset)) %>%
    mutate(Dataset = gsub("Unselected", "Unselected_", Dataset)) %>%
    mutate(Dataset = gsub("Selected", "Selected_", Dataset)) %>%
  separate(Dataset, into = c("Bulk", "Parent"), sep = "_") %>% 
  mutate_if(is.character, as.factor) -> HNGLVDRXY_CSS1_glm_prep

#SAVE FOR PERMUTATIONS
saveRDS(HNGLVDRXY_CSS1_glm_prep, file = "StandardGLM/HNGLVDRXY_CSS1_glm_prep.rds")

```

```{r, eval = FALSE}
#Test once
testdata <- HNGLVDRXY_CSS1_glm_prep %>% na.omit() %>% filter(CHROM == "II", POS == 65975) %>% arrange(Bulk, Parent)
table(testdata$Bulk, testdata$Parent)
formula = "Allele ~ Bulk * Parent"
coffee <- glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HNGLVDRXY_CSS1_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             #Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 8, outputlength = 4),
            #MAKE SURE THIS IS THE SAME LENGTH AS OUTPUT LENGTH
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HNGLVDRXY_CuSO4_CSS1_glm_all

#HNGLVDRXY_CuSO4_CSS1_glm_all %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HNGLVDRXY", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HNGLVDRXY_CSS8_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HNGLVDRXY_CSS8_merge_all

saveRDS(HNGLVDRXY_CuSO4_CSS1_glm_all, file = "StandardGLM/HNGLVDRXY_CuSO4_CSS1_glm_all.rds")
```

```{r, eval = FALSE}

# Pick out the specific groups of the dataset
xHNGLVDRXY_smoothed %>% #head() %>%
    mutate(Dataset = gsub("CuSO4_CSSI_", "", Dataset)) %>%
    mutate(Dataset = gsub("Unselected", "Unselected_", Dataset)) %>%
    mutate(Dataset = gsub("Selected", "Selected_", Dataset)) %>%
  separate(Dataset, into = c("Bulk", "Parent"), sep = "_") %>% 
  mutate_if(is.character, as.factor) -> xHNGLVDRXY_CSS1_glm_prep

#SAVE FOR PERMUTATIONS
saveRDS(xHNGLVDRXY_CSS1_glm_prep, file = "StandardGLM/xHNGLVDRXY_CSS1_glm_prep.rds")
```

```{r}
#Test once
# testdata <- xHNGLVDRXY_CSS1_glm_prep %>% na.omit() %>% filter(CHROM == "II", POS == 65975) %>% arrange(Bulk, Parent)
# table(testdata$Bulk, testdata$Parent)
# formula = "Allele ~ Bulk * Parent"
# glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
xHNGLVDRXY_CSS1_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             #Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 8, outputlength = 4),
            #MAKE SURE THIS IS THE SAME LENGTH AS OUTPUT LENGTH
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> xHNGLVDRXY_CuSO4_CSS1_glm_all

#xHNGLVDRXY_CuSO4_CSS1_glm_all %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "xHNGLVDRXY", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# xHNGLVDRXY_CSS8_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> xHNGLVDRXY_CSS8_merge_all

saveRDS(xHNGLVDRXY_CuSO4_CSS1_glm_all, file = "StandardGLM/xHNGLVDRXY_CuSO4_CSS1_glm_all.rds")
```

### CuSO4b, CSS I - DONE

```{r, eval = FALSE}

# Pick out the specific groups of the dataset
HKTFTDRX2_smoothed %>% 
  separate(Dataset, into = c("Parent", "Bulk", "Rep"), sep = "(?<=[a-z,A-Z,0-9])") %>% 
  mutate_if(is.character, as.factor) -> HKTFTDRX2_CSS1_glm_prep

saveRDS(HKTFTDRX2_CSS1_glm_prep, file = "StandardGLM/HKTFTDRX2_CSS1_glm_prep.rds")
#Test once
testdata <- HKTFTDRX2_CSS1_glm_prep %>% na.omit() %>% filter(CHROM == "I", POS == 152004) %>% arrange(Bulk, Parent)
table(testdata$Bulk, testdata$Parent)
formula = "Allele ~ Bulk * Parent + Rep"
glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HKTFTDRX2_CSS1_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>% 
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                            #numgroups = 16, 
                            outputlength = 5),
            #MAKE SURE THIS IS THE SAME LENGTH AS OUTPUT LENGTH
            Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> HKTFTDRX2_CuSO4_CSS1_glm_all


saveRDS(HKTFTDRX2_CuSO4_CSS1_glm_all, file = "StandardGLM/HKTFTDRX2_CuSO4_CSS1_glm_all.rds")

HKTFTDRX2_CuSO4_CSS1_glm_all %>% na.omit()
```




# Duplicating reps to balance them (should have fixed effect of 0)
### HGVMDRX2 = Fluconazole CSS I - DONE (Z)
 
```{r, eval = FALSE}
# Pick out the specific groups of the dataset
HGVMVDRX2_smoothed %>% separate(Dataset, into = c("Parent", "Bulk", "Rep"), sep = "_") %>% 
  mutate_if(is.character, as.factor) -> HGVMVDRX2_CSS1_glm_prep

#DUPLICATE MISSING REPLICATES (this experiment only)
HGVMVDRX2_CSS1_glm_prep %>% filter(Parent == "OakI", Bulk == "Dilute", Rep %in% c("B")) %>% 
  mutate(Rep = "C") %>% rbind(HGVMVDRX2_CSS1_glm_prep) -> HGVMVDRX2_CSS1_glm_prepd
HGVMVDRX2_CSS1_glm_prep %>% filter(Parent == "OakI", Bulk == "Dilute", Rep %in% c("B")) %>% 
  mutate(Rep = "D") %>% rbind(HGVMVDRX2_CSS1_glm_prepd) -> HGVMVDRX2_CSS1_glm_prepd
HGVMVDRX2_CSS1_glm_prep %>% filter(Parent == "WineI", Bulk == "Fluc", Rep %in% c("C")) %>% 
  mutate(Rep = "B") %>% rbind(HGVMVDRX2_CSS1_glm_prepd) -> HGVMVDRX2_CSS1_glm_prepd


#Test once
testdata <- HGVMVDRX2_CSS1_glm_prep %>% na.omit() %>% filter(CHROM == "I", POS == 128455) %>% arrange(Bulk, Parent, Rep)
table(testdata$Rep, testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Rep, testdata$Bulk, testdata$Parent))$Freq)
formula = "Allele ~ Bulk * Parent + Rep"
glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HGVMVDRX2_CSS1_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640021) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                            numgroups = 18, outputlength = 6),
            Factor = (c("Intercept", "Bulk", "Parent", "Rep1", "Rep2", "Interaction"))) -> HGVMVDRX2_CSS1_glm_all

HGVMVDRX2_CSS1_glm_all %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HGVMVDRX2", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HGVMVDRX2_CSS1_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HGVMVDRX2_CSS1_merge_all

saveRDS(HGVMVDRX2_CSS1_glm_all, file = "StandardGLM/HGVMVDRX2_CSS1_glm_all.rds")
```

# Averaging across replicates to fill in missing data (HJ5HKDRX3 and HKTMWDRX2)

### H2O2 
```{r, eval = FALSE, fig.width = 15, fig.height = 4}
HJ5HKDRX3_CSS8_glm_prep %>% group_by(CHROM, POS, Bulk, Parent, Allele) %>% na.omit() %>% summarize(AvgCount = mean(SmoothCount, na.rm = TRUE)) -> HJ5HKDRX3_CSS8_glm_prep_noreps

#Test once
testdata <- HJ5HKDRX3_CSS8_glm_prep_noreps %>% na.omit() %>% filter(CHROM == "I", POS == 30782) %>% arrange(Bulk, Parent)
table(testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Bulk, testdata$Parent))$Freq)
formula = "Allele ~ Bulk * Parent"
glm(formula = formula, family = "binomial", data = testdata, weights = AvgCount)

# Run full dataset
HJ5HKDRX3_CSS8_glm_prep_noreps %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             W = AvgCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 8, outputlength = 4),
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HJ5HKDRX3_CSS8_glm_all_noreps

HJ5HKDRX3_CSS8_glm_all_noreps %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HJ5HKDRX3", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HJ5HKDRX3_CSS8_merge_all

#saveRDS(HJ5HKDRX3_CSS8_glm_all_noreps, file = "StandardGLM/Z_HJ5HKDRX3_CSS8_glm_all_noreps.rds")

HJ5HKDRX3_CSS8_glm_all_noreps saveRDS("StandardGLM/Z_HJ5HKDRX3_CSS8_glm_all_noreps.rds")

ChromosomeScale %>% select(CHROM, POS) %>% mutate(Summary = NA, Factor = NA) -> ChromScale3

HJ5HKDRX3_CSS8_glm_all_noreps %>%  rbind(ChromScale3) %>% ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")
```

```{r, eval = FALSE, fig.width = 15, fig.height = 4}
HJ5HKDRX3_CSS1_glm_prep %>% group_by(CHROM, POS, Bulk, Parent, Allele) %>% na.omit() %>% summarize(AvgCount = mean(SmoothCount, na.rm = TRUE)) -> HJ5HKDRX3_CSS1_glm_prep_noreps

#Test once
testdata <- HJ5HKDRX3_CSS1_glm_prep_noreps %>% na.omit() %>% filter(CHROM == "I", POS == 30479) %>% arrange(Bulk, Parent)
table(testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Bulk, testdata$Parent))$Freq)
formula = "Allele ~ Bulk * Parent"
glm(formula = formula, family = "binomial", data = testdata, weights = AvgCount)

# Run full dataset
HJ5HKDRX3_CSS1_glm_prep_noreps %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640021) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             W = AvgCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 1, outputlength = 4),
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HJ5HKDRX3_CSS1_glm_all_noreps

HJ5HKDRX3_CSS1_glm_all_noreps %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HJ5HKDRX3", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HJ5HKDRX3_CSS1_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HJ5HKDRX3_CSS1_merge_all

saveRDS(HJ5HKDRX3_CSS1_glm_all_noreps, file = "StandardGLM/Z_HJ5HKDRX3_CSS1_glm_all_noreps.rds")

ChromosomeScale %>% select(CHROM, POS) %>% mutate(Summary = NA, Factor = NA) -> ChromScale3

HJ5HKDRX3_CSS1_glm_all_noreps %>%  rbind(ChromScale3) %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")
```

### Zeocin and Cycloheximide CSSI

Do the same no-rep thing for Zeocin and Cycloheximide to see if that helps

```{r, eval = FALSE, fig.width = 15, fig.height = 4}

# Zeocin
HKTMWDRX2_smoothed %>% 
  #Remove or add "Rep" and "Dose" etc
  separate(Dataset, into = c("Parent", "Bulk", "Rep"), sep="(?<=[A-Z])") %>% 
  mutate(Parent = gsub("0", "O", Parent)) %>%
  
  #Filter for Chr 1 and Bulk D or H
  filter(Bulk != "C") %>%

  #Make everything factors
  mutate_if(is.character, as.factor) -> HKTMWDRX2_Zeo_CSS1_glm_prep

# Cycloheximide
HKTMWDRX2_smoothed %>% 
  #Remove or add "Rep" and "Dose" etc
  separate(Dataset, into = c("Parent", "Bulk", "Rep"), sep="(?<=[A-Z])") %>% 
  mutate(Parent = gsub("0", "O", Parent)) %>%
  
  #Filter for Chr 1 and Bulk D or H
  filter(Bulk != "Z") %>%

  #Make everything factors
  mutate_if(is.character, as.factor) -> HKTMWDRX2_Cyc_CSS1_glm_prep
```

```{r, eval = FALSE, fig.width = 15, fig.height = 4}
HKTMWDRX2_Cyc_CSS1_glm_prep %>% group_by(CHROM, POS, Bulk, Parent, Allele) %>% na.omit() %>% summarize(AvgCount = mean(SmoothCount, na.rm = TRUE)) -> HKTMWDRX2_Cyc_CSS1_glm_prep_noreps

HKTMWDRX2_Cyc_CSS1_glm_prep_noreps %>% filter(Parent == "W", CHROM == "I")

#Test once
testdata <- HKTMWDRX2_Cyc_CSS1_glm_prep_noreps %>% na.omit() %>% filter(CHROM == "I", POS == 34991) %>% arrange(Bulk, Parent)
table(testdata$Bulk, testdata$Parent)
sum(as.data.frame(table(testdata$Bulk, testdata$Parent))$Freq)
formula = "Allele ~ Bulk * Parent"
glm(formula = formula, family = "binomial", data = testdata, weights = AvgCount)

HKTMWDRX2_Cyc_CSS1_glm_prep_noreps %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640021) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             W = AvgCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 1, outputlength = 4),
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HKTMWDRX2_Cyc_CSS1_glm_all_noreps

saveRDS(HKTMWDRX2_Cyc_CSS1_glm_all_noreps, file = "StandardGLM/HKTMWDRX2_Cyc_CSS1_glm_all_noreps.rds")

#PLOT TO CHECK
# HKTMWDRX2_Cyc_CSS1_glm_all_noreps %>% rbind(ChromScale3) %>% filter(CHROM != "I") %>% 
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() + 
#   geom_vline(data = myloci, aes(xintercept = POS, color = cat), linewidth = 1, linetype = "dashed") +
#   facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("Cycloheximide CSS I (Rep Avg)")

```

```{r, eval = FALSE, fig.width = 15, fig.height = 4}
HKTMWDRX2_Zeo_CSS1_glm_prep %>% group_by(CHROM, POS, Bulk, Parent, Allele) %>% na.omit() %>% summarize(AvgCount = mean(SmoothCount, na.rm = TRUE)) -> HKTMWDRX2_Zeo_CSS1_glm_prep_noreps

#Test once
# testdata <- HKTMWDRX2_Zeo_CSS1_glm_prep_noreps %>% na.omit() %>% filter(CHROM == "I", POS == 34991) %>% arrange(Bulk, Parent)
# table(testdata$Bulk, testdata$Parent)
# sum(as.data.frame(table(testdata$Bulk, testdata$Parent))$Freq)
# formula = "Allele ~ Bulk * Parent"
# glm(formula = formula, family = "binomial", data = testdata, weights = AvgCount)

HKTMWDRX2_Zeo_CSS1_glm_prep_noreps %>% na.omit() %>% 
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             W = AvgCount,
                             formula = "Allele ~ Bulk * Parent",
                            numgroups = 1, outputlength = 4),
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> HKTMWDRX2_Zeo_CSS1_glm_all_noreps

saveRDS(HKTMWDRX2_Zeo_CSS1_glm_all_noreps, file = "StandardGLM/HKTMWDRX2_Zeo_CSS1_glm_all_noreps.rds")

#PLOT TO CHECK
# HKTMWDRX2_Zeo_CSS1_glm_all_noreps %>% rbind(ChromScale3) %>% filter(CHROM != "I") %>% 
#   ggplot(aes(x = POS, y = abs(Summary), color = Factor)) + geom_line() +
#   geom_vline(data = myloci, aes(xintercept = POS, color = cat), linewidth = 1, linetype = "dashed") +
# facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("Zeocin CSS I (Rep Avg)")

```

