---
title: "2023 Pre-processing All Seq Runs"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: hide
  html_notebook:
    code_folding: hide
---

```{r, warning=FALSE, message=FALSE, comment=FALSE, fig.width=18, fig.height=5}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)

require(cybrBSA)


#TESTING
#glm_formula <- "Allele ~ Bulk * Parent + Rep"

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))

```

# Functions

```{r}
# glm_cb <- function(..., W, formula) {
#   data <- list(...)
#   
#   if (is.null(W) || is.null(formula)) {
#     stop("Weights (W) and formula must be provided")
#   }
#   
#   glm_formula <- as.formula(formula)
#   
#   if (!all(names(data) %in% all.vars(glm_formula))) {
#     stop("One or more variables in the formula are not provided as arguments")
#   }
#   
#   #CHANGE THIS TO ADJUST TYPE OF REGRESSION
#   glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
#   
#   #Output: Effect, Standard Error, Z-score 
#   #(ignores p-value since it can be calculated from Z score; multiply by 1 if you would like to retain it)
#   return(summary(glm_fit)$coefficients[1:((length(summary(glm_fit)$coefficients)*0.75))])
#   
# }

glm_cb2_short <- function(..., W, formula, numgroups = FALSE, outputlength = 4, return = c("Z")) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  #MAYBEWORKS
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  #Return specific ones like z-score only
  if(return %in% "Z"){
    output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients)*0.5)+1):((length(summary(glm_fit)$coefficients)*0.75))]
  }
  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
}

############################ PERMUTATIONS ######################################
# Replicates, all scrambled
cybrPermute_Rep <- function(dataset){
  
  start.time <- Sys.time()
  
  dataset %>% 
  distinct() %>% ungroup() %>%
  group_by(CHROM, POS, Allele, 
           Bulk, 
           Rep, #might not have these
           Parent) %>% 
  summarize(culprits = length((SmoothCount))) %>% 
  merge(dataset) %>% 
  filter(culprits == 1) %>% 
  ungroup() %>%
  distinct() %>% #THIS IS IMPORTANT
  pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest

  #these are now all of the ones that can be used to permute
  newnewtest %>% filter(Bulk != "F", CHROM != "I", CHROM != "III", CHROM != "VIII") %>% select(CHROM, Oak, Wine) %>% 
    mutate(rownum = sample(1:nrow(.))) %>% arrange(rownum) %>% na.omit() -> NullToSample
  
  #Trying this again
  
  dataset[1:nrow(NullToSample),] %>% ungroup() %>% select(CHROM, POS, Bulk, Parent, Rep) %>% cbind(NullToSample[,c("Oak", "Wine")]) %>%
    pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") %>%
    #Original Script
    group_by(CHROM, POS) %>%
    mutate_if(is.character, as.factor) %>%
      summarize(Summary = glm_cb2_short(Allele = Allele,
                               Bulk = Bulk,
                               Parent = Parent,
                               Rep = Rep,
                               W = SmoothCount,
                               formula = "Allele ~ Bulk * Parent + Rep",
                               outputlength = 5),
              Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> glmresult

  end.time = Sys.time()
  print(end.time - start.time)
  
  return(glmresult)
  
}

# No replicates, fix CHROM, Parent
# No replicates, fix Parent

# Replicates, fix CHROM, Parent
cybrPermute_byCHRParent_Rep <- function(dataset){
   start.time <- Sys.time()
  
   print("Make sure that dilute bulk is labeled D")
   
    dataset %>% 
      distinct() %>% ungroup() %>%
      group_by(CHROM, POS, Allele, 
               Bulk, 
               Rep, #might not have these
               Parent) %>% 
      summarize(culprits = length((SmoothCount))) %>% 
      merge(dataset) %>% 
      filter(culprits == 1) %>% 
      ungroup() %>%
      distinct() %>% #THIS IS IMPORTANT
      pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    
    #these are now all of the ones that can be used to permute
    newnewtest %>% filter(Bulk == "D") %>% select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(CHROM, Parent) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine, POS = POS,
                POS2 = sample(POS)) %>%
      group_by(CHROM, Parent, POS2) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine,
                POS = POS2, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA
    
    newnewtest %>% filter(Bulk == "D") %>% select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(CHROM, Parent) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine, POS = POS,
                POS2 = sample(POS)) %>%
      group_by(CHROM, Parent, POS2) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine,
                POS = POS2, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB
    
    rbind(shuffled_DiluteA, shuffled_DiluteB) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm
    
    #Trying this again
    
    shuffletoglm %>% na.omit() %>%
      #Original Script
      group_by(CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_short(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = "Allele ~ Bulk * Parent + Rep",
                                numgroups = 16, outputlength = 5),
                Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> glmresult
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}

# Replicates, fix Parent
cybrPermute_byParent_Rep <- function(dataset){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled D")

  dataset %>% 
  distinct() %>% ungroup() %>%
  group_by(CHROM, POS, Allele, 
           Bulk, 
           Rep, #might not have these
           Parent) %>% 
  summarize(culprits = length((SmoothCount))) %>% 
  merge(dataset) %>% 
  filter(culprits == 1) %>% 
  ungroup() %>%
  distinct() %>% #THIS IS IMPORTANT
  pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
  
  #these are now all of the ones that can be used to permute
    newnewtest %>% filter(Bulk == "D",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, 
                Loc = sample(Loc)) %>%
      group_by(Parent, Loc) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine,
                Loc = Loc, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "D",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, 
                Loc = sample(Loc)) %>%
      group_by(Parent, Loc) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine,
                Loc = Loc, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #Trying this again
    
    shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_short(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = "Allele ~ Bulk * Parent + Rep",
                                numgroups = 16, outputlength = 5),
                Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> glmresult2
  
    end.time = Sys.time()
    print(end.time - start.time)
    
    return(glmresult2)
}
  
```

See what this is actually doing

```{r}
#rawData <- RawFiles$VCF_Table[1]

cybrInputGATKTable2 <- function(rawData, yeast = TRUE){

  require(dplyr)
  require(doParallel)
  require(foreach)

  HNGLCDRXY <- read.table(rawData, header = TRUE)

  #Identify the unique values besides AD/DP/GQ/PL
  gsub(".AD", "",
       gsub(".GQ", "",
            gsub(".DP","",
                 gsub(".PL","",
                      colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
  #i <- Samples[1]

  resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
    mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
    AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
    GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
    DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
    PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
    cbind(mydf, AD , GQ , DP, PL) -> mydftotal
    colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")

    mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>%
      separate(PL, c('PL.REF','PL.ALT'), extra='drop') %>%
      #Added 10/18/23:
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf

    mycdf %>% filter(grepl(",", ALT)) %>% 
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
    
    doublecdf %>% filter(grepl(",", ALT)) %>%
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
    
    rbind(mycdf, doublecdf, triplecdf) -> newcdf
    
    newcdf
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
  }else{
    results <- resultscdf
  }
  return(results)

}

# ################################################################################
# 
# HNGLCDRXY <- read.table(rawData, header = TRUE)
# 
#   #Identify the unique values besides AD/DP/GQ/PL
#   gsub(".AD", "",
#        gsub(".GQ", "",
#             gsub(".DP","",
#                  gsub(".PL","",
#                       colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
#   i <- Samples[1]
# 
#   resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
#     mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
#     AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
#     GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
#     DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
#     PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
#     cbind(mydf, AD , GQ , DP, PL) -> mydftotal
#     colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")
# 
#     mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='merge') %>%
#       separate(PL, c('PL.REF','PL.ALT'), extra='merge') %>%
#       #Added 10/18/23:
#       select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf
# 
#     mycdf
#     
#     mycdf %>% filter(grepl(",", ALT)) %>% 
#       separate(ALT, c("A1", "A2"), extra = 'merge') %>%
#       separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
#       separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
#       
#       pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
#       pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
#       pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
#       mutate(NumAlt = gsub("A", "", NumAlt),
#              NumADAlt = gsub("AD", "", NumADAlt),
#              NumPL = gsub("P", "", NumPL)) %>%
#       filter(NumAlt == NumPL,
#              NumPL == NumADAlt) %>%
#       select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
#     
#     doublecdf %>% filter(grepl(",", ALT)) %>%
#       separate(ALT, c("A1", "A2"), extra = 'merge') %>%
#       separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
#       separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
#       
#       pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
#       pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
#       pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
#       mutate(NumAlt = gsub("A", "", NumAlt),
#              NumADAlt = gsub("AD", "", NumADAlt),
#              NumPL = gsub("P", "", NumPL)) %>%
#       filter(NumAlt == NumPL,
#              NumPL == NumADAlt) %>%
#       select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
#     
#     triplecdf
#   }

```


Note: if looking for smoothed vs unsmoothed data, the `Figures_2023_BiologyofGenomes.rmd` file has these at the beginning.

# Loading In Data

GQ Cutoff is 98 for all samples, so find out how many of these are under 98 and in which samples that tends to occur

```{r}
setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)
parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) -> pSNPs

pSNPs %>% filter(Unique == 1) 

pSNPs %>% ungroup() %>% count(Type)
# pSNPs$CHROM <- factor(pSNPs$CHROM, levels = as.character(as.roman(1:17)))
```

```{r, eval = FALSE, fig.width=12, fig.height=5}
#Checking what these actually look like for CUP1
parentSNPids %>% filter(CHROM == "VIII")

################################################################################
setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")

ParentFiles <- c("Wine_VCF.txt", "Oak_VCF.txt")
temparent <- list()
Parents = gsub("_VCF.txt","", ParentFiles)
  mergeparents <- foreach(i=1:length(ParentFiles), .combine=rbind) %dopar% {
    read.table(ParentFiles[i], header = TRUE) %>% mutate(parent = Parents[i])
  }

  
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                           "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                           CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                     "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                     "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                     "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    rbind(mergeparents) %>% arrange(CHROM, POS) %>%
      select(CHROM, POS, REF, ALT, parent) %>%
      merge(ChromKey) %>% select(-CHROM) %>%
      mutate(CHROM = chromosomes) %>% select(-chromosomes) -> ParentalVCF

  #ParentalVCF %>% pivot_wider(names_from = parent, values_from = ALT) -> SNPids
    
    #212535..212720
    
    ParentalVCF %>% filter(CHROM == "VIII") %>% filter(POS > (212535 - 10000), POS < (214720 + 10000)) %>%
      ggplot(aes(x = POS, y = ALT, color = parent)) + geom_jitter(alpha = 0.5, height = 0.1) +
      geom_vline(xintercept = c(212535,212720, 214533, 214720 )) +
      facet_grid(rows = "REF") +
      scale_color_manual(values = c("lightblue", "firebrick"))
    
    ParentalVCF %>% filter(CHROM == "VIII") %>% filter(POS > (212535 - 1000), POS < (214720 + 1000)) %>%
      ggplot(aes(x = POS, y = parent, 
                 color = ALT, shape = parent #, size = ALT %in% c("GAA", "TA", "CT")
                 )) + 
      geom_point(alpha = 0.5, size = 3) +
      geom_vline(xintercept = c(212535,212720, #CUP1-1
                                214533,214720 )) + #CUP1-2
      facet_grid(rows = "REF") 
```

## Table Processing
```{r}
MQCRuns <- read.csv("C:\\Users\\cassa\\OneDrive\\Documents\\SiegalLab\\Sequencing_Tuboweb\\AllMultiQCRuns.csv")

MQCRuns %>% select(Pool, ShortName, VCF_Table) %>% distinct() -> RawFiles

#alldata <- readRDS("allseqruns_feb24.rds")

#for(i in 1:length(RawFiles$VCF_Table)){
for(i in 1:8){
  cybrInputGATKTable2(RawFiles$VCF_Table[i]) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) %>% mutate(Pool = RawFiles$Pool[i])-> rawdata

  if(i > 1){
    alldata <- rbind(rawdata, alldata)
  }else{
    alldata <- rawdata
  }

}

alldata %>% distinct() %>%
  mutate(Dataset = gsub(".*_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub("\\.fastq$", "", Dataset)) -> alldata

saveRDS(alldata, file = "allseqruns_may24.rds")

#alldata <- readRDS("allseqruns_nov23.rds")

#unique(alldata$Pool)

unique(MQCRuns$Pool)
unique(MQCRuns$ShortName)

```

### Checking how coverage is affected by fragment length

```{r, eval = FALSE}

KeyForWhatWorked <- data.frame(ShortName = c("C1A", "C1Ar", "C1B", "C8", "F1", "HZ18", "Z1"),
                               Worked = c(T, NA, F, T, T, F, F))

MQCRuns %>% merge(KeyForWhatWorked) %>%
  filter(Pool != "xHNGLVDRXY") %>%
  select(Pool, Worked, ShortName, MQCDate) %>% distinct()

MQCRuns %>% merge(KeyForWhatWorked) %>%
  filter(Pool != "xHNGLVDRXY") %>%
  select(FragmentLength,MillionReads, Pool, ShortName, Worked) %>%
  #na.omit() %>%
  ggplot(aes(x = FragmentLength, y = MillionReads, color = paste(Pool, ShortName), fill = paste(Pool, ShortName), shape = Worked)) + 
  geom_point(size = 3, alpha = 0.4) + geom_smooth(method = "lm", alpha = 0.2)

MQCRuns %>% group_by(Pool) %>% summarize(PoolReads = sum(MillionReads),
                                              MillionReads = MillionReads,
                                              FragmentLength = FragmentLength,
                                         ShortName = ShortName) %>% 
  summarize(PercentPool = MillionReads/PoolReads,
            RelativePool = (MillionReads/PoolReads)/mean(PoolReads),
                                              MillionReads = MillionReads,
                                              FragmentLength = FragmentLength,
                                         ShortName = ShortName) %>% na.omit() %>% ungroup() -> check

check %>%
  ggplot(aes(x = (RelativePool), y = FragmentLength, color = paste(Pool, ShortName))) + geom_point() + geom_smooth(method = "lm")

```


## For any new datasets:

```{r, eval = FALSE}

#LOAD IN TABLE
MQCRuns <- read.csv("C:\\Users\\cassa\\OneDrive\\Documents\\SiegalLab\\Sequencing_Tuboweb\\AllMultiQCRuns.csv")

MQCRuns %>% select(Pool, ShortName, VCF_Table) %>% distinct() -> RawFiles

var <- 8 #CHANGE THIS FOR THE INDEX OF RawFiles YOU'RE ANALYZING
RawFiles$Pool[var] #CHECK THAT IT'S RIGHT

for(i in var){
  cybrInputGATKTable2(RawFiles$VCF_Table[i]) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) %>% mutate(Pool = RawFiles$Pool[i])-> rawdata

}

rawdata %>% distinct() %>%
  mutate(Dataset = gsub(".*_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub("\\.fastq$", "", Dataset)) -> rawdata

#SMOOTHING ETC
rawdata %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Pool, Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% group_by(Pool, Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 300, FUN = median, align = "center"))) -> rawdata_smoothed #%>% na.omit()


################################################################################
# COMBINE WITH ALLDATA FILES
################################################################################
# #LOAD IN ORIGINAL
alldata <- readRDS("allseqruns_may24.rds")
alldata <- rbind(alldata, rawdata)
# 
alldata_called <- readRDS(file = "alldata_called.rds")
alldata_called <- rbind(alldata_called, rawdata_called)
# 
# alldata_smoothed <- readRDS(file = "alldata_smoothed.rds")
# alldata_smoothed <- rbind(alldata_smoothed, rawdata_smoothed)
# 
# saveRDS(alldata, file = "allseqruns_feb24.rds")
saveRDS(alldata_called, file = "alldata_called.rds")
# saveRDS(alldata_smoothed, file = "alldata_smoothed.rds")

################################################################################
# for quick use, add in as run here:
#rawdata_smoothed -> Hsomething_smoothed
#Hsomething_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> Hsomething_counts

#alldata_smoothed %>% filter(Pool == "HTG3TDMXY")

```


This seems out of order when you add more datasets:

```{r, eval = FALSE}
alldata %>% #head(3000) %>%
  left_join(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Pool, Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> alldata_called

#6k to 4k when NAs are removed
#alldata_called %>% na.omit()

alldata_called %>% group_by(Pool, Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 200, FUN = median, align = "center"))) -> alldata_smoothed #%>% na.omit()

ChromeScale_exact <- data.frame(CHROM = as.character(as.roman(as.numeric(c(1:16)))),
                                End = c(230218,  813184,   316620, 1531933,
                                        576874,  270161,  1090940, 562643,
                                        439888,745751,666816,1078177,
                                        924431,784333,1091291,948066))
alldata_smoothed %>%
  filter(POS > 30000) %>% 
  merge(ChromeScale_exact) %>%
  # merge(ChromosomeScale[,c("CHROM", "End")]) %>% 
  # filter(End > 0) %>% 
  group_by(CHROM) %>% 
  filter(POS < (End - 30000)) -> alldata_smoothed_e

saveRDS(alldata_smoothed_e, file = "alldata_smoothed_e_may24.rds")

```

```{r}
# parentSNPids %>% count(Type)

alldata %>% head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Pool, Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> alldata_called

alldata_called %>% group_by(Pool, Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 200, FUN = median, align = "center"))) -> alldata_smoothed #%>% na.omit()

# saveRDS(alldata_called, file = "Nov23_alldata_called.rds")
# saveRDS(alldata_smoothed, file = "Nov23_alldata_smoothed.rds")

# alldata_called <- readRDS(file = "Nov23_alldata_called.rds")
# alldata_smoothed <- readRDS(file = "Nov23_alldata_smoothed.rds")

#Making HKTFTDRX2

alldata %>% filter(Pool == "HKTFTDRX2") %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Pool, Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> HKTFTDRX2_called

```

## Start Here for loading in data for subsequent analysis
Use _e dataset to ensure that the ends of the chromosomes are not there

```{r, fig.width = 15, fig.height = 4}
alldata_smoothed %>%
  filter(POS > 30000) %>% 
  merge(ChromosomeScale[,c("CHROM", "End")]) %>% 
  filter(End > 0) %>% 
  group_by(CHROM) %>% 
  filter(POS < End) -> alldata_smoothed_e

saveRDS(alldata_smoothed_e, file = "alldata_smoothed_e.rds")

#alldata_smoothed_e <- readRDS(file = "Jan24_alldata_smoothed_e.rds")

```



## For new additions:

```{r}

alldata_smoothed_e <- readRDS(file = "alldata_smoothed_e.rds") #%>% filter(Pool != "HTG3TDMXY")

rawdata_smoothed %>%
  filter(POS > 30000) %>% 
  merge(ChromosomeScale[,c("CHROM", "End")]) %>% 
  filter(End > 0) %>% 
  group_by(CHROM) %>% 
  filter(POS < End) %>%
  rbind(alldata_smoothed_e) -> alldata_smoothed_e

saveRDS(alldata_smoothed_e, file = "alldata_smoothed_e.rds")

```
**Feb 2024 update: changing all of the labels on these so that the glm runs the correct thing! **

```{r}

alldata_smoothed_e <- readRDS(file = "alldata_smoothed_e.rds") #%>% filter(Pool != "HTG3TDMXY")

alldata_smoothed_e %>% ungroup() %>% select(Dataset) %>% summarize(unique(Dataset))

alldata_smoothed_e %>% mutate(DatasetRenamed = gsub("_D", "_zD", Dataset)) -> alldata_smoothed_e_r
#saveRDS(alldata_smoothed_e, file = "alldata_smoothed_e.rds")

```

## Separate each out into independent datasets to run GLM

```{r}
# HNGLVDRXY = CuSO4a (CSS1)
alldata_smoothed_e %>% filter(Pool == "HNGLVDRXY") -> HNGLVDRXY_smoothed
#HNGLVDRXY_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HNGLVDRXY_counts

# HKTFTDRX2 = CuSO4b (CSS1) - NEW VERSION
alldata_smoothed_e %>% filter(Pool == "HKTFTDRX2") -> HKTFTDRX2_smoothed
#HKTFTDRX2_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HKTFTDRX2_counts

# HVYTYDRX2 = CSS8 Fluc and CuSO4, also CSS1
alldata_smoothed_e %>% filter(Pool == "HVYTYDRX2") -> HVYTYDRX2_smoothed
#HVYTYDRX2_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HVYTYDRX2_counts

# HGVMDRX2 = Fluconazole CSS I 
alldata_smoothed_e %>% filter(Pool == "HGVMVDRX2") -> HGVMVDRX2_smoothed
#HGVMVDRX2_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HGVMVDRX2_counts

# HJ5HKDRX3 = H2O2 data X
alldata_smoothed_e %>% filter(Pool == "HJ5HKDRX3") -> HJ5HKDRX3_smoothed
#HJ5HKDRX3_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HJ5HKDRX3_counts

# HKTMWDRX2 = Zeocin/Cycloheximide X
alldata_smoothed_e %>% filter(Pool == "HKTMWDRX2") -> HKTMWDRX2_smoothed
#HKTMWDRX2_smoothed %>% group_by(CHROM, POS, Dataset) %>% count(Allele) -> HKTMWDRX2_counts

################################################################################
# ADD NEW ONES DOWN HERE
################################################################################

# CuSO4 done again with multiple doses
alldata_smoothed_e %>% filter(Pool == "HTG3TDMXY") -> HTG3TDMXY_smoothed

```


Save all (then open the GLM script)

```{r}
saveRDS(HNGLVDRXY_smoothed, file = "StandardGLM/HNGLVDRXY_smoothed.rds")
saveRDS(HKTFTDRX2_smoothed, file = "StandardGLM/HKTFTDRX2_smoothed.rds") #NEW VERSION
saveRDS(HVYTYDRX2_smoothed, file = "StandardGLM/HVYTYDRX2_smoothed.rds")
saveRDS(HGVMVDRX2_smoothed, file = "StandardGLM/HGVMVDRX2_smoothed.rds")
saveRDS(HJ5HKDRX3_smoothed, file = "StandardGLM/HJ5HKDRX3_smoothed.rds")
saveRDS(HKTMWDRX2_smoothed, file = "StandardGLM/HKTMWDRX2_smoothed.rds")
saveRDS(xHNGLVDRXY_smoothed, file = "StandardGLM/xHNGLVDRXY_smoothed.rds")

saveRDS(HTG3TDMXY_smoothed, file = "StandardGLM/HTG3TDMXY_smoothed.rds")

```

## Change All Windows

Optional, added 2/22/24; change window size and group for ALL. This will take forever.

```{r}
alldata_called <- readRDS(file = "alldata_called.rds") #%>% rbind(HKTFTDRX2_called)

#saveRDS(alldata_called, file = "alldata_called.rds")

#Check that these are right for later
# alldata_called %>% ungroup() %>% select(Pool) %>% distinct() %>% filter(Pool %in% c("HKTFTDRX2", "HNGLVDRXY", "HTG3TDMXY", "HVYTYDRX2")) %>% summarize(uniquepool = unique(Pool))
# 
# alldata  %>% ungroup() %>% select(Pool) %>% distinct() %>% filter(Pool %in% c("HKTFTDRX2", "HNGLVDRXY", "HTG3TDMXY", "HVYTYDRX2")) %>% summarize(uniquepool = unique(Pool))

alldata_called %>% group_by(Pool, Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 300, FUN = median, align = "center"))) -> alldata_smoothed_300 #%>% na.omit()

alldata_smoothed_300 %>% filter(POS > 30000) %>% 
  merge(ChromosomeScale[,c("CHROM", "End")]) %>% 
  filter(End > 0) %>% 
  group_by(CHROM) %>% 
  filter(POS < End) %>%
  rbind(alldata_smoothed_300) -> alldata_smoothed_300_e

saveRDS(alldata_smoothed_300_e, file = "alldata_smoothed_300_e.rds")

alldata_smoothed_300_e %>% filter(Pool %in% c("HKTFTDRX2", "HNGLVDRXY", "HTG3TDMXY", "HVYTYDRX2")) -> allCuSO4_300_e

saveRDS(allCuSO4_300_e, file = "allCuSO4_300_e.rds")

```

```{r}
alldata_called %>% group_by(Pool, Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 400, FUN = median, align = "center"))) -> alldata_smoothed_400 #%>% na.omit()

alldata_smoothed_400 %>% filter(POS > 30000) %>% 
  merge(ChromosomeScale[,c("CHROM", "End")]) %>% 
  filter(End > 0) %>% 
  group_by(CHROM) %>% 
  filter(POS < End) %>%
  rbind(alldata_smoothed_400) -> alldata_smoothed_400_e

saveRDS(alldata_smoothed_400_e, file = "alldata_smoothed_400_e.rds")

alldata_smoothed_400_e %>% filter(Pool %in% c("HKTFTDRX2", "HNGLVDRXY", "HTG3TDMXY", "HVYTYDRX2")) -> allCuSO4_400_e

saveRDS(allCuSO4_400_e, file = "allCuSO4_400_e.rds")

```

* Move on to AllSequencingRuns_GLM after this point *

***

# Looking at mitochondrial genomes for each set

```{r}

alldata_called <- readRDS("alldata_called.rds")

alldata_called %>% filter(CHROM == "M") -> Mito_called

Mito_called %>% group_by(Pool, Dataset, POS) %>% distinct() %>% pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(logOdds = log(Wine/Oak))-> MitoAll

#MitoAll %>% ggplot(aes(x = logOdds, y = Pool, color = Dataset)) + geom_jitter()
MitoAll %>%   mutate(CSS8 = grepl("8", Dataset)) %>%
  mutate(CSS8 = gsub("FALSE", "CSS I", CSS8),
         CSS8 = gsub("TRUE", "CSS VIII", CSS8)) %>%
  mutate(WineParent = (grepl("W|UnselectedC|SelectedC", Dataset))) %>%
    mutate(WineParent = gsub("FALSE", "Oak", WineParent),
         WineParent = gsub("TRUE", "Wine", WineParent)) %>%
  mutate(Parent = paste(WineParent,CSS8)) %>%
  ggplot(aes(y = logOdds, x = Parent, color = Parent)) + 
  geom_jitter(alpha = 0.3) + facet_grid(~Pool) + 
  theme(legend.position = "none") +  
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

MitoAll %>% filter(Pool == "HJ5HKDRX3") %>% 
  mutate(Dataset = gsub("08", "O8", Dataset)) %>%
  separate(Dataset, into = c("Bulk", "Parent", "Replicate"), sep = "_") %>%
  ggplot(aes(x = logOdds, y = paste(Parent, Bulk, Replicate), color = Parent)) + geom_jitter(alpha = 0.3)

MitoAll %>% filter(Pool == "HVYTYDRX2") %>% 
  separate(Dataset, into = c("Parent", "Bulk"), sep = "_") %>%
  ggplot(aes(x = logOdds, y = paste(Parent, Bulk), color = Parent)) + geom_jitter(alpha = 0.3)
```

Plot by position

```{r}
MitoAll %>% filter(Pool == "HJ5HKDRX3") %>% 
    ggplot(aes(y = logOdds, x = POS, color = Dataset)) + geom_jitter(alpha = 0.3)

MitoAll %>% filter(Pool == "HVYTYDRX2") %>% 
  ggplot(aes(y = logOdds, x = POS, color = Dataset)) + geom_jitter(alpha = 0.3)
```


# Copy Number of CUP1

This actually can be used for identifying genes in peaks since it's a super easy function to write, and the labels in the ggplot might be useful...

```{r, fig.height = 4, fig.width = 12}
alldata <- readRDS("allseqruns_nov23.rds")

alldata %>% filter(grepl("8", Dataset),
                   CHROM == "VIII") -> Chr8Fixed

alldata %>% filter(grepl("8", Dataset) == FALSE,
                   CHROM == "VIII") -> Chr8_NotFixed

rm(alldata)
unique(Chr8Fixed$Dataset)

#Chr8Fixed %>% ggplot(aes(x = POS, y = Coverage, color = Dataset)) + geom_point(alpha = 0.4) + ggtitle("Chr 8 Fixed")

Chr8Fixed %>% group_by(Dataset) %>% 
  summarize(LibCoverage = mean(Coverage, na.rm = TRUE)) %>% 
  merge(Chr8Fixed) %>% mutate(Copies = log(Coverage/LibCoverage),
                              ds = "Chr8Fixed") -> Chr8Fixed_Copy

Chr8_NotFixed %>% group_by(Dataset) %>% 
  summarize(LibCoverage = mean(Coverage, na.rm = TRUE)) %>% 
  merge(Chr8_NotFixed) %>% mutate(Copies = log(Coverage/LibCoverage),
                                  ds = "Chr8Unfixed") -> Chr8_NotFixed_Copy

Chr8Fixed_Copy %>% ggplot(aes(x = POS, y = Copies, color = Dataset)) + geom_point(alpha = 0.4) + ggtitle("Chr 8 Copy Numbers")

rbind(Chr8Fixed_Copy, Chr8_NotFixed_Copy) %>%
  #filter(POS > 200000, POS < 225000) %>% #CUP1 locations
  ggplot(aes(x = POS, y = Copies, color = grepl("W", Dataset))) + 
  geom_vline(xintercept = c(212535,212720, 214533,214718)) +
  geom_point(alpha = 0.4) + facet_grid(rows = "ds")

#Maybe do a t test for each one...?
rbind(Chr8Fixed_Copy, Chr8_NotFixed_Copy) %>%
  mutate(WineParent = grepl("W", Dataset)) %>%
  na.omit()  %>% 
    filter(Copies != -Inf) %>%
  group_by(POS, ds, WineParent) %>%
  summarize(lc = length(Copies)) %>%
  filter(lc > 1) %>% 
  pivot_wider(names_from = c(ds, WineParent), values_from = lc) %>%
  na.omit() %>%
  merge(rbind(Chr8Fixed_Copy, Chr8_NotFixed_Copy)) %>%
  mutate(WineParent = grepl("W", Dataset)) %>%
  filter(Copies != -Inf) %>%
  group_by(POS, ds) %>% na.omit() %>%
  summarize(pval = t.test(Copies ~ WineParent)$p.value) -> ttests

ttests %>% filter(POS > 200000, POS < 225000) %>% ggplot(aes(x = POS, y = -log(pval), color = ds)) + geom_point(alpha = 0.5, size = 2) + ggtitle("CUP1 locus copy difference")
  
ttests %>% ggplot(aes(x = POS, y = -log(pval), color = ds)) + geom_point(alpha = 0.5, size = 0.8) + geom_vline(xintercept = c(200000, 225000)) + ggtitle("Chr 8 locus copy difference")

quantile(ttests$pval[ttests$ds == "Chr8Unfixed"], 0.05)

ttests %>% ggplot(aes(x = POS, y = -log(pval), color = ds)) + geom_line(alpha = 0.5, size = 0.8) + 
  geom_vline(xintercept = c(200000, 225000)) + 
  geom_hline(yintercept = -log(0.009027215 )) +
  ggtitle("Chr 8 locus copy difference") + facet_grid(rows = "ds")

ttests %>% filter(ds != "Chr8Unfixed") %>%
  ggplot(aes(x = POS, y = -log(pval))) + geom_line(alpha = 0.4) + geom_point(aes(color = pval < 0.009027215), alpha = 0.5, size = 0.8) + 
  geom_vline(xintercept = c(200000, 225000)) + 
  geom_hline(yintercept = -log(0.009027215 )) +
  ggtitle("Chr 8 Fixed copy difference")

ttests %>% group_by(ds) %>% arrange(POS) %>%
  reframe(POS = POS, 
          SmoothPval_3 = frollapply(pval, n = 3, FUN = mean, align = "center"),
          SmoothPval_5 = frollapply(pval, n = 5, FUN = mean, align = "center"),
          SmoothPval_7 = frollapply(pval, n = 7, FUN = mean, align = "center"),
          SmoothPval_9 = frollapply(pval, n = 9, FUN = mean, align = "center")) %>%
  pivot_longer(c(SmoothPval_3, SmoothPval_5,SmoothPval_7,SmoothPval_9))-> ttests_smoothed

ttests_smoothed %>% filter(ds != "Chr8Unfixed") %>%
  ggplot(aes(x = POS, y = -log(value))) + geom_line(alpha = 0.4) + geom_point(aes(color = value < 0.009027215), alpha = 0.5, size = 0.8) + 
  geom_vline(xintercept = c(200000, 225000)) + #CUP1
    geom_vline(xintercept = c(120091-10000,121683+10000), color = "purple") +  #ERG11
  geom_hline(yintercept = -log(0.009027215 )) +
  ggtitle("Chr 8 Fixed copy difference") +
  facet_grid(rows = "name")

ttests_smoothed %>% filter(ds == "Chr8Unfixed") %>%
  ggplot(aes(x = POS, y = -log(value))) + geom_line(alpha = 0.4) + geom_point(aes(color = value < 0.009027215), alpha = 0.5, size = 0.8) + 
  geom_vline(xintercept = c(200000, 225000)) + 
      geom_vline(xintercept = c(120091-10000,121683+10000), color = "purple") +  #ERG11

  geom_hline(yintercept = -log(0.009027215 )) +
  ggtitle("Chr 8 Unfixed copy difference") +
  facet_grid(rows = "name")

ttests_smoothed %>% filter(ds != "Chr8Unfixed",  value < 0.009027215, name == "SmoothPval_7") -> w7_peaks

#load the genes to see what they are in these
genelist <- read.csv("C:\\Users\\cassa\\Downloads//SGD_ORFs.csv") %>% mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) %>%
  select(Gene = Gene.symbol,
         Start = Gene.chromosomeLocation.start,
         End = Gene.chromosomeLocation.end,
         Strand = Gene.chromosomeLocation.strand,
         Type = Gene.featureType,
         CHROM = CHROM)

findSNP <- function(Start, End, Target){
  toreturn <- Target[Target %in% (Start-1000):(End+1000)]
  return(toreturn)
}

genelist %>% filter(Type == "ORF", CHROM == "VIII", Gene != "") %>% group_by(Gene) %>%
  summarize(SNPs = findSNP(Start, End, w7_peaks$POS),
            Start, End, Type) -> w7genes

w7genes %>% select(-SNPs) %>% distinct() -> genestoplot

#Plots
ttests_smoothed %>% filter(ds != "Chr8Unfixed", name == "SmoothPval_7") %>%
  ggplot(aes(x = POS, y = -log(value))) + 
  #All genes
  geom_vline(data = genestoplot, aes(xintercept = Start, color = Gene), alpha = 0.2, size = 2) +
  geom_label(data = genestoplot, aes(x = Start, y = sample(35:65, size = length(genestoplot$Gene), replace = FALSE), 
                                     label = Gene, color = Gene), alpha = 0.3, size = 2) +

  #All actual pvalues
  geom_line(alpha = 0.4) + 
  geom_point(alpha = 0.5, size = 0.8) + 
  geom_vline(xintercept = c(200000, 225000)) + 
  geom_hline(yintercept = -log(0.009027215 )) +
  ggtitle("Chr 8 Fixed copy difference") +

  theme(legend.position = "none", legend.key.size = unit(0.5,"line"))

```

```{r, fig.height = 4, fig.width = 15}
Chr8_NotFixed_Copy %>% 
  filter(POS > 200000, POS < 225000) %>%
  ggplot(aes(x = POS, y = Copies, color = grepl("W", Dataset))) + 
  geom_vline(xintercept = c(212535,212720, 214533,214718)) +
  geom_point(alpha = 0.4) + 
    ylim(-1,1)+
  #facet_grid(rows = "Dataset") + 
  ggtitle("Chr 8 Non Fixed Copy Numbers")

Chr8Fixed_Copy %>% 
  filter(POS > 200000, POS < 225000) %>%
  ggplot(aes(x = POS, y = Copies, color = grepl("W", Dataset))) + 
  geom_vline(xintercept = c(212535,212720, 214533,214718)) +
  geom_point(alpha = 0.4) + 
  ylim(-1,1)+
  #facet_grid(rows = "Dataset") + 
  ggtitle("Chr 8 Fixed Copy Numbers")
```

## Distribution of MQC Coverage across successful or nonsuccessful runs

It's worth noting that the best (least noisy) experiments were of the highest coverage, so that might need to be a priority next run.

```{r}
MQCRuns <- read.csv("C:\\Users\\cassa\\OneDrive\\Documents\\SiegalLab\\Sequencing_Tuboweb\\AllMultiQCRuns.csv")

MQCRuns %>% filter(grepl("undetermined", Library) == FALSE) %>% group_by(ShortName) %>%
  summarize(MinCoverage = min(MillionReads),
            MaxCoverage = max(MillionReads)) %>%
  mutate(Range = MaxCoverage - MinCoverage)

MQCRuns %>% filter(grepl("undetermined", Library) == FALSE) %>%
  ggplot(aes(x = MillionReads, y = ShortName, color = ShortName)) + geom_boxplot(outlier.shape = NA) + geom_jitter(size = 2, alpha = 0.5) +
  scale_color_manual(values = c("skyblue", "skyblue", "skyblue", "skyblue", "skyblue","darkred", "darkred")) +
  ggtitle("Coverage Colored by Success of Experiment")

MQCRuns %>% filter(grepl("undetermined", Library) == FALSE) %>%
  filter(Bulk != "Dilute") %>%
  ggplot(aes(x = SurvivalPercentage, y = ShortName, color = ShortName)) + geom_boxplot(outlier.shape = NA) + geom_jitter(size = 2, alpha = 0.5) +
  scale_color_manual(values = c("skyblue", "skyblue", "skyblue", "skyblue", "skyblue","darkred", "darkred")) +
  ggtitle("Selection % Colored by Success of Experiment")
```

## Finding SNPs in common between Oak and Wine

GQ Cutoff is 98 for all samples, so find out how many of these are under 98 and in which samples that tends to occur

```{r}
setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")

parentSNPid_alls <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = FALSE)

#20k loci where oak and wine are both the same but different from reference
parentSNPid_alls %>% filter(Type == 0) %>% select(CHROM, POS) %>% distinct() %>% mutate(Type = "both") -> pSNPs_both

#400 loci where oak and wine are both different from reference in different ways
parentSNPid_alls %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) %>% filter(Unique == 2) %>%
  select(CHROM, POS) %>% distinct() %>% mutate(Type = "diff") -> pSNPs_diff

```

Looking at what each mutation actually is? This will find those that are more than a single insertion.

```{r}
parentSNPid_alls %>% filter(nchar(REF) > 2 | nchar(Wine) > 2 | nchar(Oak) > 2) -> parentSNP_ins
```

Which of these are in CUP1?

```{r, fig.height = 4, fig.width = 15}
rbind(pSNPs_both, pSNPs_diff) %>% filter(CHROM == "VIII") %>%
  filter(POS > 200000, POS < 225000) %>%
  ggplot(aes(x = POS, y = Type)) + geom_point(size = 2, alpha = 0.4) +
  geom_vline(xintercept = c(212535, 212720)) +
  geom_vline(xintercept = c(214533,214718)) 

```


Align reads to all of the ones that are just both the same - see how many non-alt alleles we have as a measure of error

```{r}
alldata <- readRDS(file = "allseqruns_jan24.rds")

alldata %>% merge(pSNPs_diff) %>% mutate(AD.REF= as.numeric(AD.REF),
                                 AD.ALT = as.numeric(AD.ALT))  -> alldata_differentSNPs

alldata_differentSNPs %>% pivot_longer(c(AD.REF, AD.ALT), names_to = "Allele", values_to = "Reads") %>% 
  ggplot(aes(x = Reads, y = Pool, color = Allele)) + geom_boxplot()

alldata_differentSNPs %>% mutate(logOdds = log(AD.ALT/AD.REF)) %>% ggplot(aes(x = logOdds, y = Coverage, color = Pool)) + geom_point(alpha = 0.4)

alldata_differentSNPs %>% filter(AD.REF > 0) %>% ggplot(aes(x = AD.REF, y = Pool)) + geom_boxplot()

#4000 total in EVERYTHING that didn't have any reference alleles to what wasn't there
alldata_differentSNPs %>% filter(AD.REF == 0) %>% ggplot(aes(x = Pool)) + geom_bar()
```

Look at what the ratios would be for all of the ones that aren't there

```{r}
alldata %>% merge(pSNPs_both) %>% mutate(AD.REF= as.numeric(AD.REF),
                                 AD.ALT = as.numeric(AD.ALT)) -> alldata_bothSNPs

alldata_bothSNPs %>% pivot_longer(c(AD.REF, AD.ALT), names_to = "Allele", values_to = "Reads") %>% 
  ggplot(aes(x = Reads, y = Pool, color = Allele)) + geom_boxplot(outlier.alpha = 0.2)

```
```{r, fig.height=5, fig.width=12}
alldata_bothSNPs %>% mutate(logOdds = log(AD.ALT/AD.REF)) %>% ggplot(aes(x = POS, y = abs(logOdds), color = Pool)) + 
  geom_point(size = 0.5, alpha = 0.4) + facet_grid(Pool~CHROM, scales = "free", space = "free")
```

# Plot HT Chr 2?

```{r, fig.width=6, fig.height=6, eval = FALSE}

rawdata_called %>% group_by(Pool, Dataset, POS, CHROM, Allele) %>% count() %>% filter(n == 1) %>% merge(rawdata_called) -> HTG3TDMXY_called
HTG3TDMXY_called %>% pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(logOdds = log(Wine/Oak)) -> HTG3TDMXY_logOdds
HTG3TDMXY_logOdds %>% separate(Dataset, into = c("Background", "Selection", "Day"), sep = "_") -> HTG3TDMXY_logOdds

HTG3TDMXY_logOdds %>% filter(CHROM == "VII") %>% ggplot(aes(x = POS, y = logOdds, color = Selection)) + geom_point(size = 0.5, alpha = 0.5) + facet_grid(Background~CHROM, scales = "free", space = "free")
```

```{r, fig.width=6, fig.height=12, eval = FALSE}
HTG3TDMXY_logOdds %>% filter(CHROM == "II") %>% ggplot(aes(x = POS, y = logOdds)) + geom_point(size = 0.5, alpha = 0.5) + facet_grid(Dataset~CHROM, scales = "free", space = "free")
```


