---
title: "CSS Sequence Analysis Template"
date: "Jan 2023"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

#library(cybrBSA)

ggplot2::theme_set(theme_light())

################################################################################
glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
#glm with replicates
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

cybr2_rollmean <- function(dataframe){
  dataframe %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollmean(value, n = 100))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount)
}

################################################################################
#Define triple replicate function
glmfixed_rep3 <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                         HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", 
                                             "O", "O", "W", "W", "O", "O", "W", "W",
                                             "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", 
                                          "B", "B", "B", "B","B", "B", "B", "B",
                                          "C", "C", "C", "C", "C", "C","C", "C")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                                     HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                                     HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

```

# GATK Table to Data Frame

```{r}
setwd("../../../../GitHub/Sequencing/Analysis/")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) %>% filter(Unique == 1) -> pSNPs

#CHANGE THIS
#mydatatotest = "Data/HKTFTDRX2.SortedCat.vcf.output.table"
mydatatotest = "C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/Data/TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"


FilteredData <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = pSNPs, yeast = TRUE) %>% 
  na.omit()

#unique(FilteredData$CHROM)

#Using Gsub for this
# gsub(FilteredData$Dataset, "HNGLVDRXY_n01_", "") #CHANGE THIS
FilteredData %>% mutate(DS = Dataset) %>%  select(-Dataset) -> tempFilteredData

tempFilteredData$Bulk <- NA
tempFilteredData$Parent <- NA

tempFilteredData$Bulk[grep("Selected", tempFilteredData$DS)] <- "CuSO4" #CHANGE THIS
tempFilteredData$Bulk[grep("Unselected", tempFilteredData$DS)] <- "Dilute"

# tempFilteredData$Rep[grep("a", tempFilteredData$DS)] <- "A"
# tempFilteredData$Rep[grep("b", tempFilteredData$DS)] <- "B"

tempFilteredData$Parent[grep("A", tempFilteredData$DS)] <- "Oak"
tempFilteredData$Parent[grep("C", tempFilteredData$DS)] <- "Wine"

tempFilteredData$ReadCount <- as.numeric(tempFilteredData$ReadCount)
  
# #THIS IGNORES REPLICATES
tempFilteredData %>% select(CHROM, POS, PAllele, ReadCount, Bulk, Parent) %>% distinct %>% 
 pivot_wider(names_from = c(Bulk, Parent, PAllele), values_from = ReadCount) -> cybr2Data

saveRDS(object = cybr2Data, file = "Data/CuSO4_1_cybr2.rds")

#cybr2Data <- readRDS("Data/CuSO4_1_cybr2.rds")

```


Check Coverage and log(Wine/Oak) of Sequencing Run

```{r, fig.width=18, fig.height=5}
cybr2Data %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Allele"), names_sep = "_") %>%
  pivot_wider(names_from = Allele, values_from = value) %>% mutate(Coverage = Wine + Oak, logWineOak = log(Wine/Oak)) -> RawCountSummary

RawCountSummary %>% filter(CHROM == "VIII") %>%
  group_by(Bulk, Parent) %>% mutate(RelativeCoverage = Coverage/median(Coverage)) %>%
  ggplot(aes(x = POS, y = RelativeCoverage, color = paste(Bulk, Parent))) + geom_point(alpha = 0.3) +
  geom_vline(xintercept = c(212535, 212720))

RawCountSummary %>% filter(CHROM == "VIII") %>%
  group_by(Bulk, Parent) %>% mutate(RelativeCoverage = Coverage/median(Coverage)) %>%
  ungroup() %>%
  group_by(POS) %>%
  summarize(MeanRelCov = mean(RelativeCoverage),
            SD = sd(RelativeCoverage)) %>%
  filter(SD < 0.5) %>%
  ggplot(aes(x = POS, y = MeanRelCov, color = SD)) + geom_point() +
  geom_vline(xintercept = c(212535, 212720), color = "turquoise")

RawCountSummary %>% filter(CHROM == "VIII") %>%
  group_by(Bulk, Parent) %>% mutate(RelativeCoverage = Coverage/median(Coverage)) %>%
  ungroup() %>%
  ggplot(aes(x = POS, y = RelativeCoverage, color = Bulk)) + geom_point() +
  geom_vline(xintercept = c(212535, 212720), color = "turquoise")

RawCountSummary %>% filter(CHROM == "VIII") %>%
  group_by(Bulk, Parent) %>% mutate(RelativeCoverage = Coverage/median(Coverage)) %>%
  ungroup() %>%
  ggplot(aes(x = POS, y = RelativeCoverage, color = Parent)) + geom_point() +
  geom_vline(xintercept = c(212535, 212720))

RawCountSummary %>% filter(CHROM == "VIII") %>%
  group_by(Bulk, Parent) %>% mutate(RelativeCoverage = Coverage/median(Coverage)) %>%
  ungroup() %>%
  ggplot(aes(x = POS, y = RelativeCoverage, color = abs((RelativeCoverage) - 1) > 0.5)) + geom_point() +
  geom_vline(xintercept = c(212535, 212720))

```

Smooth by rolling mean or median

```{r}
#Use rolling average of 100 SNPs, finding the mean
#cybr2Data %>% cybr2_rollmean() -> rollmeanData

#Find the rolling median or change n instead
cybr2Data %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median, align = "center"))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

```

Calculate GLM from rolled data

```{r, warning=FALSE}
#Change for different datasets
mydata <- rollData
str(mydata)
#Run GLM - options are glmfixed_rep(), glmfixed_rep3(), and glmfixed()
#IF NOT USING REPS, change 1:5 to 1:4, and remove "Rep" from labels
mydata %>% group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed(HOO = CuSO4_Oak_Oak, 
                           HOW = CuSO4_Oak_Wine, 
                           HWO = CuSO4_Wine_Oak,
                           HWW = CuSO4_Wine_Wine,
                           LOO = Dilute_Oak_Oak,
                           LOW = Dilute_Oak_Wine,
                           LWO = Dilute_Wine_Oak, 
                           LWW = Dilute_Wine_Wine)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> GLMdata

#saveRDS(GLMdata, file = "Data/2023/PLACEHOLDER_glm.Rdata")
saveRDS(GLMdata, file = "Data/CuSO4_1_glm_c.rds")

```

## Basic Plots

Single GLM Plot for this Data for reference

```{r}
GLMdata %>% 
  filter(label != "intercept", CHROM != "I", CHROM != "M") %>% 
  
  ggplot(aes(x = POS, y = summary, color = label)) + geom_line() + 
  
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ggtitle("GLM of Data")
```

Log Odds of Alleles Plot for reference

```{r}
RawCountSummary %>% 
  filter(CHROM != "I", CHROM != "M") %>% 
  
  ggplot(aes(x = POS, y = logWineOak, shape = paste(Bulk, Parent, Rep, sep = "_"), color = Bulk)) + 
  geom_point(alpha = 0.3) + 
  
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  scale_color_manual(values = c("Violet", "Black"))
```

