---
title: "CSS Sequence Analysis Parent Permutations"
date: "Jan 2023"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

#library(cybrBSA)

#install.packages("lme4")
library(lme4)

ggplot2::theme_set(theme_light())

################################################################################
glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
#glm with replicates
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

cybr2_rollmean <- function(dataframe){
  dataframe %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollmean(value, n = 100))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount)
}

################################################################################
#Define triple replicate function
glmfixed_rep3 <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                         HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", 
                                             "O", "O", "W", "W", "O", "O", "W", "W",
                                             "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", 
                                          "B", "B", "B", "B","B", "B", "B", "B",
                                          "C", "C", "C", "C", "C", "C","C", "C")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                                     HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                                     HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

```

Smooth by rolling mean or median

```{r}
getwd()
cybr2Data <- readRDS("Data/Fluconazole_1_cybr2.rds")

#Use rolling average of 100 SNPs, finding the mean
#cybr2Data %>% cybr2_rollmean() -> rollmeanData

#Find the rolling median or change n instead
cybr2Data %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

saveRDS(rollData, file = "Fluc_rollData.rds")
```

## PERMUTING PARENT

So I want to swap who is oak or wine at some percentage of loci, so let's just look at several bulks and change which they are?

```{r}
cybr2Data <- readRDS("Data/CuSO4_2_cybr2.rds")

#Use rolling average of 100 SNPs, finding the mean
#cybr2Data %>% cybr2_rollmean() -> rollmeanData

#Find the rolling median or change n instead
cybr2Data %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

rollData %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_", values_to = "Reads")  -> combineddata

cybr2Data %>% ungroup() %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>%
  group_by(CHROM, label)%>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = value) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) %>%
  pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_", values_to = "Reads") %>%
  #filter(CHROM == "I") %>%
  pivot_wider(names_from = c(Bulk, Rep, Allele), values_from = Reads) %>% 
  group_by(CHROM, POS) %>%
  summarize(Parent_P = sample(c("Oak", "Wine"), replace = F, size = 2),
            CuSO4_A_Oak = CuSO4_A_Oak,
            CuSO4_B_Oak = CuSO4_B_Oak,
            CuSO4_A_Wine = CuSO4_A_Wine,
            CuSO4_B_Wine = CuSO4_B_Wine,
            Dilute_A_Oak = Dilute_A_Oak,
            Dilute_B_Oak = Dilute_B_Oak,
            Dilute_A_Wine = Dilute_A_Wine,
            Dilute_B_Wine = Dilute_B_Wine) -> ParentPerm_unrolled

ParentPerm_unrolled %>% pivot_longer(c(-CHROM, -POS, -Parent_P)) %>%
  pivot_wider(names_from = c(name, Parent_P), values_from = value) -> ParentPerm_unrolled_rev

#colnames(ParentPerm_rev)

ParentPerm_unrolled_rev %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> Permuted_rollData
```


```{r}
Permuted_rollData %>% 
  group_by(CHROM, POS) %>% summarize(Effect = glmfixed_rep(HOWa = CuSO4_A_Wine_Oak, 
                                                                HWOa = CuSO4_A_Oak_Wine,
                                                                HWWa = CuSO4_A_Wine_Wine,
                                                                HOOa = CuSO4_A_Oak_Oak,
                                                                LOOa = Dilute_A_Oak_Oak,
                                                                LOWa = Dilute_A_Wine_Oak,
                                                                LWOa = Dilute_A_Oak_Wine,
                                                                LWWa = Dilute_A_Wine_Wine,
                                                           HOWb = CuSO4_B_Wine_Oak, 
                                                                HWOb = CuSO4_B_Oak_Wine,
                                                                HWWb = CuSO4_B_Wine_Wine,
                                                                HOOb = CuSO4_B_Oak_Oak,
                                                                LOOb = Dilute_B_Oak_Oak,
                                                                LOWb = Dilute_B_Wine_Oak,
                                                                LWOb = Dilute_B_Oak_Wine,
                                                                LWWb = Dilute_B_Wine_Wine),
                                                       label = c("Z_intercept", "Z_Bulk", "Z_Parent", "Rep","Z_Interaction")) -> save_permuted

#saveRDS(save_permuted, file = "save_permuted_CuSO4.rds")

save_permuted %>% filter(label != "Z_intercept") %>% ggplot(aes(x = POS, y = abs(Effect), color = label)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free") + 
  scale_color_manual(values = c("gray", "gray20", "red", "black")) +
  theme(legend.position = "bottom")
```
Comparing to the original values

```{r}
rollData %>% group_by(CHROM, POS) %>% summarize(Effect = glmfixed_rep(HOWa = CuSO4_Oak_A_Wine, 
                                                                HWOa = CuSO4_Wine_A_Oak,
                                                                HWWa = CuSO4_Wine_A_Wine,
                                                                HOOa = CuSO4_Oak_A_Oak,
                                                                LOOa = Dilute_Oak_A_Oak,
                                                                LOWa = Dilute_Oak_A_Wine,
                                                                LWOa = Dilute_Wine_A_Oak,
                                                                LWWa = Dilute_Wine_A_Wine,
                                                           HOWb = CuSO4_Oak_B_Wine, 
                                                                HWOb = CuSO4_Wine_B_Oak,
                                                                HWWb = CuSO4_Wine_B_Wine,
                                                                HOOb = CuSO4_Oak_B_Oak,
                                                                LOOb = Dilute_Oak_B_Oak,
                                                                LOWb = Dilute_Oak_B_Wine,
                                                                LWOb = Dilute_Wine_B_Oak,
                                                                LWWb = Dilute_Wine_B_Wine),
                                                       label = c("Z_intercept", "Z_Bulk", "Z_Parent", "Rep","Z_Interaction")) -> originaldata

originaldata %>% filter(label != "Z_intercept") %>% ggplot(aes(x = POS, y = abs(Effect), color = label)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free") + 
  scale_color_manual(values = c("gray", "gray20", "red", "black")) +
  theme(legend.position = "bottom")
```

Comparing the two

```{r}
originaldata %>% mutate(Permutation = "Original") -> OG
save_permuted %>% mutate(Permutation = "Permuted") %>% rbind(OG) -> comparison

comparison %>% filter(label == "Z_Interaction") %>% ggplot(aes(x = POS, y = Effect, color = Permutation)) + geom_line(alpha = 0.4) + facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("Interaction Scores")

comparison %>% filter(label == "Z_Parent") %>% ggplot(aes(x = POS, y = abs(Effect), color = Permutation)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("Parent Scores")
```

