---
title: "CSS Sequence Analysis Template"
date: "Jan 2023"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

#library(cybrBSA)

#install.packages("lme4")
library(lme4)

ggplot2::theme_set(theme_light())

################################################################################
glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
#glm with replicates
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

cybr2_rollmean <- function(dataframe){
  dataframe %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollmean(value, n = 100))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount)
}

################################################################################
#Define triple replicate function
glmfixed_rep3 <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                         HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", 
                                             "O", "O", "W", "W", "O", "O", "W", "W",
                                             "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", 
                                          "B", "B", "B", "B","B", "B", "B", "B",
                                          "C", "C", "C", "C", "C", "C","C", "C")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                                     HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                                     HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

```

Smooth by rolling mean or median

```{r}
cybr2Data <- readRDS("Data/Fluconazole_1_cybr2.rds")

#Use rolling average of 100 SNPs, finding the mean
#cybr2Data %>% cybr2_rollmean() -> rollmeanData

#Find the rolling median or change n instead
cybr2Data %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

saveRDS(rollData, file = "Fluc_rollData.rds")
```

## NEW STUFF

```{r}

rollData %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_", values_to = "Reads") %>% mutate(RepN = paste(Bulk, Parent, Rep, sep= "_")) -> combineddata

combineddata %>% subset(CHROM == "II" & POS == 97557) -> cprelim

cprelim %>% pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(Reads = log(Wine/Oak))-> cdata

cprelim
df <- as.data.frame(lapply(cprelim, rep, cprelim$Reads)) %>% select(-Reads)

ggplot(df, aes(x = Allele, y = Bulk, color = Parent)) + geom_jitter() + facet_wrap(~Rep)

#glm(Allele ~ Bulk*Parent+Rep, data = df, family = "binomial")

```

```{r}

cdata %>% 
  ggplot(aes(x = Bulk, y = Reads, color = Parent)) + geom_jitter(size = 3, alpha = 0.6)

aov(Reads ~ Bulk*Parent+Rep, data = cdata)
summary(aov(Reads ~ Bulk*Parent+Rep, data = cdata))

```


```{r, warning=F}
ggplot(cprelim, aes(x = Allele, y = Reads, shape = Bulk, color = paste(Bulk, Parent, Rep))) + 
  geom_jitter(size = 4, width = 0.2, alpha = 0.8) + 
  scale_shape_manual(labels=c("Dilute", "Fluconazole"),
                 values = c(2,17)) +
  ggtitle("Alleles vs Reads in each Bulk")
```

Maybe I actually do need to run lmer?

```{r}

#?glmer

#REPEAT ~ SEX + PPED + MSESC + (1 + PPED|SCHOOLID)
cprelim$Bulk <- factor(cprelim$Bulk)
cprelim$Parent <- factor(cprelim$Parent)
cprelim$Rep <- factor(cprelim$Rep)
cprelim$Allele <- factor(cprelim$Allele)

glmer(Allele ~ Bulk*Parent+(1 + Bulk + Parent | Rep), weights = Reads, family = binomial, 
              data = cprelim)
```

## Workable New Code
Doing this the way the other analyses are done to see patterns

```{r}
glm_mixed_model_Fluc5_div <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(1, 1, 2,2,
                                     3,3,4, 4,
                                     5, 5, 6,6,
                                     7,7,8, 8,
                                     9, 9),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))
  
  combineddata %>% group_by(Bulk, Parent) %>% summarize(key = factor(paste(Bulk, Parent, sep = "_"))) -> combineddata

  b <- glmer(Allele ~ Bulk*Parent+(1 | key/Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  
  return(c(summary(b)$coefficients[1:12], #Effects
             as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))))
  
} 

################################################################################
glm_mixed_model_Fluc1 <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(c("B", "B", "B", "B","C", "C", "D", "D", 
                                          "B", "B", "C", "C","D", "D", "C", "C",
                                          "D", "D")),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))

  b <- glmer(Allele ~ Bulk*Parent+(1 + Bulk * Parent | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  #This seems to work
  #repstuff <- lme4::VarCorr(b)
  #repstuff$Rep
  #attr(repstuff$Rep,"stddev")
  
  # glmsd <- as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))
  
  #original code
  # zscores <- summary(b)$coefficients[
  #   ((length(summary(b)$coefficients)/4)*2+1):
  #     ((length(summary(b)$coefficients)/4)*3)]
  
  #Effects
  # effects <- summary(b)$coefficients[1:4]
  
  return(c(summary(b)$coefficients[1:4], #Effects
           summary(b)$coefficients[((length(summary(b)$coefficients)/4)*2+1):((length(summary(b)$coefficients)/4)*3)],
             as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))))
  
} 



glm_mixed_model_Fluc2 <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(c("B", "B", "B", "B","C", "C", "D", "D", 
                                          "B", "B", "C", "C","D", "D", "C", "C",
                                          "D", "D")),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))

  b <- glmer(Allele ~ Bulk*Parent+(1 | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  return(c(summary(b)$coefficients[1:4], #Effects
           summary(b)$coefficients[((length(summary(b)$coefficients)/4)*2+1):((length(summary(b)$coefficients)/4)*3)],
             as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))))
  
} 

glm_mixed_model_Fluc2b <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(c("B", "B", "B", "B","C", "C", "D", "D", 
                                          "B", "B", "C", "C","D", "D", "C", "C",
                                          "D", "D")),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))

  b <- glmer(Allele ~ Bulk*Parent+(1 | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  return(c(summary(b)$coefficients[1:8], #Effects and SE
             as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))))
  
} 

glm_mixed_model_Fluc3 <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(c("A", "B", "C", "D","E", "F", "G", "H", 
                                          "I", "J", "K", "L","M", "N", "O", "P",
                                          "Q", "R")),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))

  b <- glmer(Allele ~ Bulk*Parent+(1 | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  return(c(summary(b)$coefficients[1:4], #Effects
           summary(b)$coefficients[((length(summary(b)$coefficients)/4)*2+1):((length(summary(b)$coefficients)/4)*3)],
             as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))))
  
} 

#Now have it save the SE
glm_mixed_model_Fluc4_intbyrep <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(c("B", "B", "B", "B","C", "C", "D", "D", 
                                          "B", "B", "C", "C","D", "D", "C", "C",
                                          "D", "D")),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))

  b <- glmer(Allele ~ Bulk*Parent+(1 + Bulk * Parent | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  
  #Effects
  # effects <- summary(b)$coefficients[1:4]
  
  return(c(summary(b)$coefficients[1:12]))
  
} 

glm_mixed_model_Fluc4_unmixed <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(c("B", "B", "B", "B","C", "C", "D", "D", 
                                          "B", "B", "C", "C","D", "D", "C", "C",
                                          "D", "D")),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))

  b <- glm(Allele ~ Bulk*Parent+Rep, 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  return(c(summary(b)$coefficients[1:12]))
  
} 

glm_mixed_model_Fluc4_withSD <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(c("B", "B", "B", "B","C", "C", "D", "D", 
                                          "B", "B", "C", "C","D", "D", "C", "C",
                                          "D", "D")),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))

  b <- glmer(Allele ~ Bulk*Parent+(1 | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata, nAGQ=20)
  
  
  #Effects
  # effects <- summary(b)$coefficients[1:4]
  
  return(c(summary(b)$coefficients[1:12], 
           as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))
           ))
  
} 
```

```{r, eval = FALSE}
#Specific to Fluconazole
rollData %>% filter(CHROM == "I") %>% head(200) %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc1(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction",
                                                             "SD_intercept", "SD_Bulk", "SD_Parent", "SD_Interaction")) -> GLMdata

GLMdata %>% pivot_wider(names_from = label, values_from = summary) -> GLM_PivotW

GLM_PivotW %>% ggplot(aes(x = POS, y = E_Bulk)) + geom_line() + geom_ribbon(aes(ymin = E_Bulk - SD_Bulk, ymax = E_Bulk + SD_Bulk, alpha = 0.2)) +
  geom_line(aes(x = POS, y = E_Parent), color = "violet") + 
  geom_ribbon(aes(ymin = E_Parent - SD_Parent, ymax = E_Parent + SD_Parent, alpha = 0.2), fill = "violet") +
  geom_line(aes(x = POS, y = E_Interaction), color = "lightblue") + 
  geom_ribbon(aes(ymin = E_Interaction - SD_Interaction, ymax = E_Interaction + SD_Interaction, alpha = 0.2), fill = "lightblue")


```
Do this for Fluconazole Chr VII

```{r, eval = FALSE}
#Specific to Fluconazole
rollData %>% filter(CHROM == "VII") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc1(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction",
                                                             "SD_intercept", "SD_Bulk", "SD_Parent", "SD_Interaction")) -> GLMdata_7

#saveRDS(GLMdata_7, file = "GLMdata_7.rds")
```

```{r}
GLMdata_7 <- readRDS("GLMdata_7.rds")
GLMdata_7 %>% pivot_wider(names_from = label, values_from = summary) -> GLMdata_7_PivotW

GLMdata_7_PivotW %>% ggplot(aes(x = POS, y = E_Bulk)) + geom_line() + geom_ribbon(aes(ymin = E_Bulk - SD_Bulk, ymax = E_Bulk + SD_Bulk, alpha = 0.2)) +
  geom_line(aes(x = POS, y = E_Parent), color = "violet") + 
  geom_ribbon(aes(ymin = E_Parent - SD_Parent, ymax = E_Parent + SD_Parent, alpha = 0.2), fill = "violet") +
  geom_line(aes(x = POS, y = E_Interaction), color = "lightblue") + 
  geom_ribbon(aes(ymin = E_Interaction - SD_Interaction, ymax = E_Interaction + SD_Interaction, alpha = 0.2), fill = "lightblue")
```

A different plot

```{r}
GLMdata_7_PivotW %>% ggplot(aes(x = POS, y = Z_Bulk)) + geom_line() + geom_ribbon(aes(ymin = E_Bulk - SD_Bulk, ymax = E_Bulk + SD_Bulk, alpha = 0.2)) +
  geom_line(aes(x = POS, y = Z_Parent), color = "violet") + 
  geom_ribbon(aes(ymin = Z_Parent - SD_Parent, ymax = Z_Parent + SD_Parent, alpha = 0.2), fill = "violet") +
  geom_line(aes(x = POS, y = E_Interaction), color = "lightblue") + 
  geom_ribbon(aes(ymin = Z_Interaction - SD_Interaction, ymax = Z_Interaction + SD_Interaction, alpha = 0.2), fill = "lightblue")
```

```{r, eval = FALSE}
maketable <- function(inputdf){
  inputdf %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), 
                           names_sep = "_", values_to = "Reads") -> combineddata
    
  b <- glmer(Allele ~ Bulk*Parent+(1 + Bulk * Parent | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  #This seems to work
  #repstuff <- lme4::VarCorr(b)
  #repstuff$Rep
  #attr(repstuff$Rep,"stddev")
  
  glmsd <- as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))
  
  #original code
  zscores <- summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
  
  #Effects
  effects <- summary(b)$coefficients[1:4]
  
  return(c(effects, glmsd))
  
}

rollData %>% filter(CHROM == "III") %>% group_by(CHROM, POS) %>% summarise(summary = maketable(.),
                                                   label = c("intercept", "Bulk", "Parent", "Interaction",
                                                             "SD_intercept", "SD_Bulk", "SD_Parent", "SD_Interaction")) -> GLMdata_table

GLMdata_table %>% pivot_wider(names_from = label, values_from = summary) -> GLM_PivotW_table
```

```{r, eval = FALSE}
#Specific to Fluconazole
rollData %>% filter(CHROM == "I") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc1(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction",
                                                             "SD_intercept", "SD_Bulk", "SD_Parent", "SD_Interaction")) -> GLMdata_1

saveRDS(GLMdata_1, file = "GLMdata_1.rds")

```

```{r}
GLMdata_1 <-readRDS("GLMdata_1.rds")
GLMdata_1 %>% pivot_wider(names_from = label, values_from = summary) -> GLMdata_1_PivotW

GLMdata_1_PivotW %>% ggplot(aes(x = POS, y = E_Bulk)) + geom_line() + geom_ribbon(aes(ymin = E_Bulk - SD_Bulk, ymax = E_Bulk + SD_Bulk, alpha = 0.2)) +
  geom_line(aes(x = POS, y = E_Parent), color = "violet") + 
  geom_ribbon(aes(ymin = E_Parent - SD_Parent, ymax = E_Parent + SD_Parent, alpha = 0.2), fill = "violet") +
  geom_line(aes(x = POS, y = E_Interaction), color = "lightblue") + 
  geom_ribbon(aes(ymin = E_Interaction - SD_Interaction, ymax = E_Interaction + SD_Interaction, alpha = 0.2), fill = "lightblue") +
  ylim(c(-100,100))
```

Doing this the new way

If there are 0s, then it won't run, so have to add 1 to everything...? But that shouldn't be right.

```{r, eval = FALSE}

rollData %>% filter(CHROM == "I" & POS == 41456) %>% 
  pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_", values_to = "Reads") %>% 
  ggplot(aes(x = Bulk, y = Reads, color = Parent)) + 
  geom_jitter(size = 3, alpha = 0.6)

#Specific to Fluconazole
rollData %>% filter(CHROM == "I" & POS == 41456) -> prelim

#Add one to everything
prelim[,3:20] <- 1+ prelim[,3:20]

#Rerun with 1+ whatever
rollData %>% filter(CHROM == "VII") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc2(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction",
                                                             "RandSD")) -> GLMdata_2_VII

GLMdata_2_VII %>% pivot_wider(names_from = label, values_from = summary) -> GLM_PivotW_2_VII

#This was only done for one, so maybe do it for all?
saveRDS(GLM_PivotW_2_VII, file = "GLM_PivotW_2_VII.rds")
```

```{r, eval = FALSE}
#Rerun with 1+ whatever
rollData %>% filter(CHROM == "VII") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc2b(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "SE_intercept", "SE_Bulk", "SE_Parent", "SE_Interaction",
                                                             "RandSD")) -> GLMdata_2_VIIb

GLMdata_2_VIIb %>% pivot_wider(names_from = label, values_from = summary) -> GLM_PivotW_2_VIIb

#This was only done for one, so maybe do it for all?
saveRDS(GLM_PivotW_2_VIIb, file = "GLM_PivotW_2_VIIb.rds")

```

Plotting the only 1| Rep one

```{rm eval = FALSE}
GLM_PivotW_2_VIIb <- readRDS("GLM_PivotW_2_VIIb.rds")
GLM_PivotW_2_VIIb %>% ggplot(aes(x = POS, y = E_Bulk)) + 
  geom_ribbon(aes(ymin = E_Bulk - 2*SE_Bulk, ymax = E_Bulk + 2*SE_Bulk), alpha = 0.2) + 
  geom_ribbon(aes(ymin = E_Parent - 2*SE_Parent, ymax = E_Parent + 2*SE_Parent), fill = "orchid", alpha = 0.2) +
  geom_ribbon(aes(ymin = E_Interaction - 2*SE_Interaction, ymax = E_Interaction + 2*SE_Interaction), fill= "lightblue", alpha = 0.2) + 
  geom_line(aes(y = E_Interaction), color = "lightblue") + 
  geom_line(aes(y = E_Parent), color = "orchid") + 
  geom_line()

GLMdata_2_VII %>% filter(label %in% c("Z_Bulk", "Z_Interaction", "Z_Parent", "Z_intercept") == FALSE) %>% ggplot(aes(x = log(summary), color = label)) + geom_density() + ggtitle("Random Effect SD is MUCH smaller than all other effects")
```


Running the 3rd function on Chr 7 - this does not look right

```{r, eval = FALSE}
#Specific to Fluconazole
rollData %>% filter(CHROM == "VII") %>%  group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc3(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine) ,
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction",
                                                             "SD_all")) -> GLMdata_7_3

saveRDS(GLMdata_7_3, file = "GLMdata_7_3.rds")
```

Plotting this

```{r}
GLMdata_7_3 <- readRDS("GLMdata_7_3.rds")
GLMdata_7_3 %>% pivot_wider(names_from = label, values_from = summary) -> GLMdata_7_3_PivotW

ggplot(GLMdata_7_3_PivotW, aes(x = POS, y = E_Bulk)) + geom_line() +
  geom_line(aes(x = POS, y = E_Parent), color = "orchid") +
  geom_line(aes(x = POS, y = E_Interaction), color = "pink") +
  ylim(c(-5, 5))

```
Testing the formulas a bit more

```{r, eval = FALSE}

cdata %>% ggplot(aes(x = Bulk, y = Reads, color = Parent)) + geom_jitter(size = 3, alpha = 0.6)

aov(Reads ~ Bulk*Parent+Rep, data = cdata)
summary(aov(Reads ~ Bulk*Parent+Rep, data = cdata))

#Model 1: Just the intercept
glmer(Reads ~ 1 + (1 | Rep), cdata)
plot(glmer(Reads ~ 1 + (1 | Rep), cdata))

#Model 2: Adding in Bulk
glmer(Reads ~ 1 + Bulk + (1 + Bulk | Rep), cdata)
plot(glmer(Reads ~ 1 + Bulk + (1 + Bulk | Rep), cdata))

#Model 3: Adding in Parent - I think this is right?
glmer(Reads ~ 1 + Bulk*Parent + (1 | Rep), cdata)
summary(glmer(Reads ~ 1 + Bulk*Parent + (1 | Rep), cdata))
#plot(glmer(Reads ~ 1 + Bulk*Parent + (1 | Rep), cdata))

#Model 4: Removing intercept? in Parent*Bulk + Rep? THIS IS THE SAME so intercept isn't necessary
glmer(Reads ~ Bulk*Parent + (1 | Rep), cdata)

#GLM New functions:
#F1 - I think this is the right answer?
b <- glmer(Allele ~ Bulk*Parent+(1 + Bulk * Parent | Rep), #Adds the interaction vs rep?
             weights = Reads, family = binomial, 
              data = combineddata)
#F2 - meh
b <- glmer(Allele ~ Bulk*Parent+(1 | Rep), #just by rep?
             weights = Reads, family = binomial, 
              data = combineddata)

#F3 - doesn't seem to give the right answer though...
b <- glmer(Allele ~ Bulk*Parent+(1 | Rep), #also just by rep
             weights = Reads, family = binomial, 
              data = combineddata)
```

ChatGPT Advice

```{r}
library(lme4)

# assume your data is stored in a data frame called 'mydata', with columns 'y', 'x1', 'x2', 'replicate', and 'category'
mydata <- cdata %>% 
  filter(CHROM == "II" & POS == 97557) %>% 
  pivot_longer(cols = c(Oak, Wine), names_to = "category", values_to = "y") %>% 
  transmute(y = y, x1 = factor(Bulk), x2 = factor(Parent), replicate = RepN, category = factor(category))

# fit the GLMM with a binomial family and logit link function
model <- glmer(category ~ x1 * x2 + (1 | replicate), data = mydata, family = binomial, weights = y)

# extract the fixed effect estimates and their standard errors
fe <- fixef(model)
se <- sqrt(diag(vcov(model)))

# create a table to display the results
results <- data.frame(Fixed_Effect = c("Intercept", "x1", "x2", "x1:x2"),
                      Estimate = fe,
                      SE = se,
                      Lower_CI = fe - 1.96 * se,
                      Upper_CI = fe + 1.96 * se)

# display the results
print(results)

```

```{r}
cdata %>% 
  filter(CHROM == "II" & POS == 97557) -> testmydata

#Plot the linear model
plot(x = as.numeric(mydata$x1), y = mydata$y, col = mydata$x2)
abline(lm(y~x1, mydata))
abline(lm(y~x1, mydata, subset = x2 == "OakI"), lty = "dashed")
abline(lm(y~x1, mydata, subset = x2 == "WineI"), lty = "dashed", col = "red")
abline(lm(y~x1*x2, mydata), col = "blue")

#run glm on the one
summary(glm(category~x1*x2, mydata, family = "binomial", weights = y))

#run glm on the odds
glm(Reads ~ Bulk*Parent, testmydata, family = "binomial")
glm(Reads ~ Bulk*Parent + Rep, testmydata, family = "binomial")
b <- glmer(Reads ~ Bulk*Parent + (1| Rep), testmydata, family = "binomial")

```

Maybe actually save the SEs on the estimates? And then 2*that is the CI?

```{r, eval = FALSE}
#Specific to Fluconazole
rollData %>% filter(CHROM == "VII") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc4_intbyrep(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "SE_intercept", "SE_Bulk", "SE_Parent", "SE_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction")) -> GLMdata_test_7

saveRDS(GLMdata_test_7, file = "GLMdata_test_7.rds")
```

```{r}
GLMdata_test_7 <- readRDS("GLMdata_test_7.rds")

GLMdata_test_7 %>% pivot_wider(names_from = label, values_from = summary) -> GLMdata_test_pivot

GLMdata_test_pivot %>% ggplot(aes(x = POS, y = E_Bulk))  + geom_ribbon(aes(ymin = E_Bulk - 2*SE_Bulk, ymax = E_Bulk + 2*SE_Bulk, alpha = 0.1)) + geom_ribbon(aes(ymin = E_Parent - 2*SE_Parent, ymax = E_Parent + 2*SE_Parent, alpha = 0.1), fill = "orchid") +geom_ribbon(aes(ymin = E_Interaction - 2*SE_Interaction, ymax = E_Interaction + 2*SE_Interaction, alpha = 0.1), fill= "lightblue") + 
  geom_line(aes(y = E_Interaction), color = "lightblue") + 
geom_line(aes(y = E_Parent), color = "orchid") + geom_line()
```

Doing this with the other function

```{r, eval = FALSE}
#Specific to Fluconazole
rollData %>% filter(CHROM == "VII") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc4_unmixed(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "SE_intercept", "SE_Bulk", "SE_Parent", "SE_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction")) -> GLMdata_test2_7

saveRDS(GLMdata_test2_7, file = "GLMdata_test_72.rds")
```

```{r}
GLMdata_test2_7 <- readRDS("GLMdata_test_72.rds")

GLMdata_test2_7 %>% pivot_wider(names_from = label, values_from = summary) -> GLMdata_test2_pivot

GLMdata_test2_pivot %>% ggplot(aes(x = POS, y = E_Bulk))  + geom_ribbon(aes(ymin = E_Bulk - 2*SE_Bulk, ymax = E_Bulk + 2*SE_Bulk, alpha = 0.1)) + geom_ribbon(aes(ymin = E_Parent - 2*SE_Parent, ymax = E_Parent + 2*SE_Parent, alpha = 0.1), fill = "orchid") +geom_ribbon(aes(ymin = E_Interaction - 2*SE_Interaction, ymax = E_Interaction + 2*SE_Interaction, alpha = 0.1), fill= "lightblue") + 
  geom_line(aes(y = E_Interaction), color = "lightblue") + 
geom_line(aes(y = E_Parent), color = "orchid") + geom_line()
```

```{r, eval = FALSE}

rollData %>% filter(CHROM == "VII") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc4_withSD(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "SE_intercept", "SE_Bulk", "SE_Parent", "SE_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction", "Rep_SD")) -> GLMdata_test2_7rep

saveRDS(GLMdata_test2_7rep, file = "GLMdata_test2_7rep.rds")

```

```{r}
GLMdata_test2_7rep <- readRDS("GLMdata_test2_7rep.rds")

GLMdata_test2_7rep %>% pivot_wider(names_from = label, values_from = summary) -> GLMdata_test2_7rep_pivot

GLMdata_test2_7rep_pivot %>% ggplot(aes(x = POS, y = E_Bulk))  + geom_ribbon(aes(ymin = E_Bulk - 2*SE_Bulk, ymax = E_Bulk + 2*SE_Bulk, alpha = 0.1)) + geom_ribbon(aes(ymin = E_Parent - 2*SE_Parent, ymax = E_Parent + 2*SE_Parent, alpha = 0.1), fill = "orchid") +geom_ribbon(aes(ymin = E_Interaction - 2*SE_Interaction, ymax = E_Interaction + 2*SE_Interaction, alpha = 0.1), fill= "lightblue") + 
  geom_line(aes(y = E_Interaction), color = "lightblue") + 
geom_line(aes(y = E_Parent), color = "orchid") + geom_line()


GLMdata_test2_7rep %>% ggplot(aes(x = POS, y = summary, color = label)) + geom_line()

GLMdata_test2_7rep %>% pivot_wider(names_from = label, values_from = summary) %>%
  ggplot(aes(x = POS, y = Rep_SD)) + geom_line()
```

## Entire Genome

JK just pick a couple of chromosomes
```{r, eval = FALSE}
rollData <- readRDS("Fluc_rollData.rds")

rollData %>% filter(CHROM %in% c("I", "XIII", "XIV", "XV")) %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc4_withSD(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "SE_intercept", "SE_Bulk", "SE_Parent", "SE_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction", "Rep_SD")) -> GLMdata_fixedwithSD

saveRDS(GLMdata_fixedwithSD, file = "GLMdata_fixedwithSD.rds")
```

Do Chr III

```{r}
rollData <- readRDS("Fluc_rollData.rds")

rollData %>% filter(CHROM %in% c("V")) %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc4_withSD(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "SE_intercept", "SE_Bulk", "SE_Parent", "SE_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction", "Rep_SD")) -> GLMdata_fixedwithSD_V

saveRDS(GLMdata_fixedwithSD_V, file = "GLMdata_fixedwithSD_V.rds")
```

And now plot it
```{r}
#Load in all data with SD
GLMdata_fixedwithSD_V <- readRDS("GLMdata_fixedwithSD_V.rds")
GLMdata_fixedwithSD_III <- readRDS("GLMdata_fixedwithSD_III.rds")
GLMdata_test2_7rep <- readRDS("GLMdata_test2_7rep.rds")
GLMdata_fixedwithSD <- readRDS("GLMdata_fixedwithSD.rds")

#Combine data from different analysis runs
rbind(GLMdata_fixedwithSD, GLMdata_test2_7rep, GLMdata_fixedwithSD_III, GLMdata_fixedwithSD_V) %>% pivot_wider(names_from = label, values_from = summary) -> GLMdata_SD_Pivot

#Plot everything except Chr I
GLMdata_SD_Pivot %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = E_Bulk)) + 
  geom_ribbon(aes(ymin = E_Bulk - 2*SE_Bulk, ymax = E_Bulk + 2*SE_Bulk, alpha = 0.1)) + 
  geom_ribbon(aes(ymin = E_Parent - 2*SE_Parent, ymax = E_Parent + 2*SE_Parent, alpha = 0.1), fill = "orchid") +
  geom_ribbon(aes(ymin = E_Interaction - 2*SE_Interaction, ymax = E_Interaction + 2*SE_Interaction, alpha = 0.1), fill= "lightblue") + 
  geom_line(aes(y = E_Interaction), color = "lightblue") + 
  geom_line(aes(y = E_Parent), color = "orchid") + 
  geom_line() +
  facet_grid(~CHROM, scales = "free_x") 

#Plot only Chr
GLMdata_SD_Pivot %>% filter(CHROM == "I") %>% ggplot(aes(x = POS, y = E_Bulk))  + geom_ribbon(aes(ymin = E_Bulk - 2*SE_Bulk, ymax = E_Bulk + 2*SE_Bulk, alpha = 0.1)) + geom_ribbon(aes(ymin = E_Parent - 2*SE_Parent, ymax = E_Parent + 2*SE_Parent, alpha = 0.1), fill = "orchid") +geom_ribbon(aes(ymin = E_Interaction - 2*SE_Interaction, ymax = E_Interaction + 2*SE_Interaction, alpha = 0.1), fill= "lightblue") + 
  geom_line(aes(y = E_Interaction), color = "lightblue") + 
geom_line(aes(y = E_Parent), color = "orchid") + geom_line() +
  facet_grid(~CHROM, scales = "free") + ylim(-1000, 1000)

GLMdata_SD_Pivot %>% filter(CHROM == "I") %>% ggplot(aes(x = POS, y = E_Bulk))  + geom_ribbon(aes(ymin = E_Bulk - 2*SE_Bulk, ymax = E_Bulk + 2*SE_Bulk, alpha = 0.1)) + geom_ribbon(aes(ymin = E_Parent - 2*SE_Parent, ymax = E_Parent + 2*SE_Parent, alpha = 0.1), fill = "orchid") +geom_ribbon(aes(ymin = E_Interaction - 2*SE_Interaction, ymax = E_Interaction + 2*SE_Interaction, alpha = 0.1), fill= "lightblue") + 
  geom_line(aes(y = E_Interaction), color = "lightblue") + 
geom_line(aes(y = E_Parent), color = "orchid") + geom_line() +
  facet_grid(~CHROM, scales = "free") + ylim(-100, 100)


#Distribution of Effects
rbind(GLMdata_fixedwithSD, GLMdata_test2_7rep) %>% filter(label %in% c("E_Bulk", "E_Interaction", "E_intercept", "E_Parent")) %>% ggplot(aes(x = summary, color = label)) + geom_density() + facet_grid(~CHROM, scales = "free") + ggtitle("Effects per X by Chromosome")

#Distribution of Standard Errors
rbind(GLMdata_fixedwithSD, GLMdata_test2_7rep) %>% filter(label %in% c("SE_Bulk", "SE_Interaction", "SE_intercept", "SE_Parent")) %>% ggplot(aes(x = summary, color = label)) + geom_density() + facet_grid(~CHROM, scales = "free") + ggtitle("Standard Errors per X by Chromosome")

#Distribution of Z Scores
rbind(GLMdata_fixedwithSD, GLMdata_test2_7rep) %>% filter(label %in% c("Z_Bulk", "Z_Interaction", "Z_intercept", "Z_Parent")) %>% ggplot(aes(x = summary, color = label)) + geom_density() + facet_grid(~CHROM, scales = "free") + ggtitle("Z Scores per X by Chromosome")

#Distribution of Replicate SDs per chromosome
rbind(GLMdata_fixedwithSD, GLMdata_test2_7rep) %>% filter(label %in% c("Rep_SD")) %>% ggplot(aes(x = log(summary), color = CHROM)) + geom_density() + ggtitle("Replicate SDs per chromosome")

#Just Chr. I all
GLMdata_fixedwithSD %>% filter(CHROM == "I") %>% ggplot(aes(x = log(abs(summary)), color = label)) + geom_density()


```
Peak on Chr V
```{r}
GLMdata_fixedwithSD_V %>% mutate(abs_summary = abs(summary)) %>% filter(label == "Z_Bulk") %>% arrange(desc(abs_summary))

GLMdata_fixedwithSD_V %>% filter(label %in% c("E_intercept", "SE_intercept", "Z_intercept")==F) %>% ggplot(aes(x = POS, y = abs(summary), alpha = label == "Z_Bulk", color = label)) + geom_line() + geom_vline(xintercept = 116167)
#Bulk Effect = 373232
#Bulk Z scores = 373232

#Actual coordinates = 116167..116970

373232 - 116167

GLMdata_fixedwithSD_V %>% separate(label, sep = "_", into = c("Param", "Effect")) %>% filter(Param %in% c("E", "Rep")) %>% ggplot(aes(x = POS, y = summary, color = Effect, linetype = Param)) + geom_line()

  
```

Comparing how we do the rep thing

```{r}
GLMdata_2_VII$GLM <- "Chr7_1byRep"
GLMdata_test2_7rep$GLM <- "ChrVII_1byBulkbyParentbyRep"
rbind(GLMdata_2_VII,GLMdata_test2_7rep) -> Chr7_SDcomparison
Chr7_SDcomparison  %>% filter(label %in% c("RandSD", "Rep_SD")) %>% ggplot(aes(x = log(summary), linetype = GLM, color = GLM, fill = GLM)) + geom_density(alpha = 0.2, size = 2)

unique(Chr7_SDcomparison$label)
Chr7_SDcomparison  %>% filter(label %in% c("RandSD", "Rep_SD")) -> SDs

dim(GLMdata_2_VII) #only 30744 rows
dim(GLMdata_test2_7rep) #44408 rows

GLMdata_2_VII %>% filter(label %in% c("RandSD", "Rep_SD")) %>% ggplot(aes(x = summary)) + geom_density()
GLMdata_test2_7rep %>% filter(label %in% c("RandSD", "Rep_SD")) %>% ggplot(aes(x = summary)) + geom_density()

```

See if the newest way works

```{r}
#glm_mixed_model_Fluc5_div

cdata %>% group_by(Bulk, Parent) %>% summarize(Condition = paste(Bulk, Parent, sep="_")) -> testKEY

cdata %>% merge(testKEY) %>% select(-Reads) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> testcdata

testcdata$Allele <- factor(testcdata$Allele)

c <- glmer(Allele ~ Bulk*Parent+(1 | RepN), 
             weights = Reads, family = binomial, 
              data = testcdata)
a <- glmer(Allele ~ Bulk*Parent+(1 | Condition/RepN), 
             weights = Reads, family = binomial, 
              data = testcdata)

b <- glmer(Allele ~ Bulk*Parent+(1 +Condition | RepN), 
             weights = Reads, family = binomial, 
              data = testcdata)

d <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, data = testcdata)

e <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, data = testcdata)

summary(a)$coefficients #(1|Condition/Rep)
summary(b)$coefficients #(Condition | Rep)
summary(c)$coefficients #(1 | Rep)
summary(d)$coefficients # Bulk*Parent+Rep
summary(e)$coefficients # Bulk*Parent, no rep included


summary(b)$coefficients[1:12] #Effects
as.vector(attr(lme4::VarCorr(b)$R,"stddev"))
```

Trying this for something where the replicate *actually* has an effect

```{r}
Zeo <- readRDS("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023/Data/Zeocin_cybr2.rds")

Zeo %>% filter(CHROM == "XV") %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_", values_to = "Reads") -> Zeo_long

#Filtering for a specific point
Zeo_long %>% filter(POS > 8.65e+05 & POS < 8.75e+05) %>% filter(POS == 867041)

#Figuring out where the disparity is
Zeo_long %>% pivot_wider(names_from = Allele, values_from = Reads) %>% filter(Bulk != "NA") %>% mutate(odds = log(Oak/Wine)) %>% filter(POS > 8.65e+05 & POS < 8.75e+05) %>%
  ggplot(aes(x = POS, y = odds, color = paste(Bulk, Parent, Rep, sep = "_"))) + geom_line()

Zeo_long %>% filter(POS == 867041) -> zdata
```

```{r}
zdata %>% group_by(Bulk, Parent) %>% summarize(Condition = paste(Bulk, Parent, sep="_")) -> testKEYz

zdata %>% merge(testKEYz) %>% filter(Bulk != "NA") -> testzdata

testzdata$Allele <- factor(testzdata$Allele)

c <- glmer(Allele ~ Bulk*Parent+(1 | Rep), 
             weights = Reads, family = binomial, 
              data = testzdata)
a <- glmer(Allele ~ Bulk*Parent+(1 | Condition/Rep), 
             weights = Reads, family = binomial, 
              data = testzdata)

b <- glmer(Allele ~ Bulk*Parent+(1 +Condition | Rep), 
             weights = Reads, family = binomial, 
              data = testzdata)

d <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, data = testzdata)

e <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, data = testzdata)

summaries_zeo <- rbind(
  data.frame(summary(a)$coefficients, formula = "A (1|Condition/Rep)", param = rownames(summary(a)$coefficients)), #(1|Condition/Rep)
  data.frame(summary(b)$coefficients, formula = "B (Condition | Rep)", param = rownames(summary(b)$coefficients)), #(Condition | Rep)
  data.frame(summary(c)$coefficients, formula = "C (1 | Rep)", param = rownames(summary(c)$coefficients)), #(1 | Rep)
  data.frame(summary(d)$coefficients, formula = "D Bulk*Parent+Rep", param = rownames(summary(d)$coefficients)), # Bulk*Parent+Rep
  data.frame(summary(e)$coefficients, formula = "E Bulk*Parent, no rep included", param = rownames(summary(e)$coefficients))) # Bulk*Parent, no rep included

newkey <- data.frame(param = unique(summaries_zeo$param),
                     ParamKey = factor(c("Intercept", "Bulk", "Parent", "Interaction", "Rep"),
                                       levels = c("Bulk", "Parent", "Interaction", "Rep", "Intercept")))

summaries_zeo %>% merge(newkey) %>% ggplot(aes(x = ParamKey, y = abs(Estimate), color = formula)) + geom_point(size = 4, alpha = 0.5) +
  scale_color_manual(values = c("red", "orange", "gold", "gray", "lightblue")) + ggtitle("Zeocin Chr XV POS 867041")

lme4::VarCorr(a)
lme4::VarCorr(b)
lme4::VarCorr(c)
#lme4::VarCorr(d)
#lme4::VarCorr(e)
```

```{r}

testzdata %>% ggplot(aes(x = Condition, y = Reads, color = Allele, shape = Rep)) + geom_point(size = 4, alpha = 0.4) + ggtitle("Raw Counts Per Condition") + scale_color_manual(values = c("skyblue", "darkorange"))

testzdata %>% distinct() %>% pivot_wider(names_from = Allele, values_from = Reads) %>% 
  ggplot(aes(x = Bulk, y = log(Oak/Wine), color = Parent, shape = Rep)) + 
  geom_point(size = 4) + 
  scale_color_manual(values = c("navy", "firebrick")) + ggtitle("log(Oak/Wine) by Parent and Bulk")

#Let's see if we can make a plot just of the counts where it looks like the logistic regression to visualize
testzdata %>% distinct() -> reformattestz

counttestz <- data.frame(counts = rep(x = paste(reformattestz$Bulk, reformattestz$Parent, reformattestz$Rep, reformattestz$Allele, sep = "_"), 
    times = reformattestz$Reads))

counttestz %>% separate(counts, sep = "_", into = c("Bulk", "Parent", "Rep", "Allele")) -> cz
```

```{r, fig.width=3, fig.height=2.5}

#All factors
cz %>% ggplot(aes(y = Allele, x = Parent, color = Parent)) + 
  geom_jitter(alpha = 0.4, width = .4, height = 0.2) +
  scale_color_manual(values = c("navy", "firebrick")) + ggtitle("Zeocin Chr XV POS 867041") + facet_grid(Rep ~ Bulk)

#Not separating by Replicate
cz %>% ggplot(aes(y = Allele, x = Parent, color = Parent)) + 
  geom_jitter(alpha = 0.4, width = .4, height = 0.2) +
  scale_color_manual(values = c("navy", "firebrick")) + ggtitle("Zeocin Chr XV POS 867041") + facet_grid(~ Bulk)

#Not separating by Bulk
cz %>% ggplot(aes(y = Allele, x = Parent, color = Parent)) + 
  geom_jitter(alpha = 0.4, width = .4, height = 0.2) +
  scale_color_manual(values = c("navy", "firebrick")) + ggtitle("Zeocin Chr XV POS 867041")

#Not separating by Parent
cz %>% ggplot(aes(y = Allele, x = Bulk, color = Parent)) + 
  geom_jitter(alpha = 0.4, width = .4, height = 0.2) +
  scale_color_manual(values = c("navy", "firebrick")) + ggtitle("Zeocin Chr XV POS 867041") 

color <- c("steelblue", "orchid4")
```


## Making a new function that gets SD out


```{r}
glm_mixed_model_Fluc6_div <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(1, 1, 2,2,
                                     3,3,4, 4,
                                     5, 5, 6,6,
                                     7,7,8, 8,
                                     9, 9),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))
  
  combineddata %>% group_by(Bulk, Parent) %>% summarize(key = factor(paste(Bulk, Parent, sep = "_"))) -> combineddata

  b <- glmer(Allele ~ Bulk*Parent+(1+key | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  
  return(c(summary(b)$coefficients[1:12], #Effects
             as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))))
  
} 

summary(b)$coefficients
attr(lme4::VarCorr(b)$Rep,"stddev")


```
Conditions:
1. Intercept (must be Dilute Oak)
2. Dilute Wine
3. Zeocin Oak
4. Zeocin Wine

So if each of those has a SD, what is the variance of the Bulk vs Parent?

Bulk Zeocin
Parent Wine
Zeocin:Wine 

Trying this again in a different way

```{r}
#
glmer(Allele ~ Bulk*Parent+(1 | Rep/Condition), 
             weights = Reads, family = binomial, 
              data = testzdata)


glmer(Allele ~ Bulk*Parent+(Bulk*Parent | Rep), 
             weights = Reads, family = binomial, 
              data = testzdata)

glmer(Allele ~ Bulk*Parent+(Bulk | Rep) + (Parent | Rep), 
             weights = Reads, family = binomial, 
              data = testzdata)

```
## Finding multiple Zeocin datapoints that can be tested

```{r}
Zeo <- readRDS("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023/Data/Zeocin_cybr2.rds")

Zeo %>% filter(CHROM == "XV") %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_", values_to = "Reads") -> Zeo_long

#Figuring out where the disparity is
Zeo_long %>% filter(POS > 8e+05 & POS < 10e+05) %>% ggplot(aes(x = POS, y = Reads, color = paste(Parent, Rep, Allele))) + geom_point()

Zeo_long %>% filter(POS > 8.5e+05 & POS < 8.52e+05) %>% summarize(positions = unique(POS), cat = "HighRep") -> HighRepPositions

Zeo_long %>% filter(POS > 1.25e+05 & POS < 1.3e+05) %>% summarize(positions = unique(POS), cat = "LowRep") -> LowRepPositions


rbind(HighRepPositions, LowRepPositions) -> ZeocinRepTest

Zeo_long %>% filter(POS > 8e+05 & POS < 10e+05) %>% summarize(positions = unique(POS), cat = "EndBunch") -> EndBunchPositions
```

Looking at this for the bash script

```{r}
colnames(Zeo)
```

```{r}
for(i in 1:length(EndBunchPositions$positions)){
  Zeo_long %>% filter(POS == EndBunchPositions$positions[i]) -> zdata
  zdata %>% group_by(Bulk, Parent) %>% summarize(Condition = paste(Bulk, Parent, sep="_")) -> testKEYz

  zdata %>% merge(testKEYz) %>% filter(Bulk != "NA") -> testzdata

  testzdata$Allele <- factor(testzdata$Allele)
  
  c <- glmer(Allele ~ Bulk*Parent+(1 | Rep), 
               weights = Reads, family = binomial, 
                data = testzdata)
  a <- glmer(Allele ~ Bulk*Parent+(1 | Condition/Rep), 
               weights = Reads, family = binomial, 
                data = testzdata)
  
  b <- glmer(Allele ~ Bulk*Parent+(1 +Condition | Rep), 
               weights = Reads, family = binomial, 
                data = testzdata)
  
  d <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, data = testzdata)
  
  e <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, data = testzdata)

  summaries_zeo <- rbind(
    data.frame(summary(a)$coefficients, formula = "A (1|Condition/Rep)", param = rownames(summary(a)$coefficients)), #(1|Condition/Rep)
    data.frame(summary(b)$coefficients, formula = "B (Condition | Rep)", param = rownames(summary(b)$coefficients)), #(Condition | Rep)
    data.frame(summary(c)$coefficients, formula = "C (1 | Rep)", param = rownames(summary(c)$coefficients)), #(1 | Rep)
    data.frame(summary(d)$coefficients, formula = "D Bulk*Parent+Rep", param = rownames(summary(d)$coefficients)), # Bulk*Parent+Rep
    data.frame(summary(e)$coefficients, formula = "E Bulk*Parent, no rep included", param = rownames(summary(e)$coefficients))) # Bulk*Parent, no rep included
  
  newkey <- data.frame(param = unique(summaries_zeo$param),
                       ParamKey = factor(c("Intercept", "Bulk", "Parent", "Interaction", "Rep"),
                                         levels = c("Bulk", "Parent", "Interaction", "Rep", "Intercept")))

  if(i == 1){
    output <- summaries_zeo %>% merge(newkey) %>% mutate(POS = EndBunchPositions$positions[i])
  }else{
    tempoutput <- summaries_zeo %>% merge(newkey) %>% mutate(POS = EndBunchPositions$positions[i])
    output <- rbind(output, tempoutput)
  }
  
}

output %>% ggplot(aes(x = ParamKey, y = abs(Estimate), color = formula)) + geom_point(size = 4, alpha = 0.5) +
  scale_color_manual(values = c("red", "orange", "gold", "gray", "lightblue")) + theme(legend.position = "bottom")

#Filter all positions by replicate effect
output %>% filter(param == "RepB") %>% mutate(RepEffect = abs(Estimate) > 0.3) %>% select(POS, RepEffect) -> RepEffect

#Plot distributions of the replicate effects
output %>% filter(param == "RepB") %>% mutate(RepEffect = abs(Estimate) > 0.3) %>% 
  ggplot(aes(x = abs(Estimate), color = RepEffect, y = RepEffect)) + 
  scale_color_manual(values = c("orchid", "black")) + 
  geom_boxplot(size = 1) + 
  geom_boxplot(aes(x = abs(Estimate), y = "Combined"), color = "gray", linetype = "dashed", size = 1) + 
  ggtitle("Separation of Replicate Effects")

output %>% merge(RepEffect) %>% distinct() %>% ggplot(aes(x = abs(Estimate), y = Std..Error, color = formula)) + geom_point(alpha = 0.3) + facet_grid(~RepEffect)+ 
  scale_color_manual(values = c("red", "orange", "gold", "gray", "lightblue"))

output %>% merge(RepEffect) %>% distinct() %>% filter(ParamKey != "Rep") %>% ggplot(aes(x = RepEffect, y = Std..Error, color = formula)) + 
  geom_boxplot() + 
  scale_color_manual(values = c("red", "orange", "gold", "gray", "lightblue")) +
  facet_grid(~ParamKey) + theme(legend.position = "bottom")
```

```{r}
# output %>% merge(RepEffect) %>% ggplot(aes(x = ParamKey, y = abs(Estimate), color = formula)) + geom_point(size = 4, alpha = 0.5) +
#   scale_color_manual(values = c("red", "orange", "gold", "gray", "lightblue")) + theme(legend.position = "bottom") + facet_grid(~RepEffect)

output %>% merge(RepEffect) %>% filter(ParamKey != "Rep") %>% ggplot(aes(x = RepEffect, y = abs(Estimate), color = formula)) + geom_boxplot(size = 1, alpha = 0.5) +
  scale_color_manual(values = c("red", "orange", "gold", "gray", "lightblue")) + theme(legend.position = "bottom") + facet_grid(~ParamKey, scales = "free", space = "free") 

output %>% merge(RepEffect) %>% 
  filter(ParamKey %in% c("Bulk", "Parent", "Interaction")) %>% 
  filter(formula != "D Bulk*Parent+Rep") %>% filter(formula != "E Bulk*Parent, no rep included") %>% 
  ggplot(aes(x = ParamKey, y = abs(Estimate), color = RepEffect)) + 
  geom_boxplot(size = 1, alpha = 0.5) +
  scale_color_manual(values = c("orchid", "black")) + 
  theme(legend.position = "bottom") + facet_grid(~formula, scales = "free", space = "free") +
  ggtitle("Consequence of Rep Error (TRUE) on Effects")
```
## Consensus

The formula 1 + Condition/Rep (red above) adds the uncertainty of the replicate to the standard error, which can then be saved for all chromosomes. Use the following function:

Fluconazole model
```{r}
glm_mixed_model_Fluc6_div <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(1:18),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))
  
  combineddata %>% group_by(Bulk, Parent) %>% summarize(key = factor(paste(Bulk, Parent, sep = "_"))) %>% merge(combineddata) -> combineddata

  b <- glmer(Allele ~ Bulk*Parent+(1+key | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  
  return(c(summary(b)$coefficients[1:12]))
  
} 
```

Zeocin model
```{r}
glm_mixed_model_Zeo <- function(Dilute_Oak_A_Oak, Dilute_Oak_A_Wine, Dilute_Oak_B_Oak, Dilute_Oak_B_Wine, 
                                Zeocin_Oak_A_Oak, Zeocin_Oak_A_Wine,Zeocin_Oak_B_Oak, Zeocin_Oak_B_Wine,
                                Dilute_Wine_A_Oak ,Dilute_Wine_A_Wine, Dilute_Wine_B_Oak, Dilute_Wine_B_Wine,
                                Zeocin_Wine_A_Oak,Zeocin_Wine_A_Wine, Zeocin_Wine_B_Oak,Zeocin_Wine_B_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", 
                                             "H", "H","H", "H", 
                                             "L", "L","L", "L", 
                                             "H", "H","H", "H")),
                           Parent = factor(c("O", "O", "O", "O", 
                                             "O", "O", "O", "O",
                                             "W", "W", "W", "W", 
                                             "W", "W", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W",
                                             "O", "W", "O", "W", 
                                             "O", "W", "O", "W",
                                             "O", "W", "O", "W")),
                           Rep = factor(1:16),
                           Reads = c(Dilute_Oak_A_Oak, Dilute_Oak_A_Wine, Dilute_Oak_B_Oak, Dilute_Oak_B_Wine, 
                                Zeocin_Oak_A_Oak, Zeocin_Oak_A_Wine,Zeocin_Oak_B_Oak, Zeocin_Oak_B_Wine,
                                Dilute_Wine_A_Oak ,Dilute_Wine_A_Wine, Dilute_Wine_B_Oak, Dilute_Wine_B_Wine,
                                Zeocin_Wine_A_Oak,Zeocin_Wine_A_Wine, Zeocin_Wine_B_Oak,Zeocin_Wine_B_Wine))
  
  combineddata %>% group_by(Bulk, Parent) %>% summarize(key = factor(paste(Bulk, Parent, sep = "_"))) %>% merge(combineddata) -> combineddata

  b <- glmer(Allele ~ Bulk*Parent+(1+key | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  
  return(c(summary(b)$coefficients[1:12]))
  
} 

summary(b)$coefficients[1:12]

```

Test this on chr XV of Zeocin and chr VII of Fluconazole

```{r}
cybr2Data %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

saveRDS(rollData, file = "Fluc_rollData.rds")
```

```{r, eval=FALSE}
rd7 <- rollData %>% filter(CHROM == "VII")
#Additional Testing
D_OI_B_Oak = rd7$Dilute_OakI_B_Oak[1]
D_OI_B_Wine = rd7$Dilute_OakI_B_Wine[1]
D_WI_B_Oak = rd7$Dilute_WineI_B_Oak[1]
D_WI_B_Wine = rd7$Dilute_WineI_B_Wine[1]
D_OI_C_Oak = rd7$Dilute_WineI_C_Oak[1]
D_OI_C_Wine = rd7$Dilute_WineI_C_Wine[1]
D_OI_D_Oak = rd7$Dilute_WineI_D_Oak[1]
D_OI_D_Wine = rd7$Dilute_WineI_D_Wine[1]
F_OI_B_Oak = rd7$Fluconazole_OakI_B_Oak[1]
F_OI_B_Wine = rd7$Fluconazole_OakI_B_Wine[1]
F_OI_C_Oak = rd7$Fluconazole_OakI_C_Oak[1]
F_OI_C_Wine = rd7$Fluconazole_OakI_C_Wine[1]
F_OI_D_Oak = rd7$Fluconazole_OakI_D_Oak[1]
F_OI_D_Wine = rd7$Fluconazole_OakI_D_Wine[1]
F_WI_C_Oak = rd7$Fluconazole_WineI_C_Oak[1]
F_WI_C_Wine = rd7$Fluconazole_WineI_C_Wine[1]
F_WI_D_Oak = rd7$Fluconazole_WineI_D_Oak[1]
F_WI_D_Wine = rd7$Fluconazole_WineI_D_Wine[1]

```

```{r}
#Check how many this is
rollData %>% filter(CHROM == "VII") %>% filter(POS < 2e5) %>% dim()

#Run the function on that data (and next time, time it )
rollData %>% filter(CHROM == "VII") %>% filter(POS < 2e5) %>% group_by(CHROM, POS) %>% 
  summarise(summary = glm_mixed_model_Fluc6_div(D_OI_B_Oak = Dilute_OakI_B_Oak, 
                           D_OI_B_Wine = Dilute_OakI_B_Wine, 
                           D_WI_B_Oak = Dilute_WineI_B_Oak,
                           D_WI_B_Wine = Dilute_WineI_B_Wine,
                           
                           D_OI_C_Oak = Dilute_WineI_C_Oak,
                           D_OI_C_Wine = Dilute_WineI_C_Wine,
                           D_OI_D_Oak = Dilute_WineI_D_Oak, 
                           D_OI_D_Wine = Dilute_WineI_D_Wine,
                           
                           F_OI_B_Oak = Fluconazole_OakI_B_Oak, 
                           F_OI_B_Wine = Fluconazole_OakI_B_Wine, 
                           F_OI_C_Oak = Fluconazole_OakI_C_Oak,
                           F_OI_C_Wine = Fluconazole_OakI_C_Wine,
                           
                           F_OI_D_Oak = Fluconazole_OakI_D_Oak,
                           F_OI_D_Wine = Fluconazole_OakI_D_Wine,
                           F_WI_C_Oak = Fluconazole_WineI_C_Oak, 
                           F_WI_C_Wine = Fluconazole_WineI_C_Wine,
                           
                           F_WI_D_Oak = Fluconazole_WineI_D_Oak,
                           F_WI_D_Wine = Fluconazole_WineI_D_Wine),
                                                   label = c("E_intercept", "E_Bulk", "E_Parent", "E_Interaction",
                                                             "SE_intercept", "SE_Bulk", "SE_Parent", "SE_Interaction",
                                                             "Z_intercept", "Z_Bulk", "Z_Parent", "Z_Interaction")) -> GLMdata_F7

GLMdata_F7 %>% pivot_wider(names_from = label, values_from = summary) -> GLM_PivotW_F7

saveRDS(GLM_PivotW_F7, file = "GLM_PivotW_F7.rds")
```

```{r}
# GLM_PivotW_F7 %>% ggplot(aes(x = POS, y = E_Bulk)) + geom_line() + geom_ribbon(aes(ymin = E_Bulk - SE_Bulk, ymax = E_Bulk + SE_Bulk, alpha = 0.2)) +
#   geom_line(aes(x = POS, y = E_Parent), color = "violet") + 
#   geom_ribbon(aes(ymin = E_Parent - SE_Parent, ymax = E_Parent + SE_Parent, alpha = 0.2), fill = "violet") +
#   geom_line(aes(x = POS, y = E_Interaction), color = "lightblue") + 
#   geom_ribbon(aes(ymin = E_Interaction - SE_Interaction, ymax = E_Interaction + SE_Interaction, alpha = 0.2), fill = "lightblue")

#Look at the distribution of effects
GLM_PivotW_F7 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Param", "Effect"), names_sep = "_") %>% 
  filter(Effect != "intercept" & Param != "Z") %>% 
  ggplot(aes(x = value, color = Param)) + geom_density() + facet_grid(~Effect) + xlim(-10,10)

#GLM_PivotW_F7 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Param", "Effect"), names_sep = "_") %>% filter(abs(value) > 10)

GLM_PivotW_F7 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Param", "Effect"), names_sep = "_") %>% 
  pivot_wider(names_from = Param, values_from = value) %>% 
  ggplot(aes(x = E, y = SE, color = Effect)) + 
  geom_point(alpha = 0.5) + 
  geom_hline(yintercept = 25, linetype = "dashed") + ggtitle("No Filter")

GLM_PivotW_F7 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Param", "Effect"), names_sep = "_") %>% 
  pivot_wider(names_from = Param, values_from = value) %>% 
  ggplot(aes(x = E, y = SE, color = Effect)) + 
  geom_point(alpha = 0.5) + ylim(0, 25) + ggtitle("Filter == 25")


readRDS("C:/Users/cassa/Downloads/Fluconazole_ouput.rds")

```

```{r}
#Look at the results
GLM_PivotW_F7 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Param", "Effect"), names_sep = "_") %>% pivot_wider(names_from = Param, values_from = value) %>% filter(SE < 25 & Effect != "intercept") %>%
  ggplot(aes(x = POS, y = abs(E)))+
  geom_ribbon(aes(ymin = abs(E) - SE, 
                  ymax = abs(E) + SE, 
                  fill = Effect), alpha = 0.2) +
  geom_line(aes(color = Effect)) + xlim(0, max(rollData$POS))+
  ggtitle("Chr VII Fluconazole | Mixed Effect Model")
  
```
## Full Model

$$ y = \beta_{Bulk} + \beta_{Parent} + \beta_{Bulk*Parent} + \beta_{Day} + \epsilon_{rep} + \epsilon$$
So this would theoretically account for each day and also each replicate _within_ days. To write this would be something like the following:

```{r}
#Pseudocode formula
glmer(formula = Allele ~ Bulk*Parent + Day + (Day|Rep),
      weights = ReadCounts,
      data = mydata)

```

3D Plot (yet again)

```{r}
set.seed(417)
library(plotly)
temp <- rnorm(100, mean=30, sd=5)
pressure <- rnorm(100)
dtime <- 1:100

plot_ly(x=temp, y=pressure, z=dtime, type="scatter3d", mode="markers", color=temp)

#So let's make this for the axes of Allele (0 or 1), Bulk (0 or 1), and Parent (0 or 1) with a scatter for the number of points
threedtest <- readRDS("Fluc_rollData.rds")

threedtest %>% pivot_longer(cols = c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Replicate", "Allele"), names_sep = "_", values_to = "Reads") -> test3d

mytest <- subset(test3d, POS == 38108)

plot_ly(x = mytest$Bulk, z = mytest$Allele, y = mytest$Parent, size = mytest$Reads, 
        type = "scatter3d", shape = mytest$Replicate, color = mytest$Allele)
```

```{r}
#A different way to do this
#devtools::install_github("AckerDWM/gg3D")
library("gg3D")

ggplot(mytest, aes(x=Bulk, y=Parent, 
                   color=Allele, 
                   shape=Replicate, 
                   size = Reads)) + 
  geom_jitter(width = .2, height = .2, alpha = 0.6) +
  scale_color_manual(values = c("darkblue", "firebrick")) +
  theme_classic()
  #theme_void() +
  #axes_3D() +
  #stat_3D()
```
Another way to visualize is the log odds

```{r}
mytest2 %>% pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(logOdds = log((Wine+1)/(Oak+1))) -> mytest3

ggplot(mytest3, aes(x = Bulk, y = Parent, shape = Replicate, color = logOdds)) + 
  geom_jitter(size = 6, width = .2, height = .2) +
scale_colour_gradient2(low = muted("blue"),
  mid = "white",
  high = muted("red"),
  midpoint = 0,
  space = "Lab",
  na.value = "grey50",
  guide = "colourbar",
  aesthetics = "colour"
)

```


So to build  a model of the above:

```{r}
mytest$Bulk <- factor(mytest$Bulk)
mytest$Parent <- factor(mytest$Parent)
mytest$Replicate <- factor(mytest$Replicate)
mytest$Allele <- factor(mytest$Allele)

# mytest$Reads <- mytest$Reads + 1
# mytest2$Reads <- mytest2$Reads + 1

#Without including a replicate
glm(data = mytest,
      formula = Allele ~ Bulk*Parent, 
      weights = Reads,
      family = binomial)

#Using Replicate as a nested variable
glmer(data = mytest,
      formula = Allele ~ Bulk*Parent+(1|Replicate), 
      weights = Reads,
      family = binomial)

#Using flask as a block variable?
#mytest %>% mutate(Flask = paste(Bulk, Parent, sep = "_")) -> mytest2
glmer(data = mytest2,
      formula = Allele ~ Bulk*Parent+(1|Flask), 
      weights = Reads,
      family = binomial)
```

## Making a new function that gets SD out


```{r}
glm_mixed_model_Fluc6_div <- function(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine){
  
  combineddata <- data.frame(Bulk = factor(c("L", "L","L", "L", "L", "L","L", "L", 
                                             "H", "H","H", "H", "H", "H","H", "H", 
                                             "H", "H")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "O", "O",  
                                             "O", "O", "O", "O", "O", "O", "W", "W", 
                                             "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W")),
                           Rep = factor(1, 1, 2,2,
                                     3,3,4, 4,
                                     5, 5, 6,6,
                                     7,7,8, 8,
                                     9, 9),
                           Reads = c(D_OI_B_Oak, D_OI_B_Wine, D_WI_B_Oak,D_WI_B_Wine,
                                     D_OI_C_Oak,D_OI_C_Wine,D_OI_D_Oak, D_OI_D_Wine,
                                     F_OI_B_Oak, F_OI_B_Wine, F_OI_C_Oak,F_OI_C_Wine,
                                     F_OI_D_Oak,F_OI_D_Wine,F_WI_C_Oak, F_WI_C_Wine,
                                     F_WI_D_Oak, F_WI_D_Wine))
  
  combineddata %>% group_by(Bulk, Parent) %>% summarize(key = factor(paste(Bulk, Parent, sep = "_"))) -> combineddata

  b <- glmer(Allele ~ Bulk*Parent+(1+key | Rep), 
             weights = Reads, family = binomial, 
              data = combineddata)
  
  
  return(c(summary(b)$coefficients[1:12], #Effects
             as.vector(attr(lme4::VarCorr(b)$Rep,"stddev"))))
  
} 

summary(b)$coefficients
attr(lme4::VarCorr(b)$Rep,"stddev")


```
Conditions:
1. Intercept (must be Dilute Oak)
2. Dilute Wine
3. Zeocin Oak
4. Zeocin Wine

So if each of those has a SD, what is the variance of the Bulk vs Parent?

Bulk Zeocin
Parent Wine
Zeocin:Wine 

Trying this again in a different way

```{r}
#
glmer(Allele ~ Bulk*Parent+(1 | Rep/Condition), 
             weights = Reads, family = binomial, 
              data = testzdata)


glmer(Allele ~ Bulk*Parent+(Bulk*Parent | Rep), 
             weights = Reads, family = binomial, 
              data = testzdata)

glmer(Allele ~ Bulk*Parent+(Bulk | Rep) + (Parent | Rep), 
             weights = Reads, family = binomial, 
              data = testzdata)

```

## Finding multiple Zeocin datapoints that can be tested

```{r}
Zeo <- readRDS("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023/Data/Zeocin_cybr2.rds")

Zeo %>% filter(CHROM == "XV") %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_", values_to = "Reads") -> Zeo_long

#Figuring out where the disparity is
Zeo_long %>% filter(POS > 8e+05 & POS < 10e+05) %>% ggplot(aes(x = POS, y = Reads, color = paste(Parent, Rep, Allele))) + geom_point()

Zeo_long %>% filter(POS > 8.5e+05 & POS < 8.52e+05) %>% summarize(positions = unique(POS), cat = "HighRep") -> HighRepPositions

Zeo_long %>% filter(POS > 1.25e+05 & POS < 1.3e+05) %>% summarize(positions = unique(POS), cat = "LowRep") -> LowRepPositions


rbind(HighRepPositions, LowRepPositions) -> ZeocinRepTest

Zeo_long %>% filter(POS > 8e+05 & POS < 10e+05) %>% summarize(positions = unique(POS), cat = "EndBunch") -> EndBunchPositions
```


## I'm an idiot and CuSO4 Chr 8 is the one with weird Chr 12 so.... 

```{r, fig.height=6, fig.width=6}
# Cu <- readRDS("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023/Data/CuSO4_CSS8_cybr2.rds")
# 
# Cu %>% filter(CHROM == "XII") %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Dose", "Allele"), names_sep = "_", values_to = "Reads") %>% filter(Parent %in% c("Oak8", "Wine8"), Bulk != "Fluc") -> Cu_long
# 

Cu_long %>% ggplot(aes(x = POS, y = Reads, color = paste(Bulk, Parent, Dose), shape = Rep)) + geom_point(pch = 1, size = 2) +
  ggtitle("All Flasks Raw Read Counts")+  theme(axis.text.x=element_blank(),
            axis.ticks.x=element_blank())


Cu_long %>% pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(LOD = log(Wine/Oak)) %>%
  ggplot(aes(x = POS, y = LOD, color = paste(Bulk, Parent, Dose), shape = Rep)) + geom_point(pch = 1, size = 2) +
  ggtitle("All Flasks log(Wine/Oak)")+  theme(axis.text.x=element_blank(),
            axis.ticks.x=element_blank())


#Just plotting everything
Cu_long %>% filter(Bulk == "CuSO4", Parent == "Oak8", Dose == "C5") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(LOD = log(Wine/Oak)) %>%
  select(CHROM, POS, Bulk, Parent, Dose, LOD, Rep) %>% pivot_wider(names_from = Rep, values_from = LOD) %>% mutate(difference = abs(A - B)) %>%
  ggplot(aes(x = POS, y = difference, color = paste(Bulk, Parent, Dose))) + geom_point(pch = 1, size = 2) + geom_vline(xintercept = 5.6e5) +
  ggtitle("Oak8 C5 Difference in LOD between Replicates")+  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())


Cu_long %>% filter(Bulk == "CuSO4", Parent == "Wine8", Dose == "C6") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(LOD = log(Wine/Oak)) %>%
  select(CHROM, POS, Bulk, Parent, Dose, LOD, Rep) %>% pivot_wider(names_from = Rep, values_from = LOD) %>% mutate(difference = abs(A - B)) %>%
  ggplot(aes(x = POS, y = difference, color = paste(Bulk, Parent, Dose))) + geom_point(pch = 1, size = 2) + geom_vline(xintercept = 5.6e5) +
  ggtitle("Oak8 C5 Difference in LOD between Replicates") + scale_color_manual(values = "forestgreen")+  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())


#Plot coverage by Flask - doesn't look weird at all
Cu_long %>% filter(Bulk == "CuSO4", Parent == "Oak8", Dose == "C5") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(TotalReads = Oak+Wine) %>%
  select(CHROM, POS, Bulk, Parent, Dose, TotalReads, Rep) %>%
  ggplot(aes(x = POS, y = TotalReads, color = paste(Bulk, Parent, Dose))) + geom_point(pch = 1, size = 2)+ geom_vline(xintercept = 5.6e5) +
  ggtitle("Coverage of Oak8 C5")+  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())


#Coverage by replicate? basically no difference
Cu_long %>% filter(Bulk == "CuSO4", Parent == "Oak8", Dose == "C5") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(TotalReads = Oak+Wine) %>%
  select(CHROM, POS, Bulk, Parent, Dose, TotalReads, Rep) %>% 
  pivot_wider(names_from = Rep, values_from = TotalReads) %>% mutate(CoverageRatio = log(A/B)) %>%
  ggplot(aes(x = POS, y = abs(CoverageRatio), color = paste(Bulk, Parent, Dose))) + geom_point(pch = 1, size = 2)+ geom_vline(xintercept = 5.6e5) +
  ggtitle("Ratio of A and B Replicate Coverage (abs)")+  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())


#What do the actual counts look like just within this bulk?
Cu_long %>% filter(Bulk == "CuSO4", Parent == "Oak8", Dose == "C5") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(LOD = log(Oak/Wine)) %>%
  ggplot(aes(x = POS, y = abs(LOD), color = Rep)) + geom_point(pch = 1, size = 2)+ geom_vline(xintercept = 5.6e5) +
  ggtitle("Oak8 C5 Flask Oak/Wine LODs")+  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())


Cu_long %>% filter(Bulk == "CuSO4", Parent == "Wine8", Dose == "C6") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(LOD = log(Oak/Wine)) %>%
  ggplot(aes(x = POS, y = abs(LOD), color = Rep)) + geom_point(pch = 1, size = 2)+ geom_vline(xintercept = 5.6e5) +
  ggtitle("Wine8 C6 Flask Oak/Wine LODs") +  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

Cu_long %>% filter(Bulk == "CuSO4", Parent == "Oak8", Dose == "C5") %>%
  ggplot(aes(x = POS, y = Reads, color = paste(Rep,Allele))) + geom_point(pch = 1, size = 2) + geom_vline(xintercept = 5.6e5) +
  ggtitle("Oak8 C5 Flask Reads") + scale_color_manual(values = c("firebrick", "salmon", "#014D4E", "turquoise"))+  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())



Cu_long %>% filter(Bulk == "CuSO4", Parent == "Wine8", Dose == "C6") %>%
  ggplot(aes(x = POS, y = Reads, color = paste(Rep,Allele))) + geom_point(pch = 1, size = 2) + geom_vline(xintercept = 5.6e5) +
  ggtitle("Wine8 C6 Flask Reads") + scale_color_manual(values = c("firebrick", "salmon", "#014D4E", "turquoise"))+  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())


```



