---
title: "CSS Sequence Analysis Template"
date: "Jan 2023"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

#library(cybrBSA)

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)) + theme(legend.position = "bottom", #axis.text.x=element_blank(),
            axis.ticks.x=element_blank()))
################################################################################
glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
#glm with replicates
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

cybr2_rollmean <- function(dataframe){
  dataframe %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollmean(value, n = 100))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount)
}

################################################################################
#Define triple replicate function
glmfixed_rep3 <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                         HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", 
                                             "O", "O", "W", "W", "O", "O", "W", "W",
                                             "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", 
                                          "B", "B", "B", "B","B", "B", "B", "B",
                                          "C", "C", "C", "C", "C", "C","C", "C")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                                     HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                                     HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

```

# GATK Table to Data Frame

```{r}
setwd("../../../../GitHub/Sequencing/Analysis/")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) %>% filter(Unique == 1) -> pSNPs

#CHANGE THIS
#mydatatotest = "Data/HKTFTDRX2.SortedCat.vcf.output.table"
mydatatotest = "HGVMVDRX2.SortedCat.vcf.output.table"


FilteredData <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = pSNPs, yeast = TRUE) %>% 
  na.omit()

#Using Gsub for this
# gsub(FilteredData$Dataset, "HNGLVDRXY_n01_", "") #CHANGE THIS
FilteredData %>% mutate(DS1 = gsub("HGVMVDRX2_n01_", "",Dataset)) %>% mutate(DS = gsub(".fastq", "",DS1)) %>%  select(-Dataset, -DS1) -> tempFilteredData

unique(tempFilteredData$DS)

tempFilteredData$Bulk <- NA
tempFilteredData$Parent <- NA
tempFilteredData$Rep <- NA

tempFilteredData$Bulk[grep("Fluc", tempFilteredData$DS)] <- "Fluconazole" #CHANGE THIS
tempFilteredData$Bulk[grep("Dilute", tempFilteredData$DS)] <- "Dilute"

tempFilteredData$Rep[grep("A", tempFilteredData$DS)] <- "A"
tempFilteredData$Rep[grep("B", tempFilteredData$DS)] <- "B"
tempFilteredData$Rep[grep("C", tempFilteredData$DS)] <- "C"
tempFilteredData$Rep[grep("e_D", tempFilteredData$DS)] <- "D"
tempFilteredData$Rep[grep("c_D", tempFilteredData$DS)] <- "D"

tempFilteredData$Parent[grep("OakI", tempFilteredData$DS)] <- "OakI"
tempFilteredData$Parent[grep("WineI", tempFilteredData$DS)] <- "WineI"

tempFilteredData$ReadCount <- as.numeric(tempFilteredData$ReadCount)
  
# #THIS IGNORES REPLICATES
tempFilteredData %>% select(CHROM, POS, PAllele, ReadCount, Bulk, Parent, Rep) %>% distinct() -> Prepivot
 
Prepivot %>% pivot_wider(names_from = c(Bulk, Parent, Rep, PAllele), values_from = ReadCount) -> cybr2Data

saveRDS(object = cybr2Data, file = "Data/Fluconazole_1_cybr2.rds")

```




## Back to normal stuff
Smooth by rolling mean or median

```{r}
#Use rolling average of 100 SNPs, finding the mean
#cybr2Data %>% cybr2_rollmean() -> rollmeanData

#Find the rolling median or change n instead
cybr2Data %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

```

For Fluconazole ONLY, choose the replicates with highest coverage to compare:
```{r}
str(rollData)
as.data.frame(colSums(rollData[3:20])) %>% arrange(-colSums(rollData[3:20]))
```

Calculate GLM from rolled data

```{r, warning=FALSE}
#Change for different datasets
mydata <- rollData

#Run GLM - options are glmfixed_rep(), glmfixed_rep3(), and glmfixed()
#IF NOT USING REPS, change 1:5 to 1:4, and remove "Rep" from labels
mydata %>% group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed(HOO = Fluconazole_OakI_B_Oak, 
                           HOW = Fluconazole_OakI_B_Wine, 
                           HWO = Fluconazole_WineI_C_Oak,
                           HWW = Fluconazole_WineI_C_Wine,
                           LOO = Dilute_OakI_B_Oak,
                           LOW = Dilute_OakI_B_Wine,
                           LWO = Dilute_WineI_B_Oak, 
                           LWW = Dilute_WineI_B_Wine)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> GLMdata

#saveRDS(GLMdata, file = "Data/2023/PLACEHOLDER_glm.Rdata")
saveRDS(GLMdata, file = "Data/Fluconazole_1_glm.rds")

```

## Permute? This part does not work: 

```{r}
cybr2Data <- readRDS("Data/Fluconazole_1_cybr2.rds")
```

The flasks themselves have to be permuted rather than the points, since the coverage is different between bulks.

```{r}

glmfixed_permute <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           #Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Parent = factor(c(sample(c("O", "O", "W", "W", "O", "O", "W", "W")))),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}
```

## Permuting by flask

```{r}
rollData %>% head(1) %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_") %>% 
  pivot_wider(names_from = Allele, values_from = value) %>% mutate(NewRep = c(LETTERS[1:9])) %>% 
  group_by(Bulk) %>% mutate(NewParent = sample(Parent)) %>%  
  select(-Parent, -Rep) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele") %>%
  pivot_wider(names_from = c(Bulk, NewParent, NewRep, Allele), values_from = value) -> permutationcolumns

mydata <- rollData
colnames(mydata) <- colnames(permutationcolumns)

```

New function to just use the names of the columns; runs a LOT slower than it used to, but it's probably fine...?

```{r}
#First, define the inside of the function to make sure it works
data = mydata[1,]
columns = 4

#Copy paste inside of the function (between the ### lines)
data %>% pivot_longer(-c(CHROM, POS), 
                        names_to = c(LETTERS[1:columns]), 
                        names_sep = "_") %>% 
    mutate_if(is.character, as.factor) -> newdata
  
result <- glm(formula = as.formula(paste0("D ~ A*B+", LETTERS[3:(columns-1)])), 
      data = newdata, 
      weights = value, 
      family = binomial)

#Run this part to make sure that you know what the labels and the stats should be
label = rep(names(result$coefficients),3)
stat = colnames(summary(result)$coefficients)[1:3]

```


```{r}

glm_flex <- function(data, # = mydata[1,], 
                     columns = 4){
  ##############################################################################
  data %>% pivot_longer(-c(CHROM, POS), 
                        names_to = c(LETTERS[1:columns]), 
                        names_sep = "_") %>% 
    mutate_if(is.character, as.factor) -> newdata
  
  result <- glm(formula = as.formula(paste0("D ~ A*B+", LETTERS[3:(columns-1)])), 
        data = newdata, 
        weights = value, 
        family = binomial)
  ##############################################################################
  
  return(summary(result)$coefficients[1:(length(summary(result)$coefficients)*0.75)])
  
  #for knowing the column names run this once and print
  #label = rep(names(result$coefficients),3)
  #stat = colnames(summary(result)$coefficients)[1:3]
}

mydata %>% filter(CHROM == "II") %>% head(10) %>% group_by(CHROM, POS) %>% summarize(summary = glm_flex(.),
                                                     label = rep(names(result$coefficients),3),
                                                     stat = c(rep("Estimate", length(names(result$coefficients))),
                                                              rep("SE",  length(names(result$coefficients))),
                                                              rep("ZScore", length(names(result$coefficients))))
) -> glmflexresults_test

saveRDS(glmflexresults_II, file = "Fluc_glmflexresults_II.rds")
```


## New Functions with a different format perhaps

```{r}
data %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_") %>% mutate_if(is.character, as.factor) -> datap

glmflex_p <- function(B, P, R, A, W, ...,
    #B = datap$Bulk, P = datap$Parent, R = datap$Rep, A = datap$Allele, W = datap$value, ...,
                      myformula = "A ~ B*P+R"){
    arg_names <- as.list(match.call(expand.dots = FALSE)[-1])

    b <- glm(as.formula(myformula), weights = W, family = binomial)
  
  return(summary(b)$coefficients[1:((length(summary(b)$coefficients)*0.75))])
}

#Define the names after making sure the formula is right
mycoefficients <- names((glm(as.formula(myformula), #data = 
                             weights = W, family = binomial))$coefficients) #change the data to just the first group of your input data

#Run the function
datap %>% group_by(CHROM, POS) %>% summarize(summary = glmflex_p(Bulk, Parent, Rep, Allele, value),
                                             label = c(rep("Effect", 6),
                                                       rep("SE", 6),
                                                       rep("Zscore", 6)),
                                             coefficients = rep(mycoefficients, 3))

```

### ChatGPT idea

```{r}
#Fuck I hate that it worked
myglm <- function(..., W, formula) {
  data <- list(...)
  
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  
  glm_formula <- as.formula(formula)
  
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  return(summary(glm_fit)$coefficients[1:((length(summary(glm_fit)$coefficients)*0.75))])
  #return(glm_fit)
}

#Test on a single row
myglm(B = datap$Bulk, P = datap$Parent, R = datap$Rep, A = datap$Allele,
          W = datap$value, 
          formula = "A ~ B*P + R")

#Run using summarize
datap %>% group_by(CHROM, POS) %>% summarize(summary = myglm(Bulk = Bulk, Parent = Parent, Rep = Rep, Allele = Allele, W = value,
                                                             formula = "Allele ~ Bulk * Parent + Rep"),
                                             label = c(rep("Effect", 6),
                                                       rep("SE", 6),
                                                       rep("Zscore", 6)),
                                             coefficients = rep(mycoefficients, 3))


```

Trying this on the whole thing

```{r}
mydata %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_") %>% mutate_if(is.character, as.factor) -> mydatap

#Define the names after making sure the formula is right
myformula <- "A ~ B*P+R"
mycoefficients <- names((glm(as.formula(myformula), weights = W, family = binomial, data = subset(mydatap, CHROM == "I" & POS == 36313)))$coefficients) #change the data to just the first group of your input data

#Run the function
mydatap %>% group_by(CHROM, POS) %>% summarize(summary = glmflex_p(Bulk, Parent, Rep, Allele, value),
                                             label = c(rep("Effect", 6),
                                                       rep("SE", 6),
                                                       rep("Zscore", 6)),
                                             coefficients = rep(mycoefficients, 3)) -> flex_glm_output

saveRDS(flex_glm_output, file = "flex_glm_output_Fluc.rds")

```

Checking the plot

```{r}
flex_glm_output %>% pivot_wider(names_from = label, values_from = summary) %>% filter(coefficients != "(Intercept)", CHROM != "I") %>%
  ggplot(aes(x = POS, y = Effect, color = coefficients)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")
```
Let's try actually changing the formulas to whatever we want...

```{r}
mydatap %>% group_by(CHROM, POS) %>% summarize(summary = glmflex_p(Bulk = Bulk, Parent = Parent, Rep = Rep, Allele = Allele, 
                                                                   W = value,
                                                                   myformula = "Allele ~ Bulk*Parent + Rep")) -> flex_glm_test

glmflex_p(Bulk = datap$Bulk, Parent = datap$Parent, Rep = datap$Rep, Allele = datap$Allele,
          W = datap$value, 
          myformula = "Allele ~ Bulk*Parent + Rep")
```
## Permutations

Now that we have a formula that actually works, let's do the permutation part

```{r}
rollData %>% head(1) %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_") %>% 
  pivot_wider(names_from = Allele, values_from = value) %>% mutate(NewRep = c(LETTERS[1:9])) %>% 
  group_by(Bulk) %>% mutate(NewParent = sample(Parent)) %>%  
  select(-Parent, -Rep) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele") %>%
  pivot_wider(names_from = c(Bulk, NewParent, NewRep, Allele), values_from = value) -> permutationcolumns

mydata <- rollData
colnames(mydata) <- colnames(permutationcolumns)

#Run the original thing
myformula = "Allele ~ Bulk * Parent + Rep"
mycoefficients <- rownames(summary(glm(as.formula(myformula), 
                             weights = W, family = binomial, 
                             data = subset(md, CHROM == "I" & POS == 36313)))$coefficients) #change the data to just the first group of your input data

glm_fit<-summary((glm(as.formula(myformula), 
                             weights = W, family = binomial, 
                             data = subset(md, CHROM == "I" & POS == 36313))))

#Summarize
mydata %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_") %>% mutate_if(is.character, as.factor) -> md

md %>% filter(CHROM == "I" & POS == 36313) %>%  group_by(CHROM, POS) %>% summarize(summary = myglm(Bulk = Bulk, Parent = Parent, Rep = Rep, Allele = Allele, W = value,
                                                             formula = "Allele ~ Bulk * Parent + Rep"),
                                             label = c(rep("Effect", length(mycoefficients)),
                                                       rep("SE", length(mycoefficients)),
                                                       rep("Zscore", length(mycoefficients))),
                                             coefficients = rep(mycoefficients, 3))

md %>% filter(CHROM == "I" & POS == 36313) %>%  group_by(CHROM, POS) %>% summarize(summary = myglm(Bulk = Bulk, Parent = Parent, Rep = Rep, Allele = Allele, W = value,
                                                             formula = "Allele ~ Bulk * Parent + Rep"))

#OKAY NOW READY TO RUN THE WHOLE THING
mydata %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_") %>% mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% summarize(summary = myglm(Bulk = Bulk, Parent = Parent, Rep = Rep, Allele = Allele, W = value,
                                                             formula = "Allele ~ Bulk * Parent + Rep"),
                                             label = c(rep("Effect", length(mycoefficients)),
                                                       rep("SE", length(mycoefficients)),
                                                       rep("Zscore", length(mycoefficients))),
                                             coefficients = rep(mycoefficients, 3)) -> testpermutation

saveRDS(testpermutation, file = "testperm_fluc.RDS")

testpermutation %>% pivot_wider(names_from = label, values_from = summary) %>% filter(coefficients != "(Intercept)", CHROM != "I") %>%
  ggplot(aes(x = POS, y = Effect, color = coefficients)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")

```

