---
title: "Comparing Selection and all CuSO4 Experiments"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
library(fuzzyjoin)

require(BiocManager)
require(IRanges)

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")

# BiocManager::install("IRanges")

# ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)) +
#                      theme(#axis.title.x=element_blank(),
#         axis.text.x=element_blank(),
#         axis.ticks.x=element_blank()))

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))



## Functions for GLM all and Permutation all

glm_cb2_all <- function(..., W, formula, numgroups = FALSE, outputlength = 8) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  output <- summary(glm_fit)$coefficients[c(((length(summary(glm_fit)$coefficients)*0)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.25)),
                                            ((length(summary(glm_fit)$coefficients)*0.5)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.75)))]

  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
} 


################################################################################
# No replicates, fix Parent
cybrPermute_cb2_all <- function(dataset, 
                                R = 0, perp = 1, outlength = 4, 
                                glmform = "Allele ~ Bulk * Parent", 
                                inputlabels = c("Intercept", "Bulk", "Parent", "Interaction")){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled aD")
  print(paste("Your labels are:", inputlabels))

    if(R > 0){ #INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Rep, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == 1) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest    
    }else{ #DO NOT INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == perp) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    }
  
    #PERMUTE TWICE
  if(R > 0){
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Rep, 
             Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep, 
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, 
             Rep, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep,
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }else{
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #RUN THE GLM
    if(R > 0) {
      shuffletoglm2 %>% na.omit() %>%
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }else{
      shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }
    
    
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}


```

# Load in pre-glm data

```{r, eval = FALSE}
HTG3TDMXY_prep <- readRDS("StandardGLM/HTG3TDMXY_CSS8_glm_prep.rds") %>% 
  mutate(DoseCol = Bulk) %>%
  select(-Bulk) %>% 
  select(DoseCol = DoseCol, Bulk = B, Allele = Allele, POS = POS, Parent = Parent, SmoothCount = SmoothCount, CHROM = CHROM, Rep = Rep, Pool = Pool) %>% na.omit() %>% 
  filter(CHROM != "VIII", CHROM != "M")

unique(HTG3TDMXY_prep$Bulk)
unique(HTG3TDMXY_prep$DoseCol)

HTG3TDMXY_prep %>% mutate(Dose = gsub("51", "", DoseCol),
                          Dose = gsub("52", "", Dose),
                          Dose = paste(Parent, Dose)) -> temp

temp$Dose[temp$Dose == "O8 3"] <- "Low"
temp$Dose[temp$Dose == "W8 3"] <- "High"
temp$Dose[temp$Dose == "O8 4"] <- "High"
temp$Dose[temp$Dose == "W8 2"] <- "Low"

temp %>% filter(Dose != "Low") -> HTG3TDMXY_prep_highdose
temp %>% filter(Dose != "High") -> HTG3TDMXY_prep_lowdose

################################################################################

HVYTYDRX2_prep <- readRDS("StandardGLM/HVYTYDRX2_CuSO4_CSS8_glm_prep.rds")  %>% unite(c(Parent, Dose), col = "PD") %>% filter(PD != "O8_5") %>% separate(PD, into = c("Parent", "Dose"), sep = "_") %>%
  mutate(DoseCol = 0) %>%
  select(DoseCol = DoseCol, Bulk = Bulk, Allele = Allele, POS = POS, Parent = Parent, 
         SmoothCount = SmoothCount, CHROM = CHROM, Rep = Rep, Pool = Pool) %>% 
  filter(CHROM != "VIII", CHROM != "M") 

################################################################################

HKTFTDRX2_prep <- readRDS("StandardGLM/HKTFTDRX2_CSS1_glm_prep.rds") %>% 
  mutate(DoseCol = 0) %>%
  select(DoseCol = DoseCol, Bulk = Bulk, Allele = Allele, POS = POS, Parent = Parent, SmoothCount = SmoothCount, CHROM = CHROM, Rep = Rep, Pool = Pool) %>% na.omit() %>% 
  filter(CHROM != "I", CHROM != "M") 

################################################################################

HNGLVDRXY_prep <- readRDS("StandardGLM/HNGLVDRXY_CSS1_glm_prep.rds") %>% 
  mutate(DoseCol = 0) %>%
  select(DoseCol = DoseCol, Bulk = Bulk, Allele = Allele, POS = POS, Parent = Parent, SmoothCount = SmoothCount, CHROM = CHROM, Rep = 0 , Pool = Pool) %>% na.omit() %>% 
  filter(CHROM != "I", CHROM != "M") 

inputlabels <- c("cept", "Bulk", "Parent", "Interaction")

```

```{r}
HTG3TDMXY_prep_highdose %>% select(-Dose) %>% mutate(DoseCol = 0) %>% rbind(HVYTYDRX2_prep)  -> CSS8_CuSO4
HNGLVDRXY_prep %>% mutate(Rep = "n") %>% rbind(HKTFTDRX2_prep)-> CSS1_CuSO4

CSS1_CuSO4 %>% mutate(Bulk = gsub("Selected", "C", Bulk),
                      Bulk = gsub("Unselected", "aD", Bulk),
                      Parent = gsub("C", "W", Parent),
                      Parent = gsub("A", "O", Parent)) %>%
  group_by(Allele, POS, CHROM, Bulk, Parent) %>% 
  pivot_wider(names_from = c(Pool, Rep),
                           values_from = SmoothCount) %>%
  unnest(starts_with("H")) #458k results

CSS1_CuSO4 %>% mutate(Bulk = gsub("Selected", "C", Bulk),
                      Bulk = gsub("Unselected", "aD", Bulk),
                      Parent = gsub("C", "W", Parent),
                      Parent = gsub("A", "O", Parent)) %>%
  group_by(Allele, POS, CHROM, Bulk, Parent) %>% 
  pivot_wider(names_from = c(Pool, Rep),
                           values_from = SmoothCount) %>%
  unnest(starts_with("H")) %>% na.omit() -> AllBulks_CSS1 #214k results


AllBulks_CSS1 %>% group_by(POS, CHROM) %>% count() %>% ungroup() %>% count(n)

CSS1_CuSO4 %>% group_by(POS, CHROM) %>% select(POS, CHROM, Pool) %>% distinct() %>% count() %>% filter(n > 1)  -> bothPools_CSS1
CSS1_CuSO4 %>% group_by(POS, CHROM) %>% select(POS, CHROM, Pool) %>% distinct() %>% count() %>% filter(n == 1)  -> noPools_CSS1

#57887 with n = 1 or 2
#42815 with n = 2
#15072 of just n = 1

CSS8_CuSO4 %>% group_by(POS, CHROM) %>% select(POS, CHROM, Pool) %>% distinct() %>% count() %>% filter(n > 1)  -> bothPools_CSS8

CSS1_CuSO4 %>% filter(POS == 30032, CHROM == "XIV") %>% mutate(Bulk = gsub("Selected", "C", Bulk),
                                                               Bulk = gsub("Unselected", "aD", Bulk),
                                                               Parent = gsub("C", "W", Parent),
                                                               Parent = gsub("A", "O", Parent)) %>% mutate_if(is.character,as.factor) -> testcase

CSS1_CuSO4 %>% filter(CHROM == "XV") %>% filter(POS == 331119) %>% mutate(Bulk = gsub("Selected", "C", Bulk),
                                                               Bulk = gsub("Unselected", "aD", Bulk),
                                                               Parent = gsub("C", "W", Parent),
                                                               Parent = gsub("A", "O", Parent)) %>% mutate_if(is.character,as.factor) -> testcase
```

```{r}
require(lme4)

contrasts(testcase$Bulk)
contrasts(testcase$Parent)
contrasts(testcase$Rep)

# table(testcase$Bulk,testcase$Parent, testcase$Pool, testcase$Allele, testcase$Rep)
# table(testcase$Allele, testcase$Rep)

model_glm <- glm(formula = "Allele ~ Bulk*Parent", 
      family = binomial,
      weights = SmoothCount,
      data = testcase)

model_glmer_pool <- glmer(formula = "Allele ~ Bulk + Parent + Bulk*Parent + (1 | Pool)", 
      family = binomial,
      weights = SmoothCount,
      data = testcase)


model_glmer_rep <- glmer(formula = "Allele ~ Bulk + Parent + Bulk*Parent + (1 | Pool) + (1 | Pool:Rep)", 
      family = binomial,
      weights = SmoothCount,
      data = testcase)

model_glmer_both <- glmer(formula = "Allele ~ Bulk + Parent + Bulk*Parent + (1 | Pool:Rep)", 
      family = binomial,
      weights = SmoothCount,
      data = testcase)


print("model_glm")
summary(model_glm)
print("########################################################################")
print("model_glmer_pool")
summary(model_glmer_pool)
print("########################################################################")

print("model_glmer_rep")
summary(model_glmer_rep)

print("########################################################################")

print("model_glmer_both")
summary(model_glmer_both)

```


## Do the glm on these like normal

Maybe:  

Allele ~ Bulk + Parent + Bulk*Parent + (1 | Experiment) + (1 | Experiment:Day)

```{r, eval = FALSE}
inputlabels <- c("cept", "Bulk", "Parent", "Interaction")

CSS1_CuSO4 %>% na.omit() %>% 
      group_by(CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(
                                    Allele = Allele, Bulk = Bulk, Parent = Parent,#Rep = Rep,
                                    W = SmoothCount,
                                    formula = "Allele ~ Bulk * Parent",
                                    outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> CSS1_CuSO4_glm

CSS8_CuSO4 %>% na.omit() %>% 
      group_by(CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(
                                    Allele = Allele, Bulk = Bulk, Parent = Parent,#Rep = Rep,
                                    W = SmoothCount,
                                    formula = "Allele ~ Bulk * Parent",
                                    outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> CSS8_CuSO4_glm

saveRDS(CSS1_CuSO4_glm, file = "CSS1_CuSO4_glm.rds")
saveRDS(CSS8_CuSO4_glm, file = "CSS8_CuSO4_glm.rds")
```

```{r, fig.width=10, fig.height=5}
CSS8_CuSO4_glm %>% pivot_wider(names_from = d, values_from = Summary) %>%
  ggplot(aes(x = POS, y = Z, color = Factor)) + geom_line() + facet_grid(~CHROM, space = "free", scales = "free")

rbind(data.frame(CSS8_CuSO4_glm, CSS = "VIII"),
      data.frame(CSS1_CuSO4_glm, CSS = "I")) %>%
  filter(Factor != "cept") %>%
  pivot_wider(names_from = d, values_from = Summary) %>%
  ggplot(aes(x = POS, y = abs(Z), color = CSS)) + geom_line() + facet_grid(Factor~CHROM, space = "free", scales = "free") +
  scale_color_manual(values = c("blue","orange")) +
  theme(legend.position = "none")

rbind(data.frame(CSS8_CuSO4_glm, CSS = "VIII"),
      data.frame(CSS1_CuSO4_glm, CSS = "I")) %>%
  filter(Factor != "cept") %>%
  pivot_wider(names_from = d, values_from = Summary) %>%
  ggplot(aes(x = POS, y = abs(Effect), color = CSS)) + geom_line() + facet_grid(Factor~CHROM, space = "free", scales = "free") +
  scale_color_manual(values = c("blue","orange")) +
  theme(legend.position = "none")
```

Selecting only those that appear in both

```{r}
merge(bothPools_CSS1, CSS1_CuSO4_glm) %>%  group_by(CHROM) %>% count()

bothPools_CSS1 %>% group_by(CHROM) %>% count()

merge(bothPools_CSS1, CSS1_CuSO4_glm) %>% na.omit() %>% pivot_wider(names_from = d, values_from = Summary) -> pointsCSS1

CSS1_CuSO4_glm %>% filter(CHROM == "VII") %>% pivot_wider(names_from = d, values_from = Summary) %>%
  ggplot(aes(x = POS, y = abs(Z), color = Factor)) + 
  geom_vline(data = bothPools_CSS1[bothPools_CSS1$CHROM == "VII",], aes(xintercept = POS), size = 0.1, alpha = 0.1, color = "gray")+
  geom_point(alpha = 0.1) + 
  geom_point(data = pointsCSS1[pointsCSS1$CHROM == "VII",], aes(x = POS, y = abs(Z), color = Factor), size = 4, alpha = 0.1, shape = "triangle") +
  facet_grid(Factor~CHROM, space = "free", scales = "free") +
  theme(legend.position = "bottom")
```

## Permutations

```{r, eval = FALSE}

HTG3TDMXY_permutation <- cybrPermute_cb2_all(HTG3TDMXY_prep)


HTG3TDMXY_permutation %>% group_by(Factor) %>%
  summarize(quant095 = quantile(abs(Summary), 0.95, na.rm = TRUE),
             quant099 = quantile(abs(Summary), 0.99, na.rm = TRUE),
             quant0995 = quantile(abs(Summary), 0.995, na.rm = TRUE)) -> HTG3TDMXY_permutation_Qs


saveRDS(HTG3TDMXY_permutation_Qs, file = "HTG3TDMXY_permutation_Qs.rds")
```

## Do the GLMs and Permutations for EACH EXPERIMENT

```{r, eval = FALSE}
HVYTYDRX2_prep <- readRDS("StandardGLM/HVYTYDRX2_CuSO4_CSS8_glm_prep.rds") 

HVYTYDRX2_prep %>% unite(c(Parent, Dose), col = "PD") %>% filter(PD != "O8_5") %>% separate(PD, into = c("Parent", "Dose"), sep = "_") %>%
  mutate(DoseCol = 0) %>%
  select(DoseCol = DoseCol, Bulk = Bulk, Allele = Allele, POS = POS, Parent = Parent, 
         SmoothCount = SmoothCount, CHROM = CHROM, Rep = Rep, Pool = Pool) %>% 
  filter(CHROM != "VIII", CHROM != "M") %>% mutate_if(is.character,as.factor) -> test

test -> HVYTYDRX2_prep

table(test$Bulk, test$Parent)

inputlabels <- c("cept", "Bulk", "Parent", "Interaction")
HVYTYDRX2_prep %>% na.omit() %>% 
      group_by(Pool, CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(
                                    Allele = Allele, Bulk = Bulk, Parent = Parent, 
                                    W = SmoothCount,
                                    formula = "Allele ~ Bulk * Parent",
                                    outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> HVYTYDRX2_glm

saveRDS(HVYTYDRX2_glm, file = "HVYTYDRX2_glm_March24.rds")

HVYTYDRX2_glm %>% na.omit()

################################################################################
HVYTYDRX2_permutation <- cybrPermute_cb2_all(HVYTYDRX2_prep, perp = 2)


HVYTYDRX2_permutation %>% group_by(Factor) %>%
  summarize(quant095 = quantile(abs(Summary), 0.95, na.rm = TRUE),
             quant099 = quantile(abs(Summary), 0.99, na.rm = TRUE),
             quant0995 = quantile(abs(Summary), 0.995, na.rm = TRUE)) -> HVYTYDRX2_permutation_Qs


saveRDS(HVYTYDRX2_permutation_Qs, file = "HVYTYDRX2_permutation_Qs_M24.rds")
```

```{r, eval = FALSE}
HKTFTDRX2_prep <- readRDS("StandardGLM/HKTFTDRX2_CSS1_glm_prep.rds") %>% 
  mutate(DoseCol = 0) %>%
  select(DoseCol = DoseCol, Bulk = Bulk, Allele = Allele, POS = POS, Parent = Parent, SmoothCount = SmoothCount, CHROM = CHROM, Rep = Rep, Pool = Pool) %>% na.omit() %>% 
  filter(CHROM != "I", CHROM != "M") %>% mutate_if(is.character,as.factor)

inputlabels <- c("cept", "Bulk", "Parent", "Interaction")
HKTFTDRX2_prep %>% na.omit() %>% 
      group_by(Pool, CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(
                                    Allele = Allele, Bulk = Bulk, Parent = Parent, 
                                    W = SmoothCount,
                                    formula = "Allele ~ Bulk * Parent",
                                    outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> HKTFTDRX2_glm

saveRDS(HKTFTDRX2_glm, file = "HKTFTDRX2_glm_March24.rds")

################################################################################
HKTFTDRX2_permutation <- cybrPermute_cb2_all(HKTFTDRX2_prep)


HKTFTDRX2_permutation %>% group_by(Factor) %>%
  summarize(quant095 = quantile(abs(Summary), 0.95, na.rm = TRUE),
             quant099 = quantile(abs(Summary), 0.99, na.rm = TRUE),
             quant0995 = quantile(abs(Summary), 0.995, na.rm = TRUE)) -> HKTFTDRX2_permutation_Qs


saveRDS(HKTFTDRX2_permutation_Qs, file = "HKTFTDRX2_permutation_Qs_M24.rds")
```

```{r, eval = FALSE}
HNGLVDRXY_prep <- readRDS("StandardGLM/HNGLVDRXY_CSS1_glm_prep.rds") %>% 
  mutate(DoseCol = 0) %>%
  select(DoseCol = DoseCol, Bulk = Bulk, Allele = Allele, POS = POS, Parent = Parent, SmoothCount = SmoothCount, CHROM = CHROM, Rep = 0 , Pool = Pool) %>% na.omit() %>% 
  filter(CHROM != "I", CHROM != "M") %>% mutate_if(is.character,as.factor)

inputlabels <- c("cept", "Bulk", "Parent", "Interaction")
HNGLVDRXY_prep %>% na.omit() %>% 
      group_by(Pool, CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(
                                    Allele = Allele, Bulk = Bulk, Parent = Parent, 
                                    W = SmoothCount,
                                    formula = "Allele ~ Bulk * Parent",
                                    outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> HNGLVDRXY_glm

saveRDS(HNGLVDRXY_glm, file = "HNGLVDRXY_glm_March24.rds")

################################################################################
HNGLVDRXY_prep %>% mutate(Bulk = gsub("Unselected", "aD", Bulk)) -> HNGLVDRXY_prep_new

HNGLVDRXY_permutation <- cybrPermute_cb2_all(HNGLVDRXY_prep_new)

HNGLVDRXY_permutation %>% group_by(Factor) %>%
  summarize(quant095 = quantile(abs(Summary), 0.95, na.rm = TRUE),
             quant099 = quantile(abs(Summary), 0.99, na.rm = TRUE),
             quant0995 = quantile(abs(Summary), 0.995, na.rm = TRUE)) -> HNGLVDRXY_permutation_Qs


saveRDS(HNGLVDRXY_permutation_Qs, file = "HNGLVDRXY_permutation_Qs_M24.rds")
```


# Load in pre-run data
```{r}
HTG3TDMXY_permutation_Qs <- readRDS("HTG3TDMXY_permutation_Qs.rds")

HNGLVDRXY_permutation_Qs <- readRDS("HNGLVDRXY_permutation_Qs_M24.rds")
HKTFTDRX2_permutation_Qs <- readRDS("HKTFTDRX2_permutation_Qs_M24.rds")
HVYTYDRX2_permutation_Qs <- readRDS("HVYTYDRX2_permutation_Qs_M24.rds")

```

```{r}

HTG3TDMXY_glm_lowdose <- readRDS("HTG3TDMXY_glm_lowdose.rds")
HTG3TDMXY_glm_highdose <- readRDS("HTG3TDMXY_glm_highdose.rds")

HNGLVDRXY_glm <- readRDS("HNGLVDRXY_glm_March24.rds") 
HKTFTDRX2_glm <- readRDS("HKTFTDRX2_glm_March24.rds")
HVYTYDRX2_glm <- readRDS("HVYTYDRX2_glm_March24.rds")

HNGLVDRXY_glm %>% filter(d == "Z", Factor != "cept") %>% mutate(Pool = "HNGLVDRXY") %>% merge(HNGLVDRXY_permutation_Qs) -> HNGLVDRXY
HKTFTDRX2_glm %>% filter(d == "Z", Factor != "cept") %>% mutate(Pool = "HKTFTDRX2") %>% merge(HKTFTDRX2_permutation_Qs) -> HKTFTDRX2
HVYTYDRX2_glm %>% filter(d == "Z", Factor != "cept") %>% mutate(Pool = "HVYTYDRX2") %>% merge(HVYTYDRX2_permutation_Qs) -> HVYTYDRX2

rbind(HVYTYDRX2,
        HKTFTDRX2,
        HNGLVDRXY) %>%
  ungroup() %>% 
  group_by(Pool, CHROM, Factor, d) %>% 
  na.omit() %>%
  arrange(POS) %>%
summarize(POS = POS, Summary = frollapply(x = Summary, n = 100, FUN = mean, align = "left"),
          quant095 = quant095,
          quant099 = quant099,
          quant0995 = quant0995) %>% na.omit() %>%
   mutate(Dose = 0) %>%
  select(Factor,CHROM,Dose,d,POS,Summary,Pool,quant095,quant099,quant0995) -> Cu_Smoothed
  
```


### Dose Comparison Separation

```{r}
HTG3TDMXY_glm_lowdose %>% filter(d == "Z", Factor != "cept") %>% mutate(Dose = "Low") -> lowtemp
HTG3TDMXY_glm_highdose %>% filter(d == "Z", Factor != "cept") %>% mutate(Dose = "High") %>% rbind(lowtemp) -> DoseComparison

#smooth this to get rid of the weird ones?
DoseComparison %>% 
  ungroup() %>% 
  group_by(CHROM, Factor, Dose, d) %>% 
  na.omit() %>%
  arrange(POS) %>%
summarize(POS = POS, Summary = frollapply(x = Summary, n = 100, FUN = mean, align = "left")) %>% na.omit() %>% arrange(POS) -> DoseComp

DoseComp %>% mutate(Pool = "HTG3TDMXY") %>% merge(HTG3TDMXY_permutation_Qs) %>% 
   rbind(Cu_Smoothed) -> CuSO4_All

```

# Calling Peaks


```{r}
CuSO4_All %>% group_by(Pool, Dose, Factor, CHROM) %>% arrange(POS) %>%
  na.omit() %>%
  mutate(is_above_threshold = abs(Summary) > quant095) %>%
  
  mutate(segment_change = is_above_threshold != lag(is_above_threshold, default = FALSE)) %>%
  mutate(segment_id = cumsum(segment_change)) %>%
  # Group by chromosome and segment_id to count the points in each segment
  group_by(Pool, Dose, Factor, CHROM, segment_id) %>%
  mutate(points_in_segment = n()) %>%
  # Now filter to keep only those segments that are above the threshold and have 5 or more points
  filter(is_above_threshold, points_in_segment >= 100) %>%
 # mutate(change = is_above_threshold != lag(is_above_threshold, default = first(is_above_threshold))) %>%
  #mutate(group_id = cumsum(change)) %>%
  #group_by(Pool, Factor, CHROM, d, group_id) %>%
 # filter(is_above_threshold) %>%
  summarize(max_value = max(abs(Summary), na.rm = TRUE),
            ppeakPOS = POS[which.max(abs(Summary))]) %>%
  ungroup() -> maybepeaks

maybepeaks %>% filter(CHROM == "VIII") %>% mutate(Dist = 212535 - ppeakPOS) %>% ungroup() %>% summarize(max = max(Dist),
                                                                                                        mean = mean(Dist),
                                                                                                        sd = sd(Dist)) -> CUP1distance
#19kb distance from nearest CUP1 start
#5kb sd in this distance
```


## Bulk Peaks

This plot shows the peaks that are between 50000bp from each other within both high and low doses, and can be called overlapping peaks that way. These are also based on peaks called by the entire region above the cutoff, which in particular affects Chr XII since there are two areas where peaks seem to occur. The peaks on Chr XIII don't seem to be called at all, probably because they're too far from one another.


```{r, fig.height=4, fig.width=10}
maybepeaks %>% filter(Factor == "Bulk") %>% select(-segment_id) %>% pivot_wider(names_from = c(Pool, Dose), values_from = max_value) -> bulkpeaks

for(i in unique(bulkpeaks$CHROM)){
  bulkpeaks %>% filter(CHROM == i) -> tempdata
  tempdist <- as.matrix(dist(tempdata$ppeakPOS, upper = FALSE, diag = FALSE))
  successes <- as.data.frame(which(tempdist < 19317, arr.ind = TRUE)) %>% filter(row != col)
  
  for(k in 1:(nrow(successes)/2)){
    if(k == 1){
      all <- c(tempdata$ppeakPOS[successes$row[k]], tempdata$ppeakPOS[successes$col[k]])
    }else{
      pairs <- c(tempdata$ppeakPOS[successes$row[k]], tempdata$ppeakPOS[successes$col[k]])
      all <- rbind(all, pairs)
    }
    
  }
}
#find the common rounding, or maybe just find if there are any within 19kb?

for(i in unique(bulkpeaks$CHROM)){
  bulkpeaks %>% filter(CHROM == i) -> tempdata
  tempdist <- (dist(tempdata$ppeakPOS, upper = FALSE, diag = FALSE))
  hc <- hclust(tempdist)
  groups <- cutree(hc, h = 19317)
  tempdata$groups <- groups
  if(i == unique(bulkpeaks$CHROM)[1]){
    groupbulks <- tempdata
  }else{
    groupbulks <- rbind(groupbulks, tempdata)
  }
}

groupbulks %>% ungroup() %>% group_by(CHROM, groups) %>% summarize(num = length(ppeakPOS),
                                                                   avgPOS = mean(ppeakPOS)) %>% merge(groupbulks) -> groupbulks

groupbulks %>% 
  pivot_longer(-c(CHROM, groups, num, avgPOS, Factor, ppeakPOS), names_to = c("Pool", "Dose"), names_sep = "_") %>% 
  na.omit() %>% ungroup() %>% group_by(CHROM) -> gb

CuSO4_All %>% filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) + 
  geom_vline(data = gb, aes(xintercept = avgPOS, color = paste(num, groups)), alpha = 0.4, size = 1) +
  geom_point(data = gb, aes(x = avgPOS, color = paste(num, groups), shape = as.factor(num), y = value)) +
  geom_line(aes(linetype = Dose), alpha = 0.8, color = "black") +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Bulk Comparison of Doses") +
  theme(legend.position = "none")

CuSO4_All %>% filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) + 
  geom_vline(data = gb[gb$num > 1,], aes(xintercept = avgPOS, color = as.factor(groups)), alpha = 0.4, size = 1) +
  geom_point(data = gb[gb$num > 1,], aes(x = avgPOS, color = as.factor(groups), shape = as.factor(num), y = value)) +
  geom_line(aes(linetype = Dose), alpha = 0.8, color = "black") +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Bulk Comparison of Doses; Overlapping Peaks") +
  theme(legend.position = "none")

CuSO4_All %>% filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) + 
  geom_vline(data = gb[gb$num == 1,], aes(xintercept = avgPOS, color = as.factor(groups)), alpha = 0.4, size = 1) +
  geom_point(data = gb[gb$num == 1,], aes(x = avgPOS, color = as.factor(groups), y = value)) +
  geom_line(aes(linetype = Dose), alpha = 0.8, color = "black") +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Bulk Comparison of Doses; Non-Overlapping Peaks") +
  theme(legend.position = "none")

```

```{r, fig.height=4, fig.width=10}
CuSO4_All %>% filter(Pool == "HTG3TDMXY") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) + 
  geom_vline(data = gb[gb$Pool == "HTG3TDMXY",], aes(xintercept = avgPOS, color = as.factor(num)), alpha = 0.5, size = 1) +
  #geom_point(data = ba, aes(x = ppeakPOS, y = value, color = Dose), alpha = 0.4, size = 4) +
  geom_line(aes(linetype = Dose), alpha = 0.8, color = "black") +
  facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("Bulk Comparison of Doses") +
  scale_color_manual(values = c("gray", "lightblue", "violet", "red"))
```

```{r, fig.height=4, fig.width=10}

maybepeaks %>% filter(Factor == "Bulk") %>% 
  select(-segment_id, -max_value) %>% mutate(peakish = 50000*round(ppeakPOS/50000, 0)) %>% 
  pivot_wider(names_from = c(Pool, Dose), values_from = ppeakPOS) -> bulkpeaksa

bulkpeaks %>% pivot_longer(-c(Factor, CHROM, ppeakPOS)) %>% na.omit() %>% separate(name, into = c("Pool", "Dose"), sep = "_") -> ba

bulkpeaksa %>% group_by(Factor, CHROM, peakish) %>% 
  mutate(avg = mean(c(HKTFTDRX2_0,
                      HNGLVDRXY_0,
                      HTG3TDMXY_High,
                      HTG3TDMXY_Low,
                      HVYTYDRX2_0), na.rm = TRUE)) %>% na.omit() -> bulkpeaksavg

CuSO4_All %>% filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) + 
  geom_vline(data = bulkpeaksavg, aes(xintercept = avg), color = "red3", alpha = 0.2, size = 2) +
  geom_point(data = ba, aes(x = ppeakPOS, y = value, color = Dose), alpha = 0.4, size = 4) +
  geom_line(aes(linetype = Dose), alpha = 0.8, color = "black") +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Bulk Comparison of Doses") 

```

## Interaction Peaks

This plot shows the peaks that are between 50000bp from each other within both high and low doses, and can be called overlapping peaks that way. These are also based on peaks called by the entire region above the cutoff, which in particular affects Chr XII since there are two areas where peaks seem to occur. The peaks on Chr XIII don't seem to be called at all, probably because they're too far from one another.


```{r, fig.height=4, fig.width=10}
maybepeaks %>% filter(Factor == "Interaction") %>% select(-segment_id) %>% pivot_wider(names_from = c(Pool, Dose), values_from = max_value) -> Interactionpeaks

#find the common rounding, or maybe just find if there are any within 19kb?

for(i in unique(Interactionpeaks$CHROM[Interactionpeaks$CHROM %in% c("VIII", "II") == FALSE])){
  Interactionpeaks %>% filter(CHROM == i) -> tempdata
  tempdist <- (dist(tempdata$ppeakPOS, upper = FALSE, diag = FALSE))
  hc <- hclust(tempdist)
  groups <- cutree(hc, h = 19317)
  tempdata$groups <- groups
  if(i == unique(Interactionpeaks$CHROM[Interactionpeaks$CHROM %in% c("VIII", "II") == FALSE])[1]){
    groupInteractions <- tempdata
  }else{
    groupInteractions <- rbind(groupInteractions, tempdata)
  }
}

groupInteractions %>% ungroup() %>% group_by(CHROM, groups) %>% summarize(num = length(ppeakPOS),
                                                                   avgPOS = mean(ppeakPOS)) %>% merge(groupInteractions) -> groupInteractions

groupInteractions %>% 
  pivot_longer(-c(CHROM, groups, num, avgPOS, Factor, ppeakPOS), names_to = c("Pool", "Dose"), names_sep = "_") %>% 
  na.omit() %>% ungroup() %>% group_by(CHROM) -> gI

CuSO4_All %>% filter(Factor == "Interaction") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) + 
  geom_vline(data = gI, aes(xintercept = avgPOS, color = paste(num, groups)), alpha = 0.4, size = 1) +
  geom_point(data = gI, aes(x = avgPOS, color = paste(num, groups), shape = as.factor(num), y = value)) +
  geom_line(aes(linetype = Dose), alpha = 0.8, color = "black") +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Interaction Comparison of Doses") +
  theme(legend.position = "none")

CuSO4_All %>% filter(Factor == "Interaction") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) + 
  geom_vline(data = gI[gI$num > 1,], aes(xintercept = avgPOS, color = as.factor(groups)), alpha = 0.4, size = 1) +
  geom_point(data = gI[gI$num > 1,], aes(x = avgPOS, color = as.factor(groups), shape = as.factor(num), y = value)) +
  geom_line(aes(linetype = Dose), alpha = 0.8, color = "black") +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Interaction Comparison of Doses; Overlapping Peaks") +
  theme(legend.position = "none")

CuSO4_All %>% filter(Factor == "Interaction") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095)) + 
  geom_vline(data = gI[gI$num == 1,], aes(xintercept = avgPOS, color = as.factor(groups)), alpha = 0.4, size = 1) +
  geom_point(data = gI[gI$num == 1,], aes(x = avgPOS, color = as.factor(groups), y = value)) +
  geom_line(aes(linetype = Dose), alpha = 0.8, color = "black") +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Interaction Comparison of Doses; Non-Overlapping Peaks") +
  theme(legend.position = "none")

```

# Combining peaks to call at once

```{r}
maybepeaks %>% select(-segment_id) %>% pivot_wider(names_from = c(Pool, Dose), values_from = max_value) -> allpeaks

#find the common rounding, or maybe just find if there are any within 19kb?

for(i in unique(allpeaks$CHROM)){
  allpeaks %>% filter(CHROM == i) -> tempdata
  tempdist <- (dist(tempdata$ppeakPOS, upper = FALSE, diag = FALSE))
  hc <- hclust(tempdist)
  groups <- cutree(hc, h = 24818)
  tempdata$groups <- groups
  if(i == unique(allpeaks$CHROM)[1]){
    groupInteractions <- tempdata
  }else{
    groupInteractions <- rbind(groupInteractions, tempdata)
  }
}

groupInteractions %>% ungroup() %>% group_by(CHROM, groups) %>% summarize(num = length(ppeakPOS),
                                                                   avgPOS = mean(ppeakPOS)) %>% merge(groupInteractions) -> groupPeaks

groupPeaks %>% 
  pivot_longer(-c(CHROM, groups, num, avgPOS, Factor, ppeakPOS), names_to = c("Pool", "Dose"), names_sep = "_") %>% 
  na.omit() %>% ungroup() %>% group_by(CHROM) -> gP
```

```{r, fig.width = 12, fig.height = 5}
ggplot2::theme_set(theme_light() + theme(axis.text.x = element_blank(), legend.position = "bottom") + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "gray70"),
        panel.border = element_rect(colour = "gray70", fill=NA, linewidth=0.4))+
    theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
    theme(text = element_text(size = 10)))

# length(unique(gP$groups))

CuSO4_All %>% 
  ggplot(aes(x = POS, y = abs(Summary))) + 
  #geom_hline(aes(yintercept = quant095, linetype = paste(Pool, Dose))) + 
  #single peaks
  geom_vline(data = gP[gP$num == 1,], aes(xintercept = avgPOS), color = "gray", alpha = 0.2, size = 1) +
  geom_point(data = gP[gP$num == 1,], aes(x = ppeakPOS, y = value), alpha = 0.4, color = "gray", shape = 17) +
  #shared peaks
  geom_vline(data = gP[gP$num > 1,], aes(xintercept = avgPOS, color = as.factor(groups)), alpha = 0.4, size = 1) +
  geom_point(data = gP[gP$num > 1,], aes(x = ppeakPOS, color = as.factor(groups), y = value), size = 2, alpha = 0.8) +
  
  geom_line(aes(linetype = paste(Pool, Dose)), alpha = 0.8, color = "black", show.legend=FALSE) +
  facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("Peaks within 24kb")+ 
  theme(legend.position = "none")
```

```{r, fig.width = 12, fig.height = 5}

CuSO4_All %>% filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095, linetype = paste(Pool, Dose))) + 
  #single peaks
  geom_vline(data = gP[gP$num == 1 & gP$Factor == "Bulk",], aes(xintercept = avgPOS), color = "gray", alpha = 0.2, size = 1) +
  geom_point(data = gP[gP$num == 1 & gP$Factor == "Bulk",], aes(x = ppeakPOS, y = value), alpha = 0.4, color = "gray", shape = 17) +
  #shared peaks
  geom_vline(data = gP[gP$num > 1  & gP$Factor == "Bulk",], aes(xintercept = avgPOS, color = as.factor(groups)), alpha = 0.4, size = 1) +
  geom_point(data = gP[gP$num > 1 & gP$Factor == "Bulk",], aes(x = ppeakPOS, color = as.factor(groups), y = value), size = 2, alpha = 0.8) +
  
  geom_line(aes(linetype = paste(Pool, Dose)), alpha = 0.8, color = "black", show.legend=FALSE) +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Bulk Peaks within 24kb")+ 
  theme(legend.position = "none")

CuSO4_All %>% filter(Factor == "Interaction") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095, linetype = paste(Pool, Dose))) + 
  #single peaks
  geom_vline(data = gP[gP$num == 1 & gP$Factor == "Interaction",], aes(xintercept = avgPOS), color = "gray", alpha = 0.2, size = 1) +
  geom_point(data = gP[gP$num == 1 & gP$Factor == "Interaction",], aes(x = ppeakPOS, y = value), alpha = 0.4, color = "gray", shape = 17) +
  #shared peaks
  geom_vline(data = gP[gP$num > 1  & gP$Factor == "Interaction",], aes(xintercept = avgPOS, color = as.factor(groups)), alpha = 0.4, size = 1) +
  geom_point(data = gP[gP$num > 1 & gP$Factor == "Interaction",], aes(x = ppeakPOS, color = as.factor(groups), y = value), size = 2, alpha = 0.8) +
  
  geom_line(aes(linetype = paste(Pool, Dose)), alpha = 0.8, color = "black", show.legend=FALSE) +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Interaction Peaks within 24kb")+ 
  theme(legend.position = "none")
```

# Finding the genes in those regions

```{r, fig.width = 12, fig.height = 5}
#genes <- c("CUP1-1", "CUP1-2", "FRE1")

sgd_orfs <- read.csv("SGD_ORFs.csv") %>% 
  transmute(Gene = Gene.symbol,
            CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier),
            POS.Start = Gene.chromosomeLocation.start,
            POS.End = Gene.chromosomeLocation.end)

gP %>% mutate(avgPOS.Start = avgPOS - 24000,
              avgPOS.End = avgPOS + 24000) -> gpmerge

sgd_orfs %>% group_by(Gene, CHROM, POS.Start, POS.End) %>% summarize(Start = min(POS.Start, POS.End),
                                                                     End = max(POS.Start, POS.End)) -> sgd_orfs

genes_in_peaks <- interval_join(gpmerge, sgd_orfs, by = c("avgPOS.Start" = "Start", "avgPOS.End" = "End")) %>% filter(CHROM.x == CHROM.y) %>% select(-CHROM.y, -avgPOS.Start, -avgPOS.End)

genes_in_peaks %>% select(CHROM = CHROM.x, Factor, POS.Start, POS.End, Gene) %>% distinct() -> toplot

toplot$label_y <- rank(toplot$Gene)

CuSO4_All %>% 
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_point(data = gP, aes(x = ppeakPOS, y = value), size = 2, alpha = 0.8) +
  geom_vline(data = toplot, aes(xintercept = POS.Start, color = Gene), alpha = 0.3) +
  geom_line(aes(linetype = paste(Pool, Dose)), alpha = 0.8, color = "black", show.legend=FALSE) +
  
  facet_grid(Factor~CHROM, space = "free", scales = "free") + ggtitle("Peaks within 24kb")+ 
  theme(legend.position = "none") 


```

# Comparing which peaks actually overlap

```{r, fig.width= 12, fig.height = 4}

gP %>% filter(num > 1) %>% group_by(Pool, Factor, Dose) %>% arrange(avgPOS) %>% summarize(CHROM = CHROM,
                                                                                    avgPOS = avgPOS,
                                                                                    
                                                                                    rank = dense_rank(value),
                                                                                    max = max(dense_rank(value))) -> ranks_gP

ranks_gP %>% merge(gP) %>% filter(Factor == "Bulk") %>% ggplot(aes(x = avgPOS, y = -(rank/max), color = paste(Pool, Dose), size = num)) + 
  geom_point(alpha = 0.5) + 
  facet_grid(Factor~CHROM, scales = "free", space = "free")

gP %>% group_by(Pool, Factor, Dose) %>% arrange(avgPOS) %>% summarize(CHROM = CHROM,
                                                                                    avgPOS = avgPOS,
                                                                                    
                                                                                    rank = dense_rank(value),
                                                                                    max = max(dense_rank(value))) -> ranks_gP

ranks_gP %>% merge(gP) %>% ggplot(aes(x = avgPOS, y = -(rank/max), color = paste(Pool, Dose), size = num)) + 
  geom_point(alpha = 0.5) + 
  facet_grid(Factor~CHROM, scales = "free", space = "free")

ranks_gP %>% merge(gP) -> addtoplot
```

```{r, fig.width= 12, fig.height = 4}
CuSO4_All %>% filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095, linetype = paste(Pool, Dose))) + 
  
  #shared peaks
  geom_vline(data = addtoplot[addtoplot$num > 1  & addtoplot$Factor == "Bulk",], 
             aes(xintercept = avgPOS, color = as.factor(groups)), 
             alpha = 0.4, size = 1) +
  geom_point(data = addtoplot[addtoplot$num > 1 & addtoplot$Factor == "Bulk",], 
             aes(x = ppeakPOS, color = as.factor(groups), y = value, size = (rank/max), shape = rank == max), 
             alpha = 0.4) +
  geom_line(aes(linetype = paste(Pool, Dose)), alpha = 0.8, color = "black", show.legend=FALSE) +
  facet_grid(Pool~CHROM, space = "free", scales = "free") + ggtitle("Bulk Peaks by rank")+ 
  theme(legend.position = "none") 

```

```{r, fig.width= 12, fig.height = 4}
Selection <- data.frame(Pool = c("HTG3TDMXY", "HKTFTDRX2", "HNGLVDRXY", "HVYTYDRX2"),
                        Dose2 = c("Multi", "Low", "High", "Low"),
                        CSS = c("VIII", "I", "I", "VIII"))

addtoplot %>% merge(Selection) -> addtoplot2
#unique(CuSO4_All_doses$Dose2)

CuSO4_All %>% merge(Selection) -> CuSO4_All_doses

CuSO4_All_doses$Dose2[CuSO4_All_doses$Dose2 == "Multi"] <- CuSO4_All_doses$Dose[CuSO4_All_doses$Dose2 == "Multi"]
addtoplot2$Dose2[addtoplot2$Dose2 == "Multi"] <- addtoplot2$Dose[addtoplot2$Dose2 == "Multi"]

#table(CuSO4_All_doses$Dose, CuSO4_All_doses$Dose2)

CuSO4_All_doses %>%
  filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095, linetype = paste(Pool, Dose))) + 
  
  #shared peaks
  geom_vline(data = addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Bulk",], 
             aes(xintercept = avgPOS, color = as.factor(groups)), 
             alpha = 0.4, size = 1) +
  geom_point(data = addtoplot2[addtoplot2$num > 1 & addtoplot2$Factor == "Bulk",], 
             aes(x = ppeakPOS, color = as.factor(groups), y = value, size = (rank/max), shape = rank == max), 
             alpha = 0.4) +
  geom_line(aes(linetype = paste(Pool, Dose)), alpha = 0.8, color = "black", show.legend=FALSE) +
  facet_grid(CSS~CHROM, space = "free", scales = "free") + ggtitle("CSS Comparison | Bulk Peaks by rank")+ 
  theme(legend.position = "none") 

CuSO4_All_doses %>%
  filter(Factor == "Bulk", CHROM != "VIII", CHROM != "I") %>%
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095, linetype = paste(Pool, Dose))) + 
  
  #shared peaks
  geom_vline(data = addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Bulk",], 
             aes(xintercept = avgPOS, color = as.factor(groups)), 
             alpha = 0.4, size = 1) +
  geom_point(data = addtoplot2[addtoplot2$num > 1 & addtoplot2$Factor == "Bulk",], 
             aes(x = ppeakPOS, color = as.factor(groups), y = value, size = (rank/max), shape = rank == max), 
             alpha = 0.4) +
  geom_line(aes(linetype = paste(Pool, Dose)), alpha = 0.8, color = "black", show.legend=FALSE) +
  facet_grid(Dose2~CHROM, space = "free", scales = "free") + ggtitle("Dose Comparison | Bulk Peaks by rank")+ 
  theme(legend.position = "none") 

```

```{r, fig.width=12, fig.height=5}
#addtoplot2 %>% mutate(FacetCol = paste(CSS, Dose2)) -> addtoplot2
CuSO4_All_doses %>%
  filter(Factor == "Bulk") %>%
  mutate(FacetCol = paste(CSS, Dose2)) %>% 
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_hline(aes(yintercept = quant095, linetype = paste(Pool, Dose))) + 
  
  #shared peaks
  geom_vline(data = addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Bulk",], 
             aes(xintercept = avgPOS, color = as.factor(groups)), 
             alpha = 0.4, size = 1) +
  geom_point(data = addtoplot2[addtoplot2$num > 1 & addtoplot2$Factor == "Bulk",], 
             aes(x = ppeakPOS, color = as.factor(groups), y = value, size = (rank/max), shape = rank == max), 
             alpha = 0.4) +
  geom_line(aes(linetype = paste(Pool, Dose)), alpha = 0.8, color = "black", show.legend=FALSE) +
  facet_grid(FacetCol~CHROM, space = "free", scales = "free") + ggtitle("Bulk Peaks by rank")+ 
  theme(legend.position = "none") 


```

### Rescaling to the highest peak or cutoff

```{r, fig.width=12, fig.height=4}
CuSO4_All_doses %>% filter() %>% group_by(Pool, Factor, Dose) %>% summarize(CHROM = CHROM, POS = POS, CSS = CSS, Dose2 = Dose2,
                                                               maxpeak = max(abs(Summary)),
                                                              prop = abs(Summary)/maxpeak) -> CuSO4_All_proportions

CuSO4_All_proportions %>% filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = prop, linetype = paste(Pool,Dose2))) + geom_line() + 
  geom_vline(data = addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Bulk",], 
             aes(xintercept = avgPOS, color = as.factor(groups)), 
             alpha = 0.4, size = 1) +
  facet_grid(CSS~CHROM, space = "free", scales = "free") +
  theme(legend.position = "none") 
```
```{r, fig.width=12, fig.height=4}
CuSO4_All_doses %>% mutate(Prop95 = log(abs(Summary)/quant095)) %>%

  filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = Prop95, linetype = Dose)) + geom_line() + 
  geom_vline(data = addtoplot2[addtoplot2$num == 1  & addtoplot2$Factor == "Bulk",], 
             aes(xintercept = avgPOS), color = "gray", 
             alpha = 0.4, size = 1) +
  geom_vline(data = addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Bulk",], 
             aes(xintercept = avgPOS, color = as.factor(groups)), 
             alpha = 0.4, size = 1) +
  ylim(0, 4) +
  facet_grid(paste(CSS,Dose2)~CHROM, space = "free", scales = "free") +
  theme(legend.position = "none") + ggtitle("Proportion of 95% Cutoff | Bulk")

CuSO4_All_doses %>% mutate(Prop95 = log(abs(Summary)/quant095)) %>%

  filter(Factor == "Interaction") %>%
  ggplot(aes(x = POS, y = Prop95, linetype = Dose)) + geom_line() + 
  geom_vline(data = addtoplot2[addtoplot2$num < 2  & addtoplot2$Factor == "Interaction",], 
             aes(xintercept = avgPOS), color = "gray", 
             alpha = 0.4, size = 1) +
  geom_vline(data = addtoplot2[addtoplot2$num > 2  & addtoplot2$Factor == "Interaction",], 
             aes(xintercept = avgPOS, color = as.factor(groups)), 
             alpha = 0.4, size = 1) +
  ylim(0,2) +
  facet_grid(paste(CSS,Dose2)~CHROM, space = "free", scales = "free") +
  theme(legend.position = "none") + ggtitle("Proportion of 95% Cutoff | Interaction")
```

## Figures for (maybe) the paper

### FIGURE 2 (start of the actual data)
```{r, fig.width=9, fig.height=4}
CuSO4_All_doses %>% filter() %>% group_by(Pool, Factor, Dose) %>% summarize(CHROM = CHROM, POS = POS, CSS = CSS, Dose2 = Dose2,
                                                               maxpeak = max(abs(Summary)),
                                                              prop = abs(Summary)/maxpeak, 
                                                              quant095 = quant095,
                                                              prop095 = quant095/maxpeak) -> CuSO4_All_proportions_c

CuSO4_All_proportions_c %>% ungroup() %>% 
  mutate(peak = as.numeric(prop > prop095)) %>%
  group_by(Factor, CSS, POS) %>% summarize(allpeak = mean(peak)) %>%
  merge(CuSO4_All_proportions_c) -> CuSO4_All_proportions_cP

CuSO4_All_proportions_c$CHROM <- factor(CuSO4_All_proportions_c$CHROM, levels = as.character(as.roman(1:16)))
CuSO4_All_proportions$CHROM <- factor(CuSO4_All_proportions$CHROM, levels = as.character(as.roman(1:16)))
CuSO4_All_proportions_cP$CHROM <- factor(CuSO4_All_proportions_cP$CHROM, levels = as.character(as.roman(1:16)))

################################################################################
#BULK EFFECTS
CuSO4_All_proportions_cP %>% filter(Factor == "Bulk") %>% ggplot(aes(x = POS, y = prop, 
                                                                    linetype = paste(Pool, Dose2),
                                                                    color = Dose2)) + 
  #peaks
  geom_vline(data = CuSO4_All_proportions_cP[CuSO4_All_proportions_cP$allpeak == 1 & CuSO4_All_proportions_cP$Factor == "Bulk",], 
             aes(xintercept = POS),
             color = "#ADD8E610") +
  
  #data
  geom_line() +
  geom_hline(aes(yintercept = prop095, 
                 linetype = paste(Pool, Dose2),
                 color = Dose2)) +
  
  #aesthetics of the rest
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  ylab("Proportion of Max Selection Peak") + xlab("Genomic Position") +
  theme(legend.position = "left") +
  scale_color_manual(values = c("black", "gray30"))

#INTERACTION EFFECTS
CuSO4_All %>% filter(Factor == "Interaction") %>% ggplot(aes(x = POS, y = abs(Summary), 
                                                                    linetype = paste(Pool, Dose))) + 
  #peaks
  geom_vline(data = CuSO4_All_proportions_cP[CuSO4_All_proportions_cP$allpeak == 1 & CuSO4_All_proportions_cP$Factor == "Bulk",], 
             aes(xintercept = POS),
             color = "#ADD8E610") +
  geom_vline(data =  CuSO4_All_proportions_cP[CuSO4_All_proportions_cP$allpeak == 1 & CuSO4_All_proportions_cP$Factor == "Interaction",],
             aes(xintercept = POS),
             color = "red") +
  
  #data
  geom_line() +
  geom_hline(aes(yintercept = quant095, 
                 linetype = paste(Pool, Dose))) +
 
  
  #aesthetics of the rest
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  ylab("abs(Z-score)") + xlab("Genomic Position") +
  theme(legend.position = "left") +
  scale_color_manual(values = c("black", "gray30"))

CuSO4_All_proportions_cP %>% filter(Factor == "Bulk") %>%
  ggplot(aes(x = POS, y = prop, color = paste(Pool,Dose2))) + geom_line() + 
  
  #Peaks
  geom_vline(data = addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Bulk" ,], 
             aes(xintercept = ppeakPOS, #avgPOS, #ppeakPOS, 
                 #color = as.factor(groups), 
                 color = paste(Pool,Dose2)), 
             #linetype = "dashed",
             alpha = 0.4, size = 1) +
  
  #Cutoffs
  geom_hline(aes(yintercept = prop095, 
                 color = paste(Pool, Dose2)),
                 linetype = "dashed") +
  facet_grid(CSS~CHROM, space = "free", scales = "free") +
  theme(legend.position = "none")  +
  scale_color_manual(values = c("black", "lightblue", "darkorange", "darkred", "black"))

```

```{r, fig.height=4, fig.width=9}
genes <- c("CUP1-1", 
           #"CUP1-2", 
           "FRE1")

sgd_cup1 <- read.csv("SGD_ORFs.csv") %>% 
  transmute(Gene = Gene.symbol,
            CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier),
            POS.Start = Gene.chromosomeLocation.start,
            POS.End = Gene.chromosomeLocation.end) %>%
  filter(Gene %in% genes)
################################################################################

CuSO4_All_doses %>% filter(Factor %in% c("Bulk", "Interaction")) %>% group_by(Pool, Dose) %>% summarize(Factor = Factor, CHROM = CHROM, POS = POS, CSS = CSS, Dose2 = Dose2,
                                                               maxpeak = max(abs(Summary)),
                                                              prop = abs(Summary)/maxpeak, 
                                                              quant095 = quant095,
                                                              prop095 = quant095/maxpeak) -> CuSO4_All_proportions_f


################################################################################
CuSO4_All_proportions_f %>% filter(Factor == "Bulk") %>%
  #mutate(cutoff = prop > prop095) %>% mutate(cutoff = gsub(FALSE, NA, cutoff)) %>%
 
  ggplot(aes(x = POS, y = prop, color = paste(Pool,Dose2),
             linetype = prop < prop095)) + 
  geom_vline(data = sgd_cup1, aes(xintercept = POS.Start), color = "red", size = 2, alpha = 0.4) +
  geom_line() + 
  geom_hline(aes(yintercept = prop095), color = "gray") +
  #Peaks
  # geom_vline(data = addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Bulk" ,], 
  #            aes(xintercept = ppeakPOS, #avgPOS, #ppeakPOS, 
  #                color = paste(Pool,Dose2)), 
  #            #linetype = "dashed",
  #            alpha = 0.4, size = 1) +
  facet_grid(CSS~CHROM, space = "free", scales = "free") +
theme(legend.position = "none",
        strip.text = element_text(colour = 'black'),
    panel.background = element_rect(fill = "white", color = "gray95"),
      panel.border = element_blank(), 
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "gray95", colour = "gray95", 
                linewidth = rel(1)), complete = TRUE) +
  scale_color_manual(values = rep("black", 12)) +
  ggtitle("Selection Effects Normalized to Highest Peak")

################################################################################

CuSO4_All_proportions_f %>% #filter(Factor == "Interaction") %>%
  #mutate(cutoff = prop > prop095) %>% mutate(cutoff = gsub(FALSE, NA, cutoff)) %>%
 
  ggplot(aes(x = POS, y = prop, size = paste(Pool,Dose2),
             color = paste(Factor, CSS),
             linetype = prop < prop095)) + 
  geom_line() + 
  geom_hline(aes(yintercept = prop095, color = paste(Factor, CSS))) +
  # Peaks
  geom_vline(data = addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Interaction" ,],
             aes(xintercept = ppeakPOS, #avgPOS, #ppeakPOS,
                 size = paste(Pool,Dose2),
                 color = CSS),
             alpha = 0.2, size = 2) +
  facet_grid(CSS~CHROM, space = "free", scales = "free") +
  theme(legend.position = "none",
        strip.text = element_text(colour = 'black'),
      panel.background = element_rect(fill = "white", color = "gray95"),
        panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        strip.background = element_rect(fill = "gray95", colour = "gray95", 
                  linewidth = rel(1)), complete = TRUE) +
  scale_color_manual(values = c("gray90", #bulk I
                                "gray90", #bulk VIII
                                "#43ACC7", #CSS I
                                "#43ACC7", #Interaction I
                                "darkorange", #Interaction VIII
                                "darkorange")) + #CSS VIII
  scale_size_manual(values = rep(0.5, 20)) +
  ggtitle("Interaction Effects Normalized to Highest Additive Peak")
  

```

Organizing rankings

```{r, fig.width=2, fig.height=4}
CuSO4_All_doses %>% filter(Factor == "Bulk", d == "Z") %>%
  group_by(Pool, CHROM) %>% 
  summarize(quant095 = quant095,
            CSS = CSS,
            max_value = max(abs(Summary), na.rm = TRUE),
            ppeakPOS = POS[which.max(abs(Summary))]) %>%
  ungroup() %>%
  distinct() %>%
  filter(max_value > quant095) %>%
  arrange(desc(max_value)) %>% mutate(rank = row_number()) %>%
  ggplot(aes(x = Pool,
             y = -rank,
             color = CHROM)) +
  geom_text(aes(label = CHROM)) +
  facet_grid(~CSS, scales = "free", space = "free") +
  theme_classic() +
  theme(legend.position = "none",
        panel.background = element_rect(fill = "white", color = "black"),
        panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        strip.background = element_rect(fill = "gray95", colour = "black", 
                  linewidth = rel(1)), complete = TRUE) 

################################################################################
CuSO4_All_doses %>% filter(Factor == "Bulk", d == "Z") %>%
  group_by(Pool, CHROM) %>% 
  summarize(quant095 = quant095,
            CSS = CSS,
            max_value = max(abs(Summary), na.rm = TRUE),
            ppeakPOS = POS[which.max(abs(Summary))]) %>%
  ungroup() %>%
  distinct() %>%
  filter(max_value > quant095) %>%
  select(Pool, CHROM, CSS, max_value) %>%
  arrange(desc(max_value)) %>%
  pivot_wider(names_from = Pool, values_from = max_value) -> intermediate

intermediate %>% select(CHROM, CSS, Val = HTG3TDMXY) %>% na.omit() %>% arrange(desc(Val)) %>% mutate(rank = row_number()) %>%  mutate(Pool = "HT") -> HTG3TDMXY_rank
intermediate %>% select(CHROM, CSS, Val = HKTFTDRX2) %>% na.omit() %>% arrange(desc(Val)) %>% mutate(rank = row_number()) %>%  mutate(Pool = "HK")  -> HKTFTDRX2_rank
intermediate %>% select(CHROM, CSS, Val = HNGLVDRXY) %>% na.omit() %>% arrange(desc(Val)) %>% mutate(rank = row_number()) %>%  mutate(Pool = "HN") -> HNGLVDRXY_rank
intermediate %>% select(CHROM, CSS, Val = HVYTYDRX2) %>% na.omit() %>% arrange(desc(Val)) %>% mutate(rank = row_number()) %>%  mutate(Pool = "HV") -> HVYTYDRX2_rank
  
rbind(HTG3TDMXY_rank, HKTFTDRX2_rank, HNGLVDRXY_rank, HVYTYDRX2_rank) %>%
  ggplot(aes(x = Pool, y = -rank, color = CHROM)) +
  geom_text(aes(label = CHROM)) +
  facet_grid(~CSS, scales = "free", space = "free") +
  theme_classic() +
  #ylim(-8,-1) +
  theme(legend.position = "none",
      strip.text = element_text(colour = 'black'),
        panel.background = element_rect(fill = "white", color = "black"),
        panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        strip.background = element_rect(fill = "gray95", colour = "black", 
                  linewidth = rel(1)), complete = TRUE) +
  scale_color_manual(values = c("gray","gray","gray","gray",
                                "gray", #9
                                "red","gray","darkorange","darkturquoise",
                                #skip to ten
                                "gray","lightblue","black",
                                "gray","darkred","orchid","gray"))

```

```{r, fig.width=9, fig.height=4}
CuSO4_All_doses %>% filter(Factor == "Bulk", d == "Z") %>%
  group_by(Pool, CHROM) %>% 
  summarize(quant095 = quant095,
            CSS = CSS,
            max_value = max(abs(Summary), na.rm = TRUE),
            ppeakPOS = POS[which.max(abs(Summary))]) %>%
  ungroup() %>%
  distinct() %>%
  filter(max_value > quant095) %>%
  filter(CHROM %in% c("VIII", "V", "XII", "XV","VII", "XI", "XIV")) -> newpeaks

CuSO4_All_proportions_f %>% filter(Dose != "Low") %>% ungroup() %>% select(CHROM, CSS, Pool, maxpeak) %>% distinct() %>% merge(newpeaks) %>% mutate(prop = max_value/maxpeak) -> npm

################################################################################
CuSO4_All_proportions_f %>% filter(Factor == "Bulk") %>%
  #mutate(cutoff = prop > prop095) %>% mutate(cutoff = gsub(FALSE, NA, cutoff)) %>%
 
  ggplot() + 
  geom_vline(data = sgd_cup1, aes(xintercept = POS.Start), color = "red", size = 2, alpha = 0.2) +
  geom_line(aes(x = POS, y = prop, size = paste(Pool,Dose2),
             linetype = prop < prop095)) + 
  geom_hline(aes(yintercept = prop095), color = "gray") +
  #Peaks
  geom_point(data = npm, aes(x = ppeakPOS, y = prop, color = CHROM), size = 3, alpha = 0.5) + 
  #the rest
  facet_grid(CSS~CHROM, space = "free", scales = "free") +
theme(legend.position = "none",
       strip.text = element_text(colour = 'black'),
     panel.background = element_rect(fill = "white", color = "gray95"),
      panel.border = element_blank(), 
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "gray95", colour = "gray95", 
                linewidth = rel(1)), complete = TRUE) +
  #scale_color_manual(values = rep("black", 12)) +
  scale_size_manual(values = rep(0.5, 12)) +
  ggtitle("Selection Effects Normalized to Highest Peak") +
  scale_color_manual(values = c("red","darkorange","darkturquoise","lightblue", "black","darkred", "orchid","gray"))
```

Let's try the circle plots again...?


```{r}
cybr_circos_new <- function(d1 = "markers", d8 = "markers", peaklist1 = NULL, peaklist8 = NULL, maxy = NULL, color1 = "#24588A", color8 = "#ED7B01", opacity = "50"){
  
  markers <- rbind(data.frame(CHROM = as.character(as.roman(1:16)),
                        POS = 1,
                        summary = 0,
                        label = "Interaction"),
                   data.frame(CHROM = as.character(as.roman(1:16)),
                        POS = (c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000),
                        summary = 0,
                        label = "Interaction")
  )
  
  color1fade_50 <- paste(color1, "50", sep = "")
  color8fade_50 <- paste(color8, "50", sep = "")
  
  color1fade <- paste(color1, opacity, sep = "")
  color8fade <- paste(color8, opacity, sep = "")

  if(is.vector(d8)){
    d8 <- markers
    include8 <- FALSE
  }else{
    d8 <- rbind(d8, markers)
    include8 <- TRUE
  }
  
  if(is.vector(d1)){
    d1 <- markers
    include1 <- FALSE
  }else{
    d1 <- rbind(d1, markers)
    include1 <- TRUE
  }
    #SET UP DATA
  df8 <- data.frame(sectors = as.character(d8$CHROM),
                 x = d8$POS,
                 y = abs(d8$summary),
                  label = d8$label)

  df1 <- data.frame(sectors = as.character(d1$CHROM),
                 x = d1$POS,
                 y = abs(d1$summary),
                  label = d1$label)

  #Take just the interactions of each
  df8int <- df8 %>% filter(label == "Interaction")
  df1int <- df1 %>% filter(label == "Interaction")

    #REORDER THE CHROMOSOMES
  df8int$sectors <- factor(df8int$sectors, levels = as.character(as.roman(1:16)))
  df1int$sectors <- factor(df1int$sectors, levels = as.character(as.roman(1:16)))
  
  ##############################################################################
  dfall <- rbind(df1int, df8int) %>% na.omit()
  circos.par("track.height" = 0.3, start.degree = 90, cell.padding = c(0,0))
  circos.initialize(sectors = dfall$sectors, x = dfall$x)

  if(is.null(maxy)){
    #I think this makes the sizes?
    circos.track(ylim = c(max(c(dfall$y)), 0), dfall$sectors, y = dfall$y, 
        panel.fun = function(x, y) {
            circos.text(CELL_META$xcenter, 
                0 - mm_y(5), 
                CELL_META$sector.index,
                niceFacing = FALSE)
            circos.axis(labels.cex = 0.1)
    })
  }else{
    circos.track(ylim = c(maxy, 0), dfall$sectors, y = dfall$y, 
      panel.fun = function(x, y) {
          circos.text(CELL_META$xcenter, 
              0 - mm_y(5), 
              CELL_META$sector.index,
              niceFacing = FALSE)
          circos.axis(labels.cex = 0.1)
      })
  }
  #Makes the chromosome overlap parts
  #CHROMOSOME I
  
  if(include1 == TRUE){
    #CHROMOSOME I
    draw.sector(83.5, #RIGHT
              90, #LEFT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color1fade_50, border = FALSE)
  }
  
  if(include8 == TRUE){
    #CHROMOSOME VIII
    draw.sector(289.5, #LEFT
              305.4, #RIGHT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color8fade_50, border = FALSE)
  }
  
  
  #Makes the lines
  circos.trackPoints(df8int$sectors, df8int$x, abs(df8int$y), col = color8, pch = 16, cex = 0.1)
  circos.trackPoints(df1int$sectors, df1int$x, abs(df1int$y), col = color1, pch = 16, cex = 0.1)
  
  if(is.null(peaklist8) == FALSE){
    if(length(peaklist8$POS) >= 1){
      for(i in 1:length(peaklist8$POS)){
      circos.link(peaklist8$CHROM[i], 
                  peaklist8$POS[i], 
                  #Add 8 after
                  "VIII", c(0, max(dfall$x[dfall$sectors == "VIII"])),  
                  
                  col = color8fade, 
                  h.ratio = 0.3, 
                  border = color8fade, 
                  lwd = 1)
      }
    }else{
      print("No interactions on Chr 8")
    }
    
  }

  if(is.null(peaklist1) == FALSE){
    if(length(peaklist1$POS) >= 1){
      for(i in 1:length(peaklist1$POS)){
        circos.link("I",c(0, max(dfall$x[dfall$sectors == "I"])),  
                    #add 1 first
                    peaklist1$CHROM[i], 
                    peaklist1$POS[i], 
                    col = color1fade, 
                    h.ratio = 0.3, 
                    border = color1fade, 
                    lwd = 1)
      }
    }else{
      print("No interactions on Chr 1")
    }
  }  
  
  circos.clear()
}
```

```{r}
for(i in unique(CuSO4_All_doses$Pool)){
  CuSO4_All_doses %>% filter(Pool == i) -> templines
  
  if(length(unique(templines$Dose)) > 1){
    templines %>% filter(Dose == "High") -> templines
    addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Interaction" ,] %>% filter(Dose == "High") -> gpt
  }else{
    gpt <- addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Interaction" ,]
  }
  print(i)
  if(unique(templines$CSS) == "I"){
    int1 <- gpt %>% filter(Pool == i, Factor == "Interaction") %>% transmute(CHROM = CHROM, POS = ppeakPOS)
    templines %>% transmute(CHROM = CHROM, POS = POS, summary = abs(Summary), label = Factor) -> d1cuso4

    cybr_circos_new(peaklist1 = int1,
            #peaklist8 = int8,
            d1 = d1cuso4,
            d8 = "m",
            color1 = "#43ACC7")
  }else{
    int8 <- gpt %>% filter(Pool == i, Factor == "Interaction") %>% transmute(CHROM = CHROM, POS = ppeakPOS)
    templines %>% transmute(CHROM = CHROM, POS = POS, summary = abs(Summary), label = Factor) -> d8cuso4

    cybr_circos_new(#peaklist1 = int1,
            peaklist8 = int8,
            #d1 = d1cuso4,
            d8 = d8cuso4,
            color1 = "#43ACC7")
  }
}
```


Maybe also showing where the peaks line up?

```{r, fig.height=4, fig.width=9}
CuSO4_All_proportions_f %>% filter(Factor == "Bulk") %>%
  #mutate(cutoff = prop > prop095) %>% mutate(cutoff = gsub(FALSE, NA, cutoff)) %>%
 
  ggplot() + 
  #change to interaction ones
  geom_vline(data = addtoplot2[addtoplot2$num > 1  & addtoplot2$Factor == "Interaction" ,] , aes(xintercept = ppeakPOS, color = CSS), size = 2, alpha = 0.2) +
  geom_line(aes(x = POS, y = prop, size = paste(Pool,Dose2),
             linetype = prop < prop095)) + 
  geom_hline(aes(yintercept = prop095), color = "gray") +
  #the rest
  facet_grid(CSS~CHROM, space = "free", scales = "free") +
theme(legend.position = "none",
      strip.text = element_text(colour = 'black'),
      panel.background = element_rect(fill = "white", color = "gray95"),
      panel.border = element_blank(), 
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "gray95", colour = "gray95", 
                linewidth = rel(1)), complete = TRUE) +
  #scale_color_manual(values = rep("black", 12)) +
  scale_size_manual(values = rep(0.5, 12)) +
  ggtitle("Normalized Selection Effects Compared to Interaction Peaks") +
  scale_color_manual(values = c("#43ACC7", "darkorange"))

```

Stats for is the highest point of the interactions within the window of the additive?

```{r}
maybepeaks %>% pivot_wider(names_from = Factor, values_from = ppeakPOS) %>% select(Pool, Dose, CHROM, Bulk, Interaction) %>%
  group_by(Pool, Dose, CHROM) -> AttemptAll

mylist <- list()
for(i in unique(AttemptAll$CHROM)){
  Attempt <- AttemptAll %>% filter(CHROM == i)
  mylist[[i]] <- dist(c(Attempt$Bulk[is.na(Attempt$Bulk) == FALSE], 
     Attempt$Interaction[is.na(Attempt$Interaction) == FALSE]
    ))
}

result <- data.frame(Diff = NA, Int = NA, CHROM = NA)
for(c in unique(AttemptAll$CHROM)){
    Attempt <- AttemptAll %>% filter(CHROM == c) 
    Bulk <- Attempt$Bulk[is.na(Attempt$Bulk) == FALSE]
    Int <- Attempt$Interaction[is.na(Attempt$Interaction) == FALSE]

    if(length(Int) > 0){
      for(k in 1:length(Int)){
      res <- data.frame(Diff = Int[k] - Bulk,
                        Int = Int[k],
                        CHROM = c)
      
      result <- rbind(result, res)
    }
    }
    
}


mindist <- 24000
result %>% na.omit() %>% group_by(CHROM, Int) %>% summarize(ClosestPeak = min(abs(Diff))) %>% mutate(NumClose = ClosestPeak < mindist) -> ThesePeaks

sum(ThesePeaks$NumClose)/length(ThesePeaks$NumClose)
maybepeaks %>% filter(Factor == "Bulk") %>% group_by(Pool, CHROM) %>% count() -> BulkNums
maybepeaks %>% filter(Factor == "Interaction") %>% group_by(Pool, CHROM) %>% count() %>% mutate(Ints = n) %>% select(-n) %>% merge(BulkNums)

CuSO4_All %>% select(POS) -> Positions

simproportions <- vector()
for(k in 1:100){
  Bs <- sample(Positions$POS, size = 112)
  Is <- sample(Positions$POS, size = 46)
  
  for(i in 1:46){
    if(i == 1){
      simresults <- min(abs(Is[i] - Bs))
    }else{
      simresults <- c(simresults, min(abs(Is[i] - Bs)))
    }
  }
  
  simproportions[k] <- sum(simresults < mindist)/46
  
}

plot(c(simproportions,sum(ThesePeaks$NumClose)/length(ThesePeaks$NumClose)))
```

Things that vary:

1. Number of QTL per chromosome per pool
2. Number of interactions per chromosome per pool
3. Where those interactions actually are

Should we just randomize all of the peaks and interactions? Or within chromosomes? Or within Pools?

```{r}
maybepeaks %>% filter(Factor == "Bulk") %>% group_by(Pool, CHROM, Dose) %>% count() %>% mutate(QTL = n) %>% select(-n) -> BulkNums
maybepeaks %>% filter(Factor == "Interaction") %>% group_by(Pool, CHROM, Dose) %>% count() %>% mutate(Ints = n) %>% select(-n) %>% merge(BulkNums) -> SimStats

SimStats

sample_IntPOS <- function(CHROM, Ints, n, sims = 100, dist = 24000){
  #all available positions
  Positions <- CuSO4_All$POS[CuSO4_All$CHROM == CHROM]
  numwithinrange <- vector()
  #randomize where they are x times
  for(i in 1:sims){
    BulkPOS <- sample(Positions, size = n)
    IntPOS <- sample(Positions, size = Ints)
    
    #what's the closest peak to each interaction peak?
    if(k > 1){
      for(k in 1:length(IntPOS)){
      if(k == 1){
        diff <- min(abs(IntPOS[] - BulkPOS))
      }else{
        
      }
      diff <- c(diff, min(abs(IntPOS[] - BulkPOS)))
    }
    }else{
      diff <- min(abs(IntPOS[] - BulkPOS))

    }
    
    #How many of these are within the range?
    numwithinrange[i] <- sum(abs(diff) < dist)
  }
  
  #total in each simulation that were within the range
  return(numwithinrange)
}

sample_IntPOS(CHROM = "VIII", Ints = 1, n = 1, sims = 10)
```


```{r}
SimStats %>% group_by(CHROM, Pool) %>%
  summarize(Ints = Ints,
            n = n,
    Successes = sample_IntPOS(sims = 1000,
                              CHROM = CHROM,
                                      Ints = Ints,
                                      n = n)) -> Sims

Sims %>% ggplot(aes(x = CHROM, y = Successes/Ints, color = Pool)) + geom_jitter()
Sims %>% group_by(Pool, CHROM) %>% summarize(Ints = Ints, n = n,
                                             Sums = sum(Successes)) %>%
  distinct() %>%
  ggplot(aes(x = Ints, y = Sums/1000, size = n, color = CHROM)) + geom_point(alpha = 0.5)

```

Okay actual things that can change:

1. Position distribution (size of chromosome, 16 options)
2. Number of positions (coverage, sliding scale)
3. Number of QTL (n, infinite but realistically <10)
4. Number of interactions (Ints, infinite but realistically < n)

```{r}
SimDistribution <- function(ChrSize = 1000,
                            NumPOS = ChrSize/10,
                            n = 2,
                            Int = 2,
                            dist = 100,
                            sims = 100){
  if(n < 1){
    return("No Bulk Effects, automatic p = 0")
  }
  
  if(Int < 1){
    return("No Interactions to detect, automatic p = 0")
  }
  
  successes <- vector()
  #Simulate a genome
  for(k in 1:sims){
    POS <- sample(1:ChrSize, size = NumPOS)
    Ints <- sample(POS, size = Int)
    QTL <- sample(POS, size = n)
    
    #loop through interactions sampled
    if(length(Ints) > 1){
      for(i in 1:length(Ints)){
        if(i == 1){
          diff <- min(abs(Ints[i] - QTL))
        }else{
          diff <- c(diff, min(abs(Ints[i] - QTL)))
        }
      }
    }else{
      diff <- min(abs(Ints - QTL))
    }
    
    #Sum the differences above distance
    successes[k] <- sum(diff < dist)
  }
  
  return(successes)
  
}

as.data.frame(table(SimDistribution(n = 10, Int = 1, ChrSize =  100000, NumPOS = 100, sims = 10000))/10000)
```
```{r}
CuSO4_All %>% filter(Factor == "Bulk") %>%  
  group_by(CHROM, Pool, Dose) %>% summarize(POS = POS, max = max(POS)) %>% 
  group_by(CHROM, Pool, Dose,max) %>% count() -> ParamsToSim

merge(ParamsToSim, SimStats) %>%
  group_by(CHROM, Pool, Dose) %>%
  summarize(as.data.frame(table(SimDistribution(n = QTL,
                            Int = Ints,
                            ChrSize = max,
                            NumPOS = n,
                            dist = 24000, 
                            sims = 1000))/1000)) -> pvaltable


pvaltable %>% pivot_wider(names_from = Var1, values_from = Freq, names_prefix = "P_")

maybepeaks %>% select(Pool, Dose, Factor, CHROM, ppeakPOS) %>% filter(Factor == "Bulk") -> BulkPOS_CuSO4
maybepeaks %>% select(Pool, Dose, Factor, CHROM, ppeakPOS) %>% filter(Factor == "Interaction") -> IntPOS_CuSO4

results <- data.frame(Pool = NA, Dose = NA, Factor = NA, CHROM = NA, ppeakPOS = NA, nearest = NA)
for(i in 1:nrow(IntPOS_CuSO4)){
  myint <- IntPOS_CuSO4[i,]
  
  BulkPOS_CuSO4 %>% filter(CHROM == myint$CHROM, Pool == myint$Pool, Dose == myint$Dose) %>% 
    mutate(dist = abs(ppeakPOS - myint$ppeakPOS)) %>%
    summarize(nearest = min(dist)) -> nearestpeak
  
  results[i,] <- cbind(myint, nearestpeak)
}

results %>% group_by(Pool, Dose, CHROM) %>% summarize(Var1 = factor(sum(nearest < 24000), levels = c(0,1,2,3,4))) -> QTLpeakresults

pvaltable %>% merge(QTLpeakresults) %>% ggplot(aes(x = Var1, y = Freq, color = CHROM)) + geom_point() + geom_hline(yintercept = 0.05, linetype = "dashed")

```
 I think the fact that I'm separating by chromosome is perhaps not right... it's actually the overall genome and which are there
 
```{r}
SimGenomeDistribution <- function(dataset = NULL,
                                  n = 2,
                            Int = 2,
                            dist = 100,
                            sims = 100){
  
  #INSTRUCTIONS
  if(is.null(dataset)){return("Please include the Pool ID")}
  if(n < 1){return("No Bulk Effects, automatic p = 0")}
  if(Int < 1){return("No Interactions to detect, automatic p = 0")}
  
  #SETUP
  POS <- CuSO4_All %>% filter(Pool == dataset) %>% transmute(P = paste(CHROM, POS, sep = "_")) #HTG3TDMXY
  successes <- vector()
  
  #SIMULATE DISTRIBUTION OF INTERACTIONS
  for(k in 1:sims){
    
    #Sample and make vectors
    Ints <- sample(POS$P, size = Int)
    QTL <- sample(POS$P, size = n)
    data.frame(Ints) %>% separate(col = Ints, into =c("CHROM", "POS"), sep = "_") %>% mutate(POS = as.numeric(POS)) -> Is
    data.frame(QTL) %>% separate(col = QTL, into =c("CHROM", "POS"), sep = "_") %>% mutate(POS = as.numeric(POS))-> Qs

    ############################################################################
    
    if(sum(Qs$CHROM %in% unique(Is$CHROM)) == 0){
      #print("There are no successes this round")
      diff[1:Int] <- dist+1
    }else{
      #Check if there's more than 1 Interaction
      if(length(Is$POS) == 1){
        QTL <- Qs %>% filter(CHROM %in% Is$CHROM) %>% select(POS)
        diff <- abs(Is$POS - QTL$POS)
      }else{
        #LOOP THROUGH QTLS
        for(p in 1:length(Is$POS)){
          QTL <- Qs %>% filter(CHROM %in% Is[k,]$CHROM) %>% select(POS)
          if(as.vector(QTL %>% count()) == 0){diff[p] <- dist + 1}else{
            diff[p] <- abs(Is$POS - QTL$POS)
          }
        }
      }
    }
    
    
    #Sum the differences above distance
    successes[k] <- sum(diff < dist)
  }
  return(successes)
}

```
 
```{r}
CuSO4_All %>% filter(Factor == "Bulk") %>%  ungroup() %>%
  group_by(Pool, Dose) %>% count() -> ParamsToSim

SimStats %>% ungroup() %>% group_by(Pool, Dose) %>%
  summarize(Ints = sum(Ints), QTL = sum(QTL)) %>%
  merge(ParamsToSim) %>%
  #TROUBLESHOOT
  filter(Pool == "HKTFTDRX2", Dose == 0) %>%
  #BACK TO NORMAL
  group_by(Pool, Dose) %>%
  summarize(as.data.frame(table(SimGenomeDistribution(dataset = Pool,
                                                      n = QTL,
                                                      Int = Ints,
                                                      dist = 24000, 
                                                      sims = 10000))/10000)) -> pvaltable

pvaltable %>% ggplot(aes(x = Var1, y = Freq, color = Pool)) + geom_point()


```

```{r}
maybepeaks %>% select(Pool, Dose, Factor, ppeakPOS) %>% filter(Factor == "Bulk") -> BulkPOS_CuSO4
maybepeaks %>% select(Pool, Dose, Factor, ppeakPOS) %>% filter(Factor == "Interaction") -> IntPOS_CuSO4

results <- data.frame(Pool = NA, Dose = NA, Factor = NA, ppeakPOS = NA, nearest = NA)
for(i in 1:nrow(IntPOS_CuSO4)){
  myint <- IntPOS_CuSO4[i,]
  
  BulkPOS_CuSO4 %>% filter(Pool == myint$Pool, Dose == myint$Dose) %>% 
    mutate(dist = abs(ppeakPOS - myint$ppeakPOS)) %>%
    summarize(nearest = min(dist)) -> nearestpeak
  
  results[i,] <- cbind(myint, nearestpeak)
}
results %>% group_by(Pool, Dose) %>% summarize(Var1 = sum(nearest < 24000)) 
results %>% group_by(Pool, Dose) %>% summarize(Var1 = factor(sum(nearest < 24000), 
                                                             levels = c(0:30))) -> QTLpeakresults


pvaltable %>% merge(QTLpeakresults) %>% ggplot(aes(x = Var1, y = Freq)) + geom_point() + geom_hline(yintercept = 0.05, linetype = "dashed")
```

