---
title: "Figures 2023"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: hide
  html_notebook:
    code_folding: hide
---

```{r, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)
library(scales)
require(circlize) 
library(stringr)

#library(cybrBSA)

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 20)))

glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H","H","H", "H", 
                                             "L", "L","L", "L", 
                                             "H", "H","H", "H", 
                                             "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", 
                                             "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W",
                                             "O", "W", "O", "W", 
                                             "O", "W", "O", "W",
                                             "O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", 
                                          "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                                     HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

cybr2_rollmean <- function(dataframe){
  dataframe %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollmean(value, n = 100))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount)
}

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Interaction") %>% select(-delete)

```

Note: if looking for smoothed vs unsmoothed data, the `Figures_2023_BiologyofGenomes.rmd` file has these at the beginning.

# Processing
## Load in Data



```{r, warning=FALSE, message=FALSE}
#CSS 1
Zeocin_glm<- readRDS("Data/Zeocin_glm.rds") %>% filter(CHROM != "I")
CuSO4_glm <- readRDS("Data/CuSO4_1_glm.rds") %>% filter(CHROM != "I")
CuSO4_glm2 <- readRDS("Data/CuSO4_2_glm.rds") %>% filter(CHROM != "I")
Fluc_glm <- readRDS("Data/Fluconazole_1_glm.rds") %>% filter(CHROM != "I")

#CSS 8
CuSO4_8_glm <- readRDS("Data/CuSO4_CSS8_glm.rds") %>% filter(CHROM != "VIII")
CuSO4_8_c5_glm <- readRDS("Data/CuSO4_CSS8_C5norep_glm.rds") %>% filter(CHROM != "VIII")
CuSO4_8_D2_glm <- readRDS("Data/CuSO4_CSS8_D2_norep_glm.rds") %>% filter(CHROM != "VIII")
Fluc_8_glm <- readRDS("Data/Fluc_CSS8_glm.rds") %>% filter(CHROM != "VIII")

#New H2O2 and Zeocin 9/11/23 - might need to be reformatted to start
readRDS("H2O2_CSS8_glm_SUB_50.rds") %>% mutate(coefficient = str_replace(coefficient, "BulkH2O2", "Bulk"),
                             coefficient = str_replace(coefficient, "ParentWine8", "Parent"),
                             coefficient = str_replace(coefficient, ":", ""),
                             coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
  filter(label == "Zscore", coefficient != "(Intercept)", CHROM != "VIII") %>%
  transmute(CHROM = CHROM, POS = POS, summary = GLMResult, label = coefficient) -> H2O2_8_glmer

readRDS("H2O2_CSS1_glm50.rds") %>% mutate(coefficient = str_replace(coefficient, "BulkH2O2", "Bulk"),
                             coefficient = str_replace(coefficient, "ParentWine1", "Parent"),
                             coefficient = str_replace(coefficient, ":", ""),
                             coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
  filter(label == "Zscore", coefficient != "(Intercept)", CHROM != "I") %>%
  transmute(CHROM = CHROM, POS = POS, summary = GLMResult, label = coefficient)-> H2O2_glmer

readRDS("Zeocin_CSS8_glm50.rds") %>% mutate(coefficient = str_replace(coefficient, "BulkZeocin", "Bulk"),
                           coefficient = str_replace(coefficient, "ParentWine8", "Parent"),
                           coefficient = str_replace(coefficient, ":", ""),
                           coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
filter(label == "Zscore", coefficient != "(Intercept)", CHROM != "VIII") %>%
transmute(CHROM = CHROM, POS = POS, summary = GLMResult, label = coefficient) -> Zeocin_8_glmer

#Fluc Redone
readRDS("Fluc_CSS1_glmer.rds") %>% mutate(coefficient = str_replace(coefficient, "BulkFluc", "Bulk"),
                             coefficient = str_replace(coefficient, "ParentWine1", "Parent"),
                             coefficient = str_replace(coefficient, ":", ""),
                             coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
  filter(label == "Zscore", coefficient != "(Intercept)", CHROM != "I") %>%
  transmute(CHROM = CHROM, POS = POS, summary = GLMResult, label = coefficient) -> Fluc_1_glmer

#CuSO4 Redone
readRDS("CuSO4_CSSI_all_glmer.rds") %>% #mutate(coefficient = str_replace(Param, "BulkCuSO4", "Bulk"),
                             #coefficient = str_replace(coefficient, "ParentWine1", "Parent"),
                             #coefficient = str_replace(coefficient, ":", ""),
                             #coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
  filter(Type == "Zscore", Param != "(Intercept)", CHROM != "I") %>%
  transmute(CHROM = CHROM, POS = POS, summary = GLM, label = Param) -> CuSO4_1_glmer

################################################################################
#Reformatting
################################################################################

Zeocin_glm$label <- factor(Zeocin_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
Fluc_glm$label <- factor(Fluc_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_glm$label <- factor(CuSO4_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_glm2$label <- factor(CuSO4_glm2$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

Zeocin_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Zeocin_glm
Fluc_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Fluc_glm
CuSO4_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> CuSO4_glm
CuSO4_glm2 %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> CuSO4_glm2

#New CSS8 Ones - add back Chr I
Fluc_8_glm$label <- factor(Fluc_8_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_8_c5_glm$label <- factor(CuSO4_8_c5_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_8_D2_glm$label <- factor(CuSO4_8_D2_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_8_glm$label <- factor(CuSO4_8_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

Fluc_8_glm %>% filter(CHROM != "M", label != "intercept") -> Fluc_8_glm
CuSO4_8_c5_glm %>% filter(CHROM != "M", label != "intercept") -> CuSO4_8_c5_glm
CuSO4_8_D2_glm %>% filter(CHROM != "M", label != "intercept") -> CuSO4_8_D2_glm
CuSO4_8_glm %>% filter(CHROM != "M", label != "intercept") -> CuSO4_8_glm

#Newest: H2O2
H2O2_8_glmer %>% filter(CHROM != "VIII", CHROM != "M", label != "intercept") -> H2O2_8_glmer
H2O2_glmer %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> H2O2_glmer
H2O2_8_glmer$label <- factor(H2O2_8_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
H2O2_glmer$label <- factor(H2O2_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

#Zeocin New
Zeocin_8_glmer %>% filter(CHROM != "VIII", CHROM != "M", label != "intercept") -> Zeocin_8_glmer
Zeocin_8_glmer$label <- factor(Zeocin_8_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

#Fluc glmer
Fluc_1_glmer %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Fluc_1_glmer
Fluc_1_glmer$label <- factor(Fluc_1_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

#New CuSO4 glmer
CuSO4_1_glmer %>% filter(CHROM != "I", CHROM != "M", label != "Intercept") %>% na.omit() -> CuSO4_1_glmer
CuSO4_1_glmer$label <- factor(CuSO4_1_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

```

Parent data

```{r}
engineered <- data.frame(Locus = c("URA3", "MATalpha"),
                         CHROM = c("V", "III"),
                         POS_start = c(116167, 198671),
                         POS_end = c(116970, 201177))

```

```{r}
setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")

cybrConvertParentalAlleles <- function(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"),
                                       Parents = gsub("_VCF.txt","", ParentFiles), Truncate = TRUE, yeast = TRUE){
  temparent <- list()
  mergeparents <- foreach(i=1:length(ParentFiles), .combine=rbind) %dopar% {
    read.table(ParentFiles[i], header = TRUE) %>% mutate(parent = Parents[i])
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                           "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                           CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                     "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                     "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                     "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    rbind(mergeparents) %>% arrange(CHROM, POS) %>%
      select(CHROM, POS, REF, ALT, parent) %>%
      merge(ChromKey) %>% select(-CHROM) %>%
      mutate(CHROM = chromosomes) %>% select(-chromosomes) -> ParentalVCF

  }else{
    rbind(mergeparents) %>% arrange(CHROM, POS) %>%
      select(CHROM, POS, REF, ALT, parent) -> ParentalVCF
  }

  ParentalVCF %>% pivot_wider(names_from = parent, values_from = ALT) -> SNPids

  SNPids$Type <- 0
  for(i in Parents){

    #filter rows in which all values of columns of the parent NOT selected are NA
    select(SNPids,-i, -CHROM, -POS, -REF) -> tempdf
    tempdf$Any_NA <- apply(tempdf, 1, function(x) anyNA(x))
    SNPids$Type[which(tempdf$Any_NA)] <- i
    rm(tempdf)
  }


  #Collect it to output
  if(Truncate == TRUE){
    SNPids %>% select(CHROM, POS,  Type) %>% filter(Type != 0) -> SNPids
  }

  return(SNPids)

}

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) %>% filter(Unique == 1) -> pSNPs

pSNPs$CHROM <- factor(pSNPs$CHROM, levels = as.character(as.roman(1:17)))

pSNPs_all <- rbind(data.frame(pSNPs[,1:2], E = "A_REF"),
                   data.frame(H2O2_glmer[,1:2], E = "H1"),
                   data.frame(H2O2_8_glmer[,1:2], E = "H8"),
                   data.frame(CuSO4_8_glm[,1:2], E = "C8"),
                   data.frame(CuSO4_1_glmer[,1:2], E = "C1"),
                   data.frame(Fluc_1_glmer[,1:2], E = "F1"),
                   data.frame(Fluc_8_glm[,1:2], E = "F8"),
                   data.frame(Zeocin_8_glmer[,1:2], E = "Z8"),
                   data.frame(Zeocin_glm[,1:2], E = "Z1")) %>%
  distinct()


```

## Peak Calling

### GLM Datasets:
Zeocin_glm
Fluc_glm
CuSO4_glm
CuSO4_glm2
Fluc_8_glm
CuSO4_8_c5_glm
CuSO4_8_D2_glm
CuSO4_8_glm

#### GLMER DATASETS
H2O2_8_glmer
H2O2_glmer 
Zeocin_8_glmer 
Fluc_1_glmer 
CuSO4_1_glmer 


### Call Peaks (Function)

Works with the reformatting in the previous step

```{r}
cybr_callpeaks <- function(dataset, param = NULL, threshold = NULL, include_all = NULL){
  dataset %>% mutate(summary = abs(summary)) %>% filter(label != "intercept",
                                                        label != "Intercept",
                                                        label != "(Intercept)",
                                                        label != "NA") %>%
    na.omit() -> dataset
  #Set cutoff
  
  Cutoff <- data.frame(X = NA)
  if(is.null(threshold) == FALSE){
    Cutoff$X <- threshold
  }else{
    dataset %>% filter(CHROM == "III") %>%
    ungroup() %>%
    reframe(X = max(summary)) -> Cutoff
  }
  
  #Find peaks
  if(is.null(include_all)){
    dataset %>%
    filter(summary > Cutoff$X) %>%
    group_by(CHROM, label) %>% 
    summarize(summary = max(summary)) %>%
    ungroup() %>%
    merge(dataset) -> output
  
  }else{
    dataset %>%
    filter(summary > Cutoff$X) -> output
  }
  
  if(is.null(param) == FALSE){
    output %>% filter(label == param) -> output
  }
  return(output)
}
```

```{r}

#Chr III Cutoff Peaks
F_1_intpeaks <- cybr_callpeaks(Fluc_glm, param = "Interaction")
C_1_intpeaks <- cybr_callpeaks(CuSO4_glm, param = "Interaction")
H_8_intpeaks <- cybr_callpeaks(H2O2_8_glmer, param = "Interaction")
Z_1_intpeaks <- cybr_callpeaks(Zeocin_glm, param = "Interaction")
C_8_intpeaks <- cybr_callpeaks(CuSO4_8_glm, param = "Interaction")
F_8_intpeaks <- cybr_callpeaks(Fluc_8_glm, param = "Interaction")
Z_8_intpeaks <- cybr_callpeaks(Zeocin_8_glmer, param = "Interaction")

#Don't just find point
F_1_intpeaks_all <- cybr_callpeaks(Fluc_glm, param = "Interaction", include_all = TRUE)
C_1_intpeaks_all <- cybr_callpeaks(CuSO4_glm, param = "Interaction", include_all = TRUE)
H_8_intpeaks_all <- cybr_callpeaks(H2O2_8_glmer, param = "Interaction", include_all = TRUE)
Z_1_intpeaks_all <- cybr_callpeaks(Zeocin_glm, param = "Interaction", include_all = TRUE)
C_8_intpeaks_all <- cybr_callpeaks(CuSO4_8_glm, param = "Interaction", include_all = TRUE)
F_8_intpeaks_all <- cybr_callpeaks(Fluc_8_glm, param = "Interaction", include_all = TRUE)
Z_8_intpeaks_all <- cybr_callpeaks(Zeocin_8_glmer, param = "Interaction", include_all = TRUE)

#GLMER PEAKS
H8g_intpeaks <- cybr_callpeaks(H2O2_8_glmer, param = "Interaction")
C1g_intpeaks <- cybr_callpeaks(CuSO4_1_glmer, param = "Interaction")
H1g_intpeaks <- cybr_callpeaks(H2O2_glmer, param = "Interaction")
Z8g_intpeaks <- cybr_callpeaks(Zeocin_8_glmer, param = "Interaction")
F1g_intpeaks <- cybr_callpeaks(Fluc_1_glmer, param = "Interaction")

#UNUSED:
#CuSO4_glm2
#CuSO4_8_c5_glm
#CuSO4_8_D2_glm
```

#### Calling Peaks with dilute parent cutoffs

These were simulated by sampling from dilute bulks four random (smoothed) read counts, two from each parent, and then calculating the bulk, parent, and rep effects. The 5% and 1% FDR were then found from this, which ensures that the parent effect ranges are consistent but that all of the loci can be randomized since the assumption is the dilute bulks will not have an effect overall. This does assume that there are no QTL on the dilute bulks, which is likely false, but which is taken into consideration as the amount of noise that we won't be able to parse.

```{r, eval = FALSE}
CSSCutoff_99 <- readRDS("../9-12-23_DiluteCutoff_99.rds")

CSSCutoff_new <- readRDS("../9-20-23_DiluteCutoff_99.rds")

CutoffExpKey <- data.frame(Exp = unique(CSSCutoff_99$Exp), 
                           Dataset = c("CuSO4_glm",
                                       "CuSO4_8_glm",
                                       "Fluc_glm","Fluc_glm",
                                       "H2O2_glmer",
                                       "H2O2_8_glmer",
                                       "Zeocin_glm"))

C99 <- merge(CSSCutoff_99, CutoffExpKey)

C99 %>% filter(labels == "Interaction")

CSSCutoff_new %>% filter(labels == "Interaction") %>% filter(., grepl("glm_",Exp))

#USE FOR GLMER ONES
C_1_intpeaks_perm <- cybr_callpeaks(CuSO4_1_glmer, param = "Interaction", threshold = 2.808175)
C_8_intpeaks_perm <- cybr_callpeaks(CuSO4_8_glm, param = "Interaction", threshold = 8.691379)

F_1_intpeaks_perm <- cybr_callpeaks(Fluc_1_glmer, param = "Interaction", threshold = 1.872063)
#F_8_intpeaks_perm <- cybr_callpeaks(Fluc_8_glm, param = "Interaction", threshold = 1.96)

#H_1_intpeaks_perm <- cybr_callpeaks(H2O2_glmer, param = "Interaction", threshold = 0.3141185)

H_8_intpeaks_perm <- cybr_callpeaks(H2O2_8_glmer, param = "Interaction", threshold = 7.168438)

Z_8_intpeaks_perm <- cybr_callpeaks(Zeocin_8_glmer, param = "Interaction", threshold = 7.168438)
Z_1_intpeaks_perm <- cybr_callpeaks(Zeocin_glm, param = "Interaction", threshold = 5.626115)

################################################################################
#GLM Permutations on glm data
C_1_intpeaks_perm_b <- cybr_callpeaks(CuSO4_glm, param = "Interaction", threshold = 2.743073)
C_8_intpeaks_perm_b <- cybr_callpeaks(CuSO4_8_glm, param = "Interaction", threshold = 8.442994)
F_1_intpeaks_perm_b <- cybr_callpeaks(Fluc_glm, param = "Interaction", threshold = 1.928081)
Z_1_intpeaks_perm_b <- cybr_callpeaks(Zeocin_glm, param = "Interaction", threshold = 5.063239)


```


# Standardized Plots

```{r}
#UPDATED COLORS:
#Verdigris - chr I default
  "#7DB0B0"
#Tangerine - chr VIII default
  "#ED7B01" 
#Lapis Lazuli: 
  "#24588A90"
#Magenta Haze:
  "#954176"
#Yale Blue:
  "#0E3961"
```

### Parent SNP Loci

```{r, fig.width=16, fig.height=4}

centromeres <- data.frame(CHROM = as.character(as.roman(1:16)),
                          POS = c(151465,238207,114385,449711,151987,148510,496920,105586,
                                  355629,436307,440129,150828,268031,628758,326584,555957),
                          Type = "Oak",
                          Type2 = "Wine") %>% pivot_longer(c(Type, Type2), names_to = "label", values_to = "Type")


pSNPs %>% rbind(ChromosomeScale) %>% filter(CHROM != "M") %>%
  ggplot(aes(x = POS, y = Type, color = Type)) + 
  geom_jitter(aes(x = POS), alpha = 0.2, size =0.1, height = 0.2) +
  geom_point(data = centromeres, aes(x = POS, y = Type), size = 3, color = "black") +
  scale_color_manual(values = c("navy", "firebrick")) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none")

pSNPs %>% rbind(ChromosomeScale) %>% filter(CHROM != "M") %>%
  ggplot(aes(x = POS, y = "C")) + 
  geom_jitter(aes(x = POS), alpha = 0.05, size =0.2, height = 0.2) +
  geom_point(data = centromeres, aes(x = POS), size = 3) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none")

pSNPs %>% rbind(ChromosomeScale) %>% filter(CHROM != "M") %>%
  ggplot(aes(x = POS, color = Type)) + 
  geom_density(aes(x = POS)) +
  geom_vline(data = centromeres, aes(xintercept = POS)) +
  scale_color_manual(values = c("navy", "firebrick")) +
  facet_grid(~CHROM, scales = "free_x", space = "free") +
  theme(legend.position = "none")

#Plot all SNPs called for each experiment; this can take a while
pSNPs_all %>% filter(CHROM != "M") %>%
  ggplot(aes(x = POS, y = E)) + 
  geom_jitter(aes(x = POS, y = E), alpha = 0.05, size =0.2, height = 0.2) +
  #geom_point(data = centromeres, aes(x = POS), size = 3) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none", axis.text.x = element_blank())

```

### Function for Standard Plots

```{r, fig.width=16, fig.height=4}

cybrPurple <- function(dataset, 
                       cutoff = "III", 
                       peakslist = NULL, 
                       includeFixedGenes = TRUE, 
                       peakcolor = "#24588A90", 
                       mylim = NULL){
  #Make start and end points
  ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
    pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
    mutate(Summary = NA, Label = NA) %>% select(-delete)
  
  #Make Fixed Genes
  FG <- data.frame(Gene = c("Ura3", "MATalphastart", "MATalpha_end"),
                   CHROM = factor(c("V", "III", "III"), levels = as.character(as.roman(1:16))),
                   POS = c(116167, 198671,201177))
  
  #Make factors for coloring
  dataset$label <- factor(dataset$label, levels = c("Bulk", "Parent", "Interaction", "Rep", "Intercept"))

  dataset %>% select(CHROM, POS, Summary = summary, Label = label) %>% 
    rbind(ChromosomeScale) %>%
    ggplot(aes(x = POS, y = abs(Summary), color = Label, linetype = Label %in% c("Parent", "Rep"))) + 
    facet_grid(cols = vars(CHROM), scales = "free", space = "free") -> BasePlot
  
  #Add fixed genes to plot
  if(includeFixedGenes == TRUE){
    BasePlot <- BasePlot + geom_vline(data = FG, aes(xintercept = POS), color = "gray", size = 2, linetype = "dashed")
  }
  
  #Add peaks to plot
  if(is.null(peakslist) == FALSE){
    
    peakslist$CHROM <- factor(peakslist$CHROM, levels = as.character(as.roman(1:16)))
    BasePlot <- BasePlot + geom_vline(data = peakslist, aes(xintercept = POS), color = peakcolor, size = 2)
  }

  
  if(cutoff == "III"){
    FinalPlot <- BasePlot + 
                geom_line(size = 1.2) +
                #Cutoff
                geom_hline(aes(yintercept = max(abs(dataset$summary[dataset$CHROM == "III"]))), linetype = "dashed") +
                ylab("") + xlab("")+ ggtitle("")+
                scale_color_manual(values = c("black",  "gray40","#7030A0", "lightpink", "red")) + 
                theme(legend.position = "none", axis.text.x=element_blank(),
                          axis.ticks.x=element_blank()) 
    
  }else if(is.null(cutoff)){
    FinalPlot <- BasePlot + 
                geom_line(size = 1.2) +
                ylab("") + xlab("")+ ggtitle("")+
                # facet_grid(~CHROM, scales = "free", space = "free")  + 
                scale_color_manual(values = c("black",  "gray40","#7030A0", "lightpink", "red")) + 
                theme(legend.position = "none", axis.text.x=element_blank(),
                          axis.ticks.x=element_blank()) 
  }else{
    cutoff <- as.numeric(cutoff)
    FinalPlot <- BasePlot + 
                geom_line(size = 1.2) +
                geom_hline(aes(yintercept = cutoff)) +
                ylab("") + xlab("")+ ggtitle("")+
                # facet_grid(~CHROM, scales = "free", space = "free")  + 
                scale_color_manual(values = c("black",  "gray40","#7030A0", "lightpink", "red")) + 
                theme(legend.position = "none", axis.text.x=element_blank(),
                          axis.ticks.x=element_blank()) 
  }
  
  if(is.null(mylim) == FALSE){
    FinalPlot <- FinalPlot + ylim(0, mylim)
  }
  
  return(FinalPlot)
}
```


All GLM plots (not glmer)

```{r, fig.width=16, fig.height=4}
#GLM PLOTS
cybrPurple(Fluc_glm, peakslist = F_1_intpeaks)
cybrPurple(CuSO4_glm, peakslist = C_1_intpeaks)
cybrPurple(Zeocin_glm, peakslist = Z_1_intpeaks)
cybrPurple(CuSO4_8_glm, peakslist = C_8_intpeaks, peakcolor = "#ED7B0190")
cybrPurple(Fluc_8_glm, peakslist = F_8_intpeaks, peakcolor = "#ED7B0190")

#GLM PLOTS
cybrPurple(Fluc_glm, peakslist = F_1_intpeaks_all)
cybrPurple(CuSO4_glm, peakslist = C_1_intpeaks_all)
cybrPurple(Zeocin_glm, peakslist = Z_1_intpeaks_all)
cybrPurple(CuSO4_8_glm, peakslist = C_8_intpeaks_all, peakcolor = "#ED7B0190")
cybrPurple(Fluc_8_glm, peakslist = F_8_intpeaks_all, peakcolor = "#ED7B0190")

#GLMER PLOTS
cybrPurple(H2O2_8_glmer, peakslist = H8g_intpeaks, peakcolor = "#ED7B0190")
cybrPurple(CuSO4_1_glmer, peakslist = C1g_intpeaks)
cybrPurple(H2O2_glmer, peakslist = H1g_intpeaks)
cybrPurple(Fluc_1_glmer, peakslist = F1g_intpeaks)

```

```{r, fig.width=16, fig.height=4, eval = FALSE}

#GLM PLOTS - can't use this cutoff because the permutations were done with random effects
#cybrPurple(CuSO4_8_glm, peakcolor = "#ED7B01", cutoff = 8.691379)
#cybrPurple(Zeocin_glm, cutoff = 5.626115)

#GLMER PLOTS
cybrPurple(CuSO4_1_glmer, cutoff = 2.808175)
cybrPurple(Fluc_1_glmer, cutoff = 1.872063)
cybrPurple(H2O2_8_glmer, peakcolor = "#ED7B01", cutoff = 7.168438)
cybrPurple(Zeocin_8_glmer, peakcolor = "#ED7B01", cutoff = 7.168438)

#GLM Fixed Effect Permutation PLOTS
cybrPurple(CuSO4_glm, peakslist = C_1_intpeaks_perm_b, cutoff = 2.743073)
cybrPurple(CuSO4_8_glm, peakslist = C_8_intpeaks_perm_b, cutoff = 8.442994, peakcolor = "#ED7B01")
cybrPurple(Fluc_glm, peakslist = F_1_intpeaks_perm_b, cutoff = 1.928081)
cybrPurple(Zeocin_glm, peakslist = Z_1_intpeaks_perm_b, cutoff = 5.063239)
```

```{r, fig.width=16, fig.height=4}
Cutoffs_perm <- readRDS("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/NewCutoffs_perm.rds")
Cutoffs_perm

#CALL PEAKS WITH NEW CUTOFFS
c1p <- cybr_callpeaks(CuSO4_glm, param = "Interaction", threshold = 0.4098215, include_all = TRUE)
c8p <- cybr_callpeaks(CuSO4_8_glm, param = "Interaction", threshold = 0.4839316, include_all = TRUE)
f1p <- cybr_callpeaks(Fluc_glm, param = "Interaction", threshold = 0.4260583, include_all = TRUE)
z1p <- cybr_callpeaks(Zeocin_glm, param = "Interaction", threshold = 0.4024422, include_all = TRUE)
h8p <- cybr_callpeaks(H2O2_8_glmer, param = "Interaction", threshold = 2.0160408, include_all = TRUE)
z8p <- cybr_callpeaks(Zeocin_8_glmer, param = "Interaction", threshold = 2.0160408, include_all = TRUE)
f8p <- cybr_callpeaks(Fluc_8_glm, param = "Interaction", threshold = 0.4839316, include_all = TRUE)


#PLOT NEW CUTOFFS
cybrPurple(CuSO4_glm, peakslist = c1p, cutoff = 0.4098215)
cybrPurple(CuSO4_8_glm, peakslist = c8p, cutoff = 0.4839316, peakcolor = "#ED7B0190")
cybrPurple(Fluc_glm, peakslist = f1p, cutoff = 0.4260583)
cybrPurple(Zeocin_glm, peakslist = z1p, cutoff = 0.4024422)
cybrPurple(H2O2_8_glmer, peakslist = h8p, peakcolor = "#ED7B0190", cutoff = 2.0160408)
cybrPurple(Zeocin_8_glmer, peakslist = z8p, peakcolor = "#ED7B0190", cutoff = 2.0160408)
cybrPurple(Fluc_8_glm, peakslist = f8p, peakcolor = "#ED7B0190", cutoff = 0.4839316)


```

Not including all

```{r, fig.width=16, fig.height=4, eval = FALSE}
c1pn <- cybr_callpeaks(CuSO4_glm, param = "Interaction", threshold = 0.4098215)
c8pn <- cybr_callpeaks(CuSO4_8_glm, param = "Interaction", threshold = 0.4839316)
f1pn <- cybr_callpeaks(Fluc_glm, param = "Interaction", threshold = 0.4260583)
z1pn <- cybr_callpeaks(Zeocin_glm, param = "Interaction", threshold = 0.4024422)
h8pn <- cybr_callpeaks(H2O2_8_glmer, param = "Interaction", threshold = 2.0160408)
z8pn <- cybr_callpeaks(Zeocin_8_glmer, param = "Interaction", threshold = 2.0160408)
f8pn <- cybr_callpeaks(Fluc_8_glm, param = "Interaction", threshold = 0.4839316)

cybrPurple(CuSO4_glm, peakslist = c1pn, cutoff = 0.4098215)
cybrPurple(CuSO4_8_glm, peakslist = c8pn, cutoff = 0.4839316, peakcolor = "#ED7B01")
cybrPurple(Fluc_glm, peakslist = f1pn, cutoff = 0.4260583)
cybrPurple(Zeocin_glm, peakslist = z1pn, cutoff = 0.4024422)
cybrPurple(H2O2_8_glmer, peakslist = h8pn, peakcolor = "#ED7B01", cutoff = 2.0160408)
cybrPurple(Zeocin_8_glmer, peakslist = z8pn, peakcolor = "#ED7B01", cutoff = 2.0160408)
cybrPurple(Fluc_8_glm, peakslist = f8pn, peakcolor = "#ED7B01", cutoff = 0.4839316)


```

# Circos Plots

### Function for Circos Plots

It's SUPER annoying to have to type in each thing over and over, so how would I make a function for each of these?

```{r, warning = FALSE, message=FALSE}

cybr_circos <- function(d1, d8, peaklist1 = NULL, peaklist8 = NULL, maxy = NULL, color1 = "#24588A", color8 = "#ED7B01", opacity = "50"){
  
  color1fade_50 <- paste(color1, "50", sep = "")
  color8fade_50 <- paste(color8, "50", sep = "")
  
  color1fade <- paste(color1, opacity, sep = "")
  color8fade <- paste(color8, opacity, sep = "")

  if(is.null(d8)){
    d8 <- d1
    include8 <- FALSE
  }else{
    include8 <- TRUE
  }
    #SET UP DATA
  df8 <- data.frame(sectors = as.character(d8$CHROM),
                 x = d8$POS,
                 y = abs(d8$summary),
                  label = d8$label)

  df1 <- data.frame(sectors = as.character(d1$CHROM),
                 x = d1$POS,
                 y = abs(d1$summary),
                  label = d1$label)

  #Take just the interactions of each
  df8int <- df8 %>% filter(label == "Interaction")
  df1int <- df1 %>% filter(label == "Interaction")

    #REORDER THE CHROMOSOMES
  df8int$sectors <- factor(df8int$sectors, levels = as.character(as.roman(1:16)))
  df1int$sectors <- factor(df1int$sectors, levels = as.character(as.roman(1:16)))
  
  ##############################################################################
  dfall <- rbind(df1int, df8int) %>% na.omit()
  circos.par("track.height" = 0.3, start.degree = 90, cell.padding = c(0,0))
  circos.initialize(sectors = dfall$sectors, x = dfall$x)

  if(is.null(maxy)){
    #I think this makes the sizes?
    circos.track(ylim = c(max(c(dfall$y)), 0), dfall$sectors, y = dfall$y, 
        panel.fun = function(x, y) {
            circos.text(CELL_META$xcenter, 
                0 - mm_y(5), 
                CELL_META$sector.index,
                niceFacing = FALSE)
            circos.axis(labels.cex = 0.1)
    })
  }else{
    circos.track(ylim = c(maxy, 0), dfall$sectors, y = dfall$y, 
      panel.fun = function(x, y) {
          circos.text(CELL_META$xcenter, 
              0 - mm_y(5), 
              CELL_META$sector.index,
              niceFacing = FALSE)
          circos.axis(labels.cex = 0.1)
      })
  }
  #Makes the chromosome overlap parts
  #CHROMOSOME I
  draw.sector(83.5, #RIGHT
              90, #LEFT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color1fade_50, border = FALSE)
  if(include8 == TRUE){
    #CHROMOSOME VIII
    draw.sector(289.5, #LEFT
              305.4, #RIGHT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color8fade_50, border = FALSE)
  }
  
  
  #Makes the lines
  circos.trackPoints(df8int$sectors, df8int$x, abs(df8int$y), col = color8, pch = 16, cex = 0.1)
  circos.trackPoints(df1int$sectors, df1int$x, abs(df1int$y), col = color1, pch = 16, cex = 0.1)
  
  if(is.null(peaklist8) == FALSE){
    if(length(peaklist8$POS) >= 1){
      for(i in 1:length(peaklist8$POS)){
      circos.link(peaklist8$CHROM[i], 
                  peaklist8$POS[i], 
                  #Add 8 after
                  "VIII", c(0, max(dfall$x[dfall$sectors == "VIII"])),  
                  
                  col = color8fade, 
                  h.ratio = 0.3, 
                  border = color8fade, 
                  lwd = 1)
      }
    }else{
      print("No interactions on Chr 8")
    }
    
  }

  if(is.null(peaklist1) == FALSE){
    if(length(peaklist1$POS) >= 1){
      for(i in 1:length(peaklist1$POS)){
        circos.link("I",c(0, max(dfall$x[dfall$sectors == "I"])),  
                    #add 1 first
                    peaklist1$CHROM[i], 
                    peaklist1$POS[i], 
                    col = color1fade, 
                    h.ratio = 0.3, 
                    border = color1fade, 
                    lwd = 1)
      }
    }else{
      print("No interactions on Chr 1")
    }
  }  
  
  circos.clear()
}

```


```{r, warning=FALSE, message=FALSE, fig.width=6, fig.height=6}

circos.clear()
#H2O2 - Don't include
# cybr_circos(d1 = rbind(H2O2_glmer, ChromosomeScale2),
#             d8 = rbind(H2O2_8_glmer, ChromosomeScale2),
#             peaklist1 = H1g_intpeaks, maxy = 4,
#             peaklist8 = H8g_intpeaks)

cybr_circos(d1 = rbind(CuSO4_glm, ChromosomeScale2), 
            d8 = rbind(CuSO4_8_glm, ChromosomeScale2), 
            peaklist1 = C_1_intpeaks, 
            peaklist8 = C_8_intpeaks,
            maxy = 4)

cybr_circos(d1 = rbind(CuSO4_glm, ChromosomeScale2), 
            d8 = rbind(CuSO4_8_glm, ChromosomeScale2), 
            peaklist1 = C_1_intpeaks_all, 
            peaklist8 = C_8_intpeaks_all,
            maxy = 4)

#Fluconazole (glmer)
cybr_circos(d1 = rbind(Fluc_1_glmer, ChromosomeScale2), 
            d8 = rbind(Fluc_8_glm, ChromosomeScale2), 
            peaklist1 = F_1_intpeaks, 
            peaklist8 = F_8_intpeaks,
            maxy = 4)

cybr_circos(d1 = rbind(Fluc_1_glmer, ChromosomeScale2), 
            d8 = NULL,
            #d8 = rbind(Fluc_8_glm, ChromosomeScale2), 
            peaklist1 = F_1_intpeaks, 
            #peaklist8 = F_8_intpeaks,
            maxy = 4)

cybr_circos(d1 = rbind(Fluc_1_glmer, ChromosomeScale2), 
            d8 = rbind(Fluc_8_glm, ChromosomeScale2), 
            peaklist1 = F_1_intpeaks_all, 
            peaklist8 = F_8_intpeaks_all,
            maxy = 4)

cybr_circos(d1 = rbind(Fluc_1_glmer, ChromosomeScale2), 
            d8 = NULL,
            #d8 = rbind(Fluc_8_glm, ChromosomeScale2), 
            peaklist1 = F_1_intpeaks_all, 
            #peaklist8 = F_8_intpeaks,
            maxy = 4)

cybr_circos(d1 = rbind(Zeocin_glm, ChromosomeScale2),
            d8 = rbind(Zeocin_8_glmer, ChromosomeScale2),
            peaklist1 = Z_1_intpeaks,
            peaklist8 = Z_8_intpeaks,
            maxy = 4)

cybr_circos(d1 = rbind(Zeocin_glm, ChromosomeScale2),
            d8 = rbind(Zeocin_8_glmer, ChromosomeScale2),
            peaklist1 = Z_1_intpeaks_all,
            peaklist8 = Z_8_intpeaks_all,
            maxy = 4)

```

```{r, fig.width=6, fig.height=6}
################################################################################
#MULT PEAKS
#Fluconazole
cybr_circos(d1 = rbind(Fluc_1_glmer, ChromosomeScale2), 
            d8 = rbind(Fluc_8_glm, ChromosomeScale2), 
            peaklist1 = f1p, 
            peaklist8 = f8p,
            maxy = 4, opacity = "02")

#CuSO4 (glmer)
cybr_circos(d1 = rbind(CuSO4_1_glmer, ChromosomeScale2), 
            d8 = rbind(CuSO4_8_glm, ChromosomeScale2), 
            peaklist1 = c1p, 
            peaklist8 = c8p,
            maxy = 4, opacity = "02")

cybr_circos(d1 = rbind(CuSO4_glm, ChromosomeScale2), 
            d8 = rbind(CuSO4_8_glm, ChromosomeScale2), 
            peaklist1 = c1p, 
            peaklist8 = c8p,
            maxy = 4, opacity = "02")

#CuSO4 glm A (the good experiment)
cybr_circos(d1 = rbind(CuSO4_glm, ChromosomeScale2), 
            d8 = rbind(CuSO4_8_glm, ChromosomeScale2), 
            peaklist1 = c1p, 
            peaklist8 = c8p,
            maxy = 4, opacity = "02")

#Zeocin
cybr_circos(d1 = rbind(Zeocin_glm, ChromosomeScale2), 
            d8 = rbind(Zeocin_8_glmer, ChromosomeScale2), 
            peaklist1 = z1p, 
            peaklist8 = z8p,
            maxy = 4, opacity = "02")


```

# Specific QTL Investigation
## Comparison of CSSI and CSS VIII

CuSO4 comparison 

```{r, fig.width=17,fig.height=8}

rbind(data.frame(CuSO4_8_glm, Exp = "CSS8"),
      data.frame(CuSO4_glm, Exp = "CSS1a"),
      data.frame(CuSO4_glm2, Exp = "CSS1b")) -> CUSO4b

max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))
max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS1a"]))
max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS1b"]))


CUSO4b %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","turquoise")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)


```

Trying to find how close the suspected QTL are...

```{r, warning=FALSE, message=FALSE, fig.width=17,fig.height=8}

KnownLoci <- data.frame(Gene = c("SSU1", "MAC1", "CTR1","CTR3", "CUP1"), 
                        CHROM = c("XVI", "XIII", "XVI", "XII", "VIII"), 
                        POS = c(373793, 317165, 786208, 947253, 212535))

URA3 <- data.frame(Gene = "Ura3",
                   CHROM = "V",
                   POS = 116167)

CUSO4b %>% ggplot(aes(x = POS, y = abs(summary), color = label))+
  # geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "pink", size =2) + 
  geom_vline(data = KnownLoci, aes(xintercept = POS), color = "firebrick", size = 1.2, linetype = "dashed")+
  geom_vline(data = URA3, aes(xintercept = POS), color = "gold", size = 1.2)+
  geom_line(size = 1) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","turquoise")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)

```

## Copper QTLs: Load in datasets

```{r, warning=FALSE, message=FALSE}
copperdata <- read.csv("CopperGenes_gdoc.csv")

copperdata %>% filter(StudyType == "QTL") %>% transmute(CHROM = as.character(as.roman(CHROM)), POS = StartLocation, Study = StudyType, Gene = Gene) -> copperdata2

dim(copperdata2)

SpecificCuSO4 <- filter(copperdata2, CHROM %in% c("V", "VII", "VIII", "XI", "XII", "XVI", "XV")) %>% arrange(CHROM, POS)

subset(SpecificCuSO4, CHROM == "XII")

actualQTL <- data.frame(Gene = c("VMA4", "MTL1", "CIC1","PTR2", "IRC15", "HSP60", "CUP1", "LEU9"), 
                        CHROM = factor(c("V", "VII", "VIII", "XI", "XVI", "XII", "VIII", "XV"), 
                                       levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI")), 
                        POS = c(100451, 539938, 211292, 617558, 519854, 664127, 212535, 523702)) 

actualQTL %>% arrange(CHROM)

filteredQTL <- copperdata2 %>% filter(CHROM %in%  c("V", "VII", "VIII", "XI", "XII", "XVI") == FALSE)
filteredQTL$CHROM <- factor(filteredQTL$CHROM,
                            levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI"))
```

```{r, fig.width = 16, fig.height = 6, eval = FALSE}
rbind(data.frame(CuSO4_CSS8_PeaksB, Exp = "CSS8"),
      data.frame(CuSO4Peaks_B, Exp = "CSS1a"),
      data.frame(CuSO42Peaks_B, Exp = "CSS1b")) -> CUSO4b_peaks

CUSO4b_peaks$CHROM
actualQTL$CHROM

CUSO4b %>% #filter(CHROM %in% actualQTL$CHROM) %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label))+
  #geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "lightblue", size =2) + 
  geom_vline(data = CUSO4b_peaks, aes(xintercept = POS), color = "lightblue", size =2) + 

  geom_vline(data = actualQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed")+
  
  geom_vline(data = filteredQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed", alpha = 0.3)+

  geom_line(size = 1) + 
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink", "gray","black","#7030A0")) +
  theme(legend.position = "none", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)

CUSO4b %>% #filter(CHROM %in% actualQTL$CHROM) %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label))+
  #geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "lightblue", size =2) + 
  geom_vline(data = CUSO4b_peaks, aes(xintercept = POS), color = "lightblue", size =2) + 

  geom_vline(data = actualQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed")+
  
  #geom_vline(data = filteredQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed", alpha = 0.3)+

  geom_line(size = 1) + 
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink", "gray","black","#7030A0")) +
  theme(legend.position = "none", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)

CUSO4b %>% #filter(CHROM %in% actualQTL$CHROM) %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label))+
  #geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "lightblue", size =2) + 
  geom_vline(data = CUSO4b_peaks, aes(xintercept = POS), color = "lightblue", size =2) + 

  #geom_vline(data = actualQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed")+
  
  #geom_vline(data = filteredQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed", alpha = 0.3)+

  geom_line(size = 1) + 
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink", "gray","black","#7030A0")) +
  theme(legend.position = "none", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)
```

Fluconazole Chr 1
```{r, fig.width = 16, fig.height=4}

FlucGenes <- data.frame(Gene = c("ERG11", "LSM2", "TIM12", "RPB7", "TCP1", "RSC6", "RPL32", "RSC9", "HSP150"),
                        CHROM = as.character(as.roman(c(8,2,2,4,4,3,2,13, 10))),
                        POS = c(120091, 170623, 427155, 1276654, 887232, 214994, 45978, 17064, 120449))

Fluc_glm %>% mutate(Sub = 1) %>% rbind(data.frame(Fluc_8_glm, Sub = 8), 
                                       data.frame(ChromosomeScale2, Sub = NA)) %>%
  filter(label == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(summary), color = as.factor(Sub))) + 
    #geom_vline(data = F_1_intpeaks, aes(xintercept = POS), color = "#7DB0B0", size = 2) +
  geom_vline(data = FlucGenes, aes(xintercept = POS), color = "firebrick", linetype = "dashed", size = 2) +
  geom_vline(data = engineered, aes(xintercept = POS_start), color = "gray", size = 2, alpha = 0.6) +
  geom_vline(data = engineered, aes(xintercept = POS_end), color = "gray", size = 2, alpha = 0.6) +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Fluc_glm$summary[Fluc_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("black",  "darkorange3", "gray")) + 
  #scale_color_manual(values = c("gray",  "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,8)

```

# Exploratory Plots


```{r, fig.height=6, fig.width=16}

rbind(data.frame(CuSO4_8_glm, Exp = "CuSO4_8"),
      #data.frame(CuSO4_glm, Exp = "CuSO4_1a"),
      #data.frame(CuSO4_glm2, Exp = "CuSO4_1b"),
      data.frame(CuSO4_1_glmer, Exp = "CuSO4_1"),
      data.frame(Fluc_8_glm, Exp = "Fluc_8"),
      data.frame(Fluc_glm, Exp = "Fluc_1"),

      data.frame(Zeocin_8_glmer, Exp = "Zeo_8"),
      data.frame(Zeocin_glm, Exp = "Zeo_1"))-> All_Experiments

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1, alpha = 0.8) + 
  scale_color_manual(values = c("black","darkorange3")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + ylim(0, 20) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1, alpha = 0.8) + 
  scale_color_manual(values = c("black","black", "black", "black")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + ylim(0, 20) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")
```

## Comparison of Additive QTL in all traits

```{r, fig.height=6, fig.width=16, eval = FALSE}

rbind(data.frame(CuSO4_8_glm, Exp = "CuSO4_8"),
      data.frame(CuSO4_glm, Exp = "CuSO4_1a"),
      data.frame(CuSO4_glm2, Exp = "CuSO4_1b"),
      data.frame(Fluc_8_glm, Exp = "Fluc_8"),
      data.frame(Fluc_glm, Exp = "Fluc_1"),
      data.frame(H2O2_glmer, Exp = "H2O2_1"),
      data.frame(H2O2_8_glmer, Exp = "H2O2_8"),
      data.frame(Zeocin_8_glmer, Exp = "Zeo_8"),
      data.frame(Zeocin_glm, Exp = "Zeo_1"))-> All_Experiments

All_Experiments %>% filter(label == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(summary), color = Exp)) + geom_line(size = 1) + 
  scale_color_manual(values = c("turquoise", "turquoise3", "turquoise2", "firebrick", "firebrick2", "orchid", "darkorchid", "gold", "darkorange")) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = Selection, linetype = FixedChr)) + geom_line(size = 1) + 
  scale_color_manual(values = c("turquoise3","firebrick2", "orchid", "darkorange")) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1, alpha = 0.8) + 
  scale_color_manual(values = c("black","black", "black", "darkorange3")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + ylim(0, 15) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1, alpha = 0.8) + 
  scale_color_manual(values = c("black","black", "black", "black")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + ylim(0, 15) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  filter(Selection == "CuSO4") %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1) + 
  scale_color_manual(values = c("black","#00000090", "red3")) +
  facet_grid(~CHROM, scales = "free", space = "free") + ylim(0, 15) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("CuSO4 Additive QTL") + ylab("Bulk Z Score")


All_Experiments %>% filter(label == "Interaction") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 0.8, alpha = 0.8) + 
  scale_color_manual(values = c("#24588A","#24588A", "#24588A", "#ED7B01")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Interaction Z Score")
```



```{r}
Albert_eQTLInteractors <- data.frame(Gene = c("HAP1", "MKT1", "FLO8", "FLO11"),
                                     CHROM = c(12, 14, 5, 9),
                                     POS = c(646415, 467131, 375215, 389572))
```


