---
title: "Figures 2023"
author: "Cassandra Buzby"
date: "3/9/2023"
output:
  html_document:
    df_print: paged
    code_folding: hide
  html_notebook:
    code_folding: hide
---


```{r, warning=FALSE, message=FALSE}
#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)
library(scales)

#library(cybrBSA)

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 20)))

glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H","H","H", "H", 
                                             "L", "L","L", "L", 
                                             "H", "H","H", "H", 
                                             "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", 
                                             "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W",
                                             "O", "W", "O", "W", 
                                             "O", "W", "O", "W",
                                             "O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", 
                                          "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                                     HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

cybr2_rollmean <- function(dataframe){
  dataframe %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollmean(value, n = 100))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount)
}

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

#ChromosomeScale$CHROM

```

# Initial Scripts
## Load in Data

```{r, warning=FALSE, message=FALSE}
#Load Stuff
CSSI_CuSO4 <- readRDS("Data/CuSO4_1_cybr2.rds")
CSSI_CuSO4_2 <- readRDS("Data/CuSO4_2_cybr2.rds")
CSSI_Fluc <- readRDS("Data/Fluconazole_1_cybr2.rds")
CSSI_Zeocin <- readRDS("Data/Zeocin_cybr2.rds")

#CSS 8 Stuff
CSSVIII_CuSO4 <- readRDS("Data/CuSO4_CSS8_cybr2.rds")
```

## Non-Smoothed Data

```{r, warning=FALSE}

CSSI_CuSO4_2%>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = value) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_CuSO4_2_unsmoothed

CSSI_CuSO4 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = value) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_CuSO4_unsmoothed

CSSI_Fluc %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = value) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_Fluc_unsmoothed

CSSI_Zeocin %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = value) %>% na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_Zeocin_unsmoothed

#Merge together and rbind
CSSI_Zeocin_unsmoothed %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads")  %>% filter(Bulk != "NA") %>% 
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CSSI_Zeocin_LODu

CSSI_CuSO4_2_unsmoothed %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads")  %>% filter(Bulk != "NA") %>% 
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CSSI_CuSO4_2_LODu

CSSI_Fluc_unsmoothed %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CSSI_Fluc_LODu

CSSI_CuSO4_unsmoothed %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent","Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) %>% 
  transmute(CHROM = CHROM, POS = POS, Bulk = Bulk, Parent = Parent, Rep = NA, Oak = Oak, Wine = Wine, LOD = LOD) -> CSSI_CuSO4_LODu

rbind(data.frame(CSSI_CuSO4_LODu, Dataset = "CuSO4_1"),
      data.frame(CSSI_CuSO4_2_LODu, Dataset = "CuSO4_2"),
      data.frame(CSSI_Fluc_LODu, Dataset = "Fluconazole"),
      data.frame(CSSI_Zeocin_LODu, Dataset = "Zeocin")) -> AllLODs_unsmoothed

#Plot to see odds ratio
AllLODs_unsmoothed %>% filter(CHROM == "VIII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  #filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_point(alpha = 0.2) + 
  facet_grid(~Dataset, scales = "free", space = "free") + 
  scale_color_manual(values = c("#345F6F",  #CuSO4_1
                                        # "turquoise", #CuSO4_2
                                      "black", #Unselected
                                      "firebrick", #Fluconazole
                                      "skyblue")) + #Zeocin
  ggtitle("")

AllLODs_unsmoothed %>% filter(CHROM == "XII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  #filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, alpha = Bulk, group = Flask, color = paste(Parent, Rep))) + 
  geom_point(size = 0.2) + 
  facet_grid(~Dataset, scales = "free", space = "free") #+ 
  #scale_color_manual(values = c("#345F6F",  #CuSO4_1
                                        # "turquoise", #CuSO4_2
  #                                     "black", #Unselected
  #                                     "firebrick", #Fluconazole
  #                                     "skyblue")) + #Zeocin
  # ggtitle("")


###############
unique(AllLODs_unsmoothed$Dataset)
AllLODs_unsmoothed %>% filter(CHROM == "XII" & Dataset == "CuSO4_2") %>% 
    mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 

  #filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, group = Flask, color = paste(Bulk, Parent, Rep))) +
  geom_point() +
  scale_color_manual(values = c("#345F6F",  #CuSO4_1
                                        # "turquoise", #CuSO4_2
                                      "black", #Unselected
                                      "firebrick", #Fluconazole
                                      "skyblue",
                                      "gray",  #CuSO4_1
                                      "gray", #CuSO4_2
                                      "gray", #Unselected
                                      "gray", #Fluconazole
                                      "gray")) 
  

```

## Correlation of unsmoothed data

```{r, warning=FALSE, message=FALSE}

AllLODs_unsmoothed %>% select(-Oak, -Wine) %>% pivot_wider(names_from = Bulk, values_from = LOD) %>% select(-Fluconazole, -Zeocin) -> CorrCuSO4

subset(CorrCuSO4, CHROM == as.character(as.roman(1)))

for(i in 1:16){
  chrdata <- subset(CorrCuSO4, CHROM == as.character(as.roman(i)))
  myplot <- ggplot(chrdata, aes(x = CuSO4, y = Dilute, color = Parent)) + geom_point(alpha = 0.5) + 
    geom_smooth(method='lm', formula= y~x, alpha = 0.3) + xlim(-4,4) + ylim(-4,4)+
    geom_abline(slope = 1, intercept = 0)+
    ggtitle(i)
  print(myplot)
}
```

It would be great to see what's not included in the 1:1 correlation between CuSO4 and Dilute so... see what it would be on its own?
```{r, warning=FALSE, message=FALSE}
chrdata <- subset(CorrCuSO4, CHROM == as.character(as.roman(1)))
chrdata %>% filter(is.infinite(CuSO4) == FALSE, is.infinite(Dilute) == FALSE) -> chrdata

chr1lm <- lm(formula = chrdata$Dilute ~ chrdata$CuSO4)
summary(chr1lm)

CorrCuSO4 %>% filter(CHROM != 1) %>% 
  ggplot(aes(x = POS, y = CuSO4 - Dilute, color = Parent)) + 
  geom_point() + facet_grid(~CHROM, scales = "free")
```


## Smooth Data

```{r, warning=FALSE, message=FALSE}
#change n for window size and FUN for mean vs median smoothing
CSSI_CuSO4 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollapply(value, n = 200, FUN = median, align = "center"))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_CuSO4_cybr2

CSSI_CuSO4_2 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollapply(value, n = 200, FUN = median, align = "center"))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_CuSO4_2_cybr2

CSSI_Fluc %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollapply(value, n = 200, FUN = median, align = "center"))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_Fluc_cybr2

CSSI_Zeocin %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollapply(value, n = 200, FUN = median, align = "center"))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_Zeocin_cybr2

```

Combine datasets for plotting

```{r, warning=FALSE, message=FALSE}
CSSI_Zeocin_cybr2 %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CSSI_Zeocin_LOD


CSSI_Fluc_cybr2 %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CSSI_Fluc_LOD

CSSI_CuSO4_cybr2 %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent","Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) %>% transmute(CHROM = CHROM, POS = POS, Bulk = Bulk, Parent = Parent, Rep = NA, Oak = Oak, Wine = Wine, LOD = LOD) -> CSSI_CuSO4_LOD

CSSI_CuSO4_2_cybr2 %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) %>% transmute(CHROM = CHROM, POS = POS, Bulk = Bulk, Parent = Parent, Rep = Rep, Oak = Oak, Wine = Wine, LOD = LOD) -> CSSI_CuSO4_2_LOD

rbind(data.frame(CSSI_CuSO4_LOD, Dataset = "CuSO4_1"),
      data.frame(CSSI_CuSO4_2_LOD, Dataset = "CuSO4_2"),
      data.frame(CSSI_Fluc_LOD, Dataset = "Fluconazole"),
      data.frame(CSSI_Zeocin_LOD, Dataset = "Zeocin")) -> AllLODs

```


Plot for individual chromosome

```{r, warning=FALSE, message=FALSE}
AllLODs %>% filter(CHROM == "XII") %>% mutate(BulkBin = "Selected") -> tempAllLods
tempAllLods$BulkBin[tempAllLods$Bulk == "Dilute"] <- "Dilute"

tempAllLods %>% filter(Bulk != "NA") %>% 
  ggplot(aes(x = POS, y = LOD, alpha = BulkBin, color = paste(Parent, Rep, sep = "_"))) + 
  geom_line(size = 1) +
      ylab("") + xlab("")+
  geom_vline(xintercept = c(556591, 573671)) +
  geom_vline(xintercept = c(451575,489469), color = "orchid") +
  facet_grid(~Dataset, scales = "free") +
  #scale_color_manual(values = c("gray","black","lightblue", "firebrick")) + 
  theme(legend.position = "bottom") +
  ggtitle("CuSO4 Chr XIV")

tempAllLods[tempAllLods$LOD == min(tempAllLods$LOD),] %>% arrange(POS)

tempAllLods %>% filter(Bulk != "NA") %>% filter(POS > 556591 & POS < 573671) %>%
  ggplot(aes(x = POS, y = LOD, alpha = BulkBin, color = paste(Parent, Rep, sep = "_"))) + 
  geom_line(size = 1) +
      ylab("") + xlab("")+
  geom_vline(xintercept = c(556591, 573671)) +
  facet_grid(~Dataset, scales = "free") +
  #scale_color_manual(values = c("gray","black","lightblue", "firebrick")) + 
  theme(legend.position = "bottom") +
  ggtitle("CuSO4 Chr XIV")


###################################################

AllLODs %>% filter(CHROM == "VIII") %>% mutate(BulkBin = "Selected") -> tempAllLods8
tempAllLods8$BulkBin[tempAllLods8$Bulk == "Dilute"] <- "Dilute"

tempAllLods8 %>% filter(Bulk != "NA" & Dataset == "CuSO4_2") %>% 
  ggplot(aes(x = POS, y = LOD, alpha = BulkBin, color = paste(Parent, Rep, sep = "_"))) + 
  geom_line(size = 1) +
      ylab("") + xlab("")+
  # geom_vline(xintercept = c(556591, 573671)) +
  # geom_vline(xintercept = c(451575,489469), color = "orchid") +
  facet_grid(~Dataset, scales = "free") +
  #scale_color_manual(values = c("gray","black","lightblue", "firebrick")) + 
  theme(legend.position = "bottom") +
  ggtitle("CuSO4 Chr VIII")

#tempAllLods[tempAllLods$LOD == min(tempAllLods$LOD),] %>% arrange(POS)

```
Plot smoothed and unsmoothed data on the same axes

```{r, fig.width=7, fig.height=6}

AllLODs_unsmoothed %>% filter(CHROM == "VII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  #filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_point(alpha = 0.2, size = 0.8) + ylim(-2.5,2.5)+
      ylab("") + xlab("")+
  facet_grid(~Dataset, scales = "free", space = "free") + 
  scale_color_manual(values = c("#345F6F","black","firebrick", "#FFB05C")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("")

AllLODs %>% filter(CHROM == "VII") %>% 
  filter(Bulk != "NA") %>%
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_point(alpha = 0.2, size = 0.8) + ylim(-2.5,2.5)+
      ylab("") + xlab("")+
  facet_grid(~Dataset, scales = "free", space = "free") + 
  scale_color_manual(values = c("#345F6F","black","firebrick", "#FFB05C", "gray")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("")

```

```{r, fig.width=16,fig.height=5}

AllLODs %>% filter(CHROM != "I", CHROM != "M") %>% 
  filter(Bulk != "NA") %>%
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_line(size = 1.5) + 
      ylab("") + xlab("")+
  facet_grid(~CHROM, scales = "free", space = "free") + 
  scale_color_manual(values =  c("#345F6F", #Cu
                                        "black", #Dilute
                                        "firebrick", #Fluc
                                        "#FFB05C")) + #Zeo
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("")
```


## GLM Data plots

### Calling Peaks (Interactions) - important to reformat data
Load data

```{r, warning=FALSE, message=FALSE}
Zeocin_glm<- readRDS("Data/Zeocin_glm.rds")
CuSO4_glm <- readRDS("Data/CuSO4_1_glm.rds")
CuSO4_glm2 <- readRDS("Data/CuSO4_2_glm.rds")
Fluc_glm <- readRDS("Data/Fluconazole_1_glm.rds")

#CSS 8
CuSO4_8_glm <- readRDS("Data/CuSO4_CSS8_glm.rds")
CuSO4_8_c5_glm <- readRDS("Data/CuSO4_CSS8_C5norep_glm.rds")
CuSO4_8_D2_glm <- readRDS("Data/CuSO4_CSS8_D2_norep_glm.rds")
Fluc_8_glm <- readRDS("Data/Fluc_CSS8_glm.rds")

```

Reformat data and "call" interaction peaks for plots

```{r, fig.width=17,fig.height=5}

Zeocin_glm$label <- factor(Zeocin_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
Fluc_glm$label <- factor(Fluc_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_glm$label <- factor(CuSO4_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_glm2$label <- factor(CuSO4_glm2$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

Zeocin_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Zeocin_glm
Fluc_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Fluc_glm
CuSO4_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> CuSO4_glm
CuSO4_glm2 %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> CuSO4_glm2

#New CSS8 Ones - add back Chr I
Fluc_8_glm$label <- factor(Fluc_8_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_8_c5_glm$label <- factor(CuSO4_8_c5_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_8_D2_glm$label <- factor(CuSO4_8_D2_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_8_glm$label <- factor(CuSO4_8_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

Fluc_8_glm %>% filter(CHROM != "M", label != "intercept") -> Fluc_8_glm
CuSO4_8_c5_glm %>% filter(CHROM != "M", label != "intercept") -> CuSO4_8_c5_glm
CuSO4_8_D2_glm %>% filter(CHROM != "M", label != "intercept") -> CuSO4_8_D2_glm
CuSO4_8_glm %>% filter(CHROM != "M", label != "intercept") -> CuSO4_8_glm

################################################################################
#Peaks for INTERACTIONS
max(abs(Zeocin_glm$summary[Zeocin_glm$CHROM == "III"]))
Zeocin_glm %>% filter(label == "Interaction") %>% filter(abs(summary) > 0.7) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> ZeocinPeaks

for(i in 1:length(ZeocinPeaks$peak)){
  ZeocinPeaks$POS[i] <- Zeocin_glm$POS[Zeocin_glm$label == "Interaction" & abs(Zeocin_glm$summary) == ZeocinPeaks$peak[i] & Zeocin_glm$CHROM == ZeocinPeaks$CHROM[i]]

}

max(abs(Fluc_glm$summary[Fluc_glm$CHROM == "III"]))
Fluc_glm %>% filter(label == "Interaction") %>% filter(abs(summary) > 1.124545) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> FlucPeaks

for(i in 1:length(FlucPeaks$peak)){
  FlucPeaks$POS[i] <- Fluc_glm$POS[Fluc_glm$label == "Interaction" & abs(Fluc_glm$summary) == FlucPeaks$peak[i] & Fluc_glm$CHROM == FlucPeaks$CHROM[i]]

}

#CuSO4
max(abs(CuSO4_glm$summary[CuSO4_glm$CHROM == "III"]))
CuSO4_glm %>% filter(label == "Interaction") %>% filter(abs(summary) > 1.047225) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> CuSO4Peaks

for(i in 1:length(CuSO4Peaks$peak)){
  CuSO4Peaks$POS[i] <- CuSO4_glm$POS[CuSO4_glm$label == "Interaction" & abs(CuSO4_glm$summary) == CuSO4Peaks$peak[i] & CuSO4_glm$CHROM == CuSO4Peaks$CHROM[i]]

}

#CuSO4_2
max(abs(CuSO4_glm2$summary[CuSO4_glm2$CHROM == "III"]))
CuSO4_glm2 %>% filter(label == "Interaction") %>% filter(abs(summary) > 1.21009) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> CuSO42Peaks

for(i in 1:length(CuSO42Peaks$peak)){
  CuSO42Peaks$POS[i] <- CuSO4_glm2$POS[CuSO4_glm2$label == "Interaction" & abs(CuSO4_glm2$summary) == CuSO42Peaks$peak[i] & CuSO4_glm2$CHROM == CuSO42Peaks$CHROM[i]]

}
```

New CSS8 Cutoffs
```{r, fig.width=17,fig.height=5}
#CuSO4 CSS8
max(abs(CuSO4_8_glm$summary[CuSO4_8_glm$CHROM == "III"]))
CuSO4_8_glm %>% filter(label == "Interaction") %>% filter(abs(summary) > 1.472239) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> CuSO4_CSS8_Peaks

for(i in 1:length(CuSO4_CSS8_Peaks$peak)){
  CuSO4_CSS8_Peaks$POS[i] <- CuSO4_8_glm$POS[CuSO4_8_glm$label == "Interaction" & abs(CuSO4_8_glm$summary) == CuSO4_CSS8_Peaks$peak[i] & CuSO4_8_glm$CHROM == CuSO4_CSS8_Peaks$CHROM[i]]
}


#CuSO4 c5 CSS8
max(abs(CuSO4_8_c5_glm$summary[CuSO4_8_c5_glm$CHROM == "III"]))
CuSO4_8_c5_glm %>% filter(label == "Interaction") %>% filter(abs(summary) > 1.510061) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> CuSO4_CSS8_c5_Peaks

for(i in 1:length(CuSO4_CSS8_c5_Peaks$peak)){
  CuSO4_CSS8_c5_Peaks$POS[i] <- CuSO4_8_c5_glm$POS[CuSO4_8_c5_glm$label == "Interaction" & abs(CuSO4_8_c5_glm$summary) == CuSO4_CSS8_c5_Peaks$peak[i] & CuSO4_8_c5_glm$CHROM == CuSO4_CSS8_c5_Peaks$CHROM[i]]
}

#CuSO4 D2 CSS8

max(abs(CuSO4_8_D2_glm$summary[CuSO4_8_D2_glm$CHROM == "III"]))
CuSO4_8_D2_glm %>% filter(label == "Interaction") %>% filter(abs(summary) > 1.27264) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> CuSO4_CSS8_D2_Peaks

for(i in 1:length(CuSO4_CSS8_D2_Peaks$peak)){
  CuSO4_CSS8_D2_Peaks$POS[i] <- CuSO4_8_D2_glm$POS[CuSO4_8_D2_glm$label == "Interaction" & abs(CuSO4_8_D2_glm$summary) == CuSO4_CSS8_D2_Peaks$peak[i] & CuSO4_8_D2_glm$CHROM == CuSO4_CSS8_D2_Peaks$CHROM[i]]
}

#Fluc CSS8
max(abs(Fluc_8_glm$summary[Fluc_8_glm$CHROM == "III"]))
Fluc_8_glm %>% filter(label == "Interaction") %>% filter(abs(summary) > 1.644711) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> Fluc_CSS8_Peaks

for(i in 1:length(Fluc_CSS8_Peaks$peak)){
  Fluc_CSS8_Peaks$POS[i] <- Fluc_8_glm$POS[Fluc_8_glm$label == "Interaction" & abs(Fluc_8_glm$summary) == Fluc_CSS8_Peaks$peak[i] & Fluc_8_glm$CHROM == Fluc_CSS8_Peaks$CHROM[i]]
}

```


### Call peaks for (bulk effects)

CSS8 Data
```{r, warning=FALSE, message=FALSE}
#CuSO4 CSS8
CuSO4_8_glm %>% filter(label == "Bulk") %>% filter(summary > 1.472239 | summary < -1.472239) %>% 
  group_by(CHROM) %>% summarize(peakup = max(summary), peakdown = min(summary)) %>% pivot_longer(c(peakup, peakdown), values_to = "peak")  %>% 
  filter((name == "peakup" & peak > 1.472239) | (name == "peakdown" & peak < -1.472239))  -> CuSO4_CSS8_PeaksB

for(i in 1:length(CuSO4_CSS8_PeaksB$peak)){
  CuSO4_CSS8_PeaksB$POS[i] <- CuSO4_8_glm$POS[CuSO4_8_glm$label == "Bulk" & CuSO4_8_glm$summary == CuSO4_CSS8_PeaksB$peak[i] & CuSO4_8_glm$CHROM == CuSO4_CSS8_PeaksB$CHROM[i]]
}

#CuSO4 c5 CSS8
CuSO4_8_c5_glm %>% filter(label == "Bulk") %>% filter(summary > 1.510061 | summary < -1.510061) %>% 
  group_by(CHROM) %>% summarize(peakup = max(summary), peakdown = min(summary)) %>% pivot_longer(c(peakup, peakdown), values_to = "peak")  %>% 
  filter((name == "peakup" & peak > 1.510061) | (name == "peakdown" & peak < -1.510061))  -> CuSO4_CSS8_c5_Peaks_B

for(i in 1:length(CuSO4_CSS8_c5_Peaks_B$peak)){
  CuSO4_CSS8_c5_Peaks_B$POS[i] <- CuSO4_8_c5_glm$POS[CuSO4_8_c5_glm$label == "Bulk" & CuSO4_8_c5_glm$summary == CuSO4_CSS8_c5_Peaks_B$peak[i] & CuSO4_8_c5_glm$CHROM == CuSO4_CSS8_c5_Peaks_B$CHROM[i]]
}


#CuSO4 D2 CSS8
CuSO4_8_D2_glm %>% filter(label == "Bulk") %>% filter(summary > 1.27264 | summary < -1.27264) %>% 
  group_by(CHROM) %>% summarize(peakup = max(summary), peakdown = min(summary)) %>% pivot_longer(c(peakup, peakdown), values_to = "peak")  %>% 
  filter((name == "peakup" & peak > 1.27264) | (name == "peakdown" & peak < -1.27264))  -> CuSO4_CSS8_D2_Peaks_B

for(i in 1:length(CuSO4_CSS8_D2_Peaks_B$peak)){
  CuSO4_CSS8_D2_Peaks_B$POS[i] <- CuSO4_8_D2_glm$POS[CuSO4_8_D2_glm$label == "Bulk" & CuSO4_8_D2_glm$summary == CuSO4_CSS8_D2_Peaks_B$peak[i] & CuSO4_8_D2_glm$CHROM == CuSO4_CSS8_D2_Peaks_B$CHROM[i]]
}


#Fluc CSS8
Fluc_8_glm %>% filter(label == "Bulk") %>% filter(summary > 1.644711 | summary < -1.644711) %>% 
  group_by(CHROM) %>% summarize(peakup = max(summary), peakdown = min(summary)) %>% pivot_longer(c(peakup, peakdown), values_to = "peak") %>% 
 filter((name == "peakup" & peak > 1.644711) | (name == "peakdown" & peak < -1.644711))  -> Fluc_CSS8_Peaks_B

for(i in 1:length(Fluc_CSS8_Peaks_B$peak)){
  Fluc_CSS8_Peaks_B$POS[i] <- Fluc_8_glm$POS[Fluc_8_glm$label == "Bulk" & Fluc_8_glm$summary == Fluc_CSS8_Peaks_B$peak[i] & Fluc_8_glm$CHROM == Fluc_CSS8_Peaks_B$CHROM[i]]
}


```

CSS1 Data

```{r, warning = FALSE}
#Zeocin
Zeocin_glm %>% filter(label == "Bulk") %>% filter(summary > 0.7 | summary < -0.7) %>% group_by(CHROM) %>% 
  summarize(peakup = max(summary), peakdown = min(summary)) %>% pivot_longer(c(peakup, peakdown), values_to = "peak") %>% 
 filter((name == "peakup" & peak > 0.7) | (name == "peakdown" & peak < -0.7))  -> ZeocinPeaks_B

for(i in 1:length(ZeocinPeaks_B$peak)){
  ZeocinPeaks_B$POS[i] <- Zeocin_glm$POS[Zeocin_glm$label == "Bulk" & Zeocin_glm$summary == ZeocinPeaks_B$peak[i] & Zeocin_glm$CHROM == ZeocinPeaks_B$CHROM[i]]

}

#Fluconazole
Fluc_glm %>% filter(label == "Bulk") %>% filter(summary > 1.124545 | summary < -1.124545) %>% group_by(CHROM) %>% summarize(peakup = max(summary), peakdown = min(summary)) %>% pivot_longer(c(peakup, peakdown), values_to = "peak")  %>% filter((name == "peakup" & peak > 1.124545) | (name == "peakdown" & peak < -1.124545))  -> FlucPeaks_B

for(i in 1:length(FlucPeaks_B$peak)){
  FlucPeaks_B$POS[i] <- Fluc_glm$POS[Fluc_glm$label == "Bulk" & Fluc_glm$summary == FlucPeaks_B$peak[i] & Fluc_glm$CHROM == FlucPeaks_B$CHROM[i]]
}


#CuSO4
CuSO4_glm %>% filter(label == "Bulk") %>% filter(summary > 1.047225 | summary < -1.047225) %>% 
  group_by(CHROM) %>% 
  summarize(peakup = max(summary), peakdown = min(summary)) %>% 
  pivot_longer(c(peakup, peakdown), values_to = "peak")  %>% filter((name == "peakup" & peak > 1.047225) | (name == "peakdown" & peak < -1.047225)) -> CuSO4Peaks_B

for(i in 1:length(CuSO4Peaks_B$peak)){
  CuSO4Peaks_B$POS[i] <- CuSO4_glm$POS[CuSO4_glm$label == "Bulk" & CuSO4_glm$summary == CuSO4Peaks_B$peak[i] & CuSO4_glm$CHROM == CuSO4Peaks_B$CHROM[i]]

}


#CuSO4_2
CuSO4_glm2 %>% filter(label == "Bulk") %>% filter(summary > 1.21009 | summary < -1.21009) %>% group_by(CHROM) %>% summarize(peakup = max(summary), peakdown = min(summary)) %>% pivot_longer(c(peakup, peakdown), values_to = "peak")  %>% filter((name == "peakup" & peak > 1.21009) | (name == "peakdown" & peak < -1.21009))  -> CuSO42Peaks_B

for(i in 1:length(CuSO42Peaks_B$peak)){
  CuSO42Peaks_B$POS[i] <- CuSO4_glm2$POS[CuSO4_glm2$label == "Bulk" & CuSO4_glm2$summary == CuSO42Peaks_B$peak[i] & CuSO4_glm2$CHROM == CuSO42Peaks_B$CHROM[i]]

}

```

### Original Plots
Plot smoothed Z-scores with peaks

```{r, fig.width = 16, fig.height=6}

#Plots
Zeocin_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label, linetype = label)) +   
  geom_vline(data = ZeocinPeaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 1) + 
  geom_hline(aes(yintercept = max(abs(Zeocin_glm$summary[Zeocin_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free") + 
  scale_color_manual(values = c("gray", "gray", "black","#FFB05C")) + 
  scale_linetype_manual(values = c("dotted", "solid", "solid", "solid")) +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 
  

Fluc_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
  geom_vline(data = FlucPeaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 1) + 
  geom_hline(aes(yintercept = max(abs(Fluc_glm$summary[Fluc_glm$CHROM == "III"]))), linetype = "dashed") +

    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("gray", "black","firebrick")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 

CuSO4_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = CuSO4Peaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 1) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_glm$summary[CuSO4_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("gray","black","turquoise")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 

CuSO4_glm2 %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = CuSO42Peaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 1) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_glm$summary[CuSO4_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("red","gray",  "black","turquoise")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 
```

```{r, fig.width = 16, fig.height=6}
CuSO4_8_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = CuSO4_CSS8_Peaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 0.8) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_8_glm$summary[CuSO4_8_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink","gray",  "black","turquoise")) + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,10)

CuSO4_8_c5_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = CuSO4_CSS8_c5_Peaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 0.8) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_8_c5_glm$summary[CuSO4_8_c5_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 CSS8 low dose only (D2)")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("gray",  "black","turquoise")) + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())+ ylim(0,10)

CuSO4_8_D2_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = CuSO4_CSS8_D2_Peaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 0.8) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_8_D2_glm$summary[CuSO4_8_D2_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 CSS8 10% Doses (D2)")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("gray",  "black","turquoise")) + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())+ ylim(0,10)

```
```{r, fig.width = 16, fig.height=6}
Fluc_8_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = Fluc_CSS8_Peaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Fluc_8_glm$summary[Fluc_8_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink","gray",  "black","firebrick")) + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,10)
```


CuSO4 comparison - missing two chromosomes

```{r, fig.width=17,fig.height=5}

rbind(data.frame(CuSO4_glm, Exp = "1"),
      data.frame(CuSO4_glm2, Exp = "2")) -> CUSO4ALL

max(abs(CUSO4ALL$summary[CUSO4ALL$CHROM == "III" & CUSO4ALL$Exp == 1]))
max(abs(CUSO4ALL$summary[CUSO4ALL$CHROM == "III" & CUSO4ALL$Exp == 2]))


CUSO4ALL %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(CUSO4ALL$summary[CUSO4ALL$CHROM == "III" & CUSO4ALL$Exp == 2]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("pink3", "gray","black","#345F6F")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 


```


## Combining GLM and LOD scores

```{r, fig.width=8,fig.height=4}

# AllLODs_unsmoothed %>% filter(CHROM == "VII") %>% 
#   mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
#   #filter(Dataset == "Fluconazole") %>%
#   ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
#   geom_point(alpha = 0.2, size = 0.8) + ylim(-2.5,2.5)+
#       ylab("") + xlab("")+
#   facet_grid(~Dataset, scales = "free", space = "free") + 
#   scale_color_manual(values = c("#345F6F","black","firebrick", "#FFB05C")) + 
#   theme(legend.position = "none", axis.text.x=element_blank(),
#             axis.ticks.x=element_blank()) +
#   ggtitle("")

AllLODs_unsmoothed %>% filter(CHROM == "VII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) + 
  
  geom_point(aes(x = 940468, y = -0.7997569), alpha = 0.1, size = 3, color = "pink") +

  ylab("") + xlab("")+geom_point(size = 0.5) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
  scale_x_continuous(expand = expansion(0)) + 
  scale_y_continuous(expand = expansion(0)) + 
  ylim(-0.8,0.8)+
  ggtitle("") -> chr7_unsmoothed


AllLODs %>% filter(CHROM == "VII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) +
  ylab("") + xlab("")+geom_line(size = 2) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ylim(-0.8,0.8)+
  ggtitle("") -> chr7_glm

Fluc_glm %>% filter(CHROM == "VII") %>%
  ggplot(aes(x = POS, y = -1*(summary), color = label)) + geom_line(size = 2) + 
  ylab("") + xlab("")+geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("gray", "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ylim(-04,4)+
  ggtitle("") -> chr7_lod

plot_grid(chr7_unsmoothed, chr7_glm, chr7_lod, cols = 3)

```
Finding one dot

```{r, warning=FALSE, message=FALSE}
AllLODs_unsmoothed %>% filter(CHROM == "VII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "Fluconazole") %>%
  
  filter(POS > 850000 & POS < 1000000) %>%
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) +
    geom_point(aes(x = 940468, y = -0.7997569), alpha = 0.1, size = 3, color = "pink") +
  ylab("") + xlab("")+geom_point(size = 0.5) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  scale_x_continuous(expand = expansion(0)) + 
  scale_y_continuous(expand = expansion(0)) + 
  ylim(-0.8,0)+
  ggtitle("") 


AllLODs_unsmoothed %>% filter(CHROM == "VII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "Fluconazole") %>%
  
  filter(POS > 850000 & POS < 1000000 & Parent == "OakI" & Bulk == "Fluconazole") %>%
  filter(LOD < -0.75)
```

```{r, fig.width=2.8,fig.height=3}

AllLODs_unsmoothed %>% filter(CHROM == "VIII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "CuSO4_1") %>%
  ggplot(aes(x = POS, y = abs(LOD), color = Parent, group = Flask, alpha = Bulk)) + 
  geom_vline(aes(xintercept = 212535), size = 3, color = "turquoise") +
  ylab("") + xlab("")+geom_point(size = 1) + 
  geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
  scale_x_continuous(expand = expansion(0)) + 
  scale_y_continuous(expand = expansion(0)) + ylim(0,4)+
  ggtitle("")

```

```{r, fig.width=8,fig.height=4}

AllLODs$Bulk[AllLODs$Bulk == "CuSO4"] <- "Selected"
AllLODs_unsmoothed$Bulk[AllLODs_unsmoothed$Bulk == "CuSO4"] <- "Selected"

AllLODs_unsmoothed %>% filter(CHROM == "VIII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "CuSO4_1") %>%
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) + 
  ylab("") + xlab("")+geom_point(size = 0.5) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
  scale_x_continuous(expand = expansion(0)) + 
  scale_y_continuous(expand = expansion(0)) +
  ggtitle("")

AllLODs_unsmoothed %>% filter(CHROM == "VIII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "CuSO4_1") %>%
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) + 
  ylab("") + xlab("")+geom_point(size = 0.5) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
  scale_x_continuous(expand = expansion(0)) + 
  scale_y_continuous(expand = expansion(0)) + 
  ylim(-3,0.5)+
  ggtitle("") -> chr8_unsmoothed


AllLODs %>% filter(CHROM == "VIII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "CuSO4_1") %>%
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) + 
  ylab("") + xlab("")+geom_line(size = 2) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ylim(-3,0.5)+
  ggtitle("") -> chr8_glm

CuSO4_glm %>% filter(CHROM == "VIII") %>%
  ggplot(aes(x = POS, y = -1*(summary), color = label)) + geom_line(size = 2) + 
  ylab("") + xlab("") + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  geom_vline(aes(xintercept = 212535), color = "#00ADD0", alpha = 0.3, size = 2) +
  scale_color_manual(values = c("gray", "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ylim(-12, 2) +
  ggtitle("") -> chr8_lod


plot_grid(chr8_unsmoothed, chr8_glm, chr8_lod, cols = 3)
```

## Focus on Chr III as Baseline

Plot raw data log(oak/wine)

```{r, warning=FALSE, message=FALSE}

CScale <- ChromosomeScale %>% 
  transmute(CHROM = CHROM, POS = POS, Bulk = NA, Parent = NA, 
            Rep = NA, Oak = NA, Wine = NA, LOD = NA, Dataset = "CuSO4")

AllLODs_unsmoothed %>% rbind(CScale)  %>% filter(CHROM == "III") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  #filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_point(alpha = 0.2) + 
  geom_vline(xintercept = c(198671,201177)) +
  facet_grid(~Dataset, scales = "free", space = "free") + 
  scale_color_manual(values = c("#345F6F", "black", "firebrick", "#FFB05C")) + 
  ggtitle("") + ylab("") +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())
```
Plot overall log(alleles), with and without Chr I

```{r, fig.width=16,fig.height=5}

AllLODs %>% filter(CHROM != "I", CHROM != "M") %>% 
  filter(Bulk != "NA") %>%
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_line(size = 1.5) + 
      ylab("") + xlab("")+
  facet_grid(~CHROM, scales = "free", space = "free") + 
  scale_color_manual(values =  c("black","firebrick","#345F6F", "#FFB05C", "gray")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("")

#Sanity check for which direction each allele is; this has wine down apparently

AllLODs %>% filter(CHROM != "M") %>% 
  filter(Bulk != "NA") %>%
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_line(size = 1.5) + 
      ylab("") + xlab("")+
  facet_grid(~CHROM, scales = "free", space = "free") + 
  scale_color_manual(values =  c("black","firebrick","#345F6F", "#FFB05C", "gray")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("")

```
Plot smoothed (200 SNPs, median) log(alleles)

```{r, warning=FALSE, message=FALSE}

AllLODs %>% filter(CHROM == "III") %>% 
  filter(Bulk != "NA") %>%
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_line(size = 1.5) + 
  geom_hline(yintercept = 0, linetype = "dashed")+
      ylab("") + xlab("")+
  ylim(c(-2,1)) +
  #facet_grid(~CHROM, scales = "free", space = "free") + 
  scale_color_manual(values =  c("black","firebrick","#345F6F", "#FFB05C")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("All Datasets") -> px

AllLODs %>% filter(CHROM == "III") %>% 
  filter(Bulk != "NA") %>%
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "Fluconazole") %>%
  
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) + 
  geom_line(size = 2) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
    scale_alpha_manual(values = c(0.5, 1)) +
  theme(legend.position = "none", axis.text.x=element_blank(),axis.ticks.x=element_blank(),
        text = element_text(size = 20)) +
  ylab("") + xlab("")+ggtitle("Fluconazole") -> py

plot_grid(px, py, ncol = 2)
```

Plot Chr III GLMs (to show effects without a specific bulk effect etc)

```{r, warning=FALSE, message=FALSE}

Fluc_glm %>% filter(CHROM == "III", label != "Rep") %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 2) + 
  geom_hline(aes(yintercept = 0), linetype = "dashed") +
  geom_hline(aes(yintercept = 1), linetype = "dashed") +

    scale_color_manual(values = c("gray", "black","firebrick")) + 
  ylim(c(0, 1.2))+
  ylab("") + xlab("") + ggtitle("Fluconazole") +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank(), axis.text.y = element_blank()) -> Fluc3

CuSO4_glm %>% filter(CHROM == "III", label != "Rep") %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 2) + 
  geom_hline(aes(yintercept = 0), linetype = "dashed") +
  geom_hline(aes(yintercept = 1), linetype = "dashed") +
  scale_color_manual(values = c("gray", "black","#345F6F")) + 
  ylab("") + xlab("") + ggtitle("CuSO4") +
    ylim(c(0, 1.2))+

  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank(), axis.text.y = element_blank()) -> CuSO43

Zeocin_glm %>% filter(CHROM == "III", label != "Rep") %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 2) + 
  geom_hline(aes(yintercept = 0), linetype = "dashed") +
  geom_hline(aes(yintercept = 1), linetype = "dashed") +
  scale_color_manual(values = c("gray", "black","#FFB05C")) + 
    ylim(c(0, 1.2))+
  ylab("") + xlab("") + ggtitle("Zeocin") +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank(), axis.text.y = element_blank()) -> Zeocin3

plot_grid(CuSO43, Fluc3, Zeocin3, ncol = 3)

```

## Comparison of CSSI and CSS VIII

CuSO4 comparison 

```{r, fig.width=17,fig.height=8}

rbind(data.frame(CuSO4_8_glm, Exp = "CSS8"),
      data.frame(CuSO4_glm, Exp = "CSS1a"),
      data.frame(CuSO4_glm2, Exp = "CSS1b")) -> CUSO4b

max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))
max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS1a"]))
max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS1b"]))


CUSO4b %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","turquoise")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)


```

Finding how close the peaks are to Cup-1

```{r, warning=FALSE, message=FALSE}
max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))

CUSO4_bog <- CUSO4b
CUSO4_bog$summary <- abs(CUSO4_bog$summary)
CUSO4b %>% filter(label == "Bulk") %>% group_by(Exp, CHROM) %>% 
  summarize(summary = max(abs(summary))) %>% merge(CUSO4_bog) -> CuSO4_BulkPeaks

CuSO4_BulkPeaks %>% select(-label) %>% ggplot(aes(x = CHROM, y = POS, color = Exp)) + geom_point()

CuSO4_BulkPeaks %>% select(-summary, -label) %>% 
  filter(CHROM %in% c("V", "VII", "VIII", "XI", "XII", "XVI")) %>%
  filter((CHROM == "V" & Exp == "CSS8") == FALSE) %>%
  filter((CHROM == "VIII" & Exp == "CSS8") == FALSE) -> newBulkPeaks
  
newBulkPeaks %>% pivot_wider(names_from = Exp, values_from = POS) %>% group_by(CHROM) %>%
  mutate(mean = mean(c(CSS1a, CSS1b, CSS8), na.rm = TRUE),
         sd = floor(sd(c(CSS1a, CSS1b, CSS8), na.rm = TRUE))) %>%
  mutate(CSS1a_diff = abs(floor(CSS1a - mean))) -> summaryofprecision_CuSO4



#To know what the means are
mean(summaryofprecision_CuSO4$CSS1a_diff, na.rm = TRUE)
#35951 bp apart from the mean

#Just Chr VIII
summaryofprecision_CuSO4 %>% filter(CHROM == "VIII")

#Chr X in CSS1a has 3 numbers
newBulkPeaks %>%
    dplyr::group_by(CHROM, Exp) %>%
    dplyr::summarise(n = dplyr::n(), .groups = "drop") %>%
    dplyr::filter(n > 1L)

#Pick one
CuSO4_BulkPeaks %>%
    dplyr::group_by(CHROM, label, Exp) %>%
  filter(Exp == "CSS1a", CHROM == "X")

#Plot
CUSO4b %>% ggplot(aes(x = POS, y = abs(summary), color = label))+
  geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "pink", size =2) + geom_line(size = 1) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","turquoise")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)

#To know what the loci are
summaryofprecision_CuSO4

```
Trying to find how close the suspected QTL are...
```{r, warning=FALSE, message=FALSE}
KnownLoci <- data.frame(Gene = c("SSU1", "MAC1", "CTR1","CTR3", "CUP1"), 
                        CHROM = c("XVI", "XIII", "XVI", "XII", "VIII"), 
                        POS = c(373793, 317165, 786208, 947253, 212535))

URA3 <- data.frame(Gene = "Ura3",
                   CHROM = "V",
                   POS = 116167)

CUSO4b %>% ggplot(aes(x = POS, y = abs(summary), color = label))+
  geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "pink", size =2) + 
  geom_vline(data = KnownLoci, aes(xintercept = POS), color = "firebrick", size = 1.2, linetype = "dashed")+
  geom_vline(data = URA3, aes(xintercept = POS), color = "gold", size = 1.2)+
  geom_line(size = 1) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","turquoise")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)

```

Loading in a dataset

```{r, warning=FALSE, message=FALSE}
copperdata <- read.csv("CopperGenes_gdoc.csv")

copperdata %>% filter(StudyType == "QTL") %>% transmute(CHROM = as.character(as.roman(CHROM)), POS = StartLocation, Study = StudyType, Gene = Gene) -> copperdata2

dim(copperdata2)

SpecificCuSO4 <- filter(copperdata2, CHROM %in% c("V", "VII", "VIII", "XI", "XII", "XVI", "XV")) %>% arrange(CHROM, POS)

subset(SpecificCuSO4, CHROM == "XII")

actualQTL <- data.frame(Gene = c("VMA4", "MTL1", "CIC1","PTR2", "IRC15", "HSP60", "CUP1", "LEU9"), 
                        CHROM = factor(c("V", "VII", "VIII", "XI", "XVI", "XII", "VIII", "XV"), 
                                       levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI")), 
                        POS = c(100451, 539938, 211292, 617558, 519854, 664127, 212535, 523702)) 

actualQTL %>% arrange(CHROM)

filteredQTL <- copperdata2 %>% filter(CHROM %in%  c("V", "VII", "VIII", "XI", "XII", "XVI") == FALSE)
filteredQTL$CHROM <- factor(filteredQTL$CHROM,
                            levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI"))
```

```{r, fig.width = 16, fig.height = 6}
rbind(data.frame(CuSO4_CSS8_PeaksB, Exp = "CSS8"),
      data.frame(CuSO4Peaks_B, Exp = "CSS1a"),
      data.frame(CuSO42Peaks_B, Exp = "CSS1b")) -> CUSO4b_peaks

CUSO4b_peaks$CHROM
actualQTL$CHROM

CUSO4b %>% #filter(CHROM %in% actualQTL$CHROM) %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label))+
  #geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "lightblue", size =2) + 
  geom_vline(data = CUSO4b_peaks, aes(xintercept = POS), color = "lightblue", size =2) + 

  geom_vline(data = actualQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed")+
  
  geom_vline(data = filteredQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed", alpha = 0.3)+

  geom_line(size = 1) + 
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink", "gray","black","#7030A0")) +
  theme(legend.position = "none", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)

CUSO4b %>% #filter(CHROM %in% actualQTL$CHROM) %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label))+
  #geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "lightblue", size =2) + 
  geom_vline(data = CUSO4b_peaks, aes(xintercept = POS), color = "lightblue", size =2) + 

  geom_vline(data = actualQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed")+
  
  #geom_vline(data = filteredQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed", alpha = 0.3)+

  geom_line(size = 1) + 
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink", "gray","black","#7030A0")) +
  theme(legend.position = "none", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)

CUSO4b %>% #filter(CHROM %in% actualQTL$CHROM) %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label))+
  #geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "lightblue", size =2) + 
  geom_vline(data = CUSO4b_peaks, aes(xintercept = POS), color = "lightblue", size =2) + 

  #geom_vline(data = actualQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed")+
  
  #geom_vline(data = filteredQTL, aes(xintercept = POS), color = "firebrick2", size = 1, linetype = "dashed", alpha = 0.3)+

  geom_line(size = 1) + 
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink", "gray","black","#7030A0")) +
  theme(legend.position = "none", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)
```


A closer look at Ura3; should only have one parents' version, so that would be an additive QTL...? It was replaced so that they all have the same one, so it shouldn't differ between bulks though. Chr III is also the same thing and should have non-Wine and non-Oak loci, so that would be why it isn't a QTL??? Oh wait, I'm not actually selecting for Ura3 so that's actually something that might be a QTL. Like, I have both in the parents (Ura3 replaced in one parent but not the other) so it's just......... what the background strains did?

Wine.01(a).YPS_URA + Oak(alpha)
Oak.01(alpha) + Wine(a).YPS_URA

```{r, warning=FALSE, message=FALSE}
CUSO4b %>% filter(CHROM == "V") %>% ggplot(aes(x = POS, y = abs(summary), color = label))+
  geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "pink", size =2) + 
  geom_vline(data = URA3, aes(xintercept = POS), color = "gold", size = 1.2)+
  geom_line(size = 1) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","turquoise")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)
```


Fluconazole Comparison

```{r, fig.width=16,fig.height=8}

rbind(data.frame(Fluc_8_glm, Exp = "CSS8"),
      data.frame(Fluc_glm, Exp = "CSS1")) -> Flucb

max(abs(Flucb$summary[Flucb$CHROM == "III" & Flucb$Exp == "CSS8"]))
max(abs(Flucb$summary[Flucb$CHROM == "III" & Flucb$Exp == "CSS1"]))


Flucb %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Flucb$summary[Flucb$CHROM == "III" & Flucb$Exp == "CSS8"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","firebrick2")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 


```

CuSO4 dose comparison 

```{r, fig.width=17,fig.height=8}

rbind(data.frame(CuSO4_8_glm, Exp = "CSS8 Full"),
      data.frame(CuSO4_8_c5_glm, Exp = "CSS8 Low Dose"),
      data.frame(CuSO4_8_D2_glm, Exp = "CSS8 Day 2")) -> CUSO4c

max(abs(CUSO4c$summary[CUSO4c$CHROM == "III" & CUSO4c$Exp == "CSS8 Full"]))
max(abs(CUSO4c$summary[CUSO4c$CHROM == "III" & CUSO4c$Exp == "CSS8 Low Dose"]))
max(abs(CUSO4c$summary[CUSO4c$CHROM == "III" & CUSO4c$Exp == "CSS8 Day 2"]))


CUSO4c %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8 Low Dose"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","turquoise")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 

CUSO4c %>% filter(label != "Rep") %>% ggplot(aes(x = POS, y = abs(summary), color = paste(label,Exp))) + geom_line(size = 1, alpha = 0.8) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8 Low Dose"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(label ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("black","navyblue","darkorchid", #Bulk = black
                                       "firebrick","blue","violet", #Interaction = teal
                                       "darkgray","steelblue","plum4")) + #Parent = gray
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 

CUSO4c %>% filter(label != "Rep") %>% ggplot(aes(x = POS, y = abs(summary), color = Exp, linetype = Exp)) + geom_line(size = 1) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8 Low Dose"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(label ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("firebrick", "blue","violet")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 

```

# POSTER and now Committee Meeting

### Interaction Effects
Copper Sulfate Chr 1
```{r, fig.width = 16, fig.height=4}
CuSO4_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = CuSO4Peaks, aes(xintercept = POS), color = "orange", size = 2) +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_glm$summary[CuSO4_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("gray","black","#7030A0")) + 
  scale_y_continuous(breaks= breaks_pretty()) +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,11.5)
```

Copper Sulfate Chr 8
```{r, fig.width = 16, fig.height=4}
CuSO4_8_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = CuSO4_CSS8_Peaks, aes(xintercept = POS), color = "orange", size = 2) +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_8_glm$summary[CuSO4_8_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink","gray",  "black","#7030A0")) + 
  scale_y_continuous(breaks= breaks_pretty()) +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,11.5)
```

Fluconazole Chr 1
```{r, fig.width = 16, fig.height=4}
Fluc_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = FlucPeaks, aes(xintercept = POS), color = "orange", size = 2) +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Fluc_glm$summary[Fluc_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("gray",  "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,8)
```

Fluconazole Chr 8
```{r, fig.width = 16, fig.height=4}
Fluc_8_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = Fluc_CSS8_Peaks, aes(xintercept = POS), color = "orange", size = 2) +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Fluc_8_glm$summary[Fluc_8_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink","gray",  "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,8)
```
Adding in Zeocin Chr 1
```{r, fig.width = 16, fig.height=4}
Zeocin_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = ZeocinPeaks, aes(xintercept = POS), color = "orange", size = 2) +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Zeocin_glm$summary[Zeocin_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink","gray",  "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,8)
```

### Bulk Effects Highlighted WITH interactions

Copper Sulfate Chr 1
```{r, fig.width = 16, fig.height=4}
CuSO4_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
  #ggplot(aes(x = POS, y = summary, color = label)) + 
    geom_vline(data = CuSO4Peaks_B, aes(xintercept = POS), color = "lightblue", linewidth = 2) +
    geom_vline(data = CuSO4Peaks, aes(xintercept = POS), color = "orange", linewidth = 2, linetype = "dashed") +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_glm$summary[CuSO4_glm$CHROM == "III"]))), linetype = "dashed") +
    geom_hline(aes(yintercept = -max(abs(CuSO4_glm$summary[CuSO4_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("gray","black","#7030A0")) + 
  scale_y_continuous(breaks= breaks_pretty()) +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,11.5)
```

Copper Sulfate Chr 8
```{r, fig.width = 16, fig.height=4}
CuSO4_8_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = CuSO4_CSS8_PeaksB, aes(xintercept = POS), color = "lightblue", size = 2) +
      geom_vline(data = CuSO4_CSS8_Peaks, aes(xintercept = POS), color = "orange", size = 2, linetype = "dashed") +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_8_glm$summary[CuSO4_8_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink","gray",  "black","#7030A0")) + 
  scale_y_continuous(breaks= breaks_pretty()) +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,11.5)
```

Fluconazole Chr 1
```{r, fig.width = 16, fig.height=4}
Fluc_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = FlucPeaks_B, aes(xintercept = POS), color = "lightblue", size = 2) +
    geom_vline(data = FlucPeaks, aes(xintercept = POS), color = "orange", size = 2, linetype = "dashed") +
    geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Fluc_glm$summary[Fluc_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("gray",  "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,8)
```

Fluconazole Chr 8
```{r, fig.width = 16, fig.height=4}

ChromosomeScale$CHROM <- factor(ChromosomeScale$CHROM, levels =  levels(Fluc_8_glm$CHROM))

Fluc_8_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
    geom_vline(data = Fluc_CSS8_Peaks_B, aes(xintercept = POS), color = "lightblue", size = 2) +
    geom_vline(data = Fluc_CSS8_Peaks, aes(xintercept = POS), color = "orange", size = 2, linetype = "dashed") +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Fluc_8_glm$summary[Fluc_8_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink","gray",  "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,8)
```
Adding in Zeocin Chr 1
```{r, fig.width = 16, fig.height=4}
Zeocin_glm %>% rbind(ChromosomeScale) %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
  #ggplot(aes(x = POS, y = summary, color = label)) + 
    geom_vline(data = ZeocinPeaks_B, aes(xintercept = POS), color = "lightblue", size = 2) +
    geom_vline(data = ZeocinPeaks, aes(xintercept = POS), color = "orange", size = 2, linetype = "dashed") +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Zeocin_glm$summary[Zeocin_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("lightpink","gray",  "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())# + ylim(0,8)
```

# Post Posters
## Making circos plots

```{r, warning=FALSE, message=FALSE}
# require(devtools)
# devtools::install_github("jokergoo/circlize")

require(circlize)

# set.seed(999)
# n = 1000
# df = data.frame(sectors = sample(letters[1:8], n, replace = TRUE),
#     x = rnorm(n), y = runif(n))
# 
# plot(df[df$sectors == "f",c(2,3)])
# 
# #Initialize
# circos.par("track.height" = 0.1)
# circos.initialize(df$sectors, x = df$x)
# #I think this makes the sizes?
# circos.track(df$sectors, y = df$y,
#     panel.fun = function(x, y) {
#         circos.text(CELL_META$xcenter, 
#             CELL_META$cell.ylim[2] + mm_y(5), 
#             CELL_META$sector.index)
#         circos.axis(labels.cex = 0.6)
# })
# col = rep(c("orchid4", "violet"), 4)
# 
# circos.trackPoints(df$sectors, df$x, df$y, col = col, pch = 16, cex = 0.5)
# circos.text(-1, 0.5, "text", sector.index = "a", track.index = 1)
# 
# #The interaction part
# circos.link("a", 0, "b", 0, h = 0.4)
# circos.link("c", c(-0.5, 0.5), "d", c(-0.5,0.5), col = "red",
#     border = "blue", h = 0.2)
# circos.link("e", 0, "g", c(-1,1), col = "green", border = "black", lwd = 2, lty = 2)
```

```{r, fig.width=8, fig.height = 8, warning=FALSE, message=FALSE}
#Let's try this with chromosomes

df2 <- data.frame(sectors = as.character(Fluc_8_glm$CHROM),
                 x = Fluc_8_glm$POS,
                 y = abs(Fluc_8_glm$summary),
                  label = Fluc_8_glm$label)

df3 <- data.frame(sectors = as.character(Fluc_glm$CHROM),
                 x = Fluc_glm$POS,
                 y = abs(Fluc_glm$summary),
                  label = Fluc_glm$label)


dfint <- subset(df2, label == "Interaction")
dfint$y[dfint$sectors == "VIII"] <- 0

df3int <- subset(df3, label == "Interaction")

#plot(dfint[dfint$sectors == "II",c(2,3)])

circos.par("track.height" = 0.3, start.degree = 90)
circos.initialize(dfint$sectors, x = dfint$x)

#I think this makes the sizes?
circos.track(ylim = c(max(dfint$y), 0), dfint$sectors, y = dfint$y, 
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, 
            0 - mm_y(5), 
            CELL_META$sector.index,
            niceFacing = FALSE)
        circos.axis(labels.cex = 0.1)
        
})

draw.sector(85, 90, rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = "#00008070", border = FALSE)
draw.sector(279.5, 296.5, rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = "#da70d670", border = FALSE)
#col = rep(c("orchid4", "violet"), 8)
circos.trackPoints(dfint$sectors, dfint$x, abs(dfint$y), col = "orchid", pch = 16, cex = 0.1)
circos.trackPoints(df3int$sectors, df3int$x, abs(df3int$y), col = "#000080", pch = 16, cex = 0.1)

#circos.text(-1, 0.5, "text", sector.index = "II", track.index = 1)

#The interaction part
circos.link("XV", 910072,"VIII", c(0, max(dfint$x[dfint$sectors == "VIII"])),  col = "#da70d670", h.ratio = 0.3, border = "#da70d670", lwd = 2)
circos.link("VIII", c(0, max(dfint$x[dfint$sectors == "VIII"])),"XVI", 562911,  col = "#da70d670", h.ratio = 0.3, border = "#da70d670", lwd = 2)
circos.link("I", c(0, max(dfint$x[dfint$sectors == "I"])),
            "VII", 523005,  col = "#00008070", border = "#00008070", lwd = 2)
circos.link("I", c(0, max(dfint$x[dfint$sectors == "I"])),
            "XIII", 646703,  col = "#00008070", border = "#00008070", lwd = 2)

circos.clear()


```
The same thing but for CuSO4
```{r, fig.width=8, fig.height = 8, warning=FALSE, message=FALSE}
#Let's try this with chromosomes

df4 <- data.frame(sectors = as.character(CuSO4_8_glm$CHROM),
                 x = CuSO4_8_glm$POS,
                 y = abs(CuSO4_8_glm$summary),
                  label = CuSO4_8_glm$label)

df5 <- data.frame(sectors = as.character(CuSO4_glm$CHROM),
                 x = CuSO4_glm$POS,
                 y = abs(CuSO4_glm$summary),
                  label = CuSO4_glm$label)


df4int <- subset(df4, label == "Interaction")
df4int$y[df4int$sectors == "VIII"] <- 0

df5int <- subset(df5, label == "Interaction")

#plot(dfint[dfint$sectors == "II",c(2,3)])

################################################################################
#PLOT
################################################################################
circos.par("track.height" = 0.3, start.degree = 90)
circos.initialize(df4int$sectors, x = df4int$x)

#I think this makes the sizes?
circos.track(ylim = c(max(df4int$y), 0), df4int$sectors, y = df4int$y, 
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, 
            0 - mm_y(5), 
            CELL_META$sector.index,
            niceFacing = FALSE)
        circos.axis(labels.cex = 0.1)
})

draw.sector(85, 90, rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = "#99999970", border = FALSE)
draw.sector(279, 296.5, rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = "#00525470", border = FALSE)

#col = rep(c("orchid4", "violet"), 8)
circos.trackPoints(df4int$sectors, df4int$x, abs(df4int$y), col = "#005254", pch = 16, cex = 0.1)
circos.trackPoints(df5int$sectors, df5int$x, abs(df5int$y), col = "#99999930", pch = 16, cex = 0.1)

#The interaction part with Chr 8
circos.link("V", 106423,
            "VIII", c(0, max(df4int$x[df4int$sectors == "VIII"])),  col = "#00525470", border = "#00525470", lwd = 2)
circos.link("VIII", c(0, max(df4int$x[df4int$sectors == "VIII"])),
            "XI", 663509,  col = "#00525470", border = "#00525470", lwd = 2)
circos.link("VIII", c(0, max(df4int$x[df4int$sectors == "VIII"])),
            "XIV", 202195,  col = "#00525470", border = "#00525470", lwd = 2)

#With Chr I? Shouldn't be any
circos.clear()

```

Trying this for Zeocin as well

```{r, fig.width=8, fig.height = 8, warning=FALSE, message=FALSE}
ZeocinPeaks
df5 <- data.frame(sectors = as.character(Zeocin_glm$CHROM),
                 x = Zeocin_glm$POS,
                 y = abs(Zeocin_glm$summary),
                  label = Zeocin_glm$label)


df5int <- subset(df5, label == "Interaction")

circos.par("track.height" = 0.3, start.degree = 90)
circos.initialize(df4int$sectors, x = df4int$x)

#I think this makes the sizes?
circos.track(ylim = c(max(df5int$y), 0), df4int$sectors, y = df4int$y, 
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, 
            0 - mm_y(5), 
            CELL_META$sector.index,
            niceFacing = FALSE)
        circos.axis(labels.cex = 0.1)
        
})
#col = rep(c("orchid4", "violet"), 8)
circos.trackPoints(df5int$sectors, df5int$x, abs(df5int$y), col = "#d04c73", pch = 16, cex = 0.1)
draw.sector(85, 90, rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = "#d04c7370", border = FALSE)
#draw.sector(180, 10, rou1 = 1, rou2 = 0.7, clock.wise = FALSE, col = "#0000FF80")

#The interaction part with Chr 8
circos.link("I", c(0, max(df4int$x[df4int$sectors == "I"])),
            "IV", 428383,  col = "#d04c7370", border = "#d04c7370", lwd = 2)
circos.link("I", c(0, max(df4int$x[df4int$sectors == "I"])),
            "VII", 525205,  col = "#d04c7370", border = "#d04c7370", lwd = 2)
circos.link("I", c(0, max(df4int$x[df4int$sectors == "I"])),
            "X", 503433,  col = "#d04c7370", border = "#d04c7370", lwd = 2)
circos.link("I", c(0, max(df4int$x[df4int$sectors == "I"])),
            "XI", 665232,  col = "#d04c7370", border = "#d04c7370", lwd = 2)
circos.link("I", c(0, max(df4int$x[df4int$sectors == "I"])),
            "XV", 850507,  col = "#d04c7370", border = "#d04c7370", lwd = 2)
circos.link("I", c(0, max(df4int$x[df4int$sectors == "I"])),
            "XVI", 591540,  col = "#d04c7370", border = "#d04c7370", lwd = 2)


#With Chr I? Shouldn't be any
circos.clear()
```

```{r, warning=FALSE, message=FALSE}
# #Sample code
# par(mar = c(1, 1, 1, 1))
# plot(c(-1, 1), c(-1, 1), type = "n", axes = FALSE, ann = FALSE, asp = 1)
# draw.sector(20, 0)
# #this one
# draw.sector(30, 60, rou1 = 0.8, rou2 = 0.5, clock.wise = FALSE, col = "#FF000080")
# 
# draw.sector(350, 1000, col = "#00FF0080", border = NA)
# draw.sector(0, 180, rou1 = 0.25, center = c(-0.5, 0.5), border = 2, lwd = 2, lty = 2)
# draw.sector(0, 360, rou1 = 0.7, rou2 = 0.6, col = "#0000FF80")
```

## Comparison of Additive QTL in all traits

```{r, fig.height=6, fig.width=16}

rbind(data.frame(CuSO4_8_glm, Exp = "CuSO4_8"),
      data.frame(CuSO4_glm, Exp = "CuSO4_1a"),
      data.frame(CuSO4_glm2, Exp = "CuSO4_1b"),
      data.frame(Fluc_8_glm, Exp = "Fluc_8"),
      data.frame(Fluc_glm, Exp = "Fluc_1"),
      data.frame(Zeocin_glm, Exp = "Zeo_1"))-> All_Experiments

All_Experiments %>% filter(label == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(summary), color = Exp)) + geom_line(size = 1) + 
  scale_color_manual(values = c("turquoise", "turquoise3", "turquoise2", "firebrick", "firebrick2", "orchid")) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = Selection, linetype = FixedChr)) + geom_line(size = 1) + 
  scale_color_manual(values = c("turquoise3","firebrick2", "orchid")) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1) + 
  scale_color_manual(values = c("black","#00000080", "#00000090", "skyblue3")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + ylim(0, 15) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  filter(Selection == "CuSO4") %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1) + 
  scale_color_manual(values = c("black","#00000090", "skyblue3")) +
  facet_grid(~CHROM, scales = "free", space = "free") + ylim(0, 15) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("CuSO4 Additive QTL") + ylab("Bulk Z Score")
```

## Run a GLM on the raw data and the smoothed data for CSSI Chr 7 Fluconazole

```{r}
CSSI_Fluc <- readRDS("Data/Fluconazole_1_cybr2.rds")

CSSI_Fluc %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_Fluc_rollData

#Chr 7
CSSI_Fluc %>% filter(CHROM == "VII") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed(HOO = Fluconazole_OakI_B_Oak, 
                           HOW = Fluconazole_OakI_B_Wine, 
                           HWO = Fluconazole_WineI_C_Oak,
                           HWW = Fluconazole_WineI_C_Wine,
                           LOO = Dilute_OakI_B_Oak,
                           LOW = Dilute_OakI_B_Wine,
                           LWO = Dilute_WineI_B_Oak, 
                           LWW = Dilute_WineI_B_Wine)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Fluc_glm_7

CSSI_Fluc_rollData %>% filter(CHROM == "VII") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed(HOO = Fluconazole_OakI_B_Oak, 
                           HOW = Fluconazole_OakI_B_Wine, 
                           HWO = Fluconazole_WineI_C_Oak,
                           HWW = Fluconazole_WineI_C_Wine,
                           LOO = Dilute_OakI_B_Oak,
                           LOW = Dilute_OakI_B_Wine,
                           LWO = Dilute_WineI_B_Oak, 
                           LWW = Dilute_WineI_B_Wine)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Fluc_rollData_glm_7

#Chr 3 (for baseline)
CSSI_Fluc %>% filter(CHROM == "III") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed(HOO = Fluconazole_OakI_B_Oak, 
                           HOW = Fluconazole_OakI_B_Wine, 
                           HWO = Fluconazole_WineI_C_Oak,
                           HWW = Fluconazole_WineI_C_Wine,
                           LOO = Dilute_OakI_B_Oak,
                           LOW = Dilute_OakI_B_Wine,
                           LWO = Dilute_WineI_B_Oak, 
                           LWW = Dilute_WineI_B_Wine)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Fluc_glm_3

CSSI_Fluc_rollData %>% filter(CHROM == "III") %>% group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed(HOO = Fluconazole_OakI_B_Oak, 
                           HOW = Fluconazole_OakI_B_Wine, 
                           HWO = Fluconazole_WineI_C_Oak,
                           HWW = Fluconazole_WineI_C_Wine,
                           LOO = Dilute_OakI_B_Oak,
                           LOW = Dilute_OakI_B_Wine,
                           LWO = Dilute_WineI_B_Oak, 
                           LWW = Dilute_WineI_B_Wine)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Fluc_rollData_glm_3


plota <- CSSI_Fluc_glm_7 %>% filter(label != "intercept") %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + 
  geom_point(pch = 1) +
  geom_hline(yintercept = 4.58, color = "#00B050", size = 2) +
  geom_hline(yintercept = 1.96, color = "#4472C4", size = 2) +
  scale_color_manual(values = c("black", "orchid", "gray")) + ylim(0, max(CSSI_Fluc_glm_7$summary[CSSI_Fluc_glm_7$label != "intercept"])) +
  ylab("") + xlab("")+
  theme(legend.position = "none", axis.text.x=element_blank(),axis.text.y=element_blank(),
            axis.ticks.x=element_blank(),axis.title.y=element_blank(),
                   axis.ticks.y=element_blank())

plotb <- CSSI_Fluc_rollData_glm_7 %>% filter(label != "intercept") %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_point(pch = 1) +
  geom_hline(yintercept = 4.58, color = "#00B050", size = 2) +
  geom_hline(yintercept = 1.96, color = "#4472C4", size = 2) +
  scale_color_manual(values = c("black", "orchid", "gray"))+ ylim(0, max(CSSI_Fluc_glm_7$summary[CSSI_Fluc_glm_7$label != "intercept"])) +
  ylab("") + xlab("")+
  theme(legend.position = "none", axis.text.x=element_blank(),axis.text.y=element_blank(),
            axis.ticks.x=element_blank(),axis.title.y=element_blank(),
                   axis.ticks.y=element_blank())

plotc <- CSSI_Fluc_glm_3 %>% filter(label != "intercept") %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_point(pch = 1) +
  geom_hline(yintercept = 4.58, color = "#00B050", size = 2) +
  geom_hline(yintercept = 1.96, color = "#4472C4", size = 2) +
  geom_hline(yintercept = max(abs(CSSI_Fluc_glm_3$summary[CSSI_Fluc_glm_3$label != "intercept"])), color = "red", linetype = "dashed", size = 2)+
  scale_color_manual(values = c("black", "orchid", "gray"))+ ylim(0, max(CSSI_Fluc_glm_7$summary[CSSI_Fluc_glm_7$label != "intercept"])) +
  ylab("") + xlab("")+
  theme(legend.position = "none", axis.text.x=element_blank(),axis.text.y=element_blank(),
            axis.ticks.x=element_blank(),axis.title.y=element_blank(),
                   axis.ticks.y=element_blank())

plotd <- CSSI_Fluc_rollData_glm_3 %>% filter(label != "intercept") %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_point(pch = 1) +
  geom_hline(yintercept = 4.58, color = "#00B050", size = 2) +
  geom_hline(yintercept = 1.96, color = "#4472C4", size = 2) +
  geom_hline(yintercept = max(abs(CSSI_Fluc_rollData_glm_3$summary[CSSI_Fluc_rollData_glm_3$label != "intercept"])), color = "red", linetype = "dashed", size = 2)+
  scale_color_manual(values = c("black", "orchid", "gray"))+ ylim(0, max(CSSI_Fluc_glm_7$summary[CSSI_Fluc_glm_7$label != "intercept"])) +
  ylab("") + xlab("")+
  theme(legend.position = "none", axis.text.x=element_blank(),axis.text.y=element_blank(),
            axis.ticks.x=element_blank(),axis.title.y=element_blank(),
                   axis.ticks.y=element_blank())

plot_grid(plota, plotb, plotc, plotd, ncol = 4, rel_widths = c(2,2,1.5,1.5)) + theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))

```

## Run a GLM on the raw data and the smoothed data for CSSI Chr 7 Fluconazole

```{r, fig.width = 10, fig.height=4}
CSSI_ZeoCyc <- readRDS("Data/Zeocin_cybr2.rds")

CSSI_ZeoCyc %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_ZeoCyc_rollData

CSSI_ZeoCyc_rollData %>% pivot_longer(c(-CHROM, -POS)) %>% separate(name, into = c("Bulk", "Parent", "Rep", "Allele"), sep = "_") %>%
  pivot_wider(names_from = Allele, values_from = value) %>% mutate(LOD = log(Wine/Oak)) -> workableCycZeo

workableCycZeo %>% filter(CHROM != "I", CHROM != "M") %>% ggplot(aes(x = POS, y = LOD, color = paste(Bulk, Parent, Rep))) + 
  geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("Cycloheximide (NA) and Zeocin log(Wine/Oak)")

```

