---
title: "Figures 2023"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: hide
  html_notebook:
    code_folding: hide
---

```{r, warning=FALSE, message=FALSE, comment=FALSE, fig.width=18, fig.height=5}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
require(ggplot2)
require(tidyr)
require(reshape2)
require(cowplot)
require(data.table)

require(dplyr)
require(foreach)
require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)

require(cybrBSA)

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)))

glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H","H","H", "H", 
                                             "L", "L","L", "L", 
                                             "H", "H","H", "H", 
                                             "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", 
                                             "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W",
                                             "O", "W", "O", "W", 
                                             "O", "W", "O", "W",
                                             "O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", 
                                          "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                                     HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

cybr2_rollmean <- function(dataframe){
  dataframe %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollmean(value, n = 100))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount)
}

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

```

Note: if looking for smoothed vs unsmoothed data, the `Figures_2023_BiologyofGenomes.rmd` file has these at the beginning.

# Processing
## Load in Data


```{r, warning=FALSE, message=FALSE}
#CSS 1
Zeocin_glm<- readRDS("Data/Zeocin_glm_c.rds") %>% filter(CHROM != "I") #DONE
CuSO4_glm <- readRDS("Data/CuSO4_1_glm_c.rds") %>% filter(CHROM != "I") #DONE
CuSO4_glm2 <- readRDS("Data/CuSO4_2_glm_c.rds") %>% filter(CHROM != "I") #DONE
Fluc_glm <- readRDS("Data/Fluconazole_1_glm_c.rds") %>% filter(CHROM != "I") #DONE
CuSO4r_glm <- readRDS("Data/CuSO4_1_redone_glm_c.rds") %>% filter(CHROM != "I") #DONE


#CSS 8
CuSO4_8_glm <- readRDS("Data/CuSO4_CSS8_glm_c.rds") %>% filter(CHROM != "VIII") #DONE
CuSO4_8_c5_glm <- readRDS("Data/CuSO4_CSS8_C5norep_glm.rds") %>% filter(CHROM != "VIII") #NOT INCLUDED?
CuSO4_8_D2_glm <- readRDS("Data/CuSO4_CSS8_D2_norep_glm_c.rds") %>% filter(CHROM != "VIII") #DONE
Fluc_8_glm <- readRDS("Data/Fluc_CSS8_glm_c.rds") %>% filter(CHROM != "VIII") #DONE

#New H2O2 and Zeocin 9/11/23 - might need to be reformatted to start
readRDS("H2O2_CSS8_glm_SUB_50.rds") %>% mutate(coefficient = str_replace(coefficient, "BulkH2O2", "Bulk"),
                             coefficient = str_replace(coefficient, "ParentWine8", "Parent"),
                             coefficient = str_replace(coefficient, ":", ""),
                             coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
  filter(label == "Zscore", coefficient != "(Intercept)", CHROM != "VIII") %>%
  transmute(CHROM = CHROM, POS = POS, summary = GLMResult, label = coefficient) -> H2O2_8_glmer

readRDS("H2O2_CSS1_glm50.rds") %>% mutate(coefficient = str_replace(coefficient, "BulkH2O2", "Bulk"),
                             coefficient = str_replace(coefficient, "ParentWine1", "Parent"),
                             coefficient = str_replace(coefficient, ":", ""),
                             coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
  filter(label == "Zscore", coefficient != "(Intercept)", CHROM != "I") %>%
  transmute(CHROM = CHROM, POS = POS, summary = GLMResult, label = coefficient)-> H2O2_glmer

readRDS("Zeocin_CSS8_glm50.rds") %>% mutate(coefficient = str_replace(coefficient, "BulkZeocin", "Bulk"),
                           coefficient = str_replace(coefficient, "ParentWine8", "Parent"),
                           coefficient = str_replace(coefficient, ":", ""),
                           coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
filter(label == "Zscore", coefficient != "(Intercept)", CHROM != "VIII") %>%
transmute(CHROM = CHROM, POS = POS, summary = GLMResult, label = coefficient) -> Zeocin_8_glmer

#Fluc Redone
readRDS("Fluc_CSS1_glmer.rds") %>% mutate(coefficient = str_replace(coefficient, "BulkFluc", "Bulk"),
                             coefficient = str_replace(coefficient, "ParentWine1", "Parent"),
                             coefficient = str_replace(coefficient, ":", ""),
                             coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
  filter(label == "Zscore", coefficient != "(Intercept)", CHROM != "I") %>%
  transmute(CHROM = CHROM, POS = POS, summary = GLMResult, label = coefficient) -> Fluc_1_glmer

#CuSO4 Redone
readRDS("CuSO4_CSSI_all_glmer.rds") %>% #mutate(coefficient = str_replace(Param, "BulkCuSO4", "Bulk"),
                             #coefficient = str_replace(coefficient, "ParentWine1", "Parent"),
                             #coefficient = str_replace(coefficient, ":", ""),
                             #coefficient = str_replace(coefficient, "BulkParent", "Interaction")) %>%
  filter(Type == "Zscore", Param != "(Intercept)", CHROM != "I") %>%
  transmute(CHROM = CHROM, POS = POS, summary = GLM, label = Param) -> CuSO4_1_glmer

################################################################################
# MISC FORMATTING STANDARDIZING
################################################################################

Zeocin_glm$label <- factor(Zeocin_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
Fluc_glm$label <- factor(Fluc_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_glm$label <- factor(CuSO4_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4r_glm$label <- factor(CuSO4r_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_glm2$label <- factor(CuSO4_glm2$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

Zeocin_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Zeocin_glm
Fluc_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Fluc_glm
CuSO4_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> CuSO4_glm
CuSO4_glm2 %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> CuSO4_glm2
CuSO4r_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> CuSO4r_glm

#New CSS8 Ones - add back Chr I
Fluc_8_glm$label <- factor(Fluc_8_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_8_c5_glm$label <- factor(CuSO4_8_c5_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_8_D2_glm$label <- factor(CuSO4_8_D2_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_8_glm$label <- factor(CuSO4_8_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

Fluc_8_glm %>% filter(CHROM != "M", label != "intercept") -> Fluc_8_glm
CuSO4_8_c5_glm %>% filter(CHROM != "M", label != "intercept") -> CuSO4_8_c5_glm
CuSO4_8_D2_glm %>% filter(CHROM != "M", label != "intercept") -> CuSO4_8_D2_glm
CuSO4_8_glm %>% filter(CHROM != "M", label != "intercept") -> CuSO4_8_glm

#Newest: H2O2
H2O2_8_glmer %>% filter(CHROM != "VIII", CHROM != "M", label != "intercept") -> H2O2_8_glmer
H2O2_glmer %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> H2O2_glmer
H2O2_8_glmer$label <- factor(H2O2_8_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
H2O2_glmer$label <- factor(H2O2_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

#Zeocin New
Zeocin_8_glmer %>% filter(CHROM != "VIII", CHROM != "M", label != "intercept") -> Zeocin_8_glmer
Zeocin_8_glmer$label <- factor(Zeocin_8_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

#Fluc glmer
Fluc_1_glmer %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Fluc_1_glmer
Fluc_1_glmer$label <- factor(Fluc_1_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

#New CuSO4 glmer
CuSO4_1_glmer %>% filter(CHROM != "I", CHROM != "M", label != "Intercept") %>% na.omit() -> CuSO4_1_glmer
CuSO4_1_glmer$label <- factor(CuSO4_1_glmer$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

```

Parent data

```{r}
engineered <- data.frame(Locus = c("URA3", "MATalpha"),
                         CHROM = c("V", "III"),
                         POS_start = c(116167, 198671),
                         POS_end = c(116970, 201177))

```

```{r}
setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")

cybrConvertParentalAlleles <- function(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"),
                                       Parents = gsub("_VCF.txt","", ParentFiles), Truncate = TRUE, yeast = TRUE){
  temparent <- list()
  mergeparents <- foreach(i=1:length(ParentFiles), .combine=rbind) %dopar% {
    read.table(ParentFiles[i], header = TRUE) %>% mutate(parent = Parents[i])
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                           "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                           CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                     "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                     "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                     "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    rbind(mergeparents) %>% arrange(CHROM, POS) %>%
      select(CHROM, POS, REF, ALT, parent) %>%
      merge(ChromKey) %>% select(-CHROM) %>%
      mutate(CHROM = chromosomes) %>% select(-chromosomes) -> ParentalVCF

  }else{
    rbind(mergeparents) %>% arrange(CHROM, POS) %>%
      select(CHROM, POS, REF, ALT, parent) -> ParentalVCF
  }

  ParentalVCF %>% pivot_wider(names_from = parent, values_from = ALT) -> SNPids

  SNPids$Type <- 0
  for(i in Parents){

    #filter rows in which all values of columns of the parent NOT selected are NA
    select(SNPids,-i, -CHROM, -POS, -REF) -> tempdf
    tempdf$Any_NA <- apply(tempdf, 1, function(x) anyNA(x))
    SNPids$Type[which(tempdf$Any_NA)] <- i
    rm(tempdf)
  }


  #Collect it to output
  if(Truncate == TRUE){
    SNPids %>% select(CHROM, POS,  Type) %>% filter(Type != 0) -> SNPids
  }

  return(SNPids)

}

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) %>% filter(Unique == 1) -> pSNPs

pSNPs$CHROM <- factor(pSNPs$CHROM, levels = as.character(as.roman(1:17)))

pSNPs_all <- rbind(data.frame(pSNPs[,1:2], E = "A_REF"),
                   data.frame(H2O2_glmer[,1:2], E = "H1"),
                   data.frame(H2O2_8_glmer[,1:2], E = "H8"),
                   data.frame(CuSO4_8_glm[,1:2], E = "C8"),
                   data.frame(CuSO4_1_glmer[,1:2], E = "C1"),
                   data.frame(Fluc_1_glmer[,1:2], E = "F1"),
                   data.frame(Fluc_8_glm[,1:2], E = "F8"),
                   data.frame(Zeocin_8_glmer[,1:2], E = "Z8"),
                   data.frame(Zeocin_glm[,1:2], E = "Z1")) %>%
  distinct()

pSNPs_CuSO4Comp <- rbind(data.frame(pSNPs[,1:2], E = "A_REF"),
                   data.frame(H2O2_glmer[,1:2], E = "H1"),
                   data.frame(H2O2_8_glmer[,1:2], E = "H8"),
            
                   data.frame(CuSO4_glm[,1:2], E = "C1"),
                  
                   data.frame(CuSO4r_glm[,1:2], E = "C1r")) %>%
  distinct()
```

Compare coverage of CuSO4 real quick

```{r, fig.width=16, fig.height=4}
pSNPs_CuSO4Comp %>% filter(CHROM == "VII") %>% 
  ggplot(aes(x = POS, y = E, color = E)) + 
  geom_point(alpha = 0.1, size = 2) + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("black", "turquoise", "darkturquoise", "red", "firebrick"))
```

## Peak Calling

### GLM Datasets:
Zeocin_glm
Fluc_glm
CuSO4_glm
CuSO4_glm2
Fluc_8_glm
CuSO4_8_c5_glm
CuSO4_8_D2_glm
CuSO4_8_glm

#### GLMER DATASETS
H2O2_8_glmer
H2O2_glmer 
Zeocin_8_glmer 
Fluc_1_glmer 
CuSO4_1_glmer 


### Call Peaks (Function)

Works with the reformatting in the previous step

```{r, eval = FALSE}
cybr_callpeaks <- function(dataset, param = NULL, threshold = NULL, include_all = NULL){
  dataset %>% mutate(summary = abs(summary)) %>% filter(label != "intercept",
                                                        label != "Intercept",
                                                        label != "(Intercept)",
                                                        label != "NA") %>%
    na.omit() -> dataset
  #Set cutoff
  
  Cutoff <- data.frame(X = NA)
  if(is.null(threshold) == FALSE){
    Cutoff$X <- threshold
  }else{
    dataset %>% filter(CHROM == "III") %>%
    ungroup() %>%
    reframe(X = max(summary)) -> Cutoff
  }
  
  #Find peaks
  if(is.null(include_all)){
    dataset %>%
    filter(summary > Cutoff$X) %>%
    group_by(CHROM, label) %>% 
    summarize(summary = max(summary)) %>%
    ungroup() %>%
    merge(dataset) -> output
  
  }else{
    dataset %>%
    filter(summary > Cutoff$X) -> output
  }
  
  if(is.null(param) == FALSE){
    output %>% filter(label == param) -> output
  }
  return(output)
}

################################################################################
#dataset <- Fluc_glm
cybr_multpeaks <- function(dataset, 
                           param = NULL, 
                           threshold = NULL, 
                           width = 1000,
                           buffer = 200){
  dataset %>% mutate(summary = abs(summary)) %>% filter(label != "intercept",
                                                        label != "Intercept",
                                                        label != "(Intercept)",
                                                        label != "NA") %>%
    mutate(Window = POS %/% width) %>%
    mutate(Check = 1) %>%
    na.omit() -> dataset
  
 
  #Set cutoff

  Cutoff <- data.frame(X = NA)
  if(is.null(threshold) == FALSE){
    Cutoff$X <- threshold
  }else{
    dataset %>% filter(CHROM == "III") %>%
    ungroup() %>%
    reframe(X = max(summary)) -> Cutoff
  }

  #Select only the parameter in question
   if(is.null(param) == FALSE){
    dataset %>% filter(label == param) -> dataset
  }
  
  for(i in buffer:(nrow(dataset)-buffer)){
    if(dataset$Window[i] == dataset$Window[(i-(buffer - 1))]){
      dataset$Check[(i-(buffer - 1)):(i+(buffer - 1))] <- FALSE
    }else{
      dataset$Check[(i-(buffer - 1)):((buffer - 1))] <- TRUE
    }
  }
  
  #Find peaks
  dataset %>%
    filter(summary > Cutoff$X) %>%
    group_by(CHROM, label, Window) %>% 
    summarize(summary = max(summary),
              Check = Check) %>%
    ungroup() %>%
    merge(dataset) %>%
    distinct() -> output
  
  
  return(output)
}
```

```{r}

#Chr III Cutoff Peaks

F_1_intpeaks <- cybr_multpeaks(Fluc_glm, param = "Bulk")
C_1_intpeaks <- cybr_multpeaks(CuSO4_glm, param = "Bulk")
H_8_intpeaks <- cybr_multpeaks(H2O2_8_glmer, param = "Bulk")
Z_1_intpeaks <- cybr_multpeaks(Zeocin_glm, param = "Bulk")
C_8_intpeaks <- cybr_multpeaks(CuSO4_8_glm, param = "Bulk")
F_8_intpeaks <- cybr_multpeaks(Fluc_8_glm, param = "Bulk")
Z_8_intpeaks <- cybr_multpeaks(Zeocin_8_glmer, param = "Bulk")

# #Don't just find point
# F_1_intpeaks_all <- cybr_callpeaks(Fluc_glm, param = "Bulk", include_all = TRUE)
# C_1_intpeaks_all <- cybr_callpeaks(CuSO4_glm, param = "Bulk", include_all = TRUE)
# H_8_intpeaks_all <- cybr_callpeaks(H2O2_8_glmer, param = "Bulk", include_all = TRUE)
# Z_1_intpeaks_all <- cybr_callpeaks(Zeocin_glm, param = "Bulk", include_all = TRUE)
# C_8_intpeaks_all <- cybr_callpeaks(CuSO4_8_glm, param = "Bulk", include_all = TRUE)
# F_8_intpeaks_all <- cybr_callpeaks(Fluc_8_glm, param = "Bulk", include_all = TRUE)
# Z_8_intpeaks_all <- cybr_callpeaks(Zeocin_8_glmer, param = "Bulk", include_all = TRUE)
# 
# #GLMER PEAKS
# H8g_intpeaks <- cybr_multpeaks(H2O2_8_glmer, param = "Bulk")
# C1g_intpeaks <- cybr_multpeaks(CuSO4_1_glmer, param = "Bulk")
# H1g_intpeaks <- cybr_multpeaks(H2O2_glmer, param = "Bulk")
# Z8g_intpeaks <- cybr_multpeaks(Zeocin_8_glmer, param = "Bulk")
# F1g_intpeaks <- cybr_multpeaks(Fluc_1_glmer, param = "Bulk")

#UNUSED:
#CuSO4_glm2
#CuSO4_8_c5_glm
#CuSO4_8_D2_glm
```

#### Calling Peaks with dilute parent cutoffs

These were simulated by sampling from dilute bulks four random (smoothed) read counts, two from each parent, and then calculating the bulk, parent, and rep effects. The 5% and 1% FDR were then found from this, which ensures that the parent effect ranges are consistent but that all of the loci can be randomized since the assumption is the dilute bulks will not have an effect overall. This does assume that there are no QTL on the dilute bulks, which is likely false, but which is taken into consideration as the amount of noise that we won't be able to parse.

```{r, eval = FALSE}
CSSCutoff_99 <- readRDS("../9-12-23_DiluteCutoff_99.rds")

CSSCutoff_new <- readRDS("../9-20-23_DiluteCutoff_99.rds")

CutoffExpKey <- data.frame(Exp = unique(CSSCutoff_99$Exp), 
                           Dataset = c("CuSO4_glm",
                                       "CuSO4_8_glm",
                                       "Fluc_glm","Fluc_glm",
                                       "H2O2_glmer",
                                       "H2O2_8_glmer",
                                       "Zeocin_glm"))

C99 <- merge(CSSCutoff_99, CutoffExpKey)

C99 %>% filter(labels == "Bulk")

CSSCutoff_new %>% filter(labels == "Bulk") %>% filter(., grepl("glm_",Exp))

#USE FOR GLMER ONES
C_1_intpeaks_perm <- cybr_callpeaks(CuSO4_1_glmer, param = "Bulk", threshold = 2.808175)
C_8_intpeaks_perm <- cybr_callpeaks(CuSO4_8_glm, param = "Bulk", threshold = 8.691379)

F_1_intpeaks_perm <- cybr_callpeaks(Fluc_1_glmer, param = "Bulk", threshold = 1.872063)
#F_8_intpeaks_perm <- cybr_callpeaks(Fluc_8_glm, param = "Bulk", threshold = 1.96)

#H_1_intpeaks_perm <- cybr_callpeaks(H2O2_glmer, param = "Bulk", threshold = 0.3141185)

H_8_intpeaks_perm <- cybr_callpeaks(H2O2_8_glmer, param = "Bulk", threshold = 7.168438)

Z_8_intpeaks_perm <- cybr_callpeaks(Zeocin_8_glmer, param = "Bulk", threshold = 7.168438)
Z_1_intpeaks_perm <- cybr_callpeaks(Zeocin_glm, param = "Bulk", threshold = 5.626115)

################################################################################
#GLM Permutations on glm data
C_1_intpeaks_perm_b <- cybr_callpeaks(CuSO4_glm, param = "Bulk", threshold = 2.743073)
C_8_intpeaks_perm_b <- cybr_callpeaks(CuSO4_8_glm, param = "Bulk", threshold = 8.442994)
F_1_intpeaks_perm_b <- cybr_callpeaks(Fluc_glm, param = "Bulk", threshold = 1.928081)
Z_1_intpeaks_perm_b <- cybr_callpeaks(Zeocin_glm, param = "Bulk", threshold = 5.063239)


```


# Standardized Plots

```{r}
#UPDATED COLORS:
#Verdigris - chr I default
  "#7DB0B0"
#Tangerine - chr VIII default
  "#ED7B01" 
#Lapis Lazuli: 
  "#24588A90"
#Magenta Haze:
  "#954176"
#Yale Blue:
  "#0E3961"
```

### Parent SNP Loci

```{r, fig.width=16, fig.height=4, eval = FALSE}

centromeres <- data.frame(CHROM = as.character(as.roman(1:16)),
                          POS = c(151465,238207,114385,449711,151987,148510,496920,105586,
                                  355629,436307,440129,150828,268031,628758,326584,555957),
                          Type = "Oak",
                          Type2 = "Wine") %>% pivot_longer(c(Type, Type2), names_to = "label", values_to = "Type")


pSNPs %>% rbind(ChromosomeScale) %>% filter(CHROM != "M") %>%
  ggplot(aes(x = POS, y = Type, color = Type)) + 
  geom_jitter(aes(x = POS), alpha = 0.2, size =0.1, height = 0.2) +
  geom_point(data = centromeres, aes(x = POS, y = Type), size = 3, color = "black") +
  scale_color_manual(values = c("navy", "firebrick")) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none")

pSNPs %>% rbind(ChromosomeScale) %>% filter(CHROM != "M") %>%
  ggplot(aes(x = POS, y = "C")) + 
  geom_jitter(aes(x = POS), alpha = 0.05, size =0.2, height = 0.2) +
  geom_point(data = centromeres, aes(x = POS), size = 3) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none")

pSNPs %>% rbind(ChromosomeScale) %>% filter(CHROM != "M") %>%
  ggplot(aes(x = POS, color = Type)) + 
  geom_density(aes(x = POS)) +
  geom_vline(data = centromeres, aes(xintercept = POS)) +
  scale_color_manual(values = c("navy", "firebrick")) +
  facet_grid(~CHROM, scales = "free_x", space = "free") +
  theme(legend.position = "none")

#Plot all SNPs called for each experiment; this can take a while
pSNPs_all %>% filter(CHROM != "M") %>%
  ggplot(aes(x = POS, y = E)) + 
  geom_jitter(aes(x = POS, y = E), alpha = 0.05, size =0.2, height = 0.2) +
  #geom_point(data = centromeres, aes(x = POS), size = 3) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none", axis.text.x = element_blank())

```

#### Make a function to check number of SNPs called per region to see if there are correlations?

```{r, fig.width=16, fig.height=4, eval = FALSE}

pSNPs_all %>% group_by(CHROM, E) %>% mutate(Count = 1) %>%
  arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            Min = ceiling(frollapply(POS, n = 10, FUN = min, align = "center")),
            Max = ceiling(frollapply(POS, n = 10, FUN = max, align = "center"))) %>%
  mutate(Window = 10/(Max - Min)) -> SNPsIDed

#Look at how they are throughout the genome
SNPsIDed %>% filter(CHROM != c("I"), CHROM != "M") %>%
  ggplot(aes(x = POS, y = Window, color = E)) + geom_line() +
  facet_grid(E == "A_REF"~CHROM, scales = "free", space = "free")

#Look at the correlation
SNPsIDed %>% select(CHROM, POS, Window, E) %>% na.omit() %>% 
  pivot_wider(names_from = E, values_from = Window) %>% na.omit() %>%
  pivot_longer(c(C8, F8, H8, Z8, C1, F1, H1, Z1), names_to = "Dataset", values_to = "NonREF") %>%
  ggplot(aes(x = A_REF, y = NonREF, color = Dataset)) + geom_point(alpha = 0.3) +
  geom_smooth(method = "lm")

#From this, it looks like Chr 12, 15, and maybe 4 all have higher SNP density?
SNPsIDed %>% select(CHROM, POS, Window, E) %>% na.omit() %>% 
  pivot_wider(names_from = E, values_from = Window) %>% na.omit() %>%
  pivot_longer(c(C8, F8, H8, Z8, C1, F1, H1, Z1), names_to = "Dataset", values_to = "NonREF") %>%
  ggplot(aes(x = A_REF, y = NonREF, color = CHROM)) + geom_point(alpha = 0.3) +
  geom_smooth(method = "lm") +
  facet_wrap(facets = "CHROM", nrow = 2)#, scales = "free")

SNPsIDed %>% select(CHROM, POS, Window, E) %>% na.omit() %>% 
  pivot_wider(names_from = E, values_from = Window) %>% na.omit() %>%
  pivot_longer(c(C8, F8, H8, Z8, C1, F1, H1, Z1), names_to = "Dataset", values_to = "NonREF") %>%
  ggplot(aes(x = A_REF, y = NonREF, color = CHROM)) + geom_point(alpha = 0.3) +
  geom_smooth(method = "lm")

```

### Function for Standard Plots

```{r, fig.width=16, fig.height=4}

cybrPurple <- function(dataset, 
                       cutoff = "III", 
                       peakslist = NULL, 
                       includeFixedGenes = TRUE, 
                       peakcolor = "#24588A90", 
                       mylim = NULL){
  #Make start and end points
  ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
    pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
    mutate(Summary = NA, Label = NA) %>% select(-delete)
  
  #Make Fixed Genes
  FG <- data.frame(Gene = c("Ura3", "MATalphastart", "MATalpha_end"),
                   CHROM = factor(c("V", "III", "III"), levels = as.character(as.roman(1:16))),
                   POS = c(116167, 198671,201177))
  
  #Make factors for coloring
  dataset$label <- factor(dataset$label, levels = c("Bulk", "Parent", "Interaction", "Rep", "Intercept"))

  dataset %>% select(CHROM, POS, Summary = summary, Label = label) %>% 
    rbind(ChromosomeScale) %>%
    ggplot(aes(x = POS, y = abs(Summary), color = Label, linetype = Label %in% c("Parent", "Rep"))) + 
    facet_grid(cols = vars(CHROM), scales = "free", space = "free") -> BasePlot
  
  #Add fixed genes to plot
  if(includeFixedGenes == TRUE){
    BasePlot <- BasePlot + geom_vline(data = FG, aes(xintercept = POS), color = "gray", size = 2, linetype = "dashed")
  }
  
  #Add peaks to plot
  if(is.null(peakslist) == FALSE){
    
    peakslist$CHROM <- factor(peakslist$CHROM, levels = as.character(as.roman(1:16)))
    BasePlot <- BasePlot + geom_vline(data = peakslist, aes(xintercept = POS), color = peakcolor, size = 2)
  }

  
  if(cutoff == "III"){
    FinalPlot <- BasePlot + 
                geom_line(size = 1.2) +
                #Cutoff
                geom_hline(aes(yintercept = max(abs(dataset$summary[dataset$CHROM == "III"]))), linetype = "dashed") +
                ylab("") + xlab("")+ ggtitle("")+
                scale_color_manual(values = c("black",  "gray40","#7030A0", "lightpink", "red")) + 
                theme(legend.position = "none", axis.text.x=element_blank(),
                          axis.ticks.x=element_blank()) 
    
  }else if(is.null(cutoff)){
    FinalPlot <- BasePlot + 
                geom_line(size = 1.2) +
                ylab("") + xlab("")+ ggtitle("")+
                # facet_grid(~CHROM, scales = "free", space = "free")  + 
                scale_color_manual(values = c("black",  "gray40","#7030A0", "lightpink", "red")) + 
                theme(legend.position = "none", axis.text.x=element_blank(),
                          axis.ticks.x=element_blank()) 
  }else{
    cutoff <- as.numeric(cutoff)
    FinalPlot <- BasePlot + 
                geom_line(size = 1.2) +
                geom_hline(aes(yintercept = cutoff)) +
                ylab("") + xlab("")+ ggtitle("")+
                # facet_grid(~CHROM, scales = "free", space = "free")  + 
                scale_color_manual(values = c("black",  "gray40","#7030A0", "lightpink", "red")) + 
                theme(legend.position = "none", axis.text.x=element_blank(),
                          axis.ticks.x=element_blank()) 
  }
  
  if(is.null(mylim) == FALSE){
    FinalPlot <- FinalPlot + ylim(0, mylim)
  }
  
  return(FinalPlot)
}
```


All GLM plots (not glmer)

```{r, fig.width=16, fig.height=4}
#GLM PLOTS
cybrPurple(dataset = Fluc_glm, peakslist = F_1_intpeaks, peakcolor = "#ED7B0190")
cybrPurple(Fluc_8_glm, peakslist = F_8_intpeaks, peakcolor = "#ED7B0190")

cybrPurple(Zeocin_glm, peakslist = Z_1_intpeaks, peakcolor = "#95417690")
cybrPurple(Zeocin_8_glmer, peakslist = Z_8_intpeaks, peakcolor = "#95417690")

cybrPurple(CuSO4_8_glm, peakslist = C_8_intpeaks, peakcolor = "#7DB0B090")
cybrPurple(CuSO4_glm, peakslist = C_1_intpeaks, peakcolor = "#7DB0B090")


#GLM PLOTS
# cybrPurple(Fluc_glm, peakslist = F_1_intpeaks_all, peakcolor = "#ED7B0110")
# cybrPurple(CuSO4_glm, peakslist = C_1_intpeaks_all, peakcolor = "#7DB0B010")
# cybrPurple(Zeocin_glm, peakslist = Z_1_intpeaks_all, peakcolor = "#95417610")
# cybrPurple(CuSO4_8_glm, peakslist = C_8_intpeaks_all, peakcolor = "#7DB0B010")
# cybrPurple(Fluc_8_glm, peakslist = F_8_intpeaks_all, peakcolor = "#ED7B0110")

#GLMER PLOTS
# cybrPurple(H2O2_8_glmer, peakslist = H8g_intpeaks, peakcolor = "#24588A90")
# cybrPurple(CuSO4_1_glmer, peakslist = C1g_intpeaks, peakcolor = "#7DB0B090")
# cybrPurple(H2O2_glmer, peakslist = H1g_intpeaks, peakcolor = "#24588A90")
# cybrPurple(Fluc_1_glmer, peakslist = F1g_intpeaks, peakcolor = "#ED7B0190")

```




### Zooming in on Chr 8 for CuSO4 and Fluconazole

```{r, fig.width=18, fig.height=4}

CuSO4_glm %>% rbind(data.frame(pSNPs[,1:2], summary = 0, label = "Reference")) %>%
  filter(CHROM == "VIII", label != "Parent", label != "Bulk") %>%
  #filter(POS < 225000, POS > 2e05) %>%
  ggplot(aes(x = POS, y = summary, color = label), alpha = 0.1) +
  scale_color_manual(values = c("black", "gray")) +
  theme(legend.position = "none") +
  geom_rect(aes(xmin=210848, xmax=211978, ymin=-Inf, ymax=Inf), fill = "pink", color = "pink", alpha = 0.1) + #CIC1
  geom_rect(aes(xmin=212535, xmax=212720, ymin=-Inf, ymax=Inf), fill = "darkturquoise", color = "darkturquoise", alpha = 0.1) + #CUP1 - 1
  geom_rect(aes(xmin=214533, xmax=214718, ymin=-Inf, ymax=Inf), fill = "turquoise", color = "turquoise", alpha = 0.1) + #CUP1 - 2
  geom_rect(aes(xmin=207654, xmax=209303, ymin=-Inf, ymax=Inf), fill = "pink", color = "pink", alpha = 0.1) + #SMF2
  geom_rect(aes(xmin=209705, xmax=210151, ymin=-Inf, ymax=Inf), fill = "pink", color = "pink", alpha = 0.1) + #COX6
  #geom_rect(aes(xmin=51111, xmax=52988, ymin=-Inf, ymax=Inf), fill = "orchid4", color = "orchid4", alpha = 0.1) + #RIM101
  #geom_rect(aes(xmin=439342, xmax=440127, ymin=-Inf, ymax=Inf), fill = "orchid4", color = "orchid4", alpha = 0.1) + #THP
  geom_point(size = 3, alpha = 0.5)
```


# Loading in Chr 8 Gene Data

ERG11

```{r, fig.width=16, fig.height=4}
chr8genes <- read.csv("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023/SGD_Tables/ChromosomeRegion_AllGenes.csv", header = FALSE)

chr8genes %>% select(V3, Gene = V5, POS = V10, EndPOS = V11, Description = V6, Char = V9) -> Chr8_list

Chr8_list %>% filter(Gene == "ERG11")

max(Chr8_list$POS)

Fluc_glm %>% filter(CHROM == "VIII") %>% ggplot(aes(x = POS, y = summary, color = label)) + geom_point(alpha = 0.5) +
  geom_vline(data = Chr8_list, aes(xintercept = POS, color = Char), alpha = 0.5) + 
  geom_vline(xintercept = 120091, color = "red", linetype = "dashed") +
  scale_color_manual(values = c("white", "black", "gray", "violet", "gray", "lightgray", "#954176", "gray"))


Fluc_glm %>% filter(CHROM == "VIII") %>% ggplot(aes(x = POS, y = summary, color = label)) + 
  geom_point(alpha = 0.5, size = 4) +
  geom_vline(xintercept = c(120091, 121683)) +
  geom_jitter(data = pSNPs, aes(x = POS, y = 0, color = "X")) +
  xlim(120091-1000, 121683+1000) +
  geom_vline(xintercept = c(121700, 121782), color = "turquoise") +
    ggtitle("ERG11 Gene Region, Fluconazole CSSI Data")
```


CUP1

```{r, fig.width=16, fig.height=4}
chr8genes <- read.csv("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023/SGD_Tables/ChromosomeRegion_AllGenes.csv", header = FALSE)

chr8genes %>% select(V3, Gene = V5, POS = V10, EndPOS = V11, Description = V6, Char = V9) -> Chr8_list

Chr8_list %>% filter(Gene %in% c("CUP1-1", "CUP1-2"))

max(Chr8_list$POS)

CuSO4_glm %>% filter(CHROM == "VIII") %>% ggplot(aes(x = POS, y = summary, color = label)) + geom_point(alpha = 0.5) +
  geom_vline(data = Chr8_list, aes(xintercept = POS, color = Char), alpha = 0.5) + 
  geom_vline(xintercept = 212535, color = "red", linetype = "dashed") +
  scale_color_manual(values = c("white", "black", "gray", "violet", "gray", "lightgray", "#954176", "gray"))


CuSO4_glm %>% filter(CHROM == "VIII") %>% ggplot(aes(x = POS, y = summary, color = label)) + 
  geom_point(alpha = 0.5, size = 4) +
  geom_vline(xintercept = c(212535, 212720, 214533, 214718)) +
  geom_jitter(data = pSNPs, aes(x = POS, y = 0, color = "X")) +
  xlim(212535-1000, 214718+1000) +
    ggtitle("CUP1 Gene Regions, CuSO4 CSSI Data")
```

How many SNPs are there?

```{r}
dim(pSNPs)

#This shows them diverging ~6 every 1kb...
1000*77580/12100000


```

# Let's actually load in all of the genes

Make new cutoffs so that we can find genes in these

```{r, fig.width=16, fig.height=4, eval = FALSE}
Cutoffs_perm <- readRDS("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/NewCutoffs_perm.rds")
Cutoffs_perm


#CALL PEAKS WITH NEW CUTOFFS
c1p <- cybr_callpeaks(CuSO4_glm, param = "Bulk", threshold = 9.865824, include_all = TRUE)
c8p <- cybr_callpeaks(CuSO4_8_glm, param = "Bulk", threshold = 9.753182, include_all = TRUE)
f1p <- cybr_callpeaks(Fluc_glm, param = "Bulk", threshold = 6.031745, include_all = TRUE)

z1p <- cybr_callpeaks(Zeocin_glm, param = "Bulk", threshold = 0.4024422, include_all = TRUE)
h8p <- cybr_callpeaks(H2O2_8_glmer, param = "Bulk", threshold = 2.0160408, include_all = TRUE)
z8p <- cybr_callpeaks(Zeocin_8_glmer, param = "Bulk", threshold = 2.0160408, include_all = TRUE)
f8p <- cybr_callpeaks(Fluc_8_glm, param = "Bulk", threshold = 0.4839316, include_all = TRUE)


#PLOT NEW CUTOFFS
cybrPurple(CuSO4_glm, peakslist = c1p, cutoff = 0.4098215)
cybrPurple(CuSO4_8_glm, peakslist = c8p, cutoff = 0.4839316, peakcolor = "#ED7B0190")
cybrPurple(Fluc_glm, peakslist = f1p, cutoff = 0.4260583)
cybrPurple(Zeocin_glm, peakslist = z1p, cutoff = 0.4024422)
cybrPurple(H2O2_8_glmer, peakslist = h8p, peakcolor = "#ED7B0190", cutoff = 2.0160408)
cybrPurple(Zeocin_8_glmer, peakslist = z8p, peakcolor = "#ED7B0190", cutoff = 2.0160408)
cybrPurple(Fluc_8_glm, peakslist = f8p, peakcolor = "#ED7B0190", cutoff = 0.4839316)


```

Not including all

```{r, fig.width=16, fig.height=4}
c1pn <- cybr_multpeaks(CuSO4_glm, param = "Bulk", threshold = 0.4098215, width = 400000)
c8pn <- cybr_multpeaks(CuSO4_8_glm, param = "Bulk", threshold = 0.4839316, width = 400000)
f1pn <- cybr_multpeaks(Fluc_glm, param = "Bulk", threshold = 0.4260583, width = 400000)
z1pn <- cybr_multpeaks(Zeocin_glm, param = "Bulk", threshold = 0.4024422, width = 400000)
h8pn <- cybr_multpeaks(H2O2_8_glmer, param = "Bulk", threshold = 2.0160408, width = 400000)
z8pn <- cybr_multpeaks(Zeocin_8_glmer, param = "Bulk", threshold = 2.0160408, width = 400000)
f8pn <- cybr_multpeaks(Fluc_8_glm, param = "Bulk", threshold = 0.4839316, width = 400000)

cybrPurple(CuSO4_glm, peakslist = c1pn, cutoff = 0.4098215) + ggtitle("CuSO4 CSS 1")
cybrPurple(CuSO4_8_glm, peakslist = c8pn, cutoff = 0.4839316, peakcolor = "#ED7B01")+ ggtitle("CuSO4 CSS 8")
cybrPurple(Fluc_glm, peakslist = f1pn, cutoff = 0.4260583)+ ggtitle("Fluconazole CSS 1")
cybrPurple(Zeocin_glm, peakslist = z1pn, cutoff = 0.4024422)+ ggtitle("Zeocin CSS 1")
cybrPurple(H2O2_8_glmer, peakslist = h8pn, peakcolor = "#ED7B01", cutoff = 2.0160408)+ ggtitle("H2O2 CSS 8")
cybrPurple(Zeocin_8_glmer, peakslist = z8pn, peakcolor = "#ED7B01", cutoff = 2.0160408)+ ggtitle("Zeocin CSS 8")
cybrPurple(Fluc_8_glm, peakslist = f8pn, peakcolor = "#ED7B01", cutoff = 0.4839316)+ ggtitle("Fluconazole CSS 8")


```

```{r}
SGD_Genes <- read.csv("SGD_Tables/AllGenes_results.csv")

#Turn this into a function

cybr2_SGDGenes <- function(peaks, GeneList = SGD_Genes, mywindow = 1000){

  peaks %>% select(CHROM, POS) -> peaks
  
  #ChatGPT code adapted for filtering (what is this crossing function?)
  filtered_df <- GeneList %>%
    mutate(CHROM = gsub(pattern = "chr", replacement = "", x = Chromosome)) %>%
    select(CHROM.x = CHROM, POS_Start, POS_End) %>%
    crossing(peaks) %>%                               # Create all combinations of df and numbers
    filter(CHROM.x == CHROM) %>%                      # Match category
    rowwise() %>%
    #filter(POS %in% seq((POS_Start - window), (POS_End + window))) %>%      # Check if number is within the range
    #OR
      filter(length(intersect(seq(POS - mywindow, POS + mywindow), seq(POS_Start, POS_End))) > 0) %>%

    select(CHROM, POS, POS_Start, POS_End) %>%        # Remove redundant columns
    distinct() %>%                                    # In case there are duplicate rows
    merge(., GeneList)%>% 
    arrange(CHROM, POS) %>% 
    select(-POS) %>% distinct()
  
  rm(peaks)
  
  filtered_df %>% 
  return()
}


```

Looking at a chromosome for where the genes actually are

Error in if (dataset$Window[i] == dataset$Window[i - 1]) { : 
  missing value where TRUE/FALSE needed
  
```{r, fig.width=16, fig.height=4}
mywidth <- 400000
F_1_intpeaks_m <- cybr_multpeaks(Fluc_glm, param = "Bulk", width = mywidth, buffer = 200)

Fluc_glm %>% filter(CHROM == "XV") %>%
      ggplot() +
  geom_rect(aes(xmin = 0, xmax = mywidth, ymin = 0, ymax = Inf), fill = "lightblue", alpha = 0.2) +
      geom_vline(data = F_1_intpeaks_m[F_1_intpeaks_m$CHROM == "XV",], 
      #geom_vline(data = F_1_intpeaks_m, 
                aes(xintercept = POS, color = as.factor(Check)), alpha = 0.3, size = 2) +
  geom_line(aes(x = POS, y = abs(summary), color = label)) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("black", "red", "black", "purple", "gray"))

Fluc_glm %>% #filter(CHROM == "VIII") %>%
      ggplot() +
  geom_rect(aes(xmin = 0, xmax = mywidth, ymin = 0, ymax = Inf), fill = "lightblue", alpha = 0.2) +
      #geom_vline(data = F_1_intpeaks_m[F_1_intpeaks_m$CHROM == "VIII",], 
      geom_vline(data = F_1_intpeaks_m, 
                aes(xintercept = POS, color = as.factor(Check)), alpha = 0.7, size = 1) +
  geom_line(aes(x = POS, y = abs(summary), color = label)) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("black", "red", "black", "purple", "gray"))

Fluc_glm %>% #filter(CHROM == "VIII") %>%
      ggplot() +
  geom_rect(aes(xmin = 0, xmax = mywidth, ymin = 0, ymax = Inf), fill = "lightblue", alpha = 0.2) +
      #geom_vline(data = F_1_intpeaks_m[F_1_intpeaks_m$CHROM == "VIII",], 
      geom_vline(data = f1pn, 
                aes(xintercept = POS, color = as.factor(Check)), alpha = 0.7, size = 1) +
  geom_line(aes(x = POS, y = abs(summary), color = label)) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("black", "red", "black", "purple", "gray"))

F_1_intpeaks_m
```


```{r, fig.width=16, fig.height=4}




# f1pn_Genes <- cybr2_SGDGenes(peaks = f1pn, mywindow = 4000)
# f8pn_Genes <- cybr2_SGDGenes(peaks = f8pn, mywindow = 4000)
# 
# z1pn_Genes <- cybr2_SGDGenes(peaks = z1pn, mywindow = 4000)
# z8pn_Genes <- cybr2_SGDGenes(peaks = z8pn, mywindow = 4000)


```

## Looking for Duplicates in Genes Called

Look at both copper traces

```{r, fig.width = 16, fig.height=5}
c1pn_Genes <- cybr2_SGDGenes(peaks = c1p, mywindow = 10000) 
c8pn_Genes <- cybr2_SGDGenes(peaks = c8p, mywindow = 10000) 

#c_Genes <- c1pn_Genes
c_Genes <- rbind(data.frame(c1pn_Genes, Fixed = "I"),
                 data.frame(c8pn_Genes, Fixed = "VIII"))

C_glm <- rbind(data.frame(CuSO4_glm, Fixed = "I"),
      data.frame(CuSO4_8_glm, Fixed = "VIII"))

#Chr 7
C_glm %>% filter(label == "Bulk", CHROM == c("VIII")) %>% 
  #ggplot(aes(x = POS, y =  abs(summary), linetype = Fixed)) +
  ggplot(aes(x = POS, y =  abs(summary))) +
  geom_line() +
  #geom_vline(data = c_Genes[c_Genes$CHROM == "VIII",], aes(xintercept = POS_Start, color = StandardName, linetype = Fixed)) +
  geom_vline(data = c_Genes[c_Genes$CHROM == "VIII",], aes(xintercept = POS_Start, color = StandardName)) +
  facet_grid(~CHROM, scales = "free", space = "free") 

c_Genes[c_Genes$CHROM == "VII",] %>% arrange(POS_Start)

#Chr 12
C_glm %>% filter(label == "Bulk", CHROM == c("XII")) %>% ggplot(aes(x = POS, y =  abs(summary), linetype = Fixed)) +
  geom_line() +
  geom_vline(data = c_Genes[c_Genes$CHROM == "XII",], aes(xintercept = POS_Start, color = StandardName, linetype = Fixed)) +
  facet_grid(~CHROM, scales = "free", space = "free") 

c_Genes[c_Genes$CHROM == "XII",] %>% arrange(POS_Start)

#Chromosome XV
C_glm %>% filter(label == "Bulk", CHROM == c("XV")) %>% ggplot(aes(x = POS, y =  abs(summary), linetype = Fixed)) +
  geom_line() +
  geom_vline(data = c_Genes[c_Genes$CHROM == "XV",], aes(xintercept = POS_Start, color = StandardName, linetype = Fixed)) +
  facet_grid(~CHROM, scales = "free", space = "free") 

c_Genes[c_Genes$CHROM == "XV",] %>% arrange(POS_Start)

#Chromosome XVI
C_glm %>% filter(label == "Bulk", CHROM == c("XVI")) %>% ggplot(aes(x = POS, y =  abs(summary), linetype = Fixed)) +
  geom_line() +
  geom_vline(data = c_Genes[c_Genes$CHROM == "XVI",], aes(xintercept = POS_Start, color = StandardName, linetype = Fixed)) +
  facet_grid(~CHROM, scales = "free", space = "free")

c_Genes[c_Genes$CHROM == "XVI",] %>% arrange(POS_Start)

```

Do the same for Fluconazole except let's look for duplicates in wider windows

```{r, fig.width = 16, fig.height=5}

f1pn_Genes <- cybr2_SGDGenes(peaks = f1pn, mywindow = 10000)
f8pn_Genes <- cybr2_SGDGenes(peaks = f8pn, mywindow = 10000)

f_Genes <- rbind(data.frame(f1pn_Genes, Fixed = "I"),
                 data.frame(f8pn_Genes, Fixed = "VIII"))

f_glm <- rbind(data.frame(Fluc_glm, Fixed = "I"),
      data.frame(Fluc_8_glm, Fixed = "VIII"))


################################################################################
 
 f_Genes %>% count(SystematicName) %>% filter(n > 1) %>% merge(f_Genes) %>% select(-Fixed) %>% distinct() -> f_Genes_2
 
 f_glm %>% filter(label == "Bulk") %>% ggplot() +
  geom_line(aes(x = POS, y =  abs(summary))) +
  geom_vline(data =  f_Genes_2, aes(xintercept = POS_Start, color = StandardName), size = 2, alpha = 0.4) +
  #geom_rect(data =  f_Genes_2, aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, fill = StandardName)) +
  facet_grid(~CHROM, scales = "free", space = "free") 

for(i in unique(f_Genes_2$CHROM)){
  myplot <- f_glm %>% filter(label == "Bulk", CHROM == i) %>% ggplot() +
  geom_line(aes(x = POS, y =  abs(summary), linetype = Fixed)) +
  #geom_vline(data =  f_Genes_2, aes(xintercept = POS_Start, color = StandardName), size = 2, alpha = 0.4) +
  geom_rect(data =  f_Genes_2[f_Genes_2$CHROM == i,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, 
                fill = paste(POS_Start, StandardName)), alpha = 0.3) +
  facet_grid(~CHROM, scales = "free", space = "free") 
  
  print(myplot)
}
```

Doing this method for CuSO4 now

```{r, fig.width = 16, fig.height=5}
c1pn_Genes <- cybr2_SGDGenes(peaks = c1pn, mywindow = 10000)
c8pn_Genes <- cybr2_SGDGenes(peaks = c8pn, mywindow = 10000)

c_Genes <- rbind(data.frame(c1pn_Genes, Fixed = "I"),
                 data.frame(c8pn_Genes, Fixed = "VIII"))

c_glm <- rbind(data.frame(CuSO4_glm, Fixed = "I"),
      data.frame(CuSO4_8_glm, Fixed = "VIII"))


################################################################################
 
 c_Genes %>% count(SystematicName) %>% filter(n > 1) %>% merge(c_Genes) %>% select(-Fixed) %>% distinct() -> c_Genes_2
 
 c_glm %>% filter(label == "Bulk") %>% ggplot() +
  geom_line(aes(x = POS, y =  abs(summary))) +
  geom_vline(data =  c_Genes_2, aes(xintercept = POS_Start, color = StandardName), size = 2, alpha = 0.4) +
  #geom_rect(data =  c_Genes_2, aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, fill = StandardName)) +
  facet_grid(~CHROM, scales = "free", space = "free") 

for(i in unique(c_Genes_2$CHROM)){
  myplot <- c_glm %>% filter(label == "Bulk", CHROM == i) %>% ggplot() +
  geom_line(aes(x = POS, y =  abs(summary), linetype = Fixed)) +
  #geom_vline(data =  c_Genes_2, aes(xintercept = POS_Start, color = StandardName), size = 2, alpha = 0.4) +
  geom_rect(data =  c_Genes_2[c_Genes_2$CHROM == i,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, 
                fill = paste(POS_Start, StandardName)), alpha = 0.3) +
  facet_grid(~CHROM, scales = "free", space = "free") 
  
  print(myplot)
}
```


```{r, eval = FALSE}

CuSO4_glm %>% filter(CHROM == "VIII") %>%
  ggplot() +
  geom_line(aes(x = POS, y = abs(summary), color = label)) +
  xlim(2e05, 2.5e05) +
  geom_rect(data = c1pn_Genes[c1pn_Genes$CHROM == "VIII",], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, fill = paste(POS_Start, StandardName)), alpha = 0.1) +
  geom_vline(data = c1pn[c1pn$CHROM == "VIII",], aes(xintercept = POS), linetype = "dashed")


# CuSO4_glm %>% filter(CHROM %in% c1pn$CHROM) %>%
#   ggplot() +
#   geom_line(aes(x = POS, y = abs(summary), color = label)) +
#   geom_rect(data = c1pn_Genes,#[c1pn_Genes$CHROM == "VIII",], 
#             aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, fill = paste(POS_Start, StandardName)), alpha = 0.3) +
#   geom_vline(data = c1pn,#[c1pn$CHROM == "VIII",], 
#              aes(xintercept = POS), linetype = "dashed") +
#   facet_grid(~CHROM, scales = "free", space = "free") #+
#   theme(legend.position = "bottom")
  
#Loop through chromosomes
  for(i in as.character(unique(c1pn$CHROM))){
    CuSO4_glm %>% filter(CHROM == i) %>%
      ggplot() +
      geom_line(aes(x = POS, y = abs(summary), color = label)) +
      geom_rect(data = c1pn_Genes[c1pn_Genes$CHROM == i,], 
                aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, fill = paste(POS_Start, StandardName)), alpha = 0.3) +
      geom_vline(data = c1pn[c1pn$CHROM == i,], 
                 aes(xintercept = POS), linetype = "dashed") +
      ggtitle(paste("Copper CSS I, Chr ", i, sep = "")) -> myplot
      #facet_grid(~CHROM, scales = "free", space = "free") +
      #xlim(min(c1pn$POS[c1pn$CHROM == i]) - 10000, max(c1pn$POS[c1pn$CHROM == i]) + 10000)  
      
    print(myplot)

  }
  
#Fluconazole
  for(i in as.character(unique(f1pn$CHROM))){
    Fluc_glm %>% filter(CHROM == i) %>%
      ggplot() +
      geom_line(aes(x = POS, y = abs(summary), color = label)) +
      geom_rect(data = f1pn_Genes[f1pn_Genes$CHROM == i,], 
                aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, fill = paste(POS_Start, StandardName)), alpha = 0.3) +
      geom_vline(data = f1pn[f1pn$CHROM == i,], 
                 aes(xintercept = POS), linetype = "dashed") +
      ggtitle(paste("Fluconazole CSS I, Chr ", i, sep = "")) -> myplot
      #facet_grid(~CHROM, scales = "free", space = "free") +
      #xlim(min(c1pn$POS[c1pn$CHROM == i]) - 10000, max(c1pn$POS[c1pn$CHROM == i]) + 10000)  
      
    print(myplot)

  }
  
  #Loop through chromosomes
  for(i in as.character(unique(z1pn$CHROM))){
    Zeocin_glm %>% filter(CHROM == i) %>%
      ggplot() +
      geom_line(aes(x = POS, y = abs(summary), color = label)) +
      geom_rect(data = z1pn_Genes[z1pn_Genes$CHROM == i,], 
                aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, fill = paste(POS_Start, StandardName)), alpha = 0.3) +
      geom_vline(data = z1pn[z1pn$CHROM == i,], 
                 aes(xintercept = POS), linetype = "dashed") +
      ggtitle(paste("Zeocin CSS I, Chr ", i, sep = "")) -> myplot
      #facet_grid(~CHROM, scales = "free", space = "free") +
      #xlim(min(c1pn$POS[c1pn$CHROM == i]) - 10000, max(c1pn$POS[c1pn$CHROM == i]) + 10000)  
      
    print(myplot)

  }
# Fluc_glm %>% filter(CHROM == "VIII") %>%
#   ggplot() +
#   geom_line(aes(x = POS, y = abs(summary), color = label)) +
#   xlim(1e05, 1.5e05) +
#   geom_rect(data = f1pn_Genes[f1pn_Genes$CHROM == "VIII",], 
#             aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, fill = paste(POS_Start, StandardName)), alpha = 0.1) +
#   geom_vline(data = f1pn[f1pn$CHROM == "VIII",], aes(xintercept = POS), linetype = "dashed")
# 
# Zeocin_glm %>% filter(CHROM == "VII") %>%
#   ggplot() +
#   geom_line(aes(x = POS, y = abs(summary), color = label)) +
#   #xlim(1e05, 1.5e05) +
#   geom_rect(data = z1pn_Genes[z1pn_Genes$CHROM == "VII",], 
#             aes(xmin = POS_Start, xmax = POS_End, ymin = 0, ymax = Inf, fill = paste(POS_Start, StandardName)), alpha = 0.1) +
#   geom_vline(data = z1pn[z1pn$CHROM == "VII",], aes(xintercept = POS), linetype = "dashed")

```


# Specific QTL Investigation
## Comparison of CSSI and CSS VIII

CuSO4 comparison 

```{r, fig.width=17,fig.height=8}

rbind(data.frame(CuSO4_8_glm, Exp = "CSS8"),
      data.frame(CuSO4_glm, Exp = "CSS1a"),
      data.frame(CuSO4_glm2, Exp = "CSS1b")) -> CUSO4b

max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))
max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS1a"]))
max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS1b"]))


CUSO4b %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","turquoise")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)


```

Trying to find how close the suspected QTL are...

```{r, warning=FALSE, message=FALSE, fig.width=17,fig.height=8}

KnownLoci <- data.frame(Gene = c("SSU1", "MAC1", "CTR1","CTR3", "CUP1"), 
                        CHROM = c("XVI", "XIII", "XVI", "XII", "VIII"), 
                        POS = c(373793, 317165, 786208, 947253, 212535))

URA3 <- data.frame(Gene = "Ura3",
                   CHROM = "V",
                   POS = 116167)

CUSO4b %>% ggplot(aes(x = POS, y = abs(summary), color = label))+
  # geom_vline(data = summaryofprecision_CuSO4, aes(xintercept = mean), color = "pink", size =2) + 
  geom_vline(data = KnownLoci, aes(xintercept = POS), color = "firebrick", size = 1.2, linetype = "dashed")+
  geom_vline(data = URA3, aes(xintercept = POS), color = "gold", size = 1.2)+
  geom_line(size = 1) + 
    geom_hline(aes(yintercept = max(abs(CUSO4b$summary[CUSO4b$CHROM == "III" & CUSO4b$Exp == "CSS8"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("CuSO4 Comparison")+
  facet_grid(Exp ~ CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("plum4", "gray","black","turquoise")) + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,12)

```

## Copper QTLs: Load in datasets

```{r, warning=FALSE, message=FALSE}
copperdata <- read.csv("CopperGenes_gdoc.csv")

copperdata %>% filter(StudyType == "QTL") %>% transmute(CHROM = as.character(as.roman(CHROM)), POS = StartLocation, Study = StudyType, Gene = Gene) -> copperdata2

dim(copperdata2)

SpecificCuSO4 <- filter(copperdata2, CHROM %in% c("V", "VII", "VIII", "XI", "XII", "XVI", "XV")) %>% arrange(CHROM, POS)

subset(SpecificCuSO4, CHROM == "XII")

actualQTL <- data.frame(Gene = c("VMA4", "MTL1", "CIC1","PTR2", "IRC15", "HSP60", "CUP1", "LEU9"), 
                        CHROM = factor(c("V", "VII", "VIII", "XI", "XVI", "XII", "VIII", "XV"), 
                                       levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI")), 
                        POS = c(100451, 539938, 211292, 617558, 519854, 664127, 212535, 523702)) 

actualQTL %>% arrange(CHROM)

filteredQTL <- copperdata2 %>% filter(CHROM %in%  c("V", "VII", "VIII", "XI", "XII", "XVI") == FALSE)
filteredQTL$CHROM <- factor(filteredQTL$CHROM,
                            levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI"))
```




Fluconazole Chr 1
```{r, fig.width = 16, fig.height=4}

FlucGenes <- data.frame(Gene = c("ERG11", "LSM2", "TIM12", "RPB7", "TCP1", "RSC6", "RPL32", "RSC9", "HSP150"),
                        CHROM = as.character(as.roman(c(8,2,2,4,4,3,2,13, 10))),
                        POS = c(120091, 170623, 427155, 1276654, 887232, 214994, 45978, 17064, 120449))

Fluc_glm %>% mutate(Sub = 1) %>% rbind(data.frame(Fluc_8_glm, Sub = 8), 
                                       data.frame(ChromosomeScale2, Sub = NA)) %>%
  filter(label == "Bulk") %>%
  ggplot(aes(x = POS, y = abs(summary), color = as.factor(Sub))) + 
    #geom_vline(data = F_1_intpeaks, aes(xintercept = POS), color = "#7DB0B0", size = 2) +
  geom_vline(data = FlucGenes, aes(xintercept = POS), color = "firebrick", linetype = "dashed", size = 2) +
  geom_vline(data = engineered, aes(xintercept = POS_start), color = "gray", size = 2, alpha = 0.6) +
  geom_vline(data = engineered, aes(xintercept = POS_end), color = "gray", size = 2, alpha = 0.6) +
  geom_line(size = 1.2) + 
    geom_hline(aes(yintercept = max(abs(Fluc_glm$summary[Fluc_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("black",  "darkorange3", "gray")) + 
  #scale_color_manual(values = c("gray",  "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ylim(0,8)

```

## Exploratory Plots

```{r, fig.height=6, fig.width=16}

rbind(data.frame(CuSO4_8_glm, Exp = "CuSO4_8"),
      #data.frame(CuSO4_glm, Exp = "CuSO4_1a"),
      #data.frame(CuSO4_glm2, Exp = "CuSO4_1b"),
      data.frame(CuSO4_1_glmer, Exp = "CuSO4_1"),
      data.frame(Fluc_8_glm, Exp = "Fluc_8"),
      data.frame(Fluc_glm, Exp = "Fluc_1"),

      data.frame(Zeocin_8_glmer, Exp = "Zeo_8"),
      data.frame(Zeocin_glm, Exp = "Zeo_1"))-> All_Experiments

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1, alpha = 0.8) + 
  scale_color_manual(values = c("black","darkorange3")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + ylim(0, 20) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1, alpha = 0.8) + 
  scale_color_manual(values = c("black","black", "black", "black")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + ylim(0, 20) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")
```

## Comparison of Additive QTL in all traits

```{r, fig.height=6, fig.width=16, eval = FALSE}

rbind(data.frame(CuSO4_8_glm, Exp = "CuSO4_8"),
      data.frame(CuSO4_glm, Exp = "CuSO4_1a"),
      data.frame(CuSO4_glm2, Exp = "CuSO4_1b"),
      data.frame(Fluc_8_glm, Exp = "Fluc_8"),
      data.frame(Fluc_glm, Exp = "Fluc_1"),
      data.frame(H2O2_glmer, Exp = "H2O2_1"),
      data.frame(H2O2_8_glmer, Exp = "H2O2_8"),
      data.frame(Zeocin_8_glmer, Exp = "Zeo_8"),
      data.frame(Zeocin_glm, Exp = "Zeo_1"))-> All_Experiments

All_Experiments %>% filter(label == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(summary), color = Exp)) + geom_line(size = 1) + 
  scale_color_manual(values = c("turquoise", "turquoise3", "turquoise2", "firebrick", "firebrick2", "orchid", "darkorchid", "gold", "darkorange")) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = Selection, linetype = FixedChr)) + geom_line(size = 1) + 
  scale_color_manual(values = c("turquoise3","firebrick2", "orchid", "darkorange")) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1, alpha = 0.8) + 
  scale_color_manual(values = c("black","black", "black", "darkorange3")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + ylim(0, 15) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1, alpha = 0.8) + 
  scale_color_manual(values = c("black","black", "black", "black")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + ylim(0, 15) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")

All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  filter(Selection == "CuSO4") %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 1) + 
  scale_color_manual(values = c("black","#00000090", "red3")) +
  facet_grid(~CHROM, scales = "free", space = "free") + ylim(0, 15) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("CuSO4 Additive QTL") + ylab("Bulk Z Score")


All_Experiments %>% filter(label == "Bulk") %>% separate(Exp, sep = "_", into = c("Selection", "FixedChr")) %>%
  ggplot(aes(x = POS, y = abs(summary), color = FixedChr)) + geom_line(size = 0.8, alpha = 0.8) + 
  scale_color_manual(values = c("#24588A","#24588A", "#24588A", "#ED7B01")) +
  facet_grid(Selection~CHROM, scales = "free", space = "free") + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("Comparison of Additive QTL") + ylab("Bulk Z Score")
```



```{r}
Albert_eQTLInteractors <- data.frame(Gene = c("HAP1", "MKT1", "FLO8", "FLO11"),
                                     CHROM = c(12, 14, 5, 9),
                                     POS = c(646415, 467131, 375215, 389572))
```

Where are the ERG genes?

```{r, fig.width=17, fig.height=5}
SGD_Genes %>%  
  mutate(CHROM = gsub(pattern = "chr", replacement = "", x = Chromosome)) %>% 
  filter(., grepl("ERG",StandardName)) -> ERG_Genes

f_glm %>% filter(label == "Bulk") %>% ggplot() +
  geom_line(aes(x = POS, y =  abs(summary), linetype = Fixed)) +
  geom_vline(data =  ERG_Genes, aes(xintercept = POS_Start, color = StandardName), size = 2, alpha = 0.4) +
  facet_grid(~CHROM, scales = "free", space = "free") + theme(legend.position = "none")

for(i in c("VII", "VIII", "XIII", "XVI")){
  f_glm %>% filter(label == "Bulk", CHROM == i) %>% ggplot() +
  geom_line(aes(x = POS, y =  abs(summary), linetype = Fixed)) +
  geom_vline(data =  ERG_Genes[ERG_Genes$CHROM == i,], aes(xintercept = POS_Start, color = StandardName), size = 2, alpha = 0.4) +
  facet_grid(~CHROM, scales = "free", space = "free") -> myplot
  
  print(myplot)

}

```
Try making a plot that has all of the genes but around the peaks

Chr II of Fluconazole

```{r, fig.width=16, fig.height=6}
f_glm %>% filter(label == "Bulk", CHROM == "II") -> f_II

SGD_Genes %>%  
  mutate(CHROM.x = gsub(pattern = "chr", replacement = "", x = Chromosome)) %>% 
  filter(CHROM.x == "II") %>% crossing(f_II) %>% rowwise() %>%
    filter(POS %in% seq((POS_Start), (POS_End))) %>%
  group_by(SystematicName) %>%
  summarize(y = max(abs(summary)),
            StandardName = StandardName,
            POS_Start = POS_Start,
            POS_End = POS_End, Strand = Strand) %>% distinct()-> SGD_II_Filtered

f_II %>% ggplot() +
  geom_line(aes(x = POS, y = abs(summary), linetype = Fixed)) +
  geom_rect(data = SGD_II_Filtered[SGD_II_Filtered$y > 4 & SGD_II_Filtered$Strand == 1,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.3, ymax = y-0.1, fill = paste(Strand,POS_Start,StandardName))) +
  geom_rect(data = SGD_II_Filtered[SGD_II_Filtered$y > 4 & SGD_II_Filtered$Strand == -1,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.7, ymax = y-0.5, fill = paste(Strand,POS_Start,StandardName))) +
  theme(legend.position = "none")

f_II %>% ggplot() +
  geom_line(aes(x = POS, y = abs(summary), linetype = Fixed)) +
  geom_rect(data = SGD_II_Filtered[SGD_II_Filtered$y > 4 & SGD_II_Filtered$Strand == 1,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.3, ymax = y-0.1, fill = paste(Strand,POS_Start,StandardName))) +
  geom_rect(data = SGD_II_Filtered[SGD_II_Filtered$y > 4 & SGD_II_Filtered$Strand == -1,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.7, ymax = y-0.5, fill = paste(Strand,POS_Start,StandardName))) +
  xlim(4e05, 5e5)
```

Chromosome XV of Fluconazole

```{r, fig.width=16, fig.height=6}
f_glm %>% filter(label == "Bulk", CHROM == "XV") -> f_XV

SGD_Genes %>%  
  mutate(CHROM.x = gsub(pattern = "chr", replacement = "", x = Chromosome)) %>% 
  filter(CHROM.x == "XV") %>% crossing(f_XV) %>% rowwise() %>%
    filter(POS %in% seq((POS_Start), (POS_End))) %>%
  group_by(SystematicName) %>%
  summarize(y = max(abs(summary)),
            StandardName = StandardName,
            POS_Start = POS_Start,
            POS_End = POS_End, Strand = Strand) %>% distinct()-> SGD_XV_Filtered

SGD_Genes %>% 
  mutate(CHROM.x = gsub(pattern = "chr", replacement = "", x = Chromosome)) %>% 
  filter(CHROM.x == "XV", SystematicName %in% SGD_XV_Filtered$SystematicName == FALSE) %>% mutate(y = 8) %>% 
  select(SystematicName, y, StandardName, POS_Start, POS_End, Strand) %>%
  rbind(SGD_XV_Filtered) -> SGD_XV_Filtered_all

f_XV %>% ggplot() +
  geom_line(aes(x = POS, y = abs(summary), linetype = Fixed)) +
  geom_rect(data = SGD_XV_Filtered[SGD_XV_Filtered$y > 4 & SGD_XV_Filtered$Strand == 1,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.3, ymax = y-0.1, fill = paste(Strand,POS_Start,StandardName))) +
  geom_rect(data = SGD_XV_Filtered[SGD_XV_Filtered$y > 4 & SGD_XV_Filtered$Strand == -1,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.7, ymax = y-0.5, fill = paste(Strand,POS_Start,StandardName))) +
  theme(legend.position = "none")

# f_XV %>% ggplot() +
#   geom_line(aes(x = POS, y = abs(summary), linetype = Fixed)) +
#   geom_rect(data = SGD_XV_Filtered[SGD_XV_Filtered$y > 3.5 & SGD_XV_Filtered$Strand == 1,], 
#             aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.3, ymax = y-0.1, fill = paste(Strand,POS_Start,StandardName))) +
#   geom_rect(data = SGD_XV_Filtered[SGD_XV_Filtered$y > 3.5 & SGD_XV_Filtered$Strand == -1,], 
#             aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.7, ymax = y-0.5, fill = paste(Strand,POS_Start,StandardName))) +
#   xlim(2.5e05, 3e5)

f_XV %>% ggplot() +
  geom_point(aes(x = POS, y = abs(summary))) +
  geom_rect(data = SGD_XV_Filtered_all[SGD_XV_Filtered_all$Strand == 1 & 
                                         SGD_XV_Filtered_all$POS_End < 3.25e5 & SGD_XV_Filtered_all$POS_Start > 2.5e05,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.3, ymax = y-0.1, fill = (paste(Strand,POS_Start,StandardName)))) +
  geom_rect(data = SGD_XV_Filtered_all[SGD_XV_Filtered_all$Strand == -1 & 
                                         SGD_XV_Filtered_all$POS_End < 3.25e5 & SGD_XV_Filtered_all$POS_Start > 2.5e05,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.7, ymax = y-0.5, fill = (paste(Strand,POS_Start,StandardName)))) +
  xlim(2.5e05, 3.25e5) 

f_XV %>% ggplot() +
  geom_point(aes(x = POS, y = abs(summary))) +
  geom_rect(data = SGD_XV_Filtered_all[SGD_XV_Filtered_all$Strand == 1 & 
                                         SGD_XV_Filtered_all$POS_End < 9e5 & SGD_XV_Filtered_all$POS_Start > 8.25e05,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.3, ymax = y-0.1, fill = (paste(Strand,POS_Start,StandardName)))) +
  geom_rect(data = SGD_XV_Filtered_all[SGD_XV_Filtered_all$Strand == -1 & 
                                         SGD_XV_Filtered_all$POS_End < 9e5 & SGD_XV_Filtered_all$POS_Start > 8.25e05,], 
            aes(xmin = POS_Start, xmax = POS_End, ymin = y-0.7, ymax = y-0.5, fill = (paste(Strand,POS_Start,StandardName)))) +
  xlim(8.25e05, 9e5) 
```


