---
title: "H2O2 Testing for Resequencing"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)

require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)

require(cybrBSA)
require(lme4)

#TESTING
#glm_formula <- "Allele ~ Bulk * Parent + Rep"

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 10)))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))

```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

### Functions
```{r}
glm_cb2_short <- function(..., W, formula, numgroups = FALSE, outputlength = 4, return = c("Z")) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  #MAYBEWORKS
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  #Return specific ones like z-score only
  if(return %in% "Z"){
    output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients)*0.5)+1):((length(summary(glm_fit)$coefficients)*0.75))]
  }
  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
}

glm_cb2_short_effect <- function(..., W, formula, numgroups = FALSE, outputlength = 4) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  #MAYBEWORKS
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  #Return effects only

  output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients)*0)+1):((length(summary(glm_fit)$coefficients)*0.25))]

  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
}

glmer_cb2_short <- function(..., W, formula, numgroups = FALSE, outputlength = 4, return = c("Z")) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  #MAYBEWORKS
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glmer(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  #Return specific ones like z-score only
  if(return %in% "Z"){
    output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients)*0.5)+1):((length(summary(glm_fit)$coefficients)*0.75))]
  }
  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
}
############################ PERMUTATIONS ######################################
# Replicates, all scrambled
cybrPermute_Rep <- function(dataset){
  
  start.time <- Sys.time()
  
  dataset %>% 
  distinct() %>% ungroup() %>%
  group_by(CHROM, POS, Allele, 
           Bulk, 
           Rep, #might not have these
           Parent) %>% 
  summarize(culprits = length((SmoothCount))) %>% 
  merge(dataset) %>% 
  filter(culprits == 1) %>% 
  ungroup() %>%
  distinct() %>% #THIS IS IMPORTANT
  pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest

  #these are now all of the ones that can be used to permute
  newnewtest %>% filter(Bulk != "F", CHROM != "I", CHROM != "III", CHROM != "VIII") %>% select(CHROM, Oak, Wine) %>% 
    mutate(rownum = sample(1:nrow(.))) %>% arrange(rownum) %>% na.omit() -> NullToSample
  
  #Trying this again
  
  dataset[1:nrow(NullToSample),] %>% ungroup() %>% select(CHROM, POS, Bulk, Parent, Rep) %>% cbind(NullToSample[,c("Oak", "Wine")]) %>%
    pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") %>%
    #Original Script
    group_by(CHROM, POS) %>%
    mutate_if(is.character, as.factor) %>%
      summarize(Summary = glm_cb2_short(Allele = Allele,
                               Bulk = Bulk,
                               Parent = Parent,
                               Rep = Rep,
                               W = SmoothCount,
                               formula = "Allele ~ Bulk * Parent + Rep",
                               outputlength = 5),
              Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> glmresult

  end.time = Sys.time()
  print(end.time - start.time)
  
  return(glmresult)
  
}

# No replicates, fix CHROM, Parent
# No replicates, fix Parent

# Replicates, fix CHROM, Parent
cybrPermute_byCHRParent_Rep <- function(dataset){
   start.time <- Sys.time()
  
   print("Make sure that dilute bulk is labeled D")
   
    dataset %>% 
      distinct() %>% ungroup() %>%
      group_by(CHROM, POS, Allele, 
               Bulk, 
               Rep, #might not have these
               Parent) %>% 
      summarize(culprits = length((SmoothCount))) %>% 
      merge(dataset) %>% 
      filter(culprits == 1) %>% 
      ungroup() %>%
      distinct() %>% #THIS IS IMPORTANT
      pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    
    #these are now all of the ones that can be used to permute
    newnewtest %>% filter(Bulk == "D") %>% select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(CHROM, Parent) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine, POS = POS,
                POS2 = sample(POS)) %>%
      group_by(CHROM, Parent, POS2) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine,
                POS = POS2, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA
    
    newnewtest %>% filter(Bulk == "D") %>% select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(CHROM, Parent) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine, POS = POS,
                POS2 = sample(POS)) %>%
      group_by(CHROM, Parent, POS2) %>%
      summarize(CHROM = CHROM, Parent = Parent, Oak = Oak, Wine = Wine,
                POS = POS2, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB
    
    rbind(shuffled_DiluteA, shuffled_DiluteB) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm
    
    #Trying this again
    
    shuffletoglm %>% na.omit() %>%
      #Original Script
      group_by(CHROM, POS) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_short(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = "Allele ~ Bulk * Parent + Rep",
                                numgroups = 16, outputlength = 5),
                Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> glmresult
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}

# Replicates, fix Parent
cybrPermute_byParent_Rep <- function(dataset){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled D")

  dataset %>% 
  distinct() %>% ungroup() %>%
  group_by(CHROM, POS, Allele, 
           Bulk, 
           Rep, #might not have these
           Parent) %>% 
  summarize(culprits = length((SmoothCount))) %>% 
  merge(dataset) %>% 
  filter(culprits == 1) %>% 
  ungroup() %>%
  distinct() %>% #THIS IS IMPORTANT
  pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
  
  #these are now all of the ones that can be used to permute
    newnewtest %>% filter(Bulk == "D",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, 
                Loc = sample(Loc)) %>%
      group_by(Parent, Loc) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine,
                Loc = Loc, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "D",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, 
                Loc = sample(Loc)) %>%
      group_by(Parent, Loc) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine,
                Loc = Loc, 
                Rep = sample(c("a", "b"))) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #Trying this again
    
    shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_short(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = "Allele ~ Bulk * Parent + Rep",
                                numgroups = 16, outputlength = 5),
                Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> glmresult2
  
    end.time = Sys.time()
    print(end.time - start.time)
    
    return(glmresult2)
}
  
```

### Load Data

```{r}
alldata_called <- readRDS(file = "alldata_called.rds")

alldata_called %>% filter(Pool == "HJ5HKDRX3") -> H2O2_CSS8
rm(alldata_called)
#HJ5HKDRX3_smoothed<- readRDS(file = "StandardGLM/HJ5HKDRX3_smoothed.rds")

H2O2_CSS8 %>% group_by(Pool, Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 300, FUN = median, align = "center"))) -> alldata_smoothed_300

H2O2_CSS8 %>% filter(CHROM == "II") %>% ggplot(aes(x = POS, y = Coverage, color = Dataset)) + geom_point() + facet_grid(~CHROM)
```

### Run GLM the new way without smoothing

```{r}
H2O2_CSS8 %>% separate(Dataset, into = c("Bulk", "Parent", "Rep"), sep = "_") %>% 
  mutate(Parent = gsub("0", "O", Parent)) %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '8')) %>% 
  filter(Bulk != "Z") %>%
  mutate_if(is.character, as.factor) %>%
  filter(CHROM %in% c("M", "VIII") == FALSE) -> H2O2_glmprep


#In group 172: `CHROM = "I"`, `POS = 19750`.
H2O2_glmprep %>% #filter(CHROM == "I", POS == 19750) %>% 
  group_by(POS, CHROM) %>% 
  na.omit() %>% summarize(B = length(unique(Bulk)),
                          P = length(unique(Parent)),
                          R = length(unique(Rep))) %>%
  filter(B > 1, P > 1) %>%
  merge(H2O2_glmprep) -> H2O2_glmprep_forrealz

#SET CONTRASTS
contrasts(H2O2_glmprep_forrealz$Bulk)
contrasts(H2O2_glmprep_forrealz$Parent)

H2O2_glmprep_forrealz %>% #filter(CHROM == "II") %>%
  group_by(CHROM, POS) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             W = Reads,
                             formula = "Allele ~ Bulk * Parent", outputlength = 4),
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> H2O2_CSS8_glm_2

# for(i in unique(H2O2_CSS8_glm_2$CHROM)){
#   H2O2_CSS8_glm_2 %>% filter(CHROM == i) %>%
#     ggplot(aes(x = POS, y = Summary, color = Factor)) + geom_point() + facet_wrap("Factor") -> plot2
#   
#   print(plot2)
# 
# }

H2O2_CSS8_glm_2 %>% filter(Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(Summary))) + 
  geom_point(alpha = 0.2) + 
  facet_grid(~CHROM, space = "free", scales = "free")

```

```{r, fig.width=10, fig.height=4}
H2O2_CSS8_glm_2 %>% group_by(CHROM, Factor) %>% arrange(POS) %>%
  summarize(POS = POS, RollSummary = frollapply(Summary, 20, mean, align = "center")) -> roll20

H2O2_CSS8_glm_2 %>% group_by(CHROM, Factor) %>% arrange(POS) %>%
  summarize(POS = POS, RollSummary = frollapply(Summary, 200, mean, align = "center")) -> roll200

roll20 %>% filter(Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(RollSummary))) + 
  geom_point(alpha = 0.2) + 
  facet_grid(~CHROM, space = "free", scales = "free")

roll200 %>% filter(Factor == "Bulk") %>% 
  ggplot(aes(x = POS, y = abs(RollSummary))) + 
  geom_line() + 
  facet_grid(~CHROM, space = "free", scales = "free")

roll200 %>% filter(Factor != "Intercept") %>%
  ggplot(aes(x = POS, y = abs(RollSummary), color = Factor)) + 
  geom_line() + 
  facet_grid(~CHROM, space = "free", scales = "free")
```

### Run GLM

```{r, eval = FALSE}

# Pick out the specific groups of the dataset
HJ5HKDRX3_smoothed %>% separate(Dataset, into = c("Bulk", "Parent", "Rep"), sep = "_") %>% 
  mutate(Parent = gsub("0", "O", Parent)) %>%
  #Filter for Chr 8 and Bulk D or H
  filter(stringr::str_detect(Parent, '8')) %>% 
  filter(Bulk != "Z") %>%
  mutate_if(is.character, as.factor) -> HJ5HKDRX3_CSS8_glm_prep

#SAVE FOR PERMUTATIONS
#saveRDS(HJ5HKDRX3_CSS8_glm_prep, file = "StandardGLM/HJ5HKDRX3_CSS8_glm_prep.rds")

#Test once
# testdata <- HJ5HKDRX3_CSS8_glm_prep %>% na.omit() %>% filter(CHROM == "X", POS == 640028) %>% arrange(Bulk, Parent, Rep)
# table(testdata$Rep, testdata$Bulk, testdata$Parent)
# sum(as.data.frame(table(testdata$Rep, testdata$Bulk, testdata$Parent))$Freq) == 16
# formula = "Allele ~ Bulk * Parent + Rep"
# glm(formula = formula, family = "binomial", data = testdata, weights = SmoothCount)

# Run full dataset
HJ5HKDRX3_CSS8_glm_prep %>% na.omit() %>% 
  #filter(CHROM == "X", POS == 640028) %>%
  #REMOVE CHROMOSOME OF FIXATION (don't waste resources on that)
  filter(CHROM %in% c("M", "VIII") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(Summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent + Rep",
                            numgroups = 16, outputlength = 5),
            Factor = (c("Intercept", "Bulk", "Parent", "Rep", "Interaction"))) -> HJ5HKDRX3_CSS8_glm_all

HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>% dim()

# #Merge with GQ if you want
# alldata %>% filter(Pool == "HJ5HKDRX3", CHROM %in% c("M", "VIII") == FALSE) %>% mutate_if(is.character, as.factor) -> alldata_H2O2
# HJ5HKDRX3_CSS8_glm_all %>% na.omit() %>%
#   merge(alldata_H2O2[,c("POS", "CHROM", "GQ")]) -> HJ5HKDRX3_CSS8_merge_all

#saveRDS(HJ5HKDRX3_CSS8_glm_all, file = "StandardGLM/HJ5HKDRX3_CSS8_glm_all.rds")

#HJ5HKDRX3_CSS8_perm <- cybrPermute_Rep(dataset = HJ5HKDRX3_CSS8_glm_prep)

```

```{r}
#H2O2 CSS8
HJ5HKDRX3_CSS8_glm_prep <- readRDS("StandardGLM/HJ5HKDRX3_CSS8_glm_prep.rds")
HJ5HKDRX3_CSS8_glm_prep_noreps <- readRDS("HJ5HKDRX3_CSS8_glm_prep_noreps.rds") %>% mutate(SmoothCount = AvgCount)
```
```

