---
title: "Arrayed GLMER"
author: "Cassandra Buzby"
date: "2023-05-14"
output: html_document
---

## Overview

Because running mixed-effect models is so time- and memory-consuming, I aim to array each position within a chromosome (of smoothed data) to calculate the glmer for each and then concatenate back together in bash. This document should serve to create and test this Rscript.


```{r}
setwd("../AnalysisAndFigures_2023/")

#Load Stuff
CSSI_CuSO4 <- readRDS("Data/CuSO4_1_cybr2.rds")
CSSI_CuSO4_2 <- readRDS("Data/CuSO4_2_cybr2.rds")
CSSI_Fluc <- readRDS("Data/Fluconazole_1_cybr2.rds")
CSSI_Zeocin <- readRDS("Data/Zeocin_cybr2.rds")

#CSS 8 Stuff
CSSVIII_CuSO4 <- readRDS("Data/CuSO4_CSS8_cybr2.rds")

```

##Save each one

Chr 8 fixed (includes all, must be filtered inside script)
```{r}
#Find the rolling median or change n instead
CSSVIII_CuSO4 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData
rollData %>% filter(CHROM != "M") -> rollData

saveRDS(rollData, file = "rollData_Chr8_CuSO4.rds")

#rm(rollData)
```

Chr 1 CuSO4 (the first one)
```{r}
#Find the rolling median or change n instead
CSSI_CuSO4 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

rollData %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Dose", "Allele"), names_sep = "_", values_to = "Reads") %>% ungroup() %>% select(-CHROM, -POS, -Reads,-Allele) %>% summarize(summary = unique(Bulk))
rollData %>% filter(CHROM != "M") -> rollData

saveRDS(rollData, file = "rollData_Chr1a_CuSO4.rds")

rm(rollData)
```

Chr 1 the second CuSO4 experiment
```{r}
#Find the rolling median or change n instead
CSSI_CuSO4_2 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

rollData %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Dose", "Allele"), names_sep = "_", values_to = "Reads") %>% ungroup() %>% select(-CHROM, -POS, -Reads,-Allele) %>% summarize(summary = unique(Bulk))
rollData %>% filter(CHROM != "M") -> rollData

saveRDS(rollData, file = "rollData_Chr1b_CuSO4.rds")

rm(rollData)
```

Chr 1 Fluconazole
```{r}
#Find the rolling median or change n instead
CSSI_Fluc %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() %>% 
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

rollData %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Dose", "Allele"), names_sep = "_", values_to = "Reads") %>% ungroup() %>% select(-CHROM, -POS, -Reads,-Allele) %>% summarize(summary = unique(Bulk))

rollData %>% filter(CHROM != "M") -> rollData

saveRDS(rollData, file = "rollData_Chr1_Fluc.rds")

rm(rollData)
```

## Setting arguments for bash script

```{r}
arraynum <- 15
args <- c(arraynum)
```


## Actual bash scripts


Copper Sulfate Chr 8 - 22,000 positions, so we have to loop through 1000 20 times? D:

```{r}
getwd()
dir()
rollData <- readRDS("rollData_Chr8_CuSO4.rds")
dim(rollData)
rm(rollData)
```

```{r}
#!/usr/bin/env Rscript
#args = commandArgs(trailingOnly=TRUE)

options(warn=-1)

#Load in all packages
require(tidyr)
require(reshape2)
require(data.table)
require(dplyr)
require(foreach)
require(doParallel)
require(scales)
require(lme4)

rollData <- readRDS("rollData_Chr8_CuSO4.rds")

#Find the position necessary
rollData[args[1],] %>% pivot_longer(c(-CHROM, -POS), 
                                    names_to = c("Bulk", "Parent", "Day", "Dose", "Allele"), 
                                    names_sep = "_", 
                                    values_to = "Reads") %>% 
  filter(Bulk %in% c("Dilute", "CuSO4"), Parent %in% c("Wine8", "Oak8")) -> mydata

mydata$Bulk <- factor(mydata$Bulk)
mydata$Parent <- factor(mydata$Parent)
mydata$Allele <- factor(mydata$Allele)
mydata$Day <- factor(mydata$Day)

glmeroutput <-  glmer(Allele ~ Bulk*Parent+(1|Day),
             weights = Reads, family = binomial,
              data = mydata)

sumd <- summary(glmeroutput)

#sumd$coefficients[1:(.75*length(sumd$coefficients))]

output <- c(unique(mydata$CHROM),
            unique(mydata$POS),
            sumd$coefficients[1:(.75*length(sumd$coefficients))])


write(paste0(output, collapse = ","), stdout())

```
## Working with the output

```{r, fig.width=10, fig.height=3}
library(stringr)
require(ggplot2)

testing <- read.table("Chr8_output.txt")


testing %>% separate(col = V1, into = c("CHROM","POS", 
                                        "Intercept_C", "Bulk_C", "Parent_C", "Interaction_C", 
                                        "Intercept_SE", "Bulk_SE", "Parent_SE", "Interaction_SE", 
                                        "Intercept_Z", "Bulk_Z", "Parent_Z", "Interaction_Z",
                                        "c","d","e","f","g","h","i","j","k", "l", "m", "n", "o"), sep = ",") -> testing2

unique(testing2$CHROM)

#Process all of the extra data
accidentalmerges <- testing2[,14:27] %>% na.omit()
dim(accidentalmerges)

colnames(accidentalmerges) <- c("priorcol", "POS", 
                                        "Intercept_C", "Bulk_C", "Parent_C", "Interaction_C", 
                                        "Intercept_SE", "Bulk_SE", "Parent_SE", "Interaction_SE", 
                                        "Intercept_Z", "Bulk_Z", "Parent_Z", "Interaction_Z")
accidentalmerges$CHROM <- str_extract(accidentalmerges$priorcol, "\\d$")

#check what this looks like
unique(accidentalmerges$CHROM)
plot(accidentalmerges$CHROM)

#Add 10 to all of the values after 1000
accidentalmerges %>% mutate_at(vars(one_of("CHROM")), as.numeric) -> accidentalmerges
plot(accidentalmerges$CHROM)

accidentalmerges %>% 
  #Add 10 to those between 500-1500 less than 4
  filter(row_number() > 500 & row_number() < 1500) %>% filter(CHROM < 4) %>% 
  mutate(CHROM = CHROM +10) %>% rbind(accidentalmerges) %>% 
  #get rid of anything that was within those bounds and less than 4
  filter(CHROM < 4 & (row_number() > 500 & row_number() < 1500) == FALSE) %>%
  
  #Add 10 to those between 1500 and 2500 that are less than 6
  filter(row_number() > 1500 & row_number() < 2500) %>% filter(CHROM < 6) %>% 
  mutate(CHROM = CHROM +10) %>% rbind(accidentalmerges) %>% 
  filter(CHROM < 6 & (row_number() > 1500 & row_number() < 2500) == FALSE) %>%
  
  #Final one: greater than 2000 and greater than 5
  filter(row_number() > 2000 ) %>% filter(CHROM > 5) %>% 
  mutate(CHROM = CHROM +10) %>% rbind(accidentalmerges) %>% 
  filter(CHROM > 5 & (row_number() > 2000 & row_number() < 2500) == FALSE) -> amerges


testing3 <- rbind(testing2[,1:14], amerges[,c(15, 2:14)])

#Pivot to plot easier
testing3 %>% pivot_longer(
    cols = -c(CHROM, POS),
    names_to = c("Effect", "Factor"),
    names_pattern = "(.*)_(.*)",
    values_to = "Value"
  ) %>% mutate_at(vars(-one_of("Effect", "Factor")), as.numeric) -> thxgpt


#plot
thxgpt %>% group_by(CHROM) %>% 
  filter(CHROM != 17, CHROM != 8) %>% filter(Factor != "Z", Effect != "Intercept") %>% 
  #Separate C and SE
  pivot_wider(names_from = Factor, values_from = Value) %>%
 
  ggplot(aes(x = POS, y = abs(C))) + 
  #Plot the lines
  #Plot the standard errors
  geom_ribbon(aes(x = POS, ymin = abs(C) - 2*SE, ymax = abs(C)+2*SE, fill = Effect), alpha = 0.2, size = 0) +
  geom_line(aes(color = Effect)) + 
  geom_hline(aes(yintercept = 0)) +
  
  #colors
  scale_color_manual(values = c("black", "orchid", "gray"))+
    scale_fill_manual(values = c("black", "orchid", "gray"))+

  #Separate chromosomes
  facet_grid(~CHROM, scales = "free", space = "free") + 
  #Add in theme elements and title
  theme_minimal() +
  theme(legend.position = "bottom", axis.text.x=element_blank(), axis.ticks.x=element_blank()) + ggtitle("CuSO4 CSS8 | glmer") 



```

## Doing this for Chr I CuSO4 combined

```{r}
#Putting together the files
exp2 <- readRDS("rollData_Chr1b_CuSO4.rds")
exp1 <- readRDS("rollData_Chr1a_CuSO4.rds")

exp1 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Allele"), names_sep = "_") %>% mutate(Day = 1) -> E1

exp2 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Day", "Allele"), names_sep = "_") %>% mutate(Day = recode(Day, "A" = 2, "B" = 3)) -> E2

rbind(E1, E2) %>% pivot_wider(names_from = c(Bulk, Parent, Day, Allele), values_from = value) %>% na.omit() -> CuSO4_all_smoothed

saveRDS(CuSO4_all_smoothed, file = "CuSO4_all_smoothed.rds")

dim(CuSO4_all_smoothed)
```

```{r}
#!/usr/bin/env Rscript
#args = commandArgs(trailingOnly=TRUE)

options(warn=-1)

#Load in all packages
require(tidyr)
require(reshape2)
require(data.table)
require(dplyr)
require(foreach)
require(doParallel)
require(scales)
require(lme4)

rollData <- readRDS("CuSO4_all_smoothed.rds")

#Find the position necessary
rollData[args[1],] %>% pivot_longer(c(-CHROM, -POS), 
                                    names_to = c("Bulk", "Parent", "Day", "Allele"), 
                                    names_sep = "_", 
                                    values_to = "Reads") %>%
  mutate_at(vars(-one_of("CHROM", "POS", "Reads")), as.factor) -> mydata

glmeroutput <-  glmer(Allele ~ Bulk*Parent+(1|Day),
             weights = Reads, family = binomial,
              data = mydata)

sumd <- summary(glmeroutput)

#sumd$coefficients[1:(.75*length(sumd$coefficients))]

output <- c(unique(mydata$CHROM),
            unique(mydata$POS),
            sumd$coefficients[1:(.75*length(sumd$coefficients))],
            0)


write(paste0(output, collapse = ","), stdout())
```

Analyzing this - it looks to be inaccurate, which I can't figure out :(

```{r, fig.width=10, fig.height=3}
require(stringr)
require(ggplot2)
require(dplyr)
require(tidyr)

Chr1_CuSO4 <- read.table("Chr1_output.txt")


Chr1_CuSO4 %>% separate(col = V1, into = c("CHROM","POS", 
                                        "Intercept_C", "Bulk_C", "Parent_C", "Interaction_C", 
                                        "Intercept_SE", "Bulk_SE", "Parent_SE", "Interaction_SE", 
                                        "Intercept_Z", "Bulk_Z", "Parent_Z", "Interaction_Z",
                                        "c","d","e","f","g","h","i","j","k", "l", "m", "n", "o", "p"), sep = ",") -> Chr1_CuSO4

unique(Chr1_CuSO4$CHROM)

#Process all of the extra data
accidentalmerges <- Chr1_CuSO4[,15:28] %>% na.omit() %>% mutate_at(vars("c"), as.numeric)
dim(accidentalmerges)

colnames(accidentalmerges) <- c("CHROM", "POS", 
                                        "Intercept_C", "Bulk_C", "Parent_C", "Interaction_C", 
                                        "Intercept_SE", "Bulk_SE", "Parent_SE", "Interaction_SE", 
                                        "Intercept_Z", "Bulk_Z", "Parent_Z", "Interaction_Z")

#check what this looks like
unique(accidentalmerges$CHROM)
plot(accidentalmerges$CHROM)

#Add 10 to all of the values after 1000
accidentalmerges %>% mutate_at(vars(one_of("CHROM")), as.numeric) -> accidentalmerges

CuSO4_merged <- rbind(testing2[,1:14], accidentalmerges[]) %>% distinct()
unique(CuSO4_merged$CHROM)

CuSO4_merged %>% group_by(CHROM, POS) %>% summarize(count = n()) %>% filter(count > 1) -> duplicatedPOS

#Pivot to plot easier
CuSO4_merged %>% distinct(CHROM, POS, .keep_all = TRUE) %>% pivot_longer(
    cols = -c(CHROM, POS),
    names_to = c("Effect", "Factor"),
    names_pattern = "(.*)_(.*)",
    values_to = "Value"
  ) %>% mutate_at(vars(-one_of("Effect", "Factor")), as.numeric) %>% distinct() -> CuSO4_merged_pivot


#plot
CuSO4_merged_pivot %>%
  filter(CHROM != 17) %>% filter(Factor != "Z", Effect != "Intercept") %>% 
  #Separate C and SE
  pivot_wider(names_from = Factor, values_from = Value) %>%
 
  ggplot(aes(x = POS, y = abs(C))) + 
  #Plot the lines
  #Plot the standard errors
  geom_ribbon(aes(x = POS, ymin = abs(C) - 2*SE, ymax = abs(C)+2*SE, fill = Effect), alpha = 0.2, size = 0) +
  geom_line(aes(color = Effect)) + 
  geom_hline(aes(yintercept = 0)) +
  
  #colors
  scale_color_manual(values = c("black", "orchid", "gray"))+
    scale_fill_manual(values = c("black", "orchid", "gray"))+

  #Separate chromosomes
  facet_grid(~CHROM, scales = "free", space = "free") + 
  #Add in theme elements and title
  theme_minimal() + ylim(0, 2) +
  theme(legend.position = "bottom", axis.text.x=element_blank(), axis.ticks.x=element_blank()) + ggtitle("CuSO4 CSS1 All | glmer") 


```