---
title: "All Traits Process Jan 2025 with Effect Sizes"
author: "Cass"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))
```

IGNORE THIS
```{r, eval = FALSE}
# Number of chromosomes      : 17
# Chromosomes                : Format 'chromo_name size codon_table'
#               'IV'    1531933 Standard
#               'XV'    1091291 Standard
#               'VII'   1090940 Standard
#               'XII'   1078177 Standard
#               'XVI'   948066  Standard
#               'XIII'  924431  Standard
#               'II'    813184  Standard
#               'XIV'   784333  Standard
#               'X'     745751  Standard
#               'XI'    666816  Standard
#               'V'     576874  Standard
#               'VIII'  562643  Standard
#               'IX'    439888  Standard
#               'III'   316620  Standard
#               'VI'    270161  Standard
#               'I'     230218  Standard
#               'Mito'  85779   Standard

NC_001133.9 I
NC_001134.8 II
NC_001135.5 III
NC_001136.10 IV
NC_001137.3 V
NC_001138.5 VI
NC_001139.9 VII
NC_001140.6 VIII
NC_001141.2 IX
NC_001142.9 X
NC_001143.9 XI
NC_001144.5 XII
NC_001145.3 XIII
NC_001146.8 XIV
NC_001147.6 XV
NC_001148.4 XVI
NC_001224.1 Mito

```

# Output table to called .rds

```{r}
#Do the H2O2 and Zeocin processing here
myfile <- "C:/Users/Cassandra/Documents/Data_NYU/Sequencing/H2O2.SortedCat.vcf.output.table"

#GET RID OF EXCLUDED ONES
read.table(myfile, header = TRUE) -> testrawdata
testrawdata %>% select(CHROM, POS, ALT) %>% separate(ALT, into = c("First", "Second"), sep = ",") %>% na.omit() -> exclude

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

exclude %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> exclude

exclude %>% mutate(locus = paste(CHROM, POS)) -> exclKEY

cybrInputGATKTable2(myfile) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> rawdata

rawdata %>% merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% mutate(locus = paste(CHROM, POS)) %>% filter(locus %in% exclKEY$locus == FALSE) %>% select(-locus) -> rawdata_called_ex

saveRDS(rawdata_called_ex, file = "H2O2_called_ex.rds")

rm(testrawdata)
rm(rawdata)
rm(exclKEY)
rm(rawdata_called)
rm(rawdata_called_ex)
rm(exclude)

################################################################################

myfile <- "C:/Users/Cassandra/Documents/Data_NYU/Sequencing/Zeocin.SortedCat.vcf.output.table"

#GET RID OF EXCLUDED ONES
read.table(myfile, header = TRUE) -> testrawdata
testrawdata %>% select(CHROM, POS, ALT) %>% separate(ALT, into = c("First", "Second"), sep = ",") %>% na.omit() -> exclude

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

exclude %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> exclude

exclude %>% mutate(locus = paste(CHROM, POS)) -> exclKEY

cybrInputGATKTable2(myfile) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> rawdata

rawdata %>% merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% mutate(locus = paste(CHROM, POS)) %>% filter(locus %in% exclKEY$locus == FALSE) %>% select(-locus) -> rawdata_called_ex

saveRDS(rawdata_called_ex, file = "Zeocin_called_ex.rds")

```

# Reorganize by Selection Type

## CSS 15 + Output table to .rds

```{r}
MQC <- read.csv("C:/Users/Cassandra/Documents/Data_NYU/Sequencing/MQC_Annotated.csv")
MQC %>% select(Pool, PoolID, Dataset = Library, Parent, Bulk, Rep) -> MQC_Key

#Do the CSS15 processing here
myfile <- "C:/Users/Cassandra/Documents/Data_NYU/Sequencing/####H2O2.SortedCat.vcf.output.table"

#GET RID OF EXCLUDED ONES
read.table(myfile, header = TRUE) -> testrawdata
testrawdata %>% select(CHROM, POS, ALT) %>% separate(ALT, into = c("First", "Second"), sep = ",") %>% na.omit() -> exclude

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

exclude %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> exclude

exclude %>% mutate(locus = paste(CHROM, POS)) -> exclKEY

cybrInputGATKTable2(myfile) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> rawdata

rawdata %>% merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% mutate(locus = paste(CHROM, POS)) %>% filter(locus %in% exclKEY$locus == FALSE) %>% select(-locus) -> rawdata_called_ex

saveRDS(rawdata_called_ex, file = "CSS15_called_ex.rds")

rawdata_called_ex %>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% merge(MQC_Key) -> CSS15

saveRDS(CSS15, file = "CSS15_2025.mqc.rds")

rm(testrawdata)
rm(rawdata)
rm(exclKEY)
rm(rawdata_called)
rm(rawdata_called_ex)
rm(exclude)

```

## All others
```{r, eval = FALSE}

H2LK2DMX2_rawdata_called_ex <- readRDS("CSS8_PQ_rawdata_called_ex.rds")
H2LKCDMX2_rawdata_called_ex <- readRDS("CSS1_PQ_rawdata_called_ex_J25.rds")

HKTFTDRX2_rawdata_called_ex <- readRDS("HKTFTDRX2_rawdata_called_ex.rds")
HWMMFDMXY_rawdata_called_ex <- readRDS("HWMMFDMXY_rawdata_called_ex.rds")
HWTCNDMXY_rawdata_called_ex <- readRDS("HWTCNDMXY_rawdata_called_ex.rds") 
HGVMVDRX2_rawdata_called_ex <- readRDS("HGVMVDRX2_rawdata_called_ex.rds")
HJ5HKDRX3b_rawdata_called_ex <- readRDS("HJ5HKDRX3b_rawdata_called_ex_J25.rds") #ran this with the new output.table labeled
HKTMZDRX2_rawdata_called_ex <- readRDS("HKTMZDRX2_rawdata_called_ex.rds")
HVYTYDRX2_rawdata_called_ex <- readRDS("HVYTYDRX2_rawdata_called_ex.rds")

CuSO4_2024_rawdata_called_ex <- readRDS("CuSO4_2024_rawdata_called_ex.rds")

H2O2_rawdata_called_ex <- readRDS("H2O2_called_ex.rds")
Zeocin_rawdata_called_ex <- readRDS("Zeocin_called_ex.rds")

```

```{r, eval = FALSE}
MQC <- read.csv("C:/Users/Cassandra/Documents/Data_NYU/Sequencing/MQC_Annotated.csv")
MQC %>% select(Pool, PoolID, Dataset = Library, Parent, Bulk, Rep) -> MQC_Key
```

```{r, eval = FALSE}
H2LKCDMX2_rawdata_called_ex %>% mutate(Dataset = gsub("H2LKCDMX2_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  merge(MQC_Key) -> H2LKCDMX2

H2LK2DMX2_rawdata_called_ex %>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% 
  mutate(Dataset = gsub("H2LK2DMX2_n01_", "", Dataset)) %>% merge(MQC_Key) -> H2LK2DMX2

HKTFTDRX2_rawdata_called_ex %>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% merge(MQC_Key) -> HKTFTDRX2
#HWMMFDMXY_rawdata_called_ex %>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% merge(MQC_Key) -> HWMMFDMXY
#HWTCNDMXY_rawdata_called_ex%>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% merge(MQC_Key) -> HWTCNDMXY
HGVMVDRX2_rawdata_called_ex%>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% merge(MQC_Key) -> HGVMVDRX2
HJ5HKDRX3b_rawdata_called_ex %>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% merge(MQC_Key) -> HJ5HKDRX3b
HKTMZDRX2_rawdata_called_ex %>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% merge(MQC_Key) -> HKTMZDRX2
HVYTYDRX2_rawdata_called_ex %>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% merge(MQC_Key) -> HVYTYDRX2

HJ5HKDRX3b %>% filter(Rep == "B", grepl("8", Parent), Bulk != "H2O2") -> Z_HJ5HKDRX3b
HJ5HKDRX3b %>% filter(Bulk != "Zeocin") -> H_HJ5HKDRX3b

################################################################################
CuSO4_2024_rawdata_called_ex %>% mutate(Dataset = gsub(".fastq", "", Dataset)) %>% merge(MQC_Key) -> CuSO4_2024



################################################################################
saveRDS(HKTFTDRX2, file = "HKTFTDRX2.mqc.rds")
saveRDS(HWMMFDMXY, file = "HWMMFDMXY.mqc.rds")
saveRDS(HWTCNDMXY, file = "HWTCNDMXY.mqc.rds")
saveRDS(HGVMVDRX2, file = "HGVMVDRX2.mqc.rds")
saveRDS(HJ5HKDRX3b, file = "HJ5HKDRX3b.mqc.rds")
saveRDS(HKTMZDRX2, file = "HKTMZDRX2.mqc.rds")
saveRDS(HVYTYDRX2, file = "HVYTYDRX2.mqc.rds")
saveRDS(H2LKCDMX2, file = "H2LKCDMX2.mqc.rds")
saveRDS(H2LK2DMX2, file = "H2LK2DMX2.mqc.rds")

saveRDS(CuSO4_2024, file = "CuSO4_2024.mqc.rds")

```

Full sets on their own

```{r, eval = FALSE}
################################################################################
MQC_Key %>% filter(PoolID %in% c(11,12)) %>% mutate(Dataset = paste(Pool, "n01", Dataset, sep = "_")) -> MQC_Key_edit

MQC_Key %>% filter(PoolID %in% c(11,12) == FALSE) %>% rbind(MQC_Key_edit) %>%
  mutate(Dataset = paste(Dataset, ".fastq", sep = "")) -> MQC_Key_consistent

H2O2_rawdata_called_ex %>% merge(MQC_Key_consistent) -> H2O2_unsmoothed
Zeocin_rawdata_called_ex %>% merge(MQC_Key_consistent) -> Zeocin_unsmoothed

saveRDS(H2O2_unsmoothed, file = "H2O2.mqc.rds")
saveRDS(Zeocin_unsmoothed, file = "Zeocin.mqc.rds")
```

Remove unneeded files to clear space

```{r, eval = FALSE}
rm(HKTFTDRX2_rawdata_called_ex)
rm(HWMMFDMXY_rawdata_called_ex)
rm(HWTCNDMXY_rawdata_called_ex)
rm(HGVMVDRX2_rawdata_called_ex)
rm(HJ5HKDRX3b_rawdata_called_ex)
rm(HKTMZDRX2_rawdata_called_ex)
rm(HVYTYDRX2_rawdata_called_ex)
rm(CuSO4_2024_rawdata_called_ex)
```

Load in if needed

```{r, eval = FALSE}
HKTFTDRX2 <- readRDS("HKTFTDRX2.mqc.rds")
HWMMFDMXY <- readRDS("HWMMFDMXY.mqc.rds")
HWTCNDMXY <- readRDS("HWTCNDMXY.mqc.rds")
HGVMVDRX2 <- readRDS("HGVMVDRX2.mqc.rds")
HJ5HKDRX3b <- readRDS("HJ5HKDRX3b.mqc.rds")
HKTMZDRX2 <- readRDS("HKTMZDRX2.mqc.rds")
HVYTYDRX2 <- readRDS("HVYTYDRX2.mqc.rds")
H2LKCDMX2 <- readRDS("H2LKCDMX2.mqc.rds")
H2LK2DMX2 <- readRDS("H2LK2DMX2.mqc.rds")

CuSO4_2024 <- readRDS("CuSO4_2024.mqc.rds")
H2O2_unsmoothed <- readRDS("H2O2.mqc.rds")
Zeocin_unsmoothed <- readRDS("Zeocin.mqc.rds")
```

Remove if needed
```{r}
rm(HKTFTDRX2)
rm(HWMMFDMXY)
rm(HWTCNDMXY)
rm(HGVMVDRX2)
rm(HJ5HKDRX3b)
rm(HKTMZDRX2)
rm(HVYTYDRX2)
rm(H2LKCDMX2)
rm(H2LK2DMX2)
rm(CuSO4_2024)
```

```{r, eval = FALSE}
CuSO4_2024 %>% filter(Bulk != "Fluconazole") -> Copper_Unsmoothed

rbind(HWTCNDMXY, HWMMFDMXY) %>% filter(Bulk %in% c("Dilute", "Cycloheximide")) %>%
  group_by(CHROM, POS, Allele, Pool, Parent, Bulk)  -> Cycloheximide_Unsmoothed

rbind(HWTCNDMXY, HWMMFDMXY, H2LK2DMX2) %>% filter(Bulk %in% c("Dilute", "Ethanol", "EtOH")) %>%
  mutate(Bulk = gsub("EtOH", "Ethanol", Bulk)) %>%
  group_by(CHROM, POS, Allele, Pool, Parent, Bulk) -> Ethanol_Unsmoothed

HVYTYDRX2 %>% filter(Parent %in% c("Oak8", "Wine8"),
                     Bulk != "CuSO4") %>%
  rbind(HGVMVDRX2) -> Fluconazole_Unsmoothed


rbind(H2LK2DMX2,H2LKCDMX2) %>% filter(Bulk %in% c("Dilute", "PQ")) %>%
  group_by(CHROM, POS, Allele, Pool, Parent, Bulk) -> Paraquat_Unsmoothed


```


Make sure each only has ONE from each allele etc

```{r, eval = FALSE}
#Put these in alphabetical order so I don't go insane
Copper_Unsmoothed %>% distinct() -> Copper_Unsmoothed
Cycloheximide_Unsmoothed %>% distinct() -> Cycloheximide_Unsmoothed
Ethanol_Unsmoothed %>% distinct() -> Ethanol_Unsmoothed
Fluconazole_Unsmoothed %>% distinct() -> Fluconazole_Unsmoothed #(not actually done yet)
H2O2_unsmoothed %>% distinct() -> H2O2_Unsmoothed
Paraquat_Unsmoothed %>% na.omit() %>% distinct() -> Paraquat_Unsmoothed
Zeocin_unsmoothed %>% distinct() -> Zeocin_Unsmoothed
```

# Filter out positions with low coverage compared to average

## CSS 15

```{r}
CSS15 %>% ungroup() %>% group_by(Dataset) %>% summarize(AvgCovg = mean(Coverage, na.rm = TRUE)) -> Means

CSS15 %>% left_join(Means) %>% ungroup() %>% na.omit() %>%
  group_by(Dataset, CHROM, POS) %>% 
  select(Dataset, CHROM, POS, Coverage, AvgCovg) %>%
  distinct() %>%
  mutate(CovChange = log(Coverage/AvgCovg)) %>%
  filter(CovChange > -1) -> CSS15_Unsmoothed_cf

rm(Means)

```

## All other traits

```{r, eval = FALSE}
#COPPER
Copper_Unsmoothed %>% ungroup() %>% group_by(Dataset) %>% summarize(AvgCovg = mean(Coverage, na.rm = TRUE)) -> Means

Copper_Unsmoothed %>% left_join(Means) %>% ungroup() %>% na.omit() %>%
  group_by(Dataset, CHROM, POS) %>% 
  select(Dataset, CHROM, POS, Coverage, AvgCovg) %>%
  distinct() %>%
  mutate(CovChange = log(Coverage/AvgCovg)) %>%
  filter(CovChange > -1) -> Copper_Unsmoothed_cf

rm(Means)

################################################################################
#CYCLOHEXIMIDE
Cycloheximide_Unsmoothed %>% ungroup() %>% group_by(Dataset) %>% summarize(AvgCovg = mean(Coverage, na.rm = TRUE)) -> Means

Cycloheximide_Unsmoothed %>% left_join(Means) %>% ungroup() %>% na.omit() %>%
  group_by(Dataset, CHROM, POS) %>% 
  select(Dataset, CHROM, POS, Coverage, AvgCovg) %>%
  distinct() %>%
  mutate(CovChange = log(Coverage/AvgCovg)) %>%
  filter(CovChange > -1) -> Cycloheximide_Unsmoothed_cf

rm(Means)
################################################################################
#ETHANOL
Ethanol_Unsmoothed %>% ungroup() %>% group_by(Dataset) %>% summarize(AvgCovg = mean(Coverage, na.rm = TRUE)) -> Means

Ethanol_Unsmoothed %>% left_join(Means) %>% ungroup() %>% na.omit() %>%
  group_by(Dataset, CHROM, POS) %>% 
  select(Dataset, CHROM, POS, Coverage, AvgCovg) %>%
  distinct() %>%
  mutate(CovChange = log(Coverage/AvgCovg)) %>%
  filter(CovChange > -1) -> Ethanol_Unsmoothed_cf

rm(Means)
################################################################################
#FLUCONAZOLE
Fluconazole_Unsmoothed %>% ungroup() %>% group_by(Dataset) %>% summarize(AvgCovg = mean(Coverage, na.rm = TRUE)) -> Means

Fluconazole_Unsmoothed %>% left_join(Means) %>% ungroup() %>% na.omit() %>%
  group_by(Dataset, CHROM, POS) %>% 
  select(Dataset, CHROM, POS, Coverage, AvgCovg) %>%
  distinct() %>%
  mutate(CovChange = log(Coverage/AvgCovg)) %>%
  filter(CovChange > -1) -> Fluconazole_Unsmoothed_cf

rm(Means)
################################################################################
#H2O2
H2O2_Unsmoothed %>% ungroup() %>% group_by(Dataset) %>% summarize(AvgCovg = mean(Coverage, na.rm = TRUE)) -> Means

H2O2_Unsmoothed %>% left_join(Means) %>% ungroup() %>% na.omit() %>%
  group_by(Dataset, CHROM, POS) %>% 
  select(Dataset, CHROM, POS, Coverage, AvgCovg) %>%
  distinct() %>%
  mutate(CovChange = log(Coverage/AvgCovg)) %>%
  filter(CovChange > -1) -> H2O2_Unsmoothed_cf

rm(Means)

################################################################################
#PARAQUAT
Paraquat_Unsmoothed %>% ungroup() %>% group_by(Dataset) %>% summarize(AvgCovg = mean(Coverage, na.rm = TRUE)) -> Means

Paraquat_Unsmoothed %>% left_join(Means) %>% ungroup() %>% na.omit() %>%
  group_by(Dataset, CHROM, POS) %>% 
  select(Dataset, CHROM, POS, Coverage, AvgCovg) %>%
  distinct() %>%
  mutate(CovChange = log(Coverage/AvgCovg)) %>%
  filter(CovChange > -1) -> Paraquat_Unsmoothed_cf

rm(Means)

################################################################################
#ZEOCIN
Zeocin_Unsmoothed %>% ungroup() %>% group_by(Dataset) %>% summarize(AvgCovg = mean(Coverage, na.rm = TRUE)) -> Means

Zeocin_Unsmoothed %>% left_join(Means) %>% ungroup() %>% na.omit() %>%
  group_by(Dataset, CHROM, POS) %>% 
  select(Dataset, CHROM, POS, Coverage, AvgCovg) %>%
  distinct() %>%
  mutate(CovChange = log(Coverage/AvgCovg)) %>%
  filter(CovChange > -1) -> Zeocin_Unsmoothed_cf

rm(Means)


```

# Smoothing All

## CSS 15

```{r}
CSS15_Unsmoothed_cf %>% ungroup() %>%
  group_by(CHROM, Pool, Parent, Bulk, Allele, Dataset) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) -> CSS15_G200

saveRDS(CSS15_G200, file = "covfilt/CSS15_G200_J25.rds")
```

```{r, eval = FALSE}
Cycloheximide_Unsmoothed %>% ungroup() %>%
  group_by(CHROM, Pool, Parent, Bulk, Allele, Dataset) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) -> Cycloheximide_G200
saveRDS(Cycloheximide_G200, file = "covfilt/Cycloheximide_G200_J25.rds")

Copper_Unsmoothed %>% ungroup() %>%
  group_by(CHROM, Pool, Parent, Bulk, Allele, Dataset) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) -> Copper_G200
saveRDS(Copper_G200, file = "covfilt/Copper_G200_J25.rds")

Fluconazole_Unsmoothed %>% ungroup() %>%
  group_by(CHROM, Pool, Parent, Bulk, Allele, Dataset) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) -> Fluconazole_G200
saveRDS(Fluconazole_G200, file = "covfilt/Fluconazole_G200_J25.rds")

Ethanol_Unsmoothed %>% ungroup() %>% na.omit() %>% 
  group_by(CHROM, Pool, Parent, Bulk, Allele, Dataset) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) -> Ethanol_G200
saveRDS(Ethanol_G200, file = "covfilt/Ethanol_G200_covfilt_xNAs.rds")

#xNAs indicates that NAs were omitted from the unsmoothed data first (since those tended to have tons of copies per position?)

Paraquat_Unsmoothed %>% ungroup() %>% 
  group_by(CHROM, Pool, Parent, Bulk, Allele, Dataset) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) -> Paraquat_G200
saveRDS(Paraquat_G200, file = "covfilt/Paraquat_G200_covfilt_xNAs.rds")

## REDOING - concatenated vcf, so called correctly

Zeocin_unsmoothed %>% ungroup() %>% 
  group_by(CHROM, Pool, Parent, Bulk, Allele, Dataset) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) -> Zeocin_G200
saveRDS(Zeocin_G200, file = "covfilt/Zeocin_G200d_covfiltc.rds")

H2O2_unsmoothed %>% ungroup() %>% 
  group_by(CHROM, Pool, Parent, Bulk, Allele, Dataset) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) -> H2O2_G200
saveRDS(H2O2_G200, file = "covfilt/H2O2_G200d_covfiltc.rds")

```

##################################################
# Everything below this is new for exporting all coefficients
##################################################

```{r}
H2O2_G200 <- readRDS("covfilt/H2O2_G200d_covfiltc.rds")
Zeocin_G200 <- readRDS("covfilt/Zeocin_G200d_covfiltc.rds")
Paraquat_G200 <- readRDS("covfilt/Paraquat_G200_covfilt_xNAs.rds")
Ethanol_G200 <- readRDS("covfilt/Ethanol_G200_covfilt_xNAs.rds")
Fluconazole_G200 <- readRDS("covfilt/Fluconazole_G200_J25.rds")
Cycloheximide_G200 <- readRDS("covfilt/Cycloheximide_G200_J25.rds")
Copper_G200 <- readRDS("covfilt/Copper_G200_J25.rds")

#CSS15_G200 <- readRDS("covfilt/CSS15_G200_J25.rds")
```

# GLMs

Write new function for glms with everything output

```{r}
glm_cb2_all <- function (..., W, formula, numgroups = FALSE, outputlength = 4) 
{
    data <- list(...)
    
    if (is.null(W) || is.null(formula)) {
        stop("Weights (W) and formula must be provided")
    }
    glm_formula <- as.formula(formula)
    if (!all(names(data) %in% all.vars(glm_formula))) {
        stop("One or more variables in the formula are not provided as arguments")
    }
    for (i in all.vars(glm_formula)) {
        if (length(unique(as.data.frame(data)[, i])) < 2) {
            output <- rep(NA, outputlength*4)
            return(output)
        }
    }
    glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, 
        family = binomial)
    
    output <- c(summary(glm_fit)$coefficients)
    
    if (length(output) == outputlength*4) {
        return(output)
    }
    else {
        return(rep(NA, outputlength*4))
    }
}
```

DOUBLE CHECK that each run has the pools that it should (ie cold stress excluded)

```{r}
#Confirmed all without CSS15, which will be run on its own later
unique(Copper_G200$Pool)
unique(Cycloheximide_G200$Pool)
unique(Fluconazole_G200$Pool)
unique(Ethanol_G200$Pool)
unique(H2O2_G200$Pool)
unique(Paraquat_G200$Pool)
unique(Zeocin_G200$Pool)
```

### Chromosome XV Fixed (all)

```{r, eval = FALSE}
CSS15_G200 %>% na.omit() %>% 
  filter(grepl(1, Parent)) %>%
  mutate_if(is.character, as.factor) -> PQ_CSS15
```

```{r, eval = FALSE}
CSS15_G200 %>% na.omit() %>% 
  filter(grepl(1, Parent)) %>%
  mutate_if(is.character, as.factor) -> PQ_CSS15

#Set contrasts
contrasts(Copper_CSS1_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Copper_CSS1_factor$Bulk) <- matrix(c(1, 0))

#Run GLM
Copper_CSS1_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p")) %>%
  mutate(Selection = "Copper", CSS = "Fixed I") -> Copper_CSS1_glm

#Save
saveRDS(Copper_CSS1_glm, file = "allstats/Copper_CSS1_glm_J25.rds")

################################################################################
Copper_G200 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Copper_CSS8_factor

#Set contrasts
contrasts(Copper_CSS8_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Copper_CSS8_factor$Bulk) <- matrix(c(1, 0))

#Run GLM
Copper_CSS8_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Copper", CSS = "Fixed VIII")  -> Copper_CSS8_glm

#Save
saveRDS(Copper_CSS8_glm, file = "allstats/Copper_CSS8_glm_J25.rds")
```

Running glm rather than glmer so that the individual bulks count as noise but nothing else contributes

### Copper

```{r, eval = TRUE}
Copper_G200 %>% na.omit() %>% 
  filter(grepl(1, Parent)) %>%
  mutate_if(is.character, as.factor) -> Copper_CSS1_factor

#Set contrasts
contrasts(Copper_CSS1_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Copper_CSS1_factor$Bulk) <- matrix(c(1, 0))

#Run GLM
Copper_CSS1_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p")) %>%
  mutate(Selection = "Copper", CSS = "Fixed I") -> Copper_CSS1_glm

#Save
saveRDS(Copper_CSS1_glm, file = "allstats/Copper_CSS1_glm_J25.rds")

################################################################################
Copper_G200 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Copper_CSS8_factor

#Set contrasts
contrasts(Copper_CSS8_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Copper_CSS8_factor$Bulk) <- matrix(c(1, 0))

#Run GLM
Copper_CSS8_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Copper", CSS = "Fixed VIII")  -> Copper_CSS8_glm

#Save
saveRDS(Copper_CSS8_glm, file = "allstats/Copper_CSS8_glm_J25.rds")
```


### Cycloheximide

```{r, eval = TRUE}
Cycloheximide_G200 %>% na.omit() %>% 
  filter(grepl(1, Parent)) %>%
  mutate_if(is.character, as.factor) -> Cyc_CSS1_factor

#Set contrasts
contrasts(Cyc_CSS1_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Cyc_CSS1_factor$Bulk) <- matrix(c(1, 0))

#Run GLM
Cyc_CSS1_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Cycloheximide", CSS = "Fixed I")  -> Cyc_CSS1_glm

#Save
saveRDS(Cyc_CSS1_glm, file = "allstats/Cyc_CSS1_glm_J25.rds")

################################################################################
Cycloheximide_G200 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Cyc_CSS8_factor

#Set contrasts
contrasts(Cyc_CSS8_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Cyc_CSS8_factor$Bulk) <- matrix(c(1, 0))

#Run GLM
Cyc_CSS8_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Cycloheximide", CSS = "Fixed VIII")  -> Cyc_CSS8_glm

#Save
saveRDS(Cyc_CSS8_glm, file = "allstats/Cyc_CSS8_glm_J25.rds")
```

### Ethanol

```{r}
Ethanol_G200 %>% na.omit() %>% 
  filter(grepl(1, Parent)) %>%
  mutate_if(is.character, as.factor) -> Ethanol_CSS1_factor

#Set contrasts
contrasts(Ethanol_CSS1_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Ethanol_CSS1_factor$Bulk) <- matrix(c(0,1))

#Run GLM
Ethanol_CSS1_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Ethanol", CSS = "Fixed I")  -> Ethanol_CSS1_glm

#Save
saveRDS(Ethanol_CSS1_glm, file = "allstats/Ethanol_CSS1_glm_J25.rds")

################################################################################
Ethanol_G200 %>% na.omit() %>% 
  filter(Pool != "HWTCNDMXY") %>%
  filter(grepl(8, Parent)) %>%
  mutate(Bulk = gsub("Ethanol", "EtOH", Bulk)) %>% #needed to add this for consistency
  mutate_if(is.character, as.factor) -> Ethanol_CSS8_factor

#Set contrasts
contrasts(Ethanol_CSS8_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Ethanol_CSS8_factor$Bulk) <- matrix(c(0,1))

#Run GLM
Ethanol_CSS8_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Ethanol", CSS = "Fixed VIII")  -> Ethanol_CSS8_glm

#Save
saveRDS(Ethanol_CSS8_glm, file = "allstats/Ethanol_CSS8_glm_allstats_H2LK2only.rds")

```

### Fluconazole 

```{r, eval = TRUE}
Fluconazole_G200 %>% na.omit() %>% 
  filter(grepl(1, Parent)) %>%
  mutate_if(is.character, as.factor) -> Fluconazole_CSS1_factor

#Set contrasts
contrasts(Fluconazole_CSS1_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Fluconazole_CSS1_factor$Bulk) <- matrix(c(0,1))

#Run GLM
Fluconazole_CSS1_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Fluconazole", CSS = "Fixed I")  -> Fluconazole_CSS1_glm

#Save
saveRDS(Fluconazole_CSS1_glm, file = "allstats/Fluconazole_CSS1_glm_J25.rds")

################################################################################
Fluconazole_G200 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Fluconazole_CSS8_factor

#Set contrasts
contrasts(Fluconazole_CSS8_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Fluconazole_CSS8_factor$Bulk) <- matrix(c(0,1))

#Run GLM
Fluconazole_CSS8_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Fluconazole", CSS = "Fixed VIII")  -> Fluconazole_CSS8_glm

#Save
saveRDS(Fluconazole_CSS8_glm, file = "allstats/Fluconazole_CSS8_glm_J25.rds")

```


### Hydrogen Peroxide 

```{r}
rm(H2O2_CSS1_glm)
rm(H2O2_CSS8_glm)

H2O2_G200 %>% na.omit() %>% 
  filter(Dataset != "H2LKCDMX2_n01_Wine1_H_Ar.fastq", Dataset != "H2LKCDMX2_n01_Wine1_H_Br.fastq",
         Bulk != "Freeze") %>% 
  filter(grepl(1, Parent)) %>%
  mutate_if(is.character, as.factor) -> H2O2_CSS1_factor

#Set contrasts
contrasts(H2O2_CSS1_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(H2O2_CSS1_factor$Bulk) <- matrix(c(0,1))

#Run GLM
H2O2_CSS1_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "H2O2", CSS = "Fixed I")  -> H2O2_CSS1_glm

#Save
saveRDS(H2O2_CSS1_glm, file = "allstats/H2O2_CSS1_glm_allstats_c.rds")

################################################################################
H2O2_G200 %>% na.omit() %>% 
  filter(grepl(8, Parent), Bulk != "Freeze") %>%
  mutate_if(is.character, as.factor) -> H2O2_CSS8_factor

#Set contrasts
contrasts(H2O2_CSS8_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(H2O2_CSS8_factor$Bulk) <- matrix(c(0,1))

#Run GLM
H2O2_CSS8_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "H2O2", CSS = "Fixed VIII")  -> H2O2_CSS8_glm

#Save
saveRDS(H2O2_CSS8_glm, file = "allstats/H2O2_CSS8_glm_allstats_c.rds")

```

### Paraquat

```{r}
Paraquat_G200 %>% na.omit() %>%
  filter(grepl(1, Parent)) %>%
  mutate_if(is.character, as.factor) -> PQ_CSS1_factor

#Set contrasts
contrasts(PQ_CSS1_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(PQ_CSS1_factor$Bulk) <- matrix(c(0,1))

#Run GLM
PQ_CSS1_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>%
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Paraquat", CSS = "Fixed I")  -> PQ_CSS1_glm

#Save
saveRDS(PQ_CSS1_glm, file = "allstats/PQ_CSS1_glm_allstats_xNAs.rds")

################################################################################
Paraquat_G200 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> PQ_CSS8_factor

#Set contrasts
contrasts(PQ_CSS8_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(PQ_CSS8_factor$Bulk) <- matrix(c(0,1))

#Run GLM
PQ_CSS8_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Paraquat", CSS = "Fixed VIII")  -> PQ_CSS8_glm

#Save
saveRDS(PQ_CSS8_glm, file = "allstats/PQ_CSS8_glm_allstats_xNAs.rds")

PQ_CSS8_factor %>% filter(CHROM == "V") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) ->PQTest

PQTest %>% filter(POS > 2.2e05, POS < 2.3e05) %>% unnest(cols = c(Oak, Wine)) %>% mutate(logOdds = log(Wine/Oak)) %>%
  ggplot(aes(x = POS, y = logOdds, color = Dataset)) + geom_point(alpha = 0.3) +
  geom_vline(aes(xintercept = 225782)) +
  theme_bw()

```

### Zeocin


```{r}
unique(Zeocin_G200$Dataset)

Zeocin_G200 %>% na.omit() %>%   
  filter(Dataset != "H2LKCDMX2_n01_Wine1_Z_Ar.fastq", Dataset != "H2LKCDMX2_n01_Wine1_Z_Br.fastq") %>%
  filter(Bulk != "Freeze") %>%
  filter(grepl(1, Parent)) %>%
  mutate_if(is.character, as.factor) -> Zeocin_CSS1_factor

#Set contrasts
contrasts(Zeocin_CSS1_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Zeocin_CSS1_factor$Bulk) <- matrix(c(0,1))

#Run GLM
Zeocin_CSS1_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>%
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Zeocin", CSS = "Fixed I")  -> Zeocin_CSS1_glm

#Save
saveRDS(Zeocin_CSS1_glm, file = "allstats/Zeocin_CSS1_glm_allstats_c.rds")

################################################################################
Zeocin_G200 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
    filter(Bulk != "Freeze") %>%
  mutate_if(is.character, as.factor) -> Zeocin_CSS8_factor

#Set contrasts
contrasts(Zeocin_CSS8_factor$Parent) <- matrix(c(0.5, -0.5))
contrasts(Zeocin_CSS8_factor$Bulk) <- matrix(c(0,1))

#Run GLM
Zeocin_CSS8_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  %>%
  mutate(Selection = "Zeocin", CSS = "Fixed VIII")  -> Zeocin_CSS8_glm

#Save
saveRDS(Zeocin_CSS8_glm, file = "allstats/Zeocin_CSS8_glm_allstats_c.rds")

```

# Cutoffs

```{r}
Copper_G200<- readRDS("covfilt/Copper_G200_J25.rds")
Cycloheximide_G200<- readRDS("covfilt/Cycloheximide_G200_J25.rds")
Ethanol_G200<- readRDS("covfilt/Ethanol_G200_covfilt_xNAs.rds")
Fluconazole_G200<- readRDS("covfilt/Fluconazole_G200_J25.rds")
H2O2_G200<- readRDS("covfilt/H2O2_G200d_covfiltc.rds")
Paraquat_G200 <- readRDS("covfilt/Paraquat_G200_covfilt_xNAs.rds")
Zeocin_G200 <- readRDS("covfilt/Zeocin_G200d_covfiltc.rds")
```

## Without Excluding Chr XV (Normal)
### Cycloheximide Permutations

```{r}
Cycloheximide_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Cycloheximide_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Cyc

################################################################################
Perm_Cyc %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Cyc_CSS8_factor_p

#Set contrasts
contrasts(Cyc_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Cyc_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Cyc_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Cyc_CSS8_perm

################################
Perm_Cyc %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Cyc_CSS1_factor_p

#Set contrasts
contrasts(Cyc_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Cyc_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Cyc_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Cyc_CSS1_perm

################################

Cyc_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Cyc_CSS8_q5

Cyc_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Cyc_CSS1_q5

saveRDS(Cyc_CSS8_q5, file = "allstats/Cyc_CSS8_q5_J25.rds")
saveRDS(Cyc_CSS1_q5, file = "allstats/Cyc_CSS1_q5_J25.rds")


```

### Copper

```{r}

Copper_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Copper_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Copper

################################################################################
Perm_Copper %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Copper_CSS8_factor_p

#Set contrasts
contrasts(Copper_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Copper_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Copper_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Copper_CSS8_perm

################################
Perm_Copper %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Copper_CSS1_factor_p

#Set contrasts
contrasts(Copper_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Copper_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Copper_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Copper_CSS1_perm

################################

Copper_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Copper_CSS8_q5

Copper_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Copper_CSS1_q5

saveRDS(Copper_CSS8_q5, file = "allstats/Copper_CSS8_q5_J25.rds")
saveRDS(Copper_CSS1_q5, file = "allstats/Copper_CSS1_q5_J25.rds")


```

### H2O2

```{r}
H2O2_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

H2O2_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_H2O2

################################################################################
Perm_H2O2 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> H2O2_CSS8_factor_p

#Set contrasts
contrasts(H2O2_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(H2O2_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
H2O2_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> H2O2_CSS8_perm

################################
Perm_H2O2 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> H2O2_CSS1_factor_p

#Set contrasts
contrasts(H2O2_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(H2O2_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
H2O2_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> H2O2_CSS1_perm

################################

H2O2_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> H2O2_CSS8_q5

H2O2_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> H2O2_CSS1_q5

saveRDS(H2O2_CSS8_q5, file = "allstats/H2O2_CSS8_q5_J25.rds")
saveRDS(H2O2_CSS1_q5, file = "allstats/H2O2_CSS1_q5_J25.rds")


```

### Fluconazole

```{r}
Fluconazole_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Fluconazole_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Fluconazole

################################################################################
Perm_Fluconazole %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Fluconazole_CSS8_factor_p

#Set contrasts
contrasts(Fluconazole_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Fluconazole_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Fluconazole_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Fluconazole_CSS8_perm

################################
Perm_Fluconazole %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Fluconazole_CSS1_factor_p

#Set contrasts
contrasts(Fluconazole_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Fluconazole_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Fluconazole_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Fluconazole_CSS1_perm

################################

Fluconazole_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Fluconazole_CSS8_q5

Fluconazole_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Fluconazole_CSS1_q5

saveRDS(Fluconazole_CSS8_q5, file = "allstats/Fluconazole_CSS8_q5_J25.rds")
saveRDS(Fluconazole_CSS1_q5, file = "allstats/Fluconazole_CSS1_q5_J25.rds")

```

### Ethanol

```{r}
Ethanol_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Ethanol_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Ethanol

################################################################################
Perm_Ethanol %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Ethanol_CSS8_factor_p

#Set contrasts
contrasts(Ethanol_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Ethanol_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Ethanol_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Ethanol_CSS8_perm

################################
Perm_Ethanol %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Ethanol_CSS1_factor_p

#Set contrasts
contrasts(Ethanol_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Ethanol_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Ethanol_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Ethanol_CSS1_perm

################################

Ethanol_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Ethanol_CSS8_q5

Ethanol_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Ethanol_CSS1_q5
# 
# saveRDS(Ethanol_CSS8_q5, file = "allstats/Ethanol_CSS8_q5_J25.rds")
# saveRDS(Ethanol_CSS1_q5, file = "allstats/Ethanol_CSS1_q5_J25.rds")

saveRDS(Ethanol_CSS8_q5, file = "allstats/Ethanol_CSS8_q5_J25.rds")
saveRDS(Ethanol_CSS1_q5, file = "allstats/Ethanol_CSS1_q5_J25.rds")

```

### Zeocin

```{r}
Zeocin_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Zeocin_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Zeocin

################################################################################
Perm_Zeocin %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Zeocin_CSS8_factor_p

#Set contrasts
contrasts(Zeocin_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Zeocin_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Zeocin_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Zeocin_CSS8_perm

################################
Perm_Zeocin %>% na.omit() %>%
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Zeocin_CSS1_factor_p

#Set contrasts
contrasts(Zeocin_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Zeocin_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Zeocin_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>%
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Zeocin_CSS1_perm

################################

Zeocin_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Zeocin_CSS8_q5

Zeocin_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Zeocin_CSS1_q5


saveRDS(Zeocin_CSS8_q5, file = "allstats/Zeocin_CSS8_q5_J25.rds")
saveRDS(Zeocin_CSS1_q5, file = "allstats/Zeocin_CSS1_q5_J25.rds")

```


### Paraquat

```{r}
Paraquat_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Paraquat_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Paraquat

################################################################################
Perm_Paraquat %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Paraquat_CSS8_factor_p

#Set contrasts
contrasts(Paraquat_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Paraquat_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Paraquat_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>%
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Paraquat_CSS8_perm

################################
Perm_Paraquat %>% na.omit() %>%
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Paraquat_CSS1_factor_p

#Set contrasts
contrasts(Paraquat_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Paraquat_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Paraquat_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>%
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Paraquat_CSS1_perm

################################

Paraquat_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Paraquat_CSS8_q5

Paraquat_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Paraquat_CSS1_q5


saveRDS(Paraquat_CSS8_q5, file = "allstats/Paraquat_CSS8_q5_J25.rds")
saveRDS(Paraquat_CSS1_q5, file = "allstats/Paraquat_CSS1_q5_J25.rds")

```




## EXCLUDING Chr XV (s15) - run this overnight

### Cycloheximide Permutations

```{r}
Cycloheximide_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Cycloheximide_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Cyc

################################################################################
Perm_Cyc %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Cyc_CSS8_factor_p

#Set contrasts
contrasts(Cyc_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Cyc_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Cyc_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Cyc_CSS8_perm

################################
Perm_Cyc %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Cyc_CSS1_factor_p

#Set contrasts
contrasts(Cyc_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Cyc_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Cyc_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Cyc_CSS1_perm

################################

Cyc_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Cyc_CSS8_q5

Cyc_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Cyc_CSS1_q5

saveRDS(Cyc_CSS8_q5, file = "allstats/Cyc_CSS8_q5_allstats_s15.rds")
saveRDS(Cyc_CSS1_q5, file = "allstats/Cyc_CSS1_q5_allstats_s15.rds")


```

### Copper

```{r}

Copper_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Copper_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Copper

################################################################################
Perm_Copper %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Copper_CSS8_factor_p

#Set contrasts
contrasts(Copper_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Copper_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Copper_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Copper_CSS8_perm

################################
Perm_Copper %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Copper_CSS1_factor_p

#Set contrasts
contrasts(Copper_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Copper_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Copper_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Copper_CSS1_perm

################################

Copper_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Copper_CSS8_q5

Copper_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Copper_CSS1_q5

saveRDS(Copper_CSS8_q5, file = "allstats/Copper_CSS8_q5_allstats_s15.rds")
saveRDS(Copper_CSS1_q5, file = "allstats/Copper_CSS1_q5_allstats_s15.rds")


```

### H2O2

```{r}
H2O2_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

H2O2_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_H2O2

################################################################################
Perm_H2O2 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> H2O2_CSS8_factor_p

#Set contrasts
contrasts(H2O2_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(H2O2_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
H2O2_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> H2O2_CSS8_perm

################################
Perm_H2O2 %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> H2O2_CSS1_factor_p

#Set contrasts
contrasts(H2O2_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(H2O2_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
H2O2_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> H2O2_CSS1_perm

################################

H2O2_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> H2O2_CSS8_q5

H2O2_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> H2O2_CSS1_q5

saveRDS(H2O2_CSS8_q5, file = "allstats/H2O2_CSS8_q5_allstats_s15.rds")
saveRDS(H2O2_CSS1_q5, file = "allstats/H2O2_CSS1_q5_allstats_s15.rds")


```

### Fluconazole

```{r}
Fluconazole_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Fluconazole_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Fluconazole

################################################################################
Perm_Fluconazole %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Fluconazole_CSS8_factor_p

#Set contrasts
contrasts(Fluconazole_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Fluconazole_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Fluconazole_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Fluconazole_CSS8_perm

################################
Perm_Fluconazole %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Fluconazole_CSS1_factor_p

#Set contrasts
contrasts(Fluconazole_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Fluconazole_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Fluconazole_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Fluconazole_CSS1_perm

################################

Fluconazole_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Fluconazole_CSS8_q5

Fluconazole_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Fluconazole_CSS1_q5

saveRDS(Fluconazole_CSS8_q5, file = "allstats/Fluconazole_CSS8_q5_allstats_s15.rds")
saveRDS(Fluconazole_CSS1_q5, file = "allstats/Fluconazole_CSS1_q5_allstats_s15.rds")

```

### Ethanol

```{r}
Ethanol_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Ethanol_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Ethanol

################################################################################
Perm_Ethanol %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Ethanol_CSS8_factor_p

#Set contrasts
contrasts(Ethanol_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Ethanol_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Ethanol_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Ethanol_CSS8_perm

################################
Perm_Ethanol %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Ethanol_CSS1_factor_p

#Set contrasts
contrasts(Ethanol_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Ethanol_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Ethanol_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Ethanol_CSS1_perm

################################

Ethanol_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Ethanol_CSS8_q5

Ethanol_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Ethanol_CSS1_q5
# 
# saveRDS(Ethanol_CSS8_q5, file = "allstats/Ethanol_CSS8_q5_allstats_s15.rds")
# saveRDS(Ethanol_CSS1_q5, file = "allstats/Ethanol_CSS1_q5_allstats_s15.rds")

saveRDS(Ethanol_CSS8_q5, file = "allstats/Ethanol_CSS8_q5_allstats_s15.rds")
saveRDS(Ethanol_CSS1_q5, file = "allstats/Ethanol_CSS1_q5_allstats_s15.rds")

```

### Zeocin

```{r}
Zeocin_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Zeocin_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Zeocin

################################################################################
Perm_Zeocin %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Zeocin_CSS8_factor_p

#Set contrasts
contrasts(Zeocin_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Zeocin_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Zeocin_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,  
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Zeocin_CSS8_perm

################################
Perm_Zeocin %>% na.omit() %>%
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Zeocin_CSS1_factor_p

#Set contrasts
contrasts(Zeocin_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Zeocin_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Zeocin_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>%
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Zeocin_CSS1_perm

################################

Zeocin_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Zeocin_CSS8_q5

Zeocin_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Zeocin_CSS1_q5


saveRDS(Zeocin_CSS8_q5, file = "allstats/Zeocin_CSS8_q5_allstats_s15.rds")
saveRDS(Zeocin_CSS1_q5, file = "allstats/Zeocin_CSS1_q5_allstats_s15.rds")

```


### Paraquat

```{r}
Paraquat_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

Paraquat_G200 %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V", "XV") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount")-> Perm_Paraquat

################################################################################
Perm_Paraquat %>% na.omit() %>% 
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Paraquat_CSS8_factor_p

#Set contrasts
contrasts(Paraquat_CSS8_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Paraquat_CSS8_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Paraquat_CSS8_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>%
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                                  label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Paraquat_CSS8_perm

################################
Perm_Paraquat %>% na.omit() %>%
  filter(grepl(8, Parent)) %>%
  mutate_if(is.character, as.factor) -> Paraquat_CSS1_factor_p

#Set contrasts
contrasts(Paraquat_CSS1_factor_p$Parent) <- matrix(c(0.5, -0.5))
contrasts(Paraquat_CSS1_factor_p$Bulk) <- matrix(c(0,1))

#Run GLM
Paraquat_CSS1_factor_p %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>%
  #Select sequencing runs for Chr I fixed
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_all(Bulk = Bulk,
                                                            Parent = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Parent "),
                                                            label = rep(c("cept","Bulk", "Background", "Interaction"),4), stat = c("e","e","e","e","se","se","se","se","z","z","z","z","p","p","p","p"))  -> Paraquat_CSS1_perm

################################

Paraquat_CSS8_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Paraquat_CSS8_q5

Paraquat_CSS1_perm %>% na.omit() %>% group_by(label) %>%
  summarize(quant = quantile(abs(zscore), 0.95)) -> Paraquat_CSS1_q5


saveRDS(Paraquat_CSS8_q5, file = "allstats/Paraquat_CSS8_q5_allstats_s15.rds")
saveRDS(Paraquat_CSS1_q5, file = "allstats/Paraquat_CSS1_q5_allstats_s15.rds")

```


