---
title: "Sequencing Coverage Analysis"
author: "Cass"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)))
```

## Copper Data

Load in the unsmoothed sequences of all of the data

```{r}
CopperData <- readRDS("C:\\Users\\Cassandra\\Documents\\GitHub\\CSS_Paper\\ChromosomeFixation\\BSA_Analysis\\Output\\rawdata_called_ex.rds")
```

```{r}
CopperData %>% ggplot(aes(x = Coverage, color = Dataset)) + geom_density() +
  theme(legend.position = "none")

CopperData %>% group_by(CHROM, POS) %>%
  summarize(CovSum = sum(Coverage, na.rm = TRUE)) -> TotalCovByPOS

TotalCovByPOS %>% filter(CHROM != "M") %>% ggplot(aes(x = CovSum/6)) + geom_density(aes(color = CHROM)) +
  geom_density(data = CopperData, aes(x = Coverage)) +
  theme(legend.position = "bottom")
```

That's misleading because it's for the total of all parents and bulks

```{r}
CopperData_bybulk <- readRDS("C:\\Users\\Cassandra\\Documents\\GitHub\\CSS_Paper\\ChromosomeFixation\\BSA_Analysis\\Output\\rd_factor_G200.rds")

CopperData_bybulk %>% ungroup() %>% group_by(CHROM, POS, Bulk, Background) %>%
  summarize(SmoothedCounts = sum(SmoothCount)) -> TotalSmoothed

TotalSmoothed %>% filter(CHROM != "M") %>% 
  ggplot(aes(y = SmoothedCounts, x = CHROM, color = paste(Bulk, Background))) + 
  geom_boxplot(outliers = FALSE) +
  theme(legend.position = "bottom")

TotalSmoothed %>% filter(CHROM != "M") %>% 
  ggplot(aes(y = SmoothedCounts, x = CHROM)) + 
  geom_boxplot(outliers = FALSE) +
  theme(legend.position = "bottom")
  
```
## New Data

What about the rest of the data?

```{r}
CSS8 <- readRDS("C:\\Users\\Cassandra\\Documents\\GitHub\\Sequencing\\Analysis\\Analysis_2024\\CSSVIII_A24_rawdata_smoothed_G200x.rds") %>% 
  separate(Dataset, sep = "_n01_", into = c("Pool", "Dataset"))

CSS1 <- readRDS("C:\\Users\\Cassandra\\Documents\\GitHub\\Sequencing\\Analysis\\Analysis_2024\\CSSI_A24_rawdata_smoothed_G200x.rds") %>% 
  separate(Dataset, sep = "_n01_", into = c("Pool", "Dataset"))
```

```{r}

unique(CSS1$Dataset)

CSS1 %>% filter(grepl("AIG", Dataset) == FALSE) %>%
  separate(Dataset, sep = "_", into = c("Bulk", "Parent", "Rep", "Day")) -> CSS1adj

CSS8 %>% filter(grepl("AIG", Dataset) == FALSE) %>%
  separate(Dataset, sep = "_", into = c("Bulk", "Parent", "Rep", "Day")) %>% 
  mutate(Parent = paste(Parent, "8")) %>%
  rbind(CSS1adj) -> AllNewFiles
  
AllNewFiles %>% group_by(CHROM, POS, Bulk, Parent) %>%
  summarize(SmoothedCounts = sum(SmoothCount)) -> TotalSmoothed_newdata

TotalSmoothed_newdata %>% filter(CHROM != "M", Bulk != "undetermined.fastq") %>% 
  ggplot(aes(y = SmoothedCounts, x = CHROM)) + 
  geom_boxplot(outliers = FALSE, aes(color = Bulk)) +
  theme(legend.position = "bottom")
```

What about the old Fluconazole data? And cyc/zeo?

```{r}

```

