---
title: "Overall Sequencing Stats"
author: "Cass"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))
```

# Data

```{r}
AllTrait_glms <- readRDS(file = "Jan15_AllTrait_glms.rds")

H2LK2DMX2_mqc <- readRDS("H2LK2DMX2.mqc.rds")
```

```{r}

H2LK2DMX2_mqc %>% group_by(Dataset) %>%
  na.omit() %>% 
  summarize(Covg_mean = mean(Coverage),
            Covg_sd = sd(Coverage)) -> CoverageStats

CoverageStats %>% ggplot(aes(x = Covg_mean, y = Covg_sd, color = Dataset)) + geom_point()

# H2LK2DMX2_mqc %>% group_by(Dataset, CHROM) %>%
#   arrange(POS) %>%
#   na.omit() %>% 
#   summarize(Cov_mean = frollapply(Coverage, n = 200, FUN = mean),
#             Cov_sd = frollapply(Coverage, n = 200, FUN = sd),
#             Cov_range = frollapply(Coverage, n = 200, FUN = range)) -> test
```

If you have 500 reads to distribute across 8 bulks (4 bulks and 2 replicates each), and you only have a bulk effect of increasing wine by logOdds(1), what distribution of reads is most powerful?

```{r}
WDilute <- table(sample(c("Oak", "Wine"), size = 500, replace = TRUE, prob = c(0.5, 0.5)))
ODilute <- table(sample(c("Oak", "Wine"), size = 500, replace = TRUE, prob = c(0.5, 0.5)))
WSelected <- table(sample(c("Oak", "Wine"), size = 500, replace = TRUE, prob = c(0.5, 0.5)))
OSelected <- table(sample(c("Oak", "Wine"), size = 500, replace = TRUE, prob = c(0.5, 0.5)))

rbind(WDilute, ODilute, WSelected,OSelected) -> tempdata

tempdata %>% as.data.frame() %>% mutate(Parent = c("W","O","W","O"),
                                        Bulk = c("Dilute", "Dilute", "Selected", "Selected")) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Count") %>%
  mutate(Bulk = as.factor(Bulk),
         Parent = as.factor(Parent),
         Allele = as.factor(Allele)) %>%
  summarize(zscores = glm_cb2_short(Bulk = Bulk, Parent = Parent, Allele = Allele,
                                    formula = "Allele ~ Bulk * Parent",
                                    W = Count),
            label = c("Intercept", "Bulk", "Parent", "Interaction")) -> tempzscores
```

```{r}
# Simulate Null Data
for(i in 1:50){
  for(k in c(50,100,200,300,400,500,1000)){
    WDilute <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.5, 0.5)))
    ODilute <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.5, 0.5)))
    WSelected <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.5, 0.5)))
    OSelected <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.5, 0.5)))

    as.data.frame(rbind(WDilute, ODilute, WSelected,OSelected)) %>% mutate(ID = i,
                                                                           Coverage = k) -> tempdata
    
    if(i == 1 & k == 50){
      alldata <- tempdata
    }else{
      alldata <- rbind(alldata, tempdata)
    }
  }
}

alldata %>% mutate(Parent = rep(c("W","O","W","O"), 50*7),
                                        Bulk = rep(c("Dilute", "Dilute", "Selected", "Selected"), 50*7)) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Count") %>%
  mutate(Bulk = as.factor(Bulk),
         Parent = as.factor(Parent),
         Allele = as.factor(Allele)) %>%
  group_by(ID, Coverage) %>%
  summarize(zscores = glm_cb2_short(Bulk = Bulk, Parent = Parent, Allele = Allele,
                                    formula = "Allele ~ Bulk * Parent",
                                    W = Count),
            label = c("Intercept", "Bulk", "Parent", "Interaction")) -> null_zscores

null_zscores %>% ggplot(aes(x = factor(Coverage), y = zscores, color = factor(Coverage))) + 
  geom_boxplot() + 
  geom_jitter()
```

```{r}
# Simulate 50% Increase for Selected Data
for(i in 1:50){
  for(k in c(50,100,200,300,400,500,1000)){
    WDilute <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.5, 0.5)))
    ODilute <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.5, 0.5)))
    WSelected <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.25, 0.75)))
    OSelected <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.25, 0.75)))

    as.data.frame(rbind(WDilute, ODilute, WSelected,OSelected)) %>% mutate(ID = i,
                                                                           Coverage = k) -> tempdata
    
    if(i == 1 & k == 50){
      alldata <- tempdata
    }else{
      alldata <- rbind(alldata, tempdata)
    }
  }
}

alldata %>% mutate(Parent = rep(c("W","O","W","O"), 50*7),
                                        Bulk = rep(c("Dilute", "Dilute", "Selected", "Selected"), 50*7)) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Count") %>%
  mutate(Bulk = as.factor(Bulk),
         Parent = as.factor(Parent),
         Allele = as.factor(Allele)) %>%
  group_by(ID, Coverage) %>%
  summarize(zscores = glm_cb2_short(Bulk = Bulk, Parent = Parent, Allele = Allele,
                                    formula = "Allele ~ Bulk * Parent",
                                    W = Count),
            label = c("Intercept", "Bulk", "Parent", "Interaction")) -> bulk_zscores

bulk_zscores  %>% ggplot(aes(x = factor(Coverage), y = zscores, color = factor(Coverage))) + 
  geom_boxplot() + 
  geom_jitter()
```

```{r}
#Simulate uneven bulk sizes

# Simulate 50% Increase for Selected Data
for(i in 1:50){
  for(k in c(50,100,200,300,400,500,1000)){
    WDilute <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    ODilute <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    WSelected <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.25, 0.75)))
    OSelected <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.25, 0.75)))

    as.data.frame(rbind(WDilute, ODilute, WSelected,OSelected)) %>% mutate(ID = i,
                                                                           Coverage = k) -> tempdata
    
    if(i == 1 & k == 50){
      alldata <- tempdata
    }else{
      alldata <- rbind(alldata, tempdata)
    }
  }
}

alldata %>% mutate(Parent = rep(c("W","O","W","O"), 50*7),
                                        Bulk = rep(c("Dilute", "Dilute", "Selected", "Selected"), 50*7)) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Count") %>%
  mutate(Bulk = as.factor(Bulk),
         Parent = as.factor(Parent),
         Allele = as.factor(Allele)) %>%
  group_by(ID, Coverage) %>%
  summarize(zscores = glm_cb2_short(Bulk = Bulk, Parent = Parent, Allele = Allele,
                                    formula = "Allele ~ Bulk * Parent",
                                    W = Count),
            label = c("Intercept", "Bulk", "Parent", "Interaction")) -> bulk_sizediff_zscores

bulk_sizediff_zscores  %>% ggplot(aes(x = factor(Coverage), y = zscores, color = factor(Coverage))) + 
  geom_boxplot() + 
  geom_jitter() +
  ggtitle("")
```
```{r}
#Simulate if the overall distribution of reads matters
for(i in 1:50){
  for(k in c(50,100,200,300,400,500,600)){
    OtherSize = 650 - k
    WDilute <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    ODilute <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    WSelected <- table(sample(c("Oak", "Wine"), size = OtherSize, replace = TRUE, prob = c(0.75, 0.25)))
    OSelected <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.25, 0.75)))

    as.data.frame(rbind(WDilute, ODilute, WSelected,OSelected)) %>% mutate(ID = i,
                                                                           Coverage = k) -> tempdata
    
    if(i == 1 & k == 50){
      alldata <- tempdata
    }else{
      alldata <- rbind(alldata, tempdata)
    }
  }
}

alldata %>% mutate(Parent = rep(c("W","O","W","O"), 50*7),
                                        Bulk = rep(c("Dilute", "Dilute", "Selected", "Selected"), 50*7)) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Count") %>%
  mutate(Bulk = as.factor(Bulk),
         Parent = as.factor(Parent),
         Allele = as.factor(Allele)) -> tempalldata

contrasts(tempalldata$Parent) <- c(-0.5, 0.5)
contrasts(tempalldata$Bulk)

tempalldata %>%
  group_by(ID, Coverage) %>%
  summarize(zscores = glm_cb2_short(Bulk = Bulk, Parent = Parent, Allele = Allele,
                                    formula = "Allele ~ Bulk * Parent",
                                    W = Count),
            label = c("Intercept", "Bulk", "Parent", "Interaction")) -> sizedistr_zscores

sizedistr_zscores %>%   mutate(CovRatio = Coverage/(650-Coverage)) %>%
  ggplot(aes(x = log(CovRatio), y = zscores, color = factor(label))) + geom_point() +
  facet_wrap(facets = "label") +
  theme_bw()

sizedistr_zscores  %>% ggplot(aes(x = factor(Coverage), y = zscores, color = factor(label))) + 
  geom_boxplot() + 
  geom_jitter(position = position_jitterdodge()) +
  ggtitle("") +
  theme_bw()

```


```{r}
#Now see if having replicates helps vs not
for(i in 1:50){
  for(k in c(50,100,200,300,400,500,1000)){
    WDiluteA <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    ODiluteA <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    WSelectedA <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.25, 0.75)))
    OSelectedA <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.25, 0.75)))
    
    WDiluteB <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    ODiluteB <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    WSelectedB <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.25, 0.75)))
    OSelectedB <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.25, 0.75)))

    as.data.frame(rbind(WDiluteA, ODiluteA, WSelectedA,OSelectedA,
                        WDiluteB, ODiluteB, WSelectedB,OSelectedB)) %>% mutate(ID = i,
                                                                           Coverage = k) -> tempdata
    
    if(i == 1 & k == 50){
      alldata <- tempdata
    }else{
      alldata <- rbind(alldata, tempdata)
    }
  }
}

alldata %>% mutate(Parent = rep(c("W","O","W","O"), 50*7*2),
                                        Bulk = rep(c("Dilute", "Dilute", "Selected", "Selected"), 50*7*2),
                   Replicate = rep(c("A", "A", "A", "A", "B","B","B","B"), 50*7)) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Count") %>%
  mutate(Bulk = as.factor(Bulk),
         Parent = as.factor(Parent),
         Allele = as.factor(Allele)) -> tempalldata

contrasts(tempalldata$Parent) <- c(-0.5, 0.5)
contrasts(tempalldata$Bulk)

tempalldata %>%
  group_by(ID, Coverage) %>%
  summarize(zscores = glm_cb2_short(Bulk = Bulk, Parent = Parent, Allele = Allele,
                                    formula = "Allele ~ Bulk * Parent",
                                    W = Count),
            label = c("Intercept", "Bulk", "Parent", "Interaction")) -> rep_zscores


tempalldata %>%
  group_by(ID, Coverage, Allele, Bulk, Parent) %>%
  summarize(Count = sum(Count)) %>%
  group_by(ID, Coverage) %>%
  summarize(zscores = glm_cb2_short(Bulk = Bulk, Parent = Parent, Allele = Allele,
                                    formula = "Allele ~ Bulk * Parent",
                                    W = Count),
            label = c("Intercept", "Bulk", "Parent", "Interaction")) -> addrep_zscores

addrep_zscores  %>% ggplot(aes(x = factor(Coverage), y = zscores, color = factor(label))) + 
  geom_boxplot() + 
  geom_jitter(position = position_jitterdodge()) +
  ggtitle("Added") +
  theme_bw()

rep_zscores %>% ggplot(aes(x = factor(Coverage), y = zscores, color = factor(label))) + 
  geom_boxplot() + 
  geom_jitter(position = position_jitterdodge()) +
  ggtitle("Replicates") +
  theme_bw()
```

```{r}
#Okay BUT if the coverages are uneven between replicates, does it matter then?

for(i in 1:50){
  for(k in c(50,100,200,300,400,500,1000)){
    WDiluteA <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    ODiluteA <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    WSelectedA <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.25, 0.75)))
    OSelectedA <- table(sample(c("Oak", "Wine"), size = 1050 - k, replace = TRUE, prob = c(0.25, 0.75)))
    
    WDiluteB <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    ODiluteB <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.5, 0.5)))
    WSelectedB <- table(sample(c("Oak", "Wine"), size = 300, replace = TRUE, prob = c(0.25, 0.75)))
    OSelectedB <- table(sample(c("Oak", "Wine"), size = k, replace = TRUE, prob = c(0.25, 0.75)))

    as.data.frame(rbind(WDiluteA, ODiluteA, WSelectedA,OSelectedA,
                        WDiluteB, ODiluteB, WSelectedB,OSelectedB)) %>% mutate(ID = i,
                                                                           Coverage = k) -> tempdata
    
    if(i == 1 & k == 50){
      alldata <- tempdata
    }else{
      alldata <- rbind(alldata, tempdata)
    }
  }
}

alldata %>% mutate(Parent = rep(c("W","O","W","O"), 50*7*2),
                                        Bulk = rep(c("Dilute", "Dilute", "Selected", "Selected"), 50*7*2),
                   Replicate = rep(c("A", "A", "A", "A", "B","B","B","B"), 50*7)) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Count") %>%
  mutate(Bulk = as.factor(Bulk),
         Parent = as.factor(Parent),
         Allele = as.factor(Allele)) -> tempalldata

contrasts(tempalldata$Parent) <- c(-0.5, 0.5)
contrasts(tempalldata$Bulk)

tempalldata %>%
  group_by(ID, Coverage) %>%
  summarize(zscores = glm_cb2_short(Bulk = Bulk, Parent = Parent, Allele = Allele,
                                    formula = "Allele ~ Bulk * Parent",
                                    W = Count),
            label = c("Intercept", "Bulk", "Parent", "Interaction")) -> rep_zscores


tempalldata %>%
  group_by(ID, Coverage, Allele, Bulk, Parent) %>%
  summarize(Count = sum(Count)) %>%
  group_by(ID, Coverage) %>%
  summarize(zscores = glm_cb2_short(Bulk = Bulk, Parent = Parent, Allele = Allele,
                                    formula = "Allele ~ Bulk * Parent",
                                    W = Count),
            label = c("Intercept", "Bulk", "Parent", "Interaction")) -> addrep_zscores

addrep_zscores  %>% ggplot(aes(x = factor(Coverage), y = zscores, color = factor(label))) + 
  geom_boxplot() + 
  geom_jitter(position = position_jitterdodge()) +
  ggtitle("Added") +
  theme_bw()

rep_zscores %>% ggplot(aes(x = factor(Coverage), y = zscores, color = factor(label))) + 
  geom_boxplot() + 
  geom_jitter(position = position_jitterdodge()) +
  ggtitle("Replicates") +
  theme_bw()
```

So it looks like the coverage makes the most difference, but what is the range of coverage in a single experiment of each actual position?

```{r}
H2LK2DMX2_mqc %>% ggplot(aes(x = Coverage, y = Dataset, color = Dataset)) + geom_boxplot() 
```
Does the genome vary at all in coverage?

```{r}
H2LK2DMX2_mqc %>% group_by(Dataset) %>% summarize(AvgCovg = mean(Coverage, na.rm = TRUE)) -> Means

H2LK2DMX2_mqc %>% left_join(Means) %>% ungroup() %>% 
  na.omit() %>%
  group_by(Dataset, CHROM, POS) %>% 
  select(Dataset, CHROM, POS, Coverage, AvgCovg) %>%
  distinct() %>%
  mutate(CovChange = log(Coverage/AvgCovg)) -> Totals

Totals %>% ggplot(aes(x = POS, y = CovChange, color = Dataset)) + geom_line() +
  geom_hline(aes(yintercept = -1), linetype = "dashed") +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none")

Totals %>% filter(CovChange > -1) %>%
  ggplot(aes(x = Coverage, color = Dataset, y = Dataset)) + geom_boxplot() +
  theme(legend.position = "none")
```
```{r}
Totals %>% filter(CovChange > -1) %>% left_join(H2LK2DMX2_mqc) %>% ungroup() %>%
  group_by(CHROM, Pool, Parent, Bulk, Allele, Dataset) %>%
  na.omit() %>%
  arrange(POS) %>% 
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) -> Totals_G200

#saveRDS("Totals_G200", file = "test_smoothwithoutlowcoverage.rds")
#Zeo_c <- readRDS("J25/Zeocin_G200d_J25c.rds")

Zeo_c %>% filter(Dataset == "H2LK2DMX2_n01_Oak8_D_1.fastq") %>% mutate(Filter = "c") %>%
  ungroup()  -> subsetz

Totals_G200 %>% filter(Dataset == "Oak8_D_1") %>% mutate(Filter = "all") %>%
  rbind(subsetz) %>%
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>%
  mutate(logOdds = log(Wine/Oak)) -> Test_Oak8D1_Zeocin

Totals %>% filter(CovChange < -1) %>% left_join(H2LK2DMX2_mqc) %>% ungroup() -> Rejects

Test_Oak8D1_Zeocin %>% filter(CHROM == "XII") %>%
  ggplot(aes(x =POS, y = logOdds, color = Filter)) + geom_line(alpha = 0.8) +
  #geom_point(data = Rejects, aes(x = POS, y = Coverage), color = "gray") +
  facet_grid(~CHROM, scales = "free", space = "free")


Totals_G200 %>% filter(Dataset == "Oak8_D_1") %>% na.omit() #mutate(Filter = "all") %>%
  rbind(subsetz) %>%
  filter(CHROM == "XI") %>%
  ggplot(aes(x = POS, y = SmoothCount)) +
  geom_point(aes(color = Allele)) +
    geom_point(data = Rejects[Rejects$CHROM == "XI",], 
               aes(x = POS, y = Coverage), color = "gray", alpha = 0.2) +
    facet_grid(Filter~CHROM, scales = "free", space = "free")

Totals_G200 %>% na.omit()
```

## Comparing glm from all points vs coverage filtered points

Look at the cutoffs first
```{r}
noncfs <- readRDS("")
cfs <- readRDS("allQs_coveragefiltered.rds")
```

```{r}
cf_AllTrait_glms <- readRDS("CoverageFiltered_AllTrait_glms.rds")
nocf_AllTrait_glms <- readRDS("Jan15_AllTrait_glms.rds")

```

```{r}
cf_AllTrait_glms %>% select(label, Selection, CSS, q15, quant) %>% distinct() -> cf_qs
nocf_AllTrait_glms %>% select(label, Selection, CSS, q15, quant)%>% distinct() -> nocf_qs

nocf_qs %>% ggplot(aes(x = Selection, y = quant, color = Selection)) + 
  geom_point(alpha = 0.3, size = 3) +
  geom_point(data = cf_qs, shape = 1, size = 4) +
  facet_grid(label~CSS) 

```

```{r}
cf_AllTrait_glms %>% filter(label %in% c("Bulk", "Interaction")) %>% mutate(CF = TRUE) -> FCF

nocf_AllTrait_glms %>% filter(label %in% c("Bulk", "Interaction")) %>% mutate(CF = FALSE) -> FnoCF

rbind(FCF, FnoCF) -> CF_Comparison

CF_Comparison %>% filter(label == "Bulk", CHROM == "XV", CSS == "Fixed VIII") %>% 
  ggplot(aes(x = POS, y = zscore, color = CF)) + geom_line(alpha = 0.3) +
  facet_grid(Selection ~ CHROM, scales = "free", space = "free")


```

