---
title: "RD CNV for HPC"
author: "Cassandra Buzby, adapted from Julie Chuong"
date: "10/28/2022"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/RdataFiles/Cov_JC/")

library(tidyverse)
library(docstring) #like docstring in python for R
library(ggplot2)

ggplot2::theme_set(theme_classic())

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"),
                       x = c(1:17))

```

## Load Functions

Original script from Gresham lab: 

```{r}
#Purpose: Estimate GAP1 copy numbers and CNV breakpoints from read depth of WGS data of isolated clones from the #EE_GAP1_ArchMuts_2021. These clones were isolated from generation 79 (g79).

#Upstream of this, I ran the mapcheck.sh bash script
#which outputted `{NAME}_RD.txt` file. Column headers are not there but are chromosome number, coordinate, and read depth.

#takes output from samtools depth
#`samtools depth -a ${bam} > ${ID}_RD.txt`

#Set up working directory and load packages
#setwd("~/Documents/WGS_clones_g79")


#Note, the ref sequence used for alignment and mapping was
# /scratch/work/cgsb/genomes/In_house/Fungi/Saccharomyces_cerevisiae/Gresham/GCF_000146045.2_R64_GAP1/GCF_000146045.2_R64_genomic_GAP1.fna which contains the GFP KanMX GAP1 CNV reporter (JC manually searched for the sequences to verify)
# Coordinates in this R script use this ref sequence (i think)

#### functions ###

get_rd_data = function(x) {
  #' This function reads in a file, names columns, removes the mitochondrial chromosome, and changes chromosome names from RefSeq chromosomes to numeric
  #' @param x {name}_RD.txt file outputted from samtools depth -a ${bam} > {NAME}_RD.txt
  #' @return a tibble with three columns - chromosome number, coordinate, and read depth
  #' @example DGY2071_read_data = get_rd_data("DGY2017_RD.txt")
  rd <- read.table(x)
  names(rd) <- c("chromo", "coordinate", "depth")
  rd <- rd[rd$chromo !="NC_001224.1",] #mitochondrial
  #change chromosome names
  rd <- rd %>% mutate(chromosome = case_when(chromo == 'NC_001133.9' ~ 1,
                                             chromo == 'NC_001134.8' ~ 2,
                                             chromo == 'NC_001135.5' ~ 3,
                                             chromo == 'NC_001136.10' ~ 4,
                                             chromo == 'NC_001137.3' ~ 5,
                                             chromo == 'NC_001138.5' ~ 6,
                                             chromo == 'NC_001139.9' ~ 7,
                                             chromo == 'NC_001140.6' ~ 8,
                                             chromo == 'NC_001141.2' ~ 9,
                                             chromo == 'NC_001142.9' ~ 10,
                                             chromo == 'NC_001143.9' ~ 11,
                                             chromo == 'NC_001144.5' ~ 12,
                                             chromo == 'NC_001145.3' ~ 13,
                                             chromo == 'NC_001146.8' ~ 14,
                                             chromo == 'NC_001147.6' ~ 15,
                                             chromo == 'NC_001148.4' ~ 16)) %>%
    select(chromosome, coordinate, depth)
  return(rd)
}

#' The next set of functions calculate the relative copy number at GAP1 and other specified genes
#' Genes are specified by giving the coordinates from the reference sequence
#' @param x the tibble returned from get_rd_data()
#'
#' @return numeric; copy number at specified gene relative to mean whole genome coverage
#' @export
#'
#' @examples get_gap1_reporter_copies(DGY2071_read_data)

get_gap1_reporter_copies = function(x) {
  #coordinates are adjusted to take into account the PrACT1-GFP-KanMX
  mean((x$depth[x$chromosome=="11" & x$coordinate>=513946 & x$coordinate<=520246])/mean(x$depth))
}

#this function gets relative GFP copy number
get_GFP_copies = function(x){
  #coordinates of GFP (717bp) located upstream of GAP1
  mean((x$depth[x$chromosome=="11" & x$coordinate>=514934 & x$coordinate<=515650])/mean(x$depth))

}

#this function gets the relative copy number at GAP1
get_gap1_copies = function(x) {
  #coordinates are GAP1 CDS location our reference sequence GCF_000146045.2_R64_genomic_GAP1.fna
  mean((x$depth[x$chromosome=="11" & x$coordinate>=518438 & x$coordinate<=520246])/mean(x$depth))
}

#this function gets the relative copy number at rDNA
get_rDNA_copies = function(x) {
  mean((x$depth[x$chromosome=="12" & x$coordinate>=451000 & x$coordinate<=470000])/mean(x$depth))
}

#this function gets the relative copy number at DUR3
get_cup1_copies = function(x) {
  mean((x$depth[x$chromosome=="8" & x$coordinate>=212535 & x$coordinate<=212720])/mean(x$depth))
}

#this function gets the relative copy number at HXT 6/7
get_hxt_copies = function(x) {
  mean((x$depth[x$chromosome=="4" & x$coordinate>=1154216 & x$coordinate<=1161320])/mean(x$depth))
}


# A function that generates a table with all read depth and copy numbers for genes of interest.
# Genes of interest can be added in the mutate() argument.

make_summary_table = function(x, name) {
  #' this function generates a table with all read depth and copy numbers for genes of interest
  #' genes of interest can be added in the mutate() argument
  #' @param x should be tibble outputted from get_rd_data()
  #' @param name should be a string stating the sample name
  #'
  #' @return Tibble with read depth and relative copy numbers of each gene of interest for the sample
  #'
  #' @examples DGY2071_summary_table = make_summary_table(DGY2071_read_data, "DGY2071"))
  summary = tibble(get_gap1_copies(x))
  names(summary) = 'gap1_copies'
  summary = summary %>%
    mutate(sample = name,
           GFP_copies = get_GFP_copies(x),
           gap1_reporter_copies = get_gap1_reporter_copies(x),
           rDNA_copies = get_rDNA_copies(x),
           cup1_copies = get_cup1_copies(x),
           hxt_copies = get_hxt_copies(x),
           mean_rd = mean(x$depth), median_rd = median(x$depth), sd_rd = sd(x$depth),
           min_rd = min(x$depth), max_rd = max(x$depth)) %>%
    relocate(sample)
}


#This function makes a plots of read depth normalized to chromosome 11.
#This lets us see GAP1 CNVs.

make_rd_plot_chr11 = function(df, output, dir) {
  #' this function makes plots read depth normalized to chromosome 11 read depth
  #'
  #' @param df is a dataframe of read data,
  #' @param output name of outputted plot file
  #' @param dir is the directory in which to save it
  #'
  #' @return save a .tiff plot read depth normalized to chromosome 11
  #'
  #' @examples
  #check to see if the directory exists and if it doesn't, make it
  if(dir.exists(dir) == FALSE) {
    dir.create(dir)
  }
  #get chromosome 11 only
  chr11 <- df[which(df$chromosome == 11),]
  avg11 <- mean(chr11$depth)
  linez <- runmed(chr11$depth/avg11,25001)
  newdata <- cbind(chr11, linez)
  newdata <- data.frame(newdata)
  #downsample for ease of plotting
  chr11plot <- newdata[sample(nrow(newdata), 20000),]
  #plot
  plot <- ggplot(chr11plot, aes(x=chr11plot[,2],y=chr11plot[,3]/avg11)) +
    geom_point(size = .25) +
    theme_classic() +
    scale_y_continuous(expand = c(0, 0), 'Relative Depth (compared to mean chr11)', limits=c(-0.5, 6)) +
    scale_x_continuous(expand = c(0, 0), limits=c(5000,655000), breaks = scales::pretty_breaks(n = 5), "Position on Chromosome 11 (kb)") +
    geom_line(aes(x= chr11plot[,2], y=chr11plot[,4]), color = "red")
  ggsave(filename = paste0(output, "_rd_chr11_plot.tiff"), plot = plot, path = dir)
}


#This function makes read depth plots for chromosome 11 normalized to the entire genome.
#I think this lets us detect aneuploidies (gains/losses) of chromosome 11.

make_rd_plot_genome = function(df, output, dir) {

  #this function makes read depth plots for chromosome 11 normalized to the entire genome
  #' Title
  #'
  #' @param df
  #' @param output
  #' @param dir
  #'
  #' @return
  #' @export
  #'
  #' @examples
  #' #df is a dataframe of rd, dir is the directory in which to save it
  #check to see if the directory exists and if it doesn't, make it
  if(dir.exists(dir) == FALSE) {
    dir.create(dir)
  }
  #get mean rd for the genome
  avg <- mean(df$depth)
  #get chromosome 11 only
  chr11 <- df[which(df$chromosome == 11),]
  linez <- runmed(chr11$depth/avg,25001)
  newdata <- cbind(chr11, linez)
  newdata <- data.frame(newdata)
  #downsample for ease of plotting
  chr11plot <- newdata[sample(nrow(newdata), 20000),]
  #plot
  plot <- ggplot(chr11plot, aes(x=chr11plot[,2],y=chr11plot[,3]/avg)) +
    geom_point(size = .25) +
    theme_classic() +
    scale_y_continuous(expand = c(0, 0), 'Relative Depth (compared to mean genome)', limits=c(-0.5, 6)) +
    scale_x_continuous(expand = c(0, 0), limits=c(5000,655000), breaks = scales::pretty_breaks(n = 5), "Position on Chromosome 11 (kb)") +
    geom_line(aes(x= chr11plot[,2], y=chr11plot[,4]), color = "red")
  ggsave(filename = paste0(output, "_rd_genome_plot.tiff"), plot = plot, path = dir)
}


#This function makes read depth plots of Chromosome 4 normalized to the entire genome
#Let's us detect CNVs of the hxt6/7 gene.
make_rd_plot_genome_hxt = function(df, output, dir) {
  #df is a dataframe of rd, dir is the directory in which to save it
  #check to see if the directory exists and if it doesn't, make it
if(dir.exists(dir) == FALSE) {
  dir.create(dir)
}
#get mean rd for the genome
avg <- mean(df$depth)
#get chromosome 4 only
chr4 <- df[which(df$chromosome == 4),]
linez <- runmed(chr4$depth/avg,25001)
newdata <- cbind(chr4, linez)
newdata <- data.frame(newdata)
#downsample for ease of plotting
chr4plot <- newdata#[sample(nrow(newdata), 20000),]
#plot
plot <- ggplot(chr4plot, aes(x=chr4plot[,2],y=chr4plot[,3]/avg)) +
  geom_point(size = .25) +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0), 'Relative Depth (compared to mean genome)', limits=c(-0.5, 6)) +
  scale_x_continuous(expand = c(0, 0), limits=c(1155000,1164000), breaks = scales::pretty_breaks(n = 5), "Position on Chromosome 4 (kb)") +
  geom_line(aes(x= chr4plot[,2], y=chr4plot[,4]), color = "red")
ggsave(filename = paste0(output, "_hxt67_rd_genome_plot.tiff"), plot = plot, path = dir)
}

```

```{r, eval = FALSE}
##### run the functions #####

#get all files
files = list.files(path = './', pattern = '*RD.txt')

#perform get_rd_data() in all files
clones_rd_data=map(paste0('./', files), get_rd_data)

#make one big summary table
summary_data_list=map2(clones_rd_data, files, ~make_summary_table(.x, str_sub(.y, 1, -8)))
summary_data_table = bind_rows(summary_data_list)

#write file out, change name of outfile in needed
write_csv(summary_data_table, file = 'cnv_g79clones_rd_summary.csv') #Write a name for outfile.

#make read depth plots relative on chromosome 11 and whole genome
walk2(clones_rd_data, files, ~make_rd_plot_chr11(.x, str_sub(.y, 1, -8), 'chr11_rd_plots'))
walk2(clones_rd_data, files, ~make_rd_plot_genome(.x, str_sub(.y, 1, -8), 'genome_rd_plots'))

####################################
#CB Testing
####################################

read.table(files[1])
ggplot(summary_data_table, aes(x = sample, y = mean_rd)) + geom_point()

```

## Run this to loop through for Fluconazole samples

```{r, eval = FALSE}
CB_SummaryList <- list()

for(i in 1:8){
  files = list.files(path = './', pattern = paste('*_', i, '.bam_RD.txt', sep = ""))

  #perform get_rd_data() in all files
  clones_rd_data=map(paste0('./', files), get_rd_data)

  #make one big summary table
  summary_data_list=map2(clones_rd_data, files, ~make_summary_table(.x, str_sub(.y, 1, -8)))
  summary_data_table = bind_rows(summary_data_list)
  summary_data_table$shortsample <- i
  
  CB_SummaryList[[i]] <- summary_data_table
}

CB_SL_Fluconazole <- bind_rows(CB_SummaryList)

samplekey <- data.frame(shortsample = 1:8, Selection = c("Fluconazole", "Fluconazole", "Fluconazole", "Dilute", "Dilute", "Dilute", "Fluconazole", "Fluconazole"))

CB_SL_Fluconazole %>% left_join(samplekey, by = "shortsample") -> CB_SL_Fluconazole

save(CB_SL_Fluconazole, file = "CB_SL_Fluconazole.Rdata")
```

Plotting these

```{r}
load("CB_SL_Fluconazole.Rdata")

chroms <- gsub("HGVMVDRX2.REF_", "", CB_SL_Fluconazole$sample)
chroms <- gsub("_1.bam", "", chroms)
chroms <- gsub("_2.bam", "", chroms)
chroms <- gsub("_3.bam", "", chroms)
chroms <- gsub("_4.bam", "", chroms)
chroms <- gsub("_5.bam", "", chroms)
chroms <- gsub("_6.bam", "", chroms)
chroms <- gsub("_7.bam", "", chroms)
chroms <- gsub("_8.bam", "", chroms)

CB_SL_Fluconazole$chroms <- chroms

#Looks like chr 12 has higher coverage probably because rDNA
CB_SL_Fluconazole %>% ggplot(aes(x = mean_rd, y = chroms, color = as.factor(shortsample))) + geom_boxplot() + geom_point(position = position_jitterdodge())

CB_SL_Fluconazole %>% ggplot(aes(x = mean_rd, y = cup1_copies, color = as.factor(shortsample))) + geom_point(size = 2)

```



```{r}
load("CB_SL_Fluconazole.Rdata")

CB_SL_Fluconazole %>% ggplot(aes(x = mean_rd, y = cup1_copies, color = as.factor(Selection))) + geom_point(size = 2)

CB_SL_Fluconazole %>% ggplot(aes(x= cup1_copies, y = Selection, fill = as.factor(Selection))) + 
  geom_boxplot(size = 1, alpha = 0.2) + 
  geom_point(position = position_jitterdodge(), size = 3) +
  ggtitle("Distribution of CUP1 copy by Selection Type")

```

## Do the same thing for CuSO4 samples

```{r, eval = FALSE}
CB_SummaryList <- list()

myfilenames <- c("UnselectedC", "SelectedC", "UnselectedA", "SelectedA")
for(i in myfilenames){
  files = list.files(path = './', pattern = paste('HNGLVDRXY_n01_CuSO4_CSSI_', i, '.bqsr.output.bam_RD.txt', sep = ""))

  #perform get_rd_data() in all files
  clones_rd_data=map(paste0('./', files), get_rd_data)

  #make one big summary table
  summary_data_list=map2(clones_rd_data, files, ~make_summary_table(.x, str_sub(.y, 1, -8)))
  summary_data_table = bind_rows(summary_data_list)
  summary_data_table$shortsample <- i
  
  CB_SummaryList[[i]] <- summary_data_table
}

CB_SL_CuSO4 <- bind_rows(CB_SummaryList)

samplekeyc <- data.frame(shortsample = c("UnselectedC", "SelectedC", "UnselectedA", "SelectedA"), Selection = c("Dilute", "CuSO4", "Dilute", "CuSO4"))
CB_SL_CuSO4 %>% left_join(samplekeyc, by = "shortsample") -> CB_SL_CuSO4

save(CB_SL_CuSO4, file = "CB_SL_CuSO4.Rdata")

```

```{r}
load("CB_SL_CuSO4.Rdata")

CB_SL_CuSO4 %>% ggplot(aes(x = mean_rd, y = cup1_copies, color = as.factor(Selection))) + geom_point(size = 2)
CB_SL_CuSO4 %>% ggplot(aes(x= cup1_copies, y = Selection, fill = as.factor(Selection))) + 
  geom_boxplot(size = 1, alpha = 0.2) + 
  geom_point(position = position_jitterdodge(), size = 3) +
  ggtitle("Distribution of CUP1 copy by Selection Type")

```

## Combine Both

### CUP-1 Copy Number

```{r}
load("CB_SL_Fluconazole.Rdata")
load("CB_SL_CuSO4.Rdata")

CB_SL_Fluconazole$Dataset <- "Fluconazole"
CB_SL_CuSO4$Dataset <- "CuSO4"

All_CSSI <- rbind(CB_SL_CuSO4, CB_SL_Fluconazole)
All_CSSI$Selection <- factor(All_CSSI$Selection, levels = c("Dilute", "CuSO4", "Fluconazole"))

All_CSSI %>% ggplot(aes(y= cup1_copies, x = Selection, fill = as.factor(Selection))) + 
  geom_boxplot(size = 1, alpha = 0.2) + 
  geom_point(aes(shape = Dataset), position = position_jitterdodge(), size = 3) +
  scale_fill_manual(values = c("gray", "#345F6F", "firebrick"))+
  ggtitle("Distribution of CUP1 copy by Selection Type")

All_CSSI %>% ggplot(aes(x = mean_rd, y = cup1_copies, color = as.factor(Selection))) + 
  geom_point(size = 3, alpha = 0.6) +
  scale_color_manual(values = c("gray", "#345F6F", "firebrick"))+
  ggtitle("CUP1 copy vs Mean RD by Selection Type")

All_CSSI %>% ggplot(aes(x = median_rd, y = cup1_copies, color = as.factor(Selection))) + 
  geom_point(size = 3, alpha = 0.6) +
  scale_color_manual(values = c("gray", "#345F6F", "firebrick"))+
  ggtitle("CUP1 copy vs Median RD by Selection Type")


```

### rDNA Copy Number

```{r}

All_CSSI %>% ggplot(aes(y= rDNA_copies, x = Selection, fill = as.factor(Selection))) + 
  geom_boxplot(size = 1, alpha = 0.2) + 
  geom_point(aes(shape = Dataset), position = position_jitterdodge(), size = 3) +
  scale_fill_manual(values = c("gray", "#345F6F", "firebrick"))+
  ggtitle("Distribution of rDNA copy by Selection Type")

All_CSSI %>% ggplot(aes(x = mean_rd, y = rDNA_copies, color = as.factor(Selection), shape = Dataset)) + 
  geom_point(size = 3, alpha = 0.6) +
  scale_color_manual(values = c("gray", "#345F6F", "firebrick"))

All_CSSI %>% ggplot(aes(x = median_rd, y = rDNA_copies, color = as.factor(Selection), shape = Dataset)) + 
  geom_point(size = 3, alpha = 0.6) +
  scale_color_manual(values = c("gray", "#345F6F", "firebrick"))
```

### GAP-1 Copy Number

```{r}
All_CSSI %>% ggplot(aes(y= gap1_copies, x = Selection, fill = as.factor(Selection))) + 
  geom_boxplot(size = 1, alpha = 0.2) + 
  geom_point(aes(shape = Dataset), position = position_jitterdodge(), size = 3) +
  scale_fill_manual(values = c("gray", "#345F6F", "firebrick"))+
  ggtitle("Distribution of GAP1 copy by Selection Type")

All_CSSI %>% ggplot(aes(x = mean_rd, y = gap1_copies, color = as.factor(Selection), shape = Dataset)) + 
  geom_point(size = 3, alpha = 0.6) +
  scale_color_manual(values = c("gray", "#345F6F", "firebrick"))
All_CSSI %>% ggplot(aes(x = median_rd, y = gap1_copies, color = as.factor(Selection), shape = Dataset)) + 
  geom_point(size = 3, alpha = 0.6) +
  scale_color_manual(values = c("gray", "#345F6F", "firebrick"))
```

### HXT Copy Number

```{r}
All_CSSI %>% ggplot(aes(y= hxt_copies, x = Selection, fill = as.factor(Selection))) + 
  geom_boxplot(size = 1, alpha = 0.2) + 
  geom_point(aes(shape = Dataset),position = position_jitterdodge(), size = 3) +
  scale_fill_manual(values = c("gray", "#345F6F", "firebrick"))+
  ggtitle("Distribution of HXT copy by Selection Type")

All_CSSI %>% ggplot(aes(x = mean_rd, y = hxt_copies, color = as.factor(Selection), shape = Dataset)) + 
  geom_point(size = 3, alpha = 0.6) +
  scale_color_manual(values = c("gray", "#345F6F", "firebrick"))
All_CSSI %>% ggplot(aes(x = median_rd, y = hxt_copies, color = as.factor(Selection), shape = Dataset)) + 
  geom_point(size = 3, alpha = 0.6) +
  scale_color_manual(values = c("gray", "#345F6F", "firebrick"))
```

## Overall Copy Number

### Fluconazole Experiment

```{r, eval = FALSE}
copydata <- list()
for(i in 1:8){
  files = list.files(path = './', pattern = paste('*_', i, '.bam_RD.txt', sep = ""))
  clones_rd_data=map(paste0('./', files), get_rd_data)
  crd <- bind_rows(clones_rd_data)
  crd$sample <- i
  copydata[[i]] <- crd
  
}

cd_fluconazole <- bind_rows(copydata)

save(cd_fluconazole, file = "cd_fluconazole.rdata")
```



```{r, eval = FALSE}
## Plot this
load("cd_fluconazole.rdata")

for(i in 1){
  subcd <- subset(cd_fluconazole, sample == i)
  cpplot <- ggplot(subcd, aes(x = coordinate, y = depth)) + geom_line() + facet_grid(rows = "chromosome", scales = "free") + ggtitle(paste("Coverage of Fluconazole Sample", i, sep = " "))
  print(cpplot)
}

unique(cd_fluconazole$sample)
```

### CuSO4 Experiment

```{r, eval = FALSE}
copydata_c <- list()
myfilenames <- c("UnselectedC", "SelectedC", "UnselectedA", "SelectedA")
for(i in myfilenames){
  files = list.files(path = './', pattern = paste('HNGLVDRXY_n01_CuSO4_CSSI_', i, '.bqsr.output.bam_RD.txt', sep = ""))

  clones_rd_data=map(paste0('./', files), get_rd_data)
  crd <- bind_rows(clones_rd_data)
  crd$sample <- i
  copydata_c[[i]] <- crd
  
}

cd_cuso4 <- bind_rows(copydata_c)

save(cd_cuso4, file = "cd_cuso4.rdata")
```



```{r, eval = FALSE}
## Plot this
load("cd_cuso4.rdata")

myfilenames <- c("UnselectedC", "SelectedC", "UnselectedA", "SelectedA")

for(i in myfilenames){
  subcd <- subset(cd_cuso4, sample == i)
  cpplot <- ggplot(subcd, aes(x = coordinate, y = depth)) + geom_line() + facet_grid(rows = "chromosome", scales = "free") + ggtitle(paste("Coverage of CuSO4 Sample", i, sep = " "))
  print(cpplot)
}

```

# Find copy for each peak

Make a new function that expands the peak out from an input file and then finds copy of that in a summary

```{r}

get_peak_copies = function(x, peakvector) {
  chrom = peakvector[1]
  peak=as.numeric(peakvector[2])
  #coordinates are GAP1 CDS location our reference sequence GCF_000146045.2_R64_genomic_GAP1.fna
  mean((x$depth[x$chromosome==chrom & x$coordinate>=(peak-1000) & x$coordinate<=(peak+1000)])/mean(x$depth))
}

make_summary_table_peaks = function(x, name, peakvector) {
  #' this function generates a table with all read depth and copy numbers for genes of interest
  #' genes of interest can be added in the mutate() argument
  #' @param x should be tibble outputted from get_rd_data()
  #' @param name should be a string stating the sample name
  #'
  #' @return Tibble with read depth and relative copy numbers of each gene of interest for the sample
  #'
  #' @examples DGY2071_summary_table = make_summary_table(DGY2071_read_data, "DGY2071"))
  summary = tibble(get_peak_copies(x, peakvector))
  names(summary) = 'gap1_copies'
  summary = summary %>%
    mutate(sample = name,
           PeakCopies = get_peak_copies(x, peakvector),
           rDNA_copies = get_rDNA_copies(x),
           hxt_copies = get_hxt_copies(x),
           mean_rd = mean(x$depth), median_rd = median(x$depth), sd_rd = sd(x$depth),
           min_rd = min(x$depth), max_rd = max(x$depth)) %>%
    relocate(sample)
}

```


## Making a brand new function from these lists

```{r}

peakdf <- data.frame(name = c("GAP1","rDNA","Cup1","HXT","Peak7"), 
                     x = c(11,12,8,4,7), 
                     coordinate = c(518438, 451000,212535, 1154216, 497059))

################################################################################

make_summary_table_hxt = function(x, name, peakvector) {
  #' this function generates a table with all read depth and copy numbers for genes of interest
  #' genes of interest can be added in the mutate() argument
  #' @param x should be tibble outputted from get_rd_data()
  #' @param name should be a string stating the sample name
  #'
  #' @return Tibble with read depth and relative copy numbers of each gene of interest for the sample
  #'
  #' @examples DGY2071_summary_table = make_summary_table(DGY2071_read_data, "DGY2071"))
  summary = tibble(get_hxt_copies(x))
  names(summary) = 'hxt_copies'
  summary = summary %>%
    mutate(sample = name, 
           hxt_copies = get_hxt_copies(x),
           rDNA_copies = get_rDNA_copies(x),
           mean_rd = mean(x$depth), median_rd = median(x$depth), sd_rd = sd(x$depth),
           min_rd = min(x$depth), max_rd = max(x$depth)) %>%
    relocate(sample)
}
```

```{r, eval = FALSE}
################################################################################

CB_SummaryList <- list()
CB_PeaksList <- list()

for(i in 1:8){
  files = list.files(path = './', pattern = paste('*_', i, '.bam_RD.txt', sep = ""))

  #perform get_rd_data() in all files
  clones_rd_data=map(paste0('./', files), get_rd_data)

  #make one big summary table
  summary_data_list=map2(clones_rd_data, files, ~make_summary_table_hxt(.x, str_sub(.y, 1, -8)))
  summary_data_table = bind_rows(summary_data_list)
  summary_data_table$shortsample <- i
  
  CB_SummaryList[[i]] <- summary_data_table
  
  #make the peaks table for each also
  for(j in 1:length(peakdf$x)){
    x = clones_rd_data[[peakdf$x[j]]]
    peakdf$coverage[j] <- mean((x$depth[x$chromosome==peakdf$x[j] & x$coordinate>=(peakdf$coordinate[j]-1000) & x$coordinate<=(peakdf$coordinate[j]+1000)])/mean(x$depth))
    peakdf$coverage_total[j] <- mean((x$depth[x$chromosome==peakdf$x[j] & x$coordinate>=(peakdf$coordinate[j]-1000) & x$coordinate<=(peakdf$coordinate[j]+1000)])/mean(clones_rd_data$depth))
    
  }
  peakdf$shortsample <- i
  CB_PeaksList[[i]] <- peakdf
}

CB_SL_Fluconazole_base <- bind_rows(CB_SummaryList)
CB_SL_Fluconazole_peaks <- bind_rows(CB_PeaksList)

samplekey <- data.frame(shortsample = 1:8, Selection = c("Fluconazole", "Fluconazole", "Fluconazole", "Dilute", "Dilute", "Dilute", "Fluconazole", "Fluconazole"))

CB_SL_Fluconazole_base %>% left_join(samplekey, by = "shortsample") -> CB_SL_Fluconazole_base

head(CB_SL_Fluconazole_base)
head(CB_SL_Fluconazole_peaks)
#CB_SL_Fluconazole_peaks %>% left_join(peakdf) -> CB_SL_Fluconazole_peaks


```

```{r, eval = FALSE}
CB_SL_Fluconazole_peaks %>% ggplot(aes(x = name, y = coverage, color = as.factor(shortsample))) + geom_point()

CB_SL_Fluconazole_peaks$chromosomes <- factor(CB_SL_Fluconazole_peaks$chromosomes, levels = ChromKey$chromosomes)
CB_SL_Fluconazole_peaks$experiment <- gsub(".*CSSI_","", CB_SL_Fluconazole_peaks$name)

CB_SL_Fluconazole_peaks %>% ggplot(aes(x = coordinate, y = coverage, color = shortsample, alpha = experiment)) + geom_point(size = 2) + theme(legend.position = "bottom", axis.text.x = element_blank()) + facet_grid(~chromosomes) + scale_alpha_discrete(range = c(1, 0.2)) + ggtitle("Coverage by Chromosome")
```
```{r}
CSSI_peaks <- read.csv("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/Genes_CSSI_Experiments.csv")

CSSI_peaks %>% transmute(name = paste(CHROM, gsub("peak.", "", Type), Experiment, sep = "."),
                         chromosomes = CHROM,
                         coordinate = POS) %>% merge(., ChromKey, by = "chromosomes") -> CSSI_peak_table

```

### Copper Sulfate Peak IDs

```{r, eval = FALSE}

peakdf_c <- CSSI_peak_table
  
copydata_c <- list()
CB_PeaksList <- list()

myfilenames <- c("UnselectedC", "SelectedC", "UnselectedA", "SelectedA")
for(i in myfilenames){
  files = list.files(path = './', pattern = paste('HNGLVDRXY_n01_CuSO4_CSSI_', i, '.bqsr.output.bam_RD.txt', sep = ""))

  clones_rd_data=map(paste0('./', files), get_rd_data)
  crd <- bind_rows(clones_rd_data)
  crd$sample <- i
  copydata_c[[i]] <- crd
  
  #make the peaks table for each also
  for(j in 1:length(peakdf_c$x)){
    x = crd[crd$chromosome == peakdf_c$x[j],]
    
    peakdf_c$coverage[j] <- mean((x$depth[x$chromosome==peakdf_c$x[j] & x$coordinate>=(peakdf_c$coordinate[j]-1000) & x$coordinate<=(peakdf_c$coordinate[j]+1000)])/mean(x$depth))
    
    peakdf_c$coverage_total[j] <- mean((x$depth[x$chromosome==peakdf_c$x[j] & x$coordinate>=(peakdf_c$coordinate[j]-1000) & x$coordinate<=(peakdf_c$coordinate[j]+1000)])/mean(crd$depth))
    
  }
  peakdf_c$shortsample <- i
  CB_PeaksList[[i]] <- peakdf_c
}

cd_cuso4 <- bind_rows(copydata_c)
peakdf_cuso4 <- bind_rows(CB_PeaksList)

save(peakdf_cuso4, file = "peakdf_cuso4.Rdata")

```

Plotting these

```{r}
load("peakdf_cuso4.Rdata")
cuso4_colors <- c("blue", "blue", "black", "black")

peakdf_cuso4$chromosomes <- factor(peakdf_cuso4$chromosomes, levels = ChromKey$chromosomes)
peakdf_cuso4$experiment <- gsub(".*CSSI_","", peakdf_cuso4$name)

peakdf_cuso4 %>% ggplot(aes(x = coordinate, y = coverage, color = shortsample, alpha = experiment)) + geom_point(size = 2) + theme(legend.position = "bottom", axis.text.x = element_blank()) + facet_grid(~chromosomes) + 
    scale_color_manual(values = cuso4_colors) + 
  scale_alpha_discrete(range = c(1, 0.2)) + 
  ggtitle("Coverage by Chromosome")

peakdf_cuso4 %>% ggplot(aes(x = coordinate, y = coverage_total, color = shortsample, alpha = experiment)) + 
  geom_point(size = 2) + theme(legend.position = "bottom", axis.text.x = element_blank()) + 
  facet_grid(~chromosomes) + 
  scale_alpha_discrete(range = c(1, 0.2)) + 
  scale_color_manual(values = cuso4_colors) + 
  ggtitle("Coverage by Total")
```

## Fluconazole peak IDs

```{r, eval = FALSE}
peakdf_f <- CSSI_peak_table

CB_PeaksList_f <- list()
CB_SummaryList <- list()

for(i in 1:8){
  files = list.files(path = './', pattern = paste('*_', i, '.bam_RD.txt', sep = ""))

  #perform get_rd_data() in all files
  clones_rd_data=map(paste0('./', files), get_rd_data)
  
  total_file <- bind_rows(clones_rd_data)

  #make one big summary table
  summary_data_list=map2(clones_rd_data, files, ~make_summary_table_hxt(.x, str_sub(.y, 1, -8)))
  summary_data_table = bind_rows(summary_data_list)
  summary_data_table$shortsample <- i
  
  CB_SummaryList[[i]] <- summary_data_table
  
  #make the peaks table for each also
  for(j in 1:length(peakdf_f$x)){
    x = clones_rd_data[[peakdf_f$x[j]]]
    peakdf_f$coverage[j] <- mean((x$depth[x$chromosome==peakdf_f$x[j] & x$coordinate>=(peakdf_f$coordinate[j]-1000) & x$coordinate<=(peakdf_f$coordinate[j]+1000)])/mean(x$depth))
    
    peakdf_f$coverage_total[j] <- mean((x$depth[x$chromosome==peakdf_f$x[j] & x$coordinate>=(peakdf_f$coordinate[j]-1000) & x$coordinate<=(peakdf_f$coordinate[j]+1000)])/mean(total_file$depth))
    
  }
  peakdf_f$shortsample <- i
  CB_PeaksList_f[[i]] <- peakdf_f
}

CB_SL_Fluconazole_base <- bind_rows(CB_SummaryList)
CB_SL_Fluconazole_peaks <- bind_rows(CB_PeaksList_f)

samplekey <- data.frame(shortsample = 1:8, Selection = c("Fluconazole", "Fluconazole", "Fluconazole", "Dilute", "Dilute", "Dilute", "Fluconazole", "Fluconazole"))

save(CB_SL_Fluconazole_base, file = "CB_SL_Fluconazole_base.Rdata")
save(CB_SL_Fluconazole_peaks, file = "CB_SL_Fluconazole_peaks.Rdata")
```

Plotting

```{r}

load("CB_SL_Fluconazole_peaks.Rdata")

fluc_colors <- c("red", "red", "red", "black", "black", "black", "red", "red")
CB_SL_Fluconazole_peaks$chromosomes <- factor(peakdf_cuso4$chromosomes, levels = ChromKey$chromosomes)
CB_SL_Fluconazole_peaks$experiment <- gsub(".*CSSI_","", peakdf_cuso4$name)

CB_SL_Fluconazole_peaks %>% ggplot(aes(x = coordinate, y = coverage, color = as.factor(shortsample), alpha = experiment)) + geom_point(size = 2) + theme(legend.position = "bottom", axis.text.x = element_blank()) + facet_grid(~chromosomes) + 
    scale_color_manual(values = fluc_colors) + 
scale_alpha_discrete(range = c(0.2, 1)) + ggtitle("Coverage by Chromosome")

CB_SL_Fluconazole_peaks %>% ggplot(aes(x = coordinate, y = coverage_total, color = as.factor(shortsample), alpha = experiment)) + geom_point(size = 2) + theme(legend.position = "bottom", axis.text.x = element_blank()) + facet_grid(~chromosomes) + 
  scale_color_manual(values = fluc_colors) + 
  scale_alpha_discrete(range = c(0.2, 1)) + ggtitle("Coverage by Total")
```

## Combining these

```{r}
load("CB_SL_Fluconazole_peaks.Rdata")
load("peakdf_cuso4.Rdata")

CB_SL_Fluconazole_peaks$Dataset <- "Fluconazole"
peakdf_cuso4$Dataset <- "CuSO4"

AllPeaks_CSSI <- rbind(CB_SL_Fluconazole_peaks, peakdf_cuso4)

AllPeaks_CSSI %>% ggplot(aes(x = name, y = coverage_total, color = shortsample, shape = Dataset, alpha = experiment)) + geom_point() +   scale_color_manual(values = c(fluc_colors, cuso4_colors)) + ggtitle("Fluconazole QTLs") + 
    theme(legend.position = "blank", axis.text.x = element_text(angle = 90,hjust=1)) 

AllPeaks_CSSI %>% ggplot(aes(x = name, y = coverage_total, color = shortsample, shape = Dataset, alpha = experiment)) + geom_point() +   scale_color_manual(values = c(fluc_colors, cuso4_colors)) + ggtitle("CuSO4 QTLs") + 
  theme(legend.position = "blank", axis.text.x = element_text(angle = 90,hjust=1)) + 
  scale_alpha_discrete(range = c(1, 0.2))

```

