---
title: "CSS Analysis"
author: "Cassandra Buzby"
date: "10/27/2022"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

library(cybrBSA)

ggplot2::theme_set(theme_light())
#ggplot2::theme_set(theme_bw())


CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47",
                  "#F26430", "#0A369D", "#7EA3CC")


library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")
# load("RdataFiles/SAUA_4-13-22.Rdata")
# load("RdataFiles/SCUC_4-13-22.Rdata")
# load("RdataFiles/bsa_glm_results.Rdata")

#rawData = "HGV.SortedCat.vcf.output.table"
```

## Convert Parental Alleles

```{r, eval = FALSE}
parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

head(parentSNPids)
```

## Using cybrBSA package

CSS I Fluconazole

```{r, eval = FALSE}
mydatatotest = "HGV.SortedCat.vcf.output.table"

CSSI_Fluc <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CSSI_Fluc$Dataset)

CSSI_Fluc$Bulk <- NA
CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq"] <- "Dilute"

CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "Fluconazole"

CSSI_Fluc$Parent <- NA
CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq"] <- "OakI"

CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "WineI"

CSSI_Fluc$Parent <- factor(CSSI_Fluc$Parent)
CSSI_Fluc$Bulk <- factor(CSSI_Fluc$Bulk)
CSSI_Fluc$ReadCount <- as.numeric(CSSI_Fluc$ReadCount)

head(CSSI_Fluc)

save(CSSI_Fluc, file ="CSSI_Fluc_counts.Rdata")

CSSI_Fluc_BSA <- foreach(i=unique(CSSI_Fluc$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Fluc, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Fluc_BSA, file = "CSSI_Fluc_BSA.Rdata")
```


CSS I CuSO4

```{r, eval = FALSE}
CuSO4gatk <- "Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"

CuSO4data <- cybrInputGATKTable(CuSO4gatk) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CuSO4data$Dataset)

CuSO4data$Bulk <- NA
CuSO4data$Bulk[CuSO4data$Dataset == "SelectedA" |
              CuSO4data$Dataset == "SelectedC"] <- "CuSO4"

CuSO4data$Bulk[CuSO4data$Dataset == "UnselectedA" |
              CuSO4data$Dataset == "UnselectedC"] <- "Dilute"

CuSO4data$Parent <- NA
CuSO4data$Parent[CuSO4data$Dataset == "SelectedA" |
              CuSO4data$Dataset == "UnselectedA"] <- "OakI"

CuSO4data$Parent[CuSO4data$Dataset == "SelectedC" |
              CuSO4data$Dataset == "UnselectedC"] <- "WineI"

CuSO4data$Parent <- factor(CuSO4data$Parent)
CuSO4data$Bulk <- factor(CuSO4data$Bulk)
CuSO4data$ReadCount <- as.numeric(CuSO4data$ReadCount)

head(CuSO4data)

save(CuSO4data, file = "CuSO4data_counts.Rdata")

CSSI_CuSO4_BSA <- foreach(i=unique(CuSO4data$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuSO4data, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_CuSO4_BSA, file = "CSSI_CuSO4_BSA.Rdata")

```


## Cycloheximide / Zeocin Data

```{r, eval = FALSE}
mydatatotest = "HKTMZDRX2.SortedCat.vcf.output.table"

CSSI_CycZeo <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CSSI_CycZeo$Dataset)

CSSI_CycZeo$Bulk <- NA
CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq"] <- "Cycloheximide"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "Dilute"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq"] <- "Zeocin"

CSSI_CycZeo$Parent <- NA
CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq"] <- "OakI"

CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "WineI"

CSSI_CycZeo$Parent <- factor(CSSI_CycZeo$Parent)
CSSI_CycZeo$Bulk <- factor(CSSI_CycZeo$Bulk)
CSSI_CycZeo$ReadCount <- as.numeric(CSSI_CycZeo$ReadCount)

head(CSSI_CycZeo)

save(CSSI_CycZeo, file = "CSSI_CycZeo_counts.Rdata")

```

## Compare dilute samples for all combinations

We would expect the "dilute" group for each to not have a strong effect, so (1) find the difference in the four for the CycZeo experiment, and then (2) find the overall effect for each replicate and (3) find the overall effect for all groups at once.

```{r}
load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

CSSI_CycZeo %>% filter(Bulk == "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts
CSSI_Fluc %>% filter(Bulk == "Dilute") %>% mutate(Study = "Fluc") -> FCounts
CuSO4data %>% filter(Bulk == "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts

DiluteData <- rbind(CZCounts,FCounts,CuCounts)

save(DiluteData, file = "DiluteData.Rdata")

DiluteData %>% group_by(CHROM, POS) %>% 
  summarise(CHROM = CHROM, POS = POS, studies = length(unique(Study))) %>% 
  distinct() %>% ungroup() -> DiluteDataCounts

DiluteDataCounts %>% filter(studies < 3) -> DiluteData_nonOverlaps

# ~8% of positions are not represented in ALL groups
dim(DiluteData_nonOverlaps)[1]/dim(DiluteData)[1]

#Filter out those which don't have all three
DiluteDataCounts %>% filter(studies == 3) %>% left_join(DiluteData) %>% ungroup() -> DiluteData_filtered

save(DiluteData_nonOverlaps, file = "DiluteData_nonOverlaps.Rdata")
```

```{r, eval = FALSE}
################################################################################
#Rewrote functions to accommodate
#unique(DiluteData$Study)
CSSI_Dilute_BSA <- foreach(i=unique(DiluteData_filtered$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_filtered, chr = i, formula = "PAllele ~ Study") 
}

#CSSI_Dilute_BSA %>% cybrSmoothBSAWindows(chr = "VIX")
CSSI_Dilute_BSA_s <- foreach(i=unique(CSSI_Dilute_BSA$CHROM), .combine=rbind) %dopar%{
  cybrSmoothBSAWindows(CSSI_Dilute_BSA, chr = i) 
}

i <- "I"

mylist <- list()
#Chromosomes 6 and 9 don't work for some reason?
for(i in unique(CSSI_Dilute_BSA$CHROM)){
  mylist[[i]] <- cybrSmoothBSAWindows(CSSI_Dilute_BSA, chr = i) 
}

CSSI_Dilute_BSA_s <- bind_rows(mylist)
CSSI_Dilute_BSA_s$CHROM <- factor(CSSI_Dilute_BSA_s$CHROM, levels = ChromKey$chromosomes)
save(CSSI_Dilute_BSA_s, file = "CSSI_Dilute_BSA_s.Rdata")

################################################################################
CZDilute_BSA <- foreach(i=unique(CZCounts$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CZCounts, chr = i, formula = "PAllele ~ Dataset") %>% cybrSmoothBSAWindows()
}

save(CZDilute_BSA, file = "CZDilute_BSA.Rdata")

```


Plot these results

```{r}
load("CSSI_Dilute_BSA_s.Rdata")
DiluteData_nonOverlaps %>% select(-studies) %>% mutate('(Intercept)_Zprime' = NA, StudyCycZeo_Zprime = NA, StudyFluc_Zprime = NA, MissingData = 0) -> fillingindata
CSSI_Dilute_BSA_s %>% select(CHROM, POS, '(Intercept)_Zprime', StudyCycZeo_Zprime, StudyFluc_Zprime) %>% mutate(MissingData = NA) %>% rbind(fillingindata) -> CSSI_Dilute_BSA_u

CSSI_Dilute_BSA_u %>% pivot_longer(c('(Intercept)_Zprime', StudyCycZeo_Zprime, StudyFluc_Zprime, MissingData), names_to = "Factor", values_to = "Zprime") %>% 
  filter(Factor != "(Intercept)_Zprime") %>% 
  filter(CHROM != "I") %>%
  ggplot(aes(x = POS, y = Zprime, color = Factor)) + geom_point(alpha = 0.4, size = 0.5) + facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + scale_color_manual(values = c("black", "violet", "red")) +
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("All Studies Compared (Dilute)")
```


```{r}
load("CZDilute_BSA.Rdata")

CZDilute_BSA %>% filter(CHROM != "I") %>% pivot_longer(c('(Intercept)_Zprime', 
                                DatasetHKTMWDRX2_n01_ODB.fastq_Zprime,
                                DatasetHKTMWDRX2_n01_WDA.fastq_Zprime,
                                DatasetHKTMWDRX2_n01_WDB.fastq_Zprime), names_to = "Factor", values_to = "Zprime") %>%
  mutate(FactorName = gsub("DatasetHKTMWDRX2_n01_", "", .$Factor)) %>%
  filter(FactorName != "(Intercept)_Zprime") %>%
  ggplot(aes(x = POS, y = Zprime, color = FactorName)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("CycZeo Dilute Samples Compared")
```


## Plotting actual allele frequencies for each

I need to know what the logistic regression is doing here, so I'm going to plot each of the dilute samples in a different color and see if there's a weird trend there. That should help separate the effects of replicates as well as batches.

Load in just in case:
```{r}
load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

####################### Dilute #################################################
CSSI_CycZeo %>% filter(Bulk == "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts
CSSI_Fluc %>% filter(Bulk == "Dilute") %>% mutate(Study = "Fluc") -> FCounts
CuSO4data %>% filter(Bulk == "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts

DiluteData <- rbind(CZCounts,FCounts,CuCounts)

DatasetKey <- data.frame(Dataset = unique(DiluteData$Dataset),
                         DataName = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

DiluteData$Dataset <- as.character(DiluteData$Dataset)
DiluteData %>% left_join(DatasetKey, by = "Dataset") -> DiluteLabeled

####################### Selected ###############################################

CSSI_CycZeo %>% filter(Bulk != "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts_s
CSSI_Fluc %>% filter(Bulk != "Dilute") %>% mutate(Study = "Fluc") -> FCounts_s
CuSO4data %>% filter(Bulk != "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts_s

SelectedData <- rbind(CZCounts_s,FCounts_s,CuCounts_s)

DatasetKey_s <- data.frame(Dataset = unique(SelectedData$Dataset),
                         DataName = c("CZ_OCA", "CZ_OCB", "CZ_OZA", "CZ_OZB","CZ_WCA", "CZ_WCB","CZ_WZA", "CZ_WZB", "F_OFB", "F_OFC", "F_OFD", "F_WFC", "Cu_SA", "Cu_SC"))

SelectedData$Dataset <- as.character(SelectedData$Dataset)
SelectedData %>% left_join(DatasetKey_s, by = "Dataset") -> SelectedLabeled

####################### Colors #################################################

```

Pull out the allele frequency of each

```{r}

load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

####################### Dilute #################################################
CSSI_CycZeo %>% filter(Bulk == "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts
CSSI_Fluc %>% filter(Bulk == "Dilute") %>% mutate(Study = "Fluc") -> FCounts
CuSO4data %>% filter(Bulk == "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts

DiluteData <- rbind(CZCounts,FCounts,CuCounts)

DatasetKey <- data.frame(Dataset = unique(DiluteData$Dataset),
                         DataName = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

DiluteData$Dataset <- as.character(DiluteData$Dataset)
DiluteData %>% left_join(DatasetKey, by = "Dataset") -> DiluteLabeled
 
DatasetKey <- data.frame(Dataset = unique(DiluteData$Dataset),
                         DataName = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

DiluteData$Dataset <- as.character(DiluteData$Dataset)
DiluteData %>% left_join(DatasetKey, by = "Dataset") -> DiluteLabeled

DiluteLabeled$DataName <- factor(DiluteLabeled$DataName, 
                                   levels = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

mylabelcolorsd <- c("#91171F","#5C95FF",
                     "black", "#966B9D", "#A94C4C", "#2E1C2B", 
                     "#F2B880", "#46B1C9", "#3E517A", "#CBBAED",
                     "#8C2155", "#F87060", "#183059", "lightblue")
names(mylabelcolorsd) <- levels(DiluteLabeled$DataName)

# DiluteLabeled %>% select(-AltRef_Allele) %>% 
#     pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
#     ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.2) + 
#   facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
#   theme_minimal() +
#       theme(legend.position = "bottom", axis.text.x=element_blank(),
#             axis.ticks.x=element_blank()) + 
#       xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Read Ratio by Dilute Dataset")
# 
# DiluteLabeled %>% filter(CHROM == "VII")  %>% select(-AltRef_Allele) %>% 
#     pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
#     ggplot(aes(x = POS, y = log(Wine/Oak), color = Study, group = DataName)) + geom_point(alpha = 0.2) + 
#   facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
#   theme_minimal() + scale_color_manual(values = c("skyblue", "violet", "red")) + 
#       theme(legend.position = "bottom", axis.text.x=element_blank(),
#             axis.ticks.x=element_blank()) + 
#       xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Read Ratio by Dilute Dataset")

DiluteLabeled %>% filter(CHROM == "VII") %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.8) + 
  facet_grid(~Study, scales = "free_x", space = "free_x") + ylim(-2,2) + 
  theme_minimal() + scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Chr VII Read Ratio by Dilute Dataset")


#Smooth this dilute measure

DiluteLabeled %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
  select(CHROM, POS, Study, DataName, Oak, Wine) %>%
  mutate(logRatio = log(Wine/Oak)) %>%
  arrange(CHROM, POS) -> DiluteLogRatio

DiluteLogRatio$logRatio[DiluteLogRatio$logRatio == "Inf" | DiluteLogRatio$logRatio == "-Inf"] <- NA

save(DiluteLogRatio, file = "DiluteLogRatio.Rdata")
```

```{r, eval = FALSE}
#DiluteLogRatio %>% na.omit() %>% select(-Study, -Oak, -Wine) -> DiluteLogRatio

# smoothedlist <- list()
# 
# for(i in unique(DiluteLogRatio$DataName)){
#   smoothedlist[[i]] <-foreach(c=unique(DiluteLogRatio$CHROM), .combine=rbind) %dopar%{
#     cybrSmoothBSAWindows(Results = DiluteLogRatio[DiluteLogRatio$DataName == i,!names(DiluteLogRatio) %in% "DataName"], chr = c)
#   } 
#   smoothedlist[[i]]$DataName <- i
# }
# 
# DiluteLogRatio_ZSmooth <- bind_rows(smoothedlist)
# 
# save(DiluteLogRatio_ZSmooth, file = "DiluteLogRatio_ZSmooth.Rdata")
################################################################################
smoothedlist <- list()
# DiluteLogRatio$Oak <- DiluteLogRatio$Oak + 1
# DiluteLogRatio$Wine <- DiluteLogRatio$Wine + 1
# DiluteLogRatio$logRatio <- log(DiluteLogRatio$Wine/DiluteLogRatio$Oak)

for(i in unique(DiluteLogRatio$DataName)){
  smoothedlist[[i]] <-foreach(c=unique(DiluteLogRatio$CHROM), .combine=rbind) %dopar%{
    cybrSmoothBSAWindows(Results = DiluteLogRatio[DiluteLogRatio$DataName == i,!names(DiluteLogRatio) %in% "DataName"], chr = c)
  } 
  smoothedlist[[i]]$DataName <- i
}

DiluteLogRatio_ZSmooth_p1 <- bind_rows(smoothedlist)

save(DiluteLogRatio_ZSmooth_p1, file = "DiluteLogRatio_ZSmooth_p1.Rdata")
################################################################################
smoothedlist <- list()
# DiluteLogRatio$Oak <- DiluteLogRatio$Oak + 1
# DiluteLogRatio$Wine <- DiluteLogRatio$Wine + 1
# DiluteLogRatio$logRatio <- log(DiluteLogRatio$Wine/DiluteLogRatio$Oak)

for(i in unique(DiluteLogRatio$DataName)){
  smoothedlist[[i]] <-foreach(c=unique(DiluteLogRatio$CHROM), .combine=rbind) %dopar%{
    cybrSmoothBSAWindows_Med(Results = DiluteLogRatio[DiluteLogRatio$DataName == i,!names(DiluteLogRatio) %in% "DataName"], chr = c)
  } 
  smoothedlist[[i]]$DataName <- i
}

DiluteLogRatio_ZSmooth_med <- bind_rows(smoothedlist)

save(DiluteLogRatio_ZSmooth_med, file = "DiluteLogRatio_ZSmooth_med.Rdata")
```

Plotting smoothed without correcting for counts

```{r}
load("DiluteLogRatio_ZSmooth_med.Rdata")
load("DiluteLogRatio_ZSmooth.Rdata")
load("DiluteLogRatio.Rdata")
DiluteLogRatio_ZSmooth_med$CHROM <- factor(as.roman(DiluteLogRatio_ZSmooth$CHROM), levels = ChromKey$chromosomes)
DiluteLogRatio_ZSmooth_med <- left_join(DiluteLogRatio_ZSmooth_med, DiluteLogRatio[,1:4])

DiluteLogRatio_ZSmooth_med %>% ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-0.5,0.5)+
  theme_minimal() + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Log Ratio") + 
  ggtitle("All Smoothed ")

DiluteLogRatio_ZSmooth_med %>% filter(CHROM == "VII") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_point(alpha = 0.1, size = 1) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() + 
      theme(legend.position = "bottom",
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Log Ratio") + 
  ggtitle("All Smoothed ")

DiluteLogRatio_ZSmooth_med %>% filter(CHROM == "V") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_point(alpha = 0.1, size = 1) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() + 
      theme(legend.position = "bottom",
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Log Ratio") + 
  ggtitle("All Smoothed ")
```

Doing this with adding 1 to every count

```{r}
#Load Data
load("DiluteLogRatio_ZSmooth_p1.Rdata")

#Reformat saved data for these plots
DiluteLogRatio_ZSmooth_p1$CHROM <- factor(as.roman(DiluteLogRatio_ZSmooth_p1$CHROM), levels = ChromKey$chromosomes)
DiluteLogRatio_ZSmooth_p1 <- left_join(DiluteLogRatio_ZSmooth_p1, DiluteLogRatio[,1:4])

DiluteLogRatio_ZSmooth_p1 %>% select(-Study_Z, -Study_Zprime) %>% mutate(Oak = Oak_Z - 1, Wine = Wine_Z - 1) -> DLR_Smooth_Intermediate

DiluteLogRatio_ZSmooth_p1$Zeros <- NA
DiluteLogRatio_ZSmooth_p1$Zeros[DLR_Smooth_Intermediate$Oak == 0 | DLR_Smooth_Intermediate$Wine == 0] <- "Inf"

sum(DiluteLogRatio_ZSmooth_p1$Zeros == "Inf", na.rm = TRUE)

#Plot log ratios
DiluteLogRatio_ZSmooth_p1 %>% ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-0.5,0.5)+ geom_hline(yintercept = 0, color = "black")+
  theme_minimal() + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("All Dilute Smoothed Log Wine/Oak")

DiluteLogRatio_ZSmooth_p1 %>% filter(CHROM == "VII") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study)) + geom_point(alpha = 0.1, size = 0.7) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal() + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Chr VII Smoothed Log Ratios")

DiluteLogRatio_ZSmooth_p1 %>% filter(CHROM == "V") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study)) + geom_point(alpha = 0.1, size = 0.7) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal() + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Chr V Smoothed Log Ratios")

```

```{r, eval = FALSE}
DiluteLogRatio_ZSmooth_p1 %>% filter(CHROM == "VII") %>% filter(logRatio_Zprime > 0.3) -> VII_Peak

ggplot(VII_Peak, aes(x = POS)) + geom_density()

VII_Peak %>% filter(POS < 5e5) %>% ggplot(aes(x = POS, y = logRatio_Zprime, color = Study)) + geom_point(alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal() 

DiluteLogRatio_ZSmooth_p1 %>% filter(CHROM == "VII", POS < 500000, POS > 300000, Study == "CuSO4") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = DataName)) + geom_point(alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal()

DiluteLogRatio %>% filter(CHROM == "VII", POS < 500000, POS > 300000, Study == "CuSO4") %>% 
  ggplot(aes(x = POS, y = logRatio, color = DataName)) + geom_point(alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal()

```


### Same for selected bulks

```{r}
load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

CSSI_CycZeo %>% filter(Bulk != "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts_s
CSSI_Fluc %>% filter(Bulk != "Dilute") %>% mutate(Study = "Fluc") -> FCounts_s
CuSO4data %>% filter(Bulk != "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts_s

SelectedData <- rbind(CZCounts_s,FCounts_s,CuCounts_s)

DatasetKey_s <- data.frame(Dataset = unique(SelectedData$Dataset),
                         DataName = c("CZ_OCA", "CZ_OCB", "CZ_OZA", "CZ_OZB","CZ_WCA", "CZ_WCB","CZ_WZA", "CZ_WZB", "F_OFB", "F_OFC", "F_OFD", "F_WFC", "Cu_SA", "Cu_SC"))

SelectedData$Dataset <- as.character(SelectedData$Dataset)
SelectedData %>% left_join(DatasetKey_s, by = "Dataset") -> SelectedLabeled

SelectedLabeled$DataName <- factor(SelectedLabeled$DataName, 
                                   levels = c("CZ_OCA", "CZ_OCB", "CZ_OZA", "CZ_OZB","CZ_WCA", "CZ_WCB","CZ_WZA", "CZ_WZB", "F_OFB", "F_OFC", "F_OFD", "F_WFC", "Cu_SA", "Cu_SC"))

mylabelcolors <- c("#91171F","#5C95FF",
                     "black", "#966B9D", "#A94C4C", "#2E1C2B", 
                     "#F2B880", "#46B1C9", "#3E517A", "#CBBAED",
                     "#8C2155", "#F87060", "#183059", "lightblue")
names(mylabelcolors) <- levels(SelectedLabeled$DataName)

SelectedLabeled %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.2) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Read Ratio by Selected Dataset")

SelectedLabeled %>% filter(CHROM == "VII") %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.8) + 
  facet_grid(~Bulk) + ylim(-2,2) +
  theme_minimal() + scale_color_manual(name = "", values = mylabelcolors) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Chr VII Read Ratio by Selected Dataset")
```



# Doing actual pairwise comparisons rather than all or nothing

Z scores for pairwise comparisons of dilute data from each experiment

```{r, eval = FALSE}
#Load in data
load("DiluteData.Rdata")

#USE DIFFERENT DATASETS
DiluteData$CycZeo = ifelse(DiluteData$Study == "CycZeo", "CycZeo", NA)
DiluteData$CuSO4 = ifelse(DiluteData$Study == "CuSO4", "CuSO4", NA)
DiluteData$Fluc = ifelse(DiluteData$Study == "Fluc", "Fluc", NA)

#Make the column to compare
CuZeo_Compare <- DiluteData %>% pivot_longer(c(CycZeo, CuSO4), names_to = "Experiment", values_to = "Binary") %>% 
  filter(Binary == "CycZeo" | Binary == "CuSO4")
CuFlu_Compare <- DiluteData %>% pivot_longer(c(Fluc, CuSO4), names_to = "Experiment", values_to = "Binary") %>% 
  filter(Binary == "Fluc" | Binary == "CuSO4")
FluZeo_Compare <- DiluteData %>% pivot_longer(c(Fluc, CycZeo), names_to = "Experiment", values_to = "Binary") %>% 
  filter(Binary == "Fluc" | Binary == "CycZeo")

#Find only the rows with both
CuZeo_Compare %>% group_by(CHROM, POS) %>% summarize(CHROM = CHROM, POS = POS, Count = length(unique(Binary))) %>% distinct() %>% filter(Count == 2) %>% select(-Count) %>% left_join(CuZeo_Compare) -> CuZeo
CuFlu_Compare %>% group_by(CHROM, POS) %>% summarize(CHROM = CHROM, POS = POS, Count = length(unique(Binary))) %>% distinct() %>% filter(Count == 2) %>% select(-Count) %>% left_join(CuFlu_Compare) -> CuFlu
FluZeo_Compare %>% group_by(CHROM, POS) %>% summarize(CHROM = CHROM, POS = POS, Count = length(unique(Binary))) %>% distinct() %>% filter(Count == 2) %>% select(-Count) %>% left_join(FluZeo_Compare) -> FluZeo

#Run the analysis
#Run BSA_GLM() and SmoothBSAWindows() for each
CSSI_FluZeo <- foreach(i=unique(FluZeo$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(FluZeo, chr = i, formula = "PAllele ~ Experiment+Parent") %>% cybrSmoothBSAWindows()
}
save(CSSI_FluZeo, file = "CSSI_FluZeo.Rdata")

CSSI_CuFlu <- foreach(i=unique(CuFlu$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuFlu, chr = i, formula = "PAllele ~ Experiment+Parent") %>% cybrSmoothBSAWindows()
}
save(CSSI_CuFlu, file = "CSSI_CuFlu.Rdata")

CSSI_CuZeo <- foreach(i=unique(CuZeo$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuZeo, chr = i, formula = "PAllele ~ Experiment+Parent") %>% cybrSmoothBSAWindows()
}
save(CSSI_CuZeo, file = "CSSI_CuZeo.Rdata")

```

## Plotting

Trying this next: library(patchwork)

```{r}

load("CSSI_CuZeo.Rdata") 
load("CSSI_CuFlu.Rdata")
load("CSSI_FluZeo.Rdata")

CSSI_CuZeo %>% ggplot(aes(x = POS, y = ExperimentCycZeo_Zprime)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("CycZeo vs CuSO4 Unselected")

CSSI_CuFlu %>% ggplot(aes(x = POS, y = ExperimentFluc_Zprime)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Fluconazole vs CuSO4 Unselected")

CSSI_FluZeo %>% ggplot(aes(x = POS, y = ExperimentFluc_Zprime)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Fluconazole vs CycZeo Unselected")

## View These Together
################################################################################
CSSI_CuZeo %>% transmute(CHROM = CHROM, POS = POS, Z = ExperimentCycZeo_Z, 
                         Zprime = ExperimentCycZeo_Zprime, 
                         Parent_Z = ParentWineI_Z, 
                         Parent_Zprime = ParentWineI_Zprime,
                         Label = "CycZeo_CuSO4") -> CSSI_CuZeo_t

CSSI_CuFlu %>% transmute(CHROM = CHROM, POS = POS, Z = ExperimentFluc_Z, 
                         Zprime = ExperimentFluc_Zprime, 
                         Parent_Z = ParentWineI_Z, 
                         Parent_Zprime = ParentWineI_Zprime,
                         Label = "Fluc_CuSO4") -> CSSI_CuFlu_t

CSSI_FluZeo %>% transmute(CHROM = CHROM, POS = POS, Z = ExperimentFluc_Z, 
                         Zprime = ExperimentFluc_Zprime, 
                         Parent_Z = ParentWineI_Z, 
                         Parent_Zprime = ParentWineI_Zprime,
                         Label = "Fluc_CycZeo") -> CSSI_FluZeo_t

PairwiseDilute <- rbind(CSSI_CuZeo_t,CSSI_CuFlu_t,CSSI_FluZeo_t)
PairwiseDilute$CHROM <- factor(PairwiseDilute$CHROM, levels = ChromKey$chromosomes)

PairwiseDilute %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Zprime, color = Label)) + 
  geom_line() +
  #geom_line(aes(x = POS, y = Parent_Zprime, color = Label),linetype = "dashed", alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + scale_color_manual(values = c("brown", "#0DAD8d", "violet")) +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Pairwise Unselected")

PairwiseDilute %>% filter(CHROM == "V") %>% ggplot(aes(x = POS, y = Zprime, color = Label)) + 
  geom_line(size = 1) +
  #geom_line(aes(x = POS, y = Parent_Zprime, color = Label),linetype = "dashed", alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + scale_color_manual(values = c("brown", "#0DAD8d", "violet")) +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Pairwise Unselected")

save(PairwiseDilute, file = "PairwiseDilute.Rdata")
```

Using colorpatch

```{r}
library(patchwork)

StatColorPath <- ggproto("StatColorPath", Stat,
                         compute_group = function(data, scales, params,
                                                  n_seg = 20, n = 100, cols = c("black", "white")) {
                           # interpolate
                           d <- approx(data$x, data$y, n = n)
                           # create start and end points for segments
                           d2 <- data.frame(
                             x = head(d$x, -1), xend = d$x[-1],
                             y = head(d$y, -1), yend = d$y[-1]
                           )
                           # create vector of segment colors
                           d2$color <- rep(cols, each = ceiling((n - 1) / n_seg), length.out = n - 1)
                           d2
                         },
                         required_aes = c("x", "y")
)

geom_colorpath <- function(mapping = NULL, data = NULL, geom = "segment",
                           position = "identity", na.rm = FALSE, show.legend = NA,
                           inherit.aes = TRUE, cols = c("black", "white"),
                           n_seg = 20, n = 100, ...) {
  layer(
    stat = StatColorPath, data = data, mapping = mapping, geom = geom,
    position = position, show.legend = show.legend, inherit.aes = inherit.aes,
    params = list(na.rm = na.rm, cols = cols, n = n, n_seg = n_seg,...)
  )
}

dat <- data.frame(x = seq(2,10, 2), y = seq(4,20, 4))

# p1 <- ggplot(dat, aes(x = x, y = y)) +
#   geom_colorpath()+
#   ggtitle("Default colors")
ggplot(dat, aes(x, y)) +
  geom_colorpath(cols = c("#F8766D", "#619CFF"), size = 1)+
  ggtitle("Two colors")

ggplot(dat, aes(x, y)) +
  geom_colorpath(cols = c("#F8766D", "#00BA38"), size = 1)+
  ggtitle("Two colors")

ggplot(dat, aes(x, y)) +
  geom_colorpath(cols = c("#00BA38", "#619CFF"), size = 1)+
  ggtitle("Two colors")

# p3 <- ggplot(dat, aes(x, y)) +
#   geom_colorpath(cols = c("red", "blue", "green"))+
#   ggtitle("Three colors")
# 
# p4 <- ggplot(dat, aes(x, y)) +
#   geom_colorpath(cols = c("red", "blue", "green", "white"))+
#   ggtitle("Four colors")
# 
# wrap_plots(mget(ls(pattern = "p[1-9]")))

#Red is F
#Green is 00
#Blue is 619

ggplot(PairwiseDilute, aes(x = POS, y = Zprime)) +
  geom_colorpath(data = PairwiseDilute[PairwiseDilute$Label == "CycZeo_CuSO4",], #red green
                 cols = c("#00BA38", "#F8766D"), size = 1, alpha = 0.5) +
  geom_colorpath(data = PairwiseDilute[PairwiseDilute$Label == "Fluc_CuSO4",], #red blue
                 cols = c("#F8766D", "#619CFF"), size = 1) +
  geom_colorpath(data = PairwiseDilute[PairwiseDilute$Label == "Fluc_CycZeo",], #blue green
                 cols = c("#00BA38", "#619CFF"), size = 1, alpha = 0.1) +
facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +  ggtitle("Pairwise Dilute Z Scores | Fluc vs CuSO4")

ggplot(PairwiseDilute, aes(x = POS, y = Zprime)) +
  geom_colorpath(data = PairwiseDilute[PairwiseDilute$Label == "CycZeo_CuSO4",], #red green
                 cols = c("#00BA38", "#F8766D"), size = 1, alpha = 0.5) +
  geom_colorpath(data = PairwiseDilute[PairwiseDilute$Label == "Fluc_CuSO4",], #red blue
                 cols = c("#F8766D", "#619CFF"), size = 1, alpha = 0.1) +
  geom_colorpath(data = PairwiseDilute[PairwiseDilute$Label == "Fluc_CycZeo",], #blue green
                 cols = c("#00BA38", "#619CFF"), size = 1) +
facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +  ggtitle("Pairwise Dilute Z Scores | Fluc vs CycZeo")

ggplot(PairwiseDilute, aes(x = POS, y = Zprime)) +
  geom_colorpath(data = PairwiseDilute[PairwiseDilute$Label == "CycZeo_CuSO4",], #red green
                 cols = c("#00BA38", "#F8766D"), size = 1) +
  geom_colorpath(data = PairwiseDilute[PairwiseDilute$Label == "Fluc_CuSO4",], #red blue
                 cols = c("#F8766D", "#619CFF"), size = 1, alpha = 0.5) +
  geom_colorpath(data = PairwiseDilute[PairwiseDilute$Label == "Fluc_CycZeo",], #blue green
                 cols = c("#00BA38", "#619CFF"), size = 1, alpha = 0.1) +
facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +  ggtitle("Pairwise Dilute Z Scores | CycZeo vs CuSO4")

#CHROMOSOME VII HIGHLIGHT
PairwiseDilute_VII <- subset(PairwiseDilute, CHROM == "VII")
ggplot(PairwiseDilute_VII, aes(x = POS, y = Zprime)) +
  geom_colorpath(data = PairwiseDilute_VII[PairwiseDilute_VII$Label == "CycZeo_CuSO4",], #red green
                 cols = c("#00BA38", "#F8766D"), size = 2, alpha = 0.5) +
  geom_colorpath(data = PairwiseDilute_VII[PairwiseDilute_VII$Label == "Fluc_CuSO4",], #red blue
                 cols = c("#F8766D", "#619CFF"), size = 2) +
  geom_colorpath(data = PairwiseDilute_VII[PairwiseDilute_VII$Label == "Fluc_CycZeo",], #blue green
                 cols = c("#00BA38", "#619CFF"), size = 2, alpha = 0.1) +
facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +  ggtitle("Pairwise Dilute Z Scores | Fluc vs CuSO4")

#CHROMOSOME V HIGHLIGHT
PairwiseDilute_V <- subset(PairwiseDilute, CHROM == "V")
ggplot(PairwiseDilute_V, aes(x = POS, y = Zprime)) +
  geom_colorpath(data = PairwiseDilute_V[PairwiseDilute_V$Label == "CycZeo_CuSO4",], #red green
                 cols = c("#00BA38", "#F8766D"), size = 2, alpha = 0.5) +
  geom_colorpath(data = PairwiseDilute_V[PairwiseDilute_V$Label == "Fluc_CuSO4",], #red blue
                 cols = c("#F8766D", "#619CFF"), size = 2) +
  geom_colorpath(data = PairwiseDilute_V[PairwiseDilute_V$Label == "Fluc_CycZeo",], #blue green
                 cols = c("#00BA38", "#619CFF"), size = 2, alpha = 0.1) +
facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +  ggtitle("Pairwise Dilute Z Scores | Fluc vs CuSO4") + ylim(-0.5, 2.5)
```


## Variance along dilute samples

Sources of noise in samples include sequencing noise (should be stochastic and normally distributed), the change in expected 50/50 allele frequencies in dilute samples (harder to quantify), the agreement of the replicates (should be the same) and the agreement of the experiments in dilute samples (should be the same but might have slight differences too).  

How can we know if the signal we are getting is a product of the region having higher variance, the area by chance having a higher stochasticity, or the region actually having some biological effect? And what is the cause of Chr V having such a weird log ratio between alleles?

### Variance in sliding window regions

For every set of 100 SNPs, calculate the variance at that point.

```{r}
SNPWindowVariance <- function(Results, windowsize = 100, chr = unique(Results$CHROM)[1]){

  #extract parameters
  Results %>% select(-CHROM, -POS) %>% names() -> params

  #arrange by position
  Results %>% filter(CHROM == chr) %>% arrange(POS) -> Results

  Results %>% summarise(SNPs = length(unique(POS))) %>%
    distinct() %>%
    mutate(maxW = floor(SNPs/2)) -> tablecounts

  if(windowsize < tablecounts$maxW){
    #run window
    WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {

      smoothedparams <- vector()
      for(p in 1:length(params)){
        smoothedparams[p] <- sd(Results[[params[p]]][(i-windowsize):(i+windowsize)])
      }

      #print CHROM, index, POS, and mean
      unlist(c(chr, i, Results[i,2:(2+length(params))],smoothedparams))
    }

    #rename columns
    WResult <- as.data.frame(WResult)
    colnames(WResult) <- c("CHROM", "Index", "POS",
                           paste(params, "Z", sep = "_"),
                           paste(params, "ZVar", sep = "_"))

    #convert to numeric
    for(i in 2:length(colnames(WResult))){
      WResult[,i] <- as.numeric(WResult[,i])
    }
    return(WResult)
  }else{
    warning(print(paste("Window size of chromosome ", chr, "are larger than number of data points")),
            call. = TRUE, immediate. = FALSE, noBreaks. = FALSE,
            domain = NULL)
  }
}

```

```{r}
varcuzeo <- list()
for(i in unique(CSSI_CuZeo$CHROM)){
  CSSI_CuZeo %>% filter(CHROM == i) %>% select(CHROM, POS, ExperimentCycZeo_Z, ParentWineI_Z) %>% SNPWindowVariance() -> varcuzeo[[i]]
  
}

varcuzeo_df <- bind_rows(varcuzeo$I, varcuzeo$II,varcuzeo$IV,varcuzeo$V,varcuzeo$VII,varcuzeo$VIII,
          varcuzeo$IX, varcuzeo$X, varcuzeo$XI,varcuzeo$XII,varcuzeo$XIII,varcuzeo$XIV,varcuzeo$XV,varcuzeo$XVI)

varcuzeo_df$CHROM <- factor(varcuzeo_df$CHROM, levels = as.character(as.roman(1:16)))
```

```{r}

varcuzeo_df %>% ggplot(aes(x = POS, y = ExperimentCycZeo_Z_ZVar)) + 
  geom_point(data = CSSI_CuZeo, aes(x = POS, y = ExperimentCycZeo_Z), alpha = 0.1) +
  geom_line(alpha = 0.5, color = "red", size = 1) + 
  geom_line(data = CSSI_CuZeo, aes(x = POS, y = ExperimentCycZeo_Zprime), color = "blue", size = 1) +
  geom_hline(yintercept = 0, color = "white") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ggtitle("SD of Cyclohex vs Zeocin Dilute Z Scores")

varcuzeo_df %>% ggplot(aes(x = POS, y = ExperimentCycZeo_Z_ZVar)) + 
  geom_point(data = CSSI_CuZeo, aes(x = POS, y = ExperimentCycZeo_Z), alpha = 0.1) +
  geom_line(alpha = 0.5, color = "red", size = 1) + 
  geom_line(data = CSSI_CuZeo, aes(x = POS, y = ExperimentCycZeo_Zprime), color = "blue", size = 1) +
  geom_hline(yintercept = 0, color = "white") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ggtitle("SD of Cyclohex vs Zeocin Dilute Z Scores")

left_join(varcuzeo_df, CSSI_CuZeo[c("CHROM", "POS", "ExperimentCycZeo_Zprime")]) %>% 
  ggplot(aes(x = abs(ExperimentCycZeo_Zprime), y = ExperimentCycZeo_Z_ZVar)) + 
  geom_point(alpha = 0.3) + ggtitle("Correlation of SD and Zprime within Windows")
```


For this to be relevant, we'd need to actually subtract and add that SD to our line so:

```{r}
left_join(varcuzeo_df, CSSI_CuZeo[c("CHROM", "POS", "ExperimentCycZeo_Zprime")]) %>% 
  mutate(Upper = ExperimentCycZeo_Zprime + abs(ExperimentCycZeo_Z_ZVar),
         Lower = ExperimentCycZeo_Zprime - abs(ExperimentCycZeo_Z_ZVar)) -> CI_CuZeo

CI_CuZeo$CHROM <- factor(CI_CuZeo$CHROM, levels = as.character(as.roman(1:16)))

CI_CuZeo %>% ggplot(aes(x = POS, y = ExperimentCycZeo_Zprime)) + 
  geom_line(size = 1) + geom_line(aes(y = Upper), color = "gray") + 
  geom_line(aes(y = Lower), color = "gray") + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ggtitle("SD of Cyclohex vs Zeocin Dilute Z Scores")

CI_CuZeo %>% filter(CHROM == "IV") %>% ggplot(aes(x = POS, y = ExperimentCycZeo_Zprime)) + 
  geom_line(size = 1) + geom_line(aes(y = Upper), color = "gray") + 
  geom_line(aes(y = Lower), color = "gray") + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ggtitle("SD of Cyclohex vs Zeocin Dilute Z Scores")
  
```

If we were to do this for the actual data, what would that look like?

```{r}
load("CSSI_Fluc_BSA_med.Rdata")

sdplaceholder <- list()
for(i in unique(CSSI_Fluc_BSA$CHROM)){
  CSSI_Fluc_BSA %>% filter(CHROM == i) %>% select(CHROM, POS, BulkFluconazole_Z, ParentWineI_Z) %>% SNPWindowVariance() -> sdplaceholder[[i]]
  
}

CSSI_Fluc_BSA_sd_df <- bind_rows(sdplaceholder$I, sdplaceholder$II,sdplaceholder$IV,sdplaceholder$V,sdplaceholder$VII,sdplaceholder$VIII,
          sdplaceholder$IX, sdplaceholder$X, sdplaceholder$XI,sdplaceholder$XII,sdplaceholder$XIII,sdplaceholder$XIV,sdplaceholder$XV,sdplaceholder$XVI)

CSSI_Fluc_BSA_sd_df$CHROM <- factor(CSSI_Fluc_BSA_sd_df$CHROM, levels = as.character(as.roman(1:16)))

left_join(CSSI_Fluc_BSA_sd_df, CSSI_Fluc_BSA[c("CHROM", "POS", "BulkFluconazole_Zprime")]) %>% 
  mutate(Upper = BulkFluconazole_Zprime + abs(BulkFluconazole_Z_ZVar),
         Lower = BulkFluconazole_Zprime - abs(BulkFluconazole_Z_ZVar)) -> CI_CSSI_Fluc_BSA

CI_CSSI_Fluc_BSA %>% ggplot(aes(x = POS, y = BulkFluconazole_Zprime)) + 
  geom_line(size = 1) + geom_line(aes(y = Upper), color = "gray") + 
  geom_line(aes(y = Lower), color = "gray") + geom_hline(yintercept = 0, color = "navyblue", alpha = 0.5)+
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ggtitle("SD of Fluconazole Z Scores")

CI_CSSI_Fluc_BSA %>% ggplot(aes(x = POS, y = BulkFluconazole_Zprime)) + 
  geom_line(size = 1) + 
  geom_line(aes(y = abs(BulkFluconazole_Z_ZVar)), color = "gray") + 
  geom_line(aes(y = -abs(BulkFluconazole_Z_ZVar)), color = "gray") + 
  geom_hline(yintercept = 0, color = "navyblue", alpha = 0.5)+
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ggtitle("SD of Fluconazole Z Scores")
```


There's also more or less error in the replicates; let's look at that for our confidence in the nulls?

```{r}

```

