---
title: "CSS Analysis | Medians"
author: "Cassandra Buzby"
date: "12/2/2022"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

library(cybrBSA)

ggplot2::theme_set(theme_light())
#ggplot2::theme_set(theme_bw())


CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47",
                  "#F26430", "#0A369D", "#7EA3CC")


library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))


setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")
# load("RdataFiles/SAUA_4-13-22.Rdata")
# load("RdataFiles/SCUC_4-13-22.Rdata")
# load("RdataFiles/bsa_glm_results.Rdata")

#rawData = "HGV.SortedCat.vcf.output.table"
```

## Using Medians instead of Means

Analyzing the dilute data from all three experiments showed weird skews when using means to smooth, so checking what happens if I change the smoothing script to use medians instead.

```{r}
#CHANGE FUNCTION FOR THIS SCRIPT ONLY
cybrSmoothBSAWindows <- function(Results, windowsize = 100, chr = unique(Results$CHROM)[1]){

  #extract parameters
  Results %>% select(-CHROM, -POS) %>% names() -> params

  #arrange by position
  Results %>% filter(CHROM == chr) %>% arrange(POS) -> Results

  Results %>% summarise(SNPs = length(unique(POS))) %>%
    distinct() %>%
    mutate(maxW = floor(SNPs/2)) -> tablecounts

  if(windowsize < tablecounts$maxW){
    #run window
    WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {

      smoothedparams <- vector()
      for(p in 1:length(params)){
        smoothedparams[p] <- median(Results[[params[p]]][(i-windowsize):(i+windowsize)])
      }

      #print CHROM, index, POS, and mean
      unlist(c(chr, i, Results[i,2:(2+length(params))],smoothedparams))
    }

    #rename columns
    WResult <- as.data.frame(WResult)
    colnames(WResult) <- c("CHROM", "Index", "POS",
                           paste(params, "Z", sep = "_"),
                           paste(params, "Zprime", sep = "_"))

    #convert to numeric
    for(i in 2:length(colnames(WResult))){
      WResult[,i] <- as.numeric(WResult[,i])
    }
    return(WResult)
  }else{
    warning(print(paste("Window size of chromosome ", chr, "are larger than number of data points")),
            call. = TRUE, immediate. = FALSE, noBreaks. = FALSE,
            domain = NULL)
  }
}
```

## Convert Parental Alleles

```{r, eval = FALSE}
parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

head(parentSNPids)
```

## Using cybrBSA package

CSS I Fluconazole

```{r, eval = FALSE}
mydatatotest = "HGV.SortedCat.vcf.output.table"

CSSI_Fluc <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CSSI_Fluc$Dataset)

CSSI_Fluc$Bulk <- NA
CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq"] <- "Dilute"

CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "Fluconazole"

CSSI_Fluc$Parent <- NA
CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq"] <- "OakI"

CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "WineI"

CSSI_Fluc$Parent <- factor(CSSI_Fluc$Parent)
CSSI_Fluc$Bulk <- factor(CSSI_Fluc$Bulk)
CSSI_Fluc$ReadCount <- as.numeric(CSSI_Fluc$ReadCount)

head(CSSI_Fluc)

save(CSSI_Fluc, file ="CSSI_Fluc_counts_med.Rdata")

CSSI_Fluc_BSA <- foreach(i=unique(CSSI_Fluc$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Fluc, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Fluc_BSA, file = "CSSI_Fluc_BSA_med.Rdata")
```


CSS I CuSO4

```{r, eval = FALSE}
CuSO4gatk <- "Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"

CuSO4data <- cybrInputGATKTable(CuSO4gatk) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CuSO4data$Dataset)

CuSO4data$Bulk <- NA
CuSO4data$Bulk[CuSO4data$Dataset == "SelectedA" |
              CuSO4data$Dataset == "SelectedC"] <- "CuSO4"

CuSO4data$Bulk[CuSO4data$Dataset == "UnselectedA" |
              CuSO4data$Dataset == "UnselectedC"] <- "Dilute"

CuSO4data$Parent <- NA
CuSO4data$Parent[CuSO4data$Dataset == "SelectedA" |
              CuSO4data$Dataset == "UnselectedA"] <- "OakI"

CuSO4data$Parent[CuSO4data$Dataset == "SelectedC" |
              CuSO4data$Dataset == "UnselectedC"] <- "WineI"

CuSO4data$Parent <- factor(CuSO4data$Parent)
CuSO4data$Bulk <- factor(CuSO4data$Bulk)
CuSO4data$ReadCount <- as.numeric(CuSO4data$ReadCount)

head(CuSO4data)

save(CuSO4data, file = "CuSO4data_counts.Rdata")

CSSI_CuSO4_BSA <- foreach(i=unique(CuSO4data$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuSO4data, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_CuSO4_BSA, file = "CSSI_CuSO4_BSA_med.Rdata")

```


## Cycloheximide / Zeocin Data

```{r, eval = FALSE}
mydatatotest = "HKTMZDRX2.SortedCat.vcf.output.table"

CSSI_CycZeo <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CSSI_CycZeo$Dataset)

CSSI_CycZeo$Bulk <- NA
CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq"] <- "Cycloheximide"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "Dilute"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq"] <- "Zeocin"

CSSI_CycZeo$Parent <- NA
CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq"] <- "OakI"

CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "WineI"

CSSI_CycZeo$Parent <- factor(CSSI_CycZeo$Parent)
CSSI_CycZeo$Bulk <- factor(CSSI_CycZeo$Bulk)
CSSI_CycZeo$ReadCount <- as.numeric(CSSI_CycZeo$ReadCount)

head(CSSI_CycZeo)

save(CSSI_CycZeo, file = "CSSI_CycZeo_counts_med.Rdata")

#Separate into different datasets, since the subsequent steps will break if not
CSSI_Cycloheximide <- CSSI_CycZeo[CSSI_CycZeo$Bulk != "Zeocin",]
CSSI_Zeocin <- CSSI_CycZeo[CSSI_CycZeo$Bulk != "Cycloheximide",]

#Run BSA_GLM() and SmoothBSAWindows() for each
CSSI_Cyc_BSA <- foreach(i=unique(CSSI_Cycloheximide$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA, file = "CSSI_Cyc_BSA_med.Rdata")

CSSI_Zeo_BSA <- foreach(i=unique(CSSI_Zeocin$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA, file = "CSSI_Zeo_BSA_med.Rdata")
```

## Redoing Zeocin with replicates

Run Analysis

```{r, eval = FALSE}

CSSI_Zeocin_A <- CSSI_Zeocin[grep("A.fastq", CSSI_Zeocin$Dataset),]
CSSI_Zeocin_B <- CSSI_Zeocin[grep("B.fastq", CSSI_Zeocin$Dataset),]

CSSI_Zeo_BSA_A <- foreach(i=unique(CSSI_Zeocin_A$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin_A, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA_A, file = "CSSI_Zeo_BSA_A_med.Rdata")

CSSI_Zeo_BSA_B <- foreach(i=unique(CSSI_Zeocin_B$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin_B, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA_B, file = "CSSI_Zeo_BSA_B_med.Rdata")
```

Checking these

```{r}
colvalues = c("#345F6F", "#D7335C", "#FFB05C")

load("CSSI_Zeo_BSA_A_med.Rdata")
load("CSSI_Zeo_BSA_B_med.Rdata")

CSSI_Zeo_BSA_A$Rep = "A"
CSSI_Zeo_BSA_B$Rep = "B"

combined_Zeo <- rbind(CSSI_Zeo_BSA_A,CSSI_Zeo_BSA_B) 
colnames(combined_Zeo) <- c(colnames(combined_Zeo)[1:8], "Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime", "Rep")

combined_Zeo %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Bulk Zprime Zeocin")

combined_Zeo %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Interaction_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Interaction Zprime Zeocin")

combined_Zeo %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Parent_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Zeocin")

combined_Zeo %>% pivot_longer(c(Bulk_Zprime, Interaction_Zprime, Parent_Zprime), names_to = "Effect", values_to = "Zprime") %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Zprime, color = Effect, linetype = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + scale_color_manual(values = colvalues) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Zeocin")
```

## Redoing Cycloheximide with replicates

Run Analysis

```{r, eval = FALSE}

CSSI_Cycloheximide_A <- CSSI_Cycloheximide[grep("A.fastq", CSSI_Cycloheximide$Dataset),]
CSSI_Cycloheximide_B <- CSSI_Cycloheximide[grep("B.fastq", CSSI_Cycloheximide$Dataset),]

CSSI_Cyc_BSA_A <- foreach(i=unique(CSSI_Cycloheximide_A$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide_A, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA_A, file = "CSSI_Cyc_BSA_A_med.Rdata")

CSSI_Cyc_BSA_B <- foreach(i=unique(CSSI_Cycloheximide_B$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide_B, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA_B, file = "CSSI_Cyc_BSA_B_med.Rdata")
```

Checking these

```{r}
colvalues = c("#345F6F", "#D7335C", "#FFB05C")

load("CSSI_Cyc_BSA_A_med.Rdata")
load("CSSI_Cyc_BSA_B_med.Rdata")

CSSI_Cyc_BSA_A$Rep = "A"
CSSI_Cyc_BSA_B$Rep = "B"

combined_Cyc <- rbind(CSSI_Cyc_BSA_A,CSSI_Cyc_BSA_B) 
colnames(combined_Cyc) <- c(colnames(combined_Cyc)[1:8], "Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime", "Rep")

combined_Cyc %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Bulk Zprime Cycloheximide")

combined_Cyc %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Interaction_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Interaction Zprime Cycloheximide")

combined_Cyc %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Parent_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Cycloheximide")

combined_Cyc %>% pivot_longer(c(Bulk_Zprime, Interaction_Zprime, Parent_Zprime), names_to = "Effect", values_to = "Zprime") %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Zprime, color = Effect, linetype = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + scale_color_manual(values = colvalues) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Cycloheximide")
```


## Plotting

Load in Data

```{r}
load("CSSI_Fluc_BSA_med.Rdata")
load("CSSI_CuSO4_BSA.Rdata")

load("CSSI_Zeo_BSA_med.Rdata")
load("CSSI_Cyc_BSA_med.Rdata")
```


Plotting CSS I Fluconazole

```{r}
colnames(CSSI_Fluc_BSA)[9:11] <- c("Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime")
cybrPlotZPrime(CSSI_Fluc_BSA, columns = c("Bulk_Zprime"))

cybrPlotZScore(CSSI_Fluc_BSA, column = c("BulkFluconazole_Z"))

cybrPlotZPrime(CSSI_Fluc_BSA, chromosomes = ChromKey$chromosomes[2:16])

```
 
 Plotting CSS I CuSO4
 
```{r}
colnames(CSSI_CuSO4_BSA)[7:9] <- c("Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime")

cybrPlotZPrime(CSSI_CuSO4_BSA, columns = c("Bulk_Zprime"))

cybrPlotZPrime(CSSI_CuSO4_BSA, chromosomes = ChromKey$chromosomes[2:16])

cybrPlotZScore(CSSI_CuSO4_BSA)
```

Plotting Zeocin

```{r}
colnames(CSSI_Zeo_BSA)[9:11] <- c("Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime")

cybrPlotZPrime(CSSI_Zeo_BSA, columns = "Bulk_Zprime")

cybrPlotZPrime(CSSI_Zeo_BSA, chromosomes = ChromKey$chromosomes[2:16], title = "Zeocin Smoothed Z Scores")

```

Plotting Cycloheximide

```{r}
colnames(CSSI_Cyc_BSA)[9:11] <- c("Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime")

cybrPlotZPrime(CSSI_Cyc_BSA, columns = "Bulk_Zprime")

cybrPlotZPrime(CSSI_Cyc_BSA, chromosomes = ChromKey$chromosomes[2:16], title = "Cycloheximide Smoothed Z Scores")


```

Reasons why medians are important vs means:

```{r, eval = FALSE}
load("DiluteLogRatio.Rdata")
#DiluteLogRatio %>% na.omit() %>% select(-Study, -Oak, -Wine) -> DiluteLogRatio

# smoothedlist <- list()
# 
# for(i in unique(DiluteLogRatio$DataName)){
#   smoothedlist[[i]] <-foreach(c=unique(DiluteLogRatio$CHROM), .combine=rbind) %dopar%{
#     cybrSmoothBSAWindows(Results = DiluteLogRatio[DiluteLogRatio$DataName == i,!names(DiluteLogRatio) %in% "DataName"], chr = c)
#   } 
#   smoothedlist[[i]]$DataName <- i
# }
# 
# DiluteLogRatio_ZSmooth <- bind_rows(smoothedlist)
# 
# save(DiluteLogRatio_ZSmooth, file = "DiluteLogRatio_ZSmooth.Rdata")
################################################################################
smoothedlist <- list()
# DiluteLogRatio$Oak <- DiluteLogRatio$Oak + 1
# DiluteLogRatio$Wine <- DiluteLogRatio$Wine + 1
# DiluteLogRatio$logRatio <- log(DiluteLogRatio$Wine/DiluteLogRatio$Oak)

for(i in unique(DiluteLogRatio$DataName)){
  smoothedlist[[i]] <-foreach(c=unique(DiluteLogRatio$CHROM), .combine=rbind) %dopar%{
    cybrSmoothBSAWindows(Results = DiluteLogRatio[DiluteLogRatio$DataName == i,!names(DiluteLogRatio) %in% "DataName"], chr = c)
  } 
  smoothedlist[[i]]$DataName <- i
}

DiluteLogRatio_ZSmooth_p1 <- bind_rows(smoothedlist)

save(DiluteLogRatio_ZSmooth_p1, file = "DiluteLogRatio_ZSmooth_p1.Rdata")
################################################################################
smoothedlist <- list()
# DiluteLogRatio$Oak <- DiluteLogRatio$Oak + 1
# DiluteLogRatio$Wine <- DiluteLogRatio$Wine + 1
# DiluteLogRatio$logRatio <- log(DiluteLogRatio$Wine/DiluteLogRatio$Oak)

for(i in unique(DiluteLogRatio$DataName)){
  smoothedlist[[i]] <-foreach(c=unique(DiluteLogRatio$CHROM), .combine=rbind) %dopar%{
    cybrSmoothBSAWindows_Med(Results = DiluteLogRatio[DiluteLogRatio$DataName == i,!names(DiluteLogRatio) %in% "DataName"], chr = c)
  } 
  smoothedlist[[i]]$DataName <- i
}

DiluteLogRatio_ZSmooth_med <- bind_rows(smoothedlist)

save(DiluteLogRatio_ZSmooth_med, file = "DiluteLogRatio_ZSmooth_med.Rdata")
```

Plotting smoothed without correcting for counts

```{r}
load("DiluteLogRatio_ZSmooth_med.Rdata")
load("DiluteLogRatio.Rdata")

DiluteLogRatio_ZSmooth_med$CHROM <- factor(as.roman(DiluteLogRatio_ZSmooth_med$CHROM), levels = ChromKey$chromosomes)
DiluteLogRatio_ZSmooth_med <- left_join(DiluteLogRatio_ZSmooth_med, DiluteLogRatio[,1:4])

DiluteLogRatio_ZSmooth_med %>% ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-0.5,0.5)+
  theme_minimal() + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("All Smoothed ")

DiluteLogRatio_ZSmooth_med %>% filter(CHROM == "VII") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_point(alpha = 0.2) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() + 
      theme(legend.position = "bottom",
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("All Smoothed ")

DiluteLogRatio_ZSmooth_med %>% filter(CHROM == "V") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_point(alpha = 0.2) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() + 
      theme(legend.position = "bottom",
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("All Smoothed ")
```



## Doing actual pairwise comparisons rather than all or nothing

```{r, eval = FALSE}
#Load in data
load("DiluteData.Rdata")

#USE DIFFERENT DATASETS
DiluteData$CycZeo = ifelse(DiluteData$Study == "CycZeo", "CycZeo", NA)
DiluteData$CuSO4 = ifelse(DiluteData$Study == "CuSO4", "CuSO4", NA)
DiluteData$Fluc = ifelse(DiluteData$Study == "Fluc", "Fluc", NA)

#Make the column to compare
CuZeo_Compare <- DiluteData %>% pivot_longer(c(CycZeo, CuSO4), names_to = "Experiment", values_to = "Binary") %>% 
  filter(Binary == "CycZeo" | Binary == "CuSO4")
CuFlu_Compare <- DiluteData %>% pivot_longer(c(Fluc, CuSO4), names_to = "Experiment", values_to = "Binary") %>% 
  filter(Binary == "Fluc" | Binary == "CuSO4")
FluZeo_Compare <- DiluteData %>% pivot_longer(c(Fluc, CycZeo), names_to = "Experiment", values_to = "Binary") %>% 
  filter(Binary == "Fluc" | Binary == "CycZeo")

#Find only the rows with both
CuZeo_Compare %>% group_by(CHROM, POS) %>% summarize(CHROM = CHROM, POS = POS, Count = length(unique(Binary))) %>% distinct() %>% filter(Count == 2) %>% select(-Count) %>% left_join(CuZeo_Compare) -> CuZeo
CuFlu_Compare %>% group_by(CHROM, POS) %>% summarize(CHROM = CHROM, POS = POS, Count = length(unique(Binary))) %>% distinct() %>% filter(Count == 2) %>% select(-Count) %>% left_join(CuFlu_Compare) -> CuFlu
FluZeo_Compare %>% group_by(CHROM, POS) %>% summarize(CHROM = CHROM, POS = POS, Count = length(unique(Binary))) %>% distinct() %>% filter(Count == 2) %>% select(-Count) %>% left_join(FluZeo_Compare) -> FluZeo

#Run the analysis
#Run BSA_GLM() and SmoothBSAWindows() for each
CSSI_FluZeo <- foreach(i=unique(FluZeo$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(FluZeo, chr = i, formula = "PAllele ~ Experiment+Parent") %>% cybrSmoothBSAWindows()
}
save(CSSI_FluZeo, file = "CSSI_FluZeo_med.Rdata")

CSSI_CuFlu <- foreach(i=unique(CuFlu$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuFlu, chr = i, formula = "PAllele ~ Experiment+Parent") %>% cybrSmoothBSAWindows()
}
save(CSSI_CuFlu, file = "CSSI_CuFlu_med.Rdata")

CSSI_CuZeo <- foreach(i=unique(CuZeo$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuZeo, chr = i, formula = "PAllele ~ Experiment+Parent") %>% cybrSmoothBSAWindows()
}
save(CSSI_CuZeo, file = "CSSI_CuZeo_med.Rdata")

```

Plotting

```{r}
load("CSSI_CuZeo_med.Rdata")
load("CSSI_CuFlu_med.Rdata")
load("CSSI_FluZeo_med.Rdata")

CSSI_CuZeo %>% ggplot(aes(x = POS, y = ExperimentCycZeo_Zprime)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("CycZeo vs CuSO4 Unselected")

CSSI_CuFlu %>% ggplot(aes(x = POS, y = ExperimentFluc_Zprime)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Fluconazole vs CuSO4 Unselected")

CSSI_FluZeo %>% ggplot(aes(x = POS, y = ExperimentFluc_Zprime)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Fluconazole vs CycZeo Unselected")

## View These Together
################################################################################
CSSI_CuZeo %>% transmute(CHROM = CHROM, POS = POS, Z = ExperimentCycZeo_Z, 
                         Zprime = ExperimentCycZeo_Zprime, 
                         Parent_Z = ParentWineI_Z, 
                         Parent_Zprime = ParentWineI_Zprime,
                         Label = "CycZeo_CuSO4") -> CSSI_CuZeo_t

CSSI_CuFlu %>% transmute(CHROM = CHROM, POS = POS, Z = ExperimentFluc_Z, 
                         Zprime = ExperimentFluc_Zprime, 
                         Parent_Z = ParentWineI_Z, 
                         Parent_Zprime = ParentWineI_Zprime,
                         Label = "Fluc_CuSO4") -> CSSI_CuFlu_t

CSSI_FluZeo %>% transmute(CHROM = CHROM, POS = POS, Z = ExperimentFluc_Z, 
                         Zprime = ExperimentFluc_Zprime, 
                         Parent_Z = ParentWineI_Z, 
                         Parent_Zprime = ParentWineI_Zprime,
                         Label = "Fluc_CycZeo") -> CSSI_FluZeo_t

PairwiseDilute <- rbind(CSSI_CuZeo_t,CSSI_CuFlu_t,CSSI_FluZeo_t)
PairwiseDilute$CHROM <- factor(PairwiseDilute$CHROM, levels = ChromKey$chromosomes)

PairwiseDilute %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Zprime, color = Label)) + 
  geom_line() +
  #geom_line(aes(x = POS, y = Parent_Zprime, color = Label),linetype = "dashed", alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Pairwise Unselected")
```

Sanity check to make sure that they aren't just the same...?  

It's definitely different, but unfortunately very similar. Might need to do more in space rather than number of points?

```{r}
load("CSSI_CuZeo_med.Rdata")
CSSI_CuZeo_med <- CSSI_CuZeo

load("CSSI_CuZeo.Rdata")

head(CSSI_CuZeo_med)
head(CSSI_CuZeo)

CSSI_CuZeo_med$Test <- "Median"
CSSI_CuZeo$Test <- "Mean"

MedTest <- rbind(CSSI_CuZeo_med,CSSI_CuZeo)

ggplot(MedTest[MedTest$CHROM == "VII",], aes(x = POS, y = ParentWineI_Zprime, color = Test, shape = Test)) + 
  geom_point(alpha = 0.4) +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Pairwise Unselected")

ggplot(MedTest[MedTest$CHROM == "VII",], aes(x = POS, y = ParentWineI_Z, color = Test, shape = Test)) + 
  geom_point(alpha = 0.4) +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Pairwise Unselected")

ggplot(MedTest[MedTest$CHROM == "V",], aes(x = POS, y = ParentWineI_Zprime, color = Test)) + 
  geom_line(size = 1, alpha = 0.6) + geom_point(aes(y = ParentWineI_Z), alpha = 0.1, color = "gray")


```

# Re-writing window function for space instead of SNPs

Check to see if this works, not run yet

```{r}
load("CSSI_CuSO4_BSA.Rdata")

#TESTING 
Results <- CSSI_CuSO4_BSA[c(1,3:6)]
windowsize = 1000
chr = unique(Results$CHROM)[1]

i=unique(Results$POS)[1]
p <-  2

```

```{r}
################################################################################

#CHANGE FUNCTION FOR THIS SCRIPT ONLY
cybrSmoothBSA_Position <- function(Results, windowsize = 50000, chr = unique(Results$CHROM)[1]){

  #extract parameters
  Results %>% select(-CHROM, -POS) %>% names() -> params

  #arrange by position
  Results %>% filter(CHROM == chr) %>% arrange(POS) -> Results


  #run window
  WResult <- foreach(i=1:length(unique(Results$POS)), .combine=rbind) %dopar% {

    smoothedparams <- vector()
    for(p in 1:length(params)){
      #See how many unique positions are in the window to smooth?
      smoothedparams[p] <- median(Results[[params[p]]][Results$POS > (unique(Results$POS)[i]-windowsize) & Results$POS < (unique(Results$POS)[i]+windowsize)])
    }

    #print CHROM, index, POS, and mean
    unlist(c(chr, unique(Results$POS)[i], Results[i, 3:(2+length(params))], smoothedparams))
  }

  #rename columns
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "POS",
                         paste(params, "Z", sep = "_"),
                         paste(params, "Zprime", sep = "_"))

  #convert to numeric
  for(i in 2:length(colnames(WResult))){
    WResult[,i] <- as.numeric(WResult[,i])
  }
    
  return(WResult)
  
  }
```

Test the new function for Chr II

```{r}
testnewfunction <- cybrSmoothBSA_Position(CSSI_CuSO4_BSA[c(1,3:6)],windowsize = 50000, chr = "II")

testnewfunction %>% ggplot(aes(x = POS, y = Bulk_Z_Zprime)) + geom_line() + geom_point(aes(y = Bulk_Z_Z), alpha = 0.1)
```

## Re-running analysis with position instead of SNPs

```{r}
load("CSSI_Fluc_counts_med.Rdata")

CSSI_Fluc_BSA <- foreach(i=unique(CSSI_Fluc$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Fluc, chr = i) %>% cybrSmoothBSA_Position()
}

save(CSSI_Fluc_BSA, file = "CSSI_Fluc_BSA_medpos.Rdata")
```

Plot Fluconazole Data

```{r}
load("CSSI_Fluc_BSA_medpos.Rdata")

colnames(CSSI_Fluc_BSA)[8:10] <- c("Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime")

cybrPlotZPrime(CSSI_Fluc_BSA, columns = "Bulk_Zprime")

cybrPlotZPrime(CSSI_Fluc_BSA, chromosomes = ChromKey$chromosomes[2:16], title = "Cycloheximide Smoothed Z Scores")

```

Compare by space vs by SNPs

```{r}
load("CSSI_Fluc_BSA_medpos.Rdata")
colnames(CSSI_Fluc_BSA)[8:10] <- c("Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime")
CSSI_Fluc_Space <- CSSI_Fluc_BSA %>% mutate(Method = "Position+50k")
rm(CSSI_Fluc_BSA)

load("CSSI_Fluc_BSA.Rdata")
CSSI_Fluc_SNP <- CSSI_Fluc_BSA %>% mutate(Method = "SNPWindow")
rm(CSSI_Fluc_BSA)

FlucMethodComparison <- rbind(CSSI_Fluc_SNP[c(1,3, 7:10)], CSSI_Fluc_Space[c(1,2,8:11)])

FlucMethodComparison %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Method)) + geom_line(size = 0.8, alpha = 0.8) + scale_x_continuous(breaks = seq(from = 0, to = max(FlucMethodComparison$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Bulk Zprime Fluconazole")

FlucMethodComparison %>% filter(CHROM == "IV") %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Method)) + geom_line(size = 0.8, alpha = 0.8) 
```

Running with a larger window to compare

```{r}
load("CSSI_Fluc_counts_med.Rdata")

CSSI_Fluc_BSA_IV <- foreach(i="IV", .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Fluc, chr = i) %>% cybrSmoothBSA_Position(windowsize = 100000)
}

colnames(CSSI_Fluc_BSA_IV)[8:10] <- c("Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime")
CSSI_Fluc_Space_IV <- CSSI_Fluc_BSA_IV %>% mutate(Method = "Position+100k")

FlucMethodComparison <- rbind(FlucMethodComparison, CSSI_Fluc_Space_IV[c(1,2,8:11)])

FlucMethodComparison %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Method)) + geom_line(size = 0.8, alpha = 0.8) + scale_x_continuous(breaks = seq(from = 0, to = max(FlucMethodComparison$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Bulk Zprime Fluconazole")

FlucMethodComparison %>% filter(CHROM == "IV") %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Method)) + geom_line(size = 0.8, alpha = 0.8) 

FlucMethodComparison %>% filter(CHROM == "IV") %>% ggplot(aes(x = POS, y = Bulk_Zprime)) + geom_point(data = CSSI_Fluc_BSA_IV, aes(x = POS, y = BulkFluconazole_Z), alpha = 0.05) + geom_line(aes(color = Method),size = 0.8, alpha = 0.8) + ggtitle("Smoothing Methods Chr IV")

```

