---
title: "CSS Analysis"
author: "Cassandra Buzby"
date: "10/27/2022"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

library(cybrBSA)

ggplot2::theme_set(theme_light())
#ggplot2::theme_set(theme_bw())


CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47",
                  "#F26430", "#0A369D", "#7EA3CC")


library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")
# load("RdataFiles/SAUA_4-13-22.Rdata")
# load("RdataFiles/SCUC_4-13-22.Rdata")
# load("RdataFiles/bsa_glm_results.Rdata")

#rawData = "HGV.SortedCat.vcf.output.table"
```

## Convert Parental Alleles

```{r, eval = FALSE}
parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

head(parentSNPids)
```

## Using cybrBSA package

CSS I Fluconazole

```{r, eval = FALSE}
mydatatotest = "HGV.SortedCat.vcf.output.table"

CSSI_Fluc <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CSSI_Fluc$Dataset)

CSSI_Fluc$Bulk <- NA
CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq"] <- "Dilute"

CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "Fluconazole"

CSSI_Fluc$Parent <- NA
CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq"] <- "OakI"

CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "WineI"

CSSI_Fluc$Parent <- factor(CSSI_Fluc$Parent)
CSSI_Fluc$Bulk <- factor(CSSI_Fluc$Bulk)
CSSI_Fluc$ReadCount <- as.numeric(CSSI_Fluc$ReadCount)

head(CSSI_Fluc)

save(CSSI_Fluc, file ="CSSI_Fluc_counts.Rdata")

CSSI_Fluc_BSA <- foreach(i=unique(CSSI_Fluc$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Fluc, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Fluc_BSA, file = "CSSI_Fluc_BSA.Rdata")
```


CSS I CuSO4

```{r, eval = FALSE}
CuSO4gatk <- "Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"

CuSO4data <- cybrInputGATKTable(CuSO4gatk) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CuSO4data$Dataset)

CuSO4data$Bulk <- NA
CuSO4data$Bulk[CuSO4data$Dataset == "SelectedA" |
              CuSO4data$Dataset == "SelectedC"] <- "CuSO4"

CuSO4data$Bulk[CuSO4data$Dataset == "UnselectedA" |
              CuSO4data$Dataset == "UnselectedC"] <- "Dilute"

CuSO4data$Parent <- NA
CuSO4data$Parent[CuSO4data$Dataset == "SelectedA" |
              CuSO4data$Dataset == "UnselectedA"] <- "OakI"

CuSO4data$Parent[CuSO4data$Dataset == "SelectedC" |
              CuSO4data$Dataset == "UnselectedC"] <- "WineI"

CuSO4data$Parent <- factor(CuSO4data$Parent)
CuSO4data$Bulk <- factor(CuSO4data$Bulk)
CuSO4data$ReadCount <- as.numeric(CuSO4data$ReadCount)

head(CuSO4data)

save(CuSO4data, file = "CuSO4data_counts.Rdata")

CSSI_CuSO4_BSA <- foreach(i=unique(CuSO4data$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuSO4data, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_CuSO4_BSA, file = "CSSI_CuSO4_BSA.Rdata")

```


##Cycloheximide / Zeocin Data

```{r, eval = FALSE}
mydatatotest = "HKTMZDRX2.SortedCat.vcf.output.table"

CSSI_CycZeo <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CSSI_CycZeo$Dataset)

CSSI_CycZeo$Bulk <- NA
CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq"] <- "Cycloheximide"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "Dilute"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq"] <- "Zeocin"

CSSI_CycZeo$Parent <- NA
CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq"] <- "OakI"

CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "WineI"

CSSI_CycZeo$Parent <- factor(CSSI_CycZeo$Parent)
CSSI_CycZeo$Bulk <- factor(CSSI_CycZeo$Bulk)
CSSI_CycZeo$ReadCount <- as.numeric(CSSI_CycZeo$ReadCount)

head(CSSI_CycZeo)

save(CSSI_CycZeo, file = "CSSI_CycZeo_counts.Rdata")

#Separate into different datasets, since the subsequent steps will break if not
CSSI_Cycloheximide <- CSSI_CycZeo[CSSI_CycZeo$Bulk != "Zeocin",]
CSSI_Zeocin <- CSSI_CycZeo[CSSI_CycZeo$Bulk != "Cycloheximide",]

#Run BSA_GLM() and SmoothBSAWindows() for each
CSSI_Cyc_BSA <- foreach(i=unique(CSSI_Cycloheximide$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA, file = "CSSI_Cyc_BSA.Rdata")

CSSI_Zeo_BSA <- foreach(i=unique(CSSI_Zeocin$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA, file = "CSSI_Zeo_BSA.Rdata")
```

## Redoing Zeocin with replicates

Run Analysis

```{r, eval = FALSE}

CSSI_Zeocin_A <- CSSI_Zeocin[grep("A.fastq", CSSI_Zeocin$Dataset),]
CSSI_Zeocin_B <- CSSI_Zeocin[grep("B.fastq", CSSI_Zeocin$Dataset),]

CSSI_Zeo_BSA_A <- foreach(i=unique(CSSI_Zeocin_A$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin_A, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA_A, file = "CSSI_Zeo_BSA_A.Rdata")

CSSI_Zeo_BSA_B <- foreach(i=unique(CSSI_Zeocin_B$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin_B, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA_B, file = "CSSI_Zeo_BSA_B.Rdata")
```

Checking these

```{r}
colvalues = c("#345F6F", "#D7335C", "#FFB05C")

load("CSSI_Zeo_BSA_A.Rdata")
load("CSSI_Zeo_BSA_B.Rdata")

CSSI_Zeo_BSA_A$Rep = "A"
CSSI_Zeo_BSA_B$Rep = "B"

combined_Zeo <- rbind(CSSI_Zeo_BSA_A,CSSI_Zeo_BSA_B) 

combined_Zeo %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Bulk Zprime Zeocin")

combined_Zeo %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Interaction_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Interaction Zprime Zeocin")

combined_Zeo %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Parent_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Zeocin")

combined_Zeo %>% pivot_longer(c(Bulk_Zprime, Interaction_Zprime, Parent_Zprime), names_to = "Effect", values_to = "Zprime") %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Zprime, color = Effect, linetype = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + scale_color_manual(values = colvalues) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Zeocin")
```

## Redoing Cycloheximide with replicates

Run Analysis

```{r, eval = FALSE}

CSSI_Cycloheximide_A <- CSSI_Cycloheximide[grep("A.fastq", CSSI_Cycloheximide$Dataset),]
CSSI_Cycloheximide_B <- CSSI_Cycloheximide[grep("B.fastq", CSSI_Cycloheximide$Dataset),]

CSSI_Cyc_BSA_A <- foreach(i=unique(CSSI_Cycloheximide_A$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide_A, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA_A, file = "CSSI_Cyc_BSA_A.Rdata")

CSSI_Cyc_BSA_B <- foreach(i=unique(CSSI_Cycloheximide_B$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide_B, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA_B, file = "CSSI_Cyc_BSA_B.Rdata")
```

Checking these

```{r}
colvalues = c("#345F6F", "#D7335C", "#FFB05C")

load("CSSI_Cyc_BSA_A.Rdata")
load("CSSI_Cyc_BSA_B.Rdata")

CSSI_Cyc_BSA_A$Rep = "A"
CSSI_Cyc_BSA_B$Rep = "B"

combined_Cyc <- rbind(CSSI_Cyc_BSA_A,CSSI_Cyc_BSA_B) 

combined_Cyc %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Bulk Zprime Cycloheximide")

combined_Cyc %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Interaction_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Interaction Zprime Cycloheximide")

combined_Cyc %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Parent_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Cycloheximide")

combined_Cyc %>% pivot_longer(c(Bulk_Zprime, Interaction_Zprime, Parent_Zprime), names_to = "Effect", values_to = "Zprime") %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Zprime, color = Effect, linetype = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + scale_color_manual(values = colvalues) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Cycloheximide")
```


## Plotting

Load in Data

```{r}
load("CSSI_Fluc_BSA.Rdata")
load("CSSI_CuSO4_BSA.Rdata")

load("CSSI_Zeo_BSA.Rdata")
load("CSSI_Cyc_BSA.Rdata")
```


Plotting CSS I Fluconazole

```{r}
cybrPlotZPrime(CSSI_Fluc_BSA, columns = c("Bulk_Zprime"))

cybrPlotZScore(CSSI_Fluc_BSA)

cybrPlotZPrime(CSSI_Fluc_BSA, chromosomes = ChromKey$chromosomes[2:16])

cybrPlotZScore(CSSI_Fluc_BSA)
```
 
 Plotting CSS I CuSO4
 
```{r}
cybrPlotZPrime(CSSI_CuSO4_BSA, columns = c("Bulk_Zprime"))

cybrPlotZScore(CSSI_CuSO4_BSA)

cybrPlotZPrime(CSSI_CuSO4_BSA, chromosomes = ChromKey$chromosomes[2:16])

cybrPlotZScore(CSSI_CuSO4_BSA)
```

Plotting Zeocin

```{r}
cybrPlotZPrime(CSSI_Zeo_BSA, columns = "Bulk_Zprime")

cybrPlotZScore(CSSI_Zeo_BSA)

cybrPlotZPrime(CSSI_Zeo_BSA, chromosomes = ChromKey$chromosomes[2:16], title = "Zeocin Smoothed Z Scores")

cybrPlotZScore(CSSI_Zeo_BSA)

```

Plotting Cycloheximide

```{r}
cybrPlotZPrime(CSSI_Cyc_BSA, columns = "Bulk_Zprime")

cybrPlotZScore(CSSI_Cyc_BSA)

cybrPlotZPrime(CSSI_Cyc_BSA, chromosomes = ChromKey$chromosomes[2:16], title = "Cycloheximide Smoothed Z Scores")

cybrPlotZScore(CSSI_Cyc_BSA)

```

## Identify locations of peaks

```{r}
CSSI_CuSO4_Peaks <- data.frame(CHROM = unique(CSSI_CuSO4_BSA$CHROM),
                               peak.Bulk = NA)

for(i in 1:length(unique(CSSI_CuSO4_BSA$CHROM))){
  subsetdata <- subset(CSSI_CuSO4_BSA, CHROM == unique(CSSI_CuSO4_BSA$CHROM)[i])
  CSSI_CuSO4_Peaks$peak.Bulk[i] <- subsetdata$POS[abs(subsetdata$Bulk_Zprime) == max(abs(subsetdata$Bulk_Zprime))]
}

CSSI_CuSO4_Peaks

cybrPlotZPrime(CSSI_CuSO4_BSA, chromosomes = ChromKey$chromosomes[2:16]) + 
  geom_vline(data = CSSI_CuSO4_Peaks, aes(xintercept = peak.Bulk), color = "black", alpha = 0.5, size = 0.7, linetype = "dashed")


```

```{r, eval = FALSE}
CSSI_Fluc_peaks <- data.frame(CHROM = unique(CSSI_Fluc_BSA$CHROM),
                               peak.Bulk = NA,
                              peak.Interaction = NA,
                              peak.Parent = NA)

for(i in 1:length(unique(CSSI_Fluc_BSA$CHROM))){
  subsetdata <- subset(CSSI_Fluc_BSA, CHROM == unique(CSSI_Fluc_BSA$CHROM)[i])
  
  CSSI_Fluc_peaks$peak.Bulk[i] <- subsetdata$POS[abs(subsetdata$Bulk_Zprime) == max(abs(subsetdata$Bulk_Zprime))]
  CSSI_Fluc_peaks$peak.Interaction[i] <- subsetdata$POS[abs(subsetdata$Interaction_Zprime) == max(abs(subsetdata$Interaction_Zprime))]
  CSSI_Fluc_peaks$peak.Parent[i] <- subsetdata$POS[abs(subsetdata$Parent_Zprime) == max(abs(subsetdata$Parent_Zprime))]

}

CSSI_Fluc_peaks

cybrPlotZPrime(CSSI_Fluc_BSA, chromosomes = ChromKey$chromosomes[2:16]) + 
  geom_vline(data = CSSI_Fluc_peaks, aes(xintercept = peak.Bulk), color = "black", alpha = 0.5, size = 0.7, linetype = "dashed")


```

Find the genes nearest to each point

```{r, eval = FALSE}

colvalues = c("#345F6F", "#D7335C", "#FFB05C")

CSSI_Fluc_peaks %>% pivot_longer(c(peak.Bulk, peak.Interaction, peak.Parent), names_to = "Type", values_to = "POS") %>% mutate(Experiment = "CSSI_Fluc") %>% left_join(CSSI_Fluc_BSA, by = c("CHROM", "POS")) %>% select(-Index, -Bulk_Z, -Parent_Z, -Interaction_Z) -> CF

CSSI_CuSO4_Peaks %>% pivot_longer(c(peak.Bulk), names_to = "Type", values_to = "POS") %>% mutate(Experiment = "CSSI_CuSO4") %>% left_join(CSSI_CuSO4_BSA, by = c("CHROM", "POS")) %>% select(-Index, -Bulk_Z, -Parent_Z, -Interaction_Z) -> CC

rbind(CC, CF) -> Peaks_Combined

Peaks_Combined$Mag = NA
Peaks_Combined$Mag[Peaks_Combined$Type == "peak.Bulk"] <- Peaks_Combined$Bulk_Zprime[Peaks_Combined$Type == "peak.Bulk"] 
Peaks_Combined$Mag[Peaks_Combined$Type == "peak.Parent"] <- Peaks_Combined$Parent_Zprime[Peaks_Combined$Type == "peak.Parent"] 
Peaks_Combined$Mag[Peaks_Combined$Type == "peak.Interaction"] <- Peaks_Combined$Interaction_Zprime[Peaks_Combined$Type == "peak.Interaction"] 

cutoff <- max(na.omit(abs(Peaks_Combined$Mag[Peaks_Combined$Type == "peak.Parent"])))

Peaks_Combined %>% filter(CHROM != "M") %>% filter(abs(Mag) > cutoff) -> Shortlist

Shortlist %>% ggplot(aes(x = POS, y = Mag, shape = Experiment, color = Type)) + geom_point(size = 2) + scale_color_manual(values = c("#345F6F", "#D7335C", "#FFB05C")
) + facet_grid(~CHROM) + theme(axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

Shortlist %>% arrange(-abs(Mag)) %>% select(-Bulk_Zprime, -Parent_Zprime, -Interaction_Zprime) -> ShortlistArranged

#write.csv(ShortlistArranged, file = "Genes_CSSI_Experiments.csv")
getwd()
```


Picking out the actual SNPs from peak regions

```{r, eval = FALSE}

Shortlist %>% mutate(Lower = POS - 5000, Upper = POS + 5000) -> FilterPositions


# WindowPeaks <- foreach(i=1:length(FilterPositions$CHROM), .combine=rbind)%dopar%{
#   if(FilterPositions$Experiment[i] == "CSSI_CuSO4"){
#     stuff <- subset(CSSI_CuSO4_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
#     
#   }else{
#     stuff <- subset(CSSI_Fluc_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
#   }
#   stuff$Experiment <- FilterPositions$Experiment[i]
#   stuff
# }
# 
# WindowPeaks <- as.data.frame(WindowPeaks)

#WindowPeaks %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Experiment)) + geom_point(size = 2) + facet_wrap(~CHROM, scales = "free")

for(i in 1:length(FilterPositions$CHROM)){
  if(FilterPositions$Experiment[i] == "CSSI_CuSO4"){
    stuff <- subset(CSSI_CuSO4_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
  }else{
    stuff <- subset(CSSI_Fluc_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
  }
  
  if(FilterPositions$Type[i] == "peak.Bulk"){
    mycolor <- "blue"
    plot <- ggplot(stuff, aes(x = POS, y = Bulk_Z)) + 
    geom_vline(xintercept = FilterPositions$POS[i], alpha = 0.5, size = 2) + ylim(-10.5,10.5) +
    geom_point(size = 2, alpha = 0.5, color = mycolor) + ggtitle(paste(FilterPositions$Experiment[i], FilterPositions$CHROM[i]))
  }else{
    mycolor <- "red"
    plot <- ggplot(stuff, aes(x = POS, y = Interaction_Z)) + 
    geom_vline(xintercept = FilterPositions$POS[i], alpha = 0.5, size = 2) + ylim(-10.5,10.5) +
    geom_point(size = 2, alpha = 0.5, color = mycolor) + ggtitle(paste(FilterPositions$Experiment[i], FilterPositions$CHROM[i]))
  }
  

  print(plot)
}

```

```{r, eval = FALSE}
Shortlist %>% mutate(Lower = POS - 50000, Upper = POS + 50000) -> FilterPositions

for(i in 1:length(FilterPositions$CHROM)){
  if(FilterPositions$Experiment[i] == "CSSI_CuSO4"){
    stuff <- subset(CSSI_CuSO4_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
  }else{
    stuff <- subset(CSSI_Fluc_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
  }
  
  if(FilterPositions$Type[i] == "peak.Bulk"){
    mycolor <- "blue"
    plot <- ggplot(stuff, aes(x = POS, y = Bulk_Z)) + 
    geom_vline(xintercept = FilterPositions$POS[i], alpha = 0.5, size = 2) + ylim(-10.5,10.5) +
    geom_point(size = 2, alpha = 0.5, color = mycolor) + ggtitle(paste(FilterPositions$Experiment[i], FilterPositions$CHROM[i]))
  }else{
    mycolor <- "red"
    plot <- ggplot(stuff, aes(x = POS, y = Interaction_Z)) + 
    geom_vline(xintercept = FilterPositions$POS[i], alpha = 0.5, size = 2) + ylim(-10.5,10.5) +
    geom_point(size = 2, alpha = 0.5, color = mycolor) + ggtitle(paste(FilterPositions$Experiment[i], FilterPositions$CHROM[i]))
  }
  

  print(plot)
}

```

## Compare dilute samples for all combinations

We would expect the "dilute" group for each to not have a strong effect, so (1) find the difference in the four for the CycZeo experiment, and then (2) find the overall effect for each replicate and (3) find the overall effect for all groups at once.

```{r}
load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

CSSI_CycZeo %>% filter(Bulk == "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts
CSSI_Fluc %>% filter(Bulk == "Dilute") %>% mutate(Study = "Fluc") -> FCounts
CuSO4data %>% filter(Bulk == "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts

DiluteData <- rbind(CZCounts,FCounts,CuCounts)
DiluteData %>% group_by(CHROM, POS) %>% 
  summarise(CHROM = CHROM, POS = POS, studies = length(unique(Study))) %>% 
  distinct() %>% ungroup() -> DiluteDataCounts

DiluteDataCounts %>% filter(studies < 3) -> DiluteData_nonOverlaps

# ~8% of positions are not represented in ALL groups
dim(DiluteData_nonOverlaps)[1]/dim(DiluteData)[1]

#Filter out those which don't have all three
DiluteDataCounts %>% filter(studies == 3) %>% left_join(DiluteData) %>% ungroup() -> DiluteData_filtered

```

```{r, eval = FALSE}
################################################################################
#Rewrote functions to accommodate
#unique(DiluteData$Study)
CSSI_Dilute_BSA <- foreach(i=unique(DiluteData_filtered$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_filtered, chr = i, formula = "PAllele ~ Study") 
}

#CSSI_Dilute_BSA %>% cybrSmoothBSAWindows(chr = "VIX")
CSSI_Dilute_BSA_s <- foreach(i=unique(CSSI_Dilute_BSA$CHROM), .combine=rbind) %dopar%{
  cybrSmoothBSAWindows(CSSI_Dilute_BSA, chr = i) 
}

i <- "I"

mylist <- list()
#Chromosomes 6 and 9 don't work for some reason?
for(i in unique(CSSI_Dilute_BSA$CHROM)){
  mylist[[i]] <- cybrSmoothBSAWindows(CSSI_Dilute_BSA, chr = i) 
}

CSSI_Dilute_BSA_s <- bind_rows(mylist)
CSSI_Dilute_BSA_s$CHROM <- factor(CSSI_Dilute_BSA_s$CHROM, levels = ChromKey$chromosomes)
save(CSSI_Dilute_BSA_s, file = "CSSI_Dilute_BSA_s.Rdata")

################################################################################
CZDilute_BSA <- foreach(i=unique(CZCounts$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CZCounts, chr = i, formula = "PAllele ~ Dataset") %>% cybrSmoothBSAWindows()
}

save(CZDilute_BSA, file = "CZDilute_BSA.Rdata")

```


Plot these results

```{r}
load("CSSI_Dilute_BSA_s.Rdata")
DiluteData_nonOverlaps %>% select(-studies) %>% mutate('(Intercept)_Zprime' = NA, StudyCycZeo_Zprime = NA, StudyFluc_Zprime = NA, MissingData = 0) -> fillingindata
CSSI_Dilute_BSA_s %>% select(CHROM, POS, '(Intercept)_Zprime', StudyCycZeo_Zprime, StudyFluc_Zprime) %>% mutate(MissingData = NA) %>% rbind(fillingindata) -> CSSI_Dilute_BSA_u

CSSI_Dilute_BSA_u %>% pivot_longer(c('(Intercept)_Zprime', StudyCycZeo_Zprime, StudyFluc_Zprime, MissingData), names_to = "Factor", values_to = "Zprime") %>% 
  filter(Factor != "(Intercept)_Zprime") %>% 
  filter(CHROM != "I") %>%
  ggplot(aes(x = POS, y = Zprime, color = Factor)) + geom_point(alpha = 0.4, size = 0.5) + facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + scale_color_manual(values = c("black", "violet", "red")) +
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("All Studies Compared (Dilute)")
```


```{r}
load("CZDilute_BSA.Rdata")

CZDilute_BSA %>% filter(CHROM != "I") %>% pivot_longer(c('(Intercept)_Zprime', 
                                DatasetHKTMWDRX2_n01_ODB.fastq_Zprime,
                                DatasetHKTMWDRX2_n01_WDA.fastq_Zprime,
                                DatasetHKTMWDRX2_n01_WDB.fastq_Zprime), names_to = "Factor", values_to = "Zprime") %>%
  mutate(FactorName = gsub("DatasetHKTMWDRX2_n01_", "", .$Factor)) %>%
  filter(FactorName != "(Intercept)_Zprime") %>%
  ggplot(aes(x = POS, y = Zprime, color = FactorName)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("CycZeo Dilute Samples Compared")
```


## Plotting actual allele frequencies for each

I need to know what the logistic regression is doing here, so I'm going to plot each of the dilute samples in a different color and see if there's a weird trend there. That should help separate the effects of replicates as well as batches.

Load in just in case:
```{r}
load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

####################### Dilute #################################################
CSSI_CycZeo %>% filter(Bulk == "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts
CSSI_Fluc %>% filter(Bulk == "Dilute") %>% mutate(Study = "Fluc") -> FCounts
CuSO4data %>% filter(Bulk == "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts

DiluteData <- rbind(CZCounts,FCounts,CuCounts)

DatasetKey <- data.frame(Dataset = unique(DiluteData$Dataset),
                         DataName = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

DiluteData$Dataset <- as.character(DiluteData$Dataset)
DiluteData %>% left_join(DatasetKey, by = "Dataset") -> DiluteLabeled

####################### Selected ###############################################

CSSI_CycZeo %>% filter(Bulk != "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts_s
CSSI_Fluc %>% filter(Bulk != "Dilute") %>% mutate(Study = "Fluc") -> FCounts_s
CuSO4data %>% filter(Bulk != "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts_s

SelectedData <- rbind(CZCounts_s,FCounts_s,CuCounts_s)

DatasetKey_s <- data.frame(Dataset = unique(SelectedData$Dataset),
                         DataName = c("CZ_OCA", "CZ_OCB", "CZ_OZA", "CZ_OZB","CZ_WCA", "CZ_WCB","CZ_WZA", "CZ_WZB", "F_OFB", "F_OFC", "F_OFD", "F_WFC", "Cu_SA", "Cu_SC"))

SelectedData$Dataset <- as.character(SelectedData$Dataset)
SelectedData %>% left_join(DatasetKey_s, by = "Dataset") -> SelectedLabeled

####################### Colors #################################################

```

Pull out the allele frequency of each

```{r}
DatasetKey <- data.frame(Dataset = unique(DiluteData$Dataset),
                         DataName = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

DiluteData$Dataset <- as.character(DiluteData$Dataset)
DiluteData %>% left_join(DatasetKey, by = "Dataset") -> DiluteLabeled

DiluteLabeled$DataName <- factor(DiluteLabeled$DataName, 
                                   levels = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

mylabelcolorsd <- c("#91171F","#5C95FF",
                     "black", "#966B9D", "#A94C4C", "#2E1C2B", 
                     "#F2B880", "#46B1C9", "#3E517A", "#CBBAED",
                     "#8C2155", "#F87060", "#183059", "lightblue")
names(mylabelcolorsd) <- levels(DiluteLabeled$DataName)

DiluteLabeled %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.2) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Read Ratio by Dilute Dataset")

DiluteLabeled %>% filter(CHROM == "VII")  %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = Study, group = DataName)) + geom_point(alpha = 0.2) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() + scale_color_manual(values = c("skyblue", "violet", "red")) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Read Ratio by Dilute Dataset")

DiluteLabeled %>% filter(CHROM == "VII") %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.8) + 
  facet_grid(~Study, scales = "free_x", space = "free_x") + 
  theme_minimal() + scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Chr VII Read Ratio by Dilute Dataset")
```


```{r}
load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

CSSI_CycZeo %>% filter(Bulk != "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts_s
CSSI_Fluc %>% filter(Bulk != "Dilute") %>% mutate(Study = "Fluc") -> FCounts_s
CuSO4data %>% filter(Bulk != "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts_s

SelectedData <- rbind(CZCounts_s,FCounts_s,CuCounts_s)

DatasetKey_s <- data.frame(Dataset = unique(SelectedData$Dataset),
                         DataName = c("CZ_OCA", "CZ_OCB", "CZ_OZA", "CZ_OZB","CZ_WCA", "CZ_WCB","CZ_WZA", "CZ_WZB", "F_OFB", "F_OFC", "F_OFD", "F_WFC", "Cu_SA", "Cu_SC"))

SelectedData$Dataset <- as.character(SelectedData$Dataset)
SelectedData %>% left_join(DatasetKey_s, by = "Dataset") -> SelectedLabeled

SelectedLabeled$DataName <- factor(SelectedLabeled$DataName, 
                                   levels = c("CZ_OCA", "CZ_OCB", "CZ_OZA", "CZ_OZB","CZ_WCA", "CZ_WCB","CZ_WZA", "CZ_WZB", "F_OFB", "F_OFC", "F_OFD", "F_WFC", "Cu_SA", "Cu_SC"))

mylabelcolors <- c("#91171F","#5C95FF",
                     "black", "#966B9D", "#A94C4C", "#2E1C2B", 
                     "#F2B880", "#46B1C9", "#3E517A", "#CBBAED",
                     "#8C2155", "#F87060", "#183059", "lightblue")
names(mylabelcolors) <- levels(SelectedLabeled$DataName)

SelectedLabeled %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.2) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Read Ratio by Selected Dataset")

SelectedLabeled %>% filter(CHROM == "VII") %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.8) + 
  facet_grid(~Bulk) + ylim(-2,2) +
  theme_minimal() + scale_color_manual(name = "", values = mylabelcolors) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Chr VII Read Ratio by Selected Dataset")
```

