---
title: "CSS Analysis"
author: "Cassandra Buzby"
date: "10/27/2022"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

library(cybrBSA)

#ggplot2::theme_set(theme_light())
#ggplot2::theme_set(theme_bw())
ggplot2::theme_set(theme_cybr())


CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47",
                  "#F26430", "#0A369D", "#7EA3CC")


library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")
# load("RdataFiles/SAUA_4-13-22.Rdata")
# load("RdataFiles/SCUC_4-13-22.Rdata")
# load("RdataFiles/bsa_glm_results.Rdata")

#rawData = "HGV.SortedCat.vcf.output.table"
```

# Running cybrBSA
## Convert Parental Alleles

```{r, eval = FALSE}
parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

head(parentSNPids)
```

## Using cybrBSA package

CSS I Fluconazole

```{r, eval = FALSE}
mydatatotest = "HGV.SortedCat.vcf.output.table"

CSSI_Fluc <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CSSI_Fluc$Dataset)

CSSI_Fluc$Bulk <- NA
CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq"] <- "Dilute"

CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "Fluconazole"

CSSI_Fluc$Parent <- NA
CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq"] <- "OakI"

CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "WineI"

CSSI_Fluc$Parent <- factor(CSSI_Fluc$Parent)
CSSI_Fluc$Bulk <- factor(CSSI_Fluc$Bulk)
CSSI_Fluc$ReadCount <- as.numeric(CSSI_Fluc$ReadCount)

head(CSSI_Fluc)

save(CSSI_Fluc, file ="CSSI_Fluc_counts.Rdata")

CSSI_Fluc_BSA <- foreach(i=unique(CSSI_Fluc$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Fluc, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Fluc_BSA, file = "CSSI_Fluc_BSA.Rdata")
```


CSS I CuSO4

```{r, eval = FALSE}
CuSO4gatk <- "Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"

CuSO4data <- cybrInputGATKTable(CuSO4gatk) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CuSO4data$Dataset)

CuSO4data$Bulk <- NA
CuSO4data$Bulk[CuSO4data$Dataset == "SelectedA" |
              CuSO4data$Dataset == "SelectedC"] <- "CuSO4"

CuSO4data$Bulk[CuSO4data$Dataset == "UnselectedA" |
              CuSO4data$Dataset == "UnselectedC"] <- "Dilute"

CuSO4data$Parent <- NA
CuSO4data$Parent[CuSO4data$Dataset == "SelectedA" |
              CuSO4data$Dataset == "UnselectedA"] <- "OakI"

CuSO4data$Parent[CuSO4data$Dataset == "SelectedC" |
              CuSO4data$Dataset == "UnselectedC"] <- "WineI"

CuSO4data$Parent <- factor(CuSO4data$Parent)
CuSO4data$Bulk <- factor(CuSO4data$Bulk)
CuSO4data$ReadCount <- as.numeric(CuSO4data$ReadCount)

head(CuSO4data)

save(CuSO4data, file = "CuSO4data_counts.Rdata")

CSSI_CuSO4_BSA <- foreach(i=unique(CuSO4data$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuSO4data, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_CuSO4_BSA, file = "CSSI_CuSO4_BSA.Rdata")

```


## Cycloheximide / Zeocin Data

```{r, eval = FALSE}
mydatatotest = "HKTMZDRX2.SortedCat.vcf.output.table"

CSSI_CycZeo <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CSSI_CycZeo$Dataset)

CSSI_CycZeo$Bulk <- NA
CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq"] <- "Cycloheximide"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "Dilute"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq"] <- "Zeocin"

CSSI_CycZeo$Parent <- NA
CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq"] <- "OakI"

CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "WineI"

CSSI_CycZeo$Parent <- factor(CSSI_CycZeo$Parent)
CSSI_CycZeo$Bulk <- factor(CSSI_CycZeo$Bulk)
CSSI_CycZeo$ReadCount <- as.numeric(CSSI_CycZeo$ReadCount)

head(CSSI_CycZeo)

save(CSSI_CycZeo, file = "CSSI_CycZeo_counts.Rdata")

#Separate into different datasets, since the subsequent steps will break if not
CSSI_Cycloheximide <- CSSI_CycZeo[CSSI_CycZeo$Bulk != "Zeocin",]
CSSI_Zeocin <- CSSI_CycZeo[CSSI_CycZeo$Bulk != "Cycloheximide",]

#Run BSA_GLM() and SmoothBSAWindows() for each
CSSI_Cyc_BSA <- foreach(i=unique(CSSI_Cycloheximide$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA, file = "CSSI_Cyc_BSA.Rdata")

CSSI_Zeo_BSA <- foreach(i=unique(CSSI_Zeocin$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA, file = "CSSI_Zeo_BSA.Rdata")
```

## Redoing Zeocin with replicates

Run Analysis

```{r, eval = FALSE}

CSSI_Zeocin_A <- CSSI_Zeocin[grep("A.fastq", CSSI_Zeocin$Dataset),]
CSSI_Zeocin_B <- CSSI_Zeocin[grep("B.fastq", CSSI_Zeocin$Dataset),]

CSSI_Zeo_BSA_A <- foreach(i=unique(CSSI_Zeocin_A$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin_A, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA_A, file = "CSSI_Zeo_BSA_A.Rdata")

CSSI_Zeo_BSA_B <- foreach(i=unique(CSSI_Zeocin_B$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin_B, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA_B, file = "CSSI_Zeo_BSA_B.Rdata")
```

Checking these

```{r}
colvalues = c("#345F6F", "#D7335C", "#FFB05C")

load("CSSI_Zeo_BSA_A.Rdata")
load("CSSI_Zeo_BSA_B.Rdata")

CSSI_Zeo_BSA_A$Rep = "A"
CSSI_Zeo_BSA_B$Rep = "B"

combined_Zeo <- rbind(CSSI_Zeo_BSA_A,CSSI_Zeo_BSA_B) 

combined_Zeo %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Bulk Zprime Zeocin")

combined_Zeo %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Interaction_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Interaction Zprime Zeocin")

combined_Zeo %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Parent_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Zeocin")

combined_Zeo %>% pivot_longer(c(Bulk_Zprime, Interaction_Zprime, Parent_Zprime), names_to = "Effect", values_to = "Zprime") %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Zprime, color = Effect, linetype = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Zeo$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + scale_color_manual(values = colvalues) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Zeocin")
```

## Redoing Cycloheximide with replicates

Run Analysis

```{r, eval = FALSE}

CSSI_Cycloheximide_A <- CSSI_Cycloheximide[grep("A.fastq", CSSI_Cycloheximide$Dataset),]
CSSI_Cycloheximide_B <- CSSI_Cycloheximide[grep("B.fastq", CSSI_Cycloheximide$Dataset),]

CSSI_Cyc_BSA_A <- foreach(i=unique(CSSI_Cycloheximide_A$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide_A, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA_A, file = "CSSI_Cyc_BSA_A.Rdata")

CSSI_Cyc_BSA_B <- foreach(i=unique(CSSI_Cycloheximide_B$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide_B, chr = i) %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA_B, file = "CSSI_Cyc_BSA_B.Rdata")
```

Checking these

```{r}
colvalues = c("#345F6F", "#D7335C", "#FFB05C")

load("CSSI_Cyc_BSA_A.Rdata")
load("CSSI_Cyc_BSA_B.Rdata")

CSSI_Cyc_BSA_A$Rep = "A"
CSSI_Cyc_BSA_B$Rep = "B"

combined_Cyc <- rbind(CSSI_Cyc_BSA_A,CSSI_Cyc_BSA_B) 

combined_Cyc %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Bulk Zprime Cycloheximide")

combined_Cyc %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Interaction_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Interaction Zprime Cycloheximide")

combined_Cyc %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Parent_Zprime, color = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Cycloheximide")

combined_Cyc %>% pivot_longer(c(Bulk_Zprime, Interaction_Zprime, Parent_Zprime), names_to = "Effect", values_to = "Zprime") %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Zprime, color = Effect, linetype = Rep)) + geom_line() + scale_x_continuous(breaks = seq(from = 0, to = max(combined_Cyc$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + scale_color_manual(values = colvalues) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("Parent Zprime Cycloheximide")
```


## Plotting

Load in Data

```{r}
load("CSSI_Fluc_BSA.Rdata")
load("CSSI_CuSO4_BSA.Rdata")

load("CSSI_Zeo_BSA.Rdata")
load("CSSI_Cyc_BSA.Rdata")
```


Plotting CSS I Fluconazole

```{r}
cybrPlotZPrime(CSSI_Fluc_BSA, columns = c("Bulk_Zprime"))

cybrPlotZScore(CSSI_Fluc_BSA)

cybrPlotZPrime(CSSI_Fluc_BSA, chromosomes = ChromKey$chromosomes[2:16])

cybrPlotZScore(CSSI_Fluc_BSA)
```
 
 Plotting CSS I CuSO4
 
```{r}
cybrPlotZPrime(CSSI_CuSO4_BSA, columns = c("Bulk_Zprime"))

cybrPlotZScore(CSSI_CuSO4_BSA)

cybrPlotZPrime(CSSI_CuSO4_BSA, chromosomes = ChromKey$chromosomes[2:16])

cybrPlotZScore(CSSI_CuSO4_BSA)
```

Plotting Zeocin

```{r}
cybrPlotZPrime(CSSI_Zeo_BSA, columns = "Bulk_Zprime")

cybrPlotZScore(CSSI_Zeo_BSA)

cybrPlotZPrime(CSSI_Zeo_BSA, chromosomes = ChromKey$chromosomes[2:16], title = "Zeocin Smoothed Z Scores")

cybrPlotZScore(CSSI_Zeo_BSA)

```

Plotting Cycloheximide

```{r}
cybrPlotZPrime(CSSI_Cyc_BSA, columns = "Bulk_Zprime")

cybrPlotZScore(CSSI_Cyc_BSA)

cybrPlotZPrime(CSSI_Cyc_BSA, chromosomes = ChromKey$chromosomes[2:16], title = "Cycloheximide Smoothed Z Scores")

cybrPlotZScore(CSSI_Cyc_BSA)

```

## Identify locations of peaks

```{r}
CSSI_CuSO4_Peaks <- data.frame(CHROM = unique(CSSI_CuSO4_BSA$CHROM),
                               peak.Bulk = NA)

for(i in 1:length(unique(CSSI_CuSO4_BSA$CHROM))){
  subsetdata <- subset(CSSI_CuSO4_BSA, CHROM == unique(CSSI_CuSO4_BSA$CHROM)[i])
  CSSI_CuSO4_Peaks$peak.Bulk[i] <- subsetdata$POS[abs(subsetdata$Bulk_Zprime) == max(abs(subsetdata$Bulk_Zprime))]
}

CSSI_CuSO4_Peaks

cybrPlotZPrime(CSSI_CuSO4_BSA, chromosomes = ChromKey$chromosomes[2:16]) + 
  geom_vline(data = CSSI_CuSO4_Peaks, aes(xintercept = peak.Bulk), color = "black", alpha = 0.5, size = 0.7, linetype = "dashed")


```

```{r, eval = FALSE}
CSSI_Fluc_peaks <- data.frame(CHROM = unique(CSSI_Fluc_BSA$CHROM),
                               peak.Bulk = NA,
                              peak.Interaction = NA,
                              peak.Parent = NA)

for(i in 1:length(unique(CSSI_Fluc_BSA$CHROM))){
  subsetdata <- subset(CSSI_Fluc_BSA, CHROM == unique(CSSI_Fluc_BSA$CHROM)[i])
  
  CSSI_Fluc_peaks$peak.Bulk[i] <- subsetdata$POS[abs(subsetdata$Bulk_Zprime) == max(abs(subsetdata$Bulk_Zprime))]
  CSSI_Fluc_peaks$peak.Interaction[i] <- subsetdata$POS[abs(subsetdata$Interaction_Zprime) == max(abs(subsetdata$Interaction_Zprime))]
  CSSI_Fluc_peaks$peak.Parent[i] <- subsetdata$POS[abs(subsetdata$Parent_Zprime) == max(abs(subsetdata$Parent_Zprime))]

}

CSSI_Fluc_peaks

cybrPlotZPrime(CSSI_Fluc_BSA, chromosomes = ChromKey$chromosomes[2:16]) + 
  geom_vline(data = CSSI_Fluc_peaks, aes(xintercept = peak.Bulk), color = "black", alpha = 0.5, size = 0.7, linetype = "dashed")


```

Find the genes nearest to each point

```{r, eval = FALSE}

colvalues = c("#345F6F", "#D7335C", "#FFB05C")

CSSI_Fluc_peaks %>% pivot_longer(c(peak.Bulk, peak.Interaction, peak.Parent), names_to = "Type", values_to = "POS") %>% mutate(Experiment = "CSSI_Fluc") %>% left_join(CSSI_Fluc_BSA, by = c("CHROM", "POS")) %>% select(-Index, -Bulk_Z, -Parent_Z, -Interaction_Z) -> CF

CSSI_CuSO4_Peaks %>% pivot_longer(c(peak.Bulk), names_to = "Type", values_to = "POS") %>% mutate(Experiment = "CSSI_CuSO4") %>% left_join(CSSI_CuSO4_BSA, by = c("CHROM", "POS")) %>% select(-Index, -Bulk_Z, -Parent_Z, -Interaction_Z) -> CC

rbind(CC, CF) -> Peaks_Combined

Peaks_Combined$Mag = NA
Peaks_Combined$Mag[Peaks_Combined$Type == "peak.Bulk"] <- Peaks_Combined$Bulk_Zprime[Peaks_Combined$Type == "peak.Bulk"] 
Peaks_Combined$Mag[Peaks_Combined$Type == "peak.Parent"] <- Peaks_Combined$Parent_Zprime[Peaks_Combined$Type == "peak.Parent"] 
Peaks_Combined$Mag[Peaks_Combined$Type == "peak.Interaction"] <- Peaks_Combined$Interaction_Zprime[Peaks_Combined$Type == "peak.Interaction"] 

cutoff <- max(na.omit(abs(Peaks_Combined$Mag[Peaks_Combined$Type == "peak.Parent"])))

Peaks_Combined %>% filter(CHROM != "M") %>% filter(abs(Mag) > cutoff) -> Shortlist

Shortlist %>% ggplot(aes(x = POS, y = Mag, shape = Experiment, color = Type)) + geom_point(size = 2) + scale_color_manual(values = c("#345F6F", "#D7335C", "#FFB05C")
) + facet_grid(~CHROM) + theme(axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

Shortlist %>% arrange(-abs(Mag)) %>% select(-Bulk_Zprime, -Parent_Zprime, -Interaction_Zprime) -> ShortlistArranged

#write.csv(ShortlistArranged, file = "Genes_CSSI_Experiments.csv")
getwd()
```


Picking out the actual SNPs from peak regions

```{r, eval = FALSE}

Shortlist %>% mutate(Lower = POS - 5000, Upper = POS + 5000) -> FilterPositions


# WindowPeaks <- foreach(i=1:length(FilterPositions$CHROM), .combine=rbind)%dopar%{
#   if(FilterPositions$Experiment[i] == "CSSI_CuSO4"){
#     stuff <- subset(CSSI_CuSO4_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
#     
#   }else{
#     stuff <- subset(CSSI_Fluc_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
#   }
#   stuff$Experiment <- FilterPositions$Experiment[i]
#   stuff
# }
# 
# WindowPeaks <- as.data.frame(WindowPeaks)

#WindowPeaks %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Experiment)) + geom_point(size = 2) + facet_wrap(~CHROM, scales = "free")

for(i in 1:length(FilterPositions$CHROM)){
  if(FilterPositions$Experiment[i] == "CSSI_CuSO4"){
    stuff <- subset(CSSI_CuSO4_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
  }else{
    stuff <- subset(CSSI_Fluc_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
  }
  
  if(FilterPositions$Type[i] == "peak.Bulk"){
    mycolor <- "blue"
    plot <- ggplot(stuff, aes(x = POS, y = Bulk_Z)) + 
    geom_vline(xintercept = FilterPositions$POS[i], alpha = 0.5, size = 2) + ylim(-10.5,10.5) +
    geom_point(size = 2, alpha = 0.5, color = mycolor) + ggtitle(paste(FilterPositions$Experiment[i], FilterPositions$CHROM[i]))
  }else{
    mycolor <- "red"
    plot <- ggplot(stuff, aes(x = POS, y = Interaction_Z)) + 
    geom_vline(xintercept = FilterPositions$POS[i], alpha = 0.5, size = 2) + ylim(-10.5,10.5) +
    geom_point(size = 2, alpha = 0.5, color = mycolor) + ggtitle(paste(FilterPositions$Experiment[i], FilterPositions$CHROM[i]))
  }
  

  print(plot)
}

```

```{r, eval = FALSE}
Shortlist %>% mutate(Lower = POS - 50000, Upper = POS + 50000) -> FilterPositions

for(i in 1:length(FilterPositions$CHROM)){
  if(FilterPositions$Experiment[i] == "CSSI_CuSO4"){
    stuff <- subset(CSSI_CuSO4_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
  }else{
    stuff <- subset(CSSI_Fluc_BSA, CHROM == FilterPositions$CHROM[i] & POS > FilterPositions$Lower[i] & POS < FilterPositions$Upper[i])
  }
  
  if(FilterPositions$Type[i] == "peak.Bulk"){
    mycolor <- "blue"
    plot <- ggplot(stuff, aes(x = POS, y = Bulk_Z)) + 
    geom_vline(xintercept = FilterPositions$POS[i], alpha = 0.5, size = 2) + ylim(-10.5,10.5) +
    geom_point(size = 2, alpha = 0.5, color = mycolor) + ggtitle(paste(FilterPositions$Experiment[i], FilterPositions$CHROM[i]))
  }else{
    mycolor <- "red"
    plot <- ggplot(stuff, aes(x = POS, y = Interaction_Z)) + 
    geom_vline(xintercept = FilterPositions$POS[i], alpha = 0.5, size = 2) + ylim(-10.5,10.5) +
    geom_point(size = 2, alpha = 0.5, color = mycolor) + ggtitle(paste(FilterPositions$Experiment[i], FilterPositions$CHROM[i]))
  }
  

  print(plot)
}

```

## Compare dilute samples for all combinations

We would expect the "dilute" group for each to not have a strong effect, so (1) find the difference in the four for the CycZeo experiment, and then (2) find the overall effect for each replicate and (3) find the overall effect for all groups at once.

```{r}
load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

CSSI_CycZeo %>% filter(Bulk == "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts
CSSI_Fluc %>% filter(Bulk == "Dilute") %>% mutate(Study = "Fluc") -> FCounts
CuSO4data %>% filter(Bulk == "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts

DiluteData <- rbind(CZCounts,FCounts,CuCounts)

save(DiluteData, file = "DiluteData.Rdata")

DiluteData %>% group_by(CHROM, POS) %>% 
  summarise(CHROM = CHROM, POS = POS, studies = length(unique(Study))) %>% 
  distinct() %>% ungroup() -> DiluteDataCounts

DiluteDataCounts %>% filter(studies < 3) -> DiluteData_nonOverlaps

# ~8% of positions are not represented in ALL groups
dim(DiluteData_nonOverlaps)[1]/dim(DiluteData)[1]

#Filter out those which don't have all three
DiluteDataCounts %>% filter(studies == 3) %>% left_join(DiluteData) %>% ungroup() -> DiluteData_filtered

save(DiluteData_nonOverlaps, file = "DiluteData_nonOverlaps.Rdata")
```

```{r, eval = FALSE}
################################################################################
#Rewrote functions to accommodate
#unique(DiluteData$Study)
CSSI_Dilute_BSA <- foreach(i=unique(DiluteData_filtered$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_filtered, chr = i, formula = "PAllele ~ Study") 
}

#CSSI_Dilute_BSA %>% cybrSmoothBSAWindows(chr = "VIX")
CSSI_Dilute_BSA_s <- foreach(i=unique(CSSI_Dilute_BSA$CHROM), .combine=rbind) %dopar%{
  cybrSmoothBSAWindows(CSSI_Dilute_BSA, chr = i) 
}

i <- "I"

mylist <- list()
#Chromosomes 6 and 9 don't work for some reason?
for(i in unique(CSSI_Dilute_BSA$CHROM)){
  mylist[[i]] <- cybrSmoothBSAWindows(CSSI_Dilute_BSA, chr = i) 
}

CSSI_Dilute_BSA_s <- bind_rows(mylist)
CSSI_Dilute_BSA_s$CHROM <- factor(CSSI_Dilute_BSA_s$CHROM, levels = ChromKey$chromosomes)
save(CSSI_Dilute_BSA_s, file = "CSSI_Dilute_BSA_s.Rdata")

################################################################################
CZDilute_BSA <- foreach(i=unique(CZCounts$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CZCounts, chr = i, formula = "PAllele ~ Dataset") %>% cybrSmoothBSAWindows()
}

save(CZDilute_BSA, file = "CZDilute_BSA.Rdata")

```


Plot these results

```{r}
load("CSSI_Dilute_BSA_s.Rdata")
DiluteData_nonOverlaps %>% select(-studies) %>% mutate('(Intercept)_Zprime' = NA, StudyCycZeo_Zprime = NA, StudyFluc_Zprime = NA, MissingData = 0) -> fillingindata
CSSI_Dilute_BSA_s %>% select(CHROM, POS, '(Intercept)_Zprime', StudyCycZeo_Zprime, StudyFluc_Zprime) %>% mutate(MissingData = NA) %>% rbind(fillingindata) -> CSSI_Dilute_BSA_u

CSSI_Dilute_BSA_u %>% pivot_longer(c('(Intercept)_Zprime', StudyCycZeo_Zprime, StudyFluc_Zprime, MissingData), names_to = "Factor", values_to = "Zprime") %>% 
  filter(Factor != "(Intercept)_Zprime") %>% 
  filter(CHROM != "I") %>%
  ggplot(aes(x = POS, y = Zprime, color = Factor)) + geom_point(alpha = 0.4, size = 0.5) + facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + scale_color_manual(values = c("black", "violet", "red")) +
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("All Studies Compared (Dilute)")
```


```{r}
load("CZDilute_BSA.Rdata")

CZDilute_BSA %>% filter(CHROM != "I") %>% pivot_longer(c('(Intercept)_Zprime', 
                                DatasetHKTMWDRX2_n01_ODB.fastq_Zprime,
                                DatasetHKTMWDRX2_n01_WDA.fastq_Zprime,
                                DatasetHKTMWDRX2_n01_WDB.fastq_Zprime), names_to = "Factor", values_to = "Zprime") %>%
  mutate(FactorName = gsub("DatasetHKTMWDRX2_n01_", "", .$Factor)) %>%
  filter(FactorName != "(Intercept)_Zprime") %>%
  ggplot(aes(x = POS, y = Zprime, color = FactorName)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Zprime") + ggtitle("CycZeo Dilute Samples Compared")
```


## Plotting actual allele frequencies for each

I need to know what the logistic regression is doing here, so I'm going to plot each of the dilute samples in a different color and see if there's a weird trend there. That should help separate the effects of replicates as well as batches.

Load in just in case:
```{r}
load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

####################### Dilute #################################################
CSSI_CycZeo %>% filter(Bulk == "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts
CSSI_Fluc %>% filter(Bulk == "Dilute") %>% mutate(Study = "Fluc") -> FCounts
CuSO4data %>% filter(Bulk == "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts

DiluteData <- rbind(CZCounts,FCounts,CuCounts)

DatasetKey <- data.frame(Dataset = unique(DiluteData$Dataset),
                         DataName = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

DiluteData$Dataset <- as.character(DiluteData$Dataset)
DiluteData %>% left_join(DatasetKey, by = "Dataset") -> DiluteLabeled

####################### Selected ###############################################

CSSI_CycZeo %>% filter(Bulk != "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts_s
CSSI_Fluc %>% filter(Bulk != "Dilute") %>% mutate(Study = "Fluc") -> FCounts_s
CuSO4data %>% filter(Bulk != "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts_s

SelectedData <- rbind(CZCounts_s,FCounts_s,CuCounts_s)

DatasetKey_s <- data.frame(Dataset = unique(SelectedData$Dataset),
                         DataName = c("CZ_OCA", "CZ_OCB", "CZ_OZA", "CZ_OZB","CZ_WCA", "CZ_WCB","CZ_WZA", "CZ_WZB", "F_OFB", "F_OFC", "F_OFD", "F_WFC", "Cu_SA", "Cu_SC"))

SelectedData$Dataset <- as.character(SelectedData$Dataset)
SelectedData %>% left_join(DatasetKey_s, by = "Dataset") -> SelectedLabeled

####################### Colors #################################################

```

Pull out the allele frequency of each

```{r}

load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

####################### Dilute #################################################
CSSI_CycZeo %>% filter(Bulk == "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts
CSSI_Fluc %>% filter(Bulk == "Dilute") %>% mutate(Study = "Fluc") -> FCounts
CuSO4data %>% filter(Bulk == "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts

DiluteData <- rbind(CZCounts,FCounts,CuCounts)

DatasetKey <- data.frame(Dataset = unique(DiluteData$Dataset),
                         DataName = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

DiluteData$Dataset <- as.character(DiluteData$Dataset)
DiluteData %>% left_join(DatasetKey, by = "Dataset") -> DiluteLabeled
 
DatasetKey <- data.frame(Dataset = unique(DiluteData$Dataset),
                         DataName = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

DiluteData$Dataset <- as.character(DiluteData$Dataset)
DiluteData %>% left_join(DatasetKey, by = "Dataset") -> DiluteLabeled

DiluteLabeled$DataName <- factor(DiluteLabeled$DataName, 
                                   levels = c("CZ_ODA", "CZ_ODB", "CZ_WDA", "CZ_WDB", "F_ODB", "F_WDB", "F_WDC", "F_WDD", "Cu_DA", "Cu_DC"))

mylabelcolorsd <- c("#91171F","#5C95FF",
                     "black", "#966B9D", "#A94C4C", "#2E1C2B", 
                     "#F2B880", "#46B1C9", "#3E517A", "#CBBAED",
                     "#8C2155", "#F87060", "#183059", "lightblue")
names(mylabelcolorsd) <- levels(DiluteLabeled$DataName)

# DiluteLabeled %>% select(-AltRef_Allele) %>% 
#     pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
#     ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.2) + 
#   facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
#   theme_minimal() +
#       theme(legend.position = "bottom", axis.text.x=element_blank(),
#             axis.ticks.x=element_blank()) + 
#       xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Read Ratio by Dilute Dataset")
# 
# DiluteLabeled %>% filter(CHROM == "VII")  %>% select(-AltRef_Allele) %>% 
#     pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
#     ggplot(aes(x = POS, y = log(Wine/Oak), color = Study, group = DataName)) + geom_point(alpha = 0.2) + 
#   facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
#   theme_minimal() + scale_color_manual(values = c("skyblue", "violet", "red")) + 
#       theme(legend.position = "bottom", axis.text.x=element_blank(),
#             axis.ticks.x=element_blank()) + 
#       xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Read Ratio by Dilute Dataset")

DiluteLabeled %>% filter(CHROM == "VII") %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.8) + 
  facet_grid(~Study, scales = "free_x", space = "free_x") + ylim(-2,2) + 
  theme_minimal() + scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Chr VII Read Ratio by Dilute Dataset")


#Smooth this dilute measure

DiluteLabeled %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
  select(CHROM, POS, Study, DataName, Oak, Wine) %>%
  mutate(logRatio = log(Wine/Oak)) %>%
  arrange(CHROM, POS) -> DiluteLogRatio

DiluteLogRatio$logRatio[DiluteLogRatio$logRatio == "Inf" | DiluteLogRatio$logRatio == "-Inf"] <- NA

save(DiluteLogRatio, file = "DiluteLogRatio.Rdata")
```

```{r}
#DiluteLogRatio %>% na.omit() %>% select(-Study, -Oak, -Wine) -> DiluteLogRatio

# smoothedlist <- list()
# 
# for(i in unique(DiluteLogRatio$DataName)){
#   smoothedlist[[i]] <-foreach(c=unique(DiluteLogRatio$CHROM), .combine=rbind) %dopar%{
#     cybrSmoothBSAWindows(Results = DiluteLogRatio[DiluteLogRatio$DataName == i,!names(DiluteLogRatio) %in% "DataName"], chr = c)
#   } 
#   smoothedlist[[i]]$DataName <- i
# }
# 
# DiluteLogRatio_ZSmooth <- bind_rows(smoothedlist)
# 
# save(DiluteLogRatio_ZSmooth, file = "DiluteLogRatio_ZSmooth.Rdata")
################################################################################
smoothedlist <- list()
# DiluteLogRatio$Oak <- DiluteLogRatio$Oak + 1
# DiluteLogRatio$Wine <- DiluteLogRatio$Wine + 1
# DiluteLogRatio$logRatio <- log(DiluteLogRatio$Wine/DiluteLogRatio$Oak)

for(i in unique(DiluteLogRatio$DataName)){
  smoothedlist[[i]] <-foreach(c=unique(DiluteLogRatio$CHROM), .combine=rbind) %dopar%{
    cybrSmoothBSAWindows(Results = DiluteLogRatio[DiluteLogRatio$DataName == i,!names(DiluteLogRatio) %in% "DataName"], chr = c)
  } 
  smoothedlist[[i]]$DataName <- i
}

DiluteLogRatio_ZSmooth_p1 <- bind_rows(smoothedlist)

save(DiluteLogRatio_ZSmooth_p1, file = "DiluteLogRatio_ZSmooth_p1.Rdata")
################################################################################
smoothedlist <- list()
# DiluteLogRatio$Oak <- DiluteLogRatio$Oak + 1
# DiluteLogRatio$Wine <- DiluteLogRatio$Wine + 1
# DiluteLogRatio$logRatio <- log(DiluteLogRatio$Wine/DiluteLogRatio$Oak)

for(i in unique(DiluteLogRatio$DataName)){
  smoothedlist[[i]] <-foreach(c=unique(DiluteLogRatio$CHROM), .combine=rbind) %dopar%{
    cybrSmoothBSAWindows_Med(Results = DiluteLogRatio[DiluteLogRatio$DataName == i,!names(DiluteLogRatio) %in% "DataName"], chr = c)
  } 
  smoothedlist[[i]]$DataName <- i
}

DiluteLogRatio_ZSmooth_med <- bind_rows(smoothedlist)

save(DiluteLogRatio_ZSmooth_med, file = "DiluteLogRatio_ZSmooth_med.Rdata")
```

Plotting smoothed without correcting for counts

```{r}
load("DiluteLogRatio_ZSmooth_med.Rdata")
load("DiluteLogRatio_ZSmooth.Rdata")
load("DiluteLogRatio.Rdata")
DiluteLogRatio_ZSmooth_med$CHROM <- factor(as.roman(DiluteLogRatio_ZSmooth$CHROM), levels = ChromKey$chromosomes)
DiluteLogRatio_ZSmooth_med <- left_join(DiluteLogRatio_ZSmooth_med, DiluteLogRatio[,1:4])

DiluteLogRatio_ZSmooth_med %>% ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-0.5,0.5)+
  theme_minimal() + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Log Ratio") + 
  ggtitle("All Smoothed ")

DiluteLogRatio_ZSmooth_med %>% filter(CHROM == "VII") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_point(alpha = 0.1, size = 1) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() + 
      theme(legend.position = "bottom",
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Log Ratio") + 
  ggtitle("All Smoothed ")

DiluteLogRatio_ZSmooth_med %>% filter(CHROM == "V") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_point(alpha = 0.1, size = 1) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() + 
      theme(legend.position = "bottom",
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Log Ratio") + 
  ggtitle("All Smoothed ")
```

Doing this with adding 1 to every count

```{r}
#Load Data
load("DiluteLogRatio_ZSmooth_p1.Rdata")

#Reformat saved data for these plots
DiluteLogRatio_ZSmooth_p1$CHROM <- factor(as.roman(DiluteLogRatio_ZSmooth_p1$CHROM), levels = ChromKey$chromosomes)
DiluteLogRatio_ZSmooth_p1 <- left_join(DiluteLogRatio_ZSmooth_p1, DiluteLogRatio[,1:4])

DiluteLogRatio_ZSmooth_p1 %>% select(-Study_Z, -Study_Zprime) %>% mutate(Oak = Oak_Z - 1, Wine = Wine_Z - 1) -> DLR_Smooth_Intermediate

DiluteLogRatio_ZSmooth_p1$Zeros <- NA
DiluteLogRatio_ZSmooth_p1$Zeros[DLR_Smooth_Intermediate$Oak == 0 | DLR_Smooth_Intermediate$Wine == 0] <- "Inf"

sum(DiluteLogRatio_ZSmooth_p1$Zeros == "Inf", na.rm = TRUE)

#Plot log ratios
DiluteLogRatio_ZSmooth_p1 %>% ggplot(aes(x = POS, y = logRatio_Zprime, color = Study, group = DataName)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-0.5,0.5)+ geom_hline(yintercept = 0, color = "black")+
  theme_minimal() + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("All Dilute Smoothed Log Wine/Oak")

DiluteLogRatio_ZSmooth_p1 %>% filter(CHROM == "VII") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study)) + geom_point(alpha = 0.1, size = 0.7) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal() + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Chr VII Smoothed Log Ratios")

DiluteLogRatio_ZSmooth_p1 %>% filter(CHROM == "V") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = Study)) + geom_point(alpha = 0.1, size = 0.7) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal() + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Chr V Smoothed Log Ratios")

```

```{r}
DiluteLogRatio_ZSmooth_p1 %>% filter(CHROM == "VII") %>% filter(logRatio_Zprime > 0.3) -> VII_Peak

ggplot(VII_Peak, aes(x = POS)) + geom_density()

VII_Peak %>% filter(POS < 5e5) %>% ggplot(aes(x = POS, y = logRatio_Zprime, color = Study)) + geom_point(alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal() 

DiluteLogRatio_ZSmooth_p1 %>% filter(CHROM == "VII", POS < 500000, POS > 300000, Study == "CuSO4") %>% 
  ggplot(aes(x = POS, y = logRatio_Zprime, color = DataName)) + geom_point(alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal()

DiluteLogRatio %>% filter(CHROM == "VII", POS < 500000, POS > 300000, Study == "CuSO4") %>% 
  ggplot(aes(x = POS, y = logRatio, color = DataName)) + geom_point(alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + geom_hline(yintercept = 0, color = "black")+
  theme_minimal()

```


### Same for selected bulks

```{r}
load("CSSI_CycZeo_counts.Rdata")
load("CSSI_Fluc_counts.Rdata")
load("CuSO4data_counts.Rdata")

CSSI_CycZeo %>% filter(Bulk != "Dilute") %>% mutate(Study = "CycZeo") -> CZCounts_s
CSSI_Fluc %>% filter(Bulk != "Dilute") %>% mutate(Study = "Fluc") -> FCounts_s
CuSO4data %>% filter(Bulk != "Dilute") %>% mutate(Study = "CuSO4") -> CuCounts_s

SelectedData <- rbind(CZCounts_s,FCounts_s,CuCounts_s)

DatasetKey_s <- data.frame(Dataset = unique(SelectedData$Dataset),
                         DataName = c("CZ_OCA", "CZ_OCB", "CZ_OZA", "CZ_OZB","CZ_WCA", "CZ_WCB","CZ_WZA", "CZ_WZB", "F_OFB", "F_OFC", "F_OFD", "F_WFC", "Cu_SA", "Cu_SC"))

SelectedData$Dataset <- as.character(SelectedData$Dataset)
SelectedData %>% left_join(DatasetKey_s, by = "Dataset") -> SelectedLabeled

SelectedLabeled$DataName <- factor(SelectedLabeled$DataName, 
                                   levels = c("CZ_OCA", "CZ_OCB", "CZ_OZA", "CZ_OZB","CZ_WCA", "CZ_WCB","CZ_WZA", "CZ_WZB", "F_OFB", "F_OFC", "F_OFD", "F_WFC", "Cu_SA", "Cu_SC"))

mylabelcolors <- c("#91171F","#5C95FF",
                     "black", "#966B9D", "#A94C4C", "#2E1C2B", 
                     "#F2B880", "#46B1C9", "#3E517A", "#CBBAED",
                     "#8C2155", "#F87060", "#183059", "lightblue")
names(mylabelcolors) <- levels(SelectedLabeled$DataName)

SelectedLabeled %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.2) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Read Ratio by Selected Dataset")

SelectedLabeled %>% filter(CHROM == "VII") %>% select(-AltRef_Allele) %>% 
    pivot_wider(names_from = PAllele, values_from = ReadCount) %>% 
    ggplot(aes(x = POS, y = log(Wine/Oak), color = DataName)) + geom_point(alpha = 0.2, size = 0.8) + 
  facet_grid(~Bulk) + ylim(-2,2) +
  theme_minimal() + scale_color_manual(name = "", values = mylabelcolors) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("log(Wine/Oak) Reads") + ggtitle("Chr VII Read Ratio by Selected Dataset")
```

## Redoing analysis of each bulk together

Instead of having them be cofactors, try each paired? Or each replicate paired?  

If you use replicate as a cofactor, does each need to be in a different category? If I pair them as A:Z rather than A/B/A/B would that work? As long as there isn't just one for each, that's correct...? Right?

```{r, eval = FALSE}

DiluteData_filtered$CycZeo = ifelse(DiluteData_filtered$Study == "CycZeo", 1, 0)
DiluteData_filtered$CuSO4 = ifelse(DiluteData_filtered$Study == "CuSO4", 1, 0)
DiluteData_filtered$Fluc = ifelse(DiluteData_filtered$Study == "Fluc", 1, 0)

#issue of 'names' attribute [6] must be the same length as the vector [5] is in cybrBSA_GLM :(
#turns out it's because two columns are colinear, so it's returning an NA; this is soled by running each on its own as a binary

Dilute_BSA_CycZeo <- foreach(i=unique(DiluteData_filtered$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_filtered, chr = i, formula = "PAllele ~ CycZeo")# %>% cybrSmoothBSAWindows()
}

Dilute_BSA_CuSO4 <- foreach(i=unique(DiluteData_filtered$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_filtered, chr = i, formula = "PAllele ~ CuSO4") #%>% cybrSmoothBSAWindows()
}

Dilute_BSA_Fluc <- foreach(i=unique(DiluteData_filtered$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_filtered, chr = i, formula = "PAllele ~ Fluc") #%>% cybrSmoothBSAWindows()
}

# MERGE DATA TOGETHER
Dilute_BSA_CycZeo %>% select(-'(Intercept)') %>% merge(Dilute_BSA_CuSO4) %>% 
  select(-'(Intercept)') %>% merge(Dilute_BSA_Fluc) %>% 
  select(-'(Intercept)') -> Dilute_BSA_byParent

# RUN SMOOTHING
DiluteBSA_smoothed <- foreach(i=unique(Dilute_BSA_byParent$CHROM), .combine=rbind) %dopar%{
  Dilute_BSA_byParent %>% cybrSmoothBSAWindows2(chr = i)
}
DiluteBSA_smoothed$CHROM <- factor(DiluteBSA_smoothed$CHROM, levels = ChromKey$chromosomes)

save(DiluteBSA_smoothed, file = "DiluteBSA_smoothed.Rdata")

```

```{r}
load("DiluteBSA_smoothed.Rdata")
DiluteBSA_smoothed$CHROM <- factor(DiluteBSA_smoothed$CHROM, levels = ChromKey$chromosomes)

#Combining again with missing data
load("DiluteData_nonOverlaps.Rdata")

DiluteData_nonOverlaps %>% select(-studies) %>% transmute(CHROM = CHROM, Index = NA, POS = POS, 
                                                          CycZeo_Z = NA, CuSO4_Z = NA, Fluc_Z = NA, 
                                                       CycZeo_Zprime = NA, CuSO4_Zprime = NA, Fluc_Zprime = NA,
                                                       MissingData = 0) -> fillingindata

DiluteBSA_smoothed %>% mutate(MissingData = NA) %>% rbind(fillingindata) -> DiluteBSA_smoothed_u

################################################################################

# PIVOT FOR PLOTTING
DiluteBSA_smoothed %>% filter(CHROM != "I", CHROM != "VI", CHROM != "XI", CHROM != "V") %>% pivot_longer(c(CycZeo_Zprime, CuSO4_Zprime, Fluc_Zprime), names_to = "Experiment", values_to = "Z_Prime") %>%
  ggplot(aes(x = POS, y = Z_Prime, color = Experiment)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-5,5) +
  theme_minimal() + #scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Smoothed Z Score") + ggtitle("Effects of Dilute Samples (Excluded Chr I/V/VI/XI)")

DiluteBSA_smoothed_u %>% filter(CHROM != "I", CHROM != "VI", CHROM != "XI", CHROM != "V") %>% pivot_longer(c(CycZeo_Zprime, CuSO4_Zprime, Fluc_Zprime), names_to = "Experiment", values_to = "Z_Prime") %>%
  ggplot(aes(x = POS, y = Z_Prime, color = Experiment)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-5,5) +
  theme_minimal() + #scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Smoothed Z Score") + ggtitle("Effects of Dilute Samples (Excluded Chr I/V/VI/XI)")

DiluteBSA_smoothed_u %>% filter(CHROM != "I", CHROM != "VI", CHROM != "XI", CHROM != "V") %>% 
   pivot_longer(c(CycZeo_Z, CuSO4_Z, Fluc_Z), names_to = "Experiment", values_to = "Z_Score") %>%
  ggplot(aes(x = POS, y = Z_Score, color = Experiment)) + geom_point(size = 0.8, alpha = 0.3) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() + #scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + ggtitle("Effects of Dilute Samples (Excluded Chr I/V/VI/XI)")

DiluteBSA_smoothed_u %>% filter(CHROM != "I", CHROM != "VI", CHROM != "XI", CHROM != "V")%>% pivot_longer(c(CycZeo_Z, CuSO4_Z, Fluc_Z), names_to = "Experiment", values_to = "Z_Score") %>%
  ggplot(aes(x = POS, y = Z_Score, color = Experiment)) + geom_point(size = 0.8, alpha = 0.3) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  theme_minimal() + #scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + ggtitle("Effects of Dilute Samples (Excluded Chr I/V/VI/XI)")


DiluteBSA_smoothed_u %>% pivot_longer(c(CycZeo_Z, CuSO4_Z, Fluc_Z), names_to = "Experiment", values_to = "Z_Score") %>%
  ggplot(aes(x = POS, y = Z_Score, color = Experiment)) + geom_point(size = 0.8, alpha = 0.3) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-5,5) +
  theme_minimal() + #scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + ggtitle("Effects of Dilute Samples (Excluded Chr I/V/VI/XI)")

```

## Troubleshooting

```{r, eval = FALSE}
chr = "I"
c = chr
lrP = DiluteData_filtered
formula = "PAllele ~ CuSO4"
i = unique(lrP$POS)[1]
function(lrP, chr = "II", windowsize = 100, formula = "PAllele ~ Bulk*Parent"){

  lrP <- subset(lrP, CHROM == chr)

  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- suppressWarnings(glm(as.formula(formula), weights = ReadCount, family = binomial, data = lrP[lrP$POS == i,]))

      #Output of foreach automatically binds rows of what is printed
      c(c, i, summary(res)$coefficients[((length(summary(res)$coefficients)/2)+1):(length(summary(res)$coefficients) - length(summary(res)$coefficients)/4)])
    }

    resnames <- suppressWarnings(glm(as.formula(formula), weights = ReadCount, family = binomial, data = lrP[lrP$POS == unique(lrP$POS)[1],]))

    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", names(resnames$coefficients))
    #colnames(Results) <- c("CHROM", "POS", resultscol)

    for(i in 2:length(colnames(Results))){
      Results[,i] <- as.numeric(Results[,i])
    }
    Results %>% arrange(POS) -> Results
  }
  return(Results)
}
```

This version of the function can handle very little data being present, ie less than 2x the window size, because it will repeat the raw data in that case. Alternatively, for less data we can always decrease the window, but that would require re-analyzing everything with that decreased window. 

```{r, eval = FALSE}
Results <- Dilute_BSA_byParent
windowsize = 100
chr = "VI"

cybrSmoothBSAWindows2 <- function(Results, windowsize = 100, chr = unique(Results$CHROM)[1]){

  #extract parameters
  Results %>% select(-CHROM, -POS) %>% names() -> params

  #arrange by position
  Results %>% filter(CHROM == chr) %>% arrange(POS) -> Results

  #run window - THIS IS THE PROBLEM
  if(length(Results$POS) > 2*windowsize){
    WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {

    smoothedparams <- vector()
    for(p in 1:length(params)){
      smoothedparams[p] <- mean(Results[[params[p]]][(i-windowsize):(i+windowsize)])
    }

    #print CHROM, index, POS, and mean
    unlist(c(chr, i, Results[i,2:(2+length(params))],smoothedparams))
  }

  #rename columns
  WResult <- as.data.frame(WResult)
  }else{
    WResult <- data.frame(Results[,1], 
                          Index = 0, 
                          Results[,2],
                          Results[,3:length(colnames(Results))],
                          Results[,3:length(colnames(Results))]
                          )
    WResult[,(length(colnames(Results))+1):(length(colnames(Results)) + length(colnames(Results))-2)]
  }
  

  colnames(WResult) <- c("CHROM", "Index", "POS",
                         paste(params, "Z", sep = "_"),
                         paste(params, "Zprime", sep = "_"))

  #convert to numeric
  for(i in 2:length(colnames(WResult))){
    WResult[,i] <- as.numeric(WResult[,i])
  }

  return(WResult)
}
```

## Binary Dilute Samples with less filtering of data

Old code is commented out, but some has just been replaced

```{r, eval = FALSE}
#Load in data
load("DiluteData.Rdata")

#USE DIFFERENT DATASETS
DiluteData$CycZeo = ifelse(DiluteData$Study == "CycZeo", "CycZeo", "NotCycZeo")
DiluteData$CuSO4 = ifelse(DiluteData$Study == "CuSO4", "CuSO4", "NotCuSO4")
DiluteData$Fluc = ifelse(DiluteData$Study == "Fluc", "Fluc", "NotFluc")

DiluteData$Rep <- NA
DiluteData$Rep[grep("A.fastq", CSSI_CycZeo$Dataset)] <- "A"
DiluteData$Rep[grep("B.fastq", CSSI_CycZeo$Dataset)] <- "B"
DiluteData$Rep <- factor(DiluteData$Rep)

DiluteData %>% group_by(CHROM, POS) %>% 
  summarise(CHROM = CHROM, POS = POS, 
            cz_length = length(unique(CycZeo)),
            cu_length = length(unique(CuSO4)),
            f_length = length(unique(Fluc))) %>% 
  distinct() %>% ungroup() -> DiluteDataCounts_2

DiluteDataCounts_2r %>% pivot_longer(c(cz_length, cu_length, f_length), names_to = "Type", values_to = "Counts") %>%
  filter(Counts < 2) %>% select(CHROM, POS) %>% mutate(MissingData = 0) -> dc2

################################################################################

#Filter out those which don't have all three
DiluteDataCounts_2 %>% filter(cz_length > 1) %>% left_join(DiluteData) %>% ungroup() -> DiluteData_cz
DiluteDataCounts_2 %>% filter(cu_length > 1) %>% left_join(DiluteData) %>% ungroup() -> DiluteData_cu
DiluteDataCounts_2 %>% filter(f_length > 1) %>% left_join(DiluteData) %>% ungroup() -> DiluteData_f

################################################################################
# Without covariates
################################################################################

Dilute_BSA_CycZeo2 <- foreach(i=unique(DiluteData_cz$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_cz, chr = i, formula = "PAllele ~ CycZeo") %>% cybrSmoothBSAWindows2()
}

Dilute_BSA_CuSO42 <- foreach(i=unique(DiluteData_cu$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_cu, chr = i, formula = "PAllele ~ CuSO4") %>% cybrSmoothBSAWindows2()
}

Dilute_BSA_Fluc2 <- foreach(i=unique(DiluteData_f$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_f, chr = i, formula = "PAllele ~ Fluc") %>% cybrSmoothBSAWindows2() 
}
# Merging instead once cybrSmoothBSAWindows2 is run
Dilute_BSA_CuSO42r %>% select(-'(Intercept)_Z', -'(Intercept)_Zprime', -Index) %>% 
  full_join(Dilute_BSA_Fluc2) %>% select(-'(Intercept)_Z', -'(Intercept)_Zprime', -Index) %>% 
  full_join(Dilute_BSA_CycZeo2) %>% select(-'(Intercept)_Z', -'(Intercept)_Zprime', -Index) %>% 
  full_join(dc2) %>% 
  pivot_longer(c(CycZeoNotCycZeo_Zprime, FlucNotFluc_Zprime, CuSO4NotCuSO4_Zprime), names_to = "Experiment", values_to = "Z_Prime") %>%
  pivot_longer(c(CycZeoNotCycZeo_Z, FlucNotFluc_Z, CuSO4NotCuSO4_Z), names_to = "ExperimentZ", values_to = "Z_Score") %>%
  distinct() -> DiluteBSA_smoothed_2_char


DiluteBSA_smoothed_2_char$CHROM <- factor(DiluteBSA_smoothed_2_char$CHROM, levels = ChromKey$chromosomes)

save(DiluteBSA_smoothed_2_char, file = "DiluteBSA_smoothed_2_char.Rdata")

################################################################################
# Including Parent as covariate
################################################################################
Dilute_BSA_CycZeo2r <- foreach(i=unique(DiluteData_cz$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_cz, chr = i, formula = "PAllele ~ CycZeo*Parent") %>% cybrSmoothBSAWindows2()
}

Dilute_BSA_CuSO42r <- foreach(i=unique(DiluteData_cu$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_cu, chr = i, formula = "PAllele ~ CuSO4*Parent") %>% cybrSmoothBSAWindows2()
}

Dilute_BSA_Fluc2r <- foreach(i=unique(DiluteData_f$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(DiluteData_f, chr = i, formula = "PAllele ~ Fluc*Parent") %>% cybrSmoothBSAWindows2() 
}


# Merging instead once cybrSmoothBSAWindows2 is run
Dilute_BSA_CuSO42r %>% select(-'(Intercept)_Z', -'(Intercept)_Zprime', -Index) %>% 
  full_join(Dilute_BSA_Fluc2r) %>% select(-'(Intercept)_Z', -'(Intercept)_Zprime', -Index) %>% 
  full_join(Dilute_BSA_CycZeo2r) %>% select(-'(Intercept)_Z', -'(Intercept)_Zprime', -Index) %>% 
  full_join(dc2) -> DB_rep

save(DB_rep, file = "DB_rep.Rdata")

 colnames(DB_rep)[grep("Zprime", colnames(DB_rep))]

# FIX THIS BEFORE RUNNING
DB_rep %>% pivot_longer(colnames(DB_rep)[grep("Zprime", colnames(DB_rep))], 
                        names_to = "Experiment", 
                        values_to = "Z_Prime") -> DB_rep_intermediate
DB_rep_intermediate %>% pivot_longer(colnames(DB_rep_intermediate)[grep("_Z", colnames(DB_rep_intermediate))], 
               names_to = "ExperimentZ", values_to = "Z_Score") %>%
  distinct() -> DiluteBSA_smoothed_2_char_r


DiluteBSA_smoothed_2_char_r$CHROM <- factor(DiluteBSA_smoothed_2_char_r$CHROM, levels = ChromKey$chromosomes)

save(DiluteBSA_smoothed_2_char_r, file = "DiluteBSA_smoothed_2_char_r.Rdata")

```

Plotting the same (deprecated)

```{r, eval = FALSE}
load("DiluteBSA_smoothed_2.Rdata")

# DiluteDataCounts_2 %>% transmute(CHROM = CHROM, Index = NA, POS = POS,
#                                  CycZeo_Z = NA, CuSO4_Z = NA, Fluc_Z = NA,
#                                   CycZeo_Zprime = NA, CuSO4_Zprime = NA, Fluc_Zprime = NA,
#                                  MissingData = 0) -> MD2
# 
# DiluteBSA_smoothed_2$MissingData <- NA
# DiluteBSA_smoothed_2 %>% mutate(MissingData = NA) %>% rbind(MD2) -> DiluteBSA_smoothed2_u

DiluteBSA_smoothed_2 %>% 
  # pivot_longer(c(CycZeo_Zprime, CuSO4_Zprime, Fluc_Zprime), 
  #              names_to = "Experiment", 
  #              values_to = "Z_Prime") %>%
  ggplot(aes(x = POS, y = Z_Prime, color = Experiment)) + 
  geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  ylim(-5,5) +
  theme_minimal() + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + 
  ylab("Smoothed Z Score") + 
  ggtitle("Effects of Dilute Samples")

DiluteBSA_smoothed_2 %>% filter(CHROM != "I", CHROM != "VI") %>% 
  ggplot(aes(x = POS, y = Z_Prime, color = Experiment)) + geom_point(alpha = 0.2, size = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-5,5) +
  theme_minimal() + #scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Smoothed Z Score") + 
  ggtitle("Effects of Dilute Samples")

DiluteBSA_smoothed_2 %>% 
  ggplot(aes(x = POS, y = Z_Score, color = ExperimentZ)) + geom_point(alpha = 0.2, size = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-5,5) +
  theme_minimal() + #scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Effects of Dilute Samples | Truncated")

DiluteBSA_smoothed_2 %>% 
  ggplot(aes(x = POS, y = Z_Score, color = ExperimentZ)) + geom_point(alpha = 0.2, size = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() + #scale_color_manual(values = mylabelcolorsd) + 
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Effects of Dilute Samples | Full Range")
```

Plotting the relabeled version

```{r}
load("DiluteBSA_smoothed_2_char.Rdata")

DiluteBSA_smoothed_2_char %>% 
  ggplot(aes(x = POS, y = Z_Prime, color = Experiment)) + 
  geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  ylim(-5,5) +
  theme_minimal() + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + 
  ylab("Smoothed Z Score") + 
  ggtitle("Effects of Dilute Samples")

DiluteBSA_smoothed_2_char %>% filter(CHROM != "I", CHROM != "VI") %>% 
  ggplot(aes(x = POS, y = Z_Prime, color = Experiment)) + geom_point(alpha = 0.2, size = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-5,5) +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Smoothed Z Score") + 
  ggtitle("Effects of Dilute Samples")

DiluteBSA_smoothed_2_char %>% 
  ggplot(aes(x = POS, y = Z_Score, color = ExperimentZ)) + geom_point(alpha = 0.2, size = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-5,5) +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Effects of Dilute Samples | Truncated")

DiluteBSA_smoothed_2 %>% 
  ggplot(aes(x = POS, y = Z_Score, color = ExperimentZ)) + geom_point(alpha = 0.2, size = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Effects of Dilute Samples | Full Range")
```


Doing this with the full dataframe
```{r}
load("DiluteBSA_smoothed_2_char_r.Rdata")

DiluteBSA_smoothed_2_char_r %>% 
  ggplot(aes(x = POS, y = Z_Prime, color = Experiment)) + 
  geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
  ylim(-5,5) +
  theme_minimal() + 
  theme(legend.position = "bottom", 
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + 
  ylab("Smoothed Z Score") + 
  ggtitle("Effects of Dilute Samples")

DiluteBSA_smoothed_2_char_r %>% filter(CHROM != "I", CHROM != "VI") %>% 
  ggplot(aes(x = POS, y = Z_Prime, color = Experiment)) + geom_point(alpha = 0.2, size = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-5,5) +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Smoothed Z Score") + 
  ggtitle("Effects of Dilute Samples")

DiluteBSA_smoothed_2_char_r %>% 
  ggplot(aes(x = POS, y = Z_Score, color = ExperimentZ)) + geom_point(alpha = 0.2, size = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ylim(-5,5) +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Effects of Dilute Samples | Truncated")

DiluteBSA_smoothed_2_char_r %>% 
  ggplot(aes(x = POS, y = Z_Score, color = ExperimentZ)) + geom_point(alpha = 0.2, size = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Effects of Dilute Samples | Full Range")
```

## Doing actual pairwise comparisons rather than all or nothing

```{r}
#Load in data
load("DiluteData.Rdata")

#USE DIFFERENT DATASETS
DiluteData$CycZeo = ifelse(DiluteData$Study == "CycZeo", "CycZeo", NA)
DiluteData$CuSO4 = ifelse(DiluteData$Study == "CuSO4", "CuSO4", NA)
DiluteData$Fluc = ifelse(DiluteData$Study == "Fluc", "Fluc", NA)

#Make the column to compare
CuZeo_Compare <- DiluteData %>% pivot_longer(c(CycZeo, CuSO4), names_to = "Experiment", values_to = "Binary") %>% 
  filter(Binary == "CycZeo" | Binary == "CuSO4")
CuFlu_Compare <- DiluteData %>% pivot_longer(c(Fluc, CuSO4), names_to = "Experiment", values_to = "Binary") %>% 
  filter(Binary == "Fluc" | Binary == "CuSO4")
FluZeo_Compare <- DiluteData %>% pivot_longer(c(Fluc, CycZeo), names_to = "Experiment", values_to = "Binary") %>% 
  filter(Binary == "Fluc" | Binary == "CycZeo")

#Find only the rows with both
CuZeo_Compare %>% group_by(CHROM, POS) %>% summarize(CHROM = CHROM, POS = POS, Count = length(unique(Binary))) %>% distinct() %>% filter(Count == 2) %>% select(-Count) %>% left_join(CuZeo_Compare) -> CuZeo
CuFlu_Compare %>% group_by(CHROM, POS) %>% summarize(CHROM = CHROM, POS = POS, Count = length(unique(Binary))) %>% distinct() %>% filter(Count == 2) %>% select(-Count) %>% left_join(CuFlu_Compare) -> CuFlu
FluZeo_Compare %>% group_by(CHROM, POS) %>% summarize(CHROM = CHROM, POS = POS, Count = length(unique(Binary))) %>% distinct() %>% filter(Count == 2) %>% select(-Count) %>% left_join(FluZeo_Compare) -> FluZeo

#Run the analysis
#Run BSA_GLM() and SmoothBSAWindows() for each
CSSI_FluZeo <- foreach(i=unique(FluZeo$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(FluZeo, chr = i, formula = "PAllele ~ Experiment+Parent") %>% cybrSmoothBSAWindows()
}
save(CSSI_FluZeo, file = "CSSI_FluZeo.Rdata")

CSSI_CuFlu <- foreach(i=unique(CuFlu$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuFlu, chr = i, formula = "PAllele ~ Experiment+Parent") %>% cybrSmoothBSAWindows()
}
save(CSSI_CuFlu, file = "CSSI_CuFlu.Rdata")

CSSI_CuZeo <- foreach(i=unique(CuZeo$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CuZeo, chr = i, formula = "PAllele ~ Experiment+Parent") %>% cybrSmoothBSAWindows()
}
save(CSSI_CuZeo, file = "CSSI_CuZeo.Rdata")

```

Plotting

```{r}

load("CSSI_CuZeo.Rdata") 
load("CSSI_CuFlu.Rdata")
load("CSSI_FluZeo.Rdata")

CSSI_CuZeo %>% ggplot(aes(x = POS, y = ExperimentCycZeo_Zprime)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("CycZeo vs CuSO4 Unselected")

CSSI_CuFlu %>% ggplot(aes(x = POS, y = ExperimentFluc_Zprime)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Fluconazole vs CuSO4 Unselected")

CSSI_FluZeo %>% ggplot(aes(x = POS, y = ExperimentFluc_Zprime)) + geom_line() + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Fluconazole vs CycZeo Unselected")

## View These Together
################################################################################
CSSI_CuZeo %>% transmute(CHROM = CHROM, POS = POS, Z = ExperimentCycZeo_Z, 
                         Zprime = ExperimentCycZeo_Zprime, 
                         Parent_Z = ParentWineI_Z, 
                         Parent_Zprime = ParentWineI_Zprime,
                         Label = "CycZeo_CuSO4") -> CSSI_CuZeo_t

CSSI_CuFlu %>% transmute(CHROM = CHROM, POS = POS, Z = ExperimentFluc_Z, 
                         Zprime = ExperimentFluc_Zprime, 
                         Parent_Z = ParentWineI_Z, 
                         Parent_Zprime = ParentWineI_Zprime,
                         Label = "Fluc_CuSO4") -> CSSI_CuFlu_t

CSSI_FluZeo %>% transmute(CHROM = CHROM, POS = POS, Z = ExperimentFluc_Z, 
                         Zprime = ExperimentFluc_Zprime, 
                         Parent_Z = ParentWineI_Z, 
                         Parent_Zprime = ParentWineI_Zprime,
                         Label = "Fluc_CycZeo") -> CSSI_FluZeo_t

PairwiseDilute <- rbind(CSSI_CuZeo_t,CSSI_CuFlu_t,CSSI_FluZeo_t)
PairwiseDilute$CHROM <- factor(PairwiseDilute$CHROM, levels = ChromKey$chromosomes)

PairwiseDilute %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Zprime, color = Label)) + 
  geom_line() +
  #geom_line(aes(x = POS, y = Parent_Zprime, color = Label),linetype = "dashed", alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Pairwise Unselected")

PairwiseDilute %>% filter(CHROM == "VII") %>% ggplot(aes(x = POS, y = Zprime, color = Label)) + 
  geom_line(size = 1) +
  #geom_line(aes(x = POS, y = Parent_Zprime, color = Label),linetype = "dashed", alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + 
      xlab("Genomic Position") + ylab("Z Score") + 
  ggtitle("Pairwise Unselected")
```


# Adding in the covariate

## Redoing Analysis

```{r}
mydatatotest = "HKTMZDRX2.SortedCat.vcf.output.table"

CSSI_CycZeo <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = parentSNPids, yeast = TRUE) %>% 
  na.omit()

unique(CSSI_CycZeo$Dataset)

CSSI_CycZeo$Bulk <- NA
CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq"] <- "Cycloheximide"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "Dilute"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq"] <- "Zeocin"

CSSI_CycZeo$Parent <- NA
CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq"] <- "OakI"

CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "WineI"

CSSI_CycZeo$Rep <- NA
CSSI_CycZeo$Rep[grep("A.fastq", CSSI_CycZeo$Dataset)] <- "A"
CSSI_CycZeo$Rep[grep("B.fastq", CSSI_CycZeo$Dataset)] <- "B"

CSSI_CycZeo$Parent <- factor(CSSI_CycZeo$Parent)
CSSI_CycZeo$Bulk <- factor(CSSI_CycZeo$Bulk)
CSSI_CycZeo$ReadCount <- as.numeric(CSSI_CycZeo$ReadCount)
CSSI_CycZeo$Rep <- factor(CSSI_CycZeo$Rep)

head(CSSI_CycZeo)
CSSI_CycZeo_rep <- CSSI_CycZeo

save(CSSI_CycZeo_rep, file = "CSSI_CycZeo_rep_counts.Rdata")
```

```{r}
load("CSSI_CycZeo_rep_counts.Rdata")

#Separate into different CSSI_CycZeo_rep, since the subsequent steps will break if not
CSSI_Cycloheximide_r <- CSSI_CycZeo_rep[CSSI_CycZeo_rep$Bulk != "Zeocin",]
CSSI_Zeocin_r <- CSSI_CycZeo_rep[CSSI_CycZeo_rep$Bulk != "Cycloheximide",]

#Run BSA_GLM() and SmoothBSAWindows() for each
CSSI_Cyc_BSA_r <- foreach(i=unique(CSSI_Cycloheximide_r$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Cycloheximide_r, chr = i, formula = "PAllele ~ Bulk*Parent+Rep") %>% cybrSmoothBSAWindows()
}

save(CSSI_Cyc_BSA_r, file = "CSSI_Cyc_BSA_r.Rdata")

CSSI_Zeo_BSA_r <- foreach(i=unique(CSSI_Zeocin_r$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(CSSI_Zeocin_r, chr = i, formula = "PAllele ~ Bulk*Parent+Rep") %>% cybrSmoothBSAWindows()
}

save(CSSI_Zeo_BSA_r, file = "CSSI_Zeo_BSA_r.Rdata")
```

Plotting

```{r}
#Load in data
load("CSSI_Zeo_BSA_r.Rdata")
load("CSSI_Cyc_BSA_r.Rdata")

#Plot for Zeocin
cybrPlotZPrime(CSSI_Zeo_BSA_r, 
               chromosomes = ChromKey$chromosomes[2:16],
               columns = colnames(CSSI_Zeo_BSA_r)[10:13], 
               colvalues = c( "#D7335C","#345F6F", "#FFB05C","black"),
               title = "Zeocin Smoothed Z scores with Replicate Covariate") + 
  geom_hline(yintercept = c(-1.96, 1.96), linetype = "dashed")

#Plot for Cycloheximide
cybrPlotZPrime(CSSI_Cyc_BSA_r, 
               chromosomes = ChromKey$chromosomes[2:16],
               columns = colnames(CSSI_Cyc_BSA_r)[10:13], 
               colvalues = c( "#D7335C","#345F6F", "#FFB05C","black"),
               title = "Cycloheximide Smoothed Z scores with Replicate Covariate") + 
  geom_hline(yintercept = c(-1.96, 1.96), linetype = "dashed")
```

Using the largest replicate set per chromosome

```{r}
#Load in data
load("CSSI_Zeo_BSA_r.Rdata")
load("CSSI_Cyc_BSA_r.Rdata")

CSSI_Zeo_BSA_r %>% group_by(CHROM) %>% summarize(ChromCutoff = max(abs(RepB_Zprime))) -> Cutoffs

cybrPlotZPrime(CSSI_Zeo_BSA_r, 
               chromosomes = ChromKey$chromosomes[2:16],
               columns = colnames(CSSI_Zeo_BSA_r)[10:13], 
               colvalues = c( "#D7335C","#345F6F", "#FFB05C","black"),
               title = "Zeocin Smoothed Z scores with Replicate Covariate") + 
  geom_hline(yintercept = c(-1.96, 1.96), linetype = "dashed") +
  geom_hline(data = Cutoffs, aes(yintercept = ChromCutoff))

#INCLUDING INTERCEPT
CSSI_Zeo_BSA_r_zprimes <- CSSI_Zeo_BSA_r[,c(1,3,9:13)]

CSSI_Zeo_BSA_r_zprimes %>% pivot_longer(-c(CHROM, POS)) %>% ggplot(aes(x = POS, y = abs(value), color = name)) +
  geom_line(size = 1) +
  geom_hline(data = Cutoffs, aes(yintercept = ChromCutoff)) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_Zeo_BSA_r_zprimes$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())+
  scale_color_manual(values = c("skyblue", "red", "blue", "black", "gray"))

#EXCLUDING INTERCEPT
 CSSI_Zeo_BSA_r[,c(1,3,10:13)] %>% pivot_longer(-c(CHROM, POS)) %>% ggplot(aes(x = POS, y = abs(value), color = name)) +
  geom_line(size = 1) +
  geom_hline(data = Cutoffs, aes(yintercept = ChromCutoff)) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_Zeo_BSA_r_zprimes$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  scale_color_manual(values = c("red", "blue", "black", "gray"))

 #EXCLUDING INTERCEPT
 CSSI_Cyc_BSA_r[,c(1,3,10:13)] %>% pivot_longer(-c(CHROM, POS)) %>% ggplot(aes(x = POS, y = abs(value), color = name)) +
  geom_line(size = 1) +
  geom_hline(data = Cutoffs, aes(yintercept = ChromCutoff)) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_Zeo_BSA_r_zprimes$POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
      theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  scale_color_manual(values = c("red", "blue", "black", "gray"))

```


```{r}
load("CSSI_Cyc_BSA_r.Rdata")

CSSI_Cyc_BSA_r %>% group_by(CHROM) %>% summarize(ChromCutoff = max(abs(RepB_Zprime))) -> Cutoffs

#EXCLUDING INTERCEPT
 CSSI_Cyc_BSA_r[,c(1,3,10:13)] %>% filter(CHROM != "I") %>% pivot_longer(-c(CHROM, POS)) %>% ggplot(aes(x = POS, y = abs(value), color = name)) +
  geom_line(size = 1) +
  geom_hline(data = Cutoffs, aes(yintercept = ChromCutoff)) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_Cyc_BSA_r[,c(1,3,10:13)] $POS), by = 1e5), name = "Genomic Position") +
      facet_grid(~CHROM, scales = "free_x", space = "free_x") + 
   scale_color_manual(values = cybr_BSAcolors[c(2,1,3,4)])
 
```

# Permutations on chunks of data

```{r}
load("CSSI_Fluc_counts.Rdata")

unique(CSSI_Fluc$Dataset)

CSSI_Fluc$Rep <- NA
CSSI_Fluc$Rep[grep("B.fastq", CSSI_Fluc$Dataset)] <- "B"
CSSI_Fluc$Rep[grep("C.fastq", CSSI_Fluc$Dataset)] <- "C"
CSSI_Fluc$Rep[grep("D.fastq", CSSI_Fluc$Dataset)] <- "D"

table(CSSI_Fluc$Rep, CSSI_Fluc$Parent, CSSI_Fluc$Bulk)

#To actually make replicates from these, divide the reads in half? Then look at the interaction score(?)

CSSI_Fluc %>% select(-REF, -ALT, -Dataset, -Type) %>% 
  pivot_wider(names_from = c(Bulk, Parent, Rep), values_from = c(ReadCount)) %>%
  mutate(Dilute_OakI_C = Dilute_OakI_B,
         Dilute_OakI_D = Dilute_OakI_B,
         Fluconazole_WineI_B = Fluconazole_WineI_C,
         Fluconazole_WineI_D = Fluconazole_WineI_C) %>% pivot_longer(cols = -c(POS, CHROM, AltRef_Allele, PAllele), 
                                                                     names_to = c("Bulk", "Parent", "Rep"),
                                                                     names_pattern = "_?(.*)_(.*)_(.*)",
                                                                     values_to = "ReadCount") -> CSSI_Fluc_FalseReps

################################################################################
Fluc_7 <- cybrBSA_GLM(CSSI_Fluc_FalseReps, formula = "PAllele ~ Bulk*Parent + Rep", chr = "VII") %>% cybrSmoothBSAWindows()
Fluc_7_repint <- cybrBSA_GLM(CSSI_Fluc_FalseReps, formula = "PAllele ~ Bulk*Parent* Rep", chr = "VII") %>% cybrSmoothBSAWindows()

Fluc_15 <- cybrBSA_GLM(CSSI_Fluc_FalseReps, formula = "PAllele ~ Bulk*Parent + Rep", chr = "XV") %>% cybrSmoothBSAWindows()
Fluc_15_repint <- cybrBSA_GLM(CSSI_Fluc_FalseReps, formula = "PAllele ~ Bulk*Parent* Rep", chr = "XV") %>% cybrSmoothBSAWindows()

################################################################################
#Chr 7

colnames(Fluc_7)[11:15] <- c("Bulk", "Parent", "RepC", "RepD", "Interaction")
colnames(Fluc_7_repint)[17:27] <- c("1Bulk", "3Parent", "RepC", "RepD", 
                           "2BulkParent_Interaction", "Bulk_RepC_Interaction", "Bulk_RepD_Interaction", 
                           "Parent_RepC_Interaction", "Parent_RepD_Interaction", "BulkParentRepC", "BulkParentRepD")
#Convert to absolute values
Fluc_7[,c(-1)] <- abs(Fluc_7[,c(-1)])
Fluc_7_repint[,c(-1)] <- abs(Fluc_7_repint[,c(-1)])

p1 <- cybrPlotZPrime(Fluc_7, columns = c("Bulk", "Parent", "RepC", "RepD", "Interaction"), colvalues = cybr_BSAcolors)

p2 <- cybrPlotZPrime(Fluc_7_repint, 
               columns = c("1Bulk", "3Parent", "RepC", "RepD", 
                           "2BulkParent_Interaction", "Bulk_RepC_Interaction", "Bulk_RepD_Interaction", 
                           "Parent_RepC_Interaction", "Parent_RepD_Interaction", "BulkParentRepC", "BulkParentRepD"), 
               colvalues = c(cybr_BSAcolors, cybr_CHROMcolors))

p1
p2
p1 + theme(legend.position = "none") + ylim(0,6) + ggtitle("Fluconazole Bulk*Parent+Rep")
p2 + theme(legend.position = "none") + ylim(0,6) + ggtitle("Fluconazole Bulk*Parent*Rep")

################################################################################
#Chr 15

colnames(Fluc_15)[11:15] <- c("Bulk", "Parent", "RepC", "RepD", "Interaction")
colnames(Fluc_15_repint)[17:27] <- c("1Bulk", "3Parent", "RepC", "RepD", 
                           "2BulkParent_Interaction", "Bulk_RepC_Interaction", "Bulk_RepD_Interaction", 
                           "Parent_RepC_Interaction", "Parent_RepD_Interaction", "BulkParentRepC", "BulkParentRepD")

#Convert to absolute values
Fluc_15[,c(-1)] <- abs(Fluc_15[,c(-1)])
Fluc_15_repint[,c(-1)] <- abs(Fluc_15_repint[,c(-1)])

p3 <- cybrPlotZPrime(Fluc_15, columns = c("Bulk", "Parent", "RepC", "RepD", "Interaction"), colvalues = cybr_BSAcolors)

p4 <- cybrPlotZPrime(Fluc_15_repint, 
               columns = c("1Bulk", "3Parent", "RepC", "RepD", 
                           "2BulkParent_Interaction", "Bulk_RepC_Interaction", "Bulk_RepD_Interaction", 
                           "Parent_RepC_Interaction", "Parent_RepD_Interaction", "BulkParentRepC", "BulkParentRepD"), 
               colvalues = c(cybr_BSAcolors, cybr_CHROMcolors))

p3
p4
p3 + theme(legend.position = "none") + ylim(0,6) + ggtitle("Fluconazole Bulk*Parent+Rep")
p4 + theme(legend.position = "none") + ylim(0,6) + ggtitle("Fluconazole Bulk*Parent*Rep")
```

Okay actually doing permutations?

```{r}
### Run GLM Function
lrP <- CSSI_Fluc
chr <- "I"
i <- unique(lrP$POS)[1]
```

```{r}
cybrBSA_GLM_window <-  function(lrP, chr = "II", windowsize = 5000, formula = "PAllele~Bulk*Parent",
                         resultscol = c("Intercept", "Bulk", "Parent", "Interaction")){
  require(stringr)
  if(identical(grep(" ", formula), integer(0)) == FALSE){return(print("Remove spaces from formula or use cybrBSA_GLM()"))}
         
         
  lrP <- subset(lrP, CHROM == chr)

  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      windowdata <- lrP[lrP$POS < (i+windowsize) & lrP$POS > (i-windowsize),]
      
      #Convert to sum of counts
      mycols <- unlist(str_split(unlist(str_split(formula, pattern = c("~"))), pattern = "\\*"))
      windowdata %>% group_by(.[mycols]) %>% summarize(ReadCount = sum(ReadCount)) -> windowdata
      
      res <- suppressWarnings(glm(as.formula(formula), 
                                  weights = ReadCount, 
                                  family = binomial, 
                                  data = windowdata))

      #Output of foreach automatically binds rows of what is printed
      c(c, i, summary(res)$coefficients[((length(summary(res)$coefficients)/2)+1):(length(summary(res)$coefficients) - length(summary(res)$coefficients)/4)])
    }

    resnames <- suppressWarnings(glm(as.formula(formula), weights = ReadCount, family = binomial, data = lrP[lrP$POS == unique(lrP$POS)[1],]))

    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", names(resnames$coefficients))
    #colnames(Results) <- c("CHROM", "POS", resultscol)

    for(i in 2:length(colnames(Results))){
      Results[,i] <- as.numeric(Results[,i])
    }
    Results %>% arrange(POS) -> Results
  }
  return(Results)
  
}
```

```{r}
Fluc_I <- suppressWarnings(cybrBSA_GLM_window(CSSI_Fluc, chr = "I", windowsize = 1000))
colnames(Fluc_I)[4:6] <- c("Bulk", "Parent", "Interaction")

Fluc_I %>% pivot_longer(cols = c("Bulk", "Parent", "Interaction")) %>% ggplot(aes(x = POS, y = value, color = name)) + geom_line()
```

Do this for each sample and chromosome

```{r, eval = FALSE}
load("CSSI_Fluc_counts.Rdata")
load("CSSI_CycZeo_counts.Rdata")

CycCounts <- CSSI_CycZeo %>% filter(Bulk != "Zeocin")
ZeoCounts <- CSSI_CycZeo %>% filter(Bulk != "Cycloheximide")

Fluc_SmoothedZ <- foreach(i=unique(CSSI_Fluc$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM_window(CSSI_Fluc, chr = i) 
}
Fluc_SmoothedZ$CHROM <- factor(Fluc_SmoothedZ$CHROM, levels = as.character((as.roman(1:17))))

save(Fluc_SmoothedZ, file = "Fluc_SmoothedZ.Rdata")

Cyc_SmoothedZ <- foreach(i=unique(CycCounts$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM_window(CycCounts, chr = i)  
}
Cyc_SmoothedZ$CHROM <- factor(Cyc_SmoothedZ$CHROM, levels = as.character((as.roman(1:17))))

save(Cyc_SmoothedZ, file = "Cyc_SmoothedZ.Rdata")


Zeo_SmoothedZ <- foreach(i=unique(ZeoCounts$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM_window(ZeoCounts, chr = i) 
}
Zeo_SmoothedZ$CHROM <- factor(Zeo_SmoothedZ$CHROM, levels = as.character((as.roman(1:17))))
save(Zeo_SmoothedZ, file = "Zeo_SmoothedZ.Rdata")

```

Run this for CuSO4 as well

```{r}
load("CuSO4data_counts.Rdata")

CuSO4_SmoothedZ <- foreach(i=unique(CuSO4data$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM_window(CuSO4data, chr = i) 
}
CuSO4_SmoothedZ$CHROM <- factor(CuSO4_SmoothedZ$CHROM, levels = as.character((as.roman(1:17))))

save(CuSO4_SmoothedZ, file = "CuSO4_SmoothedZ.Rdata")

```

```{r}
#Plot both sides
load("Zeo_SmoothedZ.Rdata")
colnames(Zeo_SmoothedZ)[4:6] <- c("Bulk", "Parent", "Interaction")
Zeo_SmoothedZ %>% filter(CHROM != "I") %>% pivot_longer(cols = c("Bulk", "Parent", "Interaction")) %>% ggplot(aes(x = POS, y = value, color = name)) + geom_line() + facet_grid(~CHROM, scales = "free") +
  ggtitle("Zeocin Combined Z Scores") + ylim(-40,40)

load("Fluc_SmoothedZ.Rdata")
colnames(Fluc_SmoothedZ)[4:6] <- c("Bulk", "Parent", "Interaction")
Fluc_SmoothedZ %>% filter(CHROM != "I")%>% pivot_longer(cols = c("Bulk", "Parent", "Interaction")) %>% ggplot(aes(x = POS, y = value, color = name)) + geom_line() + facet_grid(~CHROM, scales = "free") +
    ggtitle("Fluconazole Combined Z Scores")+ ylim(-40,40)


load("Cyc_SmoothedZ.Rdata")
colnames(Cyc_SmoothedZ)[4:6] <- c("Bulk", "Parent", "Interaction")
Cyc_SmoothedZ %>% filter(CHROM != "I")%>% pivot_longer(cols = c("Bulk", "Parent", "Interaction")) %>% ggplot(aes(x = POS, y = value, color = name)) + geom_line() + facet_grid(~CHROM, scales = "free") +
    ggtitle("Cycloheximide Combined Z Scores")+ ylim(-40,40)

load("CuSO4_SmoothedZ.Rdata")
colnames(CuSO4_SmoothedZ)[4:6] <- c("Bulk", "Parent", "Interaction")
CuSO4_SmoothedZ %>% filter(CHROM != "I")%>% pivot_longer(cols = c("Bulk", "Parent", "Interaction")) %>% ggplot(aes(x = POS, y = value, color = name)) + geom_line() + facet_grid(~CHROM, scales = "free") +
    ggtitle("CuSO4 Combined Z Scores")+ ylim(-40,40)

```

```{r}
#Plot absolute values
load("Zeo_SmoothedZ.Rdata")
colnames(Zeo_SmoothedZ)[4:6] <- c("Bulk", "Parent", "Interaction")
Zeo_SmoothedZ %>% filter(CHROM != "I") %>% pivot_longer(cols = c("Bulk", "Parent", "Interaction")) %>% ggplot(aes(x = POS, y = abs(value), color = name)) + geom_line() + facet_grid(~CHROM, scales = "free") +
  ggtitle("Zeocin Combined Z Scores") + ylim(0,40)

load("Fluc_SmoothedZ.Rdata")
colnames(Fluc_SmoothedZ)[4:6] <- c("Bulk", "Parent", "Interaction")
Fluc_SmoothedZ %>% filter(CHROM != "I")%>% pivot_longer(cols = c("Bulk", "Parent", "Interaction")) %>% ggplot(aes(x = POS, y = abs(value), color = name)) + geom_line() + facet_grid(~CHROM, scales = "free") +
    ggtitle("Fluconazole Combined Z Scores")+ ylim(0,40)


load("Cyc_SmoothedZ.Rdata")
colnames(Cyc_SmoothedZ)[4:6] <- c("Bulk", "Parent", "Interaction")
Cyc_SmoothedZ %>% filter(CHROM != "I")%>% pivot_longer(cols = c("Bulk", "Parent", "Interaction")) %>% ggplot(aes(x = POS, y = abs(value), color = name)) + geom_line() + facet_grid(~CHROM, scales = "free") +
    ggtitle("Cycloheximide Combined Z Scores")+ ylim(0,40)

load("CuSO4_SmoothedZ.Rdata")
colnames(CuSO4_SmoothedZ)[4:6] <- c("Bulk", "Parent", "Interaction")
CuSO4_SmoothedZ %>% filter(CHROM != "I")%>% pivot_longer(cols = c("Bulk", "Parent", "Interaction")) %>% ggplot(aes(x = POS, y = abs(value), color = name)) + geom_line() + facet_grid(~CHROM, scales = "free") +
    ggtitle("CuSO4 Combined Z Scores") + ylim(0,40)
```

Cutoffs based on the parent effect

```{r}
quantile(abs(Cyc_SmoothedZ$Parent[Cyc_SmoothedZ$CHROM != "I"]), 0.975, na.rm = TRUE)
quantile(abs(Zeo_SmoothedZ$Parent[Zeo_SmoothedZ$CHROM != "I"]), 0.975, na.rm = TRUE)
quantile(abs(Fluc_SmoothedZ$Parent[Fluc_SmoothedZ$CHROM != "I"]), 0.975, na.rm = TRUE)

Cyc_SmoothedZ$Experiment <- "Cycloheximide"
Zeo_SmoothedZ$Experiment <- "Zeocin"
Fluc_SmoothedZ$Experiment <- "Fluconazole"

rbind(Cyc_SmoothedZ, Zeo_SmoothedZ, Fluc_SmoothedZ) %>% filter(CHROM != "I") %>% ggplot(aes(x = abs(Parent), color = Experiment)) + geom_density(size = 1)
```

