---
title: "CSS I Fluconazole Analysis from PosterScripts_Yeast2022"
author: "Cassandra Buzby"
date: "8/23/2022"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
library(ggplot2)
library(tidyr)
#library(tidyverse)
library(reshape2)
library(cowplot)
library(dplyr)

library(foreach)
library(doParallel)

ggplot2::theme_set(theme_light())
#ggplot2::theme_set(theme_bw())


CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47",
                  "#F26430", "#0A369D", "#7EA3CC")


library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")
# load("RdataFiles/SAUA_4-13-22.Rdata")
# load("RdataFiles/SCUC_4-13-22.Rdata")
# load("RdataFiles/bsa_glm_results.Rdata")

#rawData = "HGV.SortedCat.vcf.output.table"
```


## Streamlined Data Input

Function to analyze:

```{r}
#################################################
#Change this to analyze all differently
#################################################

mindepth = 50
maxdepth = 1500
minsampledepth = 50
mingq = 0.98

PipelineFunc <- function(HighBulk, LowBulk, rawData,
                         mindepth = mindepth,
                         maxdepth = maxdepth,
                         minsampledepth = minsampledepth,
                         mingq = mingq,
                         windowSize = 2e4, #5e4 works
                         Chroms = c("NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4"),
                         BulkSize = 1000
                         ){
  
  mytitle <- paste(HighBulk, LowBulk, sep = " vs ")
  HNGLCDRXY <- read.table(rawData, header = TRUE)

  df <- importFromGATK(
        file = rawData,
        highBulk = HighBulk,
        lowBulk = LowBulk,
        chromList = Chroms #"NC_001133.9" #
        )

df %>% merge(.,ChromKey) %>% 
  as.data.frame() %>% na.omit() -> df

colnames(df)[1] <- "NC_Chrom"
colnames(df)[which(colnames(df) == "chromosomes")] <- "CHROM"

df$CHROM <- factor(df$CHROM, levels = c("I", "II", "III", "IV", 
                                        "V", "VI", "VII", "VIII", 
                                        "IX", "X", "XI", "XII", 
                                        "XIII", "XIV", "XV", "XVI", "M"))
#################################

#Filter SNPs based on some criteria
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.10, #0.20
        minTotalDepth = mindepth, #100
        maxTotalDepth = maxdepth, #400
        minSampleDepth = minsampledepth, #40
        minGQ = mingq #99
    )

  #df %>% merge(.,ChromKey)
  #df_filt <- df
  
  #Run G' analysis
  df_filt <- runGprimeAnalysis(
      SNPset = df_filt,
      windowSize = windowSize, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
  
  #Run QTLseq analysis
  df_filt <- runQTLseqAnalysis(
      SNPset = df_filt,
      windowSize = windowSize,
      popStruc = "F2",
      bulkSize = 1000, #c(25, 25)
      replications = 10000,
      intervals = c(95, 99)
  )

  df_filt$idu <- row.names(df_filt)
  q <- 0.01
  fdrT <- getFDRThreshold(df_filt$pvalue, alpha = q)
  GprimeT <- df_filt[which(df_filt$pvalue == fdrT), "Gprime"]
  
  # df_filt$chromosomes <- factor(df_filt$chromosomes, levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"))

    
  return(as.data.frame(df_filt))
}
```


### Load Parental Data

```{r}
## Add parental alleles

read.table("Wine_VCF.txt", header = TRUE) %>% mutate(parent = "Wine") -> WineTemp
read.table("Oak_VCF.txt", header = TRUE) %>% mutate(parent = "Oak") -> OakTemp

ParentalVCF <- rbind(WineTemp, OakTemp) %>% arrange(CHROM, POS) %>% select(CHROM, POS, REF, ALT, parent) %>% merge(ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes)

ParentalVCF %>% pivot_wider(names_from = parent, values_from = ALT) -> SNPids

SNPids$Type <- NA
SNPids$Type[is.na(SNPids$Wine) == TRUE] <- "Oak"
SNPids$Type[is.na(SNPids$Oak) == TRUE] <- "Wine"
SNPids$Type[SNPids$Wine == SNPids$Oak] <- "Alt"

SNPids %>% select(CHROM, POS,  Type) -> SNPids
```

### Load in data using QTLSeqR

CuSO4

```{r, eval = FALSE}
rawData = "Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"
Chroms <- ChromKey$CHROM

#CHANGE THE HIGH AND LOW BULKS
OakI_CuSO4_C <- importFromGATK(
        file = rawData,
        highBulk = "SelectedA",
        lowBulk = "UnselectedA",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_CuSO4_C") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

OakI_CuSO4_D <- importFromGATK(
        file = rawData,
        highBulk = "UnselectedA",
        lowBulk = "SelectedA",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_CuSO4_D") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

WineI_CuSO4_C <- importFromGATK(
        file = rawData,
        highBulk = "SelectedC",
        lowBulk = "UnselectedA",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_CuSO4_C") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

WineI_CuSO4_D <- importFromGATK(
        file = rawData,
        highBulk = "UnselectedC",
        lowBulk = "UnselectedA",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_CuSO4_D") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

rbind(OakI_CuSO4_C, OakI_CuSO4_D, WineI_CuSO4_C, WineI_CuSO4_D) %>% merge(., ChromKey) -> CSSI_CuSO4_concat

DatasetKey <- data.frame(Dataset = c("OakI_CuSO4_C", "OakI_CuSO4_D", "WineI_CuSO4_C", "WineI_CuSO4_D"),
                         Parent = c("Oak", "Oak", "Wine", "Wine"),
                         Bulk = c("CuSO4", "Dilute", "CuSO4", "Dilute"))

CSSI_CuSO4_concat %>% left_join(DatasetKey, by = "Dataset") %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> CSSI_CuSO4_concat

CSSI_CuSO4_concat$CHROM <- factor(CSSI_CuSO4_concat$CHROM, levels = ChromKey$chromosomes)

CSSI_CuSO4_concat %>% left_join(SNPids, by = c("CHROM", "POS")) -> CSSI_CuSO4_concat
save(CSSI_CuSO4_concat, file = "RdataFiles/CSSI_CuSO4_concat.Rdata")

```

Fluconazole
```{r, eval = FALSE}
rawData = "HGVMVDRX2.SortedCat.vcf.output.table"
Chroms <- ChromKey$CHROM

OakIDB <- importFromGATK(
        file = rawData,
        highBulk = "HGVMVDRX2_n01_OakI_Dilute_B.fastq",
        lowBulk = "HGVMVDRX2_n01_WineI_Fluc_C.fastq",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_Dilute_B") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

OakIFB <- importFromGATK(
        file = rawData,
        highBulk = "HGVMVDRX2_n01_OakI_Fluc_B.fastq",
        lowBulk = "HGVMVDRX2_n01_OakI_Dilute_B.fastq",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_Fluc_B") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

OakIFC <- importFromGATK(
        file = rawData,
        highBulk = "HGVMVDRX2_n01_OakI_Fluc_C.fastq",
        lowBulk = "HGVMVDRX2_n01_OakI_Dilute_B.fastq",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_Fluc_C") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

OakIFD <- importFromGATK(
        file = rawData,
        highBulk = "HGVMVDRX2_n01_OakI_Fluc_D.fastq",
        lowBulk = "HGVMVDRX2_n01_OakI_Dilute_B.fastq",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_Fluc_D") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

WineIDC <- importFromGATK(
        file = rawData,
        highBulk = "HGVMVDRX2_n01_WineI_Dilute_C.fastq",
        lowBulk = "HGVMVDRX2_n01_OakI_Dilute_B.fastq",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_Dilute_C") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

WineIDD <- importFromGATK(
        file = rawData,
        highBulk = "HGVMVDRX2_n01_WineI_Dilute_D.fastq",
        lowBulk = "HGVMVDRX2_n01_OakI_Dilute_B.fastq",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_Dilute_D") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

WineIFC <- importFromGATK(
        file = rawData,
        highBulk = "HGVMVDRX2_n01_WineI_Fluc_C.fastq",
        lowBulk = "HGVMVDRX2_n01_OakI_Dilute_B.fastq",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_Fluc_C") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

WineIDB <- importFromGATK(
        file = rawData,
        highBulk = "HGVMVDRX2_n01_WineI_Dilute_B.fastq",
        lowBulk = "HGVMVDRX2_n01_OakI_Dilute_B.fastq",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_Dilute_B") %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

WineIFD <- importFromGATK(
        file = rawData,
        highBulk = "HGVMVDRX2_n01_WineI_Fluc_D",
        lowBulk = "HGVMVDRX2_n01_OakI_Dilute_B.fastq",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_Fluc_D") %>% select(CHROM, POS, REF, ALT,
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                Dataset)

rbind(OakIDB, OakIFB, OakIFC, OakIFD, WineIDC, WineIDD, WineIFC, WineIDB, WineIFD) %>% merge(., ChromKey) -> CSSI_Fluc_concat
#rbind(OakIDB, OakIFB, OakIFC, OakIFD, WineIDC, WineIDD, WineIFC, WineIDB) %>% merge(., ChromKey) -> CSSI_Fluc_concat

DatasetKey <- data.frame(Dataset = c("OakI_Dilute_B", "OakI_Fluc_B", "OakI_Fluc_C", "OakI_Fluc_D",
                                     "WineI_Dilute_C", "WineI_Dilute_D", "WineI_Fluc_C", "WineI_Dilute_B", "WineI_Fluc_D"),
                         Parent = c("Oak", "Oak", "Oak", "Oak", "Wine", "Wine", "Wine", "Wine", "Wine"),
                         Bulk = c("Dilute", "Fluconazole", "Fluconazole", "Fluconazole", 
                                  "Dilute", "Dilute", "Fluconazole", "Dilute", "Fluconazole"),
                         Day = c(1,1,2,2,2,2,2,1,2))

CSSI_Fluc_concat %>% left_join(DatasetKey, by = "Dataset") %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> CSSI_Fluc_concat

CSSI_Fluc_concat$CHROM <- factor(CSSI_Fluc_concat$CHROM, levels = ChromKey$chromosomes)

CSSI_Fluc_concat %>% left_join(SNPids, by = c("CHROM", "POS")) -> CSSI_Fluc_concat
save(CSSI_Fluc_concat, file = "CSSI_Fluc_concat.Rdata")
```

# Checking Data (the reason this is no longer prelim analysis)

## Looking at where the data doesn't match in some bulks

CuSO4
```{r, eval = FALSE}
load("RdataFiles/CSSI_CuSO4_concat.Rdata")

#ggplot(CSSI_Fluc_concat, aes(x = GQ.HIGH, color = Bulk)) + geom_density()
CSSI_Fluc_concat %>% #filter(GQ.HIGH > 98) %>% select(-DP.HIGH, - GQ.HIGH, -SNPindex.HIGH, -PL.HIGH, -REF, -ALT) %>%
  pivot_longer(c(AD_REF.HIGH, AD_ALT.HIGH), names_to = "ALT_REF", values_to = "ReadCount") %>% filter(Type != "Alt") %>% na.omit() -> CSSIC_Pivot3

CSSI_Fluc_concat %>% filter(GQ.HIGH > 98) %>% select(-DP.HIGH, - GQ.HIGH, -SNPindex.HIGH, -PL.HIGH, -REF, -ALT) %>%
  pivot_longer(c(AD_REF.HIGH, AD_ALT.HIGH), names_to = "ALT_REF", values_to = "ReadCount") %>% filter(Type != "Alt") %>% na.omit() -> CSSIC_Pivot2

#CSSIF_Pivot2$ReadCount <- CSSIF_Pivot2$ReadCount + 1

#Convert Alt/Ref to Oak/Wine
CSSIC_Pivot2$Allele <- NA
CSSIC_Pivot2$Allele[CSSIC_Pivot2$ALT_REF == "AD_REF.HIGH" & CSSIC_Pivot2$Type == "Wine"] <- "Wine"
CSSIC_Pivot2$Allele[CSSIC_Pivot2$ALT_REF == "AD_REF.HIGH" & CSSIC_Pivot2$Type == "Oak"] <- "Oak"
CSSIC_Pivot2$Allele[CSSIC_Pivot2$ALT_REF == "AD_ALT.HIGH" & CSSIC_Pivot2$Type == "Wine"] <- "Oak"
CSSIC_Pivot2$Allele[CSSIC_Pivot2$ALT_REF == "AD_ALT.HIGH" & CSSIC_Pivot2$Type == "Oak"] <- "Wine"

#Convert all to factors for glm
CSSIC_Pivot2$Dataset <- factor(CSSIC_Pivot2$Dataset)
CSSIC_Pivot2$Parent <- factor(CSSIC_Pivot2$Parent)
CSSIC_Pivot2$Bulk <- factor(CSSIC_Pivot2$Bulk)
CSSIC_Pivot2$Day <- factor(CSSIC_Pivot2$Day)
CSSIC_Pivot2$Allele <- factor(CSSIC_Pivot2$Allele)
CSSIC_Pivot2$CHROM <- factor(CSSIC_Pivot2$CHROM, levels = ChromKey$chromosomes)

save(CSSIC_Pivot2, file = "CSSIC_Pivot2.Rdata")

#NEW CODE
#How many positions have a read count of 0 in each chromosome?
#table(CSSIC_Pivot2$CHROM[CSSIC_Pivot2$ReadCount == 0])
round(100*table(CSSIC_Pivot2$CHROM[CSSIC_Pivot2$ReadCount == 0])/table(CSSIC_Pivot2$CHROM))

#How many of each dataset are missing?

NumbersOfDatasets <- as.data.frame(table(CSSIC_Pivot2$Dataset))
NumbersOfDatasets
NumbersOfDatasets[which(NumbersOfDatasets$Freq == max(NumbersOfDatasets$Freq)),]
NumbersOfDatasets[which(NumbersOfDatasets$Freq == min(NumbersOfDatasets$Freq)),]

CSSIC_Pivot2 %>% group_by(CHROM, POS) %>% summarize(uniqueDatasets = length(unique(Dataset))) -> testingforglm
CSSIC_Pivot2 %>% left_join(testingforglm, by = c("CHROM", "POS")) -> testforweirdnumbers

ggplot(testingforglm, aes(fill = factor(uniqueDatasets), x = CHROM)) + geom_bar(position = "dodge") + ggtitle("With Removing GQ < 99")
ggplot(testingforglm, aes(fill = factor(uniqueDatasets), x = 1)) + geom_bar(position = "dodge") + facet_wrap(~CHROM, scales = "free_y")+ ggtitle("With Removing GQ < 99")

ggplot(testingforglm, aes(fill = factor(uniqueDatasets), x = 1)) + geom_bar(position = "dodge") + facet_wrap(~CHROM, scales = "free_y")+ ggtitle("With Removing GQ < 99")

CSSIC_Pivot3 %>% group_by(CHROM, POS) %>% summarize(uniqueDatasets = length(unique(Dataset))) %>% ggplot(aes(fill = factor(uniqueDatasets), x = 1)) + geom_bar(position = "dodge") + facet_wrap(~CHROM, scales = "free_y") + ggtitle("Without Removing GQ < 99")

testingforglm %>% filter(CHROM == "I") %>% ggplot(aes(fill = factor(uniqueDatasets), x = CHROM)) + geom_bar(position = "dodge") + ggtitle("With Removing GQ < 99")

```


Fluconazole

```{r, eval = FALSE}
load("RdataFiles/CSSI_Fluc_concat.Rdata")

#ggplot(CSSI_Fluc_concat, aes(x = GQ.HIGH, color = Bulk)) + geom_density()
CSSI_Fluc_concat %>% #filter(GQ.HIGH > 98) %>% select(-DP.HIGH, - GQ.HIGH, -SNPindex.HIGH, -PL.HIGH, -REF, -ALT) %>%
  pivot_longer(c(AD_REF.HIGH, AD_ALT.HIGH), names_to = "ALT_REF", values_to = "ReadCount") %>% filter(Type != "Alt") %>% na.omit() -> CSSIF_Pivot3

CSSI_Fluc_concat %>% filter(GQ.HIGH > 98) %>% select(-DP.HIGH, - GQ.HIGH, -SNPindex.HIGH, -PL.HIGH, -REF, -ALT) %>%
  pivot_longer(c(AD_REF.HIGH, AD_ALT.HIGH), names_to = "ALT_REF", values_to = "ReadCount") %>% filter(Type != "Alt") %>% na.omit() -> CSSIF_Pivot2

#CSSIF_Pivot2$ReadCount <- CSSIF_Pivot2$ReadCount + 1

#Convert Alt/Ref to Oak/Wine
CSSIF_Pivot2$Allele <- NA
CSSIF_Pivot2$Allele[CSSIF_Pivot2$ALT_REF == "AD_REF.HIGH" & CSSIF_Pivot2$Type == "Wine"] <- "Wine"
CSSIF_Pivot2$Allele[CSSIF_Pivot2$ALT_REF == "AD_REF.HIGH" & CSSIF_Pivot2$Type == "Oak"] <- "Oak"
CSSIF_Pivot2$Allele[CSSIF_Pivot2$ALT_REF == "AD_ALT.HIGH" & CSSIF_Pivot2$Type == "Wine"] <- "Oak"
CSSIF_Pivot2$Allele[CSSIF_Pivot2$ALT_REF == "AD_ALT.HIGH" & CSSIF_Pivot2$Type == "Oak"] <- "Wine"

#Convert all to factors for glm
CSSIF_Pivot2$Dataset <- factor(CSSIF_Pivot2$Dataset)
CSSIF_Pivot2$Parent <- factor(CSSIF_Pivot2$Parent)
CSSIF_Pivot2$Bulk <- factor(CSSIF_Pivot2$Bulk)
CSSIF_Pivot2$Day <- factor(CSSIF_Pivot2$Day)
CSSIF_Pivot2$Allele <- factor(CSSIF_Pivot2$Allele)
CSSIF_Pivot2$CHROM <- factor(CSSIF_Pivot2$CHROM, levels = ChromKey$chromosomes)

#NEW CODE
#How many positions have a read count of 0 in each chromosome?
#table(CSSIF_Pivot2$CHROM[CSSIF_Pivot2$ReadCount == 0])
round(100*table(CSSIF_Pivot2$CHROM[CSSIF_Pivot2$ReadCount == 0])/table(CSSIF_Pivot2$CHROM))

#How many of each dataset are missing?

NumbersOfDatasets <- as.data.frame(table(CSSIF_Pivot2$Dataset))
NumbersOfDatasets
NumbersOfDatasets[which(NumbersOfDatasets$Freq == max(NumbersOfDatasets$Freq)),]
NumbersOfDatasets[which(NumbersOfDatasets$Freq == min(NumbersOfDatasets$Freq)),]

CSSIF_Pivot2 %>% group_by(CHROM, POS) %>% summarize(uniqueDatasets = length(unique(Dataset))) -> testingforglm
CSSIF_Pivot2 %>% left_join(testingforglm, by = c("CHROM", "POS")) -> testforweirdnumbers

ggplot(testingforglm, aes(fill = factor(uniqueDatasets), x = CHROM)) + geom_bar(position = "dodge") + ggtitle("With Removing GQ < 99")
ggplot(testingforglm, aes(fill = factor(uniqueDatasets), x = 1)) + geom_bar(position = "dodge") + facet_wrap(~CHROM, scales = "free_y")+ ggtitle("With Removing GQ < 99")

ggplot(testingforglm, aes(fill = factor(uniqueDatasets), x = 1)) + geom_bar(position = "dodge") + facet_wrap(~CHROM, scales = "free_y")+ ggtitle("With Removing GQ < 99")

CSSIF_Pivot3 %>% group_by(CHROM, POS) %>% summarize(uniqueDatasets = length(unique(Dataset))) %>% ggplot(aes(fill = factor(uniqueDatasets), x = 1)) + geom_bar(position = "dodge") + facet_wrap(~CHROM, scales = "free_y") + ggtitle("Without Removing GQ < 99")

testingforglm %>% filter(CHROM == "I") %>% ggplot(aes(fill = factor(uniqueDatasets), x = CHROM)) + geom_bar(position = "dodge") + ggtitle("With Removing GQ < 99")

```


# Run the glm

### Reformat for GLM - pivot

CuSO4

```{r, eval = FALSE}
load("RdataFiles/CSSI_CuSO4_concat.Rdata")

#ggplot(CSSI_CuSO4_concat, aes(x = GQ.HIGH, color = Bulk)) + geom_density()
CSSI_CuSO4_concat %>% filter(GQ.HIGH > 98) %>% select(-DP.HIGH, - GQ.HIGH, -SNPindex.HIGH, -PL.HIGH, -REF, -ALT) %>% pivot_longer(c(AD_REF.HIGH, AD_ALT.HIGH), names_to = "ALT_REF", values_to = "ReadCount") %>% filter(Type != "Alt") %>% na.omit() -> CSSIC_Pivot2

#REMOVE POSITIONS WHERE THERE ISN'T ALL BULKS
CSSIC_Pivot2 %>% group_by(CHROM, POS) %>% summarize(uniqueDatasets = length(unique(Dataset))) -> testingforglm
CSSIC_Pivot2 %>% left_join(testingforglm, by = c("CHROM", "POS")) %>% filter(uniqueDatasets == 4) -> CSSI_CuSO4_Pivot

#Convert Alt/Ref to Oak/Wine
CSSI_CuSO4_Pivot$Allele <- NA
CSSI_CuSO4_Pivot$Allele[CSSI_CuSO4_Pivot$ALT_REF == "AD_REF.HIGH" & CSSI_CuSO4_Pivot$Type == "Wine"] <- "Wine"
CSSI_CuSO4_Pivot$Allele[CSSI_CuSO4_Pivot$ALT_REF == "AD_REF.HIGH" & CSSI_CuSO4_Pivot$Type == "Oak"] <- "Oak"
CSSI_CuSO4_Pivot$Allele[CSSI_CuSO4_Pivot$ALT_REF == "AD_ALT.HIGH" & CSSI_CuSO4_Pivot$Type == "Wine"] <- "Oak"
CSSI_CuSO4_Pivot$Allele[CSSI_CuSO4_Pivot$ALT_REF == "AD_ALT.HIGH" & CSSI_CuSO4_Pivot$Type == "Oak"] <- "Wine"

#Convert all to factors for glm
CSSI_CuSO4_Pivot$Dataset <- factor(CSSI_CuSO4_Pivot$Dataset)
CSSI_CuSO4_Pivot$Parent <- factor(CSSI_CuSO4_Pivot$Parent)
CSSI_CuSO4_Pivot$Bulk <- factor(CSSI_CuSO4_Pivot$Bulk)
CSSI_CuSO4_Pivot$Allele <- factor(CSSI_CuSO4_Pivot$Allele)
CSSI_CuSO4_Pivot$CHROM <- factor(CSSI_CuSO4_Pivot$CHROM, levels = ChromKey$chromosomes)

save(CSSI_CuSO4_Pivot, file = "RdataFiles/CSSI_CuSO4_Pivot.Rdata")

```

Fluconazole

```{r, eval = FALSE}
load("RdataFiles/CSSI_Fluc_concat.Rdata")

#ggplot(CSSI_Fluc_concat, aes(x = GQ.HIGH, color = Bulk)) + geom_density()
CSSI_Fluc_concat %>% filter(GQ.HIGH > 98) %>% select(-DP.HIGH, - GQ.HIGH, -SNPindex.HIGH, -PL.HIGH, -REF, -ALT) %>% pivot_longer(c(AD_REF.HIGH, AD_ALT.HIGH), names_to = "ALT_REF", values_to = "ReadCount") %>% filter(Type != "Alt") %>% na.omit() -> CSSIF_Pivot2

#REMOVE POSITIONS WHERE THERE ISN'T ALL BULKS
CSSIF_Pivot2 %>% group_by(CHROM, POS) %>% summarize(uniqueDatasets = length(unique(Dataset))) -> testingforglm
CSSIF_Pivot2 %>% left_join(testingforglm, by = c("CHROM", "POS")) %>% filter(uniqueDatasets == 9) -> CSSI_Fluc_Pivot

#Convert Alt/Ref to Oak/Wine
CSSI_Fluc_Pivot$Allele <- NA
CSSI_Fluc_Pivot$Allele[CSSI_Fluc_Pivot$ALT_REF == "AD_REF.HIGH" & CSSI_Fluc_Pivot$Type == "Wine"] <- "Wine"
CSSI_Fluc_Pivot$Allele[CSSI_Fluc_Pivot$ALT_REF == "AD_REF.HIGH" & CSSI_Fluc_Pivot$Type == "Oak"] <- "Oak"
CSSI_Fluc_Pivot$Allele[CSSI_Fluc_Pivot$ALT_REF == "AD_ALT.HIGH" & CSSI_Fluc_Pivot$Type == "Wine"] <- "Oak"
CSSI_Fluc_Pivot$Allele[CSSI_Fluc_Pivot$ALT_REF == "AD_ALT.HIGH" & CSSI_Fluc_Pivot$Type == "Oak"] <- "Wine"

#Convert all to factors for glm
CSSI_Fluc_Pivot$Dataset <- factor(CSSI_Fluc_Pivot$Dataset)
CSSI_Fluc_Pivot$Parent <- factor(CSSI_Fluc_Pivot$Parent)
CSSI_Fluc_Pivot$Bulk <- factor(CSSI_Fluc_Pivot$Bulk)
CSSI_Fluc_Pivot$Day <- factor(CSSI_Fluc_Pivot$Day)
CSSI_Fluc_Pivot$Allele <- factor(CSSI_Fluc_Pivot$Allele)
CSSI_Fluc_Pivot$CHROM <- factor(CSSI_Fluc_Pivot$CHROM, levels = ChromKey$chromosomes)

save(CSSI_Fluc_Pivot, file = "RdataFiles/CSSI_Fluc_Pivot_2.Rdata")

```

```{r, eval = FALSE}
#CSSI_Fluc_Pivot$ReadCount <- CSSI_Fluc_Pivot$ReadCount + 1


#Test GLM
glmtest <- glm(Allele ~ Bulk*Parent+Day, weights = ReadCount, family = binomial, data = subset(CSSI_Fluc_Pivot, POS == 925 & CHROM == "I"))
glmtest
summary(glmtest)$coefficients [11:15]

xtabs(~ Bulk + Parent + Allele + Day, data = subset(CSSI_Fluc_Pivot, POS == 925 & CHROM == "I"))

length(unique(CSSI_Fluc_Pivot$POS[CSSI_Fluc_Pivot$CHROM == "I"]))

```

## Plot for Smoothed GLM

Data Processing: function for running glm and windows

```{r}
#testing
# lrP <- subset(CSSI_Fluc_Pivot_excl, CHROM == "I")
# c <- "I"
# chr <- "I"
# windowsize = 100

################################################################################
#function
BSA_GLM_CSSIF <-  function(lrP, chr = "II", windowsize = 100){
  
  lrP <- subset(lrP, CHROM == chr)
  
  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- glm(Allele ~ Bulk*Parent+Day, weights = ReadCount, family = binomial, data = lrP[lrP$POS == i,])
      c(c, i, summary(res)$coefficients[11:15])
    }
  
    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkFluconazole", "parentWine", "Day2", "Interaction")
    Results[,2] <- as.numeric(Results[,2])
    Results[,4] <- as.numeric(Results[,4])
    Results[,5] <- as.numeric(Results[,5])
    Results[,6] <- as.numeric(Results[,6])
    Results[,7] <- as.numeric(Results[,7])
    
    Results %>% arrange(POS) %>% select(-Intercept) -> Results
  }
  
 
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkFluconazole[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                res_day <- mean(Results$Day2[(i-windowsize):(i+windowsize)])

                #print CHROM, index, POS, and mean
                c(chr, i, Results$POS[i], 
                  Results$bulkFluconazole[i], Results$parentWine[i],Results$Interaction[i], Results$Day2[i],
                  res_bulk, res_parent, res_day, res_int)}
  
  WResult <- as.data.frame(WResult)
  
  colnames(WResult) <- c("CHROM", "Index", "POS", 
                         "Bulk_Z", "Parent_Z", "Interaction_Z", "Day_Z",
                         "Bulk_Zprime", "Parent_Zprime", "Day_Zprime","Interaction_Zprime")
  
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  WResult[,10] <- as.numeric(WResult[,10])
  WResult[,11] <- as.numeric(WResult[,11])
  
  return(WResult)
  }
```


Combine data; do not evaluate 

```{r, eval = FALSE}

load("RdataFiles/CSSI_Fluc_Pivot_2.Rdata")

zscore_1 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "I")
zscore_2 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "II")
zscore_3 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "III")
zscore_4 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "IV")
Sys.sleep(5)
zscore_5 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "V")
zscore_6 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "VI")
zscore_7 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "VII")
zscore_8 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "VIII")
Sys.sleep(5)
zscore_9 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "IX")
zscore_10 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "X")
zscore_11 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "XI")
zscore_12 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "XII")
Sys.sleep(5)
zscore_13 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "XIII")
zscore_14 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "XIV")
zscore_15 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "XV")
zscore_16 <- BSA_GLM_CSSIF(CSSI_Fluc_Pivot, "XVI")
Sys.sleep(5)
CSSI_F_ZScores_GLM <- rbind(zscore_1, zscore_2, zscore_3, zscore_4, zscore_5, zscore_6, 
                     zscore_7, zscore_8, zscore_9, zscore_10, zscore_11, 
                     zscore_12, zscore_13, zscore_14, zscore_15, zscore_16)

CSSI_F_ZScores_GLM$CHROM <- factor(CSSI_F_ZScores_GLM$CHROM, levels = ChromKey$chromosomes)

save(CSSI_F_ZScores_GLM, file = "CSSI_F_ZScores_GLM_allexcl.Rdata")
```

### Load data and plot

```{r}

load("RdataFiles/CSSI_F_ZScores_GLM_allexcl.Rdata")

JaroszKey <- data.frame(POS = c(318265, 92671, 43887, 891344, 927577, 122347, 510241, 885015, 963579, 537874, 171534, 619798),
                        CHROM = c("II","III","IV","IV","IV","VIII","XII","XII","XII","XV","XVI","XVI"),
                        Yname = c("YBR041W", "YCL017C",  "YDL229W", "YDR213W", "YDR232W", "YHR007C", 
                                 "YLR176C", "YLR382C", "YLR420W", "YOR114W", "YPL200W", "YPR026W"),
                        name = c("FAT1", "NFS1",  "SSB1", "UPC2", "HEM1", "ERG11", 
                                 "RFX1", "NAM2", "URA4", "DPI34 ", "CSM4", "ATH1"))

InteractionNetwork <- data.frame(POS = c(889751, 469092, 215987, 120091, 398628),
                                 CHROM = c("IV", "VII", "XIV", "VIII", "VII"),
                                 name = c("UPC2", "PDR1", "PDR16", "ERG11", "OLE1"))


Z_Fig_bulk <- CSSI_F_ZScores_GLM %>% ggplot(aes(x = POS, y = Bulk_Zprime)) +
  geom_point(aes(x = POS, y = Bulk_Z), color = "#345F6F", alpha= 0.1) + theme_minimal() +
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  geom_point(color = "black", size = 0.2) + facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  geom_hline(yintercept = c(-1.96, 1.96), color = "black")+
  ggtitle("CSSI Fluconazole Selection Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine") #+ ylim(-12,12)


Z_Fig_int <- CSSI_F_ZScores_GLM %>%  ggplot(aes(x = POS, y = Interaction_Zprime)) +
  geom_point(aes(x = POS, y = Interaction_Z), color = "#D7335C", alpha= 0.1) + theme_minimal() +
  geom_point(color = "black", size = 0.2) + facet_grid(~CHROM, scales = "free_x", space = "free_x") +
    scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
    geom_hline(yintercept = c(-1.96, 1.96), color = "black")+
ggtitle("CSSI Fluconazole Selection x Background Interaction Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")#+ ylim(-12,12)

Z_Fig_parent <- CSSI_F_ZScores_GLM %>%  ggplot(aes(x = POS, y = Parent_Zprime)) +
  geom_point(aes(x = POS, y = Parent_Z), color = "#FFB05C", alpha= 0.1) + theme_minimal() +
  geom_point(color = "black", size = 0.2) + facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  geom_hline(yintercept = c(-1.96, 1.96), color = "black")+
  ggtitle("CSS Fluconazole Parental Background Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")#+ ylim(-12,12)

Z_Fig_Day <- CSSI_F_ZScores_GLM  %>% ggplot(aes(x = POS, y = Day_Zprime)) +
  geom_point(aes(x = POS, y = Day_Z), color = "gray", alpha= 0.1) + theme_minimal() +
  geom_point(color = "black", size = 0.2) + facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  geom_hline(yintercept = c(-1.96, 1.96), color = "black")+
  ggtitle("CSS I Fluconazole Replicate Day scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")#+ ylim(-12,12)


 # ggsave(Z_Fig_bulk, file = "CSSI_Fluc_Z_Fig_bulk.png", width = 5, height = 4)
 # ggsave(Z_Fig_int, file = "CSSI_Fluc_Z_Fig_int.png", width = 5, height = 4)
 # ggsave(Z_Fig_parent, file = "CSSI_Fluc_Z_Fig_parent.png", width = 5, height = 4)
 # ggsave(Z_Fig_Day, file = "CSSI_Fluc_Z_Fig_Day.png", width = 5, height = 4)
 # ggsave(Z_Fig_overlay, file = "CSSI_Fluc_Z_Fig_overlay.png", width = 5, height = 4)

Z_Fig_bulk
Z_Fig_int
Z_Fig_parent
Z_Fig_Day

CSSI_F_ZScores_GLM %>% ggplot(aes(x = POS, y = Bulk_Zprime)) + 
      geom_vline(data = JaroszKey, aes(xintercept = POS), color = "black", alpha = 0.5, size = 0.7) +
      geom_vline(data = InteractionNetwork, aes(xintercept = POS), color = "red", alpha = 1, size = 0.7, linetype = "dashed") +
  
  geom_point(color = "#345F6F", size = 0.2, alpha = 0.8) + 
  geom_point(aes(x = POS, y = Parent_Zprime), color = "#FFB05C", alpha= 0.8, size = 0.2) +
  geom_point(aes(x = POS, y = Interaction_Zprime), color = "#D7335C", alpha= 0.8, size = 0.2) +
  geom_point(aes(x = POS, y = Day_Zprime), color = "gray", alpha= 0.8, size = 0.2) +
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  geom_hline(yintercept = c(-1.96, 1.96), color = "gray")+  geom_hline(yintercept = c(-1, 1), color = "black")+

  geom_text(mapping = aes(x = POS,y = -10,label = name,
                            #hjust = -1, 
                          vjust = 1,
                            angle = 90), color = "black", size = 1.8, data = JaroszKey) +
  geom_text(mapping = aes(x = POS,y = -12,label = name,
                            #hjust = -1, 
                          vjust = -0.8,
                            angle = 90), color = "red", size = 1.8, data = InteractionNetwork) +
  ggtitle("CSSI Fluconazole | Comparison of Smoothed Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
```

Removing Chr I from plot

```{r}
load("RdataFiles/CSSI_F_ZScores_GLM_allexcl.Rdata")

CSSI_F_ZScores_GLM %>% filter(CHROM != "I") %>% ggplot(aes(x = POS, y = Bulk_Zprime)) + 
  geom_point(color = "#345F6F", size = 0.2, alpha = 0.8) + 
  geom_point(aes(x = POS, y = Parent_Zprime), color = "#FFB05C", alpha= 0.8, size = 0.2) +
  geom_point(aes(x = POS, y = Interaction_Zprime), color = "#D7335C", alpha= 0.8, size = 0.2) +
    geom_point(aes(x = POS, y = Day_Zprime), color = "gray", alpha= 0.8, size = 0.2) +

  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  ggtitle("CSSI Fluconazole | Comparison of Smoothed Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  #ylim(-5, 5) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
```

# Without Using Day as a Factor

```{r}
#testing
# lrP <- subset(CSSI_CuSO4_Pivot_exclmD, CHROM == "I")
# c <- "I"
# chr <- "I"
# windowsize = 100
# i <- 925

################################################################################
#function
BSA_GLM_CSSIF_noDay <-  function(lrP, chr = "II", windowsize = 100){
  
  lrP <- subset(lrP, CHROM == chr)
  
  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- glm(Allele ~ Bulk*Parent, weights = ReadCount, family = binomial, data = lrP[lrP$POS == i,])
      c(c, i, summary(res)$coefficients[9:12])
    }
  
    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkFluconazole", "parentWine", "Interaction")
    Results[,2] <- as.numeric(Results[,2])
    Results[,3] <- as.numeric(Results[,3])
    Results[,4] <- as.numeric(Results[,4])
    Results[,5] <- as.numeric(Results[,5])
    Results[,6] <- as.numeric(Results[,6])

    Results %>% arrange(POS) %>% select(-Intercept) -> Results
  }
  
 
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkFluconazole[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])

                #print CHROM, index, POS, and mean
                c(chr, i, Results$POS[i], 
                  Results$bulkFluconazole[i], Results$parentWine[i],Results$Interaction[i], 
                  res_bulk, res_parent, res_int)}
  
  WResult <- as.data.frame(WResult)
  
  colnames(WResult) <- c("CHROM", "Index", "POS", 
                         "Bulk_Z", "Parent_Z", "Interaction_Z", 
                         "Bulk_Zprime", "Parent_Zprime", "Interaction_Zprime")
  
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  
  return(WResult)
  }
```

### Removing ones with not enough categories

CuSO4

```{r, eval = FALSE}
load("RdataFiles/CSSI_CuSO4_Pivot.Rdata")
CSSI_CuSO4_Pivot %>% group_by(CHROM, POS) %>% summarize(uniqueParents = length(unique(Parent)),
                                                uniqueBulks = length(unique(Bulk))) -> Ctestingforglm

Ctestingforglm$exclusion <- NA
Ctestingforglm$exclusion[Ctestingforglm$uniqueParents == 1 | Ctestingforglm$uniqueBulks == 1 ] <- "X"

CSSI_CuSO4_Pivot %>% left_join(Ctestingforglm, by = c("CHROM", "POS")) %>% filter(is.na(exclusion)) %>% select(-uniqueParents, -uniqueBulks, -exclusion) -> CSSI_CuSO4_Pivot_exclmD

save(CSSI_CuSO4_Pivot_exclmD, file = "RdataFiles/CSSI_CuSO4_Pivot_exclmD.Rdata")

```

Fluconazole
```{r, eval = FALSE}
CSSI_Fluc_Pivot %>% group_by(CHROM, POS) %>% summarize(uniqueParents = length(unique(Parent)),
                                                uniqueBulks = length(unique(Bulk)),
                                                uniqueDays = length(unique(Day))) -> testingforglm

testingforglm$exclusion <- NA
testingforglm$exclusion[testingforglm$uniqueParents == 1 | testingforglm$uniqueBulks == 1 ] <- "X"

CSSI_Fluc_Pivot %>% left_join(testingforglm, by = c("CHROM", "POS")) %>% filter(is.na(exclusion)) %>% select(-uniqueParents, -uniqueBulks, -uniqueDays, -exclusion) -> CSSI_Fluc_Pivot_exclmD

save(CSSI_Fluc_Pivot_exclmD, file = "CSSI_Fluc_Pivot_exclmD.Rdata")

```

Running Function

```{r, eval = FALSE}

load("RdataFiles/CSSI_CuSO4_Pivot_exclmD.Rdata")

zscore_1 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "I")
zscore_2 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "II")
Sys.sleep(5)

zscore_3 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "III")
zscore_4 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "IV")
Sys.sleep(5)

zscore_5 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "V")
zscore_6 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "VI")
Sys.sleep(5)

zscore_7 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "VII")
zscore_8 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "VIII")
Sys.sleep(5)

zscore_9 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "IX")
zscore_10 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "X")
Sys.sleep(5)

zscore_11 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "XI")
zscore_12 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "XII")
Sys.sleep(5)

zscore_13 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "XIII")
zscore_14 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "XIV")
Sys.sleep(5)

zscore_15 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "XV")
zscore_16 <- BSA_GLM_CSSIF_noDay(CSSI_CuSO4_Pivot_exclmD, "XVI")
Sys.sleep(5)

CSSI_C_ZScores_GLM_noday <- rbind(zscore_1, zscore_2, zscore_3, zscore_4, zscore_5, zscore_6, 
                     zscore_7, zscore_8, zscore_9, zscore_10, zscore_11, 
                     zscore_12, zscore_13, zscore_14, zscore_15, zscore_16)

CSSI_C_ZScores_GLM_noday$CHROM <- factor(CSSI_C_ZScores_GLM_noday$CHROM, levels = ChromKey$chromosomes)

save(CSSI_C_ZScores_GLM_noday, file = "RdataFiles/CSSI_C_ZScores_GLM_noday.Rdata")
```

```{r}
 
load("RdataFiles/CSSI_C_ZScores_GLM_noday.Rdata")


Z_Fig_bulk_CuSO4 <- CSSI_C_ZScores_GLM_noday %>% ggplot(aes(x = POS, y = Bulk_Zprime)) +
  geom_point(aes(x = POS, y = Bulk_Z), color = "#345F6F", alpha= 0.1) + theme_minimal() +
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_C_ZScores_GLM_noday$POS), by = 1e5), name = "Genomic Position") +
  geom_point(color = "black", size = 0.2) + facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  geom_hline(yintercept = c(-1.96, 1.96), color = "black")+
  ggtitle("CSSI CuSO4 Selection Z scores (-D)") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine") #+ ylim(-12,12)


Z_Fig_int_CuSO4 <- CSSI_C_ZScores_GLM_noday %>%  ggplot(aes(x = POS, y = Interaction_Zprime)) +
  geom_point(aes(x = POS, y = Interaction_Z), color = "#D7335C", alpha= 0.1) + theme_minimal() +
  geom_point(color = "black", size = 0.2) + facet_grid(~CHROM, scales = "free_x", space = "free_x") +
    scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_C_ZScores_GLM_noday$POS), by = 1e5), name = "Genomic Position") +
    geom_hline(yintercept = c(-1.96, 1.96), color = "black")+
ggtitle("CSSI CuSO4 Selection x Background Interaction Z scores (-D)") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")#+ ylim(-12,12)

Z_Fig_parent_CuSO4 <- CSSI_C_ZScores_GLM_noday %>%  ggplot(aes(x = POS, y = Parent_Zprime)) +
  geom_point(aes(x = POS, y = Parent_Z), color = "#FFB05C", alpha= 0.1) + theme_minimal() +
  geom_point(color = "black", size = 0.2) + facet_grid(~CHROM, scales = "free_x", space = "free_x") +
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_C_ZScores_GLM_noday$POS), by = 1e5), name = "Genomic Position") +
  geom_hline(yintercept = c(-1.96, 1.96), color = "black")+
  ggtitle("CSS CuSO4 Parental Background Z scores (-D)") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")#+ ylim(-12,12)

CSSI_C_ZScores_GLM_noday %>%  ggplot(aes(x = POS, y = Bulk_Zprime)) + 
  geom_point(color = "#345F6F", size = 0.2, alpha = 0.8) + 
  geom_point(aes(x = POS, y = Parent_Zprime), color = "#FFB05C", alpha= 0.8, size = 0.2) +
  geom_point(aes(x = POS, y = Interaction_Zprime), color = "#D7335C", alpha= 0.8, size = 0.2) +

  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_C_ZScores_GLM_noday$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  ggtitle("CSSI Fluconazole | Comparison of Smoothed Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  #ylim(-5, 5) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")

Z_Fig_bulk_CuSO4
Z_Fig_parent_CuSO4
Z_Fig_int_CuSO4

```

# Compare CuSO4 and Fluconazole overlays

```{r}
load("RdataFiles/CSSI_C_ZScores_GLM_noday.Rdata")
load("RdataFiles/CSSI_F_ZScores_GLM_allexcl.Rdata")


CSSI_C_ZScores_GLM_noday %>% #filter(CHROM != "I") %>% 
  ggplot(aes(x = POS, y = Bulk_Zprime)) + 
  geom_point(color = "#345F6F", size = 0.2, alpha = 0.8) + 
  geom_point(aes(x = POS, y = Parent_Zprime), color = "#FFB05C", alpha= 0.8, size = 0.2) +
  geom_point(aes(x = POS, y = Interaction_Zprime), color = "#D7335C", alpha= 0.8, size = 0.2) +

  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_C_ZScores_GLM_noday$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  ggtitle("CSSI CuSO4 | Comparison of Smoothed Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  ylim(-10, 10) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")

CSSI_F_ZScores_GLM %>% #filter(CHROM != "I") %>% 
  ggplot(aes(x = POS, y = Bulk_Zprime)) + 
  geom_point(color = "#345F6F", size = 0.2, alpha = 0.8) + 
  geom_point(aes(x = POS, y = Parent_Zprime), color = "#FFB05C", alpha= 0.8, size = 0.2) +
  geom_point(aes(x = POS, y = Interaction_Zprime), color = "#D7335C", alpha= 0.8, size = 0.2) +
    geom_point(aes(x = POS, y = Day_Zprime), color = "gray", alpha= 0.8, size = 0.2) +

  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  ggtitle("CSSI Fluconazole | Comparison of Smoothed Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  ylim(-10, 10) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")

```

# Comparing each bulk for significance between nulls

Make a new function for this

```{r}
# #testing
# lrP <- subset(WineDilute, CHROM == "I")
# c <- "I"
# chr <- "I"
# windowsize = 100
# i <- 122795 

################################################################################
#function
BSA_GLM_Nulls <-  function(lrP, chr = "II", windowsize = 100){
  
  lrP <- subset(lrP, CHROM == chr)
  #remove those that only have one factor for the dataset
  lrP %>% group_by(POS) %>% summarize(uniqueData = length(unique(Dataset))) -> testingforglm
  
  table(testingforglm$uniqueData)
  
  testingforglm$exclusion <- NA
  testingforglm$exclusion[testingforglm$uniqueData == 1] <- "X"
  
  lrP %>% left_join(testingforglm, by = c("POS")) %>% filter(is.na(exclusion)) %>% select(-uniqueData, -exclusion) -> lrP  
  #Run the glm on that chromosome
  AllResults <- list()
  Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
    res <- glm(Allele ~ Dataset, weights = ReadCount, family = binomial, data = lrP[lrP$POS == i,])
    c(chr, i, summary(res)$coefficients[5:6])
  }

  #Format the results for the next function
  Results <- as.data.frame(Results)
  colnames(Results) <- c("CHROM", "POS", "Intercept", "Dataset")
  Results[,2] <- as.numeric(Results[,2])
  Results[,4] <- as.numeric(Results[,4])
  #Results[,5] <- as.numeric(Results[,5])
  # Results[,6] <- as.numeric(Results[,6])
  # Results[,7] <- as.numeric(Results[,7])
  
  Results %>% arrange(POS) %>% select(-Intercept) -> Results

  
 
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_dataset <- mean(Results$Dataset[(i-windowsize):(i+windowsize)])
                # res_bulk <- mean(Results$bulkFluconazole[(i-windowsize):(i+windowsize)])
                # res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                # res_day <- mean(Results$Day2[(i-windowsize):(i+windowsize)])

                #print CHROM, index, POS, and mean
                c(chr, i, Results$POS[i], Results$Dataset[i], res_dataset)}
  
  WResult <- as.data.frame(WResult)
  
  colnames(WResult) <- c("CHROM", "Index", "POS", 
                         "Dataset_Z", "Dataset_Zprime")
  
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  # WResult[,6] <- as.numeric(WResult[,6])
  # WResult[,7] <- as.numeric(WResult[,7])
  # WResult[,8] <- as.numeric(WResult[,8])
  # WResult[,9] <- as.numeric(WResult[,9])
  # WResult[,10] <- as.numeric(WResult[,10])
  # WResult[,11] <- as.numeric(WResult[,11])
  
  return(WResult)
  }
```

Run for just the dilute samples

```{r, eval = FALSE}
load("RdataFiles/CSSI_Fluc_Pivot_2.Rdata")

CompareNulls <- function(TableToAnalyze = CSSI_Fluc_Pivot, D1 = "WineI_Dilute_B", D2 = "WineI_Dilute_C"){
  TableToAnalyze %>% filter(Dataset == D1 | Dataset == D2) -> WineDilute

  zscore_1 <- BSA_GLM_Nulls(WineDilute, "I")
  zscore_2 <- BSA_GLM_Nulls(WineDilute, "II")
  zscore_3 <- BSA_GLM_Nulls(WineDilute, "III")
  zscore_4 <- BSA_GLM_Nulls(WineDilute, "IV")
  Sys.sleep(5)
  zscore_5 <- BSA_GLM_Nulls(WineDilute, "V")
  zscore_6 <- BSA_GLM_Nulls(WineDilute, "VI")
  zscore_7 <- BSA_GLM_Nulls(WineDilute, "VII")
  zscore_8 <- BSA_GLM_Nulls(WineDilute, "VIII")
  Sys.sleep(5)
  zscore_9 <- BSA_GLM_Nulls(WineDilute, "IX")
  zscore_10 <- BSA_GLM_Nulls(WineDilute, "X")
  zscore_11 <- BSA_GLM_Nulls(WineDilute, "XI")
  zscore_12 <- BSA_GLM_Nulls(WineDilute, "XII")
  Sys.sleep(5)
  zscore_13 <- BSA_GLM_Nulls(WineDilute, "XIII")
  zscore_14 <- BSA_GLM_Nulls(WineDilute, "XIV")
  zscore_15 <- BSA_GLM_Nulls(WineDilute, "XV")
  zscore_16 <- BSA_GLM_Nulls(WineDilute, "XVI")
  Sys.sleep(5)
  
  CSSI_F_Wine_BC_Dilute_ZScores_GLM <- rbind(zscore_1, zscore_2, zscore_3, zscore_4, zscore_5, zscore_6, 
                       zscore_7, zscore_8, zscore_9, zscore_10, zscore_11, 
                       zscore_12, zscore_13, zscore_14, zscore_15, zscore_16)
  
  CSSI_F_Wine_BC_Dilute_ZScores_GLM$CHROM <- factor(CSSI_F_Wine_BC_Dilute_ZScores_GLM$CHROM, levels = ChromKey$chromosomes)
  return(CSSI_F_Wine_BC_Dilute_ZScores_GLM)
}

CSSI_F_Wine_BC_Dilute_ZScores_GLM <- CompareNulls(D1 = "WineI_Dilute_B", D2 = "WineI_Dilute_C")
save(CSSI_F_Wine_BC_Dilute_ZScores_GLM, file = "CSSI_F_Wine_BC_Dilute_ZScores_GLM.Rdata")

CSSI_F_Wine_BD_Dilute_ZScores_GLM <- CompareNulls(D1 = "WineI_Dilute_B", D2 = "WineI_Dilute_D")
# CSSI_F_Wine_BD_Dilute_ZScores_GLM <- CSSI_F_Wine_BC_Dilute_ZScores_GLM
save(CSSI_F_Wine_BD_Dilute_ZScores_GLM, file = "CSSI_F_Wine_BD_Dilute_ZScores_GLM.Rdata")

CSSI_F_OakFluc_BC_Dilute_ZScores_GLM <- CompareNulls(D1 = "OakI_Fluc_B", D2 = "OakI_Fluc_C")
save(CSSI_F_OakFluc_BC_Dilute_ZScores_GLM, file = "CSSI_F_OakFluc_BC_Dilute_ZScores_GLM.Rdata")

CSSI_F_WineFluc_CD_Dilute_ZScores_GLM <- CompareNulls(D1 = "WineI_Fluc_D", D2 = "WineI_Fluc_C")
save(CSSI_F_WineFluc_CD_Dilute_ZScores_GLM, file = "CSSI_F_WineFluc_CD_Dilute_ZScores_GLM.Rdata")
```

Plotting the results of Wine Dilute B and Wine Dilute C

```{r}
load("RdataFiles/CSSI_F_Wine_BC_Dilute_ZScores_GLM.Rdata")

CSSI_F_Wine_BC_Dilute_ZScores_GLM %>% ggplot(aes(x = POS, y = Dataset_Z)) + 
      
  geom_point(color = "violet", size = 0.2, alpha = 0.1) + 
  geom_point(aes(x = POS, y = Dataset_Zprime), color = "black", alpha= 0.8, size = 0.2) +
  
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_Wine_BC_Dilute_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  geom_hline(yintercept = c(-1.96, 1.96), color = "gray")+
  
  ggtitle("Wine Dilute B vs Wine Dilute C") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
```

```{r}
load("RdataFiles/CSSI_F_OakFluc_BC_Dilute_ZScores_GLM.Rdata")

CSSI_F_OakFluc_BC_Dilute_ZScores_GLM %>% ggplot(aes(x = POS, y = Dataset_Z)) + 
      
  geom_point(color = "violet", size = 0.2, alpha = 0.1) + 
  geom_point(aes(x = POS, y = Dataset_Zprime), color = "black", alpha= 0.8, size = 0.2) +
  
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_Wine_BC_Dilute_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  geom_hline(yintercept = c(-1.96, 1.96), color = "gray")+
  
  ggtitle("Oak Fluc B vs Oak Fluc C") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
```
```{r}
load("RdataFiles/CSSI_F_WineFluc_CD_Dilute_ZScores_GLM.Rdata")

CSSI_F_WineFluc_CD_Dilute_ZScores_GLM %>% ggplot(aes(x = POS, y = Dataset_Z)) + 
      
  geom_point(color = "violet", size = 0.2, alpha = 0.1) + 
  geom_point(aes(x = POS, y = Dataset_Zprime), color = "black", alpha= 0.8, size = 0.2) +
  
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_Wine_BC_Dilute_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  geom_hline(yintercept = c(-1.96, 1.96), color = "gray")+
  
  ggtitle("Wine Fluc D vs Wine Fluc C") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
```

```{r, eval = FALSE}
load("RdataFiles/CSSI_F_Wine_BD_Dilute_ZScores_GLM.Rdata")

CSSI_F_Wine_BD_Dilute_ZScores_GLM %>% ggplot(aes(x = POS, y = Dataset_Z)) + 
      
  geom_point(color = "violet", size = 0.2, alpha = 0.1) + 
  geom_point(aes(x = POS, y = Dataset_Zprime), color = "black", alpha= 0.8, size = 0.2) +
  
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_Wine_BC_Dilute_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  geom_hline(yintercept = c(-1.96, 1.96), color = "gray")+
  
  ggtitle("Wine D vs B Dilute") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
```


Combining these replicates

```{r, eval = FALSE}
load("RdataFiles/CSSI_F_WineFluc_CD_Dilute_ZScores_GLM.Rdata")
load("RdataFiles/CSSI_F_OakFluc_BC_Dilute_ZScores_GLM.Rdata")
load("RdataFiles/CSSI_F_Wine_BC_Dilute_ZScores_GLM.Rdata")
load("RdataFiles/CSSI_F_Wine_BD_Dilute_ZScores_GLM.Rdata")

CSSI_F_WineFluc_CD_Dilute_ZScores_GLM$Comparison <- "WineFluc_CD"
CSSI_F_OakFluc_BC_Dilute_ZScores_GLM$Comparison <- "OakFluc_BC"
CSSI_F_Wine_BC_Dilute_ZScores_GLM$Comparison <- "WineDilute_BC"
CSSI_F_Wine_BD_Dilute_ZScores_GLM$Comparison <- "WineDilute_BD"

NullComparison <- rbind(CSSI_F_WineFluc_CD_Dilute_ZScores_GLM,
                        CSSI_F_OakFluc_BC_Dilute_ZScores_GLM,
                        CSSI_F_Wine_BC_Dilute_ZScores_GLM,
                        CSSI_F_Wine_BD_Dilute_ZScores_GLM)

NullComparison %>% ggplot(aes(x = POS, y = Dataset_Zprime, color = Comparison)) + geom_line() + 
  scale_x_continuous(breaks = seq(from = 0, to = max(NullComparison$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  geom_hline(yintercept = c(-1.96, 1.96), color = "gray")+
  
  ggtitle("All Null Replicate Comparisons") + theme(axis.text.x=element_blank(),axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")

```

Adding in the full plot just to compare

```{r, eval = FALSE}
load("RdataFiles/CSSI_F_ZScores_GLM_allexcl.Rdata")

CSSI_F_ZScores_GLM %>% ggplot(aes(x = POS, y = Bulk_Zprime)) + 
      # geom_vline(data = JaroszKey, aes(xintercept = POS), color = "black", alpha = 0.5, size = 0.7) +
      # geom_vline(data = InteractionNetwork, aes(xintercept = POS), color = "red", alpha = 1, size = 0.7, linetype = "dashed") +
      # 
  geom_point(color = "#345F6F", size = 0.2, alpha = 0.8) + 
  geom_point(aes(x = POS, y = Parent_Zprime), color = "#FFB05C", alpha= 0.8, size = 0.2) +
  geom_point(aes(x = POS, y = Interaction_Zprime), color = "#D7335C", alpha= 0.8, size = 0.2) +
  geom_point(aes(x = POS, y = Day_Zprime), color = "gray", alpha= 0.8, size = 0.2) +
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  geom_hline(yintercept = c(max(abs(NullComparison$Dataset_Zprime)), -max(abs(NullComparison$Dataset_Zprime))), 
             color = "black", linetype = "dashed") +
  ggtitle("CSSI Fluconazole | Comparison of Smoothed Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")


```

## Actually pull out the peaks

```{r, eval = FALSE}
#Want the max value above the line for every slope, so maybe per 0 observed in the smoothed line (since there aren't any that really go the same way besides XIV and that's a bit iffy as is.)

load("RdataFiles/CSSI_F_ZScores_GLM_allexcl.Rdata")

chrompeaks <- data.frame(CHROM = unique(CSSI_F_ZScores_GLM$CHROM),
                         peak = NA,
                         trough = NA)

for(i in unique(CSSI_F_ZScores_GLM$CHROM)){
  #i <- "I"
  x <- subset(CSSI_F_ZScores_GLM, CHROM == i)
  peak = max(x$Bulk_Zprime[x$Bulk_Zprime > max(abs(NullComparison$Dataset_Zprime))]) 
  trough = min(x$Bulk_Zprime[x$Bulk_Zprime < -max(abs(NullComparison$Dataset_Zprime))])
  if(peak > 1){
      chrompeaks$peak[chrompeaks$CHROM == i] <- x$POS[x$Bulk_Zprime == peak]
  }
  if(trough < 1){
      chrompeaks$trough[chrompeaks$CHROM == i] <- x$POS[x$Bulk_Zprime == trough]
  }

}

head(chrompeaks)

chrompeaks %>% pivot_longer(c(peak, trough), values_to = "Bulk_POS") %>% select(CHROM, Bulk_POS) %>% na.omit() -> BulkPeaks

###############################################################################

chrompeaks_int <- data.frame(CHROM = unique(CSSI_F_ZScores_GLM$CHROM),
                         peak = NA,
                         trough = NA)

for(i in unique(CSSI_F_ZScores_GLM$CHROM)){
  #i <- "I"
  x <- subset(CSSI_F_ZScores_GLM, CHROM == i)
  peak = max(x$Interaction_Zprime[x$Interaction_Zprime > max(abs(NullComparison$Dataset_Zprime))]) 
  trough = min(x$Interaction_Zprime[x$Interaction_Zprime < -max(abs(NullComparison$Dataset_Zprime))])
  if(peak > 1){
      chrompeaks_int$peak[chrompeaks_int$CHROM == i] <- x$POS[x$Interaction_Zprime == peak]
  }
  if(trough < 1){
      chrompeaks_int$trough[chrompeaks_int$CHROM == i] <- x$POS[x$Interaction_Zprime == trough]
  }

}

head(chrompeaks)

chrompeaks_int %>% pivot_longer(c(peak, trough), values_to = "Int_POS") %>% select(CHROM, Int_POS) %>% na.omit() -> IntPeaks

left_join(BulkPeaks,IntPeaks) %>% pivot_longer(c(Bulk_POS, Int_POS)) %>% na.omit() -> Peaks
```

