---
title: "CuSO4 Data for Paper - JK this is all permutations"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

library(stringr)
library(lme4)

ggplot2::theme_set(theme_light())

## Function for logistic regression

glm_cb <- function(..., W, formula) {
  data <- list(...)
  
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  
  glm_formula <- as.formula(formula)
  
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #CHANGE THIS TO ADJUST TYPE OF REGRESSION
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  #Output: Effect, Standard Error, Z-score 
  #(ignores p-value since it can be calculated from Z score; multiply by 1 if you would like to retain it)
  return(coefficients(glm_fit))
  
}
```

Test to make sure the glm gives what you want 

```{r, eval = FALSE}
# mydata <- Cu_1_B_Standardized %>% filter(CHROM == "I", POS == 904) %>% mutate_if(is.character, as.factor)
# glm_fit <- glm("Allele ~ Bulk*Parent+Rep", data = as.data.frame(mydata), weights = Reads, family = binomial)
# 
# names(coefficients(glm_fit))

```

## Load in data from each experiment

```{r}
setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023")

#dir("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023/Data/")

Cu_1_A <- readRDS("Data/CuSO4_1_cybr2.rds")
Cu_1_B <- readRDS("Data/CuSO4_2_cybr2.rds")
Cu_8 <- readRDS("Data/CuSO4_CSS8_cybr2.rds")

```

## Standardize read counts for each

```{r}
Cu_1_A %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Allele"), names_sep = "_", values_to = "Reads") %>%
  pivot_wider(names_from = c(Bulk, Parent), values_from = Reads) %>% select(c(-CHROM, -POS, -Allele)) %>% colSums() %>% as.data.frame() -> overallcoverage

colnames(overallcoverage) <- "Reads"
overallcoverage$Bulk <- rownames(overallcoverage)
overallcoverage$Standard <- max(overallcoverage$Reads)/overallcoverage$Reads
overallcoverage %>% select(-Reads) %>% pivot_wider(names_from = Bulk, names_sep = "_", values_from = Standard) %>%
  pivot_longer(c(CuSO4_Oak,CuSO4_Wine,Dilute_Oak, Dilute_Wine), names_to = c("Bulk", "Parent"), names_sep = "_", values_to = "Standard") -> standardization

Cu_1_A %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Allele"), names_sep = "_", values_to = "Reads") %>%
  merge(standardization) %>%
  mutate(ReadStandard = floor(Reads * Standard)) %>%
  select(-Reads, -Standard) %>% 
  pivot_wider(names_from = c(Bulk, Parent), values_from = ReadStandard) %>%
  pivot_longer(c(-CHROM, -POS, -Allele), names_to = c("Bulk", "Parent"), names_sep = "_", values_to = "Reads") -> Cu_1_A_Standardized

Cu_1_A_Standardized %>% mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(glmresult = glm_cb(formula = "Allele ~ Bulk*Parent",
                                          W = Reads,
                                          Allele = Allele, Bulk = Bulk, Parent = Parent)) -> Cu_1_A_glm

Cu_1_A_glm$label <- rep(c("Intercept", "Bulk", "Parent", "Interaction"),length(Cu_1_A_glm$CHROM)/4)

#Cu_1_A_glm %>%  ggplot(aes(x = POS, y = glmresult, color = label)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")

```
## Mixing Parents

```{r}
Cu_1_A_Standardized  %>%
  #permute parent
  pivot_wider(names_from = c(Bulk, Allele), values_from = Reads) %>% 
  group_by(CHROM, POS) %>%
  summarize(Parent_P = sample(c("Oak", "Wine"), replace = F, size = 2),
            CuSO4_Oak = CuSO4_Oak,
            CuSO4_Wine = CuSO4_Wine,
            Dilute_Oak = Dilute_Oak,
            Dilute_Wine = Dilute_Wine) %>%
  pivot_longer(c(-CHROM, -POS, -Parent_P), names_to = c("Bulk", "Allele"), names_sep = "_", values_to = "Reads") %>%
  #then run GLM
  mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(glmresult = glm_cb(formula = "Allele ~ Bulk*Parent",
                                          W = Reads,
                                          Allele = Allele, Bulk = Bulk, Parent = Parent_P)) -> Cu_1_A_glmP

Cu_1_A_glmP$label <- rep(c("Intercept", "Bulk", "Parent", "Interaction"),length(Cu_1_A_glmP$CHROM)/4)

#Cu_1_A_glmP %>%  ggplot(aes(x = POS, y = glmresult, color = label)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")
```
## Smooth both of these

```{r}
Cu_1_A_glmP %>%
  group_by(CHROM,label) %>% arrange(POS) %>% 
  reframe(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = (frollapply(glmresult, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() -> Cu_1_A_glmP_roll

Cu_1_A_glm %>%
  group_by(CHROM, label) %>% arrange(POS) %>% 
  reframe(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = (frollapply(glmresult, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  na.omit() -> Cu_1_A_glm_roll
```

Plot these  

```{r}
Cu_1_A_glm_roll %>% filter(CHROM != "I", CHROM != "M") %>%
  ggplot(aes(x = POS, y = SmoothCount, color = label)) + 
  geom_line() + facet_grid(~CHROM, scales = "free", space = "free") +
  ggtitle("Original Scaled Data")

Cu_1_A_glmP_roll %>% filter(CHROM != "I", CHROM != "M") %>%
  ggplot(aes(x = POS, y = SmoothCount, color = label)) + 
  geom_line() + facet_grid(~CHROM, scales = "free", space = "free") +
  ggtitle("Permuted Parent Scaled Data")

```

# Replicate Studies

## CuSO4 2

Just the normal one

```{r}
Cu_1_B %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_", values_to = "Reads") %>%
  pivot_wider(names_from = c(Bulk, Parent, Rep), values_from = Reads) %>% select(c(-CHROM, -POS, -Allele)) %>% colSums() %>% as.data.frame() -> overallcoverageB

colnames(overallcoverageB) <- "Reads"
overallcoverageB$Bulk <- rownames(overallcoverageB)
overallcoverageB$Standard <- max(overallcoverageB$Reads)/overallcoverageB$Reads
overallcoverageB %>% select(-Reads) %>% pivot_wider(names_from = Bulk, names_sep = "_", values_from = Standard) %>%
  pivot_longer(c(CuSO4_Oak_A,CuSO4_Wine_A,Dilute_Oak_A, Dilute_Wine_A,
                 CuSO4_Oak_B,CuSO4_Wine_B,Dilute_Oak_B, Dilute_Wine_B), names_to = c("Bulk", "Parent", "Rep"), names_sep = "_", values_to = "Standard") -> standardization

Cu_1_B %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_", values_to = "Reads") %>%
  merge(standardization) %>%
  mutate(ReadStandard = floor(Reads * Standard)) %>%
  select(-Reads, -Standard) %>% 
  pivot_wider(names_from = c(Bulk, Parent, Rep), values_from = ReadStandard) %>%
  pivot_longer(c(-CHROM, -POS, -Allele), names_to = c("Bulk", "Parent", "Rep"), names_sep = "_", values_to = "Reads") -> Cu_1_B_Standardized

Cu_1_B_Standardized %>% mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(glmresult = glm_cb(formula = "Allele ~ Bulk*Parent + Rep",
                                          W = Reads,
                                          Allele = Allele, Bulk = Bulk, Parent = Parent, Rep = Rep)) -> Cu_1_B_glm

Cu_1_B_glm$label <- rep(c("Intercept", "Bulk", "Parent", "Rep", "Interaction"),length(Cu_1_B_glm$CHROM)/5)

```

The permuted one

```{r}

Cu_1_B_Standardized %>% #permute parent
  pivot_wider(names_from = c(Bulk, Rep, Allele), values_from = Reads) %>% 
  group_by(CHROM, POS) %>%
  summarize(Parent_P = sample(c("Oak", "Wine"), replace = F, size = 2),
            CuSO4_A_Oak = CuSO4_A_Oak,
            CuSO4_A_Wine = CuSO4_A_Wine,
            Dilute_A_Oak = Dilute_A_Oak,
            Dilute_A_Wine = Dilute_A_Wine,
            CuSO4_B_Oak = CuSO4_B_Oak,
            CuSO4_B_Wine = CuSO4_B_Wine,
            Dilute_B_Oak = Dilute_B_Oak,
            Dilute_B_Wine = Dilute_B_Wine) %>%
  pivot_longer(c(-CHROM, -POS, -Parent_P), names_to = c("Bulk", "Rep", "Allele"), names_sep = "_", values_to = "Reads") %>%
  #then run GLM
  mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(glmresult = glm_cb(formula = "Allele ~ Bulk*Parent + Rep",
                                          W = Reads,
                                          Allele = Allele, Bulk = Bulk, Parent = Parent_P, Rep = Rep)) -> Cu_1_B_glmP

Cu_1_B_glmP$label <- rep(c("Intercept", "Bulk", "Parent", "Rep", "Interaction"),length(Cu_1_B_glmP$CHROM)/5)
```

Roll these  

```{r}
Cu_1_B_glmP %>%
  group_by(CHROM,label) %>% arrange(POS) %>% 
  reframe(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = (frollapply(glmresult, n = 200, FUN = median))) %>% 
  na.omit() -> Cu_1_B_glmP_roll

Cu_1_B_glm %>%
  group_by(CHROM, label) %>% arrange(POS) %>% 
  reframe(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = (frollapply(glmresult, n = 200, FUN = median))) %>% 
  na.omit() -> Cu_1_B_glm_roll
```

Plot these  

```{r}
Cu_1_B_glm_roll %>% filter(CHROM != "I", CHROM != "M") %>%
  ggplot(aes(x = POS, y = SmoothCount, color = label)) + 
  geom_line() + facet_grid(~CHROM, scales = "free", space = "free") +
  ggtitle("Original Scaled Data")

Cu_1_B_glmP_roll %>% filter(CHROM != "I", CHROM != "M") %>%
  ggplot(aes(x = POS, y = SmoothCount, color = label)) + 
  geom_line() + facet_grid(~CHROM, scales = "free", space = "free") +
  ggtitle("Permuted Parent Scaled Data")

```

## Last one - Chr 8 instead of 1 for CuSO4

Just the normal one

```{r}
Cu_8 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Dose","Allele"), names_sep = "_", values_to = "Reads") %>%
  pivot_wider(names_from = c(Bulk, Parent, Rep, Dose), values_from = Reads) %>% select(c(-CHROM, -POS, -Allele)) %>% colSums() %>% as.data.frame() -> overallcoverageC

colnames(overallcoverageC) <- "Reads"
overallcoverageC$Bulk <- rownames(overallcoverageC)
overallcoverageC$Standard <- max(overallcoverageC$Reads)/overallcoverageC$Reads
overallcoverageC %>% select(-Reads) %>% pivot_wider(names_from = Bulk, names_sep = "_", values_from = Standard) %>%
  pivot_longer(cols = everything(), 
               names_to = c("Bulk", "Parent", "Rep", "Dose"), names_sep = "_", values_to = "Standard") -> standardization

Cu_8 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Dose", "Allele"), names_sep = "_", values_to = "Reads") %>%
  merge(standardization) %>%
  mutate(ReadStandard = floor(Reads * Standard)) %>%
  select(-Reads, -Standard) %>% 
  pivot_wider(names_from = c(Bulk, Parent, Rep, Dose), values_from = ReadStandard) %>%
  pivot_longer(c(-CHROM, -POS, -Allele), names_to = c("Bulk", "Parent", "Rep", "Dose"), names_sep = "_", values_to = "Reads") -> Cu_8_Standardized

#Actually take only the doses we want to use
Cu_8_Standardized %>% filter(Parent != "Oak1", Parent != "Wine1", Bulk != "Fluc", (Parent == "Wine8" & Dose == "C5") == FALSE) %>%
  select(-Dose) %>%

#Back to normal code

#Cu_8_Standardized %>% 
  mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(glmresult = glm_cb(formula = "Allele ~ Bulk*Parent + Rep",
                                          W = Reads,
                                          Allele = Allele, Bulk = Bulk, Parent = Parent, Rep = Rep)) -> Cu_8_glm

Cu_8_glm$label <- rep(c("Intercept", "Bulk", "Parent", "Rep", "Interaction"),length(Cu_8_glm$CHROM)/5)

```

The permuted one

```{r}

Cu_8_Standardized %>% #permute parent
  filter(Parent != "Oak1", Parent != "Wine1", Bulk != "Fluc", (Parent == "Wine8" & Dose == "C5") == FALSE) %>%
  select(-Dose) %>%
  #Back to normal code
  pivot_wider(names_from = c(Bulk, Rep, Allele), values_from = Reads) %>% 
  group_by(CHROM, POS) %>%
  summarize(Parent_P = sample(c("Oak", "Wine"), replace = F, size = 2),
            CuSO4_A_Oak = CuSO4_A_Oak,
            CuSO4_A_Wine = CuSO4_A_Wine,
            Dilute_A_Oak = Dilute_A_Oak,
            Dilute_A_Wine = Dilute_A_Wine,
            CuSO4_B_Oak = CuSO4_B_Oak,
            CuSO4_B_Wine = CuSO4_B_Wine,
            Dilute_B_Oak = Dilute_B_Oak,
            Dilute_B_Wine = Dilute_B_Wine) %>%
  pivot_longer(c(-CHROM, -POS, -Parent_P), names_to = c("Bulk", "Rep", "Allele"), names_sep = "_", values_to = "Reads") %>%
  #then run GLM
  mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(glmresult = glm_cb(formula = "Allele ~ Bulk*Parent + Rep",
                                          W = Reads,
                                          Allele = Allele, Bulk = Bulk, Parent = Parent_P, Rep = Rep)) -> Cu_8_glmP

Cu_8_glmP$label <- rep(c("Intercept", "Bulk", "Parent", "Rep", "Interaction"),length(Cu_8_glmP$CHROM)/5)
```

Roll these  

```{r}
Cu_8_glmP %>%
  group_by(CHROM,label) %>% arrange(POS) %>% 
  reframe(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = (frollapply(glmresult, n = 200, FUN = median))) %>% 
  na.omit() -> Cu_8_glmP_roll

Cu_8_glm %>%
  group_by(CHROM, label) %>% arrange(POS) %>% 
  reframe(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = (frollapply(glmresult, n = 200, FUN = median))) %>% 
  na.omit() -> Cu_8_glm_roll
```

Plot these  

```{r}
Cu_8_glm_roll %>% filter(CHROM != "VIII", CHROM != "M") %>%
  ggplot(aes(x = POS, y = SmoothCount, color = label)) + 
  geom_line() + facet_grid(~CHROM, scales = "free", space = "free") +
  ggtitle("Original Scaled Data")

Cu_8_glmP_roll %>% filter(CHROM != "VIII", CHROM != "M") %>%
  ggplot(aes(x = POS, y = SmoothCount, color = label)) + 
  geom_line() + facet_grid(~CHROM, scales = "free", space = "free") +
  ggtitle("Permuted Parent Scaled Data")

```

# Comparing the Permuted vs Non-Permuted Interaction Effects

Now that I'm thinking about it, I should probably roll the initial permuted data first rather than the actual scores, but here we are. Let's plot each of these in comparison.

```{r}
rbind(data.frame(Cu_8_glmP_roll, Exp = "8", P = "Permuted"),
  data.frame(Cu_8_glm_roll, Exp = "8", P = "Data"),
  data.frame(Cu_1_A_glmP_roll, Exp = "1", P = "Permuted"),
  data.frame(Cu_1_A_glm_roll, Exp = "1", P = "Data"),
  data.frame(Cu_1_B_glmP_roll, Exp = "1b", P = "Permuted"),
  data.frame(Cu_1_B_glm_roll, Exp = "1b", P = "Data")) %>%
  
  filter(label == "Interaction", CHROM != "I", CHROM != "VIII", CHROM != "M") %>%
  ggplot(aes(x = POS, y = SmoothCount, color = paste(Exp,P))) + geom_line() + facet_grid(Exp~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("lightblue", "black", "maroon", "black", "darkorange", "black"))

```

Look at the distribution of these

```{r}
rbind(data.frame(Cu_8_glmP_roll, Exp = "8", P = "Permuted"),
  data.frame(Cu_8_glm_roll, Exp = "8", P = "Data"),
  data.frame(Cu_1_A_glmP_roll, Exp = "1", P = "Permuted"),
  data.frame(Cu_1_A_glm_roll, Exp = "1", P = "Data"),
  data.frame(Cu_1_B_glmP_roll, Exp = "1b", P = "Permuted"),
  data.frame(Cu_1_B_glm_roll, Exp = "1b", P = "Data")) -> Permuted_data

Permuted_data  %>% filter(CHROM != "I", CHROM != "M", CHROM !="VIII") %>%
  group_by(Exp, label) %>%
  reframe(Cutoff_Abs = quantile(abs(SmoothCount), seq(0, 1, length.out = 1001), na.rm = TRUE),
            Quant = seq(0, 1, length.out = 1001)) %>%
  na.omit() -> Permuted_FDR

  #Trying new code for slicing
Permuted_FDR %>%
  group_by(Exp, label) %>% filter(label != "Intercept") %>% filter(Quant >= 0.95) %>% slice(which.min(Cutoff_Abs)) -> pCutoff_95
Permuted_FDR %>%
  group_by(Exp, label) %>% filter(label != "Intercept") %>% filter(Quant >= 0.99) %>% slice(which.min(Cutoff_Abs)) -> pCutoff_99

Permuted_FDR %>% filter(label != "Intercept") %>%
  ggplot(aes(x = Cutoff_Abs , y = Quant, color = label, linetype = Exp)) + geom_line(size= 1.2, alpha = 0.2) +
  geom_hline(yintercept = c(0.95, 0.99)) +
  geom_vline(data = pCutoff_95, aes(xintercept = Cutoff_Abs, color = label, linetype = Exp), size = 1.2)+
  ggtitle("5% False Discovery on Permuted Parent") +
  ylim(0.8, 1) +
  scale_color_manual(values = c("skyblue", "red", "lightblue", "gray"))

Permuted_FDR %>% filter(label != "Intercept") %>%
  ggplot(aes(x = Cutoff_Abs , y = Quant, color = label, linetype = Exp)) + geom_line(size= 1.2, alpha = 0.2) +
  geom_hline(yintercept = c(0.95, 0.99)) +
  geom_vline(data = pCutoff_99, aes(xintercept = Cutoff_Abs, color = label, linetype = Exp), size = 1.2)+
  ggtitle("1% False Discovery on Permuted Parent") +
  ylim(0.8, 1) +
  scale_color_manual(values = c("skyblue", "red", "lightblue", "gray"))
```

Plot these as a regression

```{r}
rbind(data.frame(Cu_8_glmP_roll, Exp = "8", P = "Permuted"),
  data.frame(Cu_8_glm_roll, Exp = "8", P = "Data"),
  data.frame(Cu_1_A_glmP_roll, Exp = "1", P = "Permuted"),
  data.frame(Cu_1_A_glm_roll, Exp = "1", P = "Data"),
  data.frame(Cu_1_B_glmP_roll, Exp = "1b", P = "Permuted"),
  data.frame(Cu_1_B_glm_roll, Exp = "1b", P = "Data")) %>%
  
  filter(label == "Interaction", CHROM != "I", CHROM != "VIII", CHROM != "M") -> permutedvsnon

permutedvsnon %>% pivot_wider(names_from = P, values_from = SmoothCount) %>%
  ggplot(aes(x = abs(Data), y = abs(Permuted), color = Exp)) + geom_point(alpha = 0.1) +
  geom_smooth(method = "lm") +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed") +
  scale_color_manual(values = c("lightblue", "maroon", "darkorange", "black"))

```

# NEXT: smooth standardized data first, and then permute multiple times to get a distribution for each position

The plots here look like I've permuted something else, so check the labeling on this tomorrow...

```{r}
Cu_8_Standardized %>% #filter(CHROM == "I") %>%
  group_by(CHROM, Bulk, Parent, Dose, Rep, Allele) %>%
  #Smooth Data
  arrange(POS) %>% 
  reframe(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(Reads, n = 200, FUN = median))) %>% 
  na.omit() %>%
  #Filter Data
  filter(Parent != "Oak1", Parent != "Wine1", Bulk != "Fluc", (Parent == "Wine8" & Dose == "C5") == FALSE) %>%
  select(-Dose) -> C8_RollData

C8_RollData %>%
  #Back to normal code
  pivot_wider(names_from = c(Bulk, Rep, Allele), values_from = SmoothCount) %>% 
  group_by(CHROM, POS) %>%
  summarize(Parent_P = sample(c("Oak", "Wine"), replace = F, size = 2),
            CuSO4_A_Oak = CuSO4_A_Oak,
            CuSO4_A_Wine = CuSO4_A_Wine,
            Dilute_A_Oak = Dilute_A_Oak,
            Dilute_A_Wine = Dilute_A_Wine,
            CuSO4_B_Oak = CuSO4_B_Oak,
            CuSO4_B_Wine = CuSO4_B_Wine,
            Dilute_B_Oak = Dilute_B_Oak,
            Dilute_B_Wine = Dilute_B_Wine) %>%
  pivot_longer(c(-CHROM, -POS, -Parent_P), names_to = c("Bulk", "Rep", "Allele"), names_sep = "_", values_to = "Reads") %>%
  #then run GLM
  mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(glmresult = glm_cb(formula = "Allele ~ Bulk*Parent + Rep",
                                          W = Reads,
                                          Allele = Allele, Bulk = Bulk, Parent = Parent_P, Rep = Rep),
                                   label = c("Intercept", "Bulk", "Parent", "Rep", "Interaction")) -> C8_roll_glmP

C8_RollData %>% 
  mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(glmresult = glm_cb(formula = "Allele ~ Bulk*Parent + Rep",
                                          W = SmoothCount,
                                          Allele = Allele, Bulk = Bulk, Parent = Parent, Rep = Rep),
                                   label = c("Intercept", "Bulk", "Parent", "Rep", "Interaction")) -> C8_roll_glm

C8_roll_glm %>% filter(CHROM != "VIII", CHROM != "M") %>%
  ggplot(aes(x = POS, y = abs(glmresult), color = label)) + 
  geom_line() + facet_grid(label~CHROM, scales = "free", space = "free") +
  ggtitle("Original Data")

C8_roll_glmP %>% filter(CHROM != "VIII", CHROM != "M") %>%
  ggplot(aes(x = POS, y = abs(glmresult), color = label)) + 
  geom_line() + facet_grid(label~CHROM, scales = "free", space = "free") +
  ggtitle("Permuted Parent Data")

```

