---
title: "SubSample CSS BSA"
author: "Cassandra Buzby"
date: "1/11/2023"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

library(cybrBSA)

ggplot2::theme_set(theme_light())

glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

```



Testing on real data

```{r}
setwd("../../../GitHub/Sequencing/Analysis/")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

#head(parentSNPids)

parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) %>% filter(Unique == 1) -> pSNPs

mydatatotest = "HGV.SortedCat.vcf.output.table"

CSSI_Fluc <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = pSNPs, yeast = TRUE) %>% 
  na.omit()

CSSI_Fluc$Bulk <- NA
CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq"] <- "Dilute"

CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "Fluconazole"

CSSI_Fluc$Parent <- NA
CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq"] <- "OakI"

CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "WineI"

CSSI_Fluc$Rep <- NA
CSSI_Fluc$Rep[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq"] <- "B"

CSSI_Fluc$Rep[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" | 
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "C"

CSSI_Fluc$Rep[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq" |
              CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq" ] <- "D"

CSSI_Fluc$ReadCount <- as.numeric(CSSI_Fluc$ReadCount)
  
# #THIS IGNORES REPLICATES
CSSI_Fluc %>% select(CHROM, POS, PAllele, ReadCount, Bulk, Parent, Rep) %>% distinct %>% 
 pivot_wider(names_from = c(Bulk, Parent, Rep, PAllele), values_from = ReadCount) -> CSSI_Fluc_cybr2

saveRDS(object = CSSI_Fluc_cybr2, file = "CSSI_Fluc_cybr2.rds")

```

```{r, eval = FALSE}
CSSI_Fluc_cybr2 <- readRDS("../../cybrBSA/cybrBSA/CSSI_Fluc_cybr2.rds")

CSSI_Fluc_cybr2 %>% mutate(H_O_O = Fluconazole_OakI_B_Oak, #Fluconazole_OakI_C_Oak, Fluconazole_OakI_D_Oak,
                           H_O_W = Fluconazole_OakI_B_Wine, #Fluconazole_OakI_C_Wine, Fluconazole_OakI_D_Wine,
                           H_W_O = Fluconazole_WineI_C_Oak,
                           H_W_W = Fluconazole_WineI_C_Wine,
                           L_O_O = Dilute_OakI_B_Oak,
                           L_O_W = Dilute_OakI_B_Wine,
                           L_W_O = Dilute_WineI_B_Oak, #Dilute_WineI_C_Oak,Dilute_WineI_D_Oak,
                           L_W_W = Dilute_WineI_B_Wine) %>% #, Dilute_WineI_C_Wine, Dilute_WineI_D_Wine)
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>%
group_by(CHROM, POS) %>% summarise(summary = glmfixed(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> testCSSIFluc_cybr2

saveRDS(object = testCSSIFluc_cybr2, file = "testCSSIFluc_cybr2.rds")

```

```{r}
testCSSIFluc_cybr2 <- readRDS("../../cybrBSA/cybrBSA/testCSSIFluc_cybr2.rds")

ggplot(testCSSIFluc_cybr2, aes(x = POS, y = summary, color = label)) + geom_line() + facet_grid(~CHROM, scales = "free")
```



```{r}
CSSIFluc_cybr2 <- readRDS("../../cybrBSA/cybrBSA/testCSSIFluc_cybr2.rds")

CSSIFluc_cybr2 %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothZ = frollmean(summary, n = 100)) -> CSSIFluc_Window100

CSSIFluc_Window100 %>% ggplot(aes(x = POS, y = SmoothZ, color = label)) + geom_line() + facet_grid(~CHROM, scales = "free")
```

## Running this with all samples

```{r, eval = FALSE}
setwd("../../../GitHub/Sequencing/Analysis/")

cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE) %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) %>% filter(Unique == 1) -> pSNPs

################################################################################
# mydatatotest = "HGV.SortedCat.vcf.output.table"
# CSSI_Fluc <- cybrInputGATKTable(mydatatotest) %>% 
#   cybrQualityFilter() %>% 
#   cybrIDAlleles(BSAdfstart = ., Parentdf = pSNPs, yeast = TRUE) %>% 
#   na.omit()
# 
# CSSI_Fluc$Bulk <- NA
# CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" | 
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq"] <- "Dilute"
# 
# CSSI_Fluc$Bulk[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq" | 
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "Fluconazole"
# 
# CSSI_Fluc$Parent <- NA
# CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" | 
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq"] <- "OakI"
# 
# CSSI_Fluc$Parent[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq" | 
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "WineI"
# 
# CSSI_Fluc$Rep <- NA
# CSSI_Fluc$Rep[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq"] <- "B"
# 
# CSSI_Fluc$Rep[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" | 
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "C"
# 
# CSSI_Fluc$Rep[CSSI_Fluc$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq" |
#               CSSI_Fluc$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq" ] <- "D"
# 
# CSSI_Fluc$ReadCount <- as.numeric(CSSI_Fluc$ReadCount)
#   
# CSSI_Fluc %>% select(CHROM, POS, PAllele, ReadCount, Bulk, Parent, Rep) %>% distinct %>% 
#  pivot_wider(names_from = c(Bulk, Parent, Rep, PAllele), values_from = ReadCount) -> CSSI_Fluc_cybr2
# 
# saveRDS(object = CSSI_Fluc_cybr2, file = "CSSI_Fluc_cybr2.rds")

################################################################################

setwd("../../../GitHub/Sequencing/Analysis/")

mydatatotest = "Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"
CSSI_CuSO4 <- cybrInputGATKTable(mydatatotest) %>%
  cybrQualityFilter() %>%
  cybrIDAlleles(BSAdfstart = ., Parentdf = pSNPs, yeast = TRUE) %>%
  na.omit()

CSSI_CuSO4$Bulk <- NA
CSSI_CuSO4$Bulk[CSSI_CuSO4$Dataset == "SelectedA" |
              CSSI_CuSO4$Dataset == "SelectedC"] <- "CuSO4"

CSSI_CuSO4$Bulk[CSSI_CuSO4$Dataset == "UnselectedA" |
              CSSI_CuSO4$Dataset == "UnselectedC"] <- "Dilute"

CSSI_CuSO4$Parent <- NA
CSSI_CuSO4$Parent[CSSI_CuSO4$Dataset == "SelectedA" |
              CSSI_CuSO4$Dataset == "UnselectedA"] <- "OakI"

CSSI_CuSO4$Parent[CSSI_CuSO4$Dataset == "SelectedC" |
              CSSI_CuSO4$Dataset == "UnselectedC"] <- "WineI"

CSSI_CuSO4$ReadCount <- as.numeric(CSSI_CuSO4$ReadCount)

CSSI_CuSO4 %>% select(CHROM, POS, PAllele, ReadCount, Bulk, Parent) %>% distinct %>%
 pivot_wider(names_from = c(Bulk, Parent, PAllele), values_from = ReadCount) -> CSSI_CuSO4_cybr2

saveRDS(object = CSSI_CuSO4_cybr2, file = "CSSI_CuSO4_cybr2.rds")

################################################################################

setwd("../../../GitHub/Sequencing/Analysis/")

mydatatotest = "HKTMZDRX2.SortedCat.vcf.output.table"

CSSI_CycZeo <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = pSNPs, yeast = TRUE) %>% 
  na.omit()

CSSI_CycZeo$Bulk <- NA
CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq"] <- "Cycloheximide"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "Dilute"

CSSI_CycZeo$Bulk[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq"] <- "Zeocin"

CSSI_CycZeo$Parent <- NA
CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_OCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_ODB.fastq"] <- "OakI"

CSSI_CycZeo$Parent[CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZA.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WZB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WCB.fastq" |
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDA.fastq" | 
              CSSI_CycZeo$Dataset == "HKTMWDRX2_n01_WDB.fastq"] <- "WineI"

CSSI_CycZeo$Rep <- NA
CSSI_CycZeo$Rep[grep("A.fastq", CSSI_CycZeo$Dataset)] <- "A"
CSSI_CycZeo$Rep[grep("B.fastq", CSSI_CycZeo$Dataset)] <- "B"

CSSI_CycZeo$ReadCount <- as.numeric(CSSI_CycZeo$ReadCount)

CSSI_Cyc <- CSSI_CycZeo[CSSI_CycZeo$Bulk != "Zeocin",]
CSSI_Zeocin <- CSSI_CycZeo[CSSI_CycZeo$Bulk != "Cycloheximide",]


CSSI_Zeocin %>% select(CHROM, POS, PAllele, ReadCount, Bulk, Parent, Rep) %>% distinct %>% 
 pivot_wider(names_from = c(Bulk, Parent, Rep, PAllele), values_from = ReadCount) -> CSSI_Zeocin_cybr2

CSSI_Cyc %>% select(CHROM, POS, PAllele, ReadCount, Bulk, Parent, Rep) %>% distinct %>% 
 pivot_wider(names_from = c(Bulk, Parent, Rep, PAllele), values_from = ReadCount) -> CSSI_Cyc_cybr2

saveRDS(object = CSSI_Zeocin_cybr2, file = "CSSI_Zeocin_cybr2.rds")
saveRDS(object = CSSI_Cyc_cybr2, file = "CSSI_Cyc_cybr2.rds")

```

```{r}
CSSI_Zeocin_cybr2 <- readRDS("../../cybrBSA/cybrBSA/CSSI_Zeocin_cybr2.rds")
CSSI_Cyc_cybr2 <- readRDS("../../cybrBSA/cybrBSA/CSSI_Cyc_cybr2.rds")
CSSI_CuSO4_cybr2 <- readRDS("../../cybrBSA/cybrBSA/CSSI_CuSO4_cybr2.rds")


```

```{r, eval = FALSE}
#Zeocin
CSSI_Zeocin_cybr2  %>% mutate(H_O_O = Zeocin_OakI_A_Oak, #Zeocin_OakI_B_Oak, 
                           H_O_W = Zeocin_OakI_A_Wine, #Zeocin_OakI_B_Wine,
                           H_W_O = Zeocin_WineI_A_Oak, #Zeocin_WineI_B_Oak, 
                           H_W_W = Zeocin_WineI_A_Wine, #Zeocin_WineI_B_Wine, 
                           L_O_O = Dilute_OakI_A_Oak, #Dilute_OakI_B_Oak,
                           L_O_W = Dilute_OakI_A_Wine,#Dilute_OakI_B_Wine,
                           L_W_O = Dilute_WineI_A_Oak, #Dilute_WineI_B_Oak,
                           L_W_W = Dilute_WineI_A_Wine) %>% #, Dilute_WineI_B_Wine)
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>%
group_by(CHROM, POS) %>% summarise(summary = glmfixed(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Zeocin_cybr2_glm
saveRDS(object = CSSI_Zeocin_cybr2_glm, file = "CSSI_Zeocin_cybr2_glm.rds")

CSSI_Zeocin_cybr2_glm %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothZ = frollmean(summary, n = 100)) -> CSSI_Zeocin_Window100

saveRDS(object = CSSI_Zeocin_Window100, file = "CSSI_Zeocin_Window100.rds")

```

```{r}
CSSI_Zeocin_Window100 <- readRDS("CSSI_Zeocin_Window100.rds")
```


```{r, eval = FALSE}

#CSSI_CuSO4_cybr2 %>% filter(CHROM == "IV", POS == 327540)

#CuSO4
CSSI_CuSO4_cybr2  %>% mutate(H_O_O = CuSO4_OakI_Oak, 
                           H_O_W = CuSO4_OakI_Wine, 
                           H_W_O = CuSO4_WineI_Oak,  
                           H_W_W = CuSO4_WineI_Wine,  
                           L_O_O = Dilute_OakI_Oak, 
                           L_O_W = Dilute_OakI_Wine,
                           L_W_O = Dilute_WineI_Oak, 
                           L_W_W = Dilute_WineI_Wine) %>% 
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>%
group_by(CHROM, POS) %>% #filter(CHROM == "IV", POS == 327540) %>% 
  summarise(summary = glmfixed(H_O_O,H_O_W,H_W_O,H_W_W,
                                                      L_O_O,L_O_W,L_W_O,L_W_W)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_CuSO4_cybr2_glm

saveRDS(object = CSSI_CuSO4_cybr2_glm, file = "CSSI_CuSO4_cybr2_glm.rds")

CSSI_CuSO4_cybr2_glm %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothZ = frollmean(summary, n = 100)) -> CSSI_CuSO4_Window100

saveRDS(object = CSSI_CuSO4_Window100, file = "CSSI_CuSO4_Window100.rds")
```

```{r, eval = FALSE}
CSSIFluc_cybr2 <- readRDS("../../cybrBSA/cybrBSA/testCSSIFluc_cybr2.rds")

CSSIFluc_cybr2 %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothZ = frollmean(summary, n = 100)) -> CSSIFluc_Window100

saveRDS(CSSIFluc_Window100, file = "CSSIFluc_Window100.rds")
```

```{r}
CSSI_Zeocin_Window100 <- readRDS("../../cybrBSA/cybrBSA/CSSI_Zeocin_Window100.rds")
CSSI_CuSO4_Window100 <- readRDS("../../cybrBSA/cybrBSA/CSSI_CuSO4_Window100.rds")
CSSIFluc_Window100 <- readRDS("../../cybrBSA/cybrBSA/CSSIFluc_Window100.rds")

```


```{r}
CSSI_CuSO4_Window100 %>% ggplot(aes(x = POS, y = abs(SmoothZ), color = label)) + geom_line() + facet_grid(~CHROM, scales = "free") + ggtitle("CuSO4") + ylim(c(0,12))
CSSI_Zeocin_Window100 %>% ggplot(aes(x = POS, y = abs(SmoothZ), color = label)) + geom_line() + facet_grid(~CHROM, scales = "free") + ggtitle("Zeocin_A") + ylim(c(0,12))
CSSIFluc_Window100 %>% ggplot(aes(x = POS, y = abs(SmoothZ), color = label)) + geom_line() + facet_grid(~CHROM, scales = "free") + ggtitle("Fluconazole (test)") + ylim(c(0,12))

```

## Scaling to Chr III Intercept peak

First, do the replicates agree on that one?

```{r}
#Zeocin B
CSSI_Zeocin_cybr2  %>% mutate(H_O_O = Zeocin_OakI_B_Oak, #Zeocin_OakI_B_Oak, 
                           H_O_W = Zeocin_OakI_B_Wine, #Zeocin_OakI_B_Wine,
                           H_W_O = Zeocin_WineI_B_Oak, #Zeocin_WineI_B_Oak, 
                           H_W_W = Zeocin_WineI_B_Wine, #Zeocin_WineI_B_Wine, 
                           L_O_O = Dilute_OakI_B_Oak, #Dilute_OakI_B_Oak,
                           L_O_W = Dilute_OakI_B_Wine,#Dilute_OakI_B_Wine,
                           L_W_O = Dilute_WineI_B_Oak, #Dilute_WineI_B_Oak,
                           L_W_W = Dilute_WineI_B_Wine) %>% #, Dilute_WineI_B_Wine)
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>%
group_by(CHROM, POS) %>% summarise(summary = glmfixed(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Zeocin_cybr2_glmB
saveRDS(object = CSSI_Zeocin_cybr2_glmB, file = "CSSI_Zeocin_cybr2_glmB.rds")

CSSI_Zeocin_cybr2_glmB %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothZ = frollmean(summary, n = 100)) -> CSSI_Zeocin_Window100B

saveRDS(object = CSSI_Zeocin_Window100B, file = "CSSI_Zeocin_Window100B.rds")

```


```{r}
CSSI_Zeocin_Window100B
CSSI_Zeocin_Window100A <- CSSI_Zeocin_Window100

CSSI_Zeocin_Window100B %>% ggplot(aes(x = POS, y = abs(SmoothZ), color = label)) + geom_line() +
  geom_line(data = CSSI_Zeocin_Window100A, aes(x = POS, y = abs(SmoothZ), color = label), linetype = "dashed") + 
  facet_grid(~CHROM, scales = "free") + ggtitle("Zeocin B Solid, A Dashed")

CSSI_Zeocin_Window100B %>% filter(CHROM == "III") %>% ggplot(aes(x = POS, y = abs(SmoothZ), color = label)) + geom_line() +
  geom_line(data = CSSI_Zeocin_Window100A[CSSI_Zeocin_Window100A$CHROM == "III",], aes(x = POS, y = abs(SmoothZ), color = label), linetype = "dashed") + 
  facet_grid(~CHROM, scales = "free") + ggtitle("Zeocin B Solid, A Dashed")

```

## Scaling by Coverage 

Next look at if coverage is significantly different between them?

```{r}

CSSI_Zeocin_cybr2 %>% pivot_longer(c(-CHROM, -POS)) -> ZeocinPivot

ZeocinPivot$Rep <- NA
ZeocinPivot$Rep[grep("_A_", ZeocinPivot$name)] <- "A"
ZeocinPivot$Rep[grep("_B_", ZeocinPivot$name)] <- "B"

################################################################################

ZeocinPivot %>% group_by(POS, CHROM, Rep) %>% summarise(TotalCoverage = sum(value)) -> Zeocin_totalrepcoverage
ZeocinPivot %>% group_by(POS, CHROM, name) %>% summarise(Rep = Rep,
                                                         BulkCoverage = sum(value)) -> Zeocin_bulkcoverage

Zeocin_bulkcoverage %>% filter(CHROM == "III") %>% ggplot(aes(x = POS, y = BulkCoverage, color = name, linetype = Rep)) + geom_line() +
    facet_grid(~CHROM, scales = "free") + ggtitle("Zeocin")

Zeocin_totalrepcoverage %>% filter(CHROM == "III") %>% ggplot(aes(x = POS, y = TotalCoverage, color = Rep)) + geom_line() +
    facet_grid(~CHROM, scales = "free") + ggtitle("Zeocin")

```

Scale by coverage since obviously the OakI A group are over-represented

```{r, eval = FALSE}

#What we want is to sample out of the total for every column so that all columns are the same overall coverage...
CSSI_Zeocin_cybr2 %>% pivot_longer(c(-CHROM, -POS)) %>% group_by(name) %>% 
  summarise(total_coverage = sum(value),
  avg_coverage = mean(value)) %>% arrange(total_coverage) %>% 
  mutate(DecreaseX = total_coverage/min(total_coverage)) -> Zeocin_SNPCoverage

tableinv <- function(x){
      y <- x[rep(rownames(x),x$value),1:(ncol(x))]
      rownames(y) <- c(1:nrow(y))
      return(y)}


CSSI_Zeocin_cybr2 %>% transmute(CHROM = CHROM, POS = POS, 
                                Zeocin_WineI_B_Wine = ceiling(Zeocin_WineI_B_Wine/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Zeocin_WineI_B_Wine"]),
                                Zeocin_WineI_B_Oak = ceiling(Zeocin_WineI_B_Oak/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Zeocin_WineI_B_Oak"]),
                                Zeocin_OakI_A_Oak = ceiling(Zeocin_OakI_A_Oak/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Zeocin_OakI_A_Oak"]),
                                Dilute_OakI_B_Oak = ceiling(Dilute_OakI_B_Oak/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Dilute_OakI_B_Oak"]),
                                Zeocin_OakI_A_Wine = ceiling(Zeocin_OakI_A_Wine/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Zeocin_OakI_A_Wine"]),
                                Dilute_OakI_B_Wine = ceiling(Dilute_OakI_B_Wine/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Dilute_OakI_B_Wine"]),
                                Zeocin_WineI_A_Wine = ceiling(Zeocin_WineI_A_Wine/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Zeocin_WineI_A_Wine"]),
                                Zeocin_WineI_A_Oak = ceiling(Zeocin_WineI_A_Oak/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Zeocin_WineI_A_Oak"]),
                                Dilute_WineI_B_Wine = ceiling(Dilute_WineI_B_Wine/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Dilute_WineI_B_Wine"]),
                                Dilute_WineI_A_Wine = ceiling(Dilute_WineI_A_Wine/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Dilute_WineI_A_Wine"]),
                                
                                Zeocin_OakI_B_Oak = ceiling(Zeocin_OakI_B_Oak/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Zeocin_OakI_B_Oak"]),
                                Dilute_WineI_B_Oak = ceiling(Dilute_WineI_B_Oak/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Dilute_WineI_B_Oak"]),
                                Zeocin_OakI_B_Wine = ceiling(Zeocin_OakI_B_Wine/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Zeocin_OakI_B_Wine"]),
                                Dilute_WineI_A_Oak = ceiling(Dilute_WineI_A_Oak/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Dilute_WineI_A_Oak"]),
                                Dilute_OakI_A_Oak = ceiling(Dilute_OakI_A_Oak/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Dilute_OakI_A_Oak"]),
                                Dilute_OakI_A_Wine = ceiling(Dilute_OakI_A_Wine/Zeocin_SNPCoverage$DecreaseX[Zeocin_SNPCoverage$name == "Dilute_OakI_A_Wine"])) -> CSSI_ZeoCin_Downsampled

CSSI_ZeoCin_Downsampled %>% mutate(H_O_O = Zeocin_OakI_B_Oak, 
                           H_O_W = Zeocin_OakI_B_Wine, #,
                           H_W_O = Zeocin_WineI_B_Oak, #, 
                           H_W_W = Zeocin_WineI_B_Wine, #, 
                           L_O_O = Dilute_OakI_B_Oak, #,
                           L_O_W = Dilute_OakI_B_Wine,#,
                           L_W_O = Dilute_WineI_B_Oak, #,
                           L_W_W = Dilute_WineI_B_Wine) %>% #, )
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>%
group_by(CHROM, POS) %>% summarise(summary = glmfixed(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Zeocin_cybr2_glmB_downsampled

CSSI_ZeoCin_Downsampled %>% mutate(H_O_O = Zeocin_OakI_A_Oak, #, 
                           H_O_W = Zeocin_OakI_A_Wine, #,
                           H_W_O = Zeocin_WineI_A_Oak, #, 
                           H_W_W = Zeocin_WineI_A_Wine, #, 
                           L_O_O = Dilute_OakI_A_Oak, #,
                           L_O_W = Dilute_OakI_A_Wine,#,
                           L_W_O = Dilute_WineI_A_Oak, #,
                           L_W_W = Dilute_WineI_A_Wine) %>% #, )
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>%
group_by(CHROM, POS) %>% summarise(summary = glmfixed(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Zeocin_cybr2_glmA_downsampled

saveRDS(rbind(data.frame(CSSI_Zeocin_cybr2_glmA_downsampled, Rep = "A"),
      data.frame(CSSI_Zeocin_cybr2_glmB_downsampled, Rep = "B")), file = "DownsampledZeocin.rds")
```

```{r}
CSSI_Zeocin_cybr2_glmA_downsampled <- readRDS("../../cybrBSA/cybrBSA/DownsampledZeocin.rds")

CSSI_Zeocin_cybr2_glmA_downsampled %>% filter(Rep == "A") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothZ = frollmean(summary, n = 100)) -> CSSI_Zeocin_cybr2_glmA_downsampled_window

CSSI_Zeocin_cybr2_glmA_downsampled %>% filter(Rep == "B") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothZ = frollmean(summary, n = 100)) -> CSSI_Zeocin_cybr2_glmB_downsampled_window

rbind(data.frame(CSSI_Zeocin_cybr2_glmA_downsampled_window, Rep = "A"),
      data.frame(CSSI_Zeocin_cybr2_glmB_downsampled_window, Rep = "B")) %>% ggplot(aes(x = POS, y = SmoothZ, color = label, linetype = Rep)) + geom_line() + facet_grid(~CHROM, scale = "free")
```

## Comparing Sub-Sampled vs Not in Zeocin

```{r}
rbind(data.frame(CSSI_Zeocin_cybr2_glmA_downsampled_window, Rep = "A"),
      data.frame(CSSI_Zeocin_cybr2_glmB_downsampled_window, Rep = "B")) %>% ggplot(aes(x = POS, y = abs(SmoothZ), color = label, linetype = Rep)) + geom_line() + facet_grid(~CHROM, scale = "free") + ggtitle("Downsampled to lowest Coverage")
```

```{r}
rbind(data.frame(CSSI_Zeocin_Window100A, Rep = "A"), 
      data.frame(CSSI_Zeocin_Window100B, Rep = "B")) %>% ggplot(aes(x = POS, y = abs(SmoothZ), color = label, linetype = Rep)) + geom_line() + facet_grid(~CHROM, scales = "free") + ggtitle("Zeocin_A")

```
```{r}
rbind(data.frame(CSSI_Zeocin_cybr2_glmA_downsampled_window, Rep = "A", Coverage = "Min"),
      data.frame(CSSI_Zeocin_cybr2_glmB_downsampled_window, Rep = "B", Coverage = "Min"),
      data.frame(CSSI_Zeocin_Window100A, Rep = "A", Coverage = "Full"), 
      data.frame(CSSI_Zeocin_Window100B, Rep = "B", Coverage = "Full")
      ) %>% filter(CHROM == "III") %>% ggplot(aes(x = POS, y = abs(SmoothZ), color = label, linetype = Rep)) + geom_line() + facet_grid(~Coverage, scales = "free") + ggtitle("Zeocin Replicate Comparison Chr III")

rbind(data.frame(CSSI_Zeocin_cybr2_glmA_downsampled_window, Rep = "A", Coverage = "Min"),
      data.frame(CSSI_Zeocin_cybr2_glmB_downsampled_window, Rep = "B", Coverage = "Min"),
      data.frame(CSSI_Zeocin_Window100A, Rep = "A", Coverage = "Full"), 
      data.frame(CSSI_Zeocin_Window100B, Rep = "B", Coverage = "Full")
      ) %>% filter(CHROM == "VII") %>% ggplot(aes(x = POS, y = abs(SmoothZ), color = label, linetype = Rep)) + geom_line() + facet_grid(~Coverage, scales = "free") + ggtitle("Zeocin Replicate Comparison Chr VII")
```

## Rollmean before glm

using confint to get confidence intervals of the model, from Eugene

```{r}
CSSI_Zeocin_cybr2 <- readRDS("../../cybrBSA/cybrBSA/CSSI_Zeocin_cybr2.rds")
CSSI_Cyc_cybr2 <- readRDS("../../cybrBSA/cybrBSA/CSSI_Cyc_cybr2.rds")
CSSI_CuSO4_cybr2 <- readRDS("../../cybrBSA/cybrBSA/CSSI_CuSO4_cybr2.rds")
```

```{r}

CSSI_Zeocin_cybr2 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothCount = frollmean(value, n = 100)) %>% pivot_wider(names_from = label,values_from = SmoothCount) %>% na.omit() %>%
  mutate(H_O_O = Zeocin_OakI_B_Oak, 
                           H_O_W = Zeocin_OakI_B_Wine,
                           H_W_O = Zeocin_WineI_B_Oak,
                           H_W_W = Zeocin_WineI_B_Wine, 
                           L_O_O = Dilute_OakI_B_Oak, 
                           L_O_W = Dilute_OakI_B_Wine,
                           L_W_O = Dilute_WineI_B_Oak, 
                           L_W_W = Dilute_WineI_B_Wine) %>% 
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>%
group_by(CHROM, POS) %>% summarise(summary = glmfixed(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Zeocin_rollmean

CSSI_Zeocin_rollmean %>% ggplot(aes(x = POS, y = summary, color = label)) + geom_line() + facet_grid(~CHROM, scales = "free")
```

```{r}
CSSI_Zeocin_cybr2 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothCount = frollmean(value, n = 100)) %>% pivot_wider(names_from = label,values_from = SmoothCount) %>% na.omit() %>%
  mutate(H_O_O = Zeocin_OakI_B_Oak, 
                           H_O_W = Zeocin_OakI_B_Wine,
                           H_W_O = Zeocin_WineI_B_Oak,
                           H_W_W = Zeocin_WineI_B_Wine, 
                           L_O_O = Dilute_OakI_B_Oak, 
                           L_O_W = Dilute_OakI_B_Wine,
                           L_W_O = Dilute_WineI_B_Oak, 
                           L_W_W = Dilute_WineI_B_Wine) %>% 
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>%
  group_by(CHROM, POS) %>% summarise(summary = glmfixed(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W)[1:4],
                                                   label = c("intercept", "Bulk", "Parent", "Interaction")) -> CSSI_Zeocin_rollmean
                                   
CSSI_Zeocin_cybr2 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, SmoothCount = frollmean(value, n = 100)) %>% pivot_wider(names_from = label,values_from = SmoothCount) %>% na.omit() %>%
  mutate(H_O_O = Zeocin_OakI_B_Oak, 
                           H_O_W = Zeocin_OakI_B_Wine,
                           H_W_O = Zeocin_WineI_B_Oak,
                           H_W_W = Zeocin_WineI_B_Wine, 
                           L_O_O = Dilute_OakI_B_Oak, 
                           L_O_W = Dilute_OakI_B_Wine,
                           L_W_O = Dilute_WineI_B_Oak, 
                           L_W_W = Dilute_WineI_B_Wine) %>% 
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>% head(1) -> test1

test1 %>% summarise(glmfixed(H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W))

combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(test1$H_O_O, test1$H_O_W, test1$H_W_O, test1$H_W_W, test1$L_O_O, test1$L_O_W, test1$L_W_O, test1$L_W_W))

test2 <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

as.vector(confint(test2))
```

```{r}
glmfixed_confint <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  c(summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)], as.vector(confint(b)))

}

glmfixed_all <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  return(b)

}

```


Implementing new function - this is SO slow and does not work

```{r, eval = FALSE}

CSSI_Zeocin_cybr2 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = frollmean(value, n = 100)) %>% pivot_wider(names_from = label,values_from = SmoothCount) %>% na.omit() %>%
  mutate(H_O_O = Zeocin_OakI_B_Oak, 
                           H_O_W = Zeocin_OakI_B_Wine,
                           H_W_O = Zeocin_WineI_B_Oak,
                           H_W_W = Zeocin_WineI_B_Wine, 
                           L_O_O = Dilute_OakI_B_Oak,  
                           L_O_W = Dilute_OakI_B_Wine,
                           L_W_O = Dilute_WineI_B_Oak, 
                           L_W_W = Dilute_WineI_B_Wine) %>% 
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>%
group_by(CHROM, POS) %>% summarise(summary = glmfixed_confint(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W),
                                                   label = c("intercept", "Bulk", "Parent", "Interaction",
                                                             "CIL_intercept", "CIL_Bulk", "CIL_Parent", "CIL_Interaction",
                                                             "CIU_intercept", "CIU_Bulk", "CIU_Parent", "CIU_Interaction")) -> CSSI_Zeocin_rollmean2
```
 

```{r}
CSSI_Zeocin_cybr2 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = frollmean(value, n = 100)) %>% pivot_wider(names_from = label,values_from = SmoothCount) %>% na.omit() %>%
  mutate(H_O_O = Zeocin_OakI_B_Oak, 
                           H_O_W = Zeocin_OakI_B_Wine,
                           H_W_O = Zeocin_WineI_B_Oak,
                           H_W_W = Zeocin_WineI_B_Wine, 
                           L_O_O = Dilute_OakI_B_Oak,  
                           L_O_W = Dilute_OakI_B_Wine,
                           L_W_O = Dilute_WineI_B_Oak, 
                           L_W_W = Dilute_WineI_B_Wine) %>% 
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>% head(50) %>%
group_by(CHROM, POS) %>% summarise(summary = glmfixed_confint(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W),
                                                   label = c("intercept", "Bulk", "Parent", "Interaction",
                                                             "CIL_intercept", "CIL_Bulk", "CIL_Parent", "CIL_Interaction",
                                                             "CIU_intercept", "CIU_Bulk", "CIU_Parent", "CIU_Interaction")) -> CSSI_Zeocin_rollmean2

CSSI_Zeocin_cybr2 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = frollmean(value, n = 100)) %>% pivot_wider(names_from = label,values_from = SmoothCount) %>% na.omit() %>%
  mutate(H_O_O = Zeocin_OakI_B_Oak, 
                           H_O_W = Zeocin_OakI_B_Wine,
                           H_W_O = Zeocin_WineI_B_Oak,
                           H_W_W = Zeocin_WineI_B_Wine, 
                           L_O_O = Dilute_OakI_B_Oak,  
                           L_O_W = Dilute_OakI_B_Wine,
                           L_W_O = Dilute_WineI_B_Oak, 
                           L_W_W = Dilute_WineI_B_Wine) %>% 
  filter(CHROM == "VII") %>% 
select(CHROM, POS, H_O_O,H_O_W,H_W_O,H_W_W,L_O_O,L_O_W,L_W_O,L_W_W) %>% 
group_by(CHROM, POS) %>% summarise(summary = glmfixed_confint(H_O_O,H_O_W,H_W_O,H_W_W,
                                                            L_O_O,L_O_W,L_W_O,L_W_W),
                                                   label = c("intercept", "Bulk", "Parent", "Interaction",
                                                             "CIL_intercept", "CIL_Bulk", "CIL_Parent", "CIL_Interaction",
                                                             "CIU_intercept", "CIU_Bulk", "CIU_Parent", "CIU_Interaction")) -> CSSI_Zeocin_rollmean_VII

saveRDS(CSSI_Zeocin_rollmean_VII, file = "CSSI_Zeocin_rollmean_VII.rds")

  CSSI_Zeocin_rollmean2 %>% pivot_wider(names_from = label, values_from = summary) 
```


```{r}
CSSI_Zeocin_rollmean_VII %>% pivot_wider(names_from = label, values_from = summary) %>% 
  ggplot(aes(x = POS, y = Interaction)) + geom_line(color = "firebrick") + geom_ribbon(aes(x = POS, ymin = Interaction+CIL_Interaction, ymax = Interaction+CIU_Interaction), alpha = 0.1, fill = "firebrick") +
  geom_line(aes(y = Bulk), color = "black") + geom_ribbon(aes(x = POS, ymin = Bulk+CIL_Bulk, ymax = Bulk+CIU_Bulk), alpha = 0.1, fill = "black") +
  geom_line(aes(x = POS, y = Parent), color = "goldenrod") + geom_ribbon(aes(x = POS, ymin = Parent+CIL_Parent, ymax = Parent+CIU_Parent), alpha = 0.1, fill = "goldenrod") + ggtitle("Window 100 + confint") + ylab("Z score")

# CSSI_Zeocin_rollmean_VII %>% pivot_wider(names_from = label, values_from = summary) %>% 
#   ggplot(aes(x = POS, y = Bulk)) + geom_line() +
#   geom_ribbon(aes(x = POS, ymin = CIL_Bulk, ymax = CIU_Bulk), alpha = 0.3)
# 
# CSSI_Zeocin_rollmean_VII %>% pivot_wider(names_from = label, values_from = summary) %>% 
#   ggplot(aes(x = POS, y = Bulk)) + geom_line() +
#     geom_line(aes(x = POS, y = CIU_Bulk), alpha = 0.3) +
#   geom_line(aes(x = POS, y = CIL_Bulk), alpha = 0.3) 
#  
#   geom_ribbon(aes(x = POS, ymin = Interaction+CIL_Interaction, ymax = Interaction+CIU_Interaction), alpha = 0.3) 
# 
# CSSI_Zeocin_rollmean_VII %>% pivot_wider(names_from = label, values_from = summary) %>% 
#   ggplot(aes(x = POS, y = Interaction)) + geom_line() +
#     geom_line(aes(x = POS, y = Interaction+CIU_Interaction), alpha = 0.3) +
#   geom_line(aes(x = POS, y = Interaction+CIL_Interaction), alpha = 0.3) 
```



