---
title: "CYBR BSA Summary"
author: "Cassandra Buzby"
date: "10/4/2022"
output: 
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(dplyr)

library(foreach)
library(doParallel)

 ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))
```

First, may the record show that the initials CYB are fun to combine with the language R to make the CYBR package. I should not be so excited about this but I am.

## Outline and Setup

This markdown contains the packages used for each step of my QTL analysis, with a sample input and output for each as a header.

### Load Package Dependencies

```{r, eval = FALSE}
#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(dplyr)

library(foreach)
library(doParallel)

```


## Convert VCF to GATK table

Script is in bash

```{bash, eval = FALSE}
module load gatk/4.2.0.0

gatk VariantsToTable \
     -V ${myfile} \
     -F CHROM -F POS -F REF -F ALT \
     -GF AD -GF DP -GF GQ -GF PL \
     -O ${myfile}.output.table

```


## Convert GATK table to Data Frame

Make a better way to do this that doesn't use QTLseqR

```{r}
cybrInputGATKTable <- function(rawData, yeast = TRUE){
  
  require(dplyr)
  require(doParallel)
  require(foreach)
  
  HNGLCDRXY <- read.table(rawData, header = TRUE)
  
  #Identify the unique values besides AD/DP/GQ/PL
  gsub(".AD", "", 
       gsub(".GQ", "", 
            gsub(".DP","",
                 gsub(".PL","",
                      colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
  i <- Samples[1]

  resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
    mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
    AD <- select(HNGLCDRXY,contains(c(i))) %>% select(., contains("AD"))
    GQ <- select(HNGLCDRXY,contains(c(i))) %>% select(., contains("GQ"))
    DP <- select(HNGLCDRXY,contains(c(i))) %>% select(., contains("DP"))
    PL <- select(HNGLCDRXY,contains(c(i))) %>% select(., contains("PL")) 
    cbind(mydf, AD , GQ , DP, PL) -> mydftotal
    colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")
    
    mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>% 
      separate(PL, c('PL.REF','PL.ALT'), extra='drop') -> mycdf
    
    mycdf
  }
  
  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))
    
    resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
  }else{
    results <- resultscdf
  }
  return(results)
  
}

```

## Filter by quality

```{r}
cybrQualityFilter <- function(gatkdf, GQcutoff = 98, cleandata = TRUE){
  
  #Filter by quality
  gatkdf %>% filter(GQ > GQcutoff) %>% 
    select(-DP, -GQ, -PL.ALT, -PL.REF) %>% 
    pivot_longer(c(AD.REF, AD.ALT), names_to = "Allele", values_to = "ReadCount") %>% 
    na.omit() -> filteredgatkdf
  
  if(cleandata == TRUE){
    #REMOVE POSITIONS WHERE THERE ISN'T ALL BULKS
    filteredgatkdf %>% group_by(CHROM, POS) %>% 
      summarize(uniqueDatasets = length(unique(Dataset))) -> testingforglm
    
    filteredgatkdf %>% left_join(testingforglm, by = c("CHROM", "POS")) %>% 
      filter(uniqueDatasets == max(uniqueDatasets)) %>% select(-uniqueDatasets) -> cleanedgatkdf
  }else{
    cleanedgatkdf <- filteredgatkdf
  }
  
  return(cleanedgatkdf)
}
```

## Convert Parental VCFs to Data Frame

```{r}
cybrConvertParentalAlleles <- function(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), 
                                   Parents = gsub("_VCF.txt","", ParentFiles), Truncate = TRUE, yeast = TRUE){
  temparent <- list()
  mergeparents <- foreach(i=1:length(ParentFiles), .combine=rbind) %dopar% {
    read.table(ParentFiles[i], header = TRUE) %>% mutate(parent = Parents[i])
  }
  
  if(yeast == TRUE){
     ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))
     
    rbind(mergeparents) %>% arrange(CHROM, POS) %>% 
    select(CHROM, POS, REF, ALT, parent) %>% 
    merge(ChromKey) %>% select(-CHROM) %>% 
    mutate(CHROM = chromosomes) %>% select(-chromosomes) -> ParentalVCF
     
  }else{
    rbind(mergeparents) %>% arrange(CHROM, POS) %>% 
    select(CHROM, POS, REF, ALT, parent) -> ParentalVCF
  }
  
  ParentalVCF %>% pivot_wider(names_from = parent, values_from = ALT) -> SNPids

  SNPids$Type <- 0
  for(i in Parents){
    
    #filter rows in which all values of columns of the parent NOT selected are NA
    select(SNPids,-i, -CHROM, -POS, -REF) -> tempdf
    tempdf$Any_NA <- apply(tempdf, 1, function(x) anyNA(x))
    SNPids$Type[which(tempdf$Any_NA)] <- i
    rm(tempdf)
  }
  
  
  #Collect it to output
  if(Truncate == TRUE){
    SNPids %>% select(CHROM, POS,  Type) %>% filter(Type != 0) -> SNPids
  }
  
  return(SNPids)
    
}

```

## Combine Parental and Experimental Variants

Combine truncated with total dataset

```{r}

cybrIDAlleles <- function(BSAdfstart = finaldf, Parentdf = test, yeast = TRUE){
  
  BSAdf <- left_join(BSAdfstart, Parentdf)
  BSAdf$PAllele <- NA
  
  Parents <- unique(Parentdf$Type)
  if(length(Parents) == 2){
      BSAdf$PAllele[BSAdf$Allele == "AD.REF" & BSAdf$Type == Parents[1]] <- Parents[1]
      BSAdf$PAllele[BSAdf$Allele == "AD.REF" & BSAdf$Type == Parents[2]] <- Parents[2]
      
      #Run if only two-parent cross
      BSAdf$PAllele[BSAdf$Allele == "AD.ALT" & BSAdf$Type == "Wine"] <- "Oak"
      BSAdf$PAllele[BSAdf$Allele == "AD.ALT" & BSAdf$Type == "Oak"] <- "Wine"
  }else{
    for(i in Parents){
      BSAdf$PAllele[BSAdf$Allele == "AD.REF" & BSAdf$Type == i] <- i
    }
  }
  
  if(yeast == TRUE){
     ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))
       BSAdf$CHROM <- factor(BSAdf$CHROM, levels = ChromKey$chromosomes)
  }
 
  
  #Convert all to factors for glm
  BSAdf$Dataset <- factor(BSAdf$Dataset)
  BSAdf$Allele <- factor(BSAdf$Allele)
  BSAdf$PAllele <- factor(BSAdf$PAllele)
  
  return(BSAdf)
}

```

## Running Pipeline of Functions

```{r, warning=FALSE, message=FALSE}
mydatatotest = "HGV.SortedCat.vcf.output.table"

mydf <- cybrInputGATKTable(mydatatotest)
head(mydf)

qualitydf <- cybrQualityFilter(mydf)
head(qualitydf)

parentSNPids <- cybrConvertParentalAlleles(Truncate = TRUE)
head(parentSNPids)

testmerge <- cybrIDAlleles(BSAdfstart = qualitydf, Parentdf = parentSNPids, yeast = TRUE)
head(testmerge)

```

Reformat data so that it has bulk etc included

```{r}
unique(testmerge$Dataset)

testmerge$Bulk <- NA
testmerge$Bulk[testmerge$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              testmerge$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              testmerge$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" | 
              testmerge$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq"] <- "Dilute"

testmerge$Bulk[testmerge$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              testmerge$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" |
              testmerge$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq" | 
              testmerge$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "Fluconazole"

testmerge$Parent <- NA
testmerge$Parent[testmerge$Dataset == "HGVMVDRX2_n01_OakI_Dilute_B.fastq" |
              testmerge$Dataset == "HGVMVDRX2_n01_OakI_Fluc_B.fastq" |
              testmerge$Dataset == "HGVMVDRX2_n01_OakI_Fluc_C.fastq" | 
              testmerge$Dataset == "HGVMVDRX2_n01_OakI_Fluc_D.fastq"] <- "OakI"

testmerge$Parent[testmerge$Dataset == "HGVMVDRX2_n01_WineI_Dilute_B.fastq" |
              testmerge$Dataset == "HGVMVDRX2_n01_WineI_Dilute_C.fastq" |
              testmerge$Dataset == "HGVMVDRX2_n01_WineI_Dilute_D.fastq" | 
              testmerge$Dataset == "HGVMVDRX2_n01_WineI_Fluc_C.fastq"] <- "WineI"

testmerge$Parent <- factor(testmerge$Parent)
testmerge$Bulk <- factor(testmerge$Bulk)
testmerge$ReadCount <- as.numeric(testmerge$ReadCount)

head(testmerge)

```


```{r, eval = FALSE}
#For testing

load("RdataFiles/CSSI_Fluc_Pivot_2.Rdata")
head(CSSI_Fluc_Pivot)
lrP <- CSSI_Fluc_Pivot
chr = "I"
windowsize = 100
POS = 925
i <- POS

lrP <- subset(lrP, CHROM == chr)

lrP[lrP$POS == i,]
```

## FUNCTION: Run GLM

```{r}
cybrBSA_GLM <-  function(lrP, chr = "II", windowsize = 100, formula = "Allele ~ Bulk*Parent", 
                           resultscol = c("Intercept", "Bulk", "Parent", "Interaction")){
  
  lrP <- subset(lrP, CHROM == chr)
  
  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      #res <- glm(Allele ~ Bulk*Parent+Day, weights = ReadCount, family = binomial, data = lrP[lrP$POS == i,])
      res <- glm(as.formula(formula), weights = ReadCount, family = binomial, data = lrP[lrP$POS == i,])

      c(c, i, summary(res)$coefficients[(length(summary(res)$coefficients)/2+1):(length(summary(res)$coefficients) - length(summary(res)$coefficients)/4)])
    }
  
    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", resultscol)
    for(i in 2:length(colnames(Results))){
      Results[,i] <- as.numeric(Results[,i])
    }
    Results %>% arrange(POS) %>% select(-Intercept) -> Results
  }
return(Results)
}

#Hard-coded windows
SmoothBSAWindows <- function(Results){
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkFluconazole[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                res_day <- mean(Results$Day2[(i-windowsize):(i+windowsize)])

                #print CHROM, index, POS, and mean
                c(chr, i, Results$POS[i], 
                  Results$bulkFluconazole[i], Results$parentWine[i],Results$Interaction[i], Results$Day2[i],
                  res_bulk, res_parent, res_day, res_int)}
  
  WResult <- as.data.frame(WResult)
  
  colnames(WResult) <- c("CHROM", "Index", "POS", 
                         "Bulk_Z", "Parent_Z", "Interaction_Z", "Day_Z",
                         "Bulk_Zprime", "Parent_Zprime", "Day_Zprime","Interaction_Zprime")
  
  for(i in 2:length(colnames(WResult))){
      Results[,i] <- as.numeric(Results[,i])
  }
  
  # WResult[,2] <- as.numeric(WResult[,2])
  # WResult[,3] <- as.numeric(WResult[,3])
  # WResult[,4] <- as.numeric(WResult[,4])
  # WResult[,5] <- as.numeric(WResult[,5])
  # WResult[,6] <- as.numeric(WResult[,6])
  # WResult[,7] <- as.numeric(WResult[,7])
  # WResult[,8] <- as.numeric(WResult[,8])
  # WResult[,9] <- as.numeric(WResult[,9])
  # WResult[,10] <- as.numeric(WResult[,10])
  # WResult[,11] <- as.numeric(WResult[,11])
  # 
  return(WResult)
}
```

New Window Script Function

```{r}
cybrSmoothBSAWindows <- function(Results, windowsize = 100, chr = unique(Results$CHROM)[1]){
  
  #extract parameters
  Results %>% select(-CHROM, -POS) %>% names() -> params
  
  #arrange by position
  Results %>% arrange(POS) -> Results
    
  #run window
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            
      smoothedparams <- vector()
       for(p in 1:length(params)){
          smoothedparams[p] <- mean(Results[[params[p]]][(i-windowsize):(i+windowsize)])
       }

      #print CHROM, index, POS, and mean
      unlist(c(chr, i, Results[i,2:(2+length(params))],smoothedparams))
      }
  
  #rename columns
  WResult <- as.data.frame(WResult)
  
  colnames(WResult) <- c("CHROM", "Index", "POS", 
                         paste(params, "Z", sep = "_"),
                         paste(params, "Zprime", sep = "_"))
  
  #convert to numeric
  for(i in 2:length(colnames(WResult))){
      WResult[,i] <- as.numeric(WResult[,i])
  }
  
  return(WResult)
}
```

Testing new cybr scripts

```{r}
testbsa <- cybrBSA_GLM(testmerge)
anothertest <- cybrSmoothBSAWindows(testbsa)

#Running the entire genome in a loop instead of just chr II
wholegenomeBSA <- foreach(i=unique(testmerge$CHROM), .combine=rbind) %dopar%{
  cybrBSA_GLM(testmerge, chr = i) %>% cybrSmoothBSAWindows()
}

length(unique(wholegenomeBSA$Parent_Zprime))
dim(wholegenomeBSA)
```
Plot this

```{r}
wholegenomeBSA$CHROM <- factor(wholegenomeBSA$CHROM, levels = ChromKey$chromosomes)

wholegenomeBSA %>% filter(CHROM != "I", CHROM != "M") %>%
  pivot_longer(c(Bulk_Zprime, Parent_Zprime, Interaction_Zprime), names_to = "Factor", values_to = "Zprime") %>% 
  ggplot(aes(x = POS, y = Zprime, color = Factor)) + 
  
  #Add Bulk, Parent, Day, and Interaction Traces
  geom_line(size = 0.5, alpha = 0.8) + 
  scale_color_manual(values = c("#345F6F", "#D7335C", "#FFB05C")) + 
  
  #Add the breaks for the facets
  scale_x_continuous(breaks = seq(from = 0, to = max(wholegenomeBSA$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
  
wholegenomeBSA %>% filter(CHROM != "I", CHROM != "M") %>%
  ggplot(aes(x = POS, y = Bulk_Z)) + 
  
  #Add Bulk, Parent, Day, and Interaction Traces
  geom_point(size = 0.5, alpha = 0.08) + 
  scale_color_manual(values = c("#345F6F", "#D7335C", "#FFB05C")) + 
  
  #Add the breaks for the facets
  scale_x_continuous(breaks = seq(from = 0, to = max(wholegenomeBSA$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
  
```


Input:

```{r, collapse=TRUE}
load("RdataFiles/CSSI_Fluc_Pivot_2.Rdata")
head(CSSI_Fluc_Pivot)
```

Output:

```{r, collapse=TRUE}
load("RdataFiles/CSSI_F_ZScores_GLM_allexcl.Rdata")
head(CSSI_F_ZScores_GLM)
```

## Plotting

Using the same data as above

```{r}
load("RdataFiles/CSSI_F_ZScores_GLM_allexcl.Rdata")

CSSI_F_ZScores_GLM %>% ggplot(aes(x = POS, y = Bulk_Zprime)) + 
  #Set Cutoff Line
  geom_hline(yintercept = c(-1.96, 1.96), color = "gray")+
  
  #Add Bulk, Parent, Day, and Interaction Traces
  geom_line(color = "#345F6F", size = 0.5, alpha = 0.8) + 
  geom_line(aes(x = POS, y = Parent_Zprime), color = "#FFB05C", alpha= 0.8, size = 0.5) +
  geom_line(aes(x = POS, y = Day_Zprime), color = "gray", alpha= 0.8, size = 0.5) +
  geom_line(aes(x = POS, y = Interaction_Zprime), color = "#D7335C", alpha= 0.8, size = 0.5) +
  
  #Add the breaks for the facets
  scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  
  #Add legends and title
  ggtitle("CSSI Fluconazole | Comparison of Smoothed Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
```

