---
title: "Unselected Comparisons of Parental Backgorunds"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

library(stringr)
#library(cybrBSA)

#install.packages("lme4")
library(lme4)

ggplot2::theme_set(theme_light())

################################################################################
glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
#glm with replicates
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

```


## Load in Data for each experiment

```{r, eval = FALSE}
setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023")

dir("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/AnalysisAndFigures_2023/Data/")

Fluc_1 <- readRDS("Data/Fluconazole_1_cybr2.rds")
Cu_1_A <- readRDS("Data/CuSO4_1_cybr2.rds")
Cu_1_B <- readRDS("Data/CuSO4_2_cybr2.rds")
CycZeo_1 <- readRDS("Data/Zeocin_cybr2.rds")
Cu_8 <- readRDS("Data/CuSO4_CSS8_cybr2.rds")

```
## Select only the dilute groups

```{r, eval = FALSE}
Fluc_1 %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), 
                        names_sep = "_",
                        values_to = "Reads") %>% filter(Bulk == "Dilute") -> Fluc_1_Dilute

Cu_1_A %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Allele"), 
                        names_sep = "_",
                        values_to = "Reads") %>% filter(Bulk == "Dilute") -> Cu_1_A_Dilute

Cu_1_B %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), 
                        names_sep = "_",
                        values_to = "Reads") %>% filter(Bulk == "Dilute") -> Cu_1_B_Dilute

CycZeo_1 %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), 
                        names_sep = "_",
                        values_to = "Reads") %>% filter(Bulk == "Dilute") -> CycZeo_1_Dilute

Cu_8 %>% pivot_longer(-c(CHROM, POS), names_to = c("Bulk", "Parent", "Rep", "Dose", "Allele"), 
                        names_sep = "_",
                        values_to = "Reads") %>% filter(Bulk == "Dilute") -> Cu_8_Dilute

```


```{r, eval = FALSE}
Fluc_1_Dilute %>% mutate(across('Parent', str_replace, "OakI" ,"O1"),
                         across('Parent', str_replace, "WineI" ,"W1")) %>%
  mutate(exp = "F1") -> F1

Cu_1_A_Dilute %>% mutate(across('Parent', str_replace, "Oak" ,"O1"),
                         across('Parent', str_replace, "Wine" ,"W1")) %>%
  mutate(Rep = NA) %>%
  mutate(exp = "C1A") %>%
  select(CHROM, POS, Bulk, Parent, Rep, Allele, Reads, exp) -> C1A

Cu_1_B_Dilute %>% mutate(across('Parent', str_replace, "Oak" ,"O1"),
                         across('Parent', str_replace, "Wine" ,"W1")) %>%
  mutate(exp = "C1B")-> C1B

CycZeo_1_Dilute %>% mutate(across('Parent', str_replace, "Oak" ,"O1"),
                         across('Parent', str_replace, "Wine" ,"W1")) %>%
  mutate(exp = "Z1")-> Z1

Cu_8_Dilute %>% mutate(across('Parent', str_replace, "Oak1" ,"O1"),
                        across('Parent', str_replace, "Wine1" ,"W1"),
                       across('Parent', str_replace, "Oak8" ,"O8"),
                        across('Parent', str_replace, "Wine8" ,"W8"))%>%
  mutate(exp = "C8") %>%
  select(-Dose)-> C8

rbind(F1, C1A, C1B, Z1, C8) -> AllDilute

saveRDS(AllDilute, file = "2023_Dilute_Consolidated.rds")
```

## Load in file from before

```{r}
AllDilute <- readRDS("2023_Dilute_Consolidated.rds")

```

#First Visualize how different the log ratios are

```{r, eval = FALSE}
AllDilute %>% pivot_wider(names_from = Allele, values_from = Reads) %>% mutate(logRatio = abs(log(Wine/Oak))) %>%
  ggplot(aes(x = POS, y = logRatio, color = Parent, group = exp)) + geom_point(alpha = 0.1, size = 0.1) + facet_grid(~CHROM, space = "free", scales = "free")

AllDilute %>% pivot_wider(names_from = Allele, values_from = Reads) %>% 
  filter(CHROM != "I", CHROM != "M", CHROM != "VIII") %>%
  mutate(logRatio = abs(log(Wine/Oak))) %>%
  ggplot(aes(x = POS, y = logRatio, color = Parent, group = exp)) + geom_point(alpha = 0.1, size = 0.1) + facet_grid(~CHROM, space = "free", scales = "free") +
  theme(legend.position = "bottom")
```

## Smooth all of these to compare

```{r}
AllDilute %>% pivot_wider(names_from = c(Bulk, Parent, Rep, exp, Allele), values_from = Reads) %>%
  pivot_longer(c(-CHROM, -POS), names_to = "label") %>% 
  group_by(CHROM, label) %>% arrange(POS) %>% 
  na.omit() %>%
  summarize(POS = POS, 
            CHROM = CHROM, 
            SmoothCount = ceiling(frollapply(value, n = 200, FUN = median))) %>% #CHANGE THIS LINE
  pivot_wider(names_from = label,values_from = SmoothCount) -> rollData

rollData %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Exp", "Allele"), names_sep = "_", values_to = "Reads")  -> dilute_combineddata

unique(dilute_combineddata$CHROM)
```

## Doing the log ratios again smoothed

```{r}
# dilute_combineddata %>% pivot_wider(names_from = Allele, values_from = Reads) %>% 
#   filter(CHROM != "I", CHROM != "M", CHROM != "VIII") %>%
#   mutate(logRatio = abs(log(Wine/Oak))) %>%
#   ggplot(aes(x = POS, y = logRatio, color = paste(Exp, Rep), linetype = Parent)) + geom_line(alpha = 0.4) + facet_grid(~CHROM, space = "free", scales = "free") +
#   theme(legend.position = "bottom") +
#   scale_color_manual(values = c("gray", "gray", "violet", "violet", "gray",
#                                 "gray", "black", "black", "black", "black"))

dilute_combineddata %>% pivot_wider(names_from = Allele, values_from = Reads) -> dilute_combined_wider

for(i in unique(dilute_combineddata$CHROM)){
  dilute_combined_wider %>%
  filter(CHROM == i) %>%
  mutate(logRatio = abs(log(Wine/Oak))) %>%
  ggplot(aes(x = POS, y = logRatio, color = Rep)) + geom_point(alpha = 0.3, size = 0.4) + facet_grid(Parent~Exp) + 
  theme(legend.position = "bottom")  + scale_color_manual(values = c("lightblue", "maroon", "navy", "darkorange", "black")) +
    ggtitle(paste("Chromosome", i, sep = " ")) -> plot
  print(plot)
}

```

Let's plot these again but try to visualize them in chromosomes

```{r}
dilute_combineddata %>% pivot_wider(names_from = Allele, values_from = Reads) %>% 
  filter(CHROM != "I", CHROM != "M", CHROM != "VIII", CHROM != "III") %>%
  mutate(logRatio = abs(log(Wine/Oak))) %>%
  ggplot(aes(x = POS, y = logRatio, color = paste(Parent, Exp, Rep))) + 
  geom_point(size = 0.2, alpha = 0.2) + 
  facet_wrap(facets = "CHROM", scales = "free") +
  theme(legend.position = "none") +
  scale_color_manual(values = c("blue1", "blue2", "blue3", "blue4", "darkblue", "gray20","black",#O1
                                "darkturquoise", "turquoise",  #O8
                                "red1", "red2", "red3", "red4", "firebrick", "darkred", "maroon", "orchid","violet", #W1
                                "gold", "orange")) + #W8
  ggtitle("Dilute Chromosomes by Parent")

#Making sure the colors line up
dilute_combineddata %>% pivot_wider(names_from = Allele, values_from = Reads) %>% 
  filter(CHROM == "IV") %>%
  mutate(logRatio = abs(log(Wine/Oak))) %>%
  ggplot(aes(x = POS, y = logRatio, color = paste(Parent, Exp, Rep))) + 
  geom_point(size = 0.2, alpha = 0.2) + 
  facet_grid(Exp~Parent) +
  theme(legend.position = "bottom") +
  scale_color_manual(values = c("blue1", "blue2", "blue3", "blue4", "darkblue", "gray20","black",#O1
                                "darkturquoise", "turquoise",  #O8
                                "red1", "red2", "red3", "red4", "firebrick", "darkred", "maroon", "orchid","violet", #W1
                                "gold", "orange")) + #W8
  ggtitle("Chr IV to check colors")
```




## Doing a GLM for the dilute samples

```{r}
dilute_combineddata %>% mutate(RepSpec = paste(Rep, Exp, sep = "")) %>% na.omit() %>% select(-Bulk, -Rep) %>%
  pivot_wider(names_from = c(Parent, Exp, RepSpec, Allele), values_from = Reads) %>%
  na.omit() -> dilute_glmready

#Test a single glm with this

dilute_glmready %>% #head(1) %>%
  pivot_longer(c(-CHROM, -POS), names_to = c("Parent", "Exp", "Rep", "Allele"), names_sep = "_", values_to = "Reads") %>% mutate_if(is.character, as.factor) -> cd

```

***

## Look at the error on the logRatios

```{r, eval = FALSE}
dilute_combineddata %>% pivot_wider(names_from = Allele, values_from = Reads) %>% 
  mutate(logRatio = abs(log(Wine/Oak))) -> logRatioData

logRatioData %>% filter(POS %in% cd$POS) %>% group_by(CHROM, POS, Parent) %>% summarize(mean_LR = mean(logRatio, na.rm = TRUE),
                                                            sd = sd(logRatio, na.rm = TRUE)) -> meanstest

meanstest %>% 
  filter(CHROM != "I", CHROM != "III", CHROM != "M", CHROM != "VIII") %>%
  ggplot(aes(x = POS, y = mean_LR)) + 
  geom_point(aes(color = Parent), size = 0.1) + 
  geom_ribbon(aes(ymin = mean_LR - sd, ymax = mean_LR + sd, fill = Parent), alpha = 0.3) + 
  facet_grid(Parent~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none")
```



Adding this in real quick since I use the cd dataset from above  

```{r}
dilute_combineddata %>% pivot_wider(names_from = Allele, values_from = Reads) %>% 
  mutate(logRatio = abs(log(Wine/Oak))) -> logRatioData

logRatioData %>% filter(POS %in% cd$POS) %>% group_by(CHROM, POS, Parent) %>% summarize(mean_LR = mean(logRatio, na.rm = TRUE),
                                                            sd = sd(logRatio, na.rm = TRUE)) -> justlogratiomeans

justlogratiomeans$CHROM <- factor(justlogratiomeans$CHROM, 
                                   levels = as.factor(c(as.character(as.roman(1:16), "M"))))

justlogratiomeans %>% 
  filter(CHROM != "I", CHROM != "III", CHROM != "M", CHROM != "VIII") %>%
  ggplot(aes(x = POS, y = mean_LR)) + 
  geom_point(aes(color = Parent), size = 0.1) + 
  geom_ribbon(aes(ymin = mean_LR - sd, ymax = mean_LR + sd, fill = Parent), alpha = 0.3) + 
  facet_grid(Parent~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none")

logRatioData %>% group_by(CHROM, POS, Parent) %>% summarize(mean_LR = mean(logRatio, na.rm = TRUE),
                                                            sd = sd(logRatio, na.rm = TRUE)) -> allpoints

allpoints$CHROM <- factor(allpoints$CHROM, 
                                   levels = as.factor(c(as.character(as.roman(1:16), "M"))))
allpoints %>% 
  filter(CHROM != "I", CHROM != "III", CHROM != "M", CHROM != "VIII") %>%
  ggplot(aes(x = POS, y = mean_LR)) + 
  geom_point(aes(color = Parent), size = 0.1, alpha = 0.4) + 
  geom_ribbon(aes(ymin = mean_LR - sd, ymax = mean_LR + sd, fill = Parent), alpha = 0.1) + 
  facet_grid(Parent~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none")
```


Checking this as lines on the same plot  

```{r}
justlogratiomeans %>% 
  filter(CHROM != "I", CHROM != "III", CHROM != "M", CHROM != "VIII") %>%
  ggplot(aes(x = POS, y = mean_LR)) + 
  geom_line(aes(color = Parent)) + 
  geom_ribbon(aes(ymin = mean_LR - sd, ymax = mean_LR + sd, fill = Parent), alpha = 0.3) + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom")

```

## Doing this the way I sent Randi

```{r}
#W is required if using a logistic regression with weights
glm_cb <- function(..., W, formula) {
  data <- list(...)
  
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  
  glm_formula <- as.formula(formula)
  
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #CHANGE THIS TO ADJUST TYPE OF REGRESSION
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  #Output: Effect, Standard Error, Z-score 
  #(ignores p-value since it can be calculated from Z score; multiply by 1 if you would like to retain it)
  return(coefficients(glm_fit))
  
}


#Define the formula based on your specific column names
myformula <- "Allele ~ Parent+Exp+Rep"

b <- glm(myformula, weights = Reads, family = binomial, 
              data = cd)
myrownames <- rownames(as.data.frame(coefficients(b)))


#Run the glm
glm_fit <- glm(formula = as.formula(myformula), 
               data = as.data.frame(cd), 
               weights = Reads, #this should be the column which has your reads or numeric data
               family = binomial)

```

```{r}
cd %>% 
  #subsetting for time; this line can be removed if you want to see the whole genome
  #filter(CHROM %in% c("I")) %>% 
  #this is how you can parallellize with more positions
  group_by(CHROM, POS) %>% 
  #reframe() is the newer version of summarize()
  reframe(GLMResult = glm_cb(formula = myformula, W = Reads,
                               Allele = Allele, Parent = Parent, Exp = Exp, Rep = Rep),
          labels = myrownames) -> glm_fulldataset

```


Looking at the results  

```{r}
glm_fulldataset$CHROM <- factor(glm_fulldataset$CHROM, 
                                   levels = as.factor(c(as.character(as.roman(1:16), "M"))))

glm_fulldataset %>% 
  filter(CHROM != "I", CHROM != "III", CHROM != "M", CHROM != "VIII", labels != "(Intercept)") %>%
  filter(str_detect(labels, 'Exp|Parent')) %>%
  ggplot(aes(x = POS, y = abs(GLMResult), color = labels)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  scale_color_manual(values = c("red", "darkred", "maroon", "orchid",
                                "turquoise", "darkturquoise", "skyblue"))
```

Testing just parent effects  

```{r}
cd %>% 
  #make your rownames first
  filter(CHROM == "I", POS == 64781) -> test

myrownames <- names(coefficients(glm("Allele ~ Parent", 
                                     data = as.data.frame(test), 
                                     weights = Reads, family = binomial)))


cd %>% 
  #this is how you can parallelize with more positions
  group_by(CHROM, POS) %>% 
  reframe(GLMResult = glm_cb(formula = "Allele ~ Parent", 
                             W = Reads,
                               Allele = Allele, Parent = Parent),
          labels = myrownames) -> glm_B

#Plot it
glm_B %>% filter(CHROM != "I", CHROM != "III", CHROM != "M", CHROM != "VIII") %>%
  ggplot(aes(x = POS, y = abs(GLMResult), color = labels)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  scale_color_manual(values = c("gray", "darkred", "darkturquoise", "black"))+
  ggtitle("Effects of each Parent")
```

Testing just replicate effects  

```{r}
cd %>% 
  #make your rownames first
  filter(CHROM == "I", POS == 64781) -> test

myrownames <- names(coefficients(glm("Allele ~ Rep", 
                                     data = as.data.frame(test), 
                                     weights = Reads, family = binomial)))


cd %>% 
  #this is how you can parallelize with more positions
  group_by(CHROM, POS) %>% 
  reframe(GLMResult = glm_cb(formula = "Allele ~ Rep", 
                             W = Reads,
                               Allele = Allele, Rep = Rep),
          labels = myrownames) -> glm_C

#Plot it
glm_C %>% filter(CHROM != "I", CHROM != "III", CHROM != "M", CHROM != "VIII", labels != "(Intercept)") %>%
  ggplot(aes(x = POS, y = abs(GLMResult), color = labels)) + geom_line(size = 1, alpha = 0.5) + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  ggtitle("Effects of each Replicate") +
  ylab("glm(Allele ~ Replicate) Estimate")

 ```

### Should any experiments be thrown out?

Check effects of each experiment, and if one is consistently significant(?) then maybe remove from all other analyses.

```{r}
cd %>% 
  #make your rownames first
  filter(CHROM == "I", POS == 64781) -> test

myrownames <- names(coefficients(glm("Allele ~ Exp", 
                                     data = as.data.frame(test), 
                                     weights = Reads, family = binomial)))


cd %>% 
  #this is how you can parallelize with more positions
  group_by(CHROM, POS) %>% 
  reframe(GLMResult = glm_cb(formula = "Allele ~ Exp", 
                             W = Reads,
                               Allele = Allele, Exp = Exp),
          labels = myrownames) -> glm_D

#Plot it
glm_D %>% filter(CHROM != "I", CHROM != "III", CHROM != "M", CHROM != "VIII", labels != "(Intercept)") %>%
  ggplot(aes(x = POS, y = abs(GLMResult), color = labels)) + geom_line(size = 1) + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  scale_color_manual(values = c("gray", "red", "darkturquoise", "black", "violet"))+
  ggtitle("Effect of each Experiment") +
  ylab("glm(Allele ~ Experiment) Estimate")

```
## What if we looked at the overall read depth of the alleles in common?

```{r}
dilute_combineddata %>% filter(POS %in% cd$POS) %>% group_by(CHROM, POS, Parent, Rep, Exp) %>%
  summarize(readdepth = sum(Reads)) -> depthincommon

depthincommon %>% ggplot(aes(x = POS, y = readdepth, color = paste(Exp,Rep))) + geom_point(alpha = 0.2, size = 0.1) + 
  facet_grid(~CHROM, scales = "free", space = "free")

justlogratiomeans %>% merge(depthincommon) %>%
  filter(CHROM != "I", CHROM != "III", CHROM != "M", CHROM != "VIII") %>%
  ggplot(aes(y = readdepth, x = mean_LR, color = paste(Exp,Rep))) + 
  geom_point(alpha = 0.2, size = 0.1) +
  theme(legend.position = "none")

```

## Using Replicates as Randomized Parent

```{r}
unique(cd$Rep)
#TESTING
cd %>% pivot_wider(names_from = Rep, values_from = Reads) %>%
  #pick which columns here
  select(CHROM, POS, Parent, Exp, Allele, 
         A = AC1B,
         B = BC1B) %>%
  pivot_longer(c(A,B), names_to = "Bulk", values_to = "Reads") %>%
  mutate_if(is.character, as.factor) %>%
  na.omit() %>%
  
  #TESTING
  filter(POS == 64781) -> Test

glm_test <- glm("Allele ~ Bulk*Parent", data = as.data.frame(Test), weights = Reads, family = binomial)
names(coefficients(glm_test))

################################################################################

cd %>% pivot_wider(names_from = Rep, values_from = Reads) %>%
  #pick which columns here
  select(CHROM, POS, Parent, Exp, Allele, 
         A = AC1B,
         B = BC1B) %>%
  pivot_longer(c(A,B), names_to = "Bulk", values_to = "Reads") %>%
  mutate_if(is.character, as.factor) %>%
  na.omit() %>%
  #run the GLM here
  group_by(CHROM, POS) %>% 
  reframe(GLMResult = glm_cb(formula = "Allele ~ Bulk*Parent", 
                             W = Reads,
                               Allele = Allele, Bulk = Bulk, Parent = Parent),
          labels = c("Intercept", "Bulk", "Parent", "Interaction")) -> glm_C1B_rand

cd %>% pivot_wider(names_from = Rep, values_from = Reads) %>%
  #pick which columns here
  select(CHROM, POS, Parent, Exp, Allele, 
         A = AZ1,
         B = BZ1) %>%
  pivot_longer(c(A,B), names_to = "Bulk", values_to = "Reads") %>%
  mutate_if(is.character, as.factor) %>%
  na.omit() %>%
  #run the GLM here
  group_by(CHROM, POS) %>% 
  reframe(GLMResult = glm_cb(formula = "Allele ~ Bulk*Parent", 
                             W = Reads,
                               Allele = Allele, Bulk = Bulk, Parent = Parent),
          labels = c("Intercept", "Bulk", "Parent", "Interaction")) -> glm_Z1_rand


cd %>% pivot_wider(names_from = Rep, values_from = Reads) %>%
  #pick which columns here
  select(CHROM, POS, Parent, Exp, Allele, 
         A = AC8,
         B = BC8) %>%
  pivot_longer(c(A,B), names_to = "Bulk", values_to = "Reads") %>%
  mutate_if(is.character, as.factor) %>%
  na.omit() %>%
  #This one only
  filter(Parent == "O8" | Parent == "W8") %>%
  #run the GLM here
  group_by(CHROM, POS) %>% 
  reframe(GLMResult = glm_cb(formula = "Allele ~ Bulk*Parent", 
                             W = Reads,
                               Allele = Allele, Bulk = Bulk, Parent = Parent),
          labels = c("Intercept", "Bulk", "Parent", "Interaction")) -> glm_C8_rand_chr8

cd %>% pivot_wider(names_from = Rep, values_from = Reads) %>%
  #pick which columns here
  select(CHROM, POS, Parent, Exp, Allele, 
         A = CF1,
         B = BF1) %>%
  pivot_longer(c(A,B), names_to = "Bulk", values_to = "Reads") %>%
  mutate_if(is.character, as.factor) %>%
  na.omit() %>%
  #run the GLM here
  group_by(CHROM, POS) %>% 
  reframe(GLMResult = glm_cb(formula = "Allele ~ Bulk*Parent", 
                             W = Reads,
                               Allele = Allele, Bulk = Bulk, Parent = Parent),
          labels = c("Intercept", "Bulk", "Parent", "Interaction")) -> glm_F1_CB_rand

# cd %>% pivot_wider(names_from = Rep, values_from = Reads) %>%
#   #pick which columns here
#   select(CHROM, POS, Parent, Exp, Allele, 
#          A = CF1,
#          B = DF1) %>%
#   pivot_longer(c(A,B), names_to = "Bulk", values_to = "Reads") %>%
#   mutate_if(is.character, as.factor) %>%
#   na.omit() %>%
#   #run the GLM here
#   group_by(CHROM, POS) %>% 
#   reframe(GLMResult = glm_cb(formula = "Allele ~ Bulk*Parent", 
#                              W = Reads,
#                                Allele = Allele, Bulk = Bulk, Parent = Parent),
#           labels = c("Intercept", "Bulk", "Parent", "Interaction")) -> glm_F1_CD_rand #BROKEN

cd %>% pivot_wider(names_from = Rep, values_from = Reads) %>%
  #pick which columns here
  select(CHROM, POS, Parent, Exp, Allele, 
         A = BF1,
         B = DF1) %>%
  pivot_longer(c(A,B), names_to = "Bulk", values_to = "Reads") %>%
  mutate_if(is.character, as.factor) %>%
  na.omit() %>%
  #run the GLM here
  group_by(CHROM, POS) %>% 
  reframe(GLMResult = glm_cb(formula = "Allele ~ Bulk*Parent", 
                             W = Reads,
                               Allele = Allele, Bulk = Bulk, Parent = Parent),
          labels = c("Intercept", "Bulk", "Parent", "Interaction")) -> glm_F1_BD_rand

```

```{r}
rbind(data.frame(glm_C1B_rand, Exp = "C1B"),
      data.frame(glm_Z1_rand, Exp = "Z1"),
      data.frame(glm_C8_rand_chr8, Exp = "C8"),
      data.frame(glm_F1_CB_rand, Exp = "F1_CB"),
      data.frame(glm_F1_BD_rand, Exp = "F1_BD")) -> RepAsBulk

```


### Bulk Effects

```{r}

RepAsBulk%>%
  filter(CHROM != "I", CHROM != "M", CHROM !="VIII") %>%
  filter(labels == "Bulk") %>%
  ggplot(aes(x = POS, y = GLMResult, color = Exp)) + geom_line() + facet_grid(~CHROM, scale = "free", space = "free") +
  ggtitle("BULK EFFECTS | Replicates as Bulk * Parent Interaction")

RepAsBulk %>% filter(CHROM != "I", CHROM != "M", CHROM !="VIII") %>%
  filter(labels == "Bulk") %>%
  group_by(Exp) %>%
  summarize(Lower = quantile(GLMResult, 0.05, na.rm = TRUE),
            Upper = quantile(GLMResult, 0.95, na.rm = TRUE),
            Abs = quantile(abs(GLMResult), 0.975, na.rm = TRUE)) %>%
  na.omit() -> RepAsBulk_Quant

RepAsBulk %>% filter(CHROM != "I", CHROM != "M", CHROM !="VIII") %>%
  na.omit() %>%
  filter(labels == "Bulk") %>%
  ggplot(aes(x = abs(GLMResult), color = Exp)) + 
  geom_density(size = 1.2) +
  geom_vline(data = RepAsBulk_Quant, aes(xintercept = Abs, color = Exp), linetype = "dashed", size = 1.2) +
  ggtitle("BULK EFFECTS | Replicates as Bulk * Parent Interaction")

```

### Interaction Effects

```{r}
RepAsBulk%>%
  filter(CHROM != "I", CHROM != "M", CHROM !="VIII") %>%
  filter(labels == "Interaction") %>%
  ggplot(aes(x = POS, y = GLMResult, color = Exp)) + geom_line() + facet_grid(~CHROM, scale = "free", space = "free") +
  ggtitle("INTERACTION EFFECTS | Replicates as Bulk * Parent Interaction")

RepAsBulk %>% filter(CHROM != "I", CHROM != "M", CHROM !="VIII") %>%
  filter(labels == "Interaction") %>%
  group_by(Exp) %>%
  summarize(Lower = quantile(GLMResult, 0.05, na.rm = TRUE),
            Upper = quantile(GLMResult, 0.95, na.rm = TRUE),
            Abs = quantile(abs(GLMResult), 0.975, na.rm = TRUE)) %>%
  na.omit() -> RepAsBulk_Quant

RepAsBulk %>% filter(CHROM != "I", CHROM != "M", CHROM !="VIII") %>%
  na.omit() %>%
  filter(labels == "Interaction") %>%
  ggplot(aes(x = abs(GLMResult), color = Exp)) + 
  geom_density(size = 1.2) +
  geom_vline(data = RepAsBulk_Quant, aes(xintercept = Abs, color = Exp), linetype = "dashed", size = 1.2) +
  ggtitle("INTERACTION EFFECTS | Replicates as Bulk * Parent Interaction")

```

#### Calculate the FDR for each cutoff from 0 to 0.6

```{r}
RepAsBulk %>% filter(CHROM != "I", CHROM != "M", CHROM !="VIII") %>%
  group_by(Exp, labels) %>%
  reframe(Cutoff_Abs = quantile(abs(GLMResult), seq(0, 1, length.out = 1001), na.rm = TRUE),
            Quant = seq(0, 1, length.out = 1001)) %>%
  na.omit() -> RepAsBulk_FDR

  #Trying new code for slicing
RepAsBulk_FDR %>%
  group_by(Exp, labels) %>% filter(labels != "Intercept", labels != "Parent") %>% filter(Quant >= 0.95) %>% slice(which.min(Cutoff_Abs)) -> Cutoff_95
RepAsBulk_FDR %>%
  group_by(Exp, labels) %>% filter(labels != "Intercept", labels != "Parent") %>% filter(Quant >= 0.99) %>% slice(which.min(Cutoff_Abs)) -> Cutoff_99

RepAsBulk_FDR %>% filter(labels != "Intercept", labels != "Parent") %>%
  ggplot(aes(x = Cutoff_Abs , y = Quant, color = Exp, linetype = labels)) + geom_line(size= 1.2, alpha = 0.2) +
  geom_hline(yintercept = c(0.95, 0.99)) +
  geom_vline(data = Cutoff_95, aes(xintercept = Cutoff_Abs, color = Exp, linetype = labels), size = 1.2)+
  ggtitle("5% False Discovery on Replicates as Bulk Effect") +
  ylim(0.8, 1)

RepAsBulk_FDR %>% filter(labels != "Intercept", labels != "Parent") %>%
  ggplot(aes(x = Cutoff_Abs , y = Quant, color = Exp, linetype = labels)) + geom_line(size= 1.2, alpha = 0.2) +
  geom_hline(yintercept = c(0.95, 0.99)) +
  geom_vline(data = Cutoff_99, aes(xintercept = Cutoff_Abs, color = Exp, linetype = labels), size = 1.2)+
  ggtitle("1% False Discovery on Replicates as Bulk Effect") +
  ylim(0.8, 1)
```

