---
title: "Figures for 2023 PopGen Poster"
author: "Cassandra Buzby"
date: "1/24/2023"
output:
  html_notebook:
    code_folding: hide
---


```{r}
#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

library(cybrBSA)

ggplot2::theme_set(theme_light() + theme(text = element_text(size = 20)))

glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

cybr2_rollmean <- function(dataframe){
  dataframe %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollmean(value, n = 100))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount)
}



```

## Load in Data

```{r}
#Load Stuff
CSSI_CuSO4 <- readRDS("RdataFiles/CSSI_CuSO4_cybr2.rds")
CSSI_Fluc <- readRDS("RdataFiles/CSSI_Fluc_cybr2.rds")
CSSI_Zeocin <- readRDS("RdataFiles/CSSI_Zeocin_cybr2.rds")
```

## Non-Smoothed Data

```{r}
CSSI_CuSO4 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = value) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_CuSO4_unsmoothed

CSSI_Fluc %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = value) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_Fluc_unsmoothed

CSSI_Zeocin %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = value) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_Zeocin_unsmoothed

CSSI_Zeocin_unsmoothed %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CSSI_Zeocin_LODu


CSSI_Fluc_unsmoothed %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CSSI_Fluc_LODu

CSSI_CuSO4_unsmoothed %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent","Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) %>% 
  transmute(CHROM = CHROM, POS = POS, Bulk = Bulk, Parent = Parent, Rep = NA, Oak = Oak, Wine = Wine, LOD = LOD) -> CSSI_CuSO4_LODu


rbind(data.frame(CSSI_CuSO4_LODu, Dataset = "CuSO4"),
data.frame(CSSI_Fluc_LODu, Dataset = "Fluconazole"),
data.frame(CSSI_Zeocin_LODu, Dataset = "Zeocin")) -> AllLODs_unsmoothed

AllLODs_unsmoothed %>% filter(CHROM == "VIII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  #filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_point(alpha = 0.2) + 
  facet_grid(~Dataset, scales = "free", space = "free") + 
  scale_color_manual(values = c("blue", "black", "firebrick", "skyblue")) + 
  ggtitle("")
```

## Smooth Data

```{r}

CSSI_CuSO4 %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollapply(value, n = 200, FUN = mean, align = "center"))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_CuSO4_cybr2

CSSI_Fluc %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollapply(value, n = 200, FUN = mean, align = "center"))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_Fluc_cybr2

CSSI_Zeocin %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>%
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollapply(value, n = 200, FUN = mean, align = "center"))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount) -> CSSI_Zeocin_cybr2

```

```{r}
CSSI_Zeocin_cybr2 %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CSSI_Zeocin_LOD


CSSI_Fluc_cybr2 %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CSSI_Fluc_LOD

CSSI_CuSO4_cybr2 %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent","Allele"),
    names_sep = "_",
  values_to = "Reads") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) %>% transmute(CHROM = CHROM, POS = POS, Bulk = Bulk, Parent = Parent, Rep = NA, Oak = Oak, Wine = Wine, LOD = LOD) -> CSSI_CuSO4_LOD

rbind(data.frame(CSSI_CuSO4_LOD, Dataset = "CuSO4"),
data.frame(CSSI_Fluc_LOD, Dataset = "Fluconazole"),
data.frame(CSSI_Zeocin_LOD, Dataset = "Zeocin")) -> AllLODs

```


Checking out XIV

```{r}
AllLODs %>% filter(CHROM == "XIV", Dataset == "CuSO4") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  ggplot(aes(x = POS, y = LOD, color = Flask)) + 
  geom_line(size = 2) +
      ylab("") + xlab("")+
  scale_color_manual(values = c("gray","black","lightblue", "firebrick")) + 
  theme(legend.position = "bottom") +
  ggtitle("CuSO4 Chr XIV")
```

```{r, fig.width=7, fig.height=6}

AllLODs_unsmoothed %>% filter(CHROM == "VII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  #filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_point(alpha = 0.2, size = 0.8) + ylim(-2.5,2.5)+
      ylab("") + xlab("")+

  facet_grid(~Dataset, scales = "free", space = "free") + 
  scale_color_manual(values = c("#345F6F","black","firebrick", "#FFB05C")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("")

AllLODs %>% filter(CHROM == "VII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  #filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_point(alpha = 0.2, size = 0.8) + ylim(-2.5,2.5)+
      ylab("") + xlab("")+
  facet_grid(~Dataset, scales = "free", space = "free") + 
  scale_color_manual(values = c("black","firebrick","#345F6F", "#FFB05C")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("")

```

```{r, fig.width=16,fig.height=5}

AllLODs %>% filter(CHROM != "I", CHROM != "M") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  ggplot(aes(x = POS, y = LOD, color = Bulk, group = Flask, linetype = Parent)) + 
  geom_line(size = 1.5) + 
      ylab("") + xlab("")+
  facet_grid(~CHROM, scales = "free", space = "free") + 
  scale_color_manual(values =  c("black","firebrick","#345F6F", "#FFB05C")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("")
```


## GLM: Load Data

```{r}
Zeocin_glm<- readRDS("2023/CSSI_Zeocin_glm.rds")
Zeocinp <- readRDS("2023/CSSI_Zeocin_cybr2_permuted.rds")
CuSO4p <- readRDS("2023/CSSI_CuSO4_cybr2_permuted.rds")
CuSO4_glm <- readRDS("2023/CSSI_CuSO4_glm.rds")
Flucp <- readRDS("2023/CSSI_Fluc_cybr2_permuted.rds")
Fluc_glm <- readRDS("2023/CSSI_Fluc_glm.rds")
```

## Find quantiles with of permutations

Merge Quantiles

```{r}
#CuSO4p %>% filter(CHROM == )
```

```{r}
CuSO4p %>% group_by(CHROM, label) %>% summarize(quant = quantile(summary, 0.975, na.rm = TRUE)) -> CuSO4q
Flucp %>% group_by(CHROM, label) %>% summarize(quant = quantile(summary, 0.975, na.rm = TRUE)) -> Flucq
Zeocinp %>% group_by(CHROM, label) %>% summarize(quant = quantile(summary, 0.975, na.rm = TRUE)) -> Zeocinq

```

## Plot with these cutoffs

```{r, fig.width=16,fig.height=6}
Zeocin_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") %>%
  ggplot(aes(x = POS, y = abs(summary), color = label, linetype = label)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("Zeocin Z-Scores") + 
  scale_color_manual(values = c("black", "#FFB05C","gray", "gray")) + 
  scale_linetype_manual(values = c("solid", "solid", "solid", "dotted")) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 
  

Fluc_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("Fluconazole Z-Scores") + 
  scale_color_manual(values = c("black", "firebrick","gray", "gray")) + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 
  

CuSO4_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") %>%
  ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") + ggtitle("CuSO4 Z-Scores") + 
  scale_color_manual(values = c("black", "#345F6F","gray", "#FFB05C")) + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 
  

```

## Using Permuted Cutoffs

This was done with 100,000 samples with several chromosomes missing in each

```{r}
CuP <- readRDS("2023/CSSI_CuSO4_cybr2_permutedWG.rds")
ZeoP <- readRDS("2023/CSSI_Zeocin_cybr2_permutedWG.rds")
FluP <- readRDS("2023/CSSI_Fluc_cybr2_permutedWG.rds")

rbind(data.frame(CuP, Dataset = "CuSO4"),
      data.frame(ZeoP, Dataset = "Zeocin"),
      data.frame(FluP, Dataset = "Fluconazole")) %>%
  filter(label != "intercept") %>%
  group_by(Dataset, label) %>%
  summarize(Quant = quantile(abs(summary), 0.975)) -> AllQuants

rbind(data.frame(CuP, Dataset = "CuSO4"),
      data.frame(ZeoP, Dataset = "Zeocin"),
      data.frame(FluP, Dataset = "Fluconazole")) %>% 
  filter(label != "intercept") %>% ggplot(aes(x = abs(summary), color = label)) + geom_density(size = 2) + 
  geom_vline(data = AllQuants, aes(xintercept = Quant, color = label), linetype= "dashed", size = 2) +
  facet_grid(rows = "Dataset") + ggtitle("Distribution of Permuted Scores")
```

```{r, fig.width=17,fig.height=5}

Zeocin_glm$label <- factor(Zeocin_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
Fluc_glm$label <- factor(Fluc_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))
CuSO4_glm$label <- factor(CuSO4_glm$label, levels = c("Rep", "Parent", "Bulk", "Interaction"))

Zeocin_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Zeocin_glm
Fluc_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> Fluc_glm
CuSO4_glm %>% filter(CHROM != "I", CHROM != "M", label != "intercept") -> CuSO4_glm

################################################################################
#Peaks
max(abs(Zeocin_glm$summary[Zeocin_glm$CHROM == "III"]))
Zeocin_glm %>% filter(label == "Interaction") %>% filter(abs(summary) > 0.7) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> ZeocinPeaks

for(i in 1:length(ZeocinPeaks$peak)){
  ZeocinPeaks$POS[i] <- Zeocin_glm$POS[Zeocin_glm$label == "Interaction" & abs(Zeocin_glm$summary) == ZeocinPeaks$peak[i] & Zeocin_glm$CHROM == ZeocinPeaks$CHROM[i]]

}

max(abs(Fluc_glm$summary[Fluc_glm$CHROM == "III"]))
Fluc_glm %>% filter(label == "Interaction") %>% filter(abs(summary) > 1.124545) %>% group_by(CHROM) %>% summarize(peak = max(abs(summary))) -> FlucPeaks

for(i in 1:length(FlucPeaks$peak)){
  FlucPeaks$POS[i] <- Fluc_glm$POS[Fluc_glm$label == "Interaction" & abs(Fluc_glm$summary) == FlucPeaks$peak[i] & Fluc_glm$CHROM == FlucPeaks$CHROM[i]]

}

################################################################################
#Plots
Zeocin_glm %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label, linetype = label)) +   
  geom_vline(data = ZeocinPeaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 1.5) + 
  geom_hline(aes(yintercept = max(abs(Zeocin_glm$summary[Zeocin_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free") + 
  scale_color_manual(values = c("gray", "gray", "black","#FFB05C")) + 
  scale_linetype_manual(values = c("dotted", "solid", "solid", "solid")) +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 
  

Fluc_glm %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
  geom_vline(data = FlucPeaks, aes(xintercept = POS), color = "#BCEAF3", size = 2) +
  geom_line(size = 1.5) + 
  geom_hline(aes(yintercept = max(abs(Fluc_glm$summary[Fluc_glm$CHROM == "III"]))), linetype = "dashed") +

    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("gray", "black","firebrick")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 

CuSO4_glm %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line(size = 1.5) + 
    geom_hline(aes(yintercept = max(abs(CuSO4_glm$summary[CuSO4_glm$CHROM == "III"]))), linetype = "dashed") +
    ylab("") + xlab("")+ ggtitle("")+
  facet_grid(~CHROM, scales = "free", space = "free")  + 
  scale_color_manual(values = c("gray", "black","#345F6F")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) 
```

## Combining GLM and LOD scores

```{r, fig.width=5,fig.height=4}

AllLODs %>% filter(CHROM == "VII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "Fluconazole") %>%
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) + 
  ylab("") + xlab("")+geom_line(size = 2) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("") -> chr7_glm

AllLODs$Bulk[AllLODs$Bulk == "CuSO4"] <- "Selected"
AllLODs %>% filter(CHROM == "VIII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "CuSO4") %>%
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) + 
  ylab("") + xlab("")+geom_line(size = 2) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("") -> chr8_glm

CuSO4_glm %>% filter(CHROM == "VIII") %>%
  ggplot(aes(x = POS, y = -1*(summary), color = label)) + geom_line(size = 2) + 
  ylab("") + xlab("") + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  geom_vline(aes(xintercept = 212535), color = "#00ADD0", alpha = 0.3, size = 2) +
  scale_color_manual(values = c("gray", "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ylim(-12, 2.5) +
  ggtitle("") -> chr8_lod

Fluc_glm %>% filter(CHROM == "VII") %>%
  ggplot(aes(x = POS, y = -1*(summary), color = label)) + geom_line(size = 2) + 
  ylab("") + xlab("")+geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("gray", "black","#7030A0")) + 
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ggtitle("") -> chr7_lod

plot_grid(chr7_glm, chr7_lod)
plot_grid(chr8_glm, chr8_lod)
```

## Correctly sizing the figures

```{r, fig.width=10,fig.height=8}

AllLODs %>% filter(CHROM == "VII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "Fluconazole") %>%
  
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) + 
  geom_line(size = 2) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
    scale_alpha_manual(values = c(0.5, 1)) +
  theme(legend.position = "none", axis.text.x=element_blank(),axis.ticks.x=element_blank(),
        text = element_text(size = 20)) +
  ylab("") + xlab("")+ggtitle("") -> chr7_glm

AllLODs$Bulk[AllLODs$Bulk == "CuSO4"] <- "Selected"
AllLODs %>% filter(CHROM == "VIII") %>% 
  mutate(Flask = paste(Bulk, Parent, Rep, Dataset, sep = "_")) %>% 
  filter(Dataset == "CuSO4") %>%
  
  ggplot(aes(x = POS, y = LOD, color = Parent, group = Flask, alpha = Bulk)) + 
  geom_line(size = 2) + geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("firebrick", "#5391C4")) + 
  scale_alpha_manual(values = c(0.5, 1)) +
  theme(legend.position = "none", axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  ylab("") + xlab("")+ ggtitle("") -> chr8_glm

CuSO4_glm %>% filter(CHROM == "VIII") %>%
  ggplot(aes(x = POS, y = -1*(summary), color = label)) + geom_line(size = 2) + 
  geom_hline(aes(yintercept = 0), linetype = "dashed") +
  geom_vline(aes(xintercept = 212535), color = "#00ADD0", alpha = 0.3, size = 2) +
  scale_color_manual(values = c("gray", "black","#7030A0")) + 
  ylab("") + xlab("") + ggtitle("") +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +
  ylim(-12, 2.5) +
  ggtitle("") -> chr8_lod

Fluc_glm %>% filter(CHROM == "VII") %>%
  ggplot(aes(x = POS, y = -1*(summary), color = label)) + geom_line(size = 2) + 
  geom_hline(aes(yintercept = 0), linetype = "dashed") +
  scale_color_manual(values = c("gray", "black","#7030A0")) + 
  ylab("") + xlab("") + ggtitle("") +
  theme(legend.position = "none", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) -> chr7_lod

plot_grid(chr7_glm, chr7_lod)
plot_grid(chr8_glm, chr8_lod)
```

