---
title: "GLM_ComparingBulks"
author: "Cassandra Buzby"
date: "7/19/2022"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
library(ggplot2)
library(tidyr)
#library(tidyverse)
library(reshape2)
library(cowplot)
library(dplyr)
library(foreach)
library(doParallel)


#ggplot2::theme_set(theme_light())
ggplot2::theme_set(theme_cowplot())


CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47",
                  "#F26430", "#0A369D", "#7EA3CC")

CBchromPalette_ext <- c("#F26430", "#0B0B45", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#1E64AA", "#32936F", "#FEC601","#8DAA9D", "#7B0828","#0F0E0E", "#522B47")
                  

library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")
load("SAUA_4-13-22.Rdata")
load("SCUC_4-13-22.Rdata")


#################################################################################################
#################################################################################################
BSA_GLM <-  function(lrP, chr = "II", windowsize = 100){
  
  lrP <- subset(lrP, CHROM == chr)
  
  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- glm(reads ~ bulk*parent, weights = value, family = binomial, data = lrP[lrP$POS == i,])
      c(c, i, res$coefficients)
    }
  
    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh", "parentWine", "Interaction")
    Results[,2] <- as.numeric(Results[,2])
    Results[,3] <- as.numeric(Results[,3])
    Results[,4] <- as.numeric(Results[,4])
    Results[,5] <- as.numeric(Results[,5])
    Results[,6] <- as.numeric(Results[,6])
    
    Results %>% arrange(POS) %>% select(-Intercept) -> Results
  }
  
 
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  return(WResult)
  }

################################################################################
perChromWindow_binom_redone <- function(lrP, chr = "II", windowsize = 100, coverage = 15634092/2, seedt = Sys.time()){
  lrP <- lrP[lrP$CHROM == chr,]
  
  set.seed(seedt)
  
  for(i in 1:length(lrP$CHROM)){
    lrP$subs[i] <- rbinom(1, coverage, prob = lrP$perc[i]) + 1
  }
  
  #Run the glm on that chromosome
  Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
    res <- glm(reads ~ bulk*parent, weights = subs, family = binomial, data = lrP[lrP$POS == i,])
    c(chr, i, res$coefficients)
  }
  
  #Format the results for the next function
  Results <- as.data.frame(Results)
  colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh", "parentWine", "Interaction")
  Results[,2] <- as.numeric(Results[,2])
  Results[,3] <- as.numeric(Results[,3])
  Results[,4] <- as.numeric(Results[,4])
  Results[,5] <- as.numeric(Results[,5])
  Results[,6] <- as.numeric(Results[,6])
  
  Results %>% arrange(POS) %>% select(-Intercept) -> Results

  
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  return(WResult)
}

#################################################################################################
#################################################################################################


pois_binom_sample <- function(lrP, chr = "II", windowsize = 100, coverage_div_factor = 2, seedt = 0){
  lrP <- lrP[lrP$CHROM == chr,]
  
  set.seed(seedt + Sys.time())
  
  #calculate reads by Poisson
   reads <- rpois(n = length(unique(lrP$id)), lambda = mean(lrP$value)/coverage_div_factor)
   readsdf <- data.frame(id = unique(lrP$id), numreads = reads)
   lrP %>% left_join(readsdf, by = "id") -> lrP
  
  #calculate binom for number of reads at position
  for(i in 1:length(lrP$CHROM)){
    lrP$subs[i] <- rbinom(1, lrP$numreads[i], prob = lrP$perc[i]) + 1
  }
  #Run the glm on that chromosome
  Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
    res <- glm(reads ~ bulk*parent, weights = subs, family = binomial, data = lrP[lrP$POS == i,])
    c(chr, i, res$coefficients)
  }
  
  #Format the results for the next function
  Results <- as.data.frame(Results)
  colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh", "parentWine", "Interaction")
  Results[,2] <- as.numeric(Results[,2])
  Results[,3] <- as.numeric(Results[,3])
  Results[,4] <- as.numeric(Results[,4])
  Results[,5] <- as.numeric(Results[,5])
  Results[,6] <- as.numeric(Results[,6])
  
  Results %>% arrange(POS) %>% select(-Intercept) -> Results

  
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  return(WResult)
}

#################################################################################################
#################################################################################################

MakeStackedPlots <- function(pb = pb_VIII_df, chrom = "VII", bar = 100000){
  
  maxChrVII <- data.frame(Index = unique(pb$Index),
                       max_bulk_zprime = NA,
                       max_parent_zprime = NA,
                       max_int_zprime = NA)

for(i in 1:length(unique(pb$Index))){
  isubset <- subset(pb, Index == as.character(i))
  maxChrVII$max_bulk_zprime[i] <- isubset$POS[abs(isubset$bulk_Zprime) == max(abs(isubset$bulk_Zprime))]
  maxChrVII$max_parent_zprime[i] <- isubset$POS[abs(isubset$parent_Zprime) == max(abs(isubset$parent_Zprime))]
  maxChrVII$max_int_zprime[i] <- isubset$POS[abs(isubset$Interaction_Zprime) == max(abs(isubset$Interaction_Zprime))]
  rm(isubset)
}

#sd(maxChrVII$max_bulk_zprime)

ggplot(maxChrVII, aes(x = max_bulk_zprime)) + geom_density(size = 1) + 
  ggtitle(paste("Maximum Z' value for Chr ",chrom, " Bulk Effect"))-> p1
ggplot(maxChrVII, aes(x = max_bulk_zprime, y = "Bulk    .")) + geom_boxplot(size = 1, color = "violet") + geom_jitter(alpha = 0.4, size = 3)  + ggtitle(paste("SD = ", round(sd(maxChrVII$max_bulk_zprime), 2))) -> p2
a1 <- plot_grid(p1, p2, nrow = 2)

ggplot(maxChrVII, aes(x = max_bulk_zprime)) + geom_density(size = 1) + 
  ggtitle(paste("Maximum Z' value for Chr ",chrom, " Bulk Effect")) -> p1
  p1 + annotate("segment", x = min(maxChrVII$max_bulk_zprime), 
           xend = min(maxChrVII$max_bulk_zprime) + bar, 
           y = layer_scales(p1)$y$range$range[1] + (layer_scales(p1)$y$range$range[2] - layer_scales(p1)$y$range$range[1])/10, 
           yend = layer_scales(p1)$y$range$range[1] + (layer_scales(p1)$y$range$range[2] - layer_scales(p1)$y$range$range[1])/10,  
           colour = "red", size = 2, alpha = 0.5) -> p1
ggplot(maxChrVII, aes(x = max_bulk_zprime, y = "Bulk    .")) + geom_boxplot(size = 1, color = "violet") + geom_jitter(alpha = 0.4, size = 3)+xlim(layer_scales(p1)$x$range$range) + ggtitle(paste("SD = ", round(sd(maxChrVII$max_bulk_zprime), 2))) -> p2
a2 <- plot_grid(p1, p2, nrow = 2)

ggplot(maxChrVII, aes(x = max_parent_zprime)) + geom_density(size = 1) + 
  ggtitle(paste("Maximum Z' value for Chr ",chrom, " Parent Effect"))-> p3
  p3 + annotate("segment", x = min(maxChrVII$max_parent_zprime), 
           xend = min(maxChrVII$max_parent_zprime) + bar, 
           y = layer_scales(p3)$y$range$range[1] + (layer_scales(p3)$y$range$range[2] - layer_scales(p3)$y$range$range[1])/10, 
           yend = layer_scales(p3)$y$range$range[1] + (layer_scales(p3)$y$range$range[2] - layer_scales(p3)$y$range$range[1])/10,  
           colour = "red", size = 2, alpha = 0.5) -> p3
ggplot(maxChrVII, aes(x = max_parent_zprime, y = "Parent .")) + geom_boxplot(size = 1, color = "darkorange") + geom_jitter(alpha = 0.4, size = 3) + ggtitle(paste("SD = ", round(sd(maxChrVII$max_parent_zprime), 2)))  -> p4
a3 <- plot_grid(p3, p4, nrow = 2)

ggplot(maxChrVII, aes(x = max_int_zprime)) + geom_density(size = 1) + ggtitle(paste("Maximum Z' value for Chr ",chrom, " Interaction Effect"))-> p5
  p5 + annotate("segment", x = min(maxChrVII$max_int_zprime), 
           xend = min(maxChrVII$max_int_zprime) + bar, 
           y = layer_scales(p5)$y$range$range[1] + (layer_scales(p5)$y$range$range[2] - layer_scales(p5)$y$range$range[1])/10, 
           yend = layer_scales(p5)$y$range$range[1] + (layer_scales(p5)$y$range$range[2] - layer_scales(p5)$y$range$range[1])/10,  
           colour = "red", size = 2, alpha = 0.5) -> p5
  
ggplot(maxChrVII, aes(x = max_int_zprime, y = "Max Int Z'")) + geom_boxplot(size = 1, color = "lightblue") + geom_jitter(alpha = 0.4, size = 3) +ylab("") + ggtitle(paste("SD = ", round(sd(maxChrVII$max_int_zprime), 2)))  -> p6
a4 <- plot_grid(p5, p6, nrow = 2)

return(list(a1, a2, a3, a4))

}

#################################################################################################
#################################################################################################

MakeZePlots <- function(pb_II_df = pb_VIII_df){
  maxChrVI <- data.frame(Index = unique(pb_II_df$Index),
                       max_bulk_zprime = NA,
                       max_parent_zprime = NA,
                       max_int_zprime = NA)

for(i in 1:length(unique(pb_II_df$Index))){
  isubset <- subset(pb_II_df, Index == as.character(i))
  maxChrVI$max_bulk_zprime[i] <- isubset$POS[abs(isubset$bulk_Zprime) == max(abs(isubset$bulk_Zprime))]
  maxChrVI$max_parent_zprime[i] <- isubset$POS[abs(isubset$parent_Zprime) == max(abs(isubset$parent_Zprime))]
  maxChrVI$max_int_zprime[i] <- isubset$POS[abs(isubset$Interaction_Zprime) == max(abs(isubset$Interaction_Zprime))]
  rm(isubset)
}


pb_II_df %>% ggplot(., aes(x = POS, y = bulk_Zprime, color = Index)) + 
  geom_line(alpha = 0.2) + theme(legend.position = "none") -> pbulkz


pb_II_df %>% ggplot(., aes(x = POS, y = parent_Zprime, color = Index)) + 
  geom_line(alpha = 0.2) + theme(legend.position = "none") -> pparentz

pb_II_df %>% ggplot(., aes(x = POS, y = Interaction_Zprime, color = Index)) + 
  geom_line(alpha = 0.2) + theme(legend.position = "none") -> pintz

ggplot(maxChrVI, aes(x = max_bulk_zprime, y = 1)) + geom_boxplot() + geom_jitter(alpha = 0.5) + theme(legend.position = "none") + ylab("") + xlab("Position") + xlim(min(pb_II_df$POS), max(pb_II_df$POS)) -> pbulkmax

ggplot(maxChrVI, aes(x = max_parent_zprime, y = 1)) + geom_boxplot() + geom_jitter(alpha = 0.5) + theme(legend.position = "none") + ylab("") + xlab("Position") + xlim(min(pb_II_df$POS), max(pb_II_df$POS)) -> pparentmax

ggplot(maxChrVI, aes(x = max_int_zprime, y = 1)) + geom_boxplot() + geom_jitter(alpha = 0.5) + theme(legend.position = "none") + ylab("") + xlab("Position") + xlim(min(pb_II_df$POS), max(pb_II_df$POS)) -> pintmax

a <- plot_grid(pbulkz, pbulkmax, nrow = 2)
b <- plot_grid(pparentz, pparentmax, nrow = 2)
c <- plot_grid(pintz, pintmax, nrow = 2)

 return(list(plotA = a, plotB = b, plotC = c))
}

#################################################################################################
#################################################################################################

getmaxvalues <- function(pb = pb_VIII_df200, idstring = NA){
  #change this for each dataset
  
  #make the df to save
  maxChrVII <- data.frame(Index = unique(pb$Index),
                         max_bulk_zprime = NA,
                         max_parent_zprime = NA,
                         max_int_zprime = NA,
                         max_bulk_zprime_pos = NA,
                         max_parent_zprime_pos = NA,
                         max_int_zprime_pos = NA)
  
  for(i in 1:length(unique(pb$Index))){
    isubset <- subset(pb, Index == as.character(i))
    maxChrVII$max_bulk_zprime_pos[i] <- isubset$POS[abs(isubset$bulk_Zprime) == max(abs(isubset$bulk_Zprime))]
    maxChrVII$max_parent_zprime_pos[i] <- isubset$POS[abs(isubset$parent_Zprime) == max(abs(isubset$parent_Zprime))]
    maxChrVII$max_int_zprime_pos[i] <- isubset$POS[abs(isubset$Interaction_Zprime) == max(abs(isubset$Interaction_Zprime))]
    maxChrVII$max_bulk_zprime[i] <- max(abs(isubset$bulk_Zprime))
    maxChrVII$max_parent_zprime[i] <- max(abs(isubset$parent_Zprime))
    maxChrVII$max_int_zprime[i] <- max(abs(isubset$Interaction_Zprime))
    rm(isubset)
  }
  
  plot_int <- maxChrVII %>% ggplot(aes(x = max_int_zprime_pos, y = max_int_zprime)) + geom_point(size = 2, alpha = 0.4) + ggtitle("Max Interaction Z primes by Position")
  plot_parent <- maxChrVII %>% ggplot(aes(x = max_parent_zprime_pos, y = max_parent_zprime)) + geom_point(size = 2, alpha = 0.4) + ggtitle("Max Parent Z primes by Position")
  plot_bulk <- maxChrVII %>% ggplot(aes(x = max_bulk_zprime_pos, y = max_bulk_zprime)) + geom_point(size = 2, alpha = 0.4) + ggtitle("Max Bulk Z primes by Position")
  
  maxChrVII$Chr <- idstring
  maxdf <- maxChrVII
  
  maxstats <- data.frame(parentz_mean = mean(maxdf$max_parent_zprime),
                         parentz_sd = sd(maxdf$max_parent_zprime),
                         parentpos_mean = mean(maxdf$max_parent_zprime_pos),
                         parentpos_sd = sd(maxdf$max_parent_zprime_pos),
                         intz_mean = mean(maxdf$max_int_zprime),
                         intz_sd = sd(maxdf$max_int_zprime),
                         intpos_mean = mean(maxdf$max_int_zprime_pos),
                         intpos_sd = sd(maxdf$max_int_zprime_pos),
                         bulkz_mean = mean(maxdf$max_bulk_zprime),
                         bulkz_sd = sd(maxdf$max_bulk_zprime),
                         bulkpos_mean = mean(maxdf$max_bulk_zprime_pos),
                         bulkpos_sd = sd(maxdf$max_bulk_zprime_pos),
                         Chr = idstring
                         )
  
  return(list(maxChrVII, maxstats, plot_bulk,plot_parent, plot_int))
}

#################################################################################################
#################################################################################################


GLM_Bulk_Only <- function(startLRP, windowsize = 100, Chr = "II"){
  if(Chr == "All"){
    appendWResult <- data.frame(CHROM = NA, Index = NA, POS = NA, bulk_Z = NA, bulk_Zprime = NA)
    for(Chrall in unique(startLRP$CHROM)){
      lrP <- subset(startLRP, CHROM == Chrall)
      #Run the glm on that chromosome
      Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
        res <- glm(reads ~ bulk, weights = value, family = binomial, data = lrP[lrP$POS == i,])
        c(Chrall, i, res$coefficients)
      }
      
      #Format the results for the next function
      Results <- as.data.frame(Results)
      colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh")
      Results[,2] <- as.numeric(Results[,2])
      Results[,3] <- as.numeric(Results[,3])
      Results[,4] <- as.numeric(Results[,4])
      
      
      Results %>% arrange(POS) %>% select(-Intercept) -> Results
    
      
      #Run the windows on that chromosome
      WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
                    res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                    #print CHROM, index, POS, and median
                    c(Chrall, i, Results$POS[i], Results$bulkHigh[i],res_bulk)}
      
      WResult <- as.data.frame(WResult)
      colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "bulk_Zprime")
      WResult[,2] <- as.numeric(WResult[,2])
      WResult[,3] <- as.numeric(WResult[,3])
      WResult[,4] <- as.numeric(WResult[,4])
      WResult[,5] <- as.numeric(WResult[,5])
      
      appendWResult <- rbind(appendWResult,WResult)
    }
    return(appendWResult)
  }else{
  lrP <- subset(lrP, CHROM == Chr)
  #Run the glm on that chromosome
  Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
    res <- glm(reads ~ bulk, weights = value, family = binomial, data = lrP[lrP$POS == i,])
    c(Chr, i, res$coefficients)
  }
  
  #Format the results for the next function
  Results <- as.data.frame(Results)
  colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh")
  Results[,2] <- as.numeric(Results[,2])
  Results[,3] <- as.numeric(Results[,3])
  Results[,4] <- as.numeric(Results[,4])
  
  
  Results %>% arrange(POS) %>% select(-Intercept) -> Results

  
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(Chr, i, Results$POS[i], Results$bulkHigh[i],res_bulk)}
  
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "bulk_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  return(WResult)
  }
  
}

load("logregParental.Rdata")
```

### Chromosome VIII

Running Chr VIII again for real

```{r}
load("logregParental.Rdata")

VIIIResults <- BSA_GLM(lrP = logregParental, chr = "VIII")


```

## New Function for Fixed Window

```{r}
BSA_GLM_fixed <-  function(lrP, chr = "II", windowsize = 100, a = 0){
  
  lrP <- subset(lrP, CHROM == chr)
  
  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- glm(reads ~ bulk*parent, weights = value, family = binomial, data = lrP[lrP$POS == i,])
      c(c, i, res$coefficients)
    }
  
    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh", "parentWine", "Interaction")
    Results[,2] <- as.numeric(Results[,2])
    Results[,3] <- as.numeric(Results[,3])
    Results[,4] <- as.numeric(Results[,4])
    Results[,5] <- as.numeric(Results[,5])
    Results[,6] <- as.numeric(Results[,6])
    
    Results %>% arrange(POS) %>% select(-Intercept) -> Results
  }
  
 
  #Run the windows on that chromosome
  WResult <- foreach(i = seq(from = (windowsize+1+a), to = (length(Results$POS) - (2*windowsize)),by = windowsize),   .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  return(WResult)
  }
```

Testing this out

```{r}
VIII_Fixed <- BSA_GLM_fixed(logregParental, chr = "VIII")
#VIIIResults <- BSA_GLM(logregParental, chr = "VIII")

ggplot(VIII_Fixed, aes(x = POS, y = bulk_Zprime))  +
  geom_point(data = VIIIResults, aes(x = POS, y = bulk_Zprime), alpha = 0.1) + geom_point(color = "red", size = 2)


dim(VIII_Fixed)
dim(VIIIResults)/100
```

## Loop through different start points

```{r, eval = FALSE}
saveaddition <- list()
for(addition in 1:100){
  saveaddition[[addition]] <- BSA_GLM_fixed(logregParental, chr = "VIII", a = addition, windowsize = 100)
}

# length(saveaddition[[1]]$CHROM)
multrows <- bind_rows(saveaddition)
# dim(multrows)
# length(unique(multrows$Index))

multrows %>% ggplot(aes(x = POS, y = bulk_Zprime)) + geom_point() +  geom_point(data = VIIIResults, aes(x = POS, y = bulk_Zprime), alpha = 0.1, color = "violet") 

```

## Using the spaced data

Since there are now positions/windowsize elements in the dataset, would these averages now be the number of hypotheses tested? 

Clearly this still actually doesn't help since the correction is too severe. Unfortunate.

```{r}
bonfcor <- 0.05/(length(unique(VIII_Fixed$POS)))
zscore_bc <- qnorm(bonfcor/2)

ggplot(VIII_Fixed, aes(x = POS, y = bulk_Zprime)) + geom_point(size = 3) + geom_line() + geom_hline(yintercept = c(zscore_bc, -zscore_bc), linetype = "dashed") + ggtitle("Chr VIII Bonferroni Correction")

```

## Recombination Rate and Window Size

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5554000/#:~:text=on%20the%20chromosome.-,S.,6%20COs%2FMb%20for%20L.

According to the internet, there are 6 crossovers/Mb in S cerevisiae. This means that there is 1 crossover per 0.167 Mb, or 167k bases. So we would expect the chances of 1/100 recombinations to be 167k/100, or 1.67k.

Magwene advised a window size of 20 cM, so that would be 1.67*20, or 33.4k bases. This is PER individual, and then must be multiplied by the number of starting population potentially, which was 1/2 million (if FuDR killed the alphas). However, this does not seem to make sense.

Calculate the distances between unique positions here:

```{r}

positions <- data.frame(POS = unique(VIIIResults$POS))

positions$Diff <- NA
for(i in 1:(length(positions$POS)-1)){
  positions$Diff[i] <- positions$POS[i+1] - positions$POS[i]
}

positions %>% filter(Diff < 500) %>% ggplot(aes(x = Diff)) + geom_density(alpha = 0.5)

positions %>% filter(Diff < 500) %>% ggplot(aes(x = Diff, y = "POS")) + geom_boxplot(size = 2, color = "violet") + geom_jitter(alpha = 0.5) + ggtitle("")

positions %>% ggplot(aes(x = Diff, y = "POS")) + geom_boxplot(size = 1, color = "violet") + geom_jitter(alpha = 0.1)

meanposdiff <- mean(positions$Diff, na.rm = T)
medposdiff <- median(positions$Diff, na.rm = T)

33400/meanposdiff

33400/medposdiff

length(positions$POS)
#I think I did the math on this wrong, since my window size should be 2826 elements...?
```

What does 33.4k bp actually look like?

```{r}

positions[which(is.na(positions$Diff)),]

positions <- positions %>% na.omit()

gap_VIII <- positions$POS[which(positions$Diff == max(positions$Diff))]

VIIIResults %>% ggplot(aes(x = POS, y = bulk_Zprime)) + geom_point() + geom_vline(xintercept = seq(from = 1, to = max(VIIIResults$POS), by = 33400), linetype = "dashed", color = "firebrick") 
```
So at 6 crossovers/Mb, that's 1Mb/6 = 0.166 Mb, or 166 Kb

```{r}
1000*1/6
```

What does every 166 Kb look like? Multiply by 20 since resolution isn't high enough, and this might(???) be 20 cM since the chance is for 1 recombination to happen in that space...????

```{r}
VIIIResults %>% ggplot(aes(x = POS, y = bulk_Zprime)) + geom_point() + geom_vline(xintercept = seq(from = 1, to = max(VIIIResults$POS), by = 166*100), linetype = "dashed", color = "firebrick") 

```

