---
title: "GLM for BSA Statistics"
author: "Cassandra Buzby"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
library(ggplot2)
library(tidyr)
library(tidyverse)
library(reshape2)
library(cowplot)

library(foreach)
library(doParallel)

#ggplot2::theme_set(theme_light())
ggplot2::theme_set(theme_cowplot())


CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47",
                  "#F26430", "#0A369D", "#7EA3CC")


library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")
load("SAUA_4-13-22.Rdata")
load("SCUC_4-13-22.Rdata")

# load("SAUAwithChrI.Rdata")
# load("SCUCwithChrI.Rdata")

#load("Selected_4-13-22.Rdata")
#load("Unselected_4-13-22.Rdata")

```

## Testing Mark's GLM idea for stats on BSA

```{r}
load("RASU_Pivot.Rdata")
```

```{r, eval = FALSE}
read.table("Wine_VCF.txt", header = TRUE) %>% mutate(parent = "Wine") -> WineTemp
read.table("Oak_VCF.txt", header = TRUE) %>% mutate(parent = "Oak") -> OakTemp

ParentalVCF <- rbind(WineTemp, OakTemp) %>% arrange(CHROM, POS) %>% select(CHROM, POS, REF, ALT, parent) %>% merge(ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes)

ParentalVCF %>% pivot_wider(names_from = parent, values_from = ALT) -> SNPids

SNPids$Type <- NA
SNPids$Type[is.na(SNPids$Wine) == TRUE] <- "Oak"
SNPids$Type[is.na(SNPids$Oak) == TRUE] <- "Wine"
SNPids$Type[SNPids$Wine == SNPids$Oak] <- "Alt"

#SNPids[is.na(SNPids$Type),]

```

## Which parent is which in the glm?

```{r, eval = FALSE}
load("ReadsParental.Rdata")
```

### Picking out just the effects of the high bulks

By dividing the alternate by the reference, the allele which is labeled as "Type" (ie the alternative) effect indicates which allele is the driving mutation.

```{r, eval = FALSE}
load("ReadsParental.Rdata")

ReadsParental %>% filter(CHROM == "VIII") %>% select(-Wine, -Oak, -REF) %>% pivot_wider(names_from = allele, values_from = value) %>% 
  ggplot(., aes(x = POS, y = ALT/REF, color = Type)) + 
    scale_color_manual(values = c("black", "blue","firebrick", "gray")) + 
  geom_point(alpha = 0.5)  + facet_grid(~bulk) + ggtitle("Chr VIII")

ReadsParental %>% filter(CHROM == "VII") %>% select(-Wine, -Oak, -REF) %>% pivot_wider(names_from = allele, values_from = value) %>% 
  ggplot(., aes(x = POS, y = ALT/REF, color = Type)) + 
    scale_color_manual(values = c("black", "blue","firebrick", "gray")) + 
  geom_point(alpha = 0.5)  + facet_grid(~bulk) + ggtitle("Chr VII")

ReadsParental %>% filter(bulk == "HIGH") %>% select(-Wine, -Oak, -REF) %>% 
  pivot_wider(names_from = allele, values_from = value) -> ReadsWider
 
ggplot(ReadsWider, aes(x = POS, y = ALT/REF, color = Type)) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(ReadsWider$POS), 
                                  by = 10^(floor(log10(max(ReadsWider$POS))))), 
                    name = "Genomic Position") +
    scale_color_manual(values = c("black", "blue","firebrick", "gray")) + 
    theme(axis.text.x = element_blank()) +
  geom_point(alpha = 0.5)  + facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle("High Bulk Allele Effects")
```


```{r, eval = FALSE}
## Converting reference and alternate into oak and wine

load("ReadsParental.Rdata")

ReadsParental %>% filter(Type == "Oak" | Type == "Wine") -> logregParental

logregParental$reads[logregParental$allele == "ALT"] <- logregParental$Type[logregParental$allele == "ALT"]
logregParental$reads[logregParental$allele == "REF" & logregParental$Type == "Oak"] <- "Wine"
logregParental$reads[logregParental$allele == "REF" & logregParental$Type == "Wine"] <- "Oak"

logregParental$reads <- factor(logregParental$reads)
head(logregParental)

save(logregParental, file = "logregParental.Rdata")
```


## Running all of this to capture the z score

This actually times out, so perhaps need to run with less in R memory or not as a for loop...?

Use summary(glm(reads~bulk*parent, weights = value, family = binomial)) for each chromosome, then merge together and plot the Z scores from this summary.

```{r, eval = FALSE}

load("logregParental.Rdata")
for(k in unique(logregParental$CHROM)){
  logregParental %>% select(CHROM, POS, value, reads, parent, bulk) %>% filter(CHROM == k) -> CHRM_logregP

  Results <- foreach (i=unique(CHRM_logregP$POS), .combine=rbind) %dopar% {
    res <- summary(glm(reads ~ bulk*parent, weights = value, family = binomial, data = CHRM_logregP[CHRM_logregP$POS == i,]))
    c(k, i, res$coefficients[,3])
  }
  
  Results <- as.data.frame(Results)

  assign(value = Results, x = paste("Results_", k, sep = ""))
}

Res_Z_Chr <- rbind(Results_II, Results_III, Results_IV, Results_V, Results_VI, Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
      Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)

rm(Results, Results_II, Results_III, Results_IV, Results_V, Results_VI, Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
      Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)

Res_Z_Chr$V2 <- as.numeric(Res_Z_Chr$V2)
Res_Z_Chr$`(Intercept)` <- as.numeric(Res_Z_Chr$`(Intercept)`)
Res_Z_Chr$bulkHIGH <- as.numeric(Res_Z_Chr$bulkHIGH)
Res_Z_Chr$parentWine <- as.numeric(Res_Z_Chr$parentWine)
Res_Z_Chr$`bulkHIGH:parentWine` <- as.numeric(Res_Z_Chr$`bulkHIGH:parentWine`)

save(Res_Z_Chr, file = "Res_Z_Chr.Rdata")
```


## Figuring out a good statistical cutoff

Two options: FDR (empirical) or adjusted p-value. Since there's 97206 observations, the p-value by bonferroni correction would be 0.05^92k, which is probably a lot...?

First, z-score at p = 0.05 is 1.96; cutoffs drawn below:

### Function uses weighted means

```{r}

#load("Res_Z_Chr_plusChrI.Rdata")

#Load function
WindowCB_meanweighted <- function(df = Res_Z_Chr[1:20,], windowsize = 5){
  
  df %>% arrange(CHROM, POS) -> df
  
  #Weights
  weightvector <- c(rep(1:(windowsize)),windowsize+1, rep((windowsize):1))

  #FOR LOOP through chromosomes
  
  Result <- foreach (k=unique(df$CHROM), .combine=rbind) %dopar% {
    
        #add reorder by position
        df[df$CHROM == k,] %>% arrange(POS)  -> dfnew
    
        #FOR LOOP from original index 1 to original end index
          foreach(i=windowsize+1:(length(dfnew$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- weighted.mean(dfnew$Interaction[(i-windowsize):(i+windowsize)], weightvector)
                res_bulk <- weighted.mean(dfnew$bulkHigh[(i-windowsize):(i+windowsize)], weightvector)
                res_parent <- weighted.mean(dfnew$parentWine[(i-windowsize):(i+windowsize)], weightvector)
                #print CHROM, index, POS, and mean
                c(k,i, dfnew$POS[i], res_bulk, res_parent, res_int)}
  
  }
  #rename everything once it's done
  
  Result <- as.data.frame(Result)
  colnames(Result) <- c("CHROM", "i", "POS", "Zprime_Bulk", "Zprime_Parent", "Zprime_Int")
  Result$POS <- as.numeric(Result$POS)
  Result$Zprime_Bulk <- as.numeric(Result$Zprime_Bulk)
  Result$Zprime_Parent <- as.numeric(Result$Zprime_Parent)
  Result$Zprime_Int <- as.numeric(Result$Zprime_Int)
  
  Result
}
```



```{r, eval = FALSE}
load("Res_Z_Chr.Rdata")

#load("Res_Z_Chr_plusChrI.Rdata")

colnames(Res_Z_Chr) <- c("CHROM", "POS", "Intercept", "bulkHigh", "parentWine", "Interaction")

results.mean.weighted.100 <- WindowCB_meanweighted(Res_Z_Chr, 100)

# dim(subset(Res_Z_Chr, CHROM == "I"))
#Plot to see
results.mean.weighted.100 %>% transmute(CHROM = CHROM, POS = POS, Bulk = Zprime_Bulk, Parent = Zprime_Parent, Int = Zprime_Int, Stat = "ZPrime") -> Zprimes
load("Zs.Rdata")
rbind(Zs, Zprimes) -> WeightedMeanZStats.100

WeightedMeanZStats.100$CHROM <- factor(WeightedMeanZStats.100$CHROM,
                              levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"))

#save(WeightedMeanZStats.100, file = "WeightedMeanZStats.100.Rdata")
```

```{r}
load("WeightedMeanZStats.100.Rdata")

WeightedMeanZStats.100 %>%  ggplot(., aes(x = POS, y = Int, color = Stat)) + geom_point(alpha = 0.3, size = 0.05) + 
  scale_color_manual(values = c("lightblue", "black")) + geom_hline(yintercept = 0, color = "gray")+
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(1.96, -1.96), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +
  ylab("Oak <- Bulk x Parent Interaction -> Wine") +
  facet_grid(~CHROM, scales = "free") + ggtitle("Interaction Z vs Zprime | Weighted Means W = 100")

WeightedMeanZStats.100 %>%  ggplot(., aes(x = POS, y = Bulk, color = Stat)) + geom_point(alpha = 0.3, size = 0.05) +  
  scale_color_manual(values = c("violet", "black")) + geom_hline(yintercept = 0, color = "gray")+ 
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(1.96, -1.96), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Bulk Effect -> Wine") +
  facet_grid(~CHROM, scales = "free") + ggtitle("Bulk Effect Z vs Zprime | Weighted Means W = 100")

WeightedMeanZStats.100 %>%  ggplot(., aes(x = POS, y = Parent, color = Stat)) + geom_point(alpha = 0.3, size = 0.05) + 
  scale_color_manual(values = c("darkorange", "black")) + geom_hline(yintercept = 0, color = "gray")+ 
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(1.96, -1.96), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Parent Effect -> Wine") +

  facet_grid(~CHROM, scales = "free") + ggtitle("Parent Effect Z vs Zprime | Weighted Means W = 100")

WeightedMeanZStats.100 %>% filter(Stat == "ZPrime") %>% pivot_longer(c("Bulk", "Parent", "Int")) %>% 
  ggplot(., aes(x = POS, y = value, color = name)) + geom_line(size = 1) + #geom_point(alpha = 0.1) + #+ 
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(1.96, -1.96), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Effect -> Wine") +
  scale_color_manual(values = c("violet", "lightblue", "darkorange"))+
  facet_grid(~CHROM, scales = "free") + ggtitle("Comparison of Zprime per Factor | Weighted Means W = 100")
```


## Using Permutations to Estimate FDR

What is the null distribution of the effect? It should be the variation that is seen in sequencing without the QTL. This is not the same as reshuffling the z-scores, though that would give the FDR within different window sizes (ie if the order didn't matter). Could reshuffle read counts and then calculate effects that way (would be longer but would probably work well).

**Do this in bash**

```{r, eval = FALSE}
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

#install.packages("foreach")
#install.packages("doParallel")

set.seed(as.numeric(args[1]))

#Load packages

library(foreach)
library(doParallel)
ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

#load("logregParental.Rdata")
#logregParental %>% select(CHROM, POS, value, reads, parent, bulk) -> logregP_bash

#save(logregP_bash, file = "logregP_bash.Rdata")
load("logregP_bash.Rdata")

  indices <- sample(logregP_bash$value, replace = FALSE)
  logregP_bash$value <- indices
    
  for(k in unique(logregP_bash$CHROM)){
    logregP_bash[logregP_bash$CHROM == k,] -> CHRM_logregP
  
    Results <- foreach (i=unique(CHRM_logregP$POS), .combine=rbind) %dopar% {
      res <- summary(glm(reads ~ bulk*parent, weights = value, family = binomial, data = CHRM_logregP[CHRM_logregP$POS == i,]))
      c(k, i, res$coefficients[,3], as.numeric(args[1]))
    }
    
    Results <- as.data.frame(Results)
  
    assign(value = Results, x = paste("Results_", k, sep = ""))
  }
  
  Res_Z_Chr <- rbind(Results_II, Results_III, Results_IV, Results_V, Results_VI, 
                     Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
        Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)
  
  rm(Results, Results_II, Results_III, Results_IV, Results_V, Results_VI, 
     Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
        Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)
  
  Res_Z_Chr$V2 <- as.numeric(Res_Z_Chr$V2)
  Res_Z_Chr$`(Intercept)` <- as.numeric(Res_Z_Chr$`(Intercept)`)
  Res_Z_Chr$bulkHIGH <- as.numeric(Res_Z_Chr$bulkHIGH)
  Res_Z_Chr$parentWine <- as.numeric(Res_Z_Chr$parentWine)
  Res_Z_Chr$`bulkHIGH:parentWine` <- as.numeric(Res_Z_Chr$`bulkHIGH:parentWine`)
  
 
  filename <- paste("Res_Z_Chr_", as.numeric(args[1]), ".csv", sep = "")
  write.csv(Res_Z_Chr, file = filename, col.names = FALSE)

```

```{r}
#the result of:
#cat GLM_QTL_Zperm/*.csv > testGLMs.csv

permRes <- read.csv("testGLMs.csv")

permRes %>% transmute(CHROM = V1, POS = as.numeric(V2), bulkHigh = as.numeric(bulkHIGH), parentWine = as.numeric(parentWine), Interaction = as.numeric(bulkHIGH.parentWine)) %>% na.omit() -> PermutationResults

quantile(PermutationResults$Bulk, c(0.025, 0.975))
quantile(PermutationResults$Parent, c(0.025, 0.975))
quantile(PermutationResults$Int, c(0.025, 0.975))

PR_Window <- WindowCB_meanweighted(df = subset(PermutationResults, CHROM == "II"), windowsize = 100)

quantile(PR_Window$Zprime_Bulk, c(0.025, 0.975))
quantile(PR_Window$Zprime_Parent, c(0.025, 0.975))
quantile(PR_Window$Zprime_Int, c(0.025, 0.975))

PR_Window %>% pivot_longer(c("Zprime_Bulk", "Zprime_Parent", "Zprime_Int")) %>% 
  ggplot(., aes(x = POS, y = value, color = name)) + geom_point(size = 0.1, alpha = 0.2) + 
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(2.1, -2.1), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Effect -> Wine") +
  scale_color_manual(values = c("violet", "lightblue", "darkorange"))+
  facet_grid(~name) + ggtitle("Comparison of Permuted Zprime per Factor | Weighted Means W = 30")

ggplot(subset(PR_Window, POS < 50000), aes(x = POS, y = Zprime_Bulk)) + geom_point(size = 0.1, alpha = 0.2, color = "violet") + 
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = quantile(PR_Window$Zprime_Bulk, c(0.025, 0.975)), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Effect -> Wine")
```


## What does the current data look like?

The range of data in the smoothed vs unsmoothed glm is variable and therefore the range that should be used for the z scores should be different from the z' scores. How exactly to do this is not clear, but let's try reshuffling the z scores to see how that changes the means.

```{r}

load("Res_Z_Chr.Rdata")

quantile(Res_Z_Chr$bulkHIGH, c(0.025, 0.975))
quantile(Res_Z_Chr$parentWine, c(0.025, 0.975))
quantile(Res_Z_Chr$`bulkHIGH:parentWine`, c(0.025, 0.975))
```

```{r}
quantile(WeightedMeanZStats.100$Bulk, c(0.025, 0.975))
quantile(WeightedMeanZStats.100$Parent, c(0.025, 0.975))
quantile(WeightedMeanZStats.100$Int, c(0.025, 0.975))

```

### Reshuffling z-scores

```{r}
indices <- sample(Res_Z_Chr$bulkHIGH)

Res_Z_Reshuffle <- Res_Z_Chr %>% mutate(CHROM = V1, POS = V2, Interaction = `bulkHIGH:parentWine`)
Res_Z_Reshuffle$bulkHigh <- indices

Reshuffle100.1 <- WindowCB_meanweighted(Res_Z_Reshuffle, 100)

quantile(Reshuffle100.1$Zprime_Bulk, c(0.025, 0.975))
quantile(Reshuffle100.1$Zprime_Parent, c(0.025, 0.975))
quantile(Reshuffle100.1$Zprime_Int, c(0.025, 0.975))

Reshuffle100.1 %>% pivot_longer(c("Zprime_Bulk", "Zprime_Parent", "Zprime_Int")) %>% 
  ggplot(., aes(x = POS, y = value, color = name)) + geom_point(size = 0.1, alpha = 0.2) + 
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(1.12, -1.16), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Effect -> Wine") +
  scale_color_manual(values = c("violet", "lightblue", "darkorange"))+
  facet_grid(~CHROM, scale = "free") + ggtitle("Comparison of Permuted Zprime per Factor | Weighted Means W = 100")
```

### Looping through to make a set of quantiles

This does not run correctly 

```{r, eval = FALSE}
Res_Z_Reshuffle <- Res_Z_Chr %>% mutate(CHROM = V1, POS = V2, Interaction = `bulkHIGH:parentWine`)

ResultsZShuffle <- foreach (i=c(1:10), .combine=rbind) %dopar% {
    
    indices <- sample(Res_Z_Chr$bulkHIGH)
    Res_Z_Reshuffle$bulkHigh <- indices
    Reshuffle100.1 <- WindowCB_meanweighted(Res_Z_Reshuffle, 100)
    
    c(quantile(Reshuffle100.1$Zprime_Bulk, c(0.025, 0.975)),
      quantile(Reshuffle100.1$Zprime_Parent, c(0.025, 0.975)),
      quantile(Reshuffle100.1$Zprime_Int, c(0.025, 0.975)))
  }

```


### Maybe I only need one 100-loci sample at a time...?

If the null distribution is what the effect is from randomly having different bulks, maybe the effect of the bulks reshuffled is how to do this? And then if it's just a sample of 100 and a single score, then that single score would be the thing to run summaries on to see where 5% or 1% FDR is. So steps:

1. Reshuffle alternate and reference alleles from each bulk
2. Calculate 100 GLMs (each trial)
3. Find the weighted mean of those 100 GLMs
4. Repeat

```{r}
load("logregParental.Rdata")

ResSummary <- data.frame(Intercept = 0, Bulk = 0, Parent = 0, Interaction = 0)
for(i in 1:1000){
  chosenpos <- sample(unique(logregParental$POS), 100)

  logreg100 <- subset(logregParental, POS %in% chosenpos)
  length(unique(logreg100$POS)) #confirmed

  logreg100 %>% select(CHROM, POS, value, reads, parent, bulk) -> CHRM_logregP

  Results <- foreach (i=unique(CHRM_logregP$POS), .combine=rbind) %dopar% {
    res <- summary(glm(reads ~ bulk*parent, weights = value, family = binomial, data = CHRM_logregP[CHRM_logregP$POS == i,]))
    c(res$coefficients[,3]) 
  }
  
  Results <- as.data.frame(Results) 
  ResSummary[i,] <- apply(Results, 2, mean)
}


ResSummary %>% pivot_longer(-Intercept) %>% 
  ggplot(., aes(x = name, y = value, color = name)) + geom_boxplot(size = 1) + 
  geom_point(alpha = 0.2, size = 2, position  = position_jitterdodge()) +  
  scale_color_manual(values = c("violet", "lightblue", "darkorange"))+ ggtitle("Mean Z score from Reshuffled Positions")

```

## What are my null assumptions?

1. There is no effect of bulk on allele frequency differences (shuffling reads across bulks)
2. There is no effect of position on smoothed significance (shuffling positions)

_This took 7 hours to run for 30k samples_

```{r, eval = FALSE}
load("logregParental.Rdata")

start.loop <- Sys.time()
ResSummary <- data.frame(Intercept = 0, Bulk = 0, Parent = 0, Interaction = 0)
for(i in 1:30000){
  
  logregParental$value <- sample(logregParental$value)
  
  chosenpos <- sample(unique(logregParental$POS), 100)
  logreg100 <- subset(logregParental, POS %in% chosenpos)
  
  logreg100 %>% select(CHROM, POS, value, reads, parent, bulk) -> CHRM_logregP

  Results <- foreach (i=unique(CHRM_logregP$POS), .combine=rbind) %dopar% {
    res <- summary(glm(reads ~ bulk*parent, weights = value, family = binomial, data = CHRM_logregP[CHRM_logregP$POS == i,]))
    c(res$coefficients[,3]) 
  }
  
  Results <- as.data.frame(Results) 
  ResSummary[i,] <- apply(Results, 2, mean)
}

Sys.time() - start.loop


ResSummary %>%  pivot_longer(-Intercept) %>% 
  ggplot(., aes(x = name, y = value, color = name)) + geom_boxplot(size = 1) + 
  geom_point(alpha = 0.2, size = 2, position  = position_jitterdodge()) +  
  scale_color_manual(values = c("violet", "lightblue", "darkorange"))+ ggtitle("Mean Z score from Reshuffled Values")

save(ResSummary, file = "ResSummary_ShuffledPOS_30000x.Rdata")

```

```{r}
load("ResSummary_ShuffledPOS_30000x.Rdata")

quantile(ResSummary$Bulk, c(0.025, 0.975))
quantile(ResSummary$Parent, c(0.025, 0.975))
quantile(ResSummary$Interaction, c(0.025, 0.975))

#If each of these was independent, it would be 90k samples instead of 30k samples, but 30k is the length of the entire dataset and then each is a score from a window of 100, so really it's the summary of each 100-position set

ResSummary %>%  pivot_longer(-Intercept) %>% 
  ggplot(., aes(x = name, y = value, color = name)) +
  geom_point(alpha = 0.1, size = 0.5, position  = position_jitterdodge()) +  
  geom_hline(yintercept = quantile(ResSummary$Bulk, c(0.025, 0.975)), linetype = "dashed") +
  geom_violin(size = 2, alpha = 0.1) + 
  ylim(c(-5,5)) + 
  scale_color_manual(values = c("violet", "lightblue", "darkorange"))+ ggtitle("Mean Z score from Reshuffled Values")

```

## More stats

So the fact that these are weighted means ensures that the sliding window will be skewed by each; so really if we see peaks that is not actually indicative of it being real, but that it's averaged.

So perhaps the slope of the peaks is what should be the null as well...?

# Updates since 6/15/22

## Running the model only on parental effects

This first round does not exclude the selected (HIGH) bulks at all, so the whole thing is based on all data but removes bulk from the model.

```{r, eval = FALSE}
load("logregParental.Rdata")
for(k in unique(logregParental$CHROM)){
  logregParental %>% select(CHROM, POS, value, reads, parent, bulk) %>% filter(CHROM == k) -> CHRM_logregP

  Results <- foreach (i=unique(CHRM_logregP$POS), .combine=rbind) %dopar% {
    res <- summary(glm(reads ~ parent, weights = value, family = binomial, data = CHRM_logregP[CHRM_logregP$POS == i,]))
    c(k, i, res$coefficients[,3])
  }
  
  Results <- as.data.frame(Results)

  assign(value = Results, x = paste("Results_", k, sep = ""))
}

Res_Z_Chr_PEffect <- rbind(Results_II, Results_III, Results_IV, Results_V, Results_VI, Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
      Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)

rm(Results, Results_II, Results_III, Results_IV, Results_V, Results_VI, Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
      Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)

Res_Z_Chr_PEffect$V2 <- as.numeric(Res_Z_Chr_PEffect$V2)
Res_Z_Chr_PEffect$`(Intercept)` <- as.numeric(Res_Z_Chr_PEffect$`(Intercept)`)
#Res_Z_Chr_PEffect$bulkHIGH <- as.numeric(Res_Z_Chr_PEffect$bulkHIGH)
Res_Z_Chr_PEffect$parentWine <- as.numeric(Res_Z_Chr_PEffect$parentWine)
#Res_Z_Chr_PEffect$`bulkHIGH:parentWine` <- as.numeric(Res_Z_Chr_PEffect$`bulkHIGH:parentWine`)

save(Res_Z_Chr_PEffect, file = "Res_Z_Chr_PEffect.Rdata")
```

```{r}

#Load function
WindowCB_meanweighted.PEffect <- function(df = Res_Z_Chr[1:20,], windowsize = 5){
  
  df %>% arrange(CHROM, POS) -> df
  
  #Weights
  weightvector <- c(rep(1:(windowsize)),windowsize+1, rep((windowsize):1))

  #FOR LOOP through chromosomes
  
  Result <- foreach (k=unique(df$CHROM), .combine=rbind) %dopar% {
    
        #add reorder by position
        df[df$CHROM == k,] %>% arrange(POS)  -> dfnew
    
        #FOR LOOP from original index 1 to original end index
          foreach(i=windowsize+1:(length(dfnew$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                #res_int <- weighted.mean(dfnew$Interaction[(i-windowsize):(i+windowsize)], weightvector)
                #res_bulk <- weighted.mean(dfnew$bulkHigh[(i-windowsize):(i+windowsize)], weightvector)
                res_parent <- weighted.mean(dfnew$parentWine[(i-windowsize):(i+windowsize)], weightvector)
                #print CHROM, index, POS, and mean
                c(k,i, dfnew$POS[i], res_parent)}
  
  }
  #rename everything once it's done
  
  Result <- as.data.frame(Result)
  colnames(Result) <- c("CHROM", "i", "POS", "Zprime_Parent")
  Result$POS <- as.numeric(Result$POS)
  #Result$Zprime_Bulk <- as.numeric(Result$Zprime_Bulk)
  Result$Zprime_Parent <- as.numeric(Result$Zprime_Parent)
  #Result$Zprime_Int <- as.numeric(Result$Zprime_Int)
  
  Result
}
```

```{r, eval = FALSE}
load("Res_Z_Chr_PEffect.Rdata")
colnames(Res_Z_Chr_PEffect) <- c("CHROM", "POS", "Intercept", "parentWine")

ParentEffect.mean.weighted.100 <- WindowCB_meanweighted.PEffect(Res_Z_Chr_PEffect, 100)

# dim(subset(Res_Z_Chr, CHROM == "I"))
#Plot to see
ParentEffect.mean.weighted.100 %>% transmute(CHROM = CHROM, POS = POS, Parent = Zprime_Parent, Stat = "ZPrime") -> Zprimes
load("Zs.Rdata")

Zs %>% select(CHROM, POS, Parent, Stat) -> Zs.P
rbind(Zs.P, Zprimes) -> P.WeightedMeanZStats.100

P.WeightedMeanZStats.100$CHROM <- factor(WeightedMeanZStats.100$CHROM,
                              levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"))

save(P.WeightedMeanZStats.100, file = "P.WeightedMeanZStats.100.Rdata")

```

```{r}
load("P.WeightedMeanZStats.100.Rdata")
unique(P.WeightedMeanZStats.100$Stat)
P.WeightedMeanZStats.100 %>%  ggplot(., aes(x = POS, y = Parent, color = Stat)) + geom_point(alpha = 0.3, size = 0.05) + 
  scale_color_manual(values = c("darkorange", "black")) + geom_hline(yintercept = 0, color = "gray")+ 
  geom_hline(yintercept = 0, color = "white") +
  geom_hline(yintercept = quantile(subset(P.WeightedMeanZStats.100, Stat == "Z")$Parent, c(0.025, 0.975)), color = "red", linetype = "dashed") +
  geom_hline(yintercept = quantile(subset(P.WeightedMeanZStats.100, Stat == "ZPrime")$Parent, c(0.025, 0.975)), color = "black", linetype = "dashed") +
    geom_hline(yintercept = c(-1.96, 1.96), color = "cyan") +
  ylim(-5, 5) +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Parent Effect -> Wine") +
  facet_grid(~CHROM, scales = "free") + ggtitle("Parent GLM Z vs Zprime | Weighted Means W = 100")

```

## Running the model only on parental effects

This next round does not INCLUDE the selected (HIGH) bulks at all, so only the unselected bulk is used.

```{r, eval = FALSE}
load("logregParental.Rdata")

logregParental %>% filter(bulk != "HIGH") -> logregParental.U

for(k in unique(logregParental.U$CHROM)){
  logregParental.U %>% select(CHROM, POS, value, reads, parent, bulk) %>% filter(CHROM == k) -> CHRM_logregP

  Results <- foreach (i=unique(CHRM_logregP$POS), .combine=rbind) %dopar% {
    res <- summary(glm(reads ~ parent, weights = value, family = binomial, data = CHRM_logregP[CHRM_logregP$POS == i,]))
    c(k, i, res$coefficients[,3])
  }
  
  Results <- as.data.frame(Results)

  assign(value = Results, x = paste("Results_", k, sep = ""))
}

Res_Z_Chr_UPEffect <- rbind(Results_II, Results_III, Results_IV, Results_V, Results_VI, Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
      Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)

rm(Results, Results_II, Results_III, Results_IV, Results_V, Results_VI, Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
      Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)

Res_Z_Chr_UPEffect$V2 <- as.numeric(Res_Z_Chr_UPEffect$V2)
Res_Z_Chr_UPEffect$`(Intercept)` <- as.numeric(Res_Z_Chr_UPEffect$`(Intercept)`)
Res_Z_Chr_UPEffect$parentWine <- as.numeric(Res_Z_Chr_UPEffect$parentWine)

save(Res_Z_Chr_UPEffect, file = "Res_Z_Chr_UPEffect.Rdata")
```

```{r, eval = FALSE}
load("Res_Z_Chr_UPEffect.Rdata")
colnames(Res_Z_Chr_UPEffect) <- c("CHROM", "POS", "Intercept", "parentWine")

UParentEffect.mean.weighted.100 <- WindowCB_meanweighted.PEffect(Res_Z_Chr_UPEffect, 100)

#Plot to see
UParentEffect.mean.weighted.100 %>% transmute(CHROM = CHROM, POS = POS, Parent = Zprime_Parent, Stat = "ZPrime") -> UParentEffect.Zprimes

Res_Z_Chr_UPEffect %>% transmute(CHROM=CHROM, POS=POS, Parent=parentWine, Stat="Z") -> Zs.UP
rbind(Zs.UP, UParentEffect.Zprimes) -> UP.WeightedMeanZStats.100

UP.WeightedMeanZStats.100$CHROM <- factor(WeightedMeanZStats.100$CHROM,
                              levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"))

save(UP.WeightedMeanZStats.100, file = "UP.WeightedMeanZStats.100.Rdata")

```

```{r}
load("UP.WeightedMeanZStats.100.Rdata")

UP.WeightedMeanZStats.100 %>%  ggplot(., aes(x = POS, y = Parent, color = Stat)) + geom_point(alpha = 0.3, size = 0.05) + 
  scale_color_manual(values = c("darkorange", "black")) + geom_hline(yintercept = 0, color = "gray")+ 
  geom_hline(yintercept = 0, color = "white") +
  geom_hline(yintercept = quantile(subset(UP.WeightedMeanZStats.100, Stat == "Z")$Parent, c(0.025, 0.975)), color = "red", linetype = "dashed") +
  geom_hline(yintercept = quantile(subset(UP.WeightedMeanZStats.100, Stat == "ZPrime")$Parent, c(0.025, 0.975)), color = "black", linetype = "dashed") +
    geom_hline(yintercept = c(-1.96, 1.96), color = "cyan") +
  ylim(-5, 5) +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Parent Effect -> Wine") +
  facet_grid(~CHROM, scales = "free") + ggtitle("[Unselected] Parent GLM Z vs Zprime | W = 100")

```

## Running the model on randomly reassigned bulks of unselected only

```{r, eval = FALSE}
load("logregParental.Rdata")

logregParental %>% filter(bulk != "HIGH") %>% group_by(value) %>% mutate(
  newHIGH = rbinom(n = 1, size = value, prob = 0.5), 
  newLOW = rbinom(n = 1, size = value, prob = 0.5)) %>% ungroup() %>%
  select(CHROM, POS, allele, parent, reads, newHIGH, newLOW) %>% pivot_longer(c("newHIGH", "newLOW")) -> logregParental.U

for(k in unique(logregParental.U$CHROM)){
  logregParental.U %>% select(CHROM, POS, value, reads, parent, name) %>% filter(CHROM == k) -> CHRM_logregP

  Results <- foreach (i=unique(CHRM_logregP$POS), .combine=rbind) %dopar% {
    res <- summary(glm(reads ~ name*parent, weights = value, family = binomial, 
                       data = CHRM_logregP[CHRM_logregP$POS == i,]))
    c(k, i, res$coefficients[,3])
  }
  
  Results <- as.data.frame(Results)

  assign(value = Results, x = paste("Results_", k, sep = ""))
}

Res_Z_Chr_ShuffledLowHigh <- rbind(Results_II, Results_III, Results_IV, Results_V, Results_VI, Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
      Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)

rm(Results, Results_II, Results_III, Results_IV, Results_V, Results_VI, Results_VII, Results_VIII, Results_IX, Results_X, Results_XI, 
      Results_XII, Results_XIII, Results_XIV, Results_XV, Results_XVI)

Res_Z_Chr_ShuffledLowHigh$POS <- as.numeric(Res_Z_Chr_ShuffledLowHigh$V2)
Res_Z_Chr_ShuffledLowHigh$`(Intercept)` <- as.numeric(Res_Z_Chr_ShuffledLowHigh$`(Intercept)`)
Res_Z_Chr_ShuffledLowHigh$interaction <- as.numeric(Res_Z_Chr_ShuffledLowHigh$`namenewLOW:parentWine`)
Res_Z_Chr_ShuffledLowHigh$parent <- as.numeric(Res_Z_Chr_ShuffledLowHigh$parentWine)
Res_Z_Chr_ShuffledLowHigh$bulk <- as.numeric(Res_Z_Chr_ShuffledLowHigh$namenewLOW)
Res_Z_Chr_ShuffledLowHigh$CHROM <- Res_Z_Chr_ShuffledLowHigh$V1

Res_Z_Chr_ShuffledLowHigh %>% select(CHROM, POS, bulk, parent, interaction) -> RZShuffledLowHigh
save(RZShuffledLowHigh, file = "Res_Z_Chr_ShuffledLowHigh.Rdata")
```

```{r, eval = FALSE}
load("Res_Z_Chr_ShuffledLowHigh.Rdata")

#load("Res_Z_Chr_plusChrI.Rdata")

colnames(RZShuffledLowHigh) <- c("CHROM", "POS", "bulkHigh", "parentWine", "Interaction")

SLH.results.mean.weighted.100 <- WindowCB_meanweighted(RZShuffledLowHigh, 100)

# dim(subset(Res_Z_Chr, CHROM == "I"))
#Plot to see
SLH.results.mean.weighted.100 %>% transmute(CHROM = CHROM, POS = POS, Bulk = Zprime_Bulk, Parent = Zprime_Parent, Int = Zprime_Int, Stat = "ZPrime") -> SLH.Zprimes

RZShuffledLowHigh %>% transmute(CHROM=CHROM, POS=POS, Bulk = bulkHigh, Parent=parentWine, Int = Interaction, Stat="Z") -> Zs.ShuffledLowHigh

rbind(Zs.ShuffledLowHigh, SLH.Zprimes) -> SLH.WeightedMeanZStats.100

SLH.WeightedMeanZStats.100$CHROM <- factor(SLH.WeightedMeanZStats.100$CHROM,
                              levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"))

save(SLH.WeightedMeanZStats.100, file = "SLH.WeightedMeanZStats.100.Rdata")
```

```{r}
load("SLH.WeightedMeanZStats.100.Rdata")

SLH.WeightedMeanZStats.100 %>%  ggplot(., aes(x = POS, y = Int, color = Stat)) + geom_point(alpha = 0.3, size = 0.05) + 
  scale_color_manual(values = c("lightblue", "black")) + geom_hline(yintercept = 0, color = "gray")+
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(1.96, -1.96), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +
  ylab("Oak <- Bulk x Parent Interaction -> Wine") +
  facet_grid(~CHROM, scales = "free") + ggtitle("Interaction Z vs Zprime | Weighted Means W = 100")

SLH.WeightedMeanZStats.100 %>%  ggplot(., aes(x = POS, y = Bulk, color = Stat)) + geom_point(alpha = 0.3, size = 0.05) +  
  scale_color_manual(values = c("violet", "black")) + geom_hline(yintercept = 0, color = "gray")+ 
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(1.96, -1.96), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Bulk Effect -> Wine") +
  facet_grid(~CHROM, scales = "free") + ggtitle("Bulk Effect Z vs Zprime | Weighted Means W = 100")

SLH.WeightedMeanZStats.100 %>%  ggplot(., aes(x = POS, y = Parent, color = Stat)) + geom_point(alpha = 0.3, size = 0.05) + 
  scale_color_manual(values = c("darkorange", "black")) + geom_hline(yintercept = 0, color = "gray")+ 
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(1.96, -1.96), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Parent Effect -> Wine") +

  facet_grid(~CHROM, scales = "free") + ggtitle("Parent Effect Z vs Zprime | Weighted Means W = 100")

SLH.WeightedMeanZStats.100 %>% filter(Stat == "ZPrime") %>% pivot_longer(c("Bulk", "Parent", "Int")) %>% 
  ggplot(., aes(x = POS, y = value, color = name)) + geom_line(size = 1) + #geom_point(alpha = 0.1) + #+ 
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = c(1.96, -1.96), color = "black", linetype = "dashed") +
  theme(axis.text.x = element_blank()) +  ylab("Oak <- Effect -> Wine") +
  scale_color_manual(values = c("violet", "lightblue", "darkorange"))+
  facet_grid(~CHROM, scales = "free") + ggtitle("Comparison of Zprime per Factor | Weighted Means W = 100")
```

## Sub-sampling using binomial sampling

```{r}
load("logregParental.Rdata")

logregParental %>% summarize(sum(value)) -> totalreads
coverage <- totalreads/length(logregParental)

# logregParental %>% select(-REF, -Wine, -Oak, -reads) %>% 
#   pivot_wider(names_from = c(allele), values_from = value) -> lrP_wider

for(i in 1:length(logregParental$CHROM)){
  logregParental$subs[i] <- rbinom(1, logregParental$value[i], 0.1)
}

str(logregParental)
```

New function for all of it with windows too

```{r}
logregParental %>% select(-REF, -Wine, -Oak, -Type) -> lrP

perChromWindow <- function(lrP, chr = "II", windowsize = 100){
  lrP <- lrP[lrP$CHROM == chr,]
  #Run the glm on that chromosome
  Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
    res <- glm(reads ~ bulk*parent, weights = value, family = binomial, data = lrP[lrP$POS == i,])
    c(chr, i, res$coefficients)
  }
  
  #Format the results for the next function
  Results <- as.data.frame(Results)
  colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh", "parentWine", "Interaction")
  Results[,2] <- as.numeric(Results[,2])
  Results[,3] <- as.numeric(Results[,3])
  Results[,4] <- as.numeric(Results[,4])
  Results[,5] <- as.numeric(Results[,5])
  Results[,6] <- as.numeric(Results[,6])
  
  Results %>% arrange(POS) %>% select(-Intercept) -> Results

  
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  return(WResult)
}

lrP_II <- perChromWindow(lrP, chr = "II")

```

## Add in sub-sampling to this function

```{r}
perChromWindow_binom <- function(lrP, chr = "II", windowsize = 100, perc_total = 0.1, seedt = Sys.time()){
  lrP <- lrP[lrP$CHROM == chr,]
  
  set.seed(seedt)
  
  for(i in 1:length(lrP$CHROM)){
    lrP$subs[i] <- rbinom(1, lrP$value[i], prob = perc_total) + 1
  }
  
  #Run the glm on that chromosome
  Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
    res <- glm(reads ~ bulk*parent, weights = subs, family = binomial, data = lrP[lrP$POS == i,])
    c(chr, i, res$coefficients)
  }
  
  #Format the results for the next function
  Results <- as.data.frame(Results)
  colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh", "parentWine", "Interaction")
  Results[,2] <- as.numeric(Results[,2])
  Results[,3] <- as.numeric(Results[,3])
  Results[,4] <- as.numeric(Results[,4])
  Results[,5] <- as.numeric(Results[,5])
  Results[,6] <- as.numeric(Results[,6])
  
  Results %>% arrange(POS) %>% select(-Intercept) -> Results

  
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  return(WResult)
}

perChromWindow_binom(logregParental, chr = "VIII", perc_total = 0.5) %>% ggplot(., aes(x = POS, y = as.numeric(bulk_Zprime))) + geom_line()

```

Running this as a loop

```{r}

ChrVIII_List <- list()
for(i in 1:100){
  ChrVIII_List[[i]] <- perChromWindow_binom(logregParental, chr = "VIII", perc_total = 0.1, seedt = i) 
}

save(ChrVIII_List, file = "ChrVIII_List.Rdata")

ChrVIII_10 <- bind_rows(ChrVIII_List, .id = "Index")

ggplot(ChrVIII_10, aes(x = POS, y = as.numeric(bulk_Zprime), color = factor(Index))) + geom_line()

```

```{r}
ChrVIII_10 %>% group_by(POS) %>% summarise(mean = mean(bulk_Zprime),
                                           sd = sd(bulk_Zprime)) %>% 
  ggplot(., aes(x = POS, y = mean, ymin = mean-sd, ymax = mean+sd)) + 
  geom_linerange(color = "gray")+ geom_line() +
  ggtitle("100 Samples of Smoothed 0.1x") +
  ylab("Bulk Zprime")

```
Now for all of the chromosomes...?

```{r}
#Chromosome Seven to check - 10% of data
ChrVII_List_10 <- list()
for(i in 1:10){
  ChrVII_List_10[[i]] <- perChromWindow_binom(logregParental, chr = "VII", perc_total = 0.1, seedt = i) 
}

save(ChrVII_List_10, file = "ChrVII_List_10.Rdata")
ChrVII_100_10perc <- bind_rows(ChrVII_List_10, .id = "Index")

#Chromosome Seven to check - 50% of data
ChrVII_List_50 <- list()
for(i in 1:10){
  ChrVII_List_50[[i]] <- perChromWindow_binom(logregParental, chr = "VII", perc_total = 0.5, seedt = i) 
}

save(ChrVII_List_50, file = "ChrVII_List_50.Rdata")
ChrVII_100_50perc <- bind_rows(ChrVII_List_50, .id = "Index")

ChrVII_100_50perc %>% ggplot(., aes(x = POS, y = parent_Zprime, color = Index)) + geom_line()

ChrVII_100_10perc %>% group_by(POS) %>% summarise(mean = mean(bulk_Zprime),
                                           sd = sd(bulk_Zprime)) %>% 
  ggplot(., aes(x = POS, y = mean, ymin = mean-sd, ymax = mean+sd)) + 
  #ylim(-1, 0.2) +
  geom_linerange(color = "violet")+ geom_line() +
  #ggtitle("10 Samples of Smoothed 0.1x") +
  ylab("Bulk Zprime") -> p1

ChrVII_100_50perc %>% group_by(POS) %>% summarise(mean = mean(bulk_Zprime),
                                           sd = sd(bulk_Zprime)) %>% 
  ggplot(., aes(x = POS, y = mean, ymin = mean-sd, ymax = mean+sd)) + 
  #ylim(-1, 0.2) +
  geom_linerange(color = "violet")+ geom_line() +
  #ggtitle("10 Samples of Smoothed 0.5x") +
  ylab("Bulk Zprime") -> p2

plot_grid(p1, p2, labels = c('Bulk Zprime 10%', 'Bulk Zprime 50%'))

##############################################################################
ChrVII_100_10perc %>% group_by(POS) %>% summarise(mean = mean(Interaction_Zprime),
                                           sd = sd(Interaction_Zprime)) %>% 
  ggplot(., aes(x = POS, y = mean, ymin = mean-sd, ymax = mean+sd)) + 
  #ylim(-1, 0.2) +
  geom_linerange(color = "lightblue")+ geom_line() +
  #ggtitle("10 Samples of Smoothed 0.1x") +
  ylab("Bulk Zprime") -> p1

ChrVII_100_50perc %>% group_by(POS) %>% summarise(mean = mean(Interaction_Zprime),
                                           sd = sd(Interaction_Zprime)) %>% 
  ggplot(., aes(x = POS, y = mean, ymin = mean-sd, ymax = mean+sd)) + 
  #ylim(-1, 0.2) +
  geom_linerange(color = "lightblue")+ geom_line() +
  #ggtitle("10 Samples of Smoothed 0.5x") +
  ylab("Bulk Zprime") -> p2

plot_grid(p1, p2, labels = c('Interaction Zprime 10%', 'Interaction Zprime 50%'))

##############################################################################
ChrVII_100_10perc %>% group_by(POS) %>% summarise(mean = mean(parent_Zprime),
                                           sd = sd(parent_Zprime)) %>% 
  ggplot(., aes(x = POS, y = mean, ymin = mean-sd, ymax = mean+sd)) + 
  #ylim(-1, 0.2) +
  geom_linerange(color = "darkorange")+ geom_line() +
  #ggtitle("10 Samples of Smoothed 0.1x") +
  ylab("Bulk Zprime") -> p1

ChrVII_100_50perc %>% group_by(POS) %>% summarise(mean = mean(parent_Zprime),
                                           sd = sd(parent_Zprime)) %>% 
  ggplot(., aes(x = POS, y = mean, ymin = mean-sd, ymax = mean+sd)) + 
  #ylim(-1, 0.2) +
  geom_linerange(color = "darkorange")+ geom_line() +
  #ggtitle("10 Samples of Smoothed 0.5x") +
  ylab("Bulk Zprime") -> p2

plot_grid(p1, p2, labels = c('Parent Zprime 10%', 'Parent Zprime 50%'))

logregParental %>% filter(CHROM == "VII") %>% group_by(POS) %>% summarise(sum = sum(value)) %>% ggplot(., aes(x = POS, y = sum)) + geom_point(alpha = 0.2) + ggtitle("Coverage of Chr VII") -> p3

logregParental %>% filter(CHROM == "VII") %>% ggplot(., aes(x = POS, y = value, color = reads)) + geom_point(alpha = 0.2) + ggtitle("Coverage of Chr VII") + scale_color_manual(values = c("lightblue", "firebrick")) + facet_grid(~bulk+parent)

logregParental %>% filter(CHROM == "VII", bulk == "HIGH") %>% ggplot(., aes(x = POS, y = value, group = c(parent))) + geom_point(alpha = 0.02) -> p4

plot_grid(p2, p4, nrow = 2)
```
******************************************************************************************

## Redoing the function for correct(?) sampling

I need to use rbinom for sampling the reads, not the size of the reads... And it should be the same sample size for all of the bulks, so reducing them down to a coverage of the same number.


```{r}
load("logregParental.Rdata")

logregParental %>% group_by(bulk, parent) %>% summarize(sum = sum(value)) %>% ungroup -> bulksums
logregParental %>% left_join(bulksums, by = c("bulk", "parent")) -> lrPbulksums
lrPbulksums$perc <- lrPbulksums$value/lrPbulksums$sum

perChromWindow_binom_redone <- function(lrP, chr = "II", windowsize = 100, coverage = 15634092/2, seedt = Sys.time()){
  lrP <- lrP[lrP$CHROM == chr,]
  
  set.seed(seedt)
  
  for(i in 1:length(lrP$CHROM)){
    lrP$subs[i] <- rbinom(1, coverage, prob = lrP$perc[i]) + 1
  }
  
  #Run the glm on that chromosome
  Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
    res <- glm(reads ~ bulk*parent, weights = subs, family = binomial, data = lrP[lrP$POS == i,])
    c(chr, i, res$coefficients)
  }
  
  #Format the results for the next function
  Results <- as.data.frame(Results)
  colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh", "parentWine", "Interaction")
  Results[,2] <- as.numeric(Results[,2])
  Results[,3] <- as.numeric(Results[,3])
  Results[,4] <- as.numeric(Results[,4])
  Results[,5] <- as.numeric(Results[,5])
  Results[,6] <- as.numeric(Results[,6])
  
  Results %>% arrange(POS) %>% select(-Intercept) -> Results

  
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  return(WResult)
}

ChrII <- perChromWindow_binom_redone(lrP = lrPbulksums)


ggplot(ChrII, aes(x = POS, y = bulk_Zprime)) + geom_line()
```

```{r}
#Chromosome Seven to check - 50% of data
ChrVII_List_50_redone <- list()
for(i in 1:100){
  ChrVII_List_50_redone[[i]] <- perChromWindow_binom_redone(lrPbulksums, chr = "VII", seedt = i) 
}

#save(ChrVII_List_50, file = "ChrVII_List_50.Rdata")
ChrVII_100_50percr <- bind_rows(ChrVII_List_50_redone, .id = "Index")
#save(ChrVII_100_50percr, file = "ChrVII_100_50percr.Rdata")

```

```{r}

load("ChrVII_100_50percr.Rdata")

ChrVII_100_50percr %>% ggplot(., aes(x = POS, y = bulk_Zprime, color = Index)) + 
  geom_line(alpha = 0.2) + theme(legend.position = "none")
ChrVII_100_50percr %>% ggplot(., aes(x = POS, y = parent_Zprime, color = Index)) + 
  geom_line(alpha = 0.2) + theme(legend.position = "none")
ChrVII_100_50percr %>% ggplot(., aes(x = POS, y = Interaction_Zprime, color = Index)) + 
  geom_line(alpha = 0.2) + theme(legend.position = "none")

ChrVII_100_50percr %>% group_by(POS) %>% summarise(meanr = mean(bulk_Zprime),
                                           sdr = sd(bulk_Zprime)) %>% 
  ggplot(., aes(x = POS, y = meanr, ymin = meanr-sdr, ymax = meanr+sdr)) + 
  #ylim(-1, 0.2) +
  geom_linerange(color = "violet")+ geom_line() +
  #ggtitle("10 Samples of Smoothed 0.1x") +
  ylab("Bulk Zprime") 
```

Looking at the change in where the peak actually lies

```{r}
# ChrVII_100_50percr %>% group_by(Index) %>% summarize(max_bulk_zprime = max(bulk_Zprime),
#                                                      max_parent_zprime = max(parent_Zprime),
#                                                      max_int_zprime = max(Interaction_Zprime)) -> maxvalues

maxChrVII <- data.frame(Index = unique(ChrVII_100_50percr$Index),
                       max_bulk_zprime = NA,
                       max_parent_zprime = NA,
                       max_int_zprime = NA)

for(i in 1:100){
  isubset <- subset(ChrVII_100_50percr, Index == as.character(i))
  maxChrVII$max_bulk_zprime[i] <- isubset$POS[abs(isubset$bulk_Zprime) == max(abs(isubset$bulk_Zprime))]
  maxChrVII$max_parent_zprime[i] <- isubset$POS[abs(isubset$parent_Zprime) == max(abs(isubset$parent_Zprime))]
  maxChrVII$max_int_zprime[i] <- isubset$POS[abs(isubset$Interaction_Zprime) == max(abs(isubset$Interaction_Zprime))]
  rm(isubset)
}

head(maxChrVII)

maxChrVII %>% pivot_longer(-Index) %>% ggplot(., aes(x = value, y = name)) + geom_boxplot() + geom_jitter(aes(color = name), alpha = 0.5) + theme(legend.position = "none") + ylab("") + xlab("Position") + ggtitle("Chr VII Max Abs(Zprime) Values")
```
Putting these plots together

```{r}
load("ChrVII_100_50percr.Rdata")

ChrVII_100_50percr %>% ggplot(., aes(x = POS, y = bulk_Zprime, color = Index)) + 
  geom_line(alpha = 0.2) + theme(legend.position = "none") -> pbulkz


ChrVII_100_50percr %>% ggplot(., aes(x = POS, y = parent_Zprime, color = Index)) + 
  geom_line(alpha = 0.2) + theme(legend.position = "none") -> pparentz

ChrVII_100_50percr %>% ggplot(., aes(x = POS, y = Interaction_Zprime, color = Index)) + 
  geom_line(alpha = 0.2) + theme(legend.position = "none") -> pintz

ggplot(maxChrVII, aes(x = max_bulk_zprime, y = 1)) + geom_boxplot() + geom_jitter(alpha = 0.5) + theme(legend.position = "none") + ylab("") + xlab("Position") + xlim(min(ChrVII_100_50percr$POS), max(ChrVII_100_50percr$POS)) -> pbulkmax

ggplot(maxChrVII, aes(x = max_parent_zprime, y = 1)) + geom_boxplot() + geom_jitter(alpha = 0.5) + theme(legend.position = "none") + ylab("") + xlab("Position") + xlim(min(ChrVII_100_50percr$POS), max(ChrVII_100_50percr$POS)) -> pparentmax

ggplot(maxChrVII, aes(x = max_int_zprime, y = 1)) + geom_boxplot() + geom_jitter(alpha = 0.5) + theme(legend.position = "none") + ylab("") + xlab("Position") + xlim(min(ChrVII_100_50percr$POS), max(ChrVII_100_50percr$POS)) -> pintmax

plot_grid(pbulkz, pbulkmax, nrow = 2)
plot_grid(pparentz, pparentmax, nrow = 2)
plot_grid(pintz, pintmax, nrow = 2)


```

