---
title: "Permutations for CSSI Fluconazole"
author: "Cassandra Buzby"
date: "9/7/2022"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

#Load packages
library(tidyr)
#library(tidyverse)
library(reshape2)
library(cowplot)
library(dplyr)
library(ggplot2)
library(foreach)
library(doParallel)

ggplot2::theme_set(theme_light())
#ggplot2::theme_set(theme_bw())


CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47",
                  "#F26430", "#0A369D", "#7EA3CC")


library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

setwd("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis")
# load("RdataFiles/SAUA_4-13-22.Rdata")
# load("RdataFiles/SCUC_4-13-22.Rdata")
# load("RdataFiles/bsa_glm_results.Rdata")

#rawData = "HGV.SortedCat.vcf.output.table"
rawData = "HGVMVDRX2.SortedCat.vcf.output.table"

load("RdataFiles/CSSI_Fluc_concat.Rdata")

```

Test the following

```{r}
# lrP <- CSSI_Fluc_concat
# chr <- "III"
# c <- "III"
# windowsize <- 100
```


## Permute the loaded in file to remove bulk effects

THIS DOESN'T WORK BECAUSE OF THE TABLE FUNCTION - need a different setup of the initial dataframe to do it.

Make a single script that can be arrayed on the HPC:

```{r, eval = FALSE}
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

#Load Libraries
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(dplyr)
library(foreach)
library(doParallel)

###################################################################################
#Load Functions
#FUNCTION
BSA_GLM_z <-  function(lrP, chr = "II", windowsize = 100){
  
  lrP <- subset(lrP, CHROM == chr)
  
  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- glm(Allele ~ Bulk*Parent, weights = value, family = binomial, data = lrP[lrP$POS == i,])
      c(c, i, summary(res)$coefficients[10:12])
    }
  
    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", "bulkHigh", "parentWine", "Interaction")
    Results[,2] <- as.numeric(Results[,2])
    Results[,3] <- as.numeric(Results[,3])
    Results[,4] <- as.numeric(Results[,4])
    Results[,5] <- as.numeric(Results[,5])
    
    Results %>% arrange(POS) -> Results
  }
  
 
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  return(WResult)
}

tableinv <- function(x){
      y <- x[rep(rownames(x),x$value),1:(ncol(x))]
      rownames(y) <- c(1:nrow(y))
      return(y)}

ShuffleLoop <- function(GLM_VIII, iterations = 2){
  
  iterativelist <- list()
  
  for(x in 1:iterations){
    #x <- 1
    #GLM_VIII <- subset(CSSI_Fluc_Pivot, CHROM == "I")
    set.seed(Sys.time())
    iterativelist[[x]] <- foreach (i=unique(GLM_VIII$POS), .combine=rbind) %dopar% {
      GLM_VIII_inv <- tableinv(subset(GLM_VIII, POS == i)) %>% select(-value, -ReadCount)
      GLM_VIII_inv$bulk <- sample(unique(GLM_VIII_inv$bulk), replace = FALSE)
      GLM_VIII_inv %>% group_by(CHROM, POS, Parent, Bulk, Allele, Type, Dataset) %>% count() -> result
      result
    }
    
  }
  finalresult <- bind_rows(iterativelist, .id = "Index") 
  return(finalresult)
}

###################################################################################
#Load Data
# load("RdataFiles/logregParental.Rdata")
load("RdataFiles/CSSI_Fluc_Pivot_2.Rdata")

#Make the subsequent functions work
CSSI_Fluc_Pivot$value <- CSSI_Fluc_Pivot$ReadCount

# str(CSSI_Fluc_Pivot)
# str(logregParental)
#i <- 1

#Run Script

VIII_it_100 <- list()
VIII_shuffle_100it <- list()
for(i in 1:100){
  VIII_it_100[[i]] <- ShuffleLoop(GLM_VIII = subset(CSSI_Fluc_Pivot, CHROM == args[1]), iterations = 1)
  #VIII_it_100[[i]] <- ShuffleLoop(GLM_VIII = subset(CSSI_Fluc_Pivot, CHROM == "I"), iterations = 1)
  VIII_it_100[[i]]$value <- VIII_it_100[[i]]$n
  VIII_shuffle_100it[[i]] <- BSA_GLM_z(subset(VIII_it_100[[i]]), chr = args[1])
  #VIII_shuffle_100it[[i]] <- BSA_GLM_z(subset(VIII_it_100[[i]]), chr = "I")

  }

Shuffled_100 <- bind_rows(VIII_shuffle_100it, .id="Index")

save(Shuffled_100, file = paste("Shuffled_100_FlucI_", args[1], ".Rdata", sep = ""))
```


```{r}
subset(logregParental, POS == 100153)
tableinv(subset(logregParental, POS == 100153))

CSSI_Fluc_Pivot$value <- CSSI_Fluc_Pivot$ReadCount
subset(CSSI_Fluc_Pivot, POS == 925)
tableinv(subset(CSSI_Fluc_Pivot, POS == 925))

```

Plot to check

```{r, eval = FALSE}

Shuffled_100 %>% ggplot(aes(x = POS, y = bulk_Zprime)) + 
      # geom_vline(data = JaroszKey, aes(xintercept = POS), color = "black", alpha = 0.5, size = 0.7) +
      # geom_vline(data = InteractionNetwork, aes(xintercept = POS), color = "red", alpha = 1, size = 0.7, linetype = "dashed") +
      # 
  geom_point(color = "#345F6F", size = 0.2, alpha = 0.8) + 
  geom_point(aes(x = POS, y = parent_Zprime), color = "#FFB05C", alpha= 0.8, size = 0.2) +
  geom_point(aes(x = POS, y = Interaction_Zprime), color = "#D7335C", alpha= 0.8, size = 0.2) +
  #geom_point(aes(x = POS, y = Day_Zprime), color = "gray", alpha= 0.8, size = 0.2) +
  # scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  # facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() +
  # 
  ggtitle("CSSI Fluconazole | Comparison of Smoothed Z scores") + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  xlab("Genomic Position") + ylab("Oak  <--    Z Score  -->  Wine")
```

## Load in permuted data to save

```{r, eval = FALSE}
load("Shuffled_100_FlucI_III.Rdata")
Shuffled_100_III <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_II.Rdata")
Shuffled_100_II <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_IV.Rdata")
Shuffled_100_IV <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_V.Rdata")
Shuffled_100_V <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_VI.Rdata")
Shuffled_100_VI <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_VII.Rdata")
Shuffled_100_VII <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_VIII.Rdata")
Shuffled_100_VIII <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_IX.Rdata")
Shuffled_100_IX <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_X.Rdata")
Shuffled_100_X <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_XI.Rdata")
Shuffled_100_XI <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_XII.Rdata")
Shuffled_100_XII <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_XIII.Rdata")
Shuffled_100_XIII <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_XIV.Rdata")
Shuffled_100_XIV <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_XV.Rdata")
Shuffled_100_XV <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_FlucI_XVI.Rdata")
Shuffled_100_XVI <- Shuffled_100
rm(Shuffled_100)

Shuffled_100_bz_CSSI_F_perm <- rbind(Shuffled_100_I, Shuffled_100_III,Shuffled_100_II,Shuffled_100_IV,Shuffled_100_V, 
                          Shuffled_100_VI, Shuffled_100_VII, Shuffled_100_VIII,Shuffled_100_IX, 
                          Shuffled_100_X, Shuffled_100_XI, Shuffled_100_XII, Shuffled_100_XIII,
                          Shuffled_100_XIV,Shuffled_100_XV,Shuffled_100_XVI)

Shuffled_100_bz_CSSI_F_perm$CHROM <- factor( Shuffled_100_bz_CSSI_F_perm$CHROM, levels = ChromKey$chromosomes)
save(Shuffled_100_bz_CSSI_F_perm, file = "Shuffled_100_bz_CSSI_F_perm.Rdata")

rm(Shuffled_100_I, Shuffled_100_III,Shuffled_100_II,Shuffled_100_IV,Shuffled_100_V, 
                          Shuffled_100_VI, Shuffled_100_VII, Shuffled_100_VIII,Shuffled_100_IX, 
                          Shuffled_100_X, Shuffled_100_XI, Shuffled_100_XII, Shuffled_100_XIII,
                          Shuffled_100_XIV,Shuffled_100_XV,Shuffled_100_XVI)
```

## Make a script for permuting parental effects

```{r, eval = FALSE}
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

#Load Libraries
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(dplyr)
library(foreach)
library(doParallel)

###################################################################################
#Load Functions
#FUNCTION
BSA_GLM_z <-  function(lrP, chr = "II", windowsize = 100){
  
  lrP <- subset(lrP, CHROM == chr)
  
  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- glm(Allele ~ Bulk*Parent, weights = value, family = binomial, data = lrP[lrP$POS == i,])
      c(c, i, summary(res)$coefficients[10:12])
    }
  
    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", "bulkHigh", "parentWine", "Interaction")
    Results[,2] <- as.numeric(Results[,2])
    Results[,3] <- as.numeric(Results[,3])
    Results[,4] <- as.numeric(Results[,4])
    Results[,5] <- as.numeric(Results[,5])
    
    Results %>% arrange(POS) -> Results
  }
  
 
  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  
  return(WResult)
}

tableinv <- function(x){
      y <- x[rep(rownames(x),x$value),1:(ncol(x))]
      rownames(y) <- c(1:nrow(y))
      return(y)}

ShuffleLoop <- function(GLM_VIII, iterations = 2){
  
  iterativelist <- list()
  
  for(x in 1:iterations){
    #x <- 1
    #GLM_VIII <- subset(CSSI_Fluc_Pivot, CHROM == "I")
    set.seed(Sys.time())
    iterativelist[[x]] <- foreach (i=unique(GLM_VIII$POS), .combine=rbind) %dopar% {
      GLM_VIII_inv <- tableinv(subset(GLM_VIII, POS == i)) %>% select(-value, -ReadCount)
      GLM_VIII_inv$Parent <- sample(unique(GLM_VIII_inv$Parent), replace = FALSE)
      GLM_VIII_inv %>% group_by(CHROM, POS, Parent, Bulk, Allele, Type, Dataset) %>% count() -> result
      result
    }
    
  }
  finalresult <- bind_rows(iterativelist, .id = "Index") 
  return(finalresult)
}

###################################################################################
#Load Data
# load("RdataFiles/logregParental.Rdata")
load("RdataFiles/CSSI_Fluc_Pivot_2.Rdata")

#Make the subsequent functions work
CSSI_Fluc_Pivot$value <- CSSI_Fluc_Pivot$ReadCount

# str(CSSI_Fluc_Pivot)
# str(logregParental)
#i <- 1

#Run Script

VIII_it_100 <- list()
VIII_shuffle_100it <- list()
for(i in 1:100){
  VIII_it_100[[i]] <- ShuffleLoop(GLM_VIII = subset(CSSI_Fluc_Pivot, CHROM == args[1]), iterations = 1)
  #VIII_it_100[[i]] <- ShuffleLoop(GLM_VIII = subset(CSSI_Fluc_Pivot, CHROM == "I"), iterations = 1)
  VIII_it_100[[i]]$value <- VIII_it_100[[i]]$n
  VIII_shuffle_100it[[i]] <- BSA_GLM_z(subset(VIII_it_100[[i]]), chr = args[1])
  #VIII_shuffle_100it[[i]] <- BSA_GLM_z(subset(VIII_it_100[[i]]), chr = "I")

  }

Shuffled_100 <- bind_rows(VIII_shuffle_100it, .id="Index")

save(Shuffled_100, file = paste("Shuffled_100_pFlucI_", args[1], ".Rdata", sep = ""))
```

## Load in permuted data to save

```{r, eval = FALSE}

load("Shuffled_100_pFlucI_I.Rdata")
Shuffled_100_I <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_III.Rdata")
Shuffled_100_III <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_II.Rdata")
Shuffled_100_II <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_IV.Rdata")
Shuffled_100_IV <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_V.Rdata")
Shuffled_100_V <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_VI.Rdata")
Shuffled_100_VI <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_VII.Rdata")
Shuffled_100_VII <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_VIII.Rdata")
Shuffled_100_VIII <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_IX.Rdata")
Shuffled_100_IX <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_X.Rdata")
Shuffled_100_X <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_XI.Rdata")
Shuffled_100_XI <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_XII.Rdata")
Shuffled_100_XII <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_XIII.Rdata")
Shuffled_100_XIII <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_XIV.Rdata")
Shuffled_100_XIV <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_XV.Rdata")
Shuffled_100_XV <- Shuffled_100
rm(Shuffled_100)

load("Shuffled_100_pFlucI_XVI.Rdata")
Shuffled_100_XVI <- Shuffled_100
rm(Shuffled_100)

Shuffled_100_pz_CSSI_F_perm <- rbind(Shuffled_100_I, Shuffled_100_III,Shuffled_100_II,Shuffled_100_IV,Shuffled_100_V, 
                          Shuffled_100_VI, Shuffled_100_VII, Shuffled_100_VIII,Shuffled_100_IX, 
                          Shuffled_100_X, Shuffled_100_XI, Shuffled_100_XII, Shuffled_100_XIII,
                          Shuffled_100_XIV,Shuffled_100_XV,Shuffled_100_XVI)

Shuffled_100_pz_CSSI_F_perm$CHROM <- factor( Shuffled_100_pz_CSSI_F_perm$CHROM, levels = ChromKey$chromosomes)
save(Shuffled_100_pz_CSSI_F_perm, file = "Shuffled_100_pz_CSSI_F_perm.Rdata")

rm(Shuffled_100_I, Shuffled_100_III,Shuffled_100_II,Shuffled_100_IV,Shuffled_100_V, 
                          Shuffled_100_VI, Shuffled_100_VII, Shuffled_100_VIII,Shuffled_100_IX, 
                          Shuffled_100_X, Shuffled_100_XI, Shuffled_100_XII, Shuffled_100_XIII,
                          Shuffled_100_XIV,Shuffled_100_XV,Shuffled_100_XVI)

```

## Making a more maleable script for running simulations

(base) [cb4097@log-2 SimShuffledParent_Fluc_CSSI]$ vi RunGLMPermutation.R

```{r, eval = FALSE}
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

#Load Libraries
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(dplyr)
library(foreach)
library(doParallel)

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"),
                       num = as.character(1:17))

cmd_args <- commandArgs(TRUE)

argchr <- ChromKey$chromosomes[ChromKey$num == args[1]]

if(length(cmd_args) < 3){
  Folder <- "None"
  print("No Folder Created")
}else{
  Folder <- args[3]
  print(Folder)
  if(file.exists(Folder)){
    print("Using existing folder")
  }else{
    dir.create(Folder)
    print("New folder created")
  }
}
#to test
#argchr <- "III"

###################################################################################
#Load Functions
#FUNCTION
BSA_GLM_CSSIF <-  function(lrP, chr = "II", windowsize = 100){

  lrP <- subset(lrP, CHROM == chr)

  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- glm(Allele ~ Bulk*Parent+Day, weights = value, family = binomial, data = lrP[lrP$POS == i,])
      c(c, i, summary(res)$coefficients[11:15])
    }

    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkFluconazole", "parentWine", "Day2", "Interaction")
    Results[,2] <- as.numeric(Results[,2])
    Results[,4] <- as.numeric(Results[,4])
    Results[,5] <- as.numeric(Results[,5])
    Results[,6] <- as.numeric(Results[,6])
    Results[,7] <- as.numeric(Results[,7])

    Results %>% arrange(POS) %>% select(-Intercept) -> Results
  }


  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkFluconazole[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                res_day <- mean(Results$Day2[(i-windowsize):(i+windowsize)])

                #print CHROM, index, POS, and mean
                c(chr, i, Results$POS[i],
                  Results$bulkFluconazole[i], Results$parentWine[i],Results$Interaction[i], Results$Day2[i],
                  res_bulk, res_parent, res_day, res_int)}

  WResult <- as.data.frame(WResult)

  colnames(WResult) <- c("CHROM", "Index", "POS",
                         "Bulk_Z", "Parent_Z", "Interaction_Z", "Day_Z",
                         "Bulk_Zprime", "Parent_Zprime", "Day_Zprime","Interaction_Zprime")

  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])
  WResult[,10] <- as.numeric(WResult[,10])
  WResult[,11] <- as.numeric(WResult[,11])

  return(WResult)
}

BSA_GLM_z <-  function(lrP, chr = "II", windowsize = 100){

  lrP <- subset(lrP, CHROM == chr)

  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- glm(Allele ~ Bulk*Parent, weights = value, family = binomial, data = lrP[lrP$POS == i,])
      c(c, i, summary(res)$coefficients[10:12])
    }

    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", "bulkHigh", "parentWine", "Interaction")
    Results[,2] <- as.numeric(Results[,2])
    Results[,3] <- as.numeric(Results[,3])
    Results[,4] <- as.numeric(Results[,4])
    Results[,5] <- as.numeric(Results[,5])

    Results %>% arrange(POS) -> Results
  }


  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_int <- mean(Results$Interaction[(i-windowsize):(i+windowsize)])
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                res_parent <- mean(Results$parentWine[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], Results$parentWine[i],Results$Interaction[i],
                  res_bulk, res_parent, res_int)}
  WResult <- as.data.frame(WResult)
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "parent_Z", "Interaction_Z",
                         "bulk_Zprime", "parent_Zprime", "Interaction_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])
  WResult[,6] <- as.numeric(WResult[,6])
  WResult[,7] <- as.numeric(WResult[,7])
  WResult[,8] <- as.numeric(WResult[,8])
  WResult[,9] <- as.numeric(WResult[,9])

  return(WResult)
}

tableinv <- function(x){
      y <- x[rep(rownames(x),x$value),1:(ncol(x))]
      rownames(y) <- c(1:nrow(y))
      return(y)}

ShuffleLoop_Day <- function(GLM_VIII, iterations = 2, COL = "Bulk"){
    iterativelist <- list()
    for(x in 1:iterations){
      #x <- 1
      #GLM_VIII <- subset(CSSI_Fluc_Pivot, CHROM == "I")
      set.seed(Sys.time())
      iterativelist[[x]] <- foreach (i=unique(GLM_VIII$POS), .combine=rbind) %dopar% {
        GLM_VIII_inv <- tableinv(subset(GLM_VIII, POS == i)) %>% select(-value, -ReadCount)
        GLM_VIII_inv[[COL]] <- sample(GLM_VIII_inv[[COL]], replace = FALSE)
        GLM_VIII_inv %>% group_by(CHROM, POS, Parent, Bulk, Day, Allele, Type, Dataset) %>% count() -> result
        result
        }

      }
      finalresult <- bind_rows(iterativelist, .id = "Index")
  return(finalresult)
}

###################################################################################
#Load Data
load("RdataFiles/CSSI_Fluc_Pivot_2.Rdata")

#Make the subsequent functions work
CSSI_Fluc_Pivot$value <- CSSI_Fluc_Pivot$ReadCount

#Print Notices
print(paste("Input Chromosome: ", args[1], " corresponding to ", argchr, sep = ""))

if(args[2] %in% colnames(CSSI_Fluc_Pivot)){
  print(paste("Using column ", args[2], " out of:", sep = ""))
  print(colnames(CSSI_Fluc_Pivot))
}else{
  print("Incorrect Column Selected")
}

#Run Script
VIII_it_100 <- list()
VIII_shuffle_100it <- list()
for(i in 1:100){
  VIII_it_100[[i]] <- ShuffleLoop_Day(GLM_VIII = subset(CSSI_Fluc_Pivot, CHROM == argchr), iterations = 1, COL = args[2])
  VIII_it_100[[i]]$value <- VIII_it_100[[i]]$n
  VIII_shuffle_100it[[i]] <- BSA_GLM_CSSIF(subset(VIII_it_100[[i]]), chr = argchr)
}

Shuffled_100 <- bind_rows(VIII_shuffle_100it, .id="Index")

if(Folder == "None"){
  save(Shuffled_100, file = paste(args[2], "_permute100_FI_", argchr, ".Rdata", sep = ""))
}else{
  save(Shuffled_100, file = paste(Folder, "/", args[2], "_permute100_FI_", argchr, ".Rdata", sep = ""))
}

```

Testing how to make the columns work

```{r}
testdf <- data.frame(A = 1:10, B = 1:10, C = 1:10)

testdf$A

selectcol <- function(D = "A", df = testdf){
  print(testdf[[D]])
  D %in% colnames(df)
}

selectcol("A", testdf)

file.exists("Selection")
dir()
```
## HPC Script to bind them all together into the correct file
```{r}
args <- list("Input")
args[1]
```

```{r, eval = FALSE}
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

#Load Libraries
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(dplyr)
library(foreach)
library(doParallel)

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

includedchromosomes <- data.frame(chr = ChromKey$chromosomes, included = "NotIncluded")
for(i in args[2:length(args)]){
  #i <- "RdataFiles/Shuffled_100_III.Rdata"
  load(i)
  newname <- unique(Shuffled_100$CHROM)
  includedchromosomes$included[includedchromosomes$chr == newname] <- "Included"
  assign(newname, Shuffled_100)
  rm(Shuffled_100)
}

combineddfs <- rbind(I, II, III, IV, V, VI, VII, VIII, IX, X, XI, XII, XIII, XIV, XV, XVI)

#combineddfs <- rbind(II, III)
#length(unique(combineddfs$CHROM))

combineddfs$CHROM <- factor(combineddfs$CHROM, levels = ChromKey$chromosomes)

# assign(as.character(args[1]), combineddfs)
# save(args[1], file = paste(as.character(args[1]), ".Rdata", sep = ""))
# save(get(as.character(args[1])), file = paste(as.character(args[1]), ".Rdata", sep = ""))

saveRDS(combineddfs, file = paste(as.character(args[1]), ".rds", sep = ""))

NewInput <- readRDS("Input.rds")
```

## Loading this stuff in

```{r}
bulkperm <- readRDS("RdataFiles/CSSIFluc_BulkPermd.rds")

#PERMUTED BULK
bulkperm %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Bulk Effects with Permuted Bulk")

bulkperm %>% ggplot(aes(x = POS, y = Parent_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Parent Effects with Permuted Bulk")

bulkperm %>% ggplot(aes(x = POS, y = Interaction_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Interactions with Permuted Bulk")

bulkperm %>% ggplot(aes(x = POS, y = Day_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Day Effects with Permuted Bulk")
```

```{r}
parentperm <- readRDS("RdataFiles/CSSIFluc_ParentPermd.rds")

#PERMUTED PARENT
parentperm %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Bulk Effects with Permuted Parent")

parentperm %>% ggplot(aes(x = POS, y = Parent_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Parent Effects with Permuted Parent")

parentperm %>% ggplot(aes(x = POS, y = Interaction_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Interactions with Permuted Parent")

parentperm %>% ggplot(aes(x = POS, y = Day_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Day Effects with Permuted Parent")
```

```{r}
dayperm <- readRDS("RdataFiles/CSSIFluc_DayPermd.rds")

#PERMUTED DAY
dayperm %>% ggplot(aes(x = POS, y = Bulk_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Bulk Effects with Permuted Day")

dayperm %>% ggplot(aes(x = POS, y = Parent_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Parent Effects with Permuted Day")

dayperm %>% ggplot(aes(x = POS, y = Interaction_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Interactions with Permuted Day")

dayperm %>% ggplot(aes(x = POS, y = Day_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(CSSI_F_ZScores_GLM$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("Day Effects with Permuted Day")
```


## Finding the q values of each shuffled bulk

```{r}
bulkperm <- readRDS("RdataFiles/CSSIFluc_BulkPermd.rds")
dayperm <- readRDS("RdataFiles/CSSIFluc_DayPermd.rds")
parentperm <- readRDS("RdataFiles/CSSIFluc_ParentPermd.rds")

bulkperm %>%  mutate(pval = pnorm(abs(Bulk_Zprime), lower.tail = F)*2) %>% arrange(pval) %>% mutate(Rank = length(pval):1) -> bulk_p
parentperm %>%  mutate(pval = pnorm(abs(Parent_Zprime), lower.tail = F)*2) %>% arrange(pval) %>% mutate(Rank = length(pval):1) -> parent_p
dayperm %>%  mutate(pval = pnorm(abs(Day_Zprime), lower.tail = F)*2) %>% arrange(pval) %>% mutate(Rank = length(pval):1) -> day_p

#bulk_p %>% group_by(Index) %>% arrange(pval) %>% summarise(quantile(0.05)) -> bulk_p_summary

bulk_p %>% ggplot(aes(x = pval, y = Rank, color = Index)) + geom_line(alpha = 0.1) + theme(legend.position = "none")

#Okay so now pick a cutoff and get the number of positives
qs <- quantile(bulk_p$Rank, c(0.05, 0.95))

bulk_p[bulk_p$Rank == 150306,]
bulk_p[bulk_p$Rank == 2855795,]

#So I want to know if I have a cutoff of 0.8, for example, how many are positive each time?
length(which(bulk_p$pval < 0.8))
```

# Testing half of each null with other half vs other null

The idea of this test is the compare a sub-sampled experiment with itself to identify the null distribution of effects, and then to compare this sub-sample to a half sample of a replicate to identify if this replicate is from the same distribution. This may help to provide a cutoff from the other nulls (ie if it's different by a certain amount, or if it's the same).

To start off, I want to ONLY compare two bulks (nulls) or one bulk (two sub-samples) to see how similar they are, which means that I should limit the comparison to two bulks and also choose which bulks ahead of time to sample.

### Update: changing the code to only sample half and half without replacement

The original is on the HPC and also github

```{r, eval = FALSE}
#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

#Load Libraries
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(dplyr)
library(foreach)
library(doParallel)

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"),
                       num = as.character(1:17))

cmd_args <- commandArgs(TRUE)

#args <- c("ChromosomeAsNumber", "Column", "Folder")
#args <- c("ChromAsNumber", "Comp1Dataset", "Comp2Dataset", "Folder")

argchr <- ChromKey$chromosomes[ChromKey$num == args[1]]

if(length(cmd_args) < 4){
  Folder <- "None"
  print("No Folder Created")
}else{
  Folder <- args[4]
  print(Folder)
  if(file.exists(Folder)){
    print("Using existing folder")
  }else{
    dir.create(Folder)
    print("New folder created")
  }
}
#to test
#argchr <- "III"

###################################################################################
#Load Functions
#FUNCTION

BSA_GLM_z_sub <-  function(lrP, chr = unique(lrP$CHROM)[1], windowsize = 100){

  lrP <- subset(lrP, CHROM == chr)

  AllResults <- list()
  for(c in unique(lrP$CHROM)){
    lrP <- subset(lrP, CHROM == c)
    #Run the glm on that chromosome
    Results <- foreach (i=unique(lrP$POS), .combine=rbind) %dopar% {
      res <- glm(Allele ~ IDcol, weights = n, family = binomial, data = lrP[lrP$POS == i,])
      c(c, i, summary(res)$coefficients[5:6])
    }

    #Format the results for the next function
    Results <- as.data.frame(Results)
    colnames(Results) <- c("CHROM", "POS", "Intercept", "bulkHigh")
    Results[,2] <- as.numeric(Results[,2])
    Results[,4] <- as.numeric(Results[,4])

    Results %>% arrange(POS) -> Results
  }


  #Run the windows on that chromosome
  WResult <- foreach(i=windowsize+1:(length(Results$POS) - (2*windowsize)), .combine=rbind) %dopar% {
            #save the median value from original index 1 to original end index
                res_bulk <- mean(Results$bulkHigh[(i-windowsize):(i+windowsize)])
                #print CHROM, index, POS, and median
                c(chr, i, Results$POS[i], Results$bulkHigh[i], res_bulk)}
  
  WResult <- as.data.frame(WResult)
  
  colnames(WResult) <- c("CHROM", "Index", "POS", "bulk_Z", "bulk_Zprime")
  WResult[,2] <- as.numeric(WResult[,2])
  WResult[,3] <- as.numeric(WResult[,3])
  WResult[,4] <- as.numeric(WResult[,4])
  WResult[,5] <- as.numeric(WResult[,5])

  return(WResult)
}

tableinv <- function(x, COL = "value"){
      y <- x[rep(rownames(x),x[[COL]]),1:(ncol(x))]
      rownames(y) <- c(1:nrow(y))
      return(y)}

#GLM_VIII <- CSSI_Fluc_Pivot
#GLM_VIII <- subset(GLM_VIII, POS < 2000)
#unique(GLM_VIII$POS)
#i <- 925

SubSample <- function(GLM_VIII, SubProportion = 0.5, BulkName = unique(GLM_VIII$Dataset)[1]){
      iterativelist <- foreach (i=unique(GLM_VIII$POS), .combine=rbind) %dopar% {
        
          #make a table for all of the reads of that position
          GLM_VIII_inv <- tableinv(subset(GLM_VIII, POS == i), COL = "ReadCount")
          
          #sample half the amount of data (or whatever proportion is given)
          GLM_VIII_inv_sampled <- GLM_VIII_inv[sample(x = 1:dim(GLM_VIII_inv)[1],
                                        size = round(dim(GLM_VIII_inv)[1]*SubProportion)),]
          
          GLM_VIII_inv_sampled$IDcol <- BulkName
          
          #add them back in 
          GLM_VIII_inv_sampled %>% group_by(CHROM, POS, Allele, IDcol) %>% count() -> result
          result
      }
      
      finalresult <- bind_rows(iterativelist)
      finalresult$ReadCount <- finalresult$n
      
      return(finalresult)

}

###################################################################################
#Load Data
load("RdataFiles/CSSI_Fluc_Pivot_2.Rdata")

#Make the subsequent functions work
CSSI_Fluc_Pivot$value <- CSSI_Fluc_Pivot$ReadCount

#argchr <- "I"
#Print Notices
print(paste("Input Chromosome: ", args[1], " corresponding to ", argchr, sep = ""))

#Run Script
VIII_it_100 <- list()
VIII_shuffle_100it <- list()
for(i in 1:100){
  comp1 <- SubSample(subset(CSSI_Fluc_Pivot, CHROM == argchr & Dataset == args[2]), SubProportion = 0.5, BulkName = paste("1_",args[2], sep = ""))
  comp2 <- SubSample(subset(CSSI_Fluc_Pivot, CHROM == argchr & Dataset == args[3]), SubProportion = 0.5, BulkName = paste("2_",args[3], sep = ""))
  VIII_it_100[[i]] <- rbind(comp1, comp2)
  
  VIII_it_100[[i]]$value <- VIII_it_100[[i]]$n
  
  VIII_shuffle_100it[[i]] <- BSA_GLM_z_sub(VIII_it_100[[i]])
}

Shuffled_100 <- bind_rows(VIII_shuffle_100it, .id="Index")

if(Folder == "None"){
  save(Shuffled_100, file = paste(args[2], args[3], "_SubSample_FI_", argchr, ".Rdata", sep = ""))
}else{
  save(Shuffled_100, file = paste(Folder, "/", args[2], args[3], "_SubSample_FI_", argchr, ".Rdata", sep = ""))
}

```

Load in data to process

```{r}
WineIFluc_D <- readRDS("RdataFiles/WineIFluc_D.rds")
OakIFluc_B <- readRDS("RdataFiles/OakIFluc_B.rds")
OakIDilute_B <- readRDS("RdataFiles/OakIDilute_B.rds")
WineIDilute_BD <- readRDS("RdataFiles/WineIDilute_BD.rds")
WineIDilute_B <- readRDS("RdataFiles/WineIDilute_B.rds")
WineIDilute_C <- readRDS("RdataFiles/WineIDilute_C.rds")

# unique(WineIFluc_D$CHROM)
# head(WineIDilute_B)
```

```{r}
WineIDilute_BD %>% ggplot(aes(x = POS, y = bulk_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(WineIDilute_BD$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("SubSampled Wine Dilute B vs D")

WineIDilute_B %>% ggplot(aes(x = POS, y = bulk_Zprime, color = Index)) + geom_line(alpha = 0.2) + scale_x_continuous(breaks = seq(from = 0, to = max(WineIDilute_B$POS), by = 1e5), name = "Genomic Position") +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + theme_minimal() + theme(legend.position = "none", axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + ggtitle("SubSampled Wine Dilute B")

```

Density of each

```{r}

WineIDilute_BD %>% ggplot(aes(x = bulk_Zprime))  +
  geom_density(data = OakIFluc_B, aes(x = bulk_Zprime), fill = "blue", alpha = 0.2, size = 1) +
  geom_density(data = OakIDilute_B, aes(x = bulk_Zprime), fill = "blue", alpha = 0.2, size = 1) +
  geom_density(data = WineIFluc_D, aes(x = bulk_Zprime), fill = "blue", alpha = 0.2, size = 1) +
  geom_density(data = WineIDilute_C, aes(x = bulk_Zprime), fill = "blue", alpha = 0.2, size = 1) +
  geom_density(data = WineIDilute_B, aes(x = bulk_Zprime), fill = "blue", alpha = 0.2, size = 1) + 
  geom_density(fill = "red", alpha = 0.2, size = 1)

```

```{r}
WineIDilute_BD %>% ggplot(aes(x = bulk_Zprime, color = Index))+ geom_density(alpha = 0.02) + theme(legend.position = "none")

```

Figuring out the confidence on the peaks

```{r}
#Find max values for each iteration

WineIDilute_BD %>% group_by(Index) %>% summarize(absmax = max(abs(bulk_Zprime)),
                                                 abs_95 = quantile(abs(bulk_Zprime), 0.975))-> Wine_BD_Max

WineIDilute_B %>% group_by(Index) %>% summarize(absmax = max(abs(bulk_Zprime)),
                                                 abs_95 = quantile(abs(bulk_Zprime), 0.975))-> Wine_B_Max

head(Wine_BD_Max)

Wine_BD_Max %>% ggplot(aes(x = absmax)) + geom_density(fill = "red", alpha = 0.2, size = 1) + 
  geom_density(aes(x = abs_95), fill = "gray", alpha = 0.2, size = 1) +
  geom_density(data = Wine_B_Max, aes(x = absmax), fill = "navyblue", 
               alpha = 0.2, size = 1, linetype = "dashed") +
  geom_density(data = Wine_B_Max, aes(x = abs_95), fill = "gray", 
               alpha = 0.2, size = 1, linetype = "dashed")+
  ggtitle("Comparison of Max (colored) and 95% (gray) of B vs BxD")

between_max <- mean(Wine_BD_Max$absmax)
between_95 <- mean(Wine_BD_Max$abs_95)
within_max <- mean(Wine_B_Max$absmax)
within_95 <- mean(Wine_B_Max$abs_95)


```

