---
title: "QTLseq Method of Bulk Segregant Analysis"
author: "Cassandra Buzby"
date: "01/17/2022"
output:
  html_document:
    df_print: paged
---

Keep in mind, use coolors.co for color palettes :)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(tidyverse)
library(reshape2)
ggplot2::theme_set(theme_light())

cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

chromPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7",
                  "#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC", "#4C4C47", "#92DCE5", 
                    "#C14953", "#2B2C28", "#FF9F1C",
                  "#8FC93A", "#F26430", "#0A369D", "#7EA3CC", 
                  "#4C4C47", "#C14953", "#2B2C28", "#FF9F1C")

library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

```

Following directions for installation

```{r, eval=FALSE}
install.packages("devtools")

##Updating Rtools from https://github.com/metrumresearchgroup/mrgsolve/issues/160 
# path <- Sys.getenv("PATH")
# path <- c("C:/rtools40/usr/bin", path) #changed from original verion to my path
# path <- paste(path,collapse=";")
# Sys.setenv(PATH=path)
# 
# Sys.which("make") #test that "make" is found in the right folder

#################
devtools::install_github("bmansfeld/QTLseqr")
```

## Example Data
```{r, eval = FALSE}
##Example Code

#load the package

#Set sample and file names
HighBulk <- "SRR834931"
LowBulk <- "SRR834927"
file <- "SNPs_from_GATK.table"

#Choose which chromosomes will be included in the analysis (i.e. exclude smaller contigs)
Chroms <- paste0(rep("Chr", 12), 1:12)

#Import SNP data from file
df <-importFromGATK(
        file = file,
        highBulk = HighBulk,
        lowBulk = LowBulk,
        chromList = Chroms
     )

df %>% merge(.,ChromKey) -> df

#Filter SNPs based on some criteria
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.20,
        minTotalDepth = 100,
        maxTotalDepth = 400,
        minSampleDepth = 40,
        minGQ = 99
    )


#Run G' analysis
df_filt <- runGprimeAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    outlierFilter = "deltaSNP")

#Run QTLseq analysis
df_filt <- runQTLseqAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    popStruc = "F2",
    bulkSize = c(25, 25),
    replications = 10000,
    intervals = c(95, 99)
)

#Plot
plotQTLStats(SNPset = df_filt, var = "Gprime", plotThreshold = TRUE, q = 0.01)
plotQTLStats(SNPset = df_filt, var = "deltaSNP", plotIntervals = TRUE)

#export summary CSV
getQTLTable(SNPset = df_filt, alpha = 0.01, export = TRUE, fileName = "my_BSA_QTL.csv")

```

```{r, eval = FALSE}
library("QTLseqr")

# devtools::install_github("bmansfeld/Yang2013data")
# library("Yang2013data")
###############################################################

#Import the data
rawData <- system.file(
"extdata",
"Yang_et_al_2013.table",
package = "Yang2013data",
mustWork = TRUE)

HighBulk <- "SRR834931"
LowBulk <- "SRR834927"
Chroms <- paste0(rep("Chr", 12), 1:12)

#import data
df <-
importFromGATK(
file = rawData,
highBulk = HighBulk,
lowBulk = LowBulk,
chromList = Chroms
)

# altdf <- importFromTable(file = rawData,
# highBulk = HighBulk,
# lowBulk = LowBulk,
# chromList = Chroms)

head(df)

```




## Using my own data for this

#### Reformat as tables first

```{r}
#choose.files()

#HNGLCDRXY <- read.table("C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\HNGLVDRXY_gR3_trim_new_merged.vcf.output.table", header = TRUE)

HNGLCDRXY <- read.table("C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\g_y.SortedCat.vcf.output.table", header = TRUE)

HNGLCDRXY_works <- read.table("C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\HNGLVDRXY_gR_merged.vcf.output.table", header = TRUE)

#Apparently this only actually saved one chromosome so that needs to be fixed...
unique(HNGLCDRXY$CHROM)

```

```{r, eval = FALSE}
HighBulk <- "Selected_OakI"
LowBulk <- "Selected_WineI"
Chroms <- unique(HNGLCDRXY$CHROM)
rawData <- "C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\g_y.SortedCat.vcf.output.table"

df <- importFromGATK(
        file = rawData,
        highBulk = HighBulk,
        lowBulk = LowBulk,
        chromList = Chroms[1:10]
        )

df %>% na.omit() -> df

```

```{r, eval = FALSE}
#Filter SNPs based on some criteria
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.01,
        minTotalDepth = 0,
        maxTotalDepth = 500,
        minSampleDepth = 0,
        minGQ = 2
    )

#Run G' analysis - HAS ERROR
df_filt <- runGprimeAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    outlierFilter = "deltaSNP")

#Run QTLseq analysis
df_filt <- runQTLseqAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    popStruc = "F2",
    bulkSize = c(25, 25),
    replications = 10000,
    intervals = c(95, 99)
)

#Plot
plotQTLStats(SNPset = df_filt, var = "Gprime", plotThreshold = TRUE, q = 0.01)
plotQTLStats(SNPset = df_filt, var = "deltaSNP", plotIntervals = TRUE)

#export summary CSV
#getQTLTable(SNPset = df_filt, alpha = 0.01, export = TRUE, fileName = "my_BSA_QTL.csv")

########MY PLOTS
plot(df_filt$Gprime)
plot(df_filt$pvalue)
plot(df_filt$negLog10Pval)

df_filt$idu <- row.names(df_filt)
ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = CHROM)) + geom_point()

ggplot(df_filt, aes(x = as.numeric(idu), y = negLog10Pval, color = CHROM, shape = CHROM == "NC_001140.6")) + geom_point(alpha = 0.4) + ggtitle("Epistasis | Selected A vs Selected C -log(p)")

```


Running gPrime Analysis by Hand

```{r, eval = FALSE}
library(dplyr)
tricubeStat <- function(POS, Stat, windowSize = 2e6, ...)
{
    if (windowSize <= 0)
        stop("A positive smoothing window is required")
    stats::predict(locfit::locfit(Stat ~ locfit::lp(POS, h = windowSize, deg = 0), ...), POS)
}

getG <- function(LowRef, HighRef, LowAlt, HighAlt)
{
    exp <- c(
        (LowRef + HighRef) * (LowRef + LowAlt) / (LowRef + HighRef + LowAlt + HighAlt),
        (LowRef + HighRef) * (HighRef + HighAlt) / (LowRef + HighRef + LowAlt + HighAlt),
        (LowRef + LowAlt) * (LowAlt + HighAlt) / (LowRef + HighRef + LowAlt + HighAlt),
        (LowAlt + HighAlt) * (HighRef + HighAlt) / (LowRef + HighRef + LowAlt + HighAlt)
    )
    obs <- c(LowRef, HighRef, LowAlt, HighAlt)
    
    G <-
        2 * (rowSums(obs * log(
            matrix(obs, ncol = 4) / matrix(exp, ncol = 4)
        )))
    return(G)
}


SNPset = df
windowSize = 1e6
outlierFilter = "deltaSNP"
filterThreshold = 0.1

message("Counting SNPs in each window...")
SNPset <- SNPset %>% dplyr::group_by(CHROM) %>% dplyr::mutate(nSNPs = countSNPs_cpp(POS = POS, 
        windowSize = windowSize))

message("Calculating tricube smoothed delta SNP index...")
SNPset <- SNPset %>% dplyr::mutate(tricubeDeltaSNP = tricubeStat(POS = POS, 
        Stat = deltaSNP, windowSize))
    
message("Calculating G and G' statistics...")
    
SNPset <- SNPset %>% dplyr::mutate(G = getG(LowRef = AD_REF.LOW, 
                                            HighRef = AD_REF.HIGH, 
                                            LowAlt = AD_ALT.LOW, 
                                            HighAlt = AD_ALT.HIGH), 
                                   Gprime = tricubeStat(POS = POS, Stat = G, windowSize = windowSize)) 
  
str(SNPset)

SNPsetPVAL <- SNPset %>% dplyr::ungroup() # %>% 
#   dplyr::mutate(pvalue = getPvals(Gprime = Gprime, 
#                                   deltaSNP = deltaSNP, 
#                                   outlierFilter = outlierFilter, 
#                                   filterThreshold = filterThreshold), 
#         negLog10Pval = -log10(pvalue), qvalue = p.adjust(p = pvalue, method = "BH"))
    


```

Trying out the getPvals function next:

```{r, eval = FALSE}
Gprime <- SNPsetPVAL$Gprime
deltaSNP = SNPsetPVAL$deltaSNP
outlierFilter = c("deltaSNP", "Hampel")
filterThreshold


    if (outlierFilter == "deltaSNP") {
        if (abs(filterThreshold) >= 0.5) {
            stop("filterThreshold should be less than 0.5")
        }
        message("Using deltaSNP-index to filter outlier regions with a threshold of ", 
            filterThreshold)
        trimGprime <- Gprime[abs(deltaSNP) < abs(filterThreshold)]
    } else {
        message("Using Hampel's rule to filter outlier regions")
        lnGprime <- log(Gprime)
        medianLogGprime <- median(lnGprime)
        MAD <- median(medianLogGprime - lnGprime[lnGprime <= 
            medianLogGprime])
        trimGprime <- Gprime[lnGprime - median(lnGprime) <= 5.2 * 
            MAD]
    }

    medianTrimGprime <- median(trimGprime)
    message("Estimating the mode of a trimmed G prime set using the 'modeest' package...")
    #Changes
    trimGprime <- trimGprime %>% na.omit()
    ###
    modeTrimGprime <- modeest::mlv(x = trimGprime, bw = 0.5, 
        method = "hsm")[1]
    muE <- log(medianTrimGprime)
    varE <- abs(muE - log(modeTrimGprime))
    message("Calculating p-values...")
    pval <- 1 - plnorm(q = Gprime, meanlog = muE, sdlog = sqrt(varE))
    # return(pval)
    
#FINISHING RUNNING THIS:
    
  SNPsetPVAL <- SNPset %>% dplyr::ungroup() %>% 
  dplyr::mutate(pvalue = pval, 
        negLog10Pval = -log10(pvalue), qvalue = p.adjust(p = pvalue, method = "BH"))
    

```

```{r, eval = FALSE}
#TRYING SOMETHING NEW - the same plot, but use the output of SNPsetPVAL above

plotQTLStats(SNPset = SNPsetPVAL, var = "Gprime", plotThreshold = TRUE, q = 0.01)

plot(SNPsetPVAL$Gprime, SNPsetPVAL$POS)
plot(SNPsetPVAL$POS)

#plot(SNPsetPVAL$pvalue)
#plot(SNPsetPVAL$negLog10Pval)


```

### What to do next:

* Clearly there is only one chromosome being listed, so make sure that the data finishes running to make the correct VCF, and then make sure that there are no filters on the table-making after

* Figure out if the data itself is being processed in a way that is making G' correspond exactly to position - this looks like it's using the wrong variable somewhere

* Run through the sample data to make sure that this works

* Test the Selected vs Unselected C, and Selected A vs Selected C data as well

## Running Selected A vs Selected C

```{r}
library("QTLseqr")

HighBulk <- "SelectedA"
LowBulk <- "SelectedC"
Chroms <- unique(HNGLCDRXY$CHROM)
rawData <- "C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\HNGLVDRXY_gR_merged.vcf.output.table"

df <- importFromGATK(
        file = rawData,
        highBulk = HighBulk,
        lowBulk = LowBulk,
        chromList = Chroms
        )

#Filter SNPs based on some criteria
#BUT WHAT DO THESE DO??? Why do they work this way? Why won't they let the rest of my data work?
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.20,
        minTotalDepth = 100,
        maxTotalDepth = 400,
        minSampleDepth = 40,
        minGQ = 99
    )

df %>% merge(.,ChromKey) -> df


#Run G' analysi
df_filt <- runGprimeAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    outlierFilter = "deltaSNP")

#Run QTLseq analysis
df_filt <- runQTLseqAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    popStruc = "F2",
    bulkSize = c(25, 25),
    replications = 10000,
    intervals = c(95, 99)
)

#Plot
# plotQTLStats(SNPset = df_filt, var = "Gprime", plotThreshold = TRUE, q = 0.01)
# plotQTLStats(SNPset = df_filt, var = "deltaSNP", plotIntervals = TRUE)

plot(df_filt$Gprime)
plot(df_filt$pvalue)
plot(df_filt$negLog10Pval)

df_filt$idu <- row.names(df_filt)
ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = CHROM)) + geom_point()

ggplot(df_filt, aes(x = as.numeric(idu), y = negLog10Pval, color = CHROM, shape = CHROM == "NC_001140.6")) + geom_point(alpha = 0.4) + ggtitle("Epistasis | Selected A vs Selected C -log(p)")

ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = chromosomes)) + 
  geom_point(size = 3, alpha = 0.8) + 
  ggtitle(title) + scale_colour_manual(values=CBchromPalette) + 
  xlab("Specific Nucleotides by Index")

```

## Running Unselected C vs Selected C

```{r}
library("QTLseqr")

HighBulk <- "SelectedC"
LowBulk <- "UnselectedC"
Chroms <- unique(HNGLCDRXY$CHROM)
rawData <- "C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\HNGLVDRXY_gR_merged.vcf.output.table"

df <- importFromGATK(
        file = rawData,
        highBulk = HighBulk,
        lowBulk = LowBulk,
        chromList = Chroms
        )

df %>% merge(.,ChromKey) -> df

#Filter SNPs based on some criteria
#BUT WHAT DO THESE DO??? Why do they work this way? Why won't they let the rest of my data work?
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.20,
        minTotalDepth = 100,
        maxTotalDepth = 400,
        minSampleDepth = 40,
        minGQ = 99
    )


#Run G' analysi
df_filt <- runGprimeAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    outlierFilter = "deltaSNP")

#Run QTLseq analysis
df_filt <- runQTLseqAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    popStruc = "F2",
    bulkSize = c(25, 25),
    replications = 10000,
    intervals = c(95, 99)
)

#Plot
plotQTLStats(SNPset = df_filt, var = "Gprime", plotThreshold = TRUE, q = 0.01)
plotQTLStats(SNPset = df_filt, var = "deltaSNP", plotIntervals = TRUE)

plot(df_filt$Gprime)
plot(df_filt$pvalue)
plot(df_filt$negLog10Pval)

df_filt$idu <- row.names(df_filt)
ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = CHROM)) + geom_point()

ggplot(df_filt, aes(x = as.numeric(idu), y = negLog10Pval, color = CHROM, shape = CHROM == "NC_001140.6")) + geom_point() + ggtitle("Unselected C vs Selected C -log(p)")

ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = chromosomes)) + 
  geom_point(size = 3, alpha = 0.8) + 
  ggtitle(title) + scale_colour_manual(values=CBchromPalette) + 
  xlab("Specific Nucleotides by Index")
```

## Running Unselected A vs Selected A

```{r}
library("QTLseqr")

HighBulk <- "SelectedA"
LowBulk <- "UnselectedA"
title <- paste(HighBulk, LowBulk, sep = " vs ")
Chroms <- unique(HNGLCDRXY$CHROM)
rawData <- "C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\HNGLVDRXY_gR_merged.vcf.output.table"

df <- importFromGATK(
        file = rawData,
        highBulk = HighBulk,
        lowBulk = LowBulk,
        chromList = Chroms
        )

df %>% merge(.,ChromKey) -> df


#Filter SNPs based on some criteria
#BUT WHAT DO THESE DO??? Why do they work this way? Why won't they let the rest of my data work?
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.20,
        minTotalDepth = 100,
        maxTotalDepth = 400,
        minSampleDepth = 40,
        minGQ = 99
    )


#Run G' analysi
df_filt <- runGprimeAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    outlierFilter = "deltaSNP")

#Run QTLseq analysis
df_filt <- runQTLseqAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    popStruc = "F2",
    bulkSize = c(25, 25),
    replications = 10000,
    intervals = c(95, 99)
)

#Plot
# plotQTLStats(SNPset = df_filt, var = "Gprime", plotThreshold = TRUE, q = 0.01)
# plotQTLStats(SNPset = df_filt, var = "deltaSNP", plotIntervals = TRUE)

plot(df_filt$Gprime)
plot(df_filt$pvalue)
plot(df_filt$negLog10Pval)

df_filt$idu <- row.names(df_filt)
ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = CHROM)) + geom_point()

ggplot(df_filt, aes(x = as.numeric(idu), y = negLog10Pval, color = CHROM, shape = CHROM == "NC_001140.6")) + geom_point() + ggtitle("Unselected A vs Selected A -log(p)")

ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = chromosomes)) + 
  geom_point(size = 3, alpha = 0.8) + 
  scale_colour_manual(values=CBchromPalette) + 
  xlab("Specific Nucleotides by Index") +
    ggtitle(title)


```

## Running Unselected A vs Selected A

```{r}
library("QTLseqr")

HighBulk <- "UnselectedA"
LowBulk <- "UnselectedC"

title <- paste(HighBulk, LowBulk, sep = " vs ")

Chroms <- unique(HNGLCDRXY$CHROM)
rawData <- "C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\HNGLVDRXY_gR_merged.vcf.output.table"

df <- importFromGATK(
        file = rawData,
        highBulk = HighBulk,
        lowBulk = LowBulk,
        chromList = Chroms
        )

df %>% merge(.,ChromKey) -> df

#Filter SNPs based on some criteria
#BUT WHAT DO THESE DO??? Why do they work this way? Why won't they let the rest of my data work?
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.20,
        minTotalDepth = 100,
        maxTotalDepth = 400,
        minSampleDepth = 40,
        minGQ = 99
    )


#Run G' analysis
df_filt <- runGprimeAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    outlierFilter = "deltaSNP")

#Run QTLseq analysis
df_filt <- runQTLseqAnalysis(
    SNPset = df_filt,
    windowSize = 1e6,
    popStruc = "F2",
    bulkSize = c(25, 25),
    replications = 10000,
    intervals = c(95, 99)
)

#Plot
# plotQTLStats(SNPset = df_filt, var = "Gprime", plotThreshold = TRUE, q = 0.01)
# plotQTLStats(SNPset = df_filt, var = "deltaSNP", plotIntervals = TRUE)

plot(df_filt$Gprime)
plot(df_filt$pvalue)
plot(df_filt$negLog10Pval)

df_filt$idu <- row.names(df_filt)
#ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = CHROM)) + geom_point()

ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = chromosomes)) + 
  geom_point(size = 3, alpha = 0.8) + 
  ggtitle(title) + scale_colour_manual(values=CBchromPalette) + 
  xlab("Specific Nucleotides by Index")


```