---
title: "QTLseq Method of Bulk Segregant Analysis"
author: "Cassandra Buzby"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
---

Keep in mind, use coolors.co for color palettes :)

Notes from Grace: use repeatmasker (isb), or trim off telomeres
* institute for systems biology, seattle
* probably also stops the transposons
* there could also be variants

SNP-indices: defined as the number of reads containing a SNP divided by the total sequencing depth at that SNP

${\Delta}$SNP:  the difference of the low value bulk SNP-index from the high value bulk SNP-index


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(tidyverse)
library(reshape2)
library(cowplot)
ggplot2::theme_set(theme_light())

CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47", 
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#8FC93A","#4C4C47",
                  "#F26430", "#0A369D", "#7EA3CC")


library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))


```

## Dataset:

```{r}
rawData = "Data\\TEST2_MergedBQSR.output.table"
#rawData = "Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"
```


### Functions for running analysis and plots on all samples

Three functions (two below) complete the analysis uniformly between each pair of bulks. I've defined the min and max depths, in sample depth, and min genome quality at the top, and then used these arguments in ```PipelineFunc``` which analyzes the data using QTLseqR functions ```filterSNPs()``` , ```runGprimeAnalysis``` , and ```runQTLseqAnalysis```. 

*removed*:
I have also adjustd ```plotQTLStatsCB()``` from QTLSeqR's function ```plotQTLStats()``` to remove errors in not having enough points to create a threshold, since this was sometimes present in a single chromosome and then prevented the remaining chromosomes from being analyzed. Now that more data is passing the filters, I will likely change this back.

The last function is ```plotallfunc()```, in which I input the output dataframe (filtered and analyzed) by ```PipelineFunc()``` to several plots, some of which are built into QTLSeqR and some of which I have created myself to better separate the data. Both functions are printed below, and each is then used on all samples (hidden) to produce the plots below.


```{r}
#################################################
#Change this to analyze all differently
#################################################

mindepth = 50
maxdepth = 1500
minsampledepth = 50
mingq = 0.98

PipelineFunc <- function(HighBulk, LowBulk, rawData,
                         mindepth = mindepth,
                         maxdepth = maxdepth,
                         minsampledepth = minsampledepth,
                         mingq = mingq,
                         windowSize = 2e4, #5e4 works
                         Chroms = c("NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4"),
                         BulkSize = 1000
                         ){
  
  mytitle <- paste(HighBulk, LowBulk, sep = " vs ")
  HNGLCDRXY <- read.table(rawData, header = TRUE)

  df <- importFromGATK(
        file = rawData,
        highBulk = HighBulk,
        lowBulk = LowBulk,
        chromList = Chroms #"NC_001133.9" #
        )

df %>% merge(.,ChromKey) %>% 
  as.data.frame() %>% na.omit() -> df

colnames(df)[1] <- "NC_Chrom"
colnames(df)[which(colnames(df) == "chromosomes")] <- "CHROM"

df$CHROM <- factor(df$CHROM, levels = c("I", "II", "III", "IV", 
                                        "V", "VI", "VII", "VIII", 
                                        "IX", "X", "XI", "XII", 
                                        "XIII", "XIV", "XV", "XVI", "M"))
#################################

#Filter SNPs based on some criteria
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.10, #0.20
        minTotalDepth = mindepth, #100
        maxTotalDepth = maxdepth, #400
        minSampleDepth = minsampledepth, #40
        minGQ = mingq #99
    )

  #df %>% merge(.,ChromKey)
  #df_filt <- df
  
  #Run G' analysis
  df_filt <- runGprimeAnalysis(
      SNPset = df_filt,
      windowSize = windowSize, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
  
  #Run QTLseq analysis
  df_filt <- runQTLseqAnalysis(
      SNPset = df_filt,
      windowSize = windowSize,
      popStruc = "F2",
      bulkSize = 1000, #c(25, 25)
      replications = 10000,
      intervals = c(95, 99)
  )

  df_filt$idu <- row.names(df_filt)
  q <- 0.01
  fdrT <- getFDRThreshold(df_filt$pvalue, alpha = q)
  GprimeT <- df_filt[which(df_filt$pvalue == fdrT), "Gprime"]
  
  # df_filt$chromosomes <- factor(df_filt$chromosomes, levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"))

    
  return(as.data.frame(df_filt))
}

#######################################################################
# Change this to plot differently #
#######################################################################

plotallfunc <- function(df_filt, HighBulk, LowBulk){
  
  mytitle <- paste(HighBulk, LowBulk, sep = " vs ")
    
  q <- 0.01
  fdrT <- getFDRThreshold(df_filt$pvalue, alpha = q)
  GprimeT <- df_filt[which(df_filt$pvalue == fdrT), "Gprime"]
  
  # plot.gprime.idu <- ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = chromosomes)) + 
  #   geom_point(size = 3, alpha = 0.8) + geom_hline(yintercept = GprimeT, linetype = "dashed") +
  #   ggtitle(mytitle) + scale_colour_manual(values=CBchromPalette)
  # 
  # plot.nlogp <- ggplot(df_filt, aes(x = as.numeric(idu), y = negLog10Pval, color = chromosomes)) + 
  #   geom_point(size = 3, alpha = 0.8) + geom_hline(yintercept = -log(0.05), linetype = "dashed") +
  #   ggtitle(mytitle) + scale_colour_manual(values=CBchromPalette)

  plot.gprime <- plotQTLStats(SNPset = df_filt, var = "Gprime", plotThreshold = TRUE, q = 0.01) #changed plotThreshold back to true to test
  plot.deltasnp <- plotQTLStats(SNPset = df_filt, var = "deltaSNP", plotIntervals = TRUE)

  
  plot.gprime.CB <- ggplot(data = df_filt, aes(x = POS, y = Gprime, color = CHROM)) + geom_line(size = 1) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(df_filt$POS), 
                                  by = 10^(floor(log10(max(df_filt$POS))))), 
                    # labels = format_genomic(), 
                    name = "Genomic Position (Mb)") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") +
    scale_colour_manual(values=CBchromPalette) + 
    #geom_vline(xintercept = 212535, linetype = "dashed", size = 1) + #CUP1 gene
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle(mytitle)
  
  #print(plot.gprime.idu)
  #print(plot.nlogp)
  
  pointsofsig <- ggplot(data = df_filt, aes(x = POS, y = Gprime, color = abs(deltaSNP) < 0.1)) + geom_point(size = 1) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(df_filt$POS), 
                                  by = 10^(floor(log10(max(df_filt$POS))))), 
                    # labels = format_genomic(), 
                    name = "Genomic Position (Mb)") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") +
    scale_colour_manual(values=CBchromPalette) + 
    #geom_vline(xintercept = 212535, linetype = "dashed", size = 1) + #CUP1 gene
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle(mytitle)
  
  #print(plot.gprime.CB)
  #print(pointsofsig)
  print(plot.gprime)
  print(plot.deltasnp + theme(legend.position = "bottom")) 
}

```

### Summaries of Files

Here I loaded in each bulk through ```importFromGATK()``` and then concatenated the data frames so that each had uniform names for their columns (ie AD, DP, PL, and GQ). These fields refer to allele depth (AD) of either reference or alternate allele, unfiltered depth (DP) of the reads, genome quality (GQ) which is apparently only up to 99, and PL I don't remember (but look this up).

The density plots show the range of each of these fields, and helps to pick filters to use in downstream steps.

```{r, eval = FALSE}
Chroms <- ChromKey$CHROM

OakI_Selected <- importFromGATK(
        file = rawData,
        highBulk = "SelectedC",
        lowBulk = "UnselectedC",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_Selected")

OakI_Unselected <- importFromGATK(
        file = rawData,
        highBulk = "UnselectedC",
        lowBulk = "SelectedC",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_Unselected")

WineI_Selected <- importFromGATK(
        file = rawData,
        highBulk = "SelectedA",
        lowBulk = "UnselectedC",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_Selected")

WineI_Unselected <- importFromGATK(
        file = rawData,
        highBulk = "UnselectedA",
        lowBulk = "UnselectedC",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_Unselected")

OakI_Selected %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                REF_FRQ, deltaSNP, Dataset) -> OakI_Selected

OakI_Unselected %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                REF_FRQ, deltaSNP, Dataset) -> OakI_Unselected

WineI_Selected %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                REF_FRQ, deltaSNP, Dataset) -> WineI_Selected

WineI_Unselected %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                REF_FRQ, deltaSNP, Dataset) -> WineI_Unselected

rbind(OakI_Selected, OakI_Unselected, WineI_Selected, WineI_Unselected) %>% merge(., ChromKey) -> HNG

#save file to load in later
save(HNG, file = "HNG_4-13-22.Rdata")
```


```{r, echo = FALSE}

load("HNG_4-13-22.Rdata")

HNG <- subset(HNG, chromosomes != "M")
ggplot(HNG, aes(x = REF_FRQ, fill = Dataset)) + geom_density(alpha = 0.5, size = 0.8) + ggtitle("Reference Allele Frequency") + facet_wrap(~chromosomes)

ggplot(HNG, aes(x = AD_REF.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 0.8) + ggtitle("Reference AD (unfiltered allele depth)") + 
  geom_vline(xintercept = c(mindepth,maxdepth), linetype = "dashed")+ facet_wrap(~chromosomes)

ggplot(HNG, aes(x = AD_ALT.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 0.8) + ggtitle("Alt AD (unfiltered allele depth)") + 
  geom_vline(xintercept = c(mindepth,maxdepth), linetype = "dashed")+ facet_wrap(~chromosomes)

ggplot(HNG, aes(x = DP.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 0.8) + ggtitle("DP (filtered depth, at the sample level)") + geom_vline(xintercept = minsampledepth, linetype = "dashed")+ facet_wrap(~chromosomes)

#ggplot(HNG[HNG$GQ.HIGH > 97,], aes(x = GQ.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 0.8) + ggtitle("Genotype Quality (Phred-scaled confidence that the genotype assignment  is correct)")+ facet_wrap(~chromosomes)

#ggplot(HNG[HNG$GQ.HIGH < 99,], aes(x = GQ.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 0.8) + ggtitle("Genotype Quality (Phred-scaled confidence that the genotype assignment  is correct)")+ facet_wrap(~chromosomes)

#table(HNG$GQ.HIGH)

ggplot(HNG, aes(x = SNPindex.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 0.8) + ggtitle("SNP Index")+ facet_wrap(~chromosomes)

```

${\Delta}$SNP:  the difference of the low value bulk SNP-index from the high value bulk SNP-index

## Run All Data Through Functions

```{r, eval = FALSE}

HNG %>% filter(CHROM != "NC_001133.9") -> HNGc

SAUA <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "UnselectedA", rawData = rawData, BulkSize = c(2.1e9, 1.26e10), Chroms = unique(HNGc$CHROM), windowSize = 1e4)
SCUC <- PipelineFunc(HighBulk = "SelectedC", LowBulk = "UnselectedC", rawData = rawData, BulkSize = c(1.8e9, 1.3e10), Chroms = unique(HNGc$CHROM))
Selected <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "SelectedC", rawData = rawData, BulkSize = c(2.1e9, 1.8e9), Chroms = unique(HNGc$CHROM))
Unselected <- PipelineFunc(HighBulk = "UnselectedA", LowBulk = "UnselectedC", rawData = rawData, BulkSize = c(1.3e10), Chroms = unique(HNGc$CHROM))

save(SAUA, file = "SAUA_4-13-22.Rdata")
save(SCUC, file = "SCUC_4-13-22.Rdata")
save(Selected, file = "Selected_4-13-22.Rdata")
save(Unselected, file = "Unselected_4-13-22.Rdata")

```

```{r, echo = FALSE}
load("SAUA_4-13-22.Rdata")
load("SCUC_4-13-22.Rdata")
load("Selected_4-13-22.Rdata")
load("Unselected_4-13-22.Rdata")
```

Raw data is defined as a vcf.output.table, downloaded from $SCRATCH/Sequencing/* where that iteration of the sequencing pipeline has been stored. Naming of these variables and the pipeline itself is stored in READ.ME in each folder.

```{r, eval = FALSE}

HNG %>% filter(CHROM != "NC_001133.9") -> HNGc

SAUA <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "UnselectedA", rawData = rawData, BulkSize = c(2.1e9, 1.26e10), Chroms = unique(HNGc$CHROM))
SCUC <- PipelineFunc(HighBulk = "SelectedC", LowBulk = "UnselectedC", rawData = rawData, BulkSize = c(1.8e9, 1.3e10), Chroms = unique(HNGc$CHROM))
Selected <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "SelectedC", rawData = rawData, BulkSize = c(2.1e9, 1.8e9), Chroms = unique(HNGc$CHROM))
Unselected <- PipelineFunc(HighBulk = "UnselectedA", LowBulk = "UnselectedC", rawData = rawData, BulkSize = c(1.3e10), Chroms = unique(HNGc$CHROM))

```

## Checking out stats for FDR on the Unselected bulks

```{r}
#Their FDR function
fdrT <- getFDRThreshold(df_filt$pvalue, alpha = q) #q is 0.01

getFDRThreshold <- function (pvalues, alpha = 0.01) 
  {
    sortedPvals <- sort(pvalues, decreasing = FALSE)
    pAdj <- p.adjust(sortedPvals, method = "BH")
    if (!any(pAdj < alpha)) {
        fdrThreshold <- NA
    }
    else {
        fdrThreshold <- sortedPvals[max(which(pAdj < alpha))]
    }
    return(fdrThreshold)
  }

fdrT_Unselected <- getFDRThreshold(Unselected$pvalue, alpha = 0.01)
GprimeT <- Unselected[which(Unselected$pvalue == fdrT_Unselected), "Gprime"]


```


## Adjusting window sizes to figure out new stats

First, let's plot the G values for each rather than the G' values (so no window is used)

```{r}
head(SAUA)
df_filt <- SAUA

ggplot(data = subset(df_filt, G < 100), aes(x = POS, y = G, color = CHROM)) + geom_line() + 
  scale_x_continuous(breaks = seq(from = 0, to = max(df_filt$POS), 
                                  by = 10^(floor(log10(max(df_filt$POS))))), 
                    name = "Genomic Position") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") +
    scale_colour_manual(values=CBchromPalette) + 
  geom_hline(yintercept = GprimeT, color = "red", alpha = 0.5) +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle("SAUA G")

ggplot(data = subset(df_filt, Gprime < 100), aes(x = POS, y = Gprime, color = CHROM)) + geom_line() + 
  scale_x_continuous(breaks = seq(from = 0, to = max(df_filt$POS), 
                                  by = 10^(floor(log10(max(df_filt$POS))))), 
                    name = "Genomic Position") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") +
    scale_colour_manual(values=CBchromPalette) + 
  geom_hline(yintercept = GprimeT, color = "red", alpha = 0.5) +
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle("SAUA Gprime")


for(i in unique(df_filt$CHROM)){
  chromplot <- ggplot(data = subset(df_filt, CHROM == i), aes(x = POS, y = G)) + geom_line() + 
    scale_x_continuous(breaks = seq(from = 0, to = max(df_filt$POS), 
                                  by = 10^(floor(log10(max(df_filt$POS))))), 
                    name = "Genomic Position") +
    theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
      geom_hline(yintercept = GprimeT, color = "red", alpha = 0.5) +
  guides(colour = "none") +
    scale_colour_manual(values=CBchromPalette) + ggtitle(paste("Chromosome ", i))
  print(chromplot)
}

for(i in unique(df_filt$CHROM)){
  chromplot <- ggplot(data = subset(df_filt, CHROM == i), aes(x = POS, y = Gprime)) + geom_line() + 
    scale_x_continuous(breaks = seq(from = 0, to = max(df_filt$POS), 
                                  by = 10^(floor(log10(max(df_filt$POS))))), 
                    name = "Genomic Position") +
    theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") + 
      geom_hline(yintercept = GprimeT, color = "red", alpha = 0.5) +
    scale_colour_manual(values=CBchromPalette) + ggtitle(paste("Chromosome ", i))
  print(chromplot)
}
  
```

## Re-evaluating the G' Window Stuff

```{r}
runGprimeAnalysis <- function (SNPset, windowSize = 1e+06, outlierFilter = "deltaSNP", 
    filterThreshold = 0.1, ...) 
{
    message("Counting SNPs in each window...")
    SNPset <- SNPset %>% dplyr::group_by(CHROM) %>% dplyr::mutate(nSNPs = countSNPs_cpp(POS = POS, 
        windowSize = windowSize))
    
    message("Calculating tricube smoothed delta SNP index...")
    SNPset <- SNPset %>% dplyr::mutate(tricubeDeltaSNP = tricubeStat(POS = POS, 
        Stat = deltaSNP, windowSize, ...))
    
    message("Calculating G and G' statistics...")
    SNPset <- SNPset %>% dplyr::mutate(G = getG(LowRef = AD_REF.LOW, 
        HighRef = AD_REF.HIGH, LowAlt = AD_ALT.LOW, HighAlt = AD_ALT.HIGH), 
        Gprime = tricubeStat(POS = POS, Stat = G, windowSize = windowSize, 
            ...)) %>% dplyr::ungroup() %>% dplyr::mutate(pvalue = getPvals(Gprime = Gprime, 
        deltaSNP = deltaSNP, outlierFilter = outlierFilter, filterThreshold = filterThreshold), 
        negLog10Pval = -log10(pvalue), qvalue = p.adjust(p = pvalue, 
            method = "BH"))
    return(as.data.frame(SNPset))
}

#So how many SNPs are within each chromosome?

ggplot(SAUA, aes(x = POS, color = CHROM)) + geom_density() + ggtitle("SNP Density by Position for Selected/Unselected Wine") +
      scale_colour_manual(values=CBchromPalette)

ggplot(SCUC, aes(x = POS, color = CHROM)) + geom_density() + ggtitle("SNP Density by Position for Selected/Unselected Oak") +
      scale_colour_manual(values=CBchromPalette)
ggplot(Unselected, aes(x = POS, color = CHROM)) + geom_density() + ggtitle("SNP Density by Position for Unselected Bulks") +
      scale_colour_manual(values=CBchromPalette)
ggplot(Selected, aes(x = POS, color = CHROM)) + geom_density() + ggtitle("SNP Density by Position for Selected Bulks") +
      scale_colour_manual(values=CBchromPalette)

ggplot(subset(SCUC, CHROM == "XII" | CHROM == "IV"), aes(x = POS, color = CHROM)) + geom_density()
```

## How big is each chromosome?

```{r}
SAUA %>% group_by(CHROM) %>% summarize(MP = max(POS)) -> maxpositions

ggplot(maxpositions, aes(x = CHROM, y = MP)) + geom_col() + ylim(0, max(maxpositions$MP)) + ggtitle("Approximate length of each Chromosome")
```


## What is the min window size you can have per chromosome?

```{r}
df <- SAUA
perm <- data.frame()

window <- 3000

temp_II <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "II"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_II, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom II")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_III <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "III"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_III, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom III")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_IV <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "IV"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_IV, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom IV")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_V <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "V"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_V, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom V") + geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_VI <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "VI"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_VI, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom VI")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_VII <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "VII"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_VII, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom VII")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_VIII <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "VIII"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_VIII, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom VIII")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_IX <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "IX"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_IX, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom IX")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_X <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "X"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_X, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom X")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_XI <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "XI"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_XI, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom XI")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_XII <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "XII"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_XII, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom XII")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_XIII <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "XIII"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_XIII, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom XIII")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_XIV <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "XIV"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_XIV, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom XIV")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_XV <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "XV"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_XV, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom XV")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

temp_XVI <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == "XVI"),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
ggplot(temp_XVI, aes(x = POS, y = Gprime)) + geom_line() + ggtitle("Chrom XVI")+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")

```

### Do this in a loop

```{r}
df <- Unselected
perm <- data.frame()

window <- 3000

perm_Unselected <- data.frame()
for(i in unique(df$CHROM)){
  temp_II <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == i),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
  plot <- ggplot(temp_II, aes(x = POS, y = Gprime)) + geom_line() + ggtitle(paste("Chrom", i))+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")
  print(plot)

  perm_Unselected <- rbind(perm_Unselected, temp_II)
}

perm_Unselected_3000 <- perm_Unselected
```
```{r}
#Selected A
df <- SAUA
perm <- data.frame()

window <- 3000

perm_A <- data.frame()
for(i in unique(df$CHROM)){
  temp_II <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == i),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
  plot <- ggplot(temp_II, aes(x = POS, y = Gprime)) + geom_line() + ggtitle(paste("Chrom", i))+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")
  print(plot)

  perm_A <- rbind(perm_A, temp_II)
}

#Selected C

df <- SCUC
perm <- data.frame()

window <- 3000

perm_C <- data.frame()
for(i in unique(df$CHROM)){
  temp_II <- runGprimeAnalysis(
      SNPset = subset(df, CHROM == i),
      windowSize = window, #1e6
      outlierFilter = "deltaSNP", maxk = 1000)
  plot <- ggplot(temp_II, aes(x = POS, y = Gprime)) + geom_line() + ggtitle(paste("Chrom", i))+ geom_hline(yintercept = 2.5, alpha = 0.5, size = 2, color = "red")
  print(plot)

  perm_C <- rbind(perm_C, temp_II)
}

```

Make a linear plot between overlapping POS values

```{r}
#Basically proving to myself why we need an interpolation instead of just taking the intersection since they're out of order :(


SelectedPositions_A <- perm_A[perm_A$POS %in% perm_C$POS,]
SelectedPositions_C <- perm_C[perm_C$POS %in% perm_A$POS,]

length(unique(SelectedPositions_A$POS))/length(SelectedPositions_A$POS)
length(unique(SelectedPositions_C$POS))/length(SelectedPositions_C$POS)

merge(SelectedPositions_A, SelectedPositions_C, by = "POS") -> AllGs

ggplot(AllGs, aes(x = G.x, y = G.y, color = CHROM.x)) + geom_point(alpha = 0.3) + facet_wrap(~CHROM.x, scales = "free")

ggplot(subset(AllGs), aes(x = Gprime.x, y = Gprime.y, color = POS)) + 
  geom_point(alpha = 0.3) +  facet_wrap(~CHROM.x, scales = "free") +
  geom_smooth(method='lm',formula=y~x)

for(i in unique(AllGs$CHROM.x)){
  pot <- ggplot(subset(AllGs, CHROM.x == i), aes(x = Gprime.x, y = Gprime.y, color = POS/max(POS))) + 
  geom_point(alpha = 0.3) +  facet_wrap(~CHROM.x, scales = "free") +
  geom_smooth(method='lm',formula=y~x) + ggtitle(paste("Chrom ", i))
  print(pot)
}

mylm <- lm(Gprime.y~Gprime.x, data = AllGs)
mylm$coefficients[1] + mylm$coefficients[2]*Gprime.y

AllGs %>% group_by(CHROM.x) %>% mutate(Pred = lm(Gprime.y~Gprime.x)$coefficients[1] + lm(Gprime.y~Gprime.x)$coefficients[2]*Gprime.y) -> AllGs

ggplot(AllGs, aes(x = Gprime.x, y = Pred)) + geom_point()
ggplot(AllGs, aes(x = POS, y = Gprime.x, col = Pred / Gprime.y)) + geom_point() + facet_wrap(~CHROM.x)

# SA <- SelectedPositions_A[SelectedPositions_A$POS == SelectedPositions_C$POS,]
# SC <- SelectedPositions_C[SelectedPositions_C$POS == SelectedPositions_A$POS,]
# 
# SA %>% transmute(CHROM.A = CHROM, POS.A = POS, Gprime.A = Gprime, G.A = G) -> SA
# SC %>% transmute(CHROM.C = CHROM, POS.C = POS, Gprime.C = Gprime, G.C = G) -> SC
# 
# SASC <- cbind(SA, SC)
# 
# ggplot(SASC, aes(x = Gprime.A, y = Gprime.C, color = CHROM.A)) + geom_point(alpha = 0.5)
# ggplot(SASC, aes(x = G.A, y = G.C, color = CHROM.A)) + geom_point(alpha = 0.5)

which(AllGs$Gprime.x < 0)

which(AllGs$Gprime.y < 0)

AllGs[AllGs$Gprime.x < 0,]
```

