---
title: "QTLseq Method of Bulk Segregant Analysis"
author: "Cassandra Buzby"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
---

Keep in mind, use coolors.co for color palettes :)

Notes from Grace: use repeatmasker (isb), or trim off telomeres
* institute for systems biology, seattle
* probably also stops the transposons
* there could also be variants



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(tidyverse)
library(reshape2)
library(cowplot)
ggplot2::theme_set(theme_light())

CBchromPalette <- c("#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#4C4C47", "#8FC93A",
                   "#F26430", "#0A369D", "#7EA3CC","#FF9F1C", "#C14953","#92DCE5", "#4C4C47", "#8FC93A",
                  "#F26430", "#0A369D", "#7EA3CC")


library("QTLseqr")

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", 
                                       "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"), 
                       CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

```

```{r, include=FALSE}
#Function to Plot
plotQTLStatsCB <- function (SNPset, subset = NULL, var = "nSNPs", scaleChroms = TRUE, 
    line = TRUE, plotThreshold = FALSE, plotIntervals = FALSE, 
    q = 0.05, ...) 
{
    if (!all(subset %in% unique(SNPset$CHROM))) {
        whichnot <- paste(subset[base::which(!subset %in% unique(SNPset$CHROM))], 
            collapse = ", ")
        stop(paste0("The following are not true chromosome names: ", 
            whichnot))
    }
    if (!var %in% c("nSNPs", "deltaSNP", "Gprime", 
        "negLog10Pval")) 
        stop("Please choose one of the following variables to plot: \"nSNPs\", \"deltaSNP\", \"Gprime\", \"negLog10Pval\"")
    if ((plotThreshold == TRUE & var == "deltaSNP") | (plotThreshold == 
        TRUE & var == "nSNPs")) {
        message("FDR threshold is not plotted in deltaSNP or nSNPs plots")
        plotThreshold <- FALSE
    }
    GprimeT <- 0
    logFdrT <- 0
    if (plotThreshold == TRUE) {
        fdrT <- getFDRThreshold(SNPset$pvalue, alpha = q)
        if (is.na(fdrT)) {
            warning("The q threshold is too low. No threshold line will be drawn")
            plotThreshold <- FALSE
        }
        else {
            logFdrT <- -log10(fdrT)
            GprimeT <- SNPset[which(SNPset$pvalue == fdrT), "Gprime"]
        }
    }
    SNPset <- if (is.null(subset)) {
        SNPset
    }
    else {
        SNPset[SNPset$CHROM %in% subset, ]
    }
    p <- ggplot2::ggplot(data = SNPset) + ggplot2::scale_x_continuous(breaks = seq(from = 0, 
        to = max(SNPset$POS), by = 10^(floor(log10(max(SNPset$POS))))), 
        #labels = format_genomic(), 
        name = "Genomic Position (Mb)") + 
        ggplot2::theme(plot.margin = ggplot2::margin(b = 10, 
            l = 20, r = 20, unit = "pt"))
    if (var == "Gprime") {
        threshold <- GprimeT
        p <- p + ggplot2::ylab("G' value")
    }
    if (var == "negLog10Pval") {
        threshold <- logFdrT
        p <- p + ggplot2::ylab(expression("-" * log[10] * 
            "(p-value)"))
    }
    if (var == "nSNPs") {
        p <- p + ggplot2::ylab("Number of SNPs in window")
    }
    if (var == "deltaSNP") {
        var <- "tricubeDeltaSNP"
        p <- p + ggplot2::ylab(expression(Delta * "(SNP-index)")) + 
            ggplot2::ylim(-0.55, 0.55) + ggplot2::geom_hline(yintercept = 0, 
            color = "black", alpha = 0.4)
        if (plotIntervals == TRUE) {
            ints_df <- dplyr::select(SNPset, CHROM, POS, dplyr::matches("CI_")) %>% 
                tidyr::gather(key = "Interval", value = "value", 
                  -CHROM, -POS)
            p <- p + ggplot2::geom_line(data = ints_df, ggplot2::aes(x = POS, 
                y = value, color = Interval)) + ggplot2::geom_line(data = ints_df, 
                ggplot2::aes(x = POS, y = -value, color = Interval))
        }
    }
    if (line) {
        p <- p + ggplot2::geom_line(ggplot2::aes_string(x = "POS", 
            y = var), ...)
    }
    if (!line) {
        p <- p + ggplot2::geom_point(ggplot2::aes_string(x = "POS", 
            y = var), ...)
    }
    if (plotThreshold == TRUE) 
        p <- p + ggplot2::geom_hline(ggplot2::aes_string(yintercept = "threshold"), 
            color = "red", size = 1, alpha = 0.4)
    if (scaleChroms == TRUE) {
        p <- p + ggplot2::facet_grid(~CHROM, scales = "free_x", 
            space = "free_x")
    }
    else {
        p <- p + ggplot2::facet_grid(~CHROM, scales = "free_x")
    }
    p
}

```

### Functions for running analysis and plots on all samples

Three functions (two below) complete the analysis uniformly between each pair of bulks. I've defined the min and max depths, in sample depth, and min genome quality at the top, and then used these arguments in ```PipelineFunc``` which analyzes the data using QTLseqR functions ```filterSNPs()``` , ```runGprimeAnalysis``` , and ```runQTLseqAnalysis```. I have also adjustd ```plotQTLStatsCB()``` from QTLSeqR's function ```plotQTLStats()``` to remove errors in not having enough points to create a threshold, since this was sometimes present in a single chromosome and then prevented the remaining chromosomes from being analyzed. Now that more data is passing the filters, I will likely change this back.

The last function is ```plotallfunc()```, in which I input the output dataframe (filtered and analyzed) by ```PipelineFunc()``` to several plots, some of which are built into QTLSeqR and some of which I have created myself to better separate the data. Both functions are printed below, and each is then used on all samples (hidden) to produce the plots below.


```{r}
#################################################
#Change this to analyze all differently
#################################################

mindepth = 50
maxdepth = 1500
minsampledepth = 50
mingq = 0.98

PipelineFunc <- function(HighBulk, LowBulk, rawData,
                         mindepth = mindepth,
                         maxdepth = maxdepth,
                         minsampledepth = minsampledepth,
                         mingq = mingq,
                         windowSize = 5e4,
                         Chroms = c("NC_001134.8", "NC_001135.5", "NC_001136.10", 
                                 "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6", 
                                 "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5", 
                                 "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4"),
                         BulkSize = 1000
                         ){
  
  mytitle <- paste(HighBulk, LowBulk, sep = " vs ")
  HNGLCDRXY <- read.table(rawData, header = TRUE)

  df <- importFromGATK(
        file = rawData,
        highBulk = HighBulk,
        lowBulk = LowBulk,
        chromList = Chroms #"NC_001133.9" #
        )

df %>% merge(.,ChromKey) %>% 
  #group_by(CHROM) %>% mutate(Start = min(POS) + 350, End = max(POS) - 350) %>% 
  as.data.frame() %>% na.omit() -> df

colnames(df)[1] <- "NC_Chrom"
colnames(df)[which(colnames(df) == "chromosomes")] <- "CHROM"

df$CHROM <- factor(df$CHROM, levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"))
#################################

#Filter SNPs based on some criteria
df_filt <-
    filterSNPs(
        SNPset = df,
        refAlleleFreq = 0.10, #0.20
        minTotalDepth = mindepth, #100
        maxTotalDepth = maxdepth, #400
        minSampleDepth = minsampledepth, #40
        minGQ = mingq #99
    )

  #df %>% merge(.,ChromKey)
  #df_filt <- df
  
  #Run G' analysi
  df_filt <- runGprimeAnalysis(
      SNPset = df_filt,
      windowSize = windowSize, #1e6
      outlierFilter = "deltaSNP")
  
  #Run QTLseq analysis
  df_filt <- runQTLseqAnalysis(
      SNPset = df_filt,
      windowSize = windowSize,
      popStruc = "F2",
      bulkSize = 1000, #c(25, 25)
      replications = 10000,
      intervals = c(95, 99)
  )

  df_filt$idu <- row.names(df_filt)
  q <- 0.01
  fdrT <- getFDRThreshold(df_filt$pvalue, alpha = q)
  GprimeT <- df_filt[which(df_filt$pvalue == fdrT), "Gprime"]
  
  # df_filt$chromosomes <- factor(df_filt$chromosomes, levels = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"))

    
  return(as.data.frame(df_filt))
}

#######################################################################
# Change this to plot differently #
#######################################################################

plotallfunc <- function(df_filt, HighBulk, LowBulk){
  
  mytitle <- paste(HighBulk, LowBulk, sep = " vs ")
    
  q <- 0.01
  fdrT <- getFDRThreshold(df_filt$pvalue, alpha = q)
  GprimeT <- df_filt[which(df_filt$pvalue == fdrT), "Gprime"]
  
  # plot.gprime.idu <- ggplot(df_filt, aes(x = as.numeric(idu), y = Gprime, color = chromosomes)) + 
  #   geom_point(size = 3, alpha = 0.8) + geom_hline(yintercept = GprimeT, linetype = "dashed") +
  #   ggtitle(mytitle) + scale_colour_manual(values=CBchromPalette)
  # 
  # plot.nlogp <- ggplot(df_filt, aes(x = as.numeric(idu), y = negLog10Pval, color = chromosomes)) + 
  #   geom_point(size = 3, alpha = 0.8) + geom_hline(yintercept = -log(0.05), linetype = "dashed") +
  #   ggtitle(mytitle) + scale_colour_manual(values=CBchromPalette)

#Fancy plots from the Ziv script adjustments

  # plot.gprime <- plotQTLStatsCB(SNPset = df_filt, var = "Gprime", plotThreshold = TRUE, q = 0.01) #changed plotThreshold back to true to test
  # plot.deltasnp <- plotQTLStatsCB(SNPset = df_filt, var = "deltaSNP", plotIntervals = TRUE)

  plot.gprime <- plotQTLStats(SNPset = df_filt, var = "Gprime", plotThreshold = TRUE, q = 0.01) #changed plotThreshold back to true to test
  plot.deltasnp <- plotQTLStats(SNPset = df_filt, var = "deltaSNP", plotIntervals = TRUE)

  
  plot.gprime.CB <- ggplot(data = df_filt, aes(x = POS, y = Gprime, color = CHROM)) + geom_line(size = 1) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(df_filt$POS), 
                                  by = 10^(floor(log10(max(df_filt$POS))))), 
                    # labels = format_genomic(), 
                    name = "Genomic Position (Mb)") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") +
    scale_colour_manual(values=CBchromPalette) + 
    #geom_vline(xintercept = 212535, linetype = "dashed", size = 1) + #CUP1 gene
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle(mytitle)
  
  #print(plot.gprime.idu)
  #print(plot.nlogp)
  
  pointsofsig <- ggplot(data = df_filt, aes(x = POS, y = Gprime, color = abs(deltaSNP) < 0.1)) + geom_point(size = 1) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(df_filt$POS), 
                                  by = 10^(floor(log10(max(df_filt$POS))))), 
                    # labels = format_genomic(), 
                    name = "Genomic Position (Mb)") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") +
    scale_colour_manual(values=CBchromPalette) + 
    #geom_vline(xintercept = 212535, linetype = "dashed", size = 1) + #CUP1 gene
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle(mytitle)
  
  print(plot.gprime.CB)
  #print(pointsofsig)
  print(plot.gprime)
  print(plot.deltasnp + theme(legend.position = "bottom")) 
}

```

### Summaries of Files

Here I loaded in each bulk through ```importFromGATK()``` and then concatenated the data frames so that each had uniform names for their columns (ie AD, DP, PL, and GQ). These fields refer to allele depth (AD) of either reference or alternate allele, unfiltered depth (DP) of the reads, genome quality (GQ) which is apparently only up to 99, and PL I don't remember (but look this up).

The density plots show the range of each of these fields, and helps to pick filters to use in downstream steps.

```{r, include = FALSE}
rawData = "C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"

Chroms <- ChromKey$CHROM

OakI_Selected <- importFromGATK(
        file = rawData,
        highBulk = "SelectedC",
        lowBulk = "UnselectedC",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_Selected")

OakI_Unselected <- importFromGATK(
        file = rawData,
        highBulk = "UnselectedC",
        lowBulk = "SelectedC",
        chromList = Chroms
        ) %>% mutate(Dataset = "OakI_Unselected")

WineI_Selected <- importFromGATK(
        file = rawData,
        highBulk = "SelectedA",
        lowBulk = "UnselectedC",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_Selected")

WineI_Unselected <- importFromGATK(
        file = rawData,
        highBulk = "UnselectedA",
        lowBulk = "UnselectedC",
        chromList = Chroms
        ) %>% mutate(Dataset = "WineI_Unselected")

OakI_Selected %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                REF_FRQ, deltaSNP, Dataset) -> OakI_Selected

OakI_Unselected %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                REF_FRQ, deltaSNP, Dataset) -> OakI_Unselected

WineI_Selected %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                REF_FRQ, deltaSNP, Dataset) -> WineI_Selected

WineI_Unselected %>% select(CHROM, POS, REF, ALT, 
                AD_REF.HIGH, AD_ALT.HIGH, DP.HIGH, GQ.HIGH, PL.HIGH, SNPindex.HIGH,
                REF_FRQ, deltaSNP, Dataset) -> WineI_Unselected

rbind(OakI_Selected, OakI_Unselected, WineI_Selected, WineI_Unselected) %>% merge(., ChromKey) -> HNG

################# Plots ###############
# mindepth = 10
# maxdepth = 1500
# minsampledepth = 40
# mingq = 0.98
```


```{r, echo = FALSE, eval = FALSE}
head(HNG)

ggplot(HNG, aes(x = REF_FRQ, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + ggtitle("Reference Allele Frequency") + facet_wrap(~CHROM)

ggplot(HNG, aes(x = AD_REF.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + ggtitle("Reference AD (unfiltered allele depth)") + 
  geom_vline(xintercept = c(mindepth,maxdepth), linetype = "dashed")

ggplot(HNG, aes(x = AD_ALT.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + ggtitle("Alt AD (unfiltered allele depth)") + 
  geom_vline(xintercept = c(mindepth,maxdepth), linetype = "dashed")

ggplot(HNG, aes(x = DP.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + ggtitle("DP (filtered depth, at the sample level)") + geom_vline(xintercept = minsampledepth, linetype = "dashed")

ggplot(HNG[HNG$GQ.HIGH > 97,], aes(x = GQ.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + ggtitle("Genotype Quality (Phred-scaled confidence that the genotype assignment  is correct)")

ggplot(HNG[HNG$GQ.HIGH < 99,], aes(x = GQ.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + ggtitle("Genotype Quality (Phred-scaled confidence that the genotype assignment  is correct)")

table(HNG$GQ.HIGH)

ggplot(HNG, aes(x = SNPindex.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + ggtitle("SNP Index") 

```

### Extracting Stats for Telomeres

```{r, echo = FALSE}
HNG %>% group_by(CHROM) %>% mutate(Start = min(POS) + 350, End = max(POS) - 350) -> HNGt

# ggplot(HNGt, aes(x = REF_FRQ, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + 
#   ggtitle("Reference Allele Frequency") 

ggplot(HNGt, aes(x = REF_FRQ, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + 
  ggtitle("Reference Allele Frequency") + 
  facet_grid(cols = vars(HNGt$POS < HNGt$Start | HNGt$POS > HNGt$End))+ theme(legend.position = "bottom")

ggplot(HNGt, aes(x = AD_REF.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + 
  ggtitle("Reference AD (unfiltered allele depth)") + 
  #geom_vline(xintercept = c(mindepth,maxdepth), linetype = "dashed") + 
  facet_grid(cols = vars(HNGt$POS < HNGt$Start | HNGt$POS > HNGt$End))+ theme(legend.position = "bottom")

ggplot(HNGt, aes(x = AD_ALT.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + 
  ggtitle("Alt AD (unfiltered allele depth)") + 
  #geom_vline(xintercept = c(mindepth,maxdepth), linetype = "dashed")  + 
  facet_grid(cols = vars(HNGt$POS < HNGt$Start | HNGt$POS > HNGt$End))+ theme(legend.position = "bottom")

ggplot(HNGt, aes(x = DP.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + 
  ggtitle("DP (filtered depth, at the sample level)") + #geom_vline(xintercept = minsampledepth, linetype = "dashed")  + 
  facet_grid(cols = vars(HNGt$POS < HNGt$Start | HNGt$POS > HNGt$End))+ theme(legend.position = "bottom")

ggplot(HNGt, aes(x = GQ.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + 
  ggtitle("Genotype Quality (Phred-scaled confidence that the genotype assignment  is correct)")  + 
  facet_grid(cols = vars(HNGt$POS < HNGt$Start | HNGt$POS > HNGt$End)) + theme(legend.position = "bottom")

ggplot(HNGt, aes(x = SNPindex.HIGH, fill = Dataset)) + geom_density(alpha = 0.5, size = 1) + 
  ggtitle("SNP Index") + 
  facet_grid(cols = vars(HNGt$POS < HNGt$Start | HNGt$POS > HNGt$End)) + theme(legend.position = "bottom")

```

## Run All Data Through Functions

```{r, include = FALSE}
rawData = "C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"
#rawData = "C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\Data\\mergedCuSO4_bqsr.REF_NC_001133.9.bam.SortedCat.vcf.output.table"

SAUA <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "UnselectedA", rawData = rawData) #BulkSize = c(2.1e9, 1.26e10),
SCUC <- PipelineFunc(HighBulk = "SelectedC", LowBulk = "UnselectedC", rawData = rawData, BulkSize = c(1.8e9, 1.3e10))
Selected <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "SelectedC", rawData = rawData, BulkSize = c(2.1e9, 1.8e9))
Unselected <- PipelineFunc(HighBulk = "UnselectedA", LowBulk = "UnselectedC", rawData = rawData, BulkSize = c(1.3e10))

```

```{r, eval = FALSE, echo = FALSE}
head(SAUA)

SAUA %>% filter(Gprime > 20) -> SAUAfilt20

write.csv(SAUAfilt20, file = "SAUAfilt20.csv")

ggplot(SAUA, aes(x = POS, y = Gprime)) + geom_line() + facet_grid(cols = vars(CHROM))

```


Raw data is defined as a vcf.output.table, downloaded from $SCRATCH/Sequencing/* where that iteration of the sequencing pipeline has been stored. Naming of these variables and the pipeline itself is stored in READ.ME in each folder.

```{r, eval = FALSE}
#rawData = "C:\\Users\\cassa\\OneDrive\\Documents\\R_SiegalLab\\SequencingPractice\\Data\\TelTrimmed_mergedCuSO4.REF.SortedCat.vcf.output.table"

SAUA <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "UnselectedA", rawData = rawData, BulkSize = c(2.1e9, 1.26e10))
SCUC <- PipelineFunc(HighBulk = "SelectedC", LowBulk = "UnselectedC", rawData = rawData, BulkSize = c(1.8e9, 1.3e10))
Selected <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "SelectedC", rawData = rawData, BulkSize = c(2.1e9, 1.8e9))
Unselected <- PipelineFunc(HighBulk = "UnselectedA", LowBulk = "UnselectedC", rawData = rawData, BulkSize = c(1.3e10))

```




### Traditional BSA (Selected vs Unselected)

Each of these sets are between Selected and Unselected A (Oak Chr I fixed) or B (Wine Chr I fixed). We would expect that the known QTL that have been identified for CuSO4 resistance (CUP1, Chr VIII) would be identified. The significance of this peak indicates the power for detection of our assay, though actual stats will also be done for this.

```{r}
plotallfunc(SAUA, HighBulk = "SelectedA", LowBulk = "UnselectedA")
plotallfunc(SCUC, HighBulk = "SelectedC", LowBulk = "UnselectedC")

```

### Unselected Bulks

These two bulks should only differ at Chr I since that is the fixed chromosome; all other loci should segregate evenly and be at approximately 50/50 frequency. One potential issue is in variation in alignment caused by the background of the fixed chromosome; those that are consistent between the unselected bulk comparison (null) and the selected bulk comparison (epistasis, below), should be removed for this reason.

```{r}
# plotallfunc(subset(Unselected, Gprime < 50), HighBulk = "UnselectedC", LowBulk = "UnselectedA")
plotallfunc(Unselected, HighBulk = "UnselectedC", LowBulk = "UnselectedA")

```

### Selected Bulks: Epistasis with Chromosome I



```{r}

#plotallfunc(subset(Selected, Gprime < 50), HighBulk = "SelectedA", LowBulk = "SelectedC")

plotallfunc(Selected, HighBulk = "SelectedA", LowBulk = "SelectedC")

```

### Testing Epistasis Data Chr. III

There looks to be one small peak on Chr III in the comparison of Selected bulks that does not appear between the Unselected bulks. To better resolve this peak, I ran ```PipelineFunc()``` on that single chromosome with a window size of 1e4, and then identified the peak. 

```{r, include = FALSE}

#Analyze All Before Plotting
Selected_3 <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "SelectedC", 
                           rawData = rawData, 
                           Chroms = "NC_001135.5", windowSize = 1e4)

#plotallfunc(Selected_3, HighBulk = "SelectedA", LowBulk = "SelectedC")
# QTL_3_Sel <- subset(Selected_3, Gprime > 5)
# dim(QTL_3_Sel)
# QTL_3_Sel

Selected_7 <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "SelectedC", 
                           rawData = rawData, 
                           Chroms = "NC_001139.9", windowSize = 2e4)

Oak_8 <- PipelineFunc(HighBulk = "SelectedA", LowBulk = "UnselectedA", 
                           rawData = rawData, 
                           Chroms = "NC_001140.6", windowSize = 1e4)


```


```{r, echo = FALSE}
plot.deltasnp <- plotQTLStatsCB(SNPset = Selected_3, var = "deltaSNP", plotIntervals = TRUE)


ggplot(data = Selected_3, aes(x = POS, y = Gprime)) + geom_line(size = 1, color = CBchromPalette[2]) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(Selected_3$POS), 
                                  by = 10^(floor(log10(max(Selected_3$POS))))), 
                    # labels = format_genomic(), 
                    name = "Genomic Position (Mb)") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") +
    geom_vline(xintercept = c(114385,114501), col = "gray", linetype = "dotted", size = 1) + #CEN3
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle("Selected, Chr III")
  

```

### Identifying CUP1 gene on Chr VIII

```{r, echo = FALSE}

plotallfunc(Oak_8, HighBulk = "SelectedA", LowBulk = "SelectedC")
plot.deltasnp <- plotQTLStatsCB(SNPset = Oak_8, var = "deltaSNP", plotIntervals = TRUE)


ggplot(data = Oak_8, aes(x = POS, y = Gprime, color = CHROM)) + geom_line(size = 1, color = CBchromPalette[7]) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(Oak_8$POS), 
                                  by = 10^(floor(log10(max(Oak_8$POS))))), 
                    # labels = format_genomic(), 
                    name = "Genomic Position (Mb)") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") +
    geom_vline(xintercept = 212535, linetype = "dashed", size = 1) + #CUP1 gene
    geom_vline(xintercept = c(105586), col = "gray", linetype = "dotted", size = 1) + #CEN8
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle("Oak Fixed, Chr VIII")

```

```{r, eval=FALSE}
## Calculating where the slope changes

library(pracma)

value <- Oak_8$Gprime
plot(value)
value_prime <- pracma::gradient(value, h1 = 2)
plot(value_prime)

ggplot(data = Oak_8, aes(x = POS, y = Gprime, color = CHROM)) + geom_line(size = 1, color = CBchromPalette[7]) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(Oak_8$POS), 
                                  by = 10^(floor(log10(max(Oak_8$POS))))), 
                    # labels = format_genomic(), 
                    name = "Genomic Position (Mb)") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  guides(colour = "none") +
    geom_vline(xintercept = 212535, linetype = "dashed", size = 1) + #CUP1 gene
    geom_vline(xintercept = c(105586), col = "gray", linetype = "dotted", size = 1) + #CEN8
    geom_vline(xintercept = Oak_8$POS[which(value_prime > 4)], col = "red") + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle("Oak Fixed, Chr VIII")

Oak_8[which(value_prime > 4),] %>% select(NC_Chrom, POS, REF, ALT, Gprime)

211825-216311
  
```

## Combining QTL Traces between Fixed Chromosome Populations

Because we care what is differentially identified between Wine-fixed and Oak-fixed experiments, I am overlaying the G' scores of the classic experiments to compare:

```{r}
# head(SAUA)
# head(SCUC)

SAUA %>% transmute(Experiment = "OakFixed", POS = POS, REF = REF, ALT = ALT, CHROM = CHROM, 
                   nSNPs = nSNPs, tricubeDeltaSNP = tricubeDeltaSNP, G = G, Gprime = Gprime,
            pvalue = pvalue, negLog10Pval = negLog10Pval, qvalue = qvalue, minDP = minDP, 
            tricupeDP = tricubeDP, CI_95 = CI_95, CI_99 = CI_99) -> OakFixed

SCUC %>% transmute(Experiment = "WineFixed", POS = POS, REF = REF, ALT = ALT, CHROM = CHROM, 
                   nSNPs = nSNPs, tricubeDeltaSNP = tricubeDeltaSNP, G = G, Gprime = Gprime,
            pvalue = pvalue, negLog10Pval = negLog10Pval, qvalue = qvalue, minDP = minDP, 
            tricupeDP = tricubeDP, CI_95 = CI_95, CI_99 = CI_99) -> WineFixed

BSAClassic <- rbind(OakFixed,WineFixed)

#plotQTLStatsCB(SNPset = BSAClassic, var = "Gprime", plotThreshold = TRUE, q = 0.01) #changed plotThreshold back to true to test

ggplot(data = BSAClassic, aes(x = POS, y = Gprime, color = Experiment)) + geom_line(size = 0.7) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(BSAClassic$POS), 
                                  by = 10^(floor(log10(max(BSAClassic$POS))))), 
                    # labels = format_genomic(), 
                    name = "Genomic Position (Mb)") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  #guides(colour = "none") +
    scale_colour_manual(values=CBchromPalette[6:5]) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle("Comparison of classic BSA")+ theme(legend.position = "bottom")

```

## Comparison of Nulls and Epistasis

```{r}
Selected %>% transmute(Experiment = "Selected", POS = POS, REF = REF, ALT = ALT, CHROM = CHROM, 
                   nSNPs = nSNPs, tricubeDeltaSNP = tricubeDeltaSNP, G = G, Gprime = Gprime,
            pvalue = pvalue, negLog10Pval = negLog10Pval, qvalue = qvalue, minDP = minDP, 
            tricupeDP = tricubeDP, CI_95 = CI_95, CI_99 = CI_99) -> Selected_All

Unselected %>% transmute(Experiment = "Unselected", POS = POS, REF = REF, ALT = ALT, CHROM = CHROM, 
                   nSNPs = nSNPs, tricubeDeltaSNP = tricubeDeltaSNP, G = G, Gprime = Gprime,
            pvalue = pvalue, negLog10Pval = negLog10Pval, qvalue = qvalue, minDP = minDP, 
            tricupeDP = tricubeDP, CI_95 = CI_95, CI_99 = CI_99) -> Unselected_All

NullComparison <- rbind(Selected_All,Unselected_All)

#plotQTLStatsCB(SNPset = BSAClassic, var = "Gprime", plotThreshold = TRUE, q = 0.01) #changed plotThreshold back to true to test

ggplot(data = subset(NullComparison, Gprime < 50), aes(x = POS, y = Gprime, color = Experiment)) + geom_line(size = 0.7) + 
  scale_x_continuous(breaks = seq(from = 0, to = max(BSAClassic$POS), 
                                  by = 10^(floor(log10(max(BSAClassic$POS))))), 
                    # labels = format_genomic(), 
                    name = "Genomic Position (Mb)") +
  theme(plot.margin = margin(b = 10, l = 20, r = 20, unit = "pt")) + 
  #guides(colour = "none") +
    scale_colour_manual(values=CBchromPalette[c(8,10)]) + 
  facet_grid(~CHROM, scales = "free_x", space = "free_x") + ggtitle("Comparison of Selected Total and Unselected Total")+ theme(legend.position = "bottom")

```


