---
title: "Rollmean Permutations"
author: "Cassandra Buzby"
date: "1/13/2023"
output:
  html_document:
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)

library(cybrBSA)

ggplot2::theme_set(theme_light())

glmfixed <- function(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W")),
                           Reads = c(HOO, HOW, HWO, HWW, LOO, LOW, LWO, LWW))

  b <- glm(Allele ~ Bulk*Parent, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

################################################################################
glmfixed_rep <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", "B", "B", "B", "B","B", "B", "B", "B")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

cybr2_rollmean <- function(dataframe){
  dataframe %>% pivot_longer(c(-CHROM, -POS), names_to = "label") %>% group_by(CHROM, label) %>% arrange(POS) %>% 
  summarize(POS = POS, CHROM = CHROM, SmoothCount = ceiling(frollmean(value, n = 100))) %>% na.omit() %>% pivot_wider(names_from = label,values_from = SmoothCount)
}


#Load Stuff
CSSI_CuSO4_cybr2 <- readRDS("RdataFiles/CSSI_CuSO4_cybr2.rds")
CSSI_CuSO4_cybr2 <- cybr2_rollmean(CSSI_CuSO4_cybr2)

CSSI_Fluc_cybr2 <- readRDS("RdataFiles/CSSI_Fluc_cybr2.rds")
CSSI_Fluc_cybr2 <- cybr2_rollmean(CSSI_Fluc_cybr2)

CSSI_Zeocin_cybr2 <- readRDS("RdataFiles/CSSI_Zeocin_cybr2.rds")
CSSI_Zeocin_cybr2 <- cybr2_rollmean(CSSI_Zeocin_cybr2)

CSSI_CuSO42_cybr2 <- readRDS("RdataFiles/CSSI_CuSO42_cybr2.rds")
CSSI_CuSO42_cybr2 <- cybr2_rollmean(CSSI_CuSO42_cybr2)



```

# CuSO4 Dataset Analysis and Comparisons

## Load Data

```{r}
CuSO4_glm <- readRDS("2023/CSSI_CuSO4_glm.rds")
CSSI_CuSO42_glm <- readRDS("2023/CSSI_CuSO42_glm.Rdata")
```

```{r, eval = FALSE}
#Analyze CuSO4 2 Data

#RUN GLM ON DATA WITH THESE COLUMNS
CSSI_CuSO42_cybr2 %>% group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed_rep(HOOa = CuSO4_Oak_A_Oak, 
                           HOWa = CuSO4_Oak_A_Wine, 
                           HWOa = CuSO4_Wine_A_Oak,
                           HWWa = CuSO4_Wine_A_Wine,
                           LOOa = Dilute_Oak_A_Oak,
                           LOWa = Dilute_Oak_A_Wine,
                           LWOa = Dilute_Wine_A_Oak, 
                           LWWa = Dilute_Wine_A_Wine,
                           HOOb = CuSO4_Oak_B_Oak, 
                           HOWb = CuSO4_Oak_B_Wine, 
                           HWOb = CuSO4_Wine_B_Oak,
                           HWWb = CuSO4_Wine_B_Wine,
                           LOOb = Dilute_Oak_B_Oak,
                           LOWb = Dilute_Oak_B_Wine,
                           LWOb = Dilute_Wine_B_Oak, 
                           LWWb = Dilute_Wine_B_Wine)[1:5],
                                                   label = c("intercept", "Bulk", "Parent", "Rep", "Interaction")) -> CSSI_CuSO42_glm

saveRDS(CSSI_CuSO42_glm, file = "2023/CSSI_CuSO42_glm.Rdata")
```

```{r}
CSSI_CuSO42_glm <- readRDS("2023/CSSI_CuSO42_glm.Rdata")
CSSI_CuSO42_glm %>% ggplot(aes(x = POS, y = summary, color = label)) + geom_line() + facet_grid(~CHROM, scales = "free") + ggtitle("Zeocin Z-Scores")

CSSI_CuSO42_glm %>% filter(label != "intercept", CHROM != "M", CHROM != "I") %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line() + 
  geom_hline(aes(yintercept = max(abs(CSSI_CuSO42_glm$summary[CSSI_CuSO42_glm$CHROM == "III" & CSSI_CuSO42_glm$label != "intercept"])))) +
  facet_grid(~CHROM, scales = "free", space = "free") + 
  ggtitle("CuSO4 Z-Scores") + 
  scale_color_manual(values = c("blue", "firebrick", "goldenrod", "gray"))


```



## Plot with these cutoffs

```{r}

CuSO4_glm %>% filter(label != "intercept", CHROM != "M", CHROM != "I") %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line() + 
  geom_hline(aes(yintercept = max(abs(CSSI_CuSO42_glm$summary[CSSI_CuSO42_glm$CHROM == "III" & CSSI_CuSO42_glm$label != "intercept"])))) +
  facet_grid(~CHROM, scales = "free", space = "free") + 
  ggtitle("CuSO4 Z-Scores 2021") + 
  scale_color_manual(values = c("firebrick", "blue", "goldenrod", "gray")) +
 theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

CSSI_CuSO42_glm %>% filter(label != "intercept", CHROM != "M", CHROM != "I") %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label)) + 
  geom_line() + 
  geom_hline(aes(yintercept = max(abs(CSSI_CuSO42_glm$summary[CSSI_CuSO42_glm$CHROM == "III" & CSSI_CuSO42_glm$label != "intercept"])))) +
  facet_grid(~CHROM, scales = "free", space = "free") + 
  ggtitle("CuSO4 Z-Scores 2023 (with Replicates)") + 
  scale_color_manual(values = c("blue", "firebrick", "goldenrod", "gray"))+
   theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

```

```{r}

CuSO4_glm %>% filter(label == "Bulk") %>% 
  ggplot(aes(x = POS, y = summary)) + geom_line() + 
  geom_line(data = subset(CSSI_CuSO42_glm, label == "Bulk"), aes(x = POS, y = summary), color = "violet") +
  facet_grid(~CHROM, scales = "free", space = "free") 

CuSO4_glm %>% filter(label == "Parent") %>% 
  ggplot(aes(x = POS, y = summary)) + geom_line() + 
  geom_line(data = subset(CSSI_CuSO42_glm, label == "Parent"), aes(x = POS, y = summary), color = "violet") +
  facet_grid(~CHROM, scales = "free", space = "free") 

CuSO4_glm %>% filter(label == "Interaction") %>% 
  ggplot(aes(x = POS, y = summary)) + geom_line() + 
  geom_line(data = subset(CSSI_CuSO42_glm, label == "Interaction"), aes(x = POS, y = summary), color = "violet") +
  facet_grid(~CHROM, scales = "free", space = "free") 
```


What is the distribution of the original reads per bulk per chromosome?

```{r}

CSSI_CuSO4_cybr2 %>% filter(CHROM != "M") %>% pivot_longer(c(-CHROM, -POS), names_to = "Bulk", values_to = "Reads") %>% ggplot(aes(x = Reads, color = Bulk)) + geom_density() + facet_wrap(~CHROM, scales = "free")
```



## Analyzing CuSO4 Replicates all together

```{r}
CSSI_CuSO4_cybr2 <- readRDS("RdataFiles/CSSI_CuSO4_cybr2.rds")
CSSI_CuSO42_cybr2 <- readRDS("RdataFiles/CSSI_CuSO42_cybr2.rds")

CuSO4_All <- full_join(CSSI_CuSO4_cybr2, CSSI_CuSO42_cybr2) %>% na.omit()
CuSO4_All_rllm <- cybr2_rollmean(CuSO4_All)
```

Look at the log of odds for each

```{r}
#Borrow scripts from LOD notebook

CuSO4_All_rllm %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") -> Intermediate

Intermediate$Allele[Intermediate$Rep == "Oak"] <- "Oak"
Intermediate$Allele[Intermediate$Rep == "Wine"] <- "Wine"
Intermediate$Rep[Intermediate$Rep == "Oak"] <- "C"
Intermediate$Rep[Intermediate$Rep == "Wine"] <- "C"

#Get rid of "OakI" as the parent
Intermediate$Parent <- gsub("I", "", Intermediate$Parent)

Intermediate %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CuSO4_A_LOD

CuSO4_A_LOD %>% filter(CHROM != "I", CHROM != "M") %>%
  ggplot(aes(x = POS, y = LOD, linetype = Rep, alpha = Bulk, color = Parent)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("firebrick", "lightblue")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

CuSO4_A_LOD %>% filter(CHROM == "VIII") %>%
  ggplot(aes(x = POS, y = LOD, linetype = Parent, alpha = Bulk, color = Rep)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

unique(CuSO4_A_LOD$CHROM)

CuSO4_A_LOD %>% filter(CHROM == "XV") %>%
  ggplot(aes(x = POS, y = LOD, linetype = Parent, alpha = Bulk, color = Rep)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom")
```
Zoom in on the weird section

```{r}
CuSO4_A_LOD %>% filter(CHROM == "XV", POS > 8e+05, POS < 1e+06) %>%
  ggplot(aes(x = POS, y = LOD, linetype = Parent, alpha = Bulk, color = Rep)) + 
  geom_vline(xintercept = c(822588,824867), color = "goldenrod", size = 2, alpha = 0.2) + #HRK1 
  geom_vline(xintercept = c(976421,976493), color = "violet", size = 2, alpha = 0.2) +
  geom_vline(xintercept = c(975913,976244), color = "lightblue", size = 2, alpha = 0.2) + #YORWdelta24 LTR 
  geom_vline(xintercept = c(849635,850054), color = "firebrick", size = 2, alpha = 0.2) + #RDL1
  geom_vline(xintercept = c(850280,850729), color = "darkorange", size = 2, alpha = 0.2) + #RDL2
  geom_vline(xintercept = c(855147,860258), color = "#138803", size = 2, alpha = 0.2) + #SNF2
  geom_vline(xintercept = c(861175,865593), color = "#00563F", size = 2, alpha = 0.2) + #YPK9 
  geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom") +
  ggtitle("Smoothed LODs")
```

Compare with CuSO4 2 GLM

```{r}
CSSI_CuSO42_glm %>% filter(label != "intercept", CHROM == "XV") %>% ggplot(aes(x = POS, y = abs(summary), color = label)) + 
  geom_hline(aes(yintercept = max(abs(CSSI_CuSO42_glm$summary[CSSI_CuSO42_glm$CHROM == "III" & CSSI_CuSO42_glm$label != "intercept"])))) +
  facet_grid(~CHROM, scales = "free", space = "free") + 
  ggtitle("CuSO4 Z-Scores (Second Replicates)") + 
  scale_color_manual(values = c("blue", "firebrick", "goldenrod", "gray")) +
  geom_vline(xintercept = c(822588,824867), color = "goldenrod", size = 2, alpha = 0.2) + #HRK1 
  geom_vline(xintercept = c(976421,976493), color = "violet", size = 2, alpha = 0.2) +
  geom_vline(xintercept = c(975913,976244), color = "lightblue", size = 2, alpha = 0.2) + #YORWdelta24 LTR 
  geom_vline(xintercept = c(849635,850054), color = "firebrick", size = 2, alpha = 0.2) + #RDL1
  geom_vline(xintercept = c(850280,850729), color = "darkorange", size = 2, alpha = 0.2) + #RDL2
  geom_vline(xintercept = c(855147,860258), color = "#138803", size = 2, alpha = 0.2) + #SNF2
  geom_vline(xintercept = c(861175,865593), color = "#00563F", size = 2, alpha = 0.2) + #YPK9  
  geom_line() + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

```


Do the GLM
```{r}
#Define triple replicate function
glmfixed_rep3 <- function(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                         HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                         HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc){
  
  combineddata <- data.frame(Bulk = factor(c("H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L", 
                                             "H", "H","H", "H", "L", "L","L", "L")),
                           Parent = factor(c("O", "O", "W", "W", "O", "O", "W", "W", 
                                             "O", "O", "W", "W", "O", "O", "W", "W",
                                             "O", "O", "W", "W", "O", "O", "W", "W")),
                           Allele = factor(c("O", "W", "O", "W","O", "W", "O", "W", 
                                             "O", "W", "O", "W","O", "W", "O", "W",
                                             "O", "W", "O", "W","O", "W", "O", "W")),
                           Rep = factor(c("A", "A", "A", "A","A", "A", "A", "A", 
                                          "B", "B", "B", "B","B", "B", "B", "B",
                                          "C", "C", "C", "C", "C", "C","C", "C")),
                           Reads = c(HOOa, HOWa, HWOa, HWWa, LOOa, LOWa, LWOa, LWWa, 
                                     HOOb, HOWb, HWOb, HWWb, LOOb, LOWb, LWOb, LWWb,
                                     HOOc, HOWc, HWOc, HWWc, LOOc, LOWc, LWOc, LWWc))

  b <- glm(Allele ~ Bulk*Parent+Rep, weights = Reads, family = binomial, 
              data = combineddata)

  summary(b)$coefficients[
    ((length(summary(b)$coefficients)/4)*2+1):
      ((length(summary(b)$coefficients)/4)*3)]
}

CuSO4_All_rllm %>% #filter(CHROM == "I", POS == 36434) %>% 
  group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed_rep3(HOOa = CuSO4_Oak_A_Oak, 
                           HOWa = CuSO4_Oak_A_Wine, 
                           HWOa = CuSO4_Wine_A_Oak,
                           HWWa = CuSO4_Wine_A_Wine,
                           LOOa = Dilute_Oak_A_Oak,
                           LOWa = Dilute_Oak_A_Wine,
                           LWOa = Dilute_Wine_A_Oak, 
                           LWWa = Dilute_Wine_A_Wine,
                           
                           HOOb = CuSO4_Oak_B_Oak, 
                           HOWb = CuSO4_Oak_B_Wine, 
                           HWOb = CuSO4_Wine_B_Oak,
                           HWWb = CuSO4_Wine_B_Wine,
                           LOOb = Dilute_Oak_B_Oak,
                           LOWb = Dilute_Oak_B_Wine,
                           LWOb = Dilute_Wine_B_Oak, 
                           LWWb = Dilute_Wine_B_Wine,
                           
                           HOOc = CuSO4_OakI_Oak, 
                           HOWc = CuSO4_OakI_Wine, 
                           HWOc = CuSO4_WineI_Oak,
                           HWWc = CuSO4_WineI_Wine,
                           LOOc = Dilute_OakI_Oak,
                           LOWc = Dilute_OakI_Wine,
                           LWOc = Dilute_WineI_Oak, 
                           LWWc = Dilute_WineI_Wine)[1:5],
                                                   label = c("intercept", "Bulk", "Parent", "Rep", "Interaction")) -> CSSI_CuSO42All_glm

saveRDS(CSSI_CuSO42All_glm, file = "2023/CSSI_CuSO4a_glm.Rdata")

```

```{r}
CSSI_CuSO42All_glm <- readRDS("2023/CSSI_CuSO4a_glm.Rdata")

CSSI_CuSO42All_glm %>% filter(label != "intercept", CHROM != "M", CHROM != "I") %>% 
  ggplot(aes(x = POS, y = abs(summary), color = label)) + geom_line() + 
  geom_hline(aes(yintercept = max(abs(CSSI_CuSO42_glm$summary[CSSI_CuSO42All_glm$CHROM == "III" & CSSI_CuSO42All_glm$label != "intercept"])))) +
  facet_grid(~CHROM, scales = "free", space = "free") + 
  ggtitle("CuSO4 Z-Scores") + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) +  
  scale_color_manual(values = c("blue", "firebrick", "goldenrod", "gray"))
```

Look at the unsmoothed values of CuSO4

```{r}
CSSI_CuSO4_cybr2 <- readRDS("RdataFiles/CSSI_CuSO4_cybr2.rds")
CSSI_CuSO42_cybr2 <- readRDS("RdataFiles/CSSI_CuSO42_cybr2.rds")

CuSO4_All_checks <- full_join(CSSI_CuSO4_cybr2, CSSI_CuSO42_cybr2)

CuSO4_All_checks %>% pivot_longer(cols = c(-CHROM, -POS),
  names_to = c("Bulk", "Parent", "Rep", "Allele"),
    names_sep = "_",
  values_to = "Reads") -> Intermediatechecks

Intermediatechecks$Allele[Intermediatechecks$Rep == "Oak"] <- "Oak"
Intermediatechecks$Allele[Intermediatechecks$Rep == "Wine"] <- "Wine"
Intermediatechecks$Rep[Intermediatechecks$Rep == "Oak"] <- "C"
Intermediatechecks$Rep[Intermediatechecks$Rep == "Wine"] <- "C"

#Get rid of "OakI" as the parent
Intermediatechecks$Parent <- gsub("I", "", Intermediatechecks$Parent)

Intermediatechecks %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(LOD = log(Wine/Oak)) -> CuSO4_A_LOD_unsmoothed

```

```{r}
#The entire region
CuSO4_A_LOD_unsmoothed %>% filter(CHROM == "XV", POS > 8e+05, POS < 1e+06) %>%
  ggplot(aes(x = POS, y = LOD, shape = Parent, alpha = Bulk, color = Rep)) + 
  geom_vline(xintercept = c(976375, 976500), linetype = "dashed") +
    geom_vline(xintercept = c(822588,824867), color = "goldenrod", size = 2, alpha = 0.5) + #HRK1 
      geom_vline(xintercept = c(976421,976493), color = "violet", size = 2, alpha = 0.5) +
  geom_vline(xintercept = c(975913,976244), color = "lightblue", size = 2, alpha = 0.5) + #YORWdelta24 LTR 
  geom_vline(xintercept = c(849635,850054), color = "firebrick", size = 2, alpha = 0.5) + #RDL1
  geom_vline(xintercept = c(850280,850729), color = "darkorange", size = 2, alpha = 0.5) + #RDL2
  geom_vline(xintercept = c(855147,860258), color = "#138803", size = 2, alpha = 0.5) + #SNF2
  geom_vline(xintercept = c(861175,865593), color = "#00563F", size = 2, alpha = 0.5) + #YPK9
  geom_point() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom") +
  ggtitle("Annotated Genes on Weird Region")

#The earlier region
CuSO4_A_LOD_unsmoothed %>% filter(CHROM == "XV", POS > 822000, POS < 824000) %>%
  ggplot(aes(x = POS, y = LOD, shape = Parent, alpha = Bulk, color = Rep)) + 
    geom_vline(xintercept = c(822588,824867), color = "goldenrod", size = 2, alpha = 0.5) + #HRK1 
  geom_point(size = 2) + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom") +
  ggtitle("HRK1 SNPs")

#The middle region
CuSO4_A_LOD_unsmoothed %>% filter(CHROM == "XV", POS > 840000, POS < 860000) %>%
  ggplot(aes(x = POS, y = LOD, shape = Parent, alpha = Bulk, color = Rep)) + 
    geom_vline(xintercept = c(849635,850054), color = "firebrick", size = 2, alpha = 0.5) + #RDL1
  geom_vline(xintercept = c(850280,850729), color = "darkorange", size = 2, alpha = 0.5) + #RDL2
  geom_point(size = 2) + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom") +
  ggtitle("RDL SNPs")

#The other middle region
CuSO4_A_LOD_unsmoothed %>% filter(CHROM == "XV", POS > 860000, POS < 870000) %>%
  ggplot(aes(x = POS, y = LOD, shape = Parent, alpha = Bulk, color = Rep)) + 
    geom_vline(xintercept = c(855147,860258), color = "#138803", size = 2, alpha = 0.5) + #SNF2
  geom_vline(xintercept = c(861175,865593), color = "#00563F", size = 2, alpha = 0.5) + #YPK9
  geom_point(size = 2) + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom") +
  ggtitle("SNPs between SNF2 and YPK9")

#Zoomed into weird part with specific genes marked
CuSO4_A_LOD_unsmoothed %>% filter(CHROM == "XV", POS > 976375, POS < 980000) %>%
  ggplot(aes(x = POS, y = LOD, shape = Parent, alpha = Bulk, color = Rep)) + geom_point(size = 2) + 
    geom_vline(xintercept = c(976421,976493), color = "violet", size = 2, alpha = 0.5) + #EMT2
   geom_vline(xintercept = c(975913,976244), color = "lightblue", size = 2, alpha = 0.5) + #YORWdelta24 LTR
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom") +
  ggtitle("SNPs between EMT2 and YORWdelta24 LTR")

```

## Look at the Log Ratios of Entire Dataset

```{r}
CuSO4_A_LOD %>% filter(CHROM != "I", CHROM != "M") %>%
  ggplot(aes(x = POS, y = LOD, linetype = Parent, alpha = Bulk, color = Rep)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

CuSO4_A_LOD %>% filter(CHROM %in% c("XIV", "XI", "XV", "VII")) %>%
  ggplot(aes(x = POS, y = LOD, linetype = Parent, alpha = Bulk, color = Rep)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("red", "blue", "black")) +
  scale_alpha_manual(values = c(1, 0.5)) +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

```
## Look at read quality along the weird regions

Load in the original data

```{r}

setwd("../../../GitHub/Sequencing/Analysis/")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)

parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) %>% filter(Unique == 1) -> pSNPs

mydatatotest = "Data/HKTFTDRX2.SortedCat.vcf.output.table"

CSSI_CuSO42_unfiltered <- cybrInputGATKTable(mydatatotest)


################################################################################

CSSI_CuSO42_unfiltered %>% filter(CHROM == "XV") %>% select(CHROM, POS, REF, ALT, Dataset, AD.REF, AD.ALT, GQ) -> CuSO42_XV

CuSO42_XV %>% ggplot(aes(x = POS, 
                         y = abs(log(as.numeric(AD.REF)/as.numeric(AD.ALT))),
                         color = GQ >= 99)) + geom_point(alpha = 0.3) +
  facet_grid(~GQ >= 99)

CuSO42_XV %>% filter(GQ >= 99) %>% ggplot(aes(x = POS, 
                         y = abs(log(as.numeric(AD.REF)/as.numeric(AD.ALT))),
                         color = GQ)) + geom_point(alpha = 0.3) + ggtitle("Chrom XV")

CuSO42_XV %>% filter(GQ >= 99, POS > 8e+05, POS < 1e+06) %>% ggplot(aes(x = POS, 
                         y = abs(log(as.numeric(AD.REF)/as.numeric(AD.ALT))),
                         color = GQ)) + geom_point(alpha = 0.3) + ggtitle("Chr XV Subset")

CuSO42_XV %>% filter(GQ >= 99, POS > 8e+05, POS < 1e+06) %>% ggplot(aes(x = POS, 
                         y = abs(log(as.numeric(AD.REF)/as.numeric(AD.ALT))),
                         color = as.numeric(AD.REF) == 0 | as.numeric(AD.ALT) == 0 )) + geom_point(alpha = 0.3) + ggtitle("Chr XV Subset")

CuSO42_XV %>% mutate(logodds = abs(log(as.numeric(AD.REF)/as.numeric(AD.ALT)))) %>% filter(logodds > 3, logodds != Inf) -> WeirdSequences

WeirdSequences %>% filter(GQ >= 99, POS > 8e+05, POS < 1e+06) %>% ggplot(aes(x = POS, 
                         y = abs(log(as.numeric(AD.REF)/as.numeric(AD.ALT))),
                         color = GQ)) + 
  geom_point(alpha = 0.3) + ggtitle("Chr XV Subset")


CuSO42_XV %>% mutate(logodds = abs(log(as.numeric(AD.REF)/as.numeric(AD.ALT)))) %>% filter(logodds < 3, logodds != Inf) -> PotentialPeaks

PotentialPeaks %>% filter(GQ >= 99, POS > 8e+05, POS < 1e+06) %>% ggplot(aes(x = POS, 
                         y = abs(log(as.numeric(AD.REF)/as.numeric(AD.ALT))),
                         color = GQ)) + 
  geom_vline(xintercept = c(822588,824867), color = "goldenrod", size = 2, alpha = 0.5) + #HRK1 
      geom_vline(xintercept = c(976421,976493), color = "violet", size = 2, alpha = 0.5) +
  geom_vline(xintercept = c(975913,976244), color = "lightblue", size = 2, alpha = 0.5) + #YORWdelta24 LTR 
  geom_vline(xintercept = c(849635,850054), color = "firebrick", size = 2, alpha = 0.5) + #RDL1
  geom_vline(xintercept = c(850280,850729), color = "darkorange", size = 2, alpha = 0.5) + #RDL2
  geom_vline(xintercept = c(855147,860258), color = "#138803", size = 2, alpha = 0.5) + #SNF2
  geom_vline(xintercept = c(861175,865593), color = "#00563F", size = 2, alpha = 0.5) + #YPK9
  geom_point(alpha = 0.3) + ggtitle("Chr XV Subset")

PotentialPeaks %>% filter(GQ >= 99, (POS > 822588 & POS < 824867) | #goldenrod
                            (POS > 975913 & POS < 976493) | #violet to light blue
                            (POS > 849635 & POS < 850729) | #firebrick to orange
                            (POS > 855147 & POS < 865593) )  -> Peaks #greens

table(Peaks$REF)
table(Peaks$ALT)

Peaks %>% ggplot(aes(x = POS, y = abs(log(as.numeric(AD.REF)/as.numeric(AD.ALT))),
                         color = REF %in% c("C", "A", "G", "T") == FALSE | 
                       ALT %in% c("C", "A", "G", "T") == FALSE)) +
  geom_point(alpha = 0.2) + theme(legend.position = "bottom")
  
```

```{r}
CSSI_CuSO42 <- cybrInputGATKTable(mydatatotest) %>% 
  cybrQualityFilter() %>% 
  cybrIDAlleles(BSAdfstart = ., Parentdf = pSNPs, yeast = TRUE) %>% 
  na.omit()

################################################################################
#Using Gsub for this

gsub(CSSI_CuSO42$Dataset, "HKTFTDRX2_n01_", "")
CSSI_CuSO42 %>% mutate(DShort = gsub("HKTFTDRX2_n01_", "", Dataset),
                       DS = gsub(".fastq", "", DShort)) %>% select(-Dataset, -DShort) -> CSSI_CuSO42

CSSI_CuSO42$Bulk <- NA
CSSI_CuSO42$Parent <- NA
CSSI_CuSO42$Rep <- NA

CSSI_CuSO42$Bulk[grep("C", CSSI_CuSO42$DS)] <- "CuSO4"
CSSI_CuSO42$Bulk[grep("D", CSSI_CuSO42$DS)] <- "Dilute"

CSSI_CuSO42$Rep[grep("a", CSSI_CuSO42$DS)] <- "A"
CSSI_CuSO42$Rep[grep("b", CSSI_CuSO42$DS)] <- "B"

CSSI_CuSO42$Parent[grep("O", CSSI_CuSO42$DS)] <- "Oak"
CSSI_CuSO42$Parent[grep("W", CSSI_CuSO42$DS)] <- "Wine"

CSSI_CuSO42$ReadCount <- as.numeric(CSSI_CuSO42$ReadCount)
 
CSSI_CuSO42 %>% select(-Type, -AltRef_Allele) %>% pivot_wider(names_from = PAllele, values_from = ReadCount) %>% mutate(logodds = log(Wine/Oak)) -> logoddsCuSO4

#Are there specific sequences that show up more in higher vs lower scores?
logoddsCuSO4 %>% ggplot(aes(x = POS, y = abs(logodds), color = REF %in% c("C", "A", "G", "T") | ALT %in% c("C", "A", "G", "T"))) +
  geom_point(alpha = 0.2) + facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

#Why might these regions be so off?
logoddsCuSO4 %>% ggplot(aes(x = POS, y = logodds, color = Bulk)) +
  geom_point(alpha = 0.2) + facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

logoddsCuSO4 %>% 
  select(CHROM, POS, DS, logodds) %>% 
  pivot_wider(names_from = DS, values_from = logodds) -> BulkComp

logoddsCuSO4 %>% 
  select(CHROM, POS, DS, logodds) %>% 
  group_by(CHROM, POS) %>% 
  summarize(PositionMean = mean(logodds, na.rm = TRUE),
            PositionSD = sd(logodds, na.rm = TRUE)) -> BulkSummary

#This shows that when there's a difference in bulks, the SD goes up
BulkSummary %>% filter(CHROM != "I", CHROM != "M") %>% ggplot(aes(x = POS, y = PositionMean, color = PositionSD)) + geom_point() + facet_grid(~CHROM, scales = "free", space = "free")

#What happens if the mean of the points is above 1?
BulkSummary %>% filter(abs(PositionMean) > 1, PositionMean != Inf) -> WeirdOnes

left_join(BulkComp, WeirdOnes) -> BulkComp_W

BulkComp_W %>% pivot_longer(c(-CHROM, -POS, -PositionMean, -PositionSD)) %>% 
ggplot(aes(x = POS, y = value, color = abs(PositionMean))) +
  geom_point(alpha = 0.2) + facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

BulkComp_W %>% na.omit() %>% pivot_longer(c(-CHROM, -POS, -PositionMean, -PositionSD)) %>% ggplot(aes(x = POS, y = value, color = name)) +
  geom_point(alpha = 0.2) + facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())
```

```{r}
# #THIS IGNORES REPLICATES
CSSI_CuSO42 %>% select(CHROM, POS, PAllele, ReadCount, Bulk, Parent, Rep) %>% distinct %>% 
 pivot_wider(names_from = c(Bulk, Parent, Rep, PAllele), values_from = ReadCount) -> CSSI_CuSO42_cybr2

saveRDS(object = CSSI_CuSO42_cybr2, file = "C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/RdataFiles/CSSI_CuSO42_cybr2.rds")
```

```{r}
CSSI_CuSO42_cybr2 <- readRDS("C:/Users/cassa/OneDrive/Documents/GitHub/Sequencing/Analysis/RdataFiles/CSSI_CuSO42_cybr2.rds")

CSSI_CuSO42_cybr2 %>% pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_") %>% 
  group_by(CHROM, POS, Bulk, Parent, Rep) %>% 
  summarize(Coverage = sum(value)) %>% ungroup() -> CuSO42_Coverage

CuSO42_Coverage %>% group_by(Bulk, Parent, Rep) %>%
  summarize(AvgC = mean(Coverage)) -> AvgCoverage

AvgCoverage %>% merge(CuSO42_Coverage) %>% mutate(RelativeCoverage = Coverage-AvgC) %>%
  ggplot(aes(x = POS, y = RelativeCoverage, color = Rep)) + 
  geom_point(alpha = 0.1) + 
  facet_grid(vars(Bulk), vars(Parent)) + 
  theme(legend.position = "bottom")

# CSSI_CuSO42_cybr2 %>% 
#   pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_") %>% 
#   group_by(CHROM, POS, Bulk, Parent, Rep) %>% 
#   summarize(Coverage = sum(value)) -> CoverageOverall

AvgCoverage %>% merge(CuSO42_Coverage) %>% mutate(RelativeCoverage = Coverage-AvgC) %>%
  select(-AvgC, -Coverage) %>%
  pivot_wider(names_from = c(Bulk, Parent, Rep), values_from = RelativeCoverage, names_sep = "_") %>% 
  pivot_longer(c(-CHROM, -POS)) %>%
  ggplot(aes(x = POS, y = value, color = name)) + 
  geom_point(alpha = 0.1) + facet_grid(~CHROM, space = "free", scale = "free") + theme(legend.position = "bottom")

AvgCoverage %>% merge(CuSO42_Coverage) %>% mutate(RelativeCoverage = Coverage-AvgC) %>%
  select(-AvgC, -Coverage) %>%
  pivot_wider(names_from = c(Bulk, Parent, Rep), values_from = RelativeCoverage, names_sep = "_") -> NormalizedCoverage

NormalizedCoverage %>% pivot_longer(c(-CHROM, -POS)) %>% summarize(mean = mean(value), sd = sd(value), absquant = quantile(abs(value), 0.95)) -> NormCov_Cutoffs

NormalizedCoverage %>% pivot_longer(c(-CHROM, -POS)) %>% filter(abs(value) < NormCov_Cutoffs$absquant) %>% pivot_wider(names_from = name, values_from = value) %>% na.omit() %>% select(CHROM, POS) -> filteredCuSO4_cov_POS


  
  
```

Re-running this glm to test

```{r}
#Unsmoothed
CSSI_CuSO42_cybr2 %>% right_join(filteredCuSO4_cov_POS) %>% group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed_rep(HOOa = CuSO4_Oak_A_Oak, 
                           HOWa = CuSO4_Oak_A_Wine, 
                           HWOa = CuSO4_Wine_A_Oak,
                           HWWa = CuSO4_Wine_A_Wine,
                           LOOa = Dilute_Oak_A_Oak,
                           LOWa = Dilute_Oak_A_Wine,
                           LWOa = Dilute_Wine_A_Oak, 
                           LWWa = Dilute_Wine_A_Wine,
                           
                           HOOb = CuSO4_Oak_B_Oak, 
                           HOWb = CuSO4_Oak_B_Wine, 
                           HWOb = CuSO4_Wine_B_Oak,
                           HWWb = CuSO4_Wine_B_Wine,
                           LOOb = Dilute_Oak_B_Oak,
                           LOWb = Dilute_Oak_B_Wine,
                           LWOb = Dilute_Wine_B_Oak, 
                           LWWb = Dilute_Wine_B_Wine)[1:5],
                                                   label = c("intercept", "Bulk", "Parent", "Rep", "Interaction")) -> CSSI_CuSO42_glm_covFILTERED

saveRDS(CSSI_CuSO42_glm_covFILTERED, file = "2023/CSSI_CuSO42_glm_covFILTERED.Rdata")

#Smoothed
CSSI_CuSO42_cybr2 %>% right_join(filteredCuSO4_cov_POS) %>% 
  cybr2_rollmean() %>%
  group_by(CHROM, POS) %>% 
  summarise(summary = glmfixed_rep(HOOa = CuSO4_Oak_A_Oak, 
                           HOWa = CuSO4_Oak_A_Wine, 
                           HWOa = CuSO4_Wine_A_Oak,
                           HWWa = CuSO4_Wine_A_Wine,
                           LOOa = Dilute_Oak_A_Oak,
                           LOWa = Dilute_Oak_A_Wine,
                           LWOa = Dilute_Wine_A_Oak, 
                           LWWa = Dilute_Wine_A_Wine,
                           
                           HOOb = CuSO4_Oak_B_Oak, 
                           HOWb = CuSO4_Oak_B_Wine, 
                           HWOb = CuSO4_Wine_B_Oak,
                           HWWb = CuSO4_Wine_B_Wine,
                           LOOb = Dilute_Oak_B_Oak,
                           LOWb = Dilute_Oak_B_Wine,
                           LWOb = Dilute_Wine_B_Oak, 
                           LWWb = Dilute_Wine_B_Wine)[1:5],
                                                   label = c("intercept", "Bulk", "Parent", "Rep", "Interaction")) -> CSSI_CuSO42_glm_covFILTEREDs

saveRDS(CSSI_CuSO42_glm_covFILTEREDs, file = "2023/CSSI_CuSO42_glm_covFILTEREDss.Rdata")

```

Look at LODs of the filtered data

```{r}
CSSI_CuSO42_cybr2 %>% right_join(filteredCuSO4_cov_POS) %>%
  cybr2_rollmean() %>%
  pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_") -> test

test %>% pivot_wider(names_from = Allele, values_from = value) %>% mutate(logOdds = log(Wine/Oak)) -> CuSO4_filtcov_logodds

CuSO4_filtcov_logodds %>% mutate(Flask = paste(Bulk, Parent, Rep, sep = "_")) %>% filter(CHROM != "I") %>%
  ggplot(aes(x = POS, y = logOdds, color = Flask)) + geom_point() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

CuSO4_filtcov_logodds %>% mutate(Flask = paste(Bulk, Parent, Rep, sep = "_")) %>% filter(CHROM == "XV") %>%
  ggplot(aes(x = POS, y = logOdds, color = Flask)) + geom_point() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

#Without Rollmeans
CSSI_CuSO42_cybr2 %>% right_join(filteredCuSO4_cov_POS) %>%
  pivot_longer(c(-CHROM, -POS), names_to = c("Bulk", "Parent", "Rep", "Allele"), names_sep = "_")  %>%
  pivot_wider(names_from = Allele, values_from = value) %>% mutate(logOdds = log(Wine/Oak)) %>%
  mutate(Flask = paste(Bulk, Parent, Rep, sep = "_")) %>% filter(CHROM != "I") -> woRollmean_Filter

woRollmean_Filter %>% filter(CHROM == "XV") %>%
  ggplot(aes(x = POS, y = logOdds, shape = Flask, color = Bulk)) + geom_point() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

woRollmean_Filter %>%
  ggplot(aes(x = POS, y = logOdds, shape = Flask, color = Bulk)) + geom_point(alpha = 0.4) + 
  facet_grid(~CHROM, scales = "free", space = "free") + 
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())
```

Look at the GLMs of the filtered data

```{r}
CSSI_CuSO42_glm_covFILTERED %>% filter(label != "intercept", CHROM != "I") %>% ggplot(aes(x = POS, y = summary, color = label)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank())

CSSI_CuSO42_glm_covFILTEREDs %>% filter(label != "intercept", CHROM != "I") %>% ggplot(aes(x = POS, y = summary, color = label)) + geom_line() + 
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom", axis.text.x=element_blank(),
            axis.ticks.x=element_blank()) + ggtitle("CuSO4 Filtered by Coverage")
```

