---
title: "QTLComparison_July24"
author: "Cass"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))



## Functions for GLM all and Permutation all

glm_cb2_all <- function(..., W, formula, numgroups = FALSE, outputlength = 8) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  output <- summary(glm_fit)$coefficients[c(((length(summary(glm_fit)$coefficients)*0)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.25)),
                                            ((length(summary(glm_fit)$coefficients)*0.5)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.75)))]

  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
} 


################################################################################
# No replicates, fix Parent
cybrPermute_cb2_all <- function(dataset, 
                                R = 0, perp = 1, outlength = 4, 
                                glmform = "Allele ~ Bulk * Parent", 
                                inputlabels = c("Intercept", "Bulk", "Parent", "Interaction")){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled aD")
  print(paste("Your labels are:", inputlabels))

    if(R > 0){ #INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Rep, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == 1) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest    
    }else{ #DO NOT INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == perp) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    }
  
    #PERMUTE TWICE
  if(R > 0){
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Rep, 
             Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep, 
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, 
             Rep, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep,
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }else{
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #RUN THE GLM
    if(R > 0) {
      shuffletoglm2 %>% na.omit() %>%
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }else{
      shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }
    
    
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}

cybrInputGATKTable2 <- function(rawData, yeast = TRUE){

  require(dplyr)
  require(doParallel)
  require(foreach)

  HNGLCDRXY <- read.table(rawData, header = TRUE)

  #Identify the unique values besides AD/DP/GQ/PL
  gsub(".AD", "",
       gsub(".GQ", "",
            gsub(".DP","",
                 gsub(".PL","",
                      colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
  #i <- Samples[1]

  resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
    mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
    AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
    GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
    DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
    PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
    cbind(mydf, AD , GQ , DP, PL) -> mydftotal
    colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")

    mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>%
      separate(PL, c('PL.REF','PL.ALT'), extra='drop') %>%
      #Added 10/18/23:
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf

    mycdf %>% filter(grepl(",", ALT)) %>% 
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
    
    doublecdf %>% filter(grepl(",", ALT)) %>%
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
    
    rbind(mycdf, doublecdf, triplecdf) -> newcdf
    
    newcdf
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
  }else{
    results <- resultscdf
  }
  return(results)

}

#CIRCLIZE
cybr_circos <- function(d1, d8, peaklist1 = NULL, peaklist8 = NULL, maxy = NULL, color1 = "#24588A", color8 = "#ED7B01", opacity = "50"){
  
  color1fade_50 <- paste(color1, "50", sep = "")
  color8fade_50 <- paste(color8, "50", sep = "")
  
  color1fade <- paste(color1, opacity, sep = "")
  color8fade <- paste(color8, opacity, sep = "")

  if(is.null(d8)){
    d8 <- d1
    include8 <- FALSE
  }else{
    include8 <- TRUE
  }
    #SET UP DATA
  df8 <- data.frame(sectors = as.character(d8$CHROM),
                 x = d8$POS,
                 y = abs(d8$summary),
                  label = d8$label)

  df1 <- data.frame(sectors = as.character(d1$CHROM),
                 x = d1$POS,
                 y = abs(d1$summary),
                  label = d1$label)

  #Take just the interactions of each
  df8int <- df8 %>% filter(label == "Interaction")
  df1int <- df1 %>% filter(label == "Interaction")

    #REORDER THE CHROMOSOMES
  df8int$sectors <- factor(df8int$sectors, levels = as.character(as.roman(1:16)))
  df1int$sectors <- factor(df1int$sectors, levels = as.character(as.roman(1:16)))
  
  ##############################################################################
  dfall <- rbind(df1int, df8int) %>% na.omit()
  circos.par("track.height" = 0.3, start.degree = 90, cell.padding = c(0,0))
  circos.initialize(sectors = dfall$sectors, x = dfall$x)

  if(is.null(maxy)){
    #I think this makes the sizes?
    circos.track(ylim = c(max(c(dfall$y)), 0), dfall$sectors, y = dfall$y, 
        panel.fun = function(x, y) {
            circos.text(CELL_META$xcenter, 
                0 - mm_y(5), 
                CELL_META$sector.index,
                niceFacing = FALSE)
            circos.axis(labels.cex = 0.1)
    })
  }else{
    circos.track(ylim = c(maxy, 0), dfall$sectors, y = dfall$y, 
      panel.fun = function(x, y) {
          circos.text(CELL_META$xcenter, 
              0 - mm_y(5), 
              CELL_META$sector.index,
              niceFacing = FALSE)
          circos.axis(labels.cex = 0.1)
      })
  }
  #Makes the chromosome overlap parts
  #CHROMOSOME I
  draw.sector(83.5, #RIGHT
              90, #LEFT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color1fade_50, border = FALSE)
  if(include8 == TRUE){
    #CHROMOSOME VIII
    draw.sector(289.5, #LEFT
              305.4, #RIGHT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color8fade_50, border = FALSE)
  }
  
  
  #Makes the lines
  circos.trackPoints(df8int$sectors, df8int$x, abs(df8int$y), col = color8, pch = 16, cex = 0.1)
  circos.trackPoints(df1int$sectors, df1int$x, abs(df1int$y), col = color1, pch = 16, cex = 0.1)
  
  if(is.null(peaklist8) == FALSE){
    if(length(peaklist8$POS) >= 1){
      for(i in 1:length(peaklist8$POS)){
      circos.link(peaklist8$CHROM[i], 
                  peaklist8$POS[i], 
                  #Add 8 after
                  "VIII", c(0, max(dfall$x[dfall$sectors == "VIII"])),  
                  
                  col = color8fade, 
                  h.ratio = 0.3, 
                  border = color8fade, 
                  lwd = 1)
      }
    }else{
      print("No interactions on Chr 8")
    }
    
  }

  if(is.null(peaklist1) == FALSE){
    if(length(peaklist1$POS) >= 1){
      for(i in 1:length(peaklist1$POS)){
        circos.link("I",c(0, max(dfall$x[dfall$sectors == "I"])),  
                    #add 1 first
                    peaklist1$CHROM[i], 
                    peaklist1$POS[i], 
                    col = color1fade, 
                    h.ratio = 0.3, 
                    border = color1fade, 
                    lwd = 1)
      }
    }else{
      print("No interactions on Chr 1")
    }
  }  
  
  circos.clear()
} 
```


# Load in Datasets

```{r, eval = FALSE}
Bloom <- read.csv("C:/Users/Cassandra/Documents/R_FileTransfer/Bloom_CuSO4Interactions.csv")
Costanzo <- read.table("C:/Users/Cassandra/Documents/R_FileTransfer/SGA_NxN.txt", header = TRUE, sep = "\t")
Nguyen <- read.csv("C:/Users/Cassandra/Documents/R_FileTransfer/Nguyen_Cu.csv")
Matsui <-  read.table("C:/Users/Cassandra/Downloads/Two-locus_interactions.txt", header = TRUE, sep = "\t")

################################################################################
Costanzo %>% filter(grepl("cup1", Query.allele.name)) -> CUP1_query

CUP1_query %>% ggplot(aes(x = P.value, y = Genetic.interaction.score..ε.)) + geom_point(aes(color = P.value < 0.05)) +
  theme(axis.text.x = element_text())

sgd_orfs %>% filter(Gene.chromosome.primaryIdentifier == "chrVIII") -> ChrVIII_Genes


CUP1_query %>% filter(P.value < 0.05) %>% select(Array.allele.name, E = Genetic.interaction.score..ε.) -> CUP1Interactors


CUP1Interactors %>% 
  #OPTIONAL FILTER
  #filter(abs(E) > 0.2) %>% 
  mutate(Gene.symbol = toupper(Array.allele.name)) %>% merge(sgd_orfs) %>% 
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) %>%
  group_by(Array.allele.name) %>%
  mutate(POS = mean(c(Gene.chromosomeLocation.end, Gene.chromosomeLocation.start))) %>%
  select(CHROM, POS, E) %>% mutate(C = "VIII") -> CUP1_Interactions

CUP1_Interactions %>% ggplot(aes(x = CHROM)) + geom_bar() + theme(axis.text.x = element_text())

# Costanzo %>% mutate(Gene.symbol = toupper(Array.allele.name)) %>% merge(sgd_orfs) %>%
#   ggplot(aes(x = Gene.chromosome.primaryIdentifier)) + geom_bar() + theme(axis.text.x = element_text())

saveRDS(CUP1_Interactions, file = "Costanzo_CUP1_Interactions.rds")


################################################################################

Bloom %>% filter(Q1_chr == "chrVIII") %>% select(POS = Q2_pos, CHROM = Q2_chr, Chr8POS = Q1_pos) %>% mutate(CHROM = gsub("chr", "", CHROM)) -> Q1_Bloom
Bloom %>% filter(Q2_chr == "chrVIII") %>% select(POS = Q1_pos, CHROM = Q1_chr, Chr8POS = Q2_pos) %>% mutate(CHROM = gsub("chr", "", CHROM)) -> Q2_Bloom

rbind(Q1_Bloom, Q2_Bloom) %>% mutate(C = "VIII") -> BloomVIII_Interactions

saveRDS(BloomVIII_Interactions, file = "BloomVIII_Interactions.rds")

################################################################################

Nguyen %>% filter(Chromosome.1 == 8) %>% 
  select(CHROM = Chromosome.2, POS = Lead.SNP.2.Position, Chr8POS = Lead.SNP.1.Position, Gene = Lead.SNP.2.Gene) %>% 
  mutate(C = "VIII",                                                                              
         CHROM = as.character(as.roman(CHROM))) -> Q1_Nguyen

Nguyen %>% filter(Chromosome.2 == 8) %>% 
  select(CHROM = Chromosome.1, POS = Lead.SNP.1.Position, Chr8POS = Lead.SNP.2.Position, Gene = Lead.SNP.1.Gene) %>% 
  mutate(C = "VIII",
                                                                                                            
         CHROM = as.character(as.roman(CHROM))) -> Q2_Nguyen

rbind(Q1_Nguyen, Q2_Nguyen) -> NguyenVIII_Interactions

saveRDS(NguyenVIII_Interactions, file = "NguyenVIII_Interactions.rds")

################################################################################

Matsui %>% filter(c1 == 8, Environment == "CuSO4") %>% select(CHROM = c2, POS = p2, C8_POS = p1, LOD) %>% mutate(C = "VIII") -> MC1
Matsui %>% filter(c2 == 8, Environment == "CuSO4") %>% select(CHROM = c1, POS = p1, C8_POS = p2, LOD) %>% mutate(C = "VIII") -> MC2

rbind(MC1, MC2) %>% mutate(CHROM = as.character(as.roman(CHROM))) -> MatsuiVIII_Interactions

saveRDS(MatsuiVIII_Interactions, file = "MatsuiVIII_Interactions.rds")

rm(Costanzo)
rm(Bloom)
rm(Nguyen)
```

```{r}
#Load data for interaction networks
CUP1_Interactions <- readRDS("Costanzo_CUP1_Interactions.rds")
BloomVIII_Interactions <- readRDS("BloomVIII_Interactions.rds")
NguyenVIII_Interactions <- readRDS("NguyenVIII_Interactions.rds")
MatsuiVIII_Interactions <- readRDS("MatsuiVIII_Interactions.rds")
```


```{r}
CuSO4_CSS1_glmer_1byRep <- readRDS("CuSO4_CSS1_glmer_1byRepG.rds")
CuSO4_CSS8_glmer_1byRep <- readRDS("CuSO4_CSS8_glmer_1byRepG.rds")
q5_all <- readRDS("q5_CuSO4_CSS.rds")


CuSO4_CSS1_glmer_1byRep %>% filter(CHROM != "I") %>% mutate(CSS = "I") -> temp1
CuSO4_CSS8_glmer_1byRep %>% mutate(CSS = "VIII") %>%
  rbind(temp1) -> AllCuSO4_glmer_1byRep

rm(temp1)

AllCuSO4_glmer_1byRep %>% filter(CHROM == "VIII", label == "Bulk") %>% ggplot(aes(x = POS, y = zscore)) +
  geom_line() +
  geom_vline(data = BloomVIII_Interactions, aes(xintercept = Chr8POS), color = "gray", linetype = "dashed") +
  geom_vline(data = NguyenVIII_Interactions, aes(xintercept = Chr8POS), color = "red") +
  theme(axis.text.x = element_text())

```

### Plot overlaps between datasets

```{r, fig.width=16, fig.height=5}


AllCuSO4_glmer_1byRep %>% filter(label %in% c("Bulk", "Interaction"), CSS == "VIII") %>% ggplot(aes(x = POS, y = abs(zscore), linetype = CSS, color = label)) +
  geom_vline(data = CUP1_Interactions, aes(xintercept = POS, alpha = abs(E)), color = "lightblue") +
  geom_line() +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("black", "purple")) +
  ggtitle("Costanzo Data by E")

AllCuSO4_glmer_1byRep %>% filter(label %in% c("Bulk", "Interaction"), CSS == "VIII") %>% ggplot(aes(x = POS, y = abs(zscore), linetype = CSS, color = label)) +
  geom_vline(data = MatsuiVIII_Interactions, aes(xintercept = POS, size = LOD), alpha = 0.3, color = "lightblue") +
  geom_line() +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("black", "purple")) +
  ggtitle("Matsui data by LOD")

# 
# AllCuSO4_glmer_1byRep %>% filter(label == "Interaction") %>% ggplot(aes(x = POS, y = abs(zscore), linetype = CSS)) +
#   geom_line() +
#   geom_vline(data = BloomVIII_Interactions, aes(xintercept = POS), color = "purple", linetype = "dashed") +
#   geom_vline(data = NguyenVIII_Interactions, aes(xintercept = POS), color = "red") +
#   facet_grid(~CHROM, scales = "free", space = "free")
# 
# NguyenVIII_Interactions %>% filter(Chr8POS < 2.5e5, Chr8POS > 1.5e5)


AllCuSO4_glmer_1byRep %>% filter(label %in% c("Bulk"), CSS == "VIII") %>% 
  ggplot(aes(x = POS, y = CHROM)) +
  geom_tile(aes(color = abs(zscore)),na.rm = TRUE) +
  geom_vline(data = BloomVIII_Interactions, aes(xintercept = POS), color = "orange",  size = 1) +
  geom_vline(data = NguyenVIII_Interactions, aes(xintercept = POS), color = "red",size = 1) +
  facet_grid(rows = "CHROM", scales = "free", space = "free") +
  scale_color_gradient2(low = "gray", mid = "black", high = "cyan", midpoint = 6)+
  ggtitle("Bulk Effects")

AllCuSO4_glmer_1byRep %>% filter(label %in% c("Interaction"), CSS == "VIII") %>% 
  ggplot(aes(x = POS, y = CHROM)) +
  geom_tile(aes(color = abs(zscore)),na.rm = TRUE) +
  geom_vline(data = BloomVIII_Interactions, aes(xintercept = POS), color = "orange",  size = 1) +
  geom_vline(data = NguyenVIII_Interactions, aes(xintercept = POS), color = "red",size = 1) +
  facet_grid(rows = "CHROM", scales = "free", space = "free") +
  scale_color_gradient2(low = "gray", mid = "black", high = "cyan", midpoint = 2)+
  ggtitle("Interaction Effects")

AllCuSO4_glmer_1byRep %>% filter(label %in% c("Bulk", "Interaction"), CSS == "VIII") %>% ggplot(aes(x = POS, y = abs(zscore), linetype = CSS, color = label)) +
  geom_vline(data = BloomVIII_Interactions, aes(xintercept = POS, size = CHROM == "XI"), color = "orange") +
  geom_vline(data = NguyenVIII_Interactions, aes(xintercept = POS, size = CHROM %in% c("XVI", "XV")), color = "red") +
  geom_vline(data = MatsuiVIII_Interactions, aes(xintercept = POS, size = CHROM == "XV"), color = "skyblue") +
  geom_line() +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("black", "purple"))+
  scale_size_manual(values = c(0.5, 2))


```

### What genes are these?

**XII | LMO1**: Homolog of mammalian ELMO (Engulfment and celL MOtility); upstream component for regulation through the small GTPase Rho5p; may form a complex with Dck1p that acts as a GEF for Rho5p; cytoplasmic protein that relocates to mitochondria under **oxidative stress**; implicated in mitophagy; not an essential protein

**XIV | rps7b**: Protein component of the small (40S) ribosomal subunit; interacts with Kti11p; deletion causes hypersensitivity to zymocin; homologous to mammalian ribosomal protein S7, no bacterial homolog; RPS7B has a paralog, RPS7A, that arose from the whole genome duplication; protein abundance increases in response to **DNA replication stress**

**XV | IRA2**: GTPase-activating protein; negatively regulates RAS by converting it from the GTP- to the GDP-bound inactive form, required for **reducing cAMP levels under nutrient limiting conditions**; IRA2 has a paralog, IRA1, that arose from the whole genome duplication; defects in human homolog NF1 are associated with neurofibromatosis

**XV | LSC1**: Alpha subunit of succinyl-CoA ligase; succinyl-CoA ligase is a mitochondrial enzyme of the TCA cycle that catalyzes the nucleotide-dependent conversion of succinyl-CoA to succinate; phosphorylated

**XVI | SKS1**: Putative serine/threonine protein kinase; involved in the adaptation to **low concentrations of glucose** independent of the SNF3 regulated pathway; SKS1 has a paralog, VHS1, that arose from the whole genome duplication  

* Is under additive but not interaction peak


```{r}
NguyenVIII_Interactions %>% filter(CHROM == "XVI") #SKS1

NguyenVIII_Interactions %>% filter(CHROM == "XV") #IRA2,LSC1

NguyenVIII_Interactions %>% filter(CHROM == "XIV") #rps7b

NguyenVIII_Interactions %>% filter(CHROM == "XII") %>% arrange(POS) #LMO1

```

