---
title: "MergedVCFs_CuSO4_Gaussian_June24"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))


```

# Pre-processing

## Load in table and match to oak or wine parents

```{r}
# setwd("C:/Users/Cassandra/Documents/GitHub/Sequencing/Analysis/")
# dir()

parentSNPids <- cybrConvertParentalAlleles(Truncate = FALSE)
parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) -> pSNPs

merge(pSNPs, parentSNPids) %>% filter(Unique == 1) %>% select(-Unique) -> parentSNPids

unique(pSNPs$Unique)

```

## Exclude duplicates 

```{r, message = FALSE, warning=FALSE, fig.width=12, fig.height=4}
myfile <- "C:/Users/Cassandra/Documents/Data_NYU/Sequencing/HWMMFDMXY.SortedCat.vcf.output.table"

#GET RID OF EXCLUDED ONES
read.table(myfile, header = TRUE) -> testrawdata
testrawdata %>% select(CHROM, POS, ALT) %>% separate(ALT, into = c("First", "Second"), sep = ",") %>% na.omit() -> exclude

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

exclude %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> exclude

exclude %>% mutate(locus = paste(CHROM, POS)) -> exclKEY

cybrInputGATKTable2(myfile) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> rawdata

saveRDS(rawdata, file = "CSSI_A24_rawdataG.rds")

rawdata %>% filter(POS == 801680, CHROM == "IV")

################################################################################
rawdata %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% mutate(locus = paste(CHROM, POS)) %>% filter(locus %in% exclKEY$locus == FALSE) %>% select(-locus) -> rawdata_called_ex


################################################################################

saveRDS(rawdata_called_ex, file = "CSSI_A24_rawdata_called_ex.rds")
```

```{r, fig.width=12, fig.height=5, eval = FALSE}
#Testing what this data actually looks like
rawdata_called_ex %>% pivot_wider(names_from = Allele, values_from = Reads) %>% unnest() %>% 
  mutate(logodds = log(Wine/Oak)) -> test_rawdatacalledex

test_rawdatacalledex %>% filter(grepl("_AIG_", Dataset) == FALSE, CHROM %in% c("I", "M", "III") == FALSE,
                                grepl("_D_", Dataset)) %>%
  ggplot(aes(x = POS, y = logodds, color = grepl("_W_", Dataset))) + geom_point(alpha = 0.3, size = 0.1) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom")

for(i in unique(test_rawdatacalledex$CHROM)){
  test_rawdatacalledex %>% filter(grepl("_AIG_", Dataset) == FALSE, CHROM == i,
                                grepl("_D_", Dataset)) %>%
  ggplot(aes(x = POS, y = logodds, color = grepl("_W_", Dataset))) + geom_point(alpha = 0.3, size = 0.1) +
  facet_grid(~CHROM, scales = "free", space = "free") + ggtitle(i) +
  theme(legend.position = "bottom") -> newplot
  
  print(newplot)
}

```

Narrowing segments that look weird specifically (segregation issues)

```{r}
# test_rawdatacalledex %>% filter(CHROM == "XVI") %>% filter(POS < 150000, POS > 120000) %>%
#   ggplot(aes(x = POS, y = Oak, color = grepl("_W_", Dataset))) + geom_point(alpha = 0.3) +
#   theme_minimal()+ ggtitle("Chr Oak XVI Coverage Zoomed In")
# 
# test_rawdatacalledex %>% filter(CHROM == "XI") %>% filter(POS > 3.05e5, POS < 3.2e5) %>%
#   ggplot(aes(x = POS, y = Wine, color = grepl("_W_", Dataset))) + geom_point(alpha = 0.3) +
#   theme_minimal()+ ggtitle("Chr Oak XI Coverage Zoomed In")
```

### Filter out weird coverage and stuff

```{r}
# exclusion_POS <- data.frame(CHROM = c("XI", "XVI"),
#                             Start = c(3.06e5, 128000),
#                             End = c(3.2e5, 142000))
# 
# rawdata_called_ex %>% filter((CHROM == "XI" & POS < 320000 & POS > 306000)== FALSE,
#                              (CHROM == "XVI" & POS < 142000 & POS > 128000)== FALSE) -> rawdata_called_exx
# 
# saveRDS(rawdata_called_exx, file = "CSSI_A24_rawdata_called_exx.rds")

```

### Separate out what could be smoothed here

```{r}

# cybr_weightedgauss <- function(myx){
#   myy <- dnorm(1:length(myx), mean = length(myx)/2, sd = 10)
#   return(weighted.mean(x = myx, y = myy, na.rm = TRUE))
# } 

# plot(dnorm(1:200, mean = 200/2, sd = 10), main = "SD 10")
# plot(dnorm(1:200, mean = 200/2, sd = 20), main = "SD 20")
# plot(dnorm(1:200, mean = 200/2, sd = 100), main = "SD 100") #choosing this one for including all of the points
# plot(dnorm(1:200, mean = 200/2, sd = 1000), main = "SD 1000")

rawdata_called_ex <- readRDS("CSSI_A24_rawdata_called_ex.rds")
# 
# unique(rawdata_called_exx$Dataset)

rawdata_called_ex %>% ungroup() %>% 
  group_by(Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) %>% mutate(smooth = 200) -> rawdata_smoothed_G200x

rawdata_smoothed_G200x %>% mutate(SmoothCount = ceiling(SmoothCount)) %>% na.omit() -> rawdata_smoothed_G200x

saveRDS(rawdata_smoothed_G200x, file = "CSSI_A24_rawdata_smoothed_G200x.rds")

```


## Separate into bulks, make ready for glms

```{r}

# rawdata_called_ex <- readRDS("CSSI_A24_rawdata_smoothed_G200.rds")
#rawdata_smoothed_G200 <- readRDS("CSSI_A24_rawdata_smoothed_G200.rds")
rawdata_smoothed_G200 <- readRDS("CSSI_A24_rawdata_smoothed_G200x.rds")

################################################################################
#FIXING MIXUP

# rawdata_smoothed_G200 %>% mutate(Dataset = gsub("E_O_B_x", "blank", Dataset),
#                                  Dataset = gsub("H_O_B_x", "E_O_B_x", Dataset)) %>%
#   mutate(Dataset = gsub("blank", "H_O_B_x", Dataset)) -> rawdata_smoothed_G200_corrected
                                   
  
################################################################################
# rawdata_called_ex %>% filter(grepl("AIG", Dataset) == FALSE) %>%
#   mutate(Dataset = gsub("HWMMFDMXY_n01_", "", Dataset)) %>%
#   mutate(Dataset = gsub(".fastq", "", Dataset)) %>% 
#   separate(Dataset, into = c("Bulk","Parent","Rep","Day"), sep = "_") -> rd_named

rawdata_smoothed_G200 %>% filter(grepl("AIG", Dataset) == FALSE) %>%
  mutate(Dataset = gsub("HWMMFDMXY_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>% 
  separate(Dataset, into = c("Bulk","Parent","Rep","Day"), sep = "_") -> smoothdata_named

smoothdata_named %>% mutate(Day = toupper(Day)) -> smoothdata_named

# 
# 
# unique(rawdata_smoothed_G200$Dataset) #Found out I was missing all of these whoops
# 
# unique(smoothdata_named$Bulk)
# unique(smoothdata_named$Parent)
# unique(smoothdata_named$Rep)
# 
# table(smoothdata_named$Bulk,smoothdata_named$Day)
# 
# table(smoothdata_named$Bulk)
```


## Change into factors

```{r}

smoothdata_named %>% filter(Bulk %in% c("D", "C")) %>% mutate(Bulk = gsub("D", "A", Bulk))  %>% na.omit() %>%
  mutate_if(is.character, as.factor) %>% mutate(Sel = "C") -> rd_factor_cyclohex

smoothdata_named %>% filter(Bulk %in% c("D", "E")) %>% mutate(Bulk = gsub("D", "A", Bulk))  %>% na.omit() %>%
  mutate_if(is.character, as.factor) %>% mutate(Sel = "E") -> rd_factor_ethanol

smoothdata_named %>% filter(Bulk %in% c("D", "H")) %>% mutate(Bulk = gsub("D", "A", Bulk))  %>% na.omit() %>%
  mutate_if(is.character, as.factor) %>% mutate(Sel = "H") -> rd_factor_H2O2


contrasts(rd_factor_cyclohex$Parent) <- matrix(c(0.5, -0.5))
contrasts(rd_factor_ethanol$Parent) <- matrix(c(0.5, -0.5))
contrasts(rd_factor_H2O2$Parent) <- matrix(c(0.5, -0.5))

contrasts(rd_factor_cyclohex$Bulk) <- matrix(c(0, 1))
contrasts(rd_factor_ethanol$Bulk) <- matrix(c(0, 1))
contrasts(rd_factor_H2O2$Bulk) <- matrix(c(0, 1))

saveRDS(rd_factor_cyclohex, "CSSI_A24_rd_factor_cyclohex.rds")
saveRDS(rd_factor_ethanol, "CSSI_A24_rd_factor_ethanol.rds")
saveRDS(rd_factor_H2O2, "CSSI_A24_rd_factor_H2O2.rds")

```

################################################################################

# Running GLMs

```{r}
rd_factor_cyclohex <- readRDS("CSSI_A24_rd_factor_cyclohex.rds") 
rd_factor_ethanol <- readRDS("CSSI_A24_rd_factor_ethanol.rds")
rd_factor_H2O2 <- readRDS("CSSI_A24_rd_factor_H2O2.rds")

rd_factor <- rbind(rd_factor_cyclohex,rd_factor_ethanol,rd_factor_H2O2)

```

## Checking raw data for sanity

```{r, eval = FALSE}
rd_factor %>% pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(logodds = log(Wine/Oak)) -> testing_rdfactor

testing_rdfactor %>% na.omit()

# testing_rdfactor %>% filter(CHROM %in% c("M", "VIII", "I") == FALSE) %>%
#   mutate(logodds = logodds + 1) %>%
#   ggplot(aes(x = POS, y = logodds, color = paste(Parent, Day, Rep))) + geom_line() +
#   facet_grid(Bulk~CHROM, scales = "free", space = "free") +
#   theme(legend.position = "bottom")


testing_rdfactor %>% filter(CHROM  == "XVI") %>%
  ggplot(aes(x = POS, y = abs(logodds), color = paste(Day, Rep))) + geom_line() +
  facet_grid(Bulk~Parent, scales = "free", space = "free") +
  theme(legend.position = "bottom")+ ggtitle("Chr XVI")

testing_rdfactor %>% filter(CHROM  == "XV") %>%
  ggplot(aes(x = POS, y = abs(logodds), color = paste(Day, Rep))) + geom_line() +
  facet_grid(Bulk~Parent, scales = "free", space = "free") +
  theme(legend.position = "bottom") + ggtitle("Chr XV")

testing_rdfactor %>% filter(CHROM  == "VII") %>%
  ggplot(aes(x = POS, y = abs(logodds), color = paste(Day, Rep))) + geom_line() +
  facet_grid(Bulk~Parent, scales = "free", space = "free") +
  theme(legend.position = "bottom")+ ggtitle("Chr VII")

############################################
testing_rdfactor %>% filter(CHROM %in% c("M", "VIII", "I", "III") == FALSE, Bulk == "A") %>%
  ggplot(aes(x = POS, y = abs(logodds), color = Parent)) + geom_line() +
  facet_grid(paste(Day, Rep)~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom")

```

### Glmer for when we have more time would add to this

```{r}
#glm test
# rd_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
#   filter(POS == 5627) -> test_5627
# 
# table(test_5627$Bulk)
# table(test_5627$Bulk, test_5627$Day)

###############

rd_factor %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #filter(POS == 5627) %>%
  group_by(CHROM, POS, Sel) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Parent,
                                                            #Rep = Rep,
                                                            #Day = Day,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> Aug24_CSS1_glm_day

saveRDS(Aug24_CSS1_glm_day, file = "Aug24_CSS1_glm_day.rds")

```

### Permutations using glmer

```{r}
rd_factor %>% filter(Bulk == "A",
                     CHROM %in% c("I", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  #group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Parent,
            Day = Day, Rep = Rep) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

rd_factor %>% filter(Bulk == "A",
                     CHROM %in% c("I", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  #group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Parent,
            Day = Day, Rep = Rep) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Unselected") -> rd_shuffled_unselected

Perm8 <- rbind(rd_shuffled_unselected, rd_shuffled_selected)
Perm8 %>% ungroup() %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") %>% unnest() %>%
  mutate(POS = as.numeric(POS)) %>%
  mutate_if(is.character, as.factor) -> Perm8contrasts

contrasts(Perm8contrasts$Background) <- matrix(c(-0.5, 0.5))
contrasts(Perm8contrasts$Bulk) <- matrix(c(1,0))


Perm8 %>% mutate(logodds = log(Wine/Oak)) %>%
  ggplot(aes(x = logodds, color = Background)) + geom_density() +
  geom_vline(xintercept = 0)


Perm8contrasts %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            #Rep = Rep,
                                                            #Day = Day,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> Aug24_CSS1_glm_PERM_day

saveRDS(Aug24_CSS1_glm_PERM_day, file = "Aug24_CSS1_glm_PERM_day.rds")

Aug24_CSS1_glm_PERM_day %>% ungroup() %>% group_by(label) %>% summarize(q5 = quantile(abs(zscore), 0.95)) -> q5_CSSI_Aug24

q5_CSSI_Aug24
```


# Plot New Data

```{r, fig.width=12, fig.height=5}
#Aug24_CSS1_glm_day <- readRDS("Aug24_CSS1_glm_day.rds")

Aug24_CSS1_glm_day %>% merge(q5_CSSI_Aug24) %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = Sel)) + geom_line() + 
  geom_hline(aes(yintercept = q5),linetype = "dashed") +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  ggtitle("CSS I")
```



```{r}

rd_factor_cyclohex %>% filter(CHROM == "II")
rd_factor_cyclohex %>% filter(POS == 19019, CHROM == "II") %>%
  ggplot(aes(x = Bulk, y = Allele, size = SmoothCount)) + geom_jitter(width = 0.2, height = 0.2, alpha = 0.5) +
  theme_bw()

rd_factor_cyclohex %>% filter(POS == 19019, CHROM == "II") %>%
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>%
  mutate(logOdds = log(Wine/Oak)) %>%
  ggplot(aes(x = Bulk, y = logOdds, color = Parent)) + geom_jitter(width = 0.2, height = 0.2, alpha = 0.5) +
  theme_bw()

rd_factor_cyclohex %>% filter(POS == 19019, CHROM == "II") %>%
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>%
  mutate(logOdds = log(Wine/Oak)) %>%
  ggplot(aes(x = Parent, y = logOdds, color = Parent)) + geom_jitter(width = 0.2, height = 0.2, alpha = 0.5) +
  theme_bw()

rd_factor_cyclohex %>% filter(POS == 19019, CHROM == "II") -> test_cyclohex

#Bulk only effect
test_bulk <- glm(formula = Allele ~ Bulk, data = test_cyclohex, weights = SmoothCount, family = "binomial")
summary(test_bulk)

#Parent only effect
test_parent <- glm(formula = Allele ~ Parent, data = test_cyclohex, weights = SmoothCount, family = "binomial")
summary(test_parent)

contrasts(test_cyclohex$Bulk)
contrasts(test_cyclohex$Parent)

#Bulk x Parent  effect
test_both <- glm(formula = Allele ~ Bulk*Parent, data = test_cyclohex, weights = SmoothCount, family = "binomial")
results <- as.data.frame(summary(test_both)$coefficient)

results %>% ggplot(aes(x = 1, y = `z value`)) + geom_point()

results
Aug24_CSS1_glm_day %>% 
  filter(Selection == "Cyclohex", POS == 19019, CHROM == "II") %>%
  ggplot(aes(x = POS, y = abs(zscore), color = label)) + geom_point() + 
  ggtitle("CSS I") +
  theme(legend.position = "none") -> p1
                   
Aug24_CSS1_glm_day %>% 
  filter(Selection == "Cyclohex", CHROM == "II") %>%
  ggplot(aes(x = POS, y = abs(zscore), color = label)) + geom_point() + 
  geom_vline(xintercept = 19019) +
  ggtitle("CSS I") -> p2


cowplot::plot_grid(p1, p2, rel_widths = c(1,6))
```
# Trying to redo analysis but with added amts

If separating day and replicate are making the confidence worse on these, we may as well add them together to see if overall coverage would help. 

```{r}
rawdata_called_ex <- readRDS("CSSI_A24_rawdata_called_ex.rds")

rawdata_called_ex %>% mutate(Dataset = gsub("HWMMFDMXY_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Selection", "Parent", "Rep", "Day"), sep = "_") -> rawdata_renamed

rawdata_renamed %>% group_by(Selection, Parent, POS, CHROM, Allele) %>%
  summarize(AddedReads = sum(Reads)) -> rawdata_summed

rawdata_summed %>% filter(Selection != "AIG") -> rawdata_summed_saig

unique(rawdata_summed_saig$Selection)
unique(rawdata_summed_saig$Parent)
unique(rawdata_summed_saig$Allele)

rawdata_summed_saig %>% filter(Selection != "undetermined") -> rawdata_summed_saig

  
```
```{r}
rawdata_summed_saig %>% pivot_wider(names_from = Allele, values_from = AddedReads) %>%
  mutate(logOdds = log(Wine/Oak)) %>%
  ggplot(aes(x = POS, y = logOdds, color = Selection)) +
  geom_point() +
  facet_grid(Parent ~ CHROM, space = "free", scales = "free")
```

```{r}

saveRDS(rawdata_summed_saig, file = "CSSI_rawdata_summed_saig.rds")

rawdata_summed_saig %>% ungroup() %>% 
  group_by(Parent, Selection, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(AddedReads, n = 200, FUN = cybr_weightedgauss, align = "center")) %>% mutate(smooth = 200) -> rawdata_summed_g200

rawdata_summed_g200 %>% mutate(SmoothCount = ceiling(SmoothCount)) %>% na.omit() -> rawdata_summed_g200

saveRDS(rawdata_summed_g200, file = "CSSI_rawdata_summed_g200.rds")

rawdata_summed_g200 %>% pivot_wider(names_from = Allele, values_from = SmoothCount) %>%
  mutate(logOdds = log(Wine/Oak)) %>%
  ggplot(aes(x = POS, y = logOdds, color = Parent)) +
  geom_line() +
  facet_grid(Selection ~ CHROM, space = "free", scales = "free")
```
```{r}
rawdata_summed_g200 %>% filter(Selection %in% c("D", "C")) %>% mutate(Bulk = gsub("D", "A", Selection))  %>% na.omit() %>%
  mutate_if(is.character, as.factor) %>% mutate(Sel = "C") -> rd_factor_cyclohex

rawdata_summed_g200 %>% filter(Selection %in% c("D", "E")) %>% mutate(Bulk = gsub("D", "A", Selection))  %>% na.omit() %>%
  mutate_if(is.character, as.factor) %>% mutate(Sel = "E") -> rd_factor_ethanol

rawdata_summed_g200 %>% filter(Selection %in% c("D", "H")) %>% mutate(Bulk = gsub("D", "A", Selection))  %>% na.omit() %>%
  mutate_if(is.character, as.factor) %>% mutate(Sel = "H") -> rd_factor_H2O2


contrasts(rd_factor_cyclohex$Parent) <- matrix(c(0.5, -0.5))
contrasts(rd_factor_ethanol$Parent) <- matrix(c(0.5, -0.5))
contrasts(rd_factor_H2O2$Parent) <- matrix(c(0.5, -0.5))

contrasts(rd_factor_cyclohex$Bulk) <- matrix(c(0, 1))
contrasts(rd_factor_ethanol$Bulk) <- matrix(c(0, 1))
contrasts(rd_factor_H2O2$Bulk) <- matrix(c(0, 1))
```

```{r}
rd_factor_cyclohex %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #filter(POS == 5627) %>%
  group_by(CHROM, POS, Sel) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background"),
                                                                  label = c("cept","Bulk", "Background", "Interaction")) -> glm_summed_cyclohex

rd_factor_ethanol %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #filter(POS == 5627) %>%
  group_by(CHROM, POS, Sel) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background"),
                                                                  label = c("cept","Bulk", "Background", "Interaction")) -> glm_summed_ethanol

rd_factor_H2O2 %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  #filter(POS == 5627) %>%
  group_by(CHROM, POS, Sel) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Parent,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background"),
                                                                  label = c("cept","Bulk", "Background", "Interaction")) -> glm_summed_H2O2


rbind(glm_summed_cyclohex,glm_summed_ethanol,glm_summed_H2O2) -> summed_glm_g200

saveRDS(summed_glm_g200, file = "11-12-24_summed_glm_g200.rds")

```

```{r}
summed_glm_g200 %>% filter(label != "cept") %>%
  ggplot(aes(x = POS, y = zscore, color = Sel)) +
  geom_line() +
  facet_grid(label ~ CHROM, space = "free", scales = "free")

rawdata_summed_g200 %>% filter(Selection == "D") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) %>%
  mutate(logOdds = log(Wine/Oak)) -> DiluteLogOdds

summed_glm_g200 %>% filter(label == "Background") %>% merge(DiluteLogOdds) %>%
  ggplot(aes(x = POS)) +
  geom_line(aes(y = zscore), color = "black") +
  geom_line(aes(y = logOdds, color = Parent)) +
  facet_grid(~CHROM, space = "free", scales = "free") +
  ggtitle("Unselected by Parent vs Background z-score")

```

```{r, fig.width=10, fig.height=5}
################################################################################
rawdata_summed_g200 %>% filter(Selection %in% c("C", "D")) %>% pivot_wider(names_from = Allele, values_from = SmoothCount) %>%
  mutate(logOdds = log(Wine/Oak)) -> CycLogOdds

summed_glm_g200 %>% filter(label %in% c("Interaction", "Background"), Sel == "C") %>% filter(CHROM != "III") %>%
  merge(CycLogOdds) %>%
  ggplot(aes(x = POS)) +
  geom_line(aes(y = zscore, color = label), alpha = 0.5) +
  geom_line(aes(y = logOdds, color = Parent), size = 1) +
  facet_grid(Selection~CHROM, space = "free", scales = "free") +
  ggtitle("Cyclohex by Parent vs Interaction z-score") +
  theme(legend.position = "bottom") +
  scale_color_manual(values = c("black", "turquoise", "darkorange", "maroon"))

summed_glm_g200 %>% filter(label %in% c("Interaction", "Background"), Sel == "C") %>% filter(CHROM == "IV") %>%
  merge(CycLogOdds) %>%
  ggplot(aes(x = POS)) +
  geom_line(aes(y = zscore, color = label), alpha = 0.5) +
  geom_line(aes(y = logOdds, color = Parent), size = 1) +
  facet_grid(Selection~CHROM, space = "free", scales = "free") +
  ggtitle("Cyclohex by Parent vs Interaction z-score") +
  theme(legend.position = "bottom") +
  scale_color_manual(values = c("black", "turquoise", "darkorange", "maroon"))
```

```{r}
#Do the days differ in parent effect?

rawdata_renamed %>% filter(Selection == "D") %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>%
  mutate(logOdds = log(Wine/Oak)) -> daycheck

daycheck %>% filter(CHROM != "I", CHROM != "M") %>% 
  ggplot(aes(x = POS, y = abs(logOdds), color = Parent)) +
  geom_point(size = 0.5, alpha = 0.5) +
  facet_grid(Day ~ CHROM, scales = "free", space = "free")

daycheck %>% filter(CHROM != "I", CHROM != "M", CHROM != "III") %>%
  select(CHROM, POS, Parent, Rep, Day, logOdds) %>%
  pivot_wider(names_from = Parent, values_from = logOdds) -> testDiff

testDiff %>% mutate(Diff = abs(W - O)) %>%
  ggplot(aes(x = W, y = O, color = Day)) + 
  geom_point(alpha = 0.3, size = 0.5) +
  geom_smooth(method = "lm") +
  geom_abline(slope = 1, intercept = 0, color = "black")

testDiff %>% mutate(Diff = abs(W - O)) %>%
  ggplot(aes(x = Diff, y = Day, color = Day)) + geom_boxplot()
```

