---
title: "MergedVCFs_CuSO4_Gaussian_June24"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))

# ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
#                                              levels = as.character(as.roman(1:16))),
#                               start = rep(1, 16),
#                               end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
#   pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
#   mutate(Summary = NA, Label = NA) %>% select(-delete)
# 
# ChromosomeScale$Start <- 30000
# ChromosomeScale$End <- ChromosomeScale$POS - 30000
# 
# ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
#                                              levels = as.character(as.roman(1:16))),
#                               start = rep(1, 16),
#                               end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
#   pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
#   mutate(summary = 0, label = "Bulk") %>% select(-delete)
# 
# # Loci to look at specifically
# 
# myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
#                      POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
#                      name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
#                      cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))
# 
# 
# 
# ## Functions for GLM all and Permutation all
# 
# glm_cb2_all <- function(..., W, formula, numgroups = FALSE, outputlength = 8) {
#   data <- list(...)
#   
#   #Ensure that there is a formula and W parameter
#   if (is.null(W) || is.null(formula)) {
#     stop("Weights (W) and formula must be provided")
#   }
#   #Set formula
#   glm_formula <- as.formula(formula)
#   #Ensure that formula works for the data provided
#   if (!all(names(data) %in% all.vars(glm_formula))) {
#     stop("One or more variables in the formula are not provided as arguments")
#   }
#   
#   #########################
#   for(i in all.vars(glm_formula)){
#     if(length(unique(as.data.frame(data)[,i])) < 2){
#       output <- rep(NA, outputlength)
#       #print("Not enough levels within groups")
#       
#       return(output)
#     }
#   }
#   
#   glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
#   
#   output <- summary(glm_fit)$coefficients[c(((length(summary(glm_fit)$coefficients)*0)+1):
#                                             ((length(summary(glm_fit)$coefficients)*0.25)),
#                                             ((length(summary(glm_fit)$coefficients)*0.5)+1):
#                                             ((length(summary(glm_fit)$coefficients)*0.75)))]
# 
#   
#   if(length(output) == outputlength){
#     return(output)
#   }else{
#     return(rep(NA, outputlength))
#   }
#   
# } 
# 
# 
# ################################################################################
# # No replicates, fix Parent
# cybrPermute_cb2_all <- function(dataset, 
#                                 R = 0, perp = 1, outlength = 4, 
#                                 glmform = "Allele ~ Bulk * Parent", 
#                                 inputlabels = c("Intercept", "Bulk", "Parent", "Interaction")){
#   start.time <- Sys.time()
#     
#   print("Make sure that dilute bulk is labeled aD")
#   print(paste("Your labels are:", inputlabels))
# 
#     if(R > 0){ #INCLUDE REPLICATES
#         dataset %>% 
#         distinct() %>% ungroup() %>%
#         group_by(CHROM, POS, Allele, Bulk, Rep, Parent) %>% 
#         summarize(culprits = length((SmoothCount))) %>% 
#         merge(dataset) %>% 
#         filter(culprits == 1) %>% 
#         ungroup() %>%
#         distinct() %>% #THIS IS IMPORTANT
#         pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest    
#     }else{ #DO NOT INCLUDE REPLICATES
#         dataset %>% 
#         distinct() %>% ungroup() %>%
#         group_by(CHROM, POS, Allele, Bulk, Parent) %>% 
#         summarize(culprits = length((SmoothCount))) %>% 
#         merge(dataset) %>% 
#         filter(culprits == perp) %>% 
#         ungroup() %>%
#         distinct() %>% #THIS IS IMPORTANT
#         pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
#     }
#   
#     #PERMUTE TWICE
#   if(R > 0){
#     newnewtest %>% filter(Bulk == "aD",
#                           CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
#       select(CHROM, POS, Rep, 
#              Parent, Oak, Wine) %>% 
#       group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
#       summarize(Parent = Parent, Rep = Rep, 
#                 Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
#       mutate(Bulk = "A") -> shuffled_DiluteA2
#     
#     newnewtest %>% filter(Bulk == "aD",
#                           CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
#       select(CHROM, POS, 
#              Rep, Parent, Oak, Wine) %>% 
#       group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
#       summarize(Parent = Parent, Rep = Rep,
#                 Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
#       mutate(Bulk = "B") -> shuffled_DiluteB2
#     
#   }else{
#     newnewtest %>% filter(Bulk == "aD",
#                           CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
#       select(CHROM, POS, Parent, Oak, Wine) %>% 
#       group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
#       summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
#       mutate(Bulk = "A") -> shuffled_DiluteA2
#     
#     newnewtest %>% filter(Bulk == "aD",
#                           CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
#       select(CHROM, POS, Parent, Oak, Wine) %>% 
#       group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
#       summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
#       mutate(Bulk = "B") -> shuffled_DiluteB2
#     
#   }
#     
#     rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
#     
#     #RUN THE GLM
#     if(R > 0) {
#       shuffletoglm2 %>% na.omit() %>%
#       group_by(Loc) %>%
#       mutate_if(is.character, as.factor) %>%
#         summarize(Summary = glm_cb2_all(Allele = Allele,
#                                  Bulk = Bulk,
#                                  Parent = Parent,
#                                  Rep = Rep,
#                                  W = SmoothCount,
#                                  formula = glmform,
#                                 outputlength = length(inputlabels)*2),
#                 Factor = rep(inputlabels, 2),
#                 d = c(rep("Effect", length(inputlabels)),
#                       rep("Z", length(inputlabels)))) -> glmresult
#     }else{
#       shuffletoglm2 %>% na.omit() %>%
#       #Original Script
#       group_by(Loc) %>%
#       mutate_if(is.character, as.factor) %>%
#         summarize(Summary = glm_cb2_all(Allele = Allele,
#                                  Bulk = Bulk,
#                                  Parent = Parent,
#                                  W = SmoothCount,
#                                  formula = glmform,
#                                 outputlength = length(inputlabels)*2),
#                 Factor = rep(inputlabels, 2),
#                 d = c(rep("Effect", length(inputlabels)),
#                       rep("Z", length(inputlabels)))) -> glmresult
#     }
#     
#     
#     end.time = Sys.time()
#     print(end.time - start.time)
#     return(glmresult)
# }
# 
# cybrInputGATKTable2 <- function(rawData, yeast = TRUE){
# 
#   require(dplyr)
#   require(doParallel)
#   require(foreach)
# 
#   HNGLCDRXY <- read.table(rawData, header = TRUE)
# 
#   #Identify the unique values besides AD/DP/GQ/PL
#   gsub(".AD", "",
#        gsub(".GQ", "",
#             gsub(".DP","",
#                  gsub(".PL","",
#                       colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
#   #i <- Samples[1]
# 
#   resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
#     mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
#     AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
#     GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
#     DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
#     PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
#     cbind(mydf, AD , GQ , DP, PL) -> mydftotal
#     colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")
# 
#     mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>%
#       separate(PL, c('PL.REF','PL.ALT'), extra='drop') %>%
#       #Added 10/18/23:
#       select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf
# 
#     mycdf %>% filter(grepl(",", ALT)) %>% 
#       separate(ALT, c("A1", "A2"), extra = 'merge') %>%
#       separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
#       separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
#       
#       pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
#       pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
#       pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
#       mutate(NumAlt = gsub("A", "", NumAlt),
#              NumADAlt = gsub("AD", "", NumADAlt),
#              NumPL = gsub("P", "", NumPL)) %>%
#       filter(NumAlt == NumPL,
#              NumPL == NumADAlt) %>%
#       select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
#     
#     doublecdf %>% filter(grepl(",", ALT)) %>%
#       separate(ALT, c("A1", "A2"), extra = 'merge') %>%
#       separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
#       separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
#       
#       pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
#       pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
#       pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
#       mutate(NumAlt = gsub("A", "", NumAlt),
#              NumADAlt = gsub("AD", "", NumADAlt),
#              NumPL = gsub("P", "", NumPL)) %>%
#       filter(NumAlt == NumPL,
#              NumPL == NumADAlt) %>%
#       select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
#     
#     rbind(mycdf, doublecdf, triplecdf) -> newcdf
#     
#     newcdf
#   }
# 
#   if(yeast == TRUE){
#     ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
#                                          "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
#                          CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
#                                    "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
#                                    "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
#                                    "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))
# 
#     resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
#   }else{
#     results <- resultscdf
#   }
#   return(results)
# 
# }

```

# Pre-processing

## Load in table and match to oak or wine parents

```{r}
# setwd("C:/Users/Cassandra/Documents/GitHub/Sequencing/Analysis/")
# dir()

parentSNPids <- cybrConvertParentalAlleles(Truncate = FALSE)
parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) -> pSNPs

merge(pSNPs, parentSNPids) %>% filter(Unique == 1) %>% select(-Unique) -> parentSNPids

unique(pSNPs$Unique)

```

## Exclude duplicates 

```{r, message = FALSE, warning=FALSE, fig.width=12, fig.height=4}
myfile <- "C:/Users/Cassandra/Documents/Data_NYU/Sequencing/HWTCNDMXY.SortedCat.vcf.output.table"

#GET RID OF EXCLUDED ONES
read.table(myfile, header = TRUE) -> testrawdata
testrawdata %>% select(CHROM, POS, ALT) %>% separate(ALT, into = c("First", "Second"), sep = ",") %>% na.omit() -> exclude

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

exclude %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> exclude

exclude %>% mutate(locus = paste(CHROM, POS)) -> exclKEY

cybrInputGATKTable2(myfile) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> rawdata

saveRDS(rawdata, file = "CSSVIII_A24_rawdataG.rds")

rawdata %>% filter(POS == 801680, CHROM == "IV")

################################################################################
rawdata %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% mutate(locus = paste(CHROM, POS)) %>% filter(locus %in% exclKEY$locus == FALSE) %>% select(-locus) -> rawdata_called_ex


################################################################################

saveRDS(rawdata_called_ex, file = "CSSVIII_A24_rawdata_called_ex.rds")
```

```{r}
#Testing what this data actually looks like
rawdata_called_ex %>% pivot_wider(names_from = Allele, values_from = Reads) %>% unnest() %>% 
  mutate(logodds = log(Wine/Oak)) -> test_rawdatacalledex

```

Narrowing segments that look weird specifically (segregation issues)

```{r}
test_rawdatacalledex %>% filter(CHROM == "XVI") %>% filter(POS < 150000, POS > 120000) %>%
  ggplot(aes(x = POS, y = Oak, color = grepl("_W_", Dataset))) + geom_point(alpha = 0.3) +
  theme_minimal()+ ggtitle("Chr Oak XVI Coverage Zoomed In")

test_rawdatacalledex %>% filter(CHROM == "XI") %>% filter(POS > 3.05e5, POS < 3.2e5) %>%
  ggplot(aes(x = POS, y = Wine, color = grepl("_W_", Dataset))) + geom_point(alpha = 0.3) +
  theme_minimal()+ ggtitle("Chr Oak XI Coverage Zoomed In")
```

### Filter out weird coverage and stuff

```{r}
exclusion_POS <- data.frame(CHROM = c("XI", "XVI"),
                            Start = c(3.06e5, 128000),
                            End = c(3.2e5, 142000))

rawdata_called_ex %>% filter((CHROM == "XI" & POS < 320000 & POS > 306000)== FALSE,
                             (CHROM == "XVI" & POS < 142000 & POS > 128000)== FALSE) -> rawdata_called_exx

saveRDS(rawdata_called_exx, file = "CSSVIII_A24_rawdata_called_exx.rds")

```

### Separate out what could be smoothed here

```{r}

# cybr_weightedgauss <- function(myx){
#   myy <- dnorm(1:length(myx), mean = length(myx)/2, sd = 10)
#   return(weighted.mean(x = myx, y = myy, na.rm = TRUE))
# } 

# plot(dnorm(1:200, mean = 200/2, sd = 10), main = "SD 10")
# plot(dnorm(1:200, mean = 200/2, sd = 20), main = "SD 20")
# plot(dnorm(1:200, mean = 200/2, sd = 100), main = "SD 100") #choosing this one for including all of the points
# plot(dnorm(1:200, mean = 200/2, sd = 1000), main = "SD 1000")

rawdata_called_exx <- readRDS("CSSVIII_A24_rawdata_called_exx.rds")

unique(rawdata_called_exx$Dataset)

rawdata_called_exx %>% ungroup() %>% 
  group_by(Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) %>% mutate(smooth = 200) -> rawdata_smoothed_G200x

rawdata_smoothed_G200x %>% mutate(SmoothCount = ceiling(SmoothCount)) %>% na.omit() -> rawdata_smoothed_G200x

saveRDS(rawdata_smoothed_G200x, file = "CSSVIII_A24_rawdata_smoothed_G200x.rds")

```

```{r}
#Checking to make sure this actually works
rawdata_smoothed_G200 %>% filter(CHROM == "XVI") %>% filter(POS < 150000, POS > 120000) %>%
  ggplot(aes(x = POS, y = SmoothCount, color = grepl("_W_", Dataset))) + geom_point(alpha = 0.3) +
  theme_minimal()+ ggtitle("Chr Oak XVI Coverage Zoomed In")
```

## Separate into bulks, make ready for glms

```{r}

# rawdata_called_ex <- readRDS("CSSVIII_A24_rawdata_smoothed_G200.rds")
#rawdata_smoothed_G200 <- readRDS("CSSVIII_A24_rawdata_smoothed_G200.rds")
rawdata_smoothed_G200 <- readRDS("CSSVIII_A24_rawdata_smoothed_G200x.rds")

################################################################################
#FIXING MIXUP

rawdata_smoothed_G200 %>% mutate(Dataset = gsub("E_O_B_x", "blank", Dataset),
                                 Dataset = gsub("H_O_B_x", "E_O_B_x", Dataset)) %>%
  mutate(Dataset = gsub("blank", "H_O_B_x", Dataset)) -> rawdata_smoothed_G200_corrected
                                   
  
################################################################################
# rawdata_called_ex %>% filter(grepl("AIG", Dataset) == FALSE) %>%
#   mutate(Dataset = gsub("HWMMFDMXY_n01_", "", Dataset)) %>%
#   mutate(Dataset = gsub(".fastq", "", Dataset)) %>% 
#   separate(Dataset, into = c("Bulk","Parent","Rep","Day"), sep = "_") -> rd_named

rawdata_smoothed_G200_corrected %>% filter(grepl("AIG", Dataset) == FALSE) %>%
  mutate(Dataset = gsub("HWTCNDMXY_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>% 
  separate(Dataset, into = c("Bulk","Parent","Rep","Day"), sep = "_") -> smoothdata_named

smoothdata_named %>% mutate(Day = toupper(Day)) -> smoothdata_named

# 
# 
# unique(rawdata_smoothed_G200$Dataset) #Found out I was missing all of these whoops
# 
# unique(smoothdata_named$Bulk)
# unique(smoothdata_named$Parent)
# unique(smoothdata_named$Rep)
# 
# table(smoothdata_named$Bulk,smoothdata_named$Day)
# 
# table(smoothdata_named$Bulk)
```


## Change into factors

```{r}

smoothdata_named %>% filter(Bulk %in% c("D", "C")) %>% mutate(Bulk = gsub("D", "A", Bulk))  %>% na.omit() %>%
  mutate_if(is.character, as.factor) %>% mutate(Sel = "C") -> rd_factor_cyclohex

smoothdata_named %>% filter(Bulk %in% c("D", "E")) %>% mutate(Bulk = gsub("D", "A", Bulk))  %>% na.omit() %>%
  mutate_if(is.character, as.factor) %>% mutate(Sel = "E") -> rd_factor_ethanol

smoothdata_named %>% filter(Bulk %in% c("D", "H")) %>% mutate(Bulk = gsub("D", "A", Bulk))  %>% na.omit() %>%
  mutate_if(is.character, as.factor) %>% mutate(Sel = "H") -> rd_factor_H2O2


contrasts(rd_factor_cyclohex$Parent) <- matrix(c(0.5, -0.5))
contrasts(rd_factor_ethanol$Parent) <- matrix(c(0.5, -0.5))
contrasts(rd_factor_H2O2$Parent) <- matrix(c(0.5, -0.5))

contrasts(rd_factor_cyclohex$Bulk) <- matrix(c(0, 1))
contrasts(rd_factor_ethanol$Bulk) <- matrix(c(0, 1))
contrasts(rd_factor_H2O2$Bulk) <- matrix(c(0, 1))

saveRDS(rd_factor_cyclohex, "CSSVIII_A24_rd_factor_cyclohex.rds")
saveRDS(rd_factor_ethanol, "CSSVIII_A24_rd_factor_ethanol.rds")
saveRDS(rd_factor_H2O2, "CSSVIII_A24_rd_factor_H2O2.rds")

```

################################################################################

# Running GLMs

```{r}
rd_factor_cyclohex <- readRDS("CSSVIII_A24_rd_factor_cyclohex.rds") 
rd_factor_ethanol <- readRDS("CSSVIII_A24_rd_factor_ethanol.rds")
rd_factor_H2O2 <- readRDS("CSSVIII_A24_rd_factor_H2O2.rds")

rd_factor <- rbind(rd_factor_cyclohex,rd_factor_ethanol,rd_factor_H2O2)
```

## Checking raw data for sanity

```{r}
rd_factor %>% pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(logodds = log(Wine/Oak)) -> testing_rdfactor

testing_rdfactor %>% na.omit()

# testing_rdfactor %>% filter(CHROM %in% c("M", "VIII", "I") == FALSE) %>%
#   mutate(logodds = logodds + 1) %>%
#   ggplot(aes(x = POS, y = logodds, color = paste(Parent, Day, Rep))) + geom_line() +
#   facet_grid(Bulk~CHROM, scales = "free", space = "free") +
#   theme(legend.position = "bottom")


testing_rdfactor %>% filter(CHROM  == "XVI") %>%
  ggplot(aes(x = POS, y = abs(logodds), color = paste(Day, Rep))) + geom_line() +
  facet_grid(Bulk~Parent, scales = "free", space = "free") +
  theme(legend.position = "bottom")+ ggtitle("Chr XVI")

testing_rdfactor %>% filter(CHROM  == "XV") %>%
  ggplot(aes(x = POS, y = abs(logodds), color = paste(Day, Rep))) + geom_line() +
  facet_grid(Bulk~Parent, scales = "free", space = "free") +
  theme(legend.position = "bottom") + ggtitle("Chr XV")

testing_rdfactor %>% filter(CHROM  == "VII") %>%
  ggplot(aes(x = POS, y = abs(logodds), color = paste(Day, Rep))) + geom_line() +
  facet_grid(Bulk~Parent, scales = "free", space = "free") +
  theme(legend.position = "bottom")+ ggtitle("Chr VII")

############################################
testing_rdfactor %>% filter(CHROM %in% c("M", "VIII", "I", "III") == FALSE, Bulk == "A") %>%
  ggplot(aes(x = POS, y = abs(logodds), color = Parent)) + geom_line() +
  facet_grid(paste(Day, Rep)~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom")

```

### Glmer for when we have more time would add to this

```{r}
rd_factor %>% filter(CHROM != "M", CHROM != "VIII") %>% distinct() %>% 
  group_by(CHROM, POS, Sel) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Parent,
                                                            #Rep = Rep,
                                                            Day = Day,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 5,
                                                            formula = "Allele ~ Bulk*Background + Day "),
                                                                  label = c("cept","Bulk", "Background", "Interaction", "Day"))  -> Aug24_CSS8_glm_day

saveRDS(Aug24_CSS8_glm_day, file = "Aug24_CSS8_glm_day.rds")

```

### Permutations using glmer

```{r}
rd_factor %>% filter(Bulk == "A",
                     CHROM %in% c("VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  #group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Parent,
            Day = Day, Rep = Rep) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

rd_factor %>% filter(Bulk == "A",
                     CHROM %in% c("VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  #group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Parent,
            Day = Day, Rep = Rep) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Unselected") -> rd_shuffled_unselected

Perm8 <- rbind(rd_shuffled_unselected, rd_shuffled_selected)
Perm8 %>% ungroup() %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") %>% unnest() %>%
  mutate(POS = as.numeric(POS)) %>%
  mutate_if(is.character, as.factor) -> Perm8contrasts

contrasts(Perm8contrasts$Background) <- matrix(c(-0.5, 0.5))
contrasts(Perm8contrasts$Bulk) <- matrix(c(1,0))


Perm8 %>% mutate(logodds = log(Wine/Oak)) %>%
  ggplot(aes(x = logodds, color = Background)) + geom_density() +
  geom_vline(xintercept = 0)


Perm8contrasts %>% filter(CHROM != "M", CHROM != "I") %>% distinct() %>% 
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            #Rep = Rep,
                                                            Day = Day,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 5,
                                                            formula = "Allele ~ Bulk*Background + Day "),
                                                                  label = c("cept","Bulk", "Background", "Interaction", "Day"))  -> Aug24_CSS8_glm_PERM_day

saveRDS(Aug24_CSS8_glm_PERM_day, file = "Aug24_CSS8_glm_PERM_day.rds")

Aug24_CSS8_glm_PERM_day %>% ungroup() %>% group_by(label) %>% summarize(q5 = quantile(abs(zscore), 0.95)) -> q5_CSSVIII_Aug24

q5_CSSVIII_Aug24
```


# Load previous glms

```{r, eval = FALSE}

#With 2 replicates included
Fluc_CSS1_glm <- readRDS("C:/Users/Cassandra/Documents/R_FileTransfer/StandardGLM/HGVMVDRX2_CSS1_glm_alld.rds")

#With no replicates
Fluc_CSS1_glm <- readRDS("C:/Users/Cassandra/Documents/R_FileTransfer/StandardGLM/HGVMVDRX2norep_CSS1_glm_all.rds") %>% 
  mutate(zscore = Summary, label = Factor, Selection = "Fluconazole") %>%
  select(-Summary, -Factor) %>%
  na.omit() 

CuSO4_CSS1_glmer_1byRep <- readRDS("CuSO4_CSS1_glmer_1byRepG.rds")
Aug24_CSS1_glm_H2O2 <- readRDS("Aug24_CSS1_glm_H2O2.rds")
Aug24_CSS1_glm_cyclohex <- readRDS("Aug24_CSS1_glm_cyclohex.rds")
Aug24_CSS1_glm_ethanol <- readRDS("Aug24_CSS1_glm_ethanol.rds")

Aug24_CSS1_glm_H2O2 %>% mutate(Selection = "H2O2") -> temp1
Aug24_CSS1_glm_ethanol %>% mutate(Selection = "EtOH") -> temp2
Aug24_CSS1_glm_cyclohex %>% mutate(Selection = "Cycloheximide") %>%
  rbind(temp1) %>%
  rbind(temp2) %>%
  mutate(CSS = 1) -> CSS1_Aug2024_glm

Aug24_CSS8_glm_H2O2 <- readRDS("Aug24_CSS8_glm_H2O2.rds")
Aug24_CSS8_glm_ethanol <- readRDS("Aug24_CSS8_glm_ethanol.rds")
Aug24_CSS8_glm_cyclohex <- readRDS("Aug24_CSS8_glm_cyclohex.rds")

Aug24_CSS8_glm_H2O2 %>% mutate(Selection = "H2O2") -> temp1
Aug24_CSS8_glm_ethanol %>% mutate(Selection = "EtOH") -> temp2
Aug24_CSS8_glm_cyclohex %>% mutate(Selection = "Cycloheximide") %>%
  rbind(temp1) %>%
  rbind(temp2)%>%
  mutate(CSS = 8) -> CSSVIII_Aug2024_glm

rm(temp1)
rm(temp2)

```

# Plot New Data
```{r, fig.width=12, fig.height=5}
Aug24_CSS8_glm_day <- readRDS("Aug24_CSS8_glm_day.rds")

Aug24_CSS8_glm_day %>% #merge(q5_CSSVIII_Aug24) %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = Sel)) + geom_line() + 
  #geom_hline(aes(yintercept = q5),linetype = "dashed") +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  ggtitle("CSS VIII")
```



