---
title: "CUP1 Region Realignment"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))


```

# Pre-processing

## Load in table and match to oak or wine parents

```{r}
# setwd("C:/Users/Cassandra/Documents/GitHub/Sequencing/Analysis/")
# dir()

parentSNPids <- cybrConvertParentalAlleles(Truncate = FALSE)
parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) -> pSNPs

merge(pSNPs, parentSNPids) %>% filter(Unique == 1) %>% select(-Unique) -> parentSNPids

unique(pSNPs$Unique)

```

## Exclude duplicates 

```{r, message = FALSE, warning=FALSE, fig.width=12, fig.height=4}
myfile <- "C:/Users/Cassandra/Documents/Data_NYU/C8_HNGLVDRXY.SortedCat.vcf.output.table"

#GET RID OF EXCLUDED ONES
read.table(myfile, header = TRUE) -> testrawdata
testrawdata %>% select(CHROM, POS, ALT) %>% separate(ALT, into = c("First", "Second"), sep = ",") %>% na.omit() -> exclude

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

exclude %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> exclude

exclude %>% mutate(locus = paste(CHROM, POS)) -> exclKEY

cybrInputGATKTable2(myfile) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> rawdata

saveRDS(rawdata, file = "Chr8_rawdataG.rds")

################################################################################
rawdata %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% mutate(locus = paste(CHROM, POS)) %>% filter(locus %in% exclKEY$locus == FALSE) %>% select(-locus) -> rawdata_called_ex


################################################################################

saveRDS(rawdata_called_ex, file = "Chr8_rawdata_called_ex.rds")
```

Apparently there are 1741 Ns at 212408, but that is not the number of blanked out sequences that we actually have here...

```{r}
rawdata %>% ggplot(aes(x = POS, y = Coverage, color = Dataset)) + geom_point(alpha = 0.3) +
  geom_vline(xintercept = c(212535,212720)) +
  theme(legend.position = "bottom")


rawdata %>% mutate(logodds = abs(log(as.numeric(AD.ALT)/as.numeric(AD.REF)))) %>%
  ggplot(aes(x = POS, y = logodds, color = Dataset)) + geom_point(alpha = 0.3) +
  geom_vline(xintercept = c(212535,212720, 776803)) +
  theme(legend.position = "bottom")

rawdata %>% summarize(uniquePOS = unique(POS)) %>% arrange((uniquePOS))

212720 - 5390
430471 - 212720
#430471
```


### Separate out what could be smoothed here

```{r}

# cybr_weightedgauss <- function(myx){
#   myy <- dnorm(1:length(myx), mean = length(myx)/2, sd = 10)
#   return(weighted.mean(x = myx, y = myy, na.rm = TRUE))
# } 

# plot(dnorm(1:200, mean = 200/2, sd = 10), main = "SD 10")
# plot(dnorm(1:200, mean = 200/2, sd = 20), main = "SD 20")
# plot(dnorm(1:200, mean = 200/2, sd = 100), main = "SD 100") #choosing this one for including all of the points
# plot(dnorm(1:200, mean = 200/2, sd = 1000), main = "SD 1000")

rawdata_called_ex <- readRDS("Chr8_rawdata_called_ex.rds")
# 
# unique(rawdata_called_exx$Dataset)

rawdata_called_ex %>% ungroup() %>% 
  group_by(Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) %>% mutate(smooth = 200) -> rawdata_smoothed_G200x

rawdata_smoothed_G200x %>% mutate(SmoothCount = ceiling(SmoothCount)) %>% na.omit() -> rawdata_smoothed_G200x

saveRDS(rawdata_smoothed_G200x, file = "Chr8_rawdata_smoothed_G200x.rds")

```


## Separate into bulks, make ready for glms

```{r}

rawdata_smoothed_G200 <- readRDS("Chr8_rawdata_smoothed_G200x.rds")


rawdata_smoothed_G200 %>% filter(grepl("AIG", Dataset) == FALSE) %>%
  mutate(Dataset = gsub("HWMMFDMXY_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>% 
  separate(Dataset, into = c("Bulk","Parent","Rep","Day"), sep = "_") -> smoothdata_named

smoothdata_named %>% mutate(Day = toupper(Day)) -> smoothdata_named

 
# unique(smoothdata_named$Bulk)
# unique(smoothdata_named$Parent)
# unique(smoothdata_named$Rep)
# 
# table(smoothdata_named$Bulk,smoothdata_named$Day)
# 
# table(smoothdata_named$Bulk)
```


## Change into factors

```{r}


```




