---
title: "MergedVCFs_CuSO4_Gaussian_June24"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))



## Functions for GLM all and Permutation all

glm_cb2_all <- function(..., W, formula, numgroups = FALSE, outputlength = 8) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  output <- summary(glm_fit)$coefficients[c(((length(summary(glm_fit)$coefficients)*0)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.25)),
                                            ((length(summary(glm_fit)$coefficients)*0.5)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.75)))]

  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
} 


################################################################################
# No replicates, fix Parent
cybrPermute_cb2_all <- function(dataset, 
                                R = 0, perp = 1, outlength = 4, 
                                glmform = "Allele ~ Bulk * Parent", 
                                inputlabels = c("Intercept", "Bulk", "Parent", "Interaction")){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled aD")
  print(paste("Your labels are:", inputlabels))

    if(R > 0){ #INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Rep, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == 1) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest    
    }else{ #DO NOT INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == perp) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    }
  
    #PERMUTE TWICE
  if(R > 0){
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Rep, 
             Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep, 
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, 
             Rep, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep,
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }else{
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #RUN THE GLM
    if(R > 0) {
      shuffletoglm2 %>% na.omit() %>%
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }else{
      shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }
    
    
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}

cybrInputGATKTable2 <- function(rawData, yeast = TRUE){

  require(dplyr)
  require(doParallel)
  require(foreach)

  HNGLCDRXY <- read.table(rawData, header = TRUE)

  #Identify the unique values besides AD/DP/GQ/PL
  gsub(".AD", "",
       gsub(".GQ", "",
            gsub(".DP","",
                 gsub(".PL","",
                      colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
  #i <- Samples[1]

  resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
    mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
    AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
    GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
    DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
    PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
    cbind(mydf, AD , GQ , DP, PL) -> mydftotal
    colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")

    mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>%
      separate(PL, c('PL.REF','PL.ALT'), extra='drop') %>%
      #Added 10/18/23:
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf

    mycdf %>% filter(grepl(",", ALT)) %>% 
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
    
    doublecdf %>% filter(grepl(",", ALT)) %>%
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
    
    rbind(mycdf, doublecdf, triplecdf) -> newcdf
    
    newcdf
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
  }else{
    results <- resultscdf
  }
  return(results)

}

################################################################################
library(zoo)

gaussian_kernel <- function(x, sigma) {
  exp(-(x^2) / (2 * sigma^2)) / (sqrt(2 * pi) * sigma)
}

window_size <- 300  # Size of the smoothing window
sigma <- 50         # Standard deviation for the Gaussian kernel

#Example of how to use this:
# rawdata_called %>%
#   group_by(Pool, Dataset, CHROM, Allele) %>%
#   #filter(CHROM == "I") %>%
#   arrange(POS) %>%
#   reframe(POS = POS, SmoothCount = rollapply(Reads, width = window_size, align = "center", 
#                                 FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
#                                 fill = NA)) -> rawdata_gaussian

plot(gaussian_kernel(c(100, 100, 100, NA, 10:100), 30))

z <- 1:10
sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma))
```

# Pre-processing

## Load in table and match to oak or wine parents

```{r}
setwd("C:/Users/Cassandra/Documents/GitHub/Sequencing/Analysis/")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)
parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) -> pSNPs

```

## Exclude duplicates 

```{r, message = FALSE, warning=FALSE}
#testfile <- "C:/Users/cassa/Downloads/AllCuSO4.REF_NC_001143.9.SortedCat.vcf.output.table"


myfile <- "C:/Users/Cassandra/Documents/Data_NYU/Sequencing/AllCuSO4.REF_.SortedCat.vcf.output.table"

#GET RID OF EXCLUDED ONES
read.table(myfile, header = TRUE) -> testrawdata
testrawdata %>% select(CHROM, POS, ALT) %>% separate(ALT, into = c("First", "Second"), sep = ",") %>% na.omit() -> exclude

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

exclude %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> exclude

exclude %>% mutate(locus = paste(CHROM, POS)) -> exclKEY

```

Changing the order of this to make it easier

```{r}
rawdata <- readRDS("rawdata.rds")

#NORMAL RAW DATA
rawdata %>% 
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Pool, Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

rawdata_called %>% mutate(Dataset = gsub("n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Pool", "Dataset"), sep = "_", extra = "merge") -> rd_unnamed

expnames <- read.csv("experiment_names_524.csv")
rd_unnamed %>% left_join(expnames) -> rd

rd %>% filter(CHROM == "IX") -> rd

rd %>% ungroup() %>% select(Pool, Dataset) %>% distinct() -> PoolKey
```

### TESTS

Changing variables

```{r}

#EXCLUSIONS
rd %>% mutate(locus = paste(CHROM, POS)) %>% filter(locus %in% exclKEY$locus == FALSE) %>% select(-locus) -> rd_ex

################################################################################

#MEDIAN ONLY
rd %>%
  group_by(Pool, Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 300, FUN = median, align = "center", na.rm = TRUE))) %>%
  mutate(TEST = "rawdata_median") -> rawdata_median

#GAUSSIAN ONLY
rd %>%
  group_by(Pool, Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = window_size, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "rawdata_gauss") -> rawdata_gauss

#EXCLUSIONS + MEDIAN
rd_ex %>%
  group_by(Pool, Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 300, FUN = median, align = "center", na.rm = TRUE))) %>%
  mutate(TEST = "rawdata_median_ex") -> rawdata_median_ex

#EXCLUSIONS + GAUSSIAN
rd_ex %>%
  group_by(Pool, Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = window_size, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "rawdata_gauss_ex") -> rawdata_gauss_ex

rawdata_gauss_ex %>% mutate(SmoothCount = ceiling(SmoothCount)) %>%
  mutate(TEST = "rawdata_gauss_ex_ceiling") -> rawdata_gauss_ex_ceiling

rawdata_gauss %>% mutate(SmoothCount = ceiling(SmoothCount)) %>%
  mutate(TEST = "rawdata_gauss_ceiling") -> rawdata_gauss_ceiling


#ADDING ONE MORE IN: NAs omitted before gaussian
window_size = 300
sigma = 10
rd_ex %>% select(-Selection) %>% na.omit() %>%
  group_by(Pool, Dataset, CHROM, Allele) %>% 
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = window_size, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "rawdata_gauss_rmNA") -> rawdata_gauss_rmNA

rawdata_gauss_rmNA %>% select(-Pool) -> rawdata_gauss_rmNA

```

## Combine

```{r}
rbind(rawdata_gauss_ceiling,
      rawdata_gauss_ex_ceiling,
      rawdata_gauss_ex,
      rawdata_median_ex,
      rawdata_gauss,
      rawdata_median,
      rawdata_gauss_rmNA) -> testing_smoothing

```


## Change into factors

```{r}
testing_smoothing %>% left_join(expnames) -> testing_smoothing

testing_smoothing %>% 
  filter(Bulk != "Fluconazole") %>%
  select(-Selection, -Replicate) %>% na.omit() %>%
  mutate_if(is.character, as.factor) -> rd_factor

rd_factor %>% filter(Dataset %in% c("O1_aD", "W1_aD") == FALSE) -> rd_factor

contrasts(rd_factor$Background) <- matrix(c(0.5, -0.5))
contrasts(rd_factor$Bulk) <- matrix(c(1, 0))

rd_factor %>% merge(PoolKey) -> rd_factor
```



## Make function to run glmer

```{r}
glmer_cb2_short <- function (..., W, formula, numgroups = FALSE, outputlength = 4, 
    return = c("Z")) 
{
    data <- list(...)
    
    require(lme4)
    if (is.null(W) || is.null(formula)) {
        stop("Weights (W) and formula must be provided")
    }
    glm_formula <- as.formula(formula)
    if (!all(names(data) %in% all.vars(glm_formula))) {
        stop("One or more variables in the formula are not provided as arguments")
    }
    for (i in all.vars(glm_formula)) {
        if (length(unique(as.data.frame(data)[, i])) < 2) {
            output <- rep(NA, outputlength)
            return(output)
        }
    }
    glm_fit <- glmer(glm_formula, data = as.data.frame(data), weights = W, 
        family = binomial)
    if (return %in% "Z") {
        output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients) * 
            0.5) + 1):((length(summary(glm_fit)$coefficients) * 
            0.75))]
    }
    if (length(output) == outputlength) {
        return(output)
    }
    else {
        return(rep(NA, outputlength))
    }
}
```

# Testing differences in Chr IX

### Glmer for all of these combined

```{r}
rd_factor %>% filter(CHROM == "IX") %>% distinct() %>% filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2")) %>%
  group_by(TEST, CHROM, POS) %>% summarize(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background + (1 | Rep) "),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> testing_smoothing_glmer_1byRep

```

```{r}
testing_smoothing_glmer_1byRep %>%
  mutate(TEST = gsub("rawdata_", "", TEST)) %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = TEST)) +
  geom_point(alpha = 0.4, size = 1) +
  facet_grid(TEST~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  ylim(0,3)
```

From this, we can see that exclusions weirdly HELP the Gaussian smoother, and the medians, and that the ceiling makes no difference to the glmer.

```{r}
#How many do we lose from exclusions?
dim(rd)
dim(rd_ex)

rawdata_gauss_ex %>% na.omit() %>% dim()
rawdata_gauss_ex %>% dim()

Reads <- 1:100
Reads <- Reads + rnorm(100, 0, 20)

sigma <- 1
rollapply(Reads, width = 10, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA) -> roll_s1
rollapply(Reads, width = 50, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA) -> roll_s1_w50

sigma <- 10
rollapply(Reads, width = 10, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA) -> roll_s10
rollapply(Reads, width = 50, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA) -> roll_s10_w50

sigma <- 50
rollapply(Reads, width = 10, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA) -> roll_s50

sigma <- 100
rollapply(Reads, width = 10, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA) -> roll_s100

plot(Reads, main = "Reads")
abline(1,1)

plot(roll_s1, main = "roll_s1")
abline(1,1)

plot(roll_s1_w50, main = "roll_s1_w50")
abline(1,1)

plot(roll_s10, main = "roll_s10")
abline(1,1)

plot(roll_s10_w50, main = "roll_s10_w50")
abline(1,1)

plot(roll_s50, main = "roll_s50")
abline(1,1)

plot(roll_s100, main = "roll_s100")
abline(1,1)

```

```{r}
#why the fuck do the exclusions actually matter for this kernel?
window_size = 50
sigma = 10

rd_ex %>%
  group_by(Pool, Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = window_size, align = "center",
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "Window_50_Sigma_10") -> TEST

window_size = 100
sigma = 10

rd_ex %>%
  group_by(Pool, Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = window_size, align = "center",
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "Window_100_Sigma_10") -> TEST2

window_size = 300
sigma = 10

rd_ex %>%
  group_by(Pool, Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = window_size, align = "center",
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "Window_300_Sigma_10") -> TEST3

rd_ex %>% select(-Selection) %>% na.omit() %>% 
  group_by(Pool, Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = 1000, align = "center",
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "Window_1000_Sigma_10_NAOMIT") -> TEST4

rbind(TEST, TEST2, TEST3, TEST4) %>%
  filter(Dataset == "OCa", Allele == "Wine") %>% ggplot(aes(x = POS, y = SmoothCount, color = TEST)) + geom_point(alpha = 0.1) +
  facet_wrap("TEST") +
  theme(legend.position = "bottom")
# rd_ex %>% filter(Dataset == "OCa") %>% ggplot(aes(x = POS, y = Reads, color = Allele)) + geom_point()
# TEST %>% filter(Dataset == "OCa") %>% ggplot(aes(x = POS, y = SmoothCount, color = Allele)) + geom_point()

```


# Window size test

```{r}
#ADDING ONE MORE IN: NAs omitted before gaussian

rd_ex %>% select(-Selection) %>% na.omit() -> rd_ex_rmna
sigma = 10

rd_ex_rmna %>%
  group_by(Pool, Dataset, CHROM, Allele) %>% 
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = 200, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "w300_s10") -> w300_s10


rd_ex_rmna %>%
  group_by(Pool, Dataset, CHROM, Allele) %>% 
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = 500, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "w500_s10") -> w500_s10


rd_ex_rmna %>%
  group_by(Pool, Dataset, CHROM, Allele) %>% 
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = 800, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "w800_s10") -> w800_s10


rd_ex_rmna %>%
  group_by(Pool, Dataset, CHROM, Allele) %>% 
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = rollapply(Reads, width = 1000, align = "center", 
                                FUN = function(z) sum(z * gaussian_kernel(seq_along(z) - (length(z) + 1) / 2, sigma)),
                                fill = NA)) %>%
  mutate(TEST = "w1000_s10") -> w1000_s10

rd_ex_rmna %>% select(Pool, Dataset, CHROM, Allele, POS, SmoothCount = Reads) %>% mutate(TEST = "RawData") %>%
  rbind(w300_s10,
      w500_s10,
      w1000_s10) -> GaussianWindowTest

GaussianWindowTest %>%
filter(Dataset == "OCa", Allele == "Wine") %>% ggplot(aes(x = POS, y = SmoothCount, color = TEST)) + geom_point(alpha = 0.1) +
  #facet_wrap("TEST") +
  theme(legend.position = "bottom")
```

```{r}
################################################################################
GaussianWindowTest %>%
  left_join(expnames)  %>% 
  filter(Bulk != "Fluconazole") %>%
  select(-Selection, -Replicate) %>% na.omit() %>%
   filter(Dataset %in% c("O1_aD", "W1_aD") == FALSE) %>%
  mutate_if(is.character, as.factor) -> GaussianWindowTest_factor

contrasts(GaussianWindowTest_factor$Background) <- matrix(c(0.5, -0.5))
contrasts(GaussianWindowTest_factor$Bulk) <- matrix(c(1, 0))

GaussianWindowTest_factor %>% filter(CHROM == "IX") %>% distinct() %>% filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2"),
                                                                              TEST != "RawData") %>%
  group_by(TEST, CHROM, POS) %>% summarize(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background + (1 | Rep) "),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> smoothGaussian_GLM

smoothGaussian_GLM %>%
  mutate(TEST = gsub("rawdata_", "", TEST)) %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = TEST)) +
  geom_point(alpha = 0.4, size = 1) +
  facet_grid(TEST~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") 
```



