---
title: "MergedVCFs_CuSO4_Gaussian_June24"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))



## Functions for GLM all and Permutation all

glm_cb2_all <- function(..., W, formula, numgroups = FALSE, outputlength = 8) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  output <- summary(glm_fit)$coefficients[c(((length(summary(glm_fit)$coefficients)*0)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.25)),
                                            ((length(summary(glm_fit)$coefficients)*0.5)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.75)))]

  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
} 


################################################################################
# No replicates, fix Parent
cybrPermute_cb2_all <- function(dataset, 
                                R = 0, perp = 1, outlength = 4, 
                                glmform = "Allele ~ Bulk * Parent", 
                                inputlabels = c("Intercept", "Bulk", "Parent", "Interaction")){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled aD")
  print(paste("Your labels are:", inputlabels))

    if(R > 0){ #INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Rep, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == 1) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest    
    }else{ #DO NOT INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == perp) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    }
  
    #PERMUTE TWICE
  if(R > 0){
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Rep, 
             Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep, 
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, 
             Rep, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep,
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }else{
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #RUN THE GLM
    if(R > 0) {
      shuffletoglm2 %>% na.omit() %>%
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }else{
      shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }
    
    
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}

cybrInputGATKTable2 <- function(rawData, yeast = TRUE){

  require(dplyr)
  require(doParallel)
  require(foreach)

  HNGLCDRXY <- read.table(rawData, header = TRUE)

  #Identify the unique values besides AD/DP/GQ/PL
  gsub(".AD", "",
       gsub(".GQ", "",
            gsub(".DP","",
                 gsub(".PL","",
                      colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
  #i <- Samples[1]

  resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
    mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
    AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
    GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
    DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
    PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
    cbind(mydf, AD , GQ , DP, PL) -> mydftotal
    colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")

    mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>%
      separate(PL, c('PL.REF','PL.ALT'), extra='drop') %>%
      #Added 10/18/23:
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf

    mycdf %>% filter(grepl(",", ALT)) %>% 
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
    
    doublecdf %>% filter(grepl(",", ALT)) %>%
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
    
    rbind(mycdf, doublecdf, triplecdf) -> newcdf
    
    newcdf
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
  }else{
    results <- resultscdf
  }
  return(results)

}

```

# The reason for redoing this

```{r}
## Plot Chr VIII Comparison

CB_AIG_glm_VIII_unsm %>% filter(label == "Bulk") %>%
  ungroup() %>%
  reframe(POS = POS, Z = frollapply(zscore, n = 50, FUN = cybr_weightedgauss, align = "center")) -> AIG_50
  
AllCuSO4_g200 %>% filter(label %in% c("Bulk"), CHROM == "VIII") %>%
  ggplot(aes(x = POS, y = abs(zscore))) + geom_line() +
  geom_vline(xintercept = 212535, color = "darkturquoise", size = 2, alpha = 0.4) +
  geom_point(data = CB_AIG_glm_VIII[CB_AIG_glm_VIII$label == "Bulk",], aes(x = POS, y = abs(zscore)), alpha = 0.2, color = "orange") +
  geom_point(data = AIG_50, aes(x = POS, y = Z), color = "firebrick") +
  geom_hline(aes(yintercept = q5))

```


# Test Permutations

Load in raw data with exclusions

```{r}
rawdata_called_ex <- readRDS("rawdata_called_ex.rds")
```


### Separate out what could be smoothed here

This is what the windows look like

```{r}

cybr_weightedgauss <- function(myx){
  myy <- dnorm(1:length(myx), mean = length(myx)/2, sd = 10)
  return(weighted.mean(x = myx, y = myy, na.rm = TRUE))
} 

plot(dnorm(1:200, mean = 200/2, sd = 10), main = "SD 10")
plot(dnorm(1:200, mean = 200/2, sd = 20), main = "SD 20")
plot(dnorm(1:200, mean = 200/2, sd = 100), main = "SD 100") #choosing this one for including all of the points
plot(dnorm(1:200, mean = 200/2, sd = 1000), main = "SD 1000")

```


This is how we get the original data with Gaussian of n 200

```{r, eval = FALSE}

rawdata_called_ex %>% ungroup() %>% #filter(Dataset == "HTG3TDMXY_n01_O8_451_E_2.fastq", CHROM == "VII") %>%
  group_by(Dataset, CHROM, Allele) %>%
  arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center")) %>% mutate(smooth = 200) -> rawdata_called_ex_G200

#saveRDS(rawdata_called_ex_G200, file = "rawdata_called_ex_G200.rds")

rawdata_G200 <- readRDS("rawdata_called_ex_G200.rds")
```


## Separate into bulks, make ready for glms

```{r}

rawdata_called_ex %>% mutate(Dataset = gsub("n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Pool", "Dataset"), sep = "_", extra = "merge") -> rd

expnames <- read.csv("experiment_names_524.csv")

rd %>% left_join(expnames) -> rd_named

```

## Use only unselected bulks and change into factors

```{r}
rd_named %>% filter(Bulk == "Dilute") %>%
  select(-Selection) %>% na.omit() -> rd_factor

#Not sure why these were excluded? Unless that experiment was just wrong or had no reps?
rd_factor %>% filter(Dataset %in% c("O1_aD", "W1_aD") == FALSE) -> rd_factor

# contrasts(rd_factor$Background) <- matrix(c(0.5, -0.5))
# contrasts(rd_factor$Bulk) <- matrix(c(1, 0))

unique(rd_factor$Dataset)

rd_factor %>% ungroup() %>% select(Pool, Dataset, Bulk, Background) %>% distinct()
```

################################################################################

# Run Permutations

```{r}
rd_factor %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = Reads) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Background) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

head(rd_shuffled_selected)

rd_factor %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = Reads) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Background) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) -> Perm1

Perm1 %>% ungroup() %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") %>% unnest() %>%
  mutate(POS = as.numeric(POS)) %>%
  mutate_if(is.character, as.factor) -> Perm1contrasts

contrasts(Perm1contrasts$Background) <- matrix(c(-0.5, 0.5))
contrasts(Perm1contrasts$Bulk)
```

# Run glmer on permuted values

## Make function to run glmer

```{r}
glmer_cb2_short <- function (..., W, formula, numgroups = FALSE, outputlength = 4, 
    return = c("Z")) 
{
    data <- list(...)
    
    require(lme4)
    if (is.null(W) || is.null(formula)) {
        stop("Weights (W) and formula must be provided")
    }
    glm_formula <- as.formula(formula)
    if (!all(names(data) %in% all.vars(glm_formula))) {
        stop("One or more variables in the formula are not provided as arguments")
    }
    for (i in all.vars(glm_formula)) {
        if (length(unique(as.data.frame(data)[, i])) < 2) {
            output <- rep(NA, outputlength)
            return(output)
        }
    }
    glm_fit <- glmer(glm_formula, data = as.data.frame(data), weights = W, 
        family = binomial)
    if (return %in% "Z") {
        output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients) * 
            0.5) + 1):((length(summary(glm_fit)$coefficients) * 
            0.75))]
    }
    if (length(output) == outputlength) {
        return(output)
    }
    else {
        return(rep(NA, outputlength))
    }
}
```



```{r}

Perm1contrasts %>%  filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2")) %>% distinct() %>%
  ungroup() %>% 
  filter((CHROM == "II" & POS == 791816) == FALSE,
         (CHROM == "IV" & POS == 1296744) == FALSE,
         (CHROM == "VII" & POS == 1014774) == FALSE) %>%
  group_by(CHROM, POS) %>% reframe(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = Reads, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background +  (1 | Rep)"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_CSS1_glmer_PERM_unsmoothed


# Perm1contrasts %>%  filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2")) %>% distinct() %>% 
#   
#   group_by(CHROM, POS) %>% reframe(zscore = glmer_cb2_short(Bulk = Bulk,  
#                                                             Background = Background,
#                                                             Rep = Dataset,
#                                                             Allele = Allele,
#                                                             W = Reads, outputlength = 4,
#                                                             formula = "Allele ~ Bulk*Background +  (1 | Rep)"),
#                                                                   label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_CSS1_glmer_PERM_unsmoothed

# saveRDS(CuSO4_CSS1_glmer_1byRep_PERM, "CuSO4_CSS1_glmer_1byRep_PERMG.rds")

# Perm1contrasts %>%
#   filter(Pool %in% c("HVYTYDRX2","HTG3TDMXY")) %>% distinct() %>% 
#   mutate(POS = as.numeric(POS)) %>%
#   mutate_if(is.character, as.factor) %>%
#   group_by(CHROM, POS) %>% reframe(zscore = glmer_cb2_short(Bulk = Bulk,  
#                                                             Background = Background,
#                                                             Rep = Dataset,
#                                                             Allele = Allele,
#                                                             W = Reads, outputlength = 4,
#                                                             formula = "Allele ~ Bulk*Background + (1 | Rep)"),
#                                                                   label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_CSS8_glmer_PERM_unsmoothed
# CuSO4_CSS8_glmer_PERM_unsmoothed %>% na.omit() %>% group_by(label) %>% summarize(q5 = quantile(abs(zscore), 0.95)) %>% mutate(CSS = "I") %>% rbind(q5_css8) -> q5_CuSO4_CSS

# saveRDS(CuSO4_CSS8_glmer_1byRep_PERM, "CuSO4_CSS8_glmer_1byRep_PERMG.rds")
#Look at the quantiles without smoothing
CuSO4_CSS1_glmer_PERM_unsmoothed %>% na.omit() %>% group_by(label) %>% summarize(q5 = quantile(abs(zscore), 0.95)) %>% mutate(CSS = "Unsmoothed") -> c

#Look at the plot without smoothing
CuSO4_CSS1_glmer_PERM_unsmoothed %>% filter(abs(zscore) < 2e5) %>% ggplot(aes(x = POS, y = abs(zscore), color = label)) + geom_point(alpha = 0.2) + facet_grid(~CHROM, scales = "free", space = "free")

#Smooth with the weighted gaussian?
CuSO4_CSS1_glmer_PERM_unsmoothed %>% ungroup() %>% na.omit() %>% group_by(CHROM, label) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = frollapply(abs(zscore), n = 200, FUN = cybr_weightedgauss, align = "center"))-> CuSO41_Perm_smoothed_I_200

#Look at the plot with smoothing
CuSO41_Perm_smoothed_I %>% na.omit() %>% ungroup() %>% group_by(label) %>% summarize(q5 = quantile(abs(SmoothCount), 0.95)) %>% mutate(CSS = "Window 50") -> a
CuSO41_Perm_smoothed_I_200 %>% na.omit() %>% ungroup() %>% group_by(label) %>% summarize(q5 = quantile(abs(SmoothCount), 0.95)) %>% mutate(CSS = "Window 200") -> b
#CuSO4_CSS8_glmer_1byRep_PERM <- readRDS("CuSO4_CSS8_glmer_1byRep_PERMG.rds")
CuSO4_CSS8_glmer_1byRep_PERM %>% na.omit() %>% ungroup() %>% group_by(label) %>% summarize(q5 = quantile(abs(zscore), 0.95)) %>% mutate(CSS = "Smoothed Counts") %>%
  rbind(a) %>%
  rbind(b) %>%
  rbind(c) %>%
  ggplot(aes(x = CSS, y = q5, color = label)) + geom_point(size = 4, alpha = 0.4) + ylim(0,3)

CuSO41_Perm_smoothed_VIII %>% filter(SmoothCount < 3000) %>% ggplot(aes(x = POS, y = SmoothCount, color = label)) + geom_point(alpha = 0.2) + facet_grid(~CHROM, scales = "free", space = "free")

```

# Quick plots to compare

```{r, eval = FALSE}
q5_CuSO4_CSS <- readRDS("q5_CuSO4_CSS.rds")
CuSO4_CSS1_glmer_1byRep <- readRDS("CuSO4_CSS1_glmer_1byRepG.rds")
CuSO4_CSS8_glmer_1byRep <- readRDS("CuSO4_CSS8_glmer_1byRepG.rds")

CuSO4_CSS1_glmer_1byRep %>% mutate(CSS = "I") -> temp
CuSO4_CSS8_glmer_1byRep %>% mutate(CSS = "VIII") %>%
  rbind(temp) %>%
  merge(q5_CuSO4_CSS) -> AllCuSO4_g200

AllCuSO4_g200 %>% filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = zscore, color = CSS)) + geom_line() +
  geom_hline(aes(yintercept = q5, color = CSS)) +
  geom_hline(aes(yintercept = -q5, color = CSS)) +
  facet_grid(label~CHROM, scales = "free", space = "free")

AllCuSO4_g200 %>% filter(label %in% c("Bulk"), CHROM == "VIII") %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CSS)) + geom_line() +
  geom_hline(aes(yintercept = q5, color = CSS))


```

# Quick process of Chr VIII of AIGs

```{r}
CB_AIGs <- readRDS("AIG_CFZ_rawdata.rds")

CB_AIGs %>% filter(CHROM == "VIII") %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") %>%
  ungroup() %>%
  group_by(Dataset, CHROM, Allele) %>%
  arrange(POS) %>% #
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 50, FUN = cybr_weightedgauss, align = "center"))-> CB_AIGs_smoothed_VIII

CB_AIGs %>% filter(CHROM == "XII") %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") %>%
  ungroup() %>%
  group_by(Dataset, CHROM, Allele) %>%
  arrange(POS) %>% #
  reframe(POS = POS, SmoothCount = frollapply(Reads, n = 50, FUN = cybr_weightedgauss, align = "center"))-> CB_AIGs_smoothed_XII

CB_AIGs_smoothed_VIII %>%
  mutate(Dataset = gsub("H5LHYDRX5_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Bulk", "Rep")) %>%
  filter(Bulk %in% c("C", "F", "Z", "U")) %>%
  distinct() %>% 
  mutate_if(is.character, as.factor) -> CB_AIG_readytoglm_unsm

CB_AIGs_smoothed_XII %>%
  mutate(Dataset = gsub("H5LHYDRX5_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Bulk", "Rep")) %>%
  filter(Bulk %in% c("C", "F", "Z", "U")) %>%
  distinct() %>% 
  mutate_if(is.character, as.factor) -> CB_AIG_readytoglm_XII

glmer_cb2_short_adj <- function (..., W, formula, numgroups = FALSE, outputlength = 4, 
    return = c("Z")) 
{
    data <- list(...)
    
    require(lme4)
    if (is.null(W) || is.null(formula)) {
        stop("Weights (W) and formula must be provided")
    }
    glm_formula <- as.formula(formula)
    if (!all(names(data) %in% all.vars(glm_formula))) {
        stop("One or more variables in the formula are not provided as arguments")
    }
    for (i in all.vars(glm_formula)) {
        if (length(unique(as.data.frame(data)[, i])) < 2) {
            output <- rep(1, outputlength)
            return(output)
        }
    }
    glm_fit <- glmer(glm_formula, data = as.data.frame(data), weights = W, 
        family = binomial)
    if (return %in% "Z") {
        output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients) * 
            0.5) + 1):((length(summary(glm_fit)$coefficients) * 
            0.75))]
    }
    if (length(output) == outputlength) {
        return(output)
    }
    else {
        return(rep(2, outputlength))
    }
}

CB_AIG_readytoglm %>% 
  #mutate(SmoothCount = Reads + 1) %>%
  filter(Bulk %in% c("C", "U")) %>%
  ungroup() %>%
  na.omit() %>%
  mutate_if(is.character, as.factor) %>%
  group_by(POS) %>% summarize(zscore = glmer_cb2_short(Bulk = Bulk, 
                                                              Rep = Rep,
                                                            Allele = Allele,
                                                            W = SmoothCount, 
                                                            outputlength = 2,
                                                            formula = "Allele ~ Bulk + (1 | Rep)"),
                                    label = c("cept","Bulk"))  -> CB_AIG_glm_VIII
CB_AIG_readytoglm_XII %>% 
  #mutate(SmoothCount = Reads + 1) %>%
  filter(Bulk %in% c("C", "U")) %>%
  ungroup() %>%
  na.omit() %>%
  mutate_if(is.character, as.factor) %>%
  group_by(POS) %>% summarize(zscore = glmer_cb2_short(Bulk = Bulk, 
                                                              Rep = Rep,
                                                            Allele = Allele,
                                                            W = SmoothCount, 
                                                            outputlength = 2,
                                                            formula = "Allele ~ Bulk + (1 | Rep)"),
                                    label = c("cept","Bulk"))  -> CB_AIG_glm_XII
```


NOTES FROM MARK

- weight by distance AND total counts?
- or just by total counts?
- to do permutations, sample with probabilites that are the weights?
- don't just scramble? I don't think that's how this works...
- scramble WITHIN the window, then compute z score?
- normal ANOVA - just what is the variance that I have to beat?
- permute by SNP and then smooth by abs z score...?
- compare thresholds with that

- change cutoffs and iterate to find the right balance of FDR that gives what % they would be under a peak?...???
- look at how they do it in RNAseq
- call number of positives and false positives...? Check if someone has solved this already
- then use the rest of the data to say if positives are positive......? Tunable thing, check this out (at least know what he's talking about)