---
title: "Merged CuSO4 Figures (Gaussian 200 window)"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)



# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))



## Functions for GLM all and Permutation all

glm_cb2_all <- function(..., W, formula, numgroups = FALSE, outputlength = 8) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  output <- summary(glm_fit)$coefficients[c(((length(summary(glm_fit)$coefficients)*0)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.25)),
                                            ((length(summary(glm_fit)$coefficients)*0.5)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.75)))]

  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
} 


################################################################################
# No replicates, fix Parent
cybrPermute_cb2_all <- function(dataset, 
                                R = 0, perp = 1, outlength = 4, 
                                glmform = "Allele ~ Bulk * Parent", 
                                inputlabels = c("Intercept", "Bulk", "Parent", "Interaction")){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled aD")
  print(paste("Your labels are:", inputlabels))

    if(R > 0){ #INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Rep, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == 1) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest    
    }else{ #DO NOT INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == perp) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    }
  
    #PERMUTE TWICE
  if(R > 0){
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Rep, 
             Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep, 
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, 
             Rep, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep,
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }else{
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #RUN THE GLM
    if(R > 0) {
      shuffletoglm2 %>% na.omit() %>%
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }else{
      shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }
    
    
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}

cybrInputGATKTable2 <- function(rawData, yeast = TRUE){

  require(dplyr)
  require(doParallel)
  require(foreach)

  HNGLCDRXY <- read.table(rawData, header = TRUE)

  #Identify the unique values besides AD/DP/GQ/PL
  gsub(".AD", "",
       gsub(".GQ", "",
            gsub(".DP","",
                 gsub(".PL","",
                      colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
  #i <- Samples[1]

  resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
    mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
    AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
    GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
    DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
    PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
    cbind(mydf, AD , GQ , DP, PL) -> mydftotal
    colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")

    mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>%
      separate(PL, c('PL.REF','PL.ALT'), extra='drop') %>%
      #Added 10/18/23:
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf

    mycdf %>% filter(grepl(",", ALT)) %>% 
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
    
    doublecdf %>% filter(grepl(",", ALT)) %>%
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
    
    rbind(mycdf, doublecdf, triplecdf) -> newcdf
    
    newcdf
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
  }else{
    results <- resultscdf
  }
  return(results)

}

theme_classic
```

# Load in data

```{r}
#rd_factor <- readRDS("rd_factorG.rds")
q5_css_G200 <- readRDS("q5_css_G200_100.rds")
AllCu_plot <- readRDS("AllCu_plot.rds")
AIG_CU_cybrG50 <- readRDS("AIG_CU_cybrG50.rds")


#AllCu_plot %>% filter(abs(zscore) < 20) -> AllCu_plot

```


```{r, eval = FALSE}
CuSO4_CSS1_glmer_unsm <- readRDS("CuSO4_CSS1_glmer_unsmG.rds")
CuSO4_CSS8_glmer_unsm <- readRDS("CuSO4_CSS8_glmer_unsmG.rds")

CuSO4_CSS1_glmer_unsm %>% mutate(CSS = "I") -> temp
CuSO4_CSS8_glmer_unsm %>% mutate(CSS = "VIII") %>%
  rbind(temp) -> AllCuSO4_unsm

AllCuSO4_unsm %>% ggplot(aes(x = abs(zscore))) + geom_density() +
  theme(axis.text.x = element_text()) +
  xlim(0, 40)

AllCuSO4_unsm %>% na.omit() %>% arrange(POS) %>% group_by(CSS, CHROM, label) %>%
  filter(abs(zscore) < 100) %>%
  
  #filter(CHROM == "VIII", label == "Bulk") %>%
  reframe(POS = POS, SmoothCount = frollapply(zscore, n = 200, FUN = cybr_weightedgauss, align = "center")) -> AllCuSO4_cybrG200_100

AllCuSO4_cybrG200_100 %>% merge(q5_css_G200_100) -> AllCu_plot

AllCu_plot %>% mutate(zscore = SmoothCount) -> AllCu_plot

saveRDS(AllCu_plot, file = "AllCu_plot.rds")
```


What genes are nearby?

```{r, fig.width=10, fig.height=10}
#file.choose()
eQTL_hubs <- read.csv("C:\\Users\\Cassandra\\Downloads\\SI_Data_08_hotspotTableForPaper_topGenes100_170420.csv")
eQTL_hubs %>% select(hotspotMarker, numberGenesInHotspot, numberEQTLInHotspot, bestGeneBootstraps, bestGenePeak) %>%
  separate(hotspotMarker, c("CHROM", "POS_SNP"), sep = c(":")) %>%
  separate(POS_SNP, c("POS", "SNP"), sep = c("_")) -> eQTL_hubs_POS

# ChromosomeScale %>% select(CHROM, POS) %>% filter(POS != 1) %>% mutate(CHROM = as.character(CHROM)) -> ChromLength
# eQTL_hubs_POS %>% count(CHROM) %>% mutate(CHROM = gsub("chr", "", CHROM)) %>% merge(ChromLength) %>%
#   ggplot(aes(x = POS, y = n)) + geom_smooth(method = "lm", color = "black") + geom_point(aes(color = CHROM), size = 4) 

eQTL_hubs_POS %>% mutate(CHROM = gsub("chr", "", CHROM)) %>% filter(bestGeneBootstraps == bestGenePeak) %>%
  select(-bestGeneBootstraps) %>%
  separate(bestGenePeak, c("G1", "G2", "G3", "G4"), sep = ";") %>%
  pivot_longer(starts_with("G"), names_to = "NumGenes", values_to = "GeneName") %>%
  na.omit() -> eQTL_hubs_Genes

eQTL_hubs_Genes %>% arrange(desc(numberGenesInHotspot))
eQTL_hubs_Genes %>% arrange(desc(numberEQTLInHotspot))

```

```{r, fig.width=10, fig.height=5}
sgd_orfs <- read.csv("C:/Users/Cassandra/Documents/R_FileTransfer/SGD_ORFs.csv")

genes <- c( "CUP1-1", "CUP1-2", "FRE1") #,
           # "CRS5", "SOD1", "MET17", "FRE7", "MUP1", "PCL1", "OYE3",  "CTR1", "IRC7", "FET3", "PTK1", "CTR3", "MAC1")
sgd_orfs %>% filter(Gene.symbol %in% genes) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_copper

genes_CB <- c("MATALPHA2", "MATA", "URA3", "HO")
sgd_orfs %>% filter(Gene.symbol %in% genes_CB) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_CB

genes_eh <- c("VMA4", "MTL1", "PTR2", "LEU9", "IRC15")
sgd_orfs %>% filter(Gene.symbol %in% genes_eh) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_copper_eh

#HUBS
sgd_orfs %>% filter(Gene.symbol %in% c("PTR3", "VPS70", "CUP1", "YLR257W")) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) %>% mutate(Hub = "Matsui") -> sgd_hubs_Matsui

sgd_orfs %>% filter(Gene.symbol %in% eQTL_hubs_Genes$GeneName) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) %>% mutate(Hub = "Albert") %>% arrange(CHROM) -> sgd_hubs_Albert

#Sporulation Genes
sgd_orfs %>% filter(Gene.symbol %in% c("RME1", "MKT1", "TAO3")) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) %>% mutate(Hub = "Sporulation") -> sgd_sporulation

AllCu_plot %>% filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) + 
  geom_line(aes(linetype = CSS)) +
  geom_vline(data = sgd_sporulation, 
             aes(xintercept = Gene.chromosomeLocation.start),
             size = 1, alpha = 0.3, color = "black", linetype = "dashed") +
  geom_vline(data = sgd_hubs_Matsui, 
             aes(xintercept = Gene.chromosomeLocation.start),
             size = 1, alpha = 0.3, color = "red", linetype = "dashed") +
  geom_vline(data = eQTL_hubs_Genes[eQTL_hubs_Genes$numberGenesInHotspot > 5,],
             aes(xintercept = as.numeric(POS), color = CHROM),
             size = 1, alpha = 0.3) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0)) + ggtitle("eQTL Hub Locations")
```



# Comparison with AIGs

```{r, fig.height=2, fig.width=4}
#CUP1
AllCu_plot %>% filter(label %in% c("Bulk"), CHROM == "VIII") -> Chr8Isolated

AIG_CU_cybrG50 %>% filter(label == "Bulk", CHROM == "VIII") %>% ggplot(aes(x = POS, y = (SmoothCount))) + geom_line(color = "darkorange") +
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = -1.311763, color = "red", alpha = 0.4) +
  geom_hline(yintercept = 1.311763, color = "red", alpha = 0.4) +
  geom_vline(xintercept = 212535, color = "skyblue3") + ggtitle("Chr VIII") + ylab("Z-Score") + xlab("Chromosome VIII Position") +
  geom_line(data = Chr8Isolated, aes(x = POS, y = (zscore))) +
  theme(legend.position = "none")

#FRE1
AllCu_plot %>% filter(label %in% c("Bulk"), CHROM == "XII") -> Chr12Isolated

AIG_CU_cybrG200 %>% filter(label == "Bulk", CHROM == "XII") %>% ggplot(aes(x = POS, y = (SmoothCount))) + geom_line(color = "darkorange") +
  geom_hline(yintercept = 0, color = "gray") +
  geom_hline(yintercept = -1.311763, color = "red", alpha = 0.4) +
  geom_hline(yintercept = 1.311763, color = "red", alpha = 0.4) +
  geom_line(data = Chr12Isolated, aes(x = POS, y = (zscore), linetype = CSS)) +
  geom_vline(xintercept = 568567, color = "skyblue3") +  ggtitle("Chr XII") + ylab("Z-Score") + xlab("Chromosome XII Position") +
    theme(legend.position = "none")

```

```{r}
# AIG_CU_cybrG50 %>% filter(label == "Bulk", CHROM == "XII") %>% ggplot(aes(x = POS, y = (SmoothCount))) + geom_line(color = "skyblue2") +
#   geom_line(data = Chr12Isolated, aes(x = POS, y = (zscore), linetype = CSS)) +
#   geom_hline(yintercept = 0) +
#   geom_vline(xintercept = 568567, color = "red") + ggtitle("FRE1 F2 and F12 G50")

#CHROM 7
AllCu_plot %>% filter(label %in% c("Bulk", "Interaction"), CHROM == "VII") -> Chr7Isolated

AIG_CU_cybrG200 %>% filter(label == "Bulk", CHROM == "VII") %>% ggplot(aes(x = POS, y = (SmoothCount))) + geom_line(color = "black") +
  geom_line(data = Chr7Isolated, aes(x = POS, y = (zscore), linetype = CSS, color = label)) +
  geom_hline(yintercept = 0) + ggtitle("Chr VII, Window 200")

# AIG_CU_cybrG50 %>% filter(label == "Bulk", CHROM == "VII") %>% ggplot(aes(x = POS, y = (SmoothCount))) + geom_line(color = "gray") +
#   geom_line(data = Chr7Isolated, aes(x = POS, y = (zscore), linetype = CSS, color = label)) +
#   geom_line(data = AIG_CU_cybrG200[AIG_CU_cybrG200$CHROM == "VII" & AIG_CU_cybrG200$label == "Bulk",], color = "black") +
#   geom_hline(yintercept = 0) + ggtitle("Chr VII, Window 50")


#CHROM 10
AllCu_plot %>% filter(label %in% c("Bulk", "Interaction"), CHROM == "X") -> Chr10Isolated

AIG_CU_cybrG200 %>% filter(label == "Bulk", CHROM == "X") %>% ggplot(aes(x = POS, y = (SmoothCount))) + geom_line(color = "black") +
  geom_line(data = Chr10Isolated, aes(x = POS, y = (zscore), linetype = CSS, color = label)) +
  geom_hline(yintercept = 0) + ggtitle("Chr X, Window 200")


#CHROM 12
AllCu_plot %>% filter(label %in% c("Bulk", "Interaction"), CHROM == "XIV") -> Chr12Isolated

AIG_CU_cybrG200 %>% filter(label == "Bulk", CHROM == "XIV") %>% ggplot(aes(x = POS, y = (SmoothCount))) + geom_line(color = "black") +
  geom_line(data = Chr12Isolated, aes(x = POS, y = (zscore), linetype = CSS, color = label)) +
  geom_hline(yintercept = 0) + ggtitle("Chr XIV, Window 200")


```

```{r, fig.width=12, fig.height=5}

#Do this all at once
AllCu_plot %>% filter(label %in% c("Bulk")) %>%
  ggplot() +
  geom_line(data = AIG_CU_cybrG50[AIG_CU_cybrG50$label == "Bulk" & AIG_CU_cybrG50$CHROM != "M",], aes(x = POS, y = SmoothCount), color = "gray", alpha = 0.3) +
  geom_line(data = AIG_CU_cybrG200[AIG_CU_cybrG200$label == "Bulk"& AIG_CU_cybrG200$CHROM != "M",], aes(x = POS, y = SmoothCount), color = "red", alpha = 0.3) +
  geom_line(aes(x = POS, y = zscore, linetype = CSS)) +
  facet_grid(~CHROM, space = "free", scales = "free") +
  geom_vline(data = sgd_copper, aes(xintercept = Gene.chromosomeLocation.start, color = Gene.symbol)) +
  geom_vline(data = sgd_CB, aes(xintercept = Gene.chromosomeLocation.start, color = Gene.symbol), linetype = "dashed") +
  theme(legend.position = "bottom") 

#Do this all at once
AllCu_plot$CHROM <- factor(AllCu_plot$CHROM, levels = as.character(as.roman(1:16)))
AllCu_plot %>% filter(label %in% c("Bulk")) %>%
  ggplot() +
  # geom_line(data = AIG_CU_cybrG50[AIG_CU_cybrG50$label == "Bulk" & AIG_CU_cybrG50$CHROM != "M",], aes(x = POS, y = SmoothCount), color = "gray", alpha = 0.3) +
  # geom_line(data = AIG_CU_cybrG200[AIG_CU_cybrG200$label == "Bulk"& AIG_CU_cybrG200$CHROM != "M",], aes(x = POS, y = SmoothCount), color = "red", alpha = 0.3) +
  geom_hline(yintercept = 0) +
  geom_hline(aes(yintercept = -q5), alpha = 0.4, color = "red") +
  geom_hline(aes(yintercept = q5), alpha = 0.4, color = "red") +
  geom_line(aes(x = POS, y = zscore, linetype = CSS)) +
  geom_vline(data = sgd_copper, aes(xintercept = Gene.chromosomeLocation.start, color = Gene.symbol)) +
  facet_grid(~CHROM, space = "free", scales = "free") +
  theme(legend.position = "bottom") +
  scale_color_manual(values = c("skyblue3", "skyblue3", "skyblue3"))
  
  
```


# Final Plotting

Ensure filters happen

```{r}
AllCu_plot$CHROM <- factor(AllCu_plot$CHROM, levels=as.character(as.roman(c(1:16,1000))))
sgd_copper$CHROM <- factor(sgd_copper$CHROM, levels=as.character(as.roman(c(1:16,1000))))
sgd_copper_eh$CHROM <- factor(sgd_copper_eh$CHROM, levels=as.character(as.roman(c(1:16,1000))))

AllCu_plot %>% filter(label %in% c("Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) + 
  geom_point(aes(alpha = abs(zscore) > q5), color = "gray", size = 0.1) +
  geom_point(aes(alpha = abs(zscore) < q5), color = "black", size = 0.1) +
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
  geom_vline(data = sgd_copper_eh,
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             linetype = "dashed", size = 1, alpha = 0.3) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") + ggtitle("Interaction points") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))

AllCu_plot %>% filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) + 
  geom_point(aes(alpha = abs(zscore) > q5), color = "gray", size = 0.1) +
  geom_point(aes(alpha = abs(zscore) < q5), color = "black", size = 0.1) +
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
  geom_vline(data = sgd_copper_eh,
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             linetype = "dashed", size = 1, alpha = 0.3) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") + ggtitle("Bulk points") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))
```

```{r, fig.height=6, fig.width=12}
AllCu_plot %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), linetype = CSS)) + 
  geom_line() +
  geom_hline(aes(yintercept = q5)) +
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))
  
```

Looking at FRE1

```{r, fig.height=4, fig.width=8}
AllCu_plot %>% filter(CHROM == "XII") %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), linetype = CSS)) + 
  geom_line() +
  geom_vline(aes(xintercept = 568567, color = "FRE1"),
             size = 1, alpha = 0.3) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))

AllCu_plot %>% filter(CHROM == "VIII") %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), linetype = CSS)) + 
  geom_line() +
  geom_vline(aes(xintercept = 212535, color = "CUP1"),
             size = 1, alpha = 0.3) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))

```

```{r}
AllCu_plot %>% 
  ungroup() %>% na.omit() %>%
  group_by(CHROM, CSS, label) %>%
  arrange(desc(abs(zscore))) %>%
  summarize(peakPOS = head(POS, 1),
            zscore = max(abs(zscore))) %>%
  merge(q5_css_G200) %>%
  filter(abs(zscore) > q5) -> peaks

knowngenes <- data.frame(Gene = c("CUP1", "FRE1", "FRE1"),
                         POS = c(212535, 568567, 570627),
                         CHROM = c("VIII", "XII", "XII"))

AllCu_plot %>% 
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), linetype = CSS)) +
  #cutoffs and genes
  geom_vline(data = knowngenes, aes(xintercept = POS), color = "red", size = 1) +
  geom_vline(data = peaks[peaks$label %in% c("Bulk", "Interaction"),], 
             aes(xintercept = peakPOS, color = CHROM, linetype = CSS),
             size = 1, alpha = 0.8) +
  #actual data
  geom_line() +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = "CSS", alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))

#peaks %>% filter(CHROM == "VIII")
#rd_logratios %>% filter(POS == 141525) %>% distinct()

AllCu_plot %>% 
  filter(CHROM == "XII") %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), shape = CSS)) +
  #cutoffs and genes
  geom_rect(aes(xmin = 568567, xmax = 570627, ymin = -Inf, ymax = Inf), alpha = 0.2, fill = "lightblue") +
  geom_vline(data = peaks[peaks$label %in% c("Bulk") & peaks$CHROM == "XII",], 
             aes(xintercept = peakPOS, color = CHROM, linetype = CSS),
             size = 1, alpha = 0.8) +
  #actual data
  geom_point(alpha = 0.4, size = 1) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = "CSS", alpha = NULL) +
  theme(legend.position = "bottom") +
  xlim(520000, 600000) +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))+
  theme(axis.text.x=element_text())

AllCu_plot %>% 
  filter(CHROM == "VIII") %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), shape = CSS)) +
  #cutoffs and genes
  geom_rect(aes(xmin = 214533, xmax = 214718, ymin = -Inf, ymax = Inf), alpha = 0.2, fill = "lightblue") +
  geom_vline(data = peaks[peaks$label %in% c("Bulk") & peaks$CHROM == "VIII",], 
             aes(xintercept = peakPOS, color = CHROM, linetype = CSS),
             size = 1, alpha = 0.8) +
  #actual data
  geom_point(alpha = 0.4, size = 1) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = "CSS", alpha = NULL) +
  theme(legend.position = "bottom") +
  xlim(c(2e05, 2.25e05))+
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))+
  theme(axis.text.x=element_text())

```

```{r}

peaks %>% filter(CHROM == "XII") %>%
  mutate(Dist = peakPOS - 570627) %>%
  arrange(abs(Dist)) -> FRE1_dist

FRE1_dist

peaks %>% filter(CHROM == "VIII") %>%
  mutate(Dist = peakPOS - 212535) %>%
  arrange(abs(Dist)) -> CUP1_dist

CUP1_dist


```

# Peak finder based on actual change in direction

```{r}

findchange <- function(x){
  t <- length(x)
  diff <- x[t] - x[1]
  if(is.na(diff)){
    return(NA)
  }else if(diff == 0){
    return(0)
  }else{
    return(diff > 0)  
  }
}

findpeak <- function(x){
  t <- length(x)
  diff <- x[t] - x[1]
  return(diff)
}

  
```

Applying this to everythingg

```{r}
AllCu_plot %>% 
  ungroup() %>% 
  na.omit() %>%
  arrange(POS) %>%
  group_by(label, CSS, CHROM) %>%
  mutate(SmoothZ = zscore) %>%
  summarize(POS = POS,
            zscore = zscore,
            SmoothZ = frollapply(zscore, n = 100, FUN = mean, align = "center", na.rm = TRUE)) %>%
  summarize(POS = POS,
                    zscore = zscore,
                    SmoothZ = SmoothZ,
                    change = frollapply(abs(SmoothZ), n = 100, findchange, align = "center")) %>%
  summarize(POS = POS,
                    zscore = zscore,
                    SmoothZ = SmoothZ,
          change = change,
                    peak = frollapply(abs(change), n = 5, findpeak, align = "center")) -> smoothpeak

# smoothpeak %>% filter(label == "Bulk") %>% ggplot(aes(x = POS, y = zscore)) + geom_line() + geom_line(aes(y = SmoothZ), color = "lightblue") +
#   facet_grid(CSS~CHROM, scales = "free", space = "free") +
#   theme(legend.position = "bottom")
  
smoothpeak%>%
  filter(peak == -1) %>% merge(q5_css_G200) %>% filter(abs(zscore) > q5) -> TruePeaks

################################################################################
#TESTING
AllCu_plot %>% filter(label == "Bulk", CHROM == "VII") %>%
  ungroup() %>% 
  na.omit() %>%
  arrange(POS) %>%
  group_by(label, CSS, CHROM) %>%
  #mutate(SmoothZ = zscore) %>%
  summarize(POS = POS,
            zscore = zscore,
            SmoothZ = frollapply(zscore, n = 200, FUN = mean, align = "center", na.rm = TRUE)) %>%
    ggplot(aes(x = POS, y = zscore)) + geom_point() + geom_point(aes(y = SmoothZ), color = "lightblue") +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom")
  
AllCu_plot %>% filter(label == "Bulk", CHROM == "VII") %>% ggplot(aes(x = zscore, color = CSS, y = CSS)) + geom_boxplot(size = 3) + geom_jitter(alpha = 0.3) +
  theme(axis.text.x = element_text())

################################################################################
cbyr_ksmooth <- function(z, p, b = 1000){
  result <- smoothr::smooth_ksmooth(matrix(c(z,p), ncol = 2), bandwidth = b)
  return(data.frame(result))
}

#Using ksmooth
AllCu_plot %>% filter(label == "Bulk", CHROM == "VII") %>%
  ungroup() %>% 
  na.omit() %>%
  arrange(POS) %>%
  group_by(label, CSS, CHROM) %>%
  summarize(cbyr_ksmooth(z = zscore, p = POS, b = 20000)) -> anattempt

anattempt %>%
  ggplot(aes(y = X1, x = X2)) + geom_line() +
  geom_line(data = AllCu_plot[AllCu_plot$label == "Bulk" & AllCu_plot$CHROM == "VII",], aes(x = POS, y = zscore), color = "red") +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom")


```

```{r, fig.width=8, fig.height=4}

AllCu_plot %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CSS)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk", "Interaction"),],
             aes(xintercept = POS, color = CSS), alpha = 0.2, linetype = "dashed")+
  geom_line(alpha = 0.4, size = 1) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  scale_color_manual(values = c("navyblue", "darkorange")) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

AllCu_plot %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CSS)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk"),],
             aes(xintercept = POS, color = CSS), alpha = 0.2, linetype = "dashed")+
  geom_line(alpha = 0.4, size = 1) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") + #ylim(0,12) +
  scale_color_manual(values = c("navyblue", "darkorange")) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("Bulk Effect Peaks")

```

```{r, fig.width=8, fig.height=4}
AllCu_plot %>% 
  filter(label %in% c("Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CHROM %in% c("I", "III", "V", "VII", "IX", "XI", "XIII", "XV"))) + 
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Interaction"),], 
             aes(xintercept = POS, color = CHROM %in% c("I", "III", "V", "VII", "IX", "XI", "XIII", "XV")), alpha = 0.2, linetype = "dashed")+
  geom_line(alpha = 0.4, size = 1) +
  labs(color = NULL)+
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none") + #ylim(0,12) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("Interaction Effect Peaks")

AllCu_plot %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CHROM %in% c("I", "III", "V", "VII", "IX", "XI", "XIII", "XV"))) + 
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk"),], 
             aes(xintercept = POS, color = CHROM %in% c("I", "III", "V", "VII", "IX", "XI", "XIII", "XV")), alpha = 0.2, linetype = "dashed")+
  geom_line(alpha = 0.4, size = 1) +
  labs(color = NULL)+
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none") + #ylim(0,12) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("Bulk Effect Peaks")

```

## Overlaps with specific genes in all peaks

```{r, fig.width=10, fig.height=4}
AllCu_plot %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CSS)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk", "Interaction"),],
             aes(xintercept = POS, color = CSS), alpha = 0.2)+
  geom_line(alpha = 0.4, size = 1) +
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, linetype = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3)+ 
  labs(linetype = NULL) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_linetype_manual(values = c("dotted", "dotted","dotted", "dotted","dotted", "dotted","dotted", "dotted","dotted", 
                                   "dotted","dotted", "dotted","dotted", "dotted","dotted", "dotted","dotted", "dotted",
                                   "dotted", "dotted","dotted", "dotted","dotted", "dotted","dotted", "dotted","dotted", 
                                   "dotted","dotted", "dotted","dotted", "dotted"))

################################################################################
#Separate by chromosomes
################################################################################

chromlist <- c(c("VII", "X", "XII", "XIV", "XV"))
AllCu_plot %>%
  filter(CHROM %in% chromlist) %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CSS)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk", "Interaction") & TruePeaks$CHROM %in% chromlist,],
             aes(xintercept = POS, color = CSS), alpha = 0.2)+
  geom_line(alpha = 0.4, size = 1) +
  geom_hline(aes(yintercept = q5)) +
  geom_vline(data = sgd_copper[sgd_copper$CHROM %in% chromlist,], 
             aes(xintercept = Gene.chromosomeLocation.start, linetype = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3)+ 
  labs(linetype = NULL) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_linetype_manual(values = c("dotted", "dotted","solid", "dotted","dotted", "dotted","dotted", "dotted","dotted", 
                                   "dotted","dotted", "dotted","dotted", "dotted","dotted", "dotted","dotted", "dotted",
                                   "dotted", "dotted","dotted", "dotted","dotted", "dotted","dotted", "dotted","dotted", 
                                   "dotted","dotted", "dotted","dotted", "dotted")) +
  ggtitle("Chr Containing Interactions")

################################################################################
chromlist <- c(c("IV", "VII", "IX", "XI", "XII", "XIV", "XVI"))
AllCu_plot %>%
  filter(CHROM %in% chromlist) %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CSS)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk") & TruePeaks$CHROM %in% chromlist,],
             aes(xintercept = POS, color = CSS), alpha = 0.05, size = 2)+
  geom_line(alpha = 0.4, size = 1) +
  geom_hline(aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("CSS Bulk Agreements")

chromlist <- c(c("IV", "VII", "IX", "XI", "XII", "XIV", "XVI"))
AllCu_plot %>%
  filter(CHROM %in% chromlist) %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CHROM)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk") & TruePeaks$CHROM %in% chromlist,],
             aes(xintercept = POS, color = CHROM), alpha = 0.05, size = 2)+
  geom_line(alpha = 0.4, size = 1) +
  geom_hline(aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("CSS Bulk Agreements")

chromlist <- c(c("IV", "VII", "IX", "XI", "XII", "XIV", "XVI"))
AllCu_plot %>%
  filter(CHROM %in% chromlist) %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CHROM)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk", "Interaction") & TruePeaks$CHROM %in% chromlist,],
             aes(xintercept = POS, color = CHROM), alpha = 0.05, size = 2)+
  geom_line(alpha = 0.4, size = 1) +
  geom_hline(aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("CSS Bulk Agreements")




```


# How close are the peaks?

```{r, fig.width=10, fig.height=4}
TruePeaks %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  select(label, CHROM, CSS, POS) %>%
  pivot_wider(names_from = CSS, values_from = POS) -> TotalList

TotalList %>% select(-VIII) %>% unnest() -> chr1
TotalList %>% select(-I) %>% unnest() -> chr8

merge(chr1, chr8) %>% mutate(diff = abs(I - VIII)) %>% arrange(diff) %>%
  filter(diff < 8000) %>%
  ggplot() + geom_point(aes(y = diff, x = CHROM)) +
  theme(axis.text.x=element_text())

merge(chr1, chr8) %>% mutate(diff = abs(I - VIII)) %>% arrange(diff) %>%
  filter(diff < 8000) %>%
  group_by(label, CHROM, diff) %>%
  summarize(maxpoint = max(I, VIII),
            minpoint = min(I, VIII)) %>%
  ggplot(aes(color = CHROM)) + geom_linerange(aes(xmax = maxpoint, xmin = minpoint, y = diff)) +
  facet_grid(rows = "CHROM") +
  theme(axis.text.x=element_text())

merge(chr1, chr8) %>% mutate(diff = abs(I - VIII)) %>% arrange(diff) %>%
  filter(diff < 1500) %>%
  group_by(label, CHROM, diff) %>%
  summarize(maxpoint = max(I, VIII),
            minpoint = min(I, VIII)) %>%
  ggplot(aes(color = CHROM)) + geom_point(aes(x = CHROM, y = diff), alpha = 0.4, size = 2) +
  theme(axis.text.x=element_text())

merge(chr1, chr8) %>% mutate(diff = abs(I - VIII)) %>% arrange(diff) %>%
  filter(diff < 1500) %>%
  pivot_longer(c(I, VIII)) ->listofdiffs

chromlist <- c(c("IX", "XII", "XIV", "XVI"))
AllCu_plot %>%
  filter(CHROM %in% chromlist) %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CSS)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk") & TruePeaks$CHROM %in% chromlist & TruePeaks$POS %in% listofdiffs$value,],
             aes(xintercept = POS, color = CSS), alpha = 0.2, size = 2)+
  geom_line(alpha = 0.4, size = 1) +
  geom_hline(aes(yintercept = q5)) +
  labs(linetype = NULL) +
  #facet_grid(label~CHROM, scales = "free", space = "free") +
  facet_wrap(facets = "CHROM", scales = "free") +
  theme(legend.position = "bottom") +
  ggtitle("CSS I and VIII agreements")

chromlist <- c(c("IV", "VII", "IX", "XI", "XII", "XIV", "XVI"))
AllCu_plot %>%
  filter(CHROM %in% chromlist) %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CHROM, linetype = CSS)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk") & TruePeaks$CHROM %in% chromlist & TruePeaks$POS %in% listofdiffs$value,],
             aes(xintercept = POS, color = CSS), alpha = 0.2, size = 2, color = "black")+
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk", "Interaction") & TruePeaks$CHROM %in% chromlist,],
             aes(xintercept = POS, color = CHROM), alpha = 0.05, size = 2)+
  geom_line(alpha = 0.4, size = 1) +
    
  geom_hline(aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("CSS Bulk Agreements within 1.5 kb")

AllCu_plot %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CHROM, linetype = CSS)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk") & TruePeaks$CHROM %in% chromlist & TruePeaks$POS %in% listofdiffs$value,],
             aes(xintercept = POS, color = CHROM), size = 1)+

  geom_line(alpha = 0.4, size = 1) +
    
  geom_hline(aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("CSS Bulk Agreements within 1.5 kb")

################################################################################
#Change the distance

merge(chr1, chr8) %>% mutate(diff = abs(I - VIII)) %>% arrange(diff) %>%
  filter(diff < 8000) %>%
  pivot_longer(c(I, VIII)) ->listofdiffs5k

AllCu_plot %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CHROM, linetype = CSS)) +
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk") & TruePeaks$POS %in% listofdiffs5k$value,],
             aes(xintercept = POS, color = CHROM), size = 1)+

  geom_line(alpha = 0.4, size = 1) +
    
  geom_hline(aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("CSS Bulk Agreements within 8 kb")
```


# The Classics

```{r}
AllCu_plot %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = zscore, color = label)) + 
  geom_hline(aes(yintercept = c(q5)), linetype = "dashed") +
  geom_hline(aes(yintercept = c(-q5)), linetype = "dashed") +
  geom_line() +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  ggtitle("Bulk and Interaction Effects") +
  scale_color_manual(values = c("black", "purple"))

AllCu_plot %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = zscore, color = label)) + 
  geom_line() +
  geom_rect(aes(ymin = -q5, ymax = q5, xmin = -Inf, xmax = Inf), color = "gray", alpha = 0.01) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  ggtitle("Bulk and Interaction Effects") +
  scale_color_manual(values = c("black", "purple"))

AllCu_plot %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = label)) + 
  geom_hline(aes(yintercept = c(q5)), linetype = "dashed") +
  geom_line() +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom") +
  ggtitle("Bulk and Interaction Effects") +
  scale_color_manual(values = c("black", "purple"))

```

# Circos plot 

Since I haven't added these in yet :(

Function first
```{r}
require(circlize)

cybr_circos <- function(d1, d8, peaklist1 = NULL, peaklist8 = NULL, maxy = NULL, color1 = "#24588A", color8 = "#ED7B01", opacity = "50"){
  
  color1fade_50 <- paste(color1, "50", sep = "")
  color8fade_50 <- paste(color8, "50", sep = "")
  
  color1fade <- paste(color1, opacity, sep = "")
  color8fade <- paste(color8, opacity, sep = "")

  if(is.null(d8)){
    d8 <- d1
    include8 <- FALSE
  }else{
    include8 <- TRUE
  }
    #SET UP DATA
  df8 <- data.frame(sectors = as.character(d8$CHROM),
                 x = d8$POS,
                 y = abs(d8$summary),
                  label = d8$label)

  df1 <- data.frame(sectors = as.character(d1$CHROM),
                 x = d1$POS,
                 y = abs(d1$summary),
                  label = d1$label)

  #Take just the interactions of each
  df8int <- df8 %>% filter(label == "Interaction")
  df1int <- df1 %>% filter(label == "Interaction")

    #REORDER THE CHROMOSOMES
  df8int$sectors <- factor(df8int$sectors, levels = as.character(as.roman(1:16)))
  df1int$sectors <- factor(df1int$sectors, levels = as.character(as.roman(1:16)))
  
  ##############################################################################
  dfall <- rbind(df1int, df8int) %>% na.omit()
  circos.par("track.height" = 0.3, start.degree = 90, cell.padding = c(0,0))
  circos.initialize(sectors = dfall$sectors, x = dfall$x)

  if(is.null(maxy)){
    #I think this makes the sizes?
    circos.track(ylim = c(max(c(dfall$y)), 0), dfall$sectors, y = dfall$y, 
        panel.fun = function(x, y) {
            circos.text(CELL_META$xcenter, 
                0 - mm_y(5), 
                CELL_META$sector.index,
                niceFacing = FALSE)
            circos.axis(labels.cex = 0.1)
    })
  }else{
    circos.track(ylim = c(maxy, 0), dfall$sectors, y = dfall$y, 
      panel.fun = function(x, y) {
          circos.text(CELL_META$xcenter, 
              0 - mm_y(5), 
              CELL_META$sector.index,
              niceFacing = FALSE)
          circos.axis(labels.cex = 0.1)
      })
  }
  #Makes the chromosome overlap parts
  #CHROMOSOME I
  draw.sector(83.5, #RIGHT
              90, #LEFT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color1fade_50, border = FALSE)
  if(include8 == TRUE){
    #CHROMOSOME VIII
    draw.sector(289.5, #LEFT
              305.4, #RIGHT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color8fade_50, border = FALSE)
  }
  
  
  #Makes the lines
  circos.trackPoints(df8int$sectors, df8int$x, abs(df8int$y), col = color8, pch = 16, cex = 0.1)
  circos.trackPoints(df1int$sectors, df1int$x, abs(df1int$y), col = color1, pch = 16, cex = 0.1)
  
  if(is.null(peaklist8) == FALSE){
    if(length(peaklist8$POS) >= 1){
      for(i in 1:length(peaklist8$POS)){
      circos.link(peaklist8$CHROM[i], 
                  peaklist8$POS[i], 
                  #Add 8 after
                  "VIII", c(0, max(dfall$x[dfall$sectors == "VIII"])),  
                  
                  col = color8fade, 
                  h.ratio = 0.3, 
                  border = color8fade, 
                  lwd = 1)
      }
    }else{
      print("No interactions on Chr 8")
    }
    
  }

  if(is.null(peaklist1) == FALSE){
    if(length(peaklist1$POS) >= 1){
      for(i in 1:length(peaklist1$POS)){
        circos.link("I",c(0, max(dfall$x[dfall$sectors == "I"])),  
                    #add 1 first
                    peaklist1$CHROM[i], 
                    peaklist1$POS[i], 
                    col = color1fade, 
                    h.ratio = 0.3, 
                    border = color1fade, 
                    lwd = 1)
      }
    }else{
      print("No interactions on Chr 1")
    }
  }  
  
  circos.clear()
}

```

Run the actual function to produce plots
```{r}
ChromosomeScale3 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(SmoothCount = 0, label = "Interaction", q5 = 0, zscore = 0, CSS = "VIII") %>% select(-delete) %>%
  select(CSS, label, CHROM, POS, SmoothCount, q5, zscore) 

C_1_intpeaks <- AllCu_plot %>% filter(label == "Interaction", CSS == "VIII") %>%
  filter(abs(SmoothCount) > q5) %>%
  group_by(CHROM) %>% arrange(SmoothCount) %>%
  summarize(POS = head(POS, 1))

C_8_intpeaks <- AllCu_plot %>% filter(label == "Interaction", CSS == "VIII") %>%
  filter(abs(SmoothCount) > q5) %>%
  group_by(CHROM) %>% arrange(SmoothCount) %>%
  summarize(POS = head(POS, 1))
  
d1new <- rbind(AllCu_plot, ChromosomeScale3) %>% mutate(summary = SmoothCount) %>% filter(CSS == "I")
d8new <- rbind(AllCu_plot, ChromosomeScale3) %>% mutate(summary = SmoothCount) %>% filter(CSS == "VIII")

cybr_circos(d1 = d1new, 
            d8 = d8new, 
            peaklist8 = C_8_intpeaks)
```

