---
title: "MergedVCFs_CuSO4_May24"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))



## Functions for GLM all and Permutation all

glm_cb2_all <- function(..., W, formula, numgroups = FALSE, outputlength = 8) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  output <- summary(glm_fit)$coefficients[c(((length(summary(glm_fit)$coefficients)*0)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.25)),
                                            ((length(summary(glm_fit)$coefficients)*0.5)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.75)))]

  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
} 


################################################################################
# No replicates, fix Parent
cybrPermute_cb2_all <- function(dataset, 
                                R = 0, perp = 1, outlength = 4, 
                                glmform = "Allele ~ Bulk * Parent", 
                                inputlabels = c("Intercept", "Bulk", "Parent", "Interaction")){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled aD")
  print(paste("Your labels are:", inputlabels))

    if(R > 0){ #INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Rep, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == 1) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest    
    }else{ #DO NOT INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == perp) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    }
  
    #PERMUTE TWICE
  if(R > 0){
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Rep, 
             Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep, 
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, 
             Rep, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep,
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }else{
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #RUN THE GLM
    if(R > 0) {
      shuffletoglm2 %>% na.omit() %>%
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }else{
      shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }
    
    
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}

cybrInputGATKTable2 <- function(rawData, yeast = TRUE){

  require(dplyr)
  require(doParallel)
  require(foreach)

  HNGLCDRXY <- read.table(rawData, header = TRUE)

  #Identify the unique values besides AD/DP/GQ/PL
  gsub(".AD", "",
       gsub(".GQ", "",
            gsub(".DP","",
                 gsub(".PL","",
                      colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
  #i <- Samples[1]

  resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
    mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
    AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
    GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
    DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
    PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
    cbind(mydf, AD , GQ , DP, PL) -> mydftotal
    colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")

    mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>%
      separate(PL, c('PL.REF','PL.ALT'), extra='drop') %>%
      #Added 10/18/23:
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf

    mycdf %>% filter(grepl(",", ALT)) %>% 
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
    
    doublecdf %>% filter(grepl(",", ALT)) %>%
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
    
    rbind(mycdf, doublecdf, triplecdf) -> newcdf
    
    newcdf
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
  }else{
    results <- resultscdf
  }
  return(results)

}

theme_classic
```

# Pre-processing

## Load in table and match to oak or wine parents

```{r}
setwd("C:/Users/Cassandra/Documents/GitHub/Sequencing/Analysis/")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)
parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) -> pSNPs

```

```{r, message = FALSE, warning=FALSE}
#testfile <- "C:/Users/cassa/Downloads/AllCuSO4.REF_NC_001143.9.SortedCat.vcf.output.table"


myfile <- "C:/Users/Cassandra/Documents/Data_NYU/Sequencing/AllCuSO4.REF_.SortedCat.vcf.output.table"

# unique(read.table("C:/Users/Cassandra/Documents/Data_NYU/Sequencing/AllCuSO4.REF_NC_001133.9.bam.vcf.SortedCat.vcf.output.table", header = TRUE)$CHROM)
# unique(read.table("C:/Users/Cassandra/Documents/Data_NYU/Sequencing/AllCuSO4.REF_NC_001134.8.bam.vcf.SortedCat.vcf.output.table", header = TRUE)$CHROM)
# unique(read.table("C:/Users/Cassandra/Documents/Data_NYU/Sequencing/AllCuSO4.REF_NC_001135.5.bam.vcf.SortedCat.vcf.output.table", header = TRUE)$CHROM)

read.table(myfile, header = TRUE) -> testrawdata

unique(testrawdata$CHROM)

cybrInputGATKTable2(myfile) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> rawdata

saveRDS(rawdata, file = "rawdata.rds")

```

## Actually determine which read is which

```{r}

#SMOOTHING ETC
rawdata %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

# rawdata_called %>% filter(CHROM == "II") %>%
#   ggplot(aes(x = POS, y = Dataset, color = Reads)) + geom_tile()

################################################################################
rawdata_called %>% group_by(Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 300, FUN = median, align = "center", na.rm = TRUE))) -> rawdata_smoothed #%>% na.omit()

# rawdata_smoothed %>% filter(CHROM == "II") %>%
#   ggplot(aes(x = POS, y = Dataset, color = SmoothCount)) + geom_tile()
```

## Separate into bulks, make ready for glms

```{r}

rawdata_smoothed %>% mutate(Dataset = gsub("n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Pool", "Dataset"), sep = "_", extra = "merge") -> rd

# unique(rd$Pool)
# unique(rd$Dataset)

#write.csv(data.frame(Dataset = unique(rd$Dataset)), file = "experiment_names.csv")
expnames <- read.csv("experiment_names_524.csv")

rd %>% filter(CHROM == "II") %>%
  ggplot(aes(x = POS, y = Dataset, color = SmoothCount)) + geom_tile()

rd %>% left_join(expnames) -> rd_named
```


```{r, eval = FALSE}
#Checking to see how many of these are missing
rd_named %>% filter(CHROM == "II") %>%
  ggplot(aes(x = POS, y = Dataset, color = SmoothCount)) + geom_tile()

```

## Change into factors

```{r,eval = FALSE}
rd_named %>% filter(Bulk != "Fluconazole") %>%
  select(-Selection, -Replicate) %>% na.omit() %>%
  mutate_if(is.character, as.factor) -> rd_factor

rd_factor %>% filter(Dataset %in% c("O1_aD", "W1_aD") == FALSE) -> rd_factor

contrasts(rd_factor$Background) <- matrix(c(0.5, -0.5))
contrasts(rd_factor$Bulk) <- matrix(c(1, 0))

saveRDS(rd_factor, "rd_factor.rds")
```

################################################################################

# Running GLMs

```{r}
rd_factor <- readRDS("rd_factor.rds")

```

```{r, eval = FALSE}
#THIS IS USELESS BECAUSE IT DOESN'T SEPARATE THEM
rd_factor %>% 
  group_by(CHROM, POS) %>% summarize(zscore = glm_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_all_glm

unique(CuSO4_all_glm$CHROM)

saveRDS(CuSO4_all_glm, file = "CuSO4_all_glm_May24.rds")
```

```{r, eval = FALSE}
CuSO4_all_glm %>% filter(CHROM %in% c("I", "VIII") == FALSE) %>% ggplot(aes(x = POS, y = abs(zscore), color = label)) + geom_line() + facet_grid(label~CHROM, space = "free", scales = "free")

CuSO4_all_glm %>% filter(CHROM == "VIII") %>% ggplot(aes(x = POS, y = abs(zscore), color = label)) + geom_line() + facet_grid(label~CHROM, space = "free", scales = "free") + ylim(c(0, 100))
```

## Make function to run glmer

```{r}
glmer_cb2_short <- function (..., W, formula, numgroups = FALSE, outputlength = 4, 
    return = c("Z")) 
{
    data <- list(...)
    
    require(lme4)
    if (is.null(W) || is.null(formula)) {
        stop("Weights (W) and formula must be provided")
    }
    glm_formula <- as.formula(formula)
    if (!all(names(data) %in% all.vars(glm_formula))) {
        stop("One or more variables in the formula are not provided as arguments")
    }
    for (i in all.vars(glm_formula)) {
        if (length(unique(as.data.frame(data)[, i])) < 2) {
            output <- rep(NA, outputlength)
            return(output)
        }
    }
    glm_fit <- glmer(glm_formula, data = as.data.frame(data), weights = W, 
        family = binomial)
    if (return %in% "Z") {
        output <- summary(glm_fit)$coefficients[((length(summary(glm_fit)$coefficients) * 
            0.5) + 1):((length(summary(glm_fit)$coefficients) * 
            0.75))]
    }
    if (length(output) == outputlength) {
        return(output)
    }
    else {
        return(rep(NA, outputlength))
    }
}
```

## Separate Chromosomes

```{r, eval = FALSE}
#Test the new one

rd_factor %>% filter(CHROM == "III") %>% filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2")) %>%
  group_by(CHROM, POS) %>% summarize(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Pool = Pool, Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background + (1 | Pool) + (Pool | Rep)"),
                                                                  label = c("cept","Bulk", "Background", "Interaction")) 
```


```{r, eval = FALSE}
rd_factor %>% filter(CHROM != "M") %>% distinct() %>% filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2")) %>%
  group_by(CHROM, POS) %>% summarize(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background + (1 | Rep) "),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_CSS1_glmer_1byRep

saveRDS(CuSO4_CSS1_glmer_1byRep, file = "CuSO4_CSS1_glmer_1byRep.rds")

################################################################################
rd_factor %>% filter(CHROM != "M", CHROM != "VIII")%>% distinct() %>% filter(Pool %in% c("HVYTYDRX2","HTG3TDMXY")) %>%
  group_by(CHROM, POS) %>% summarize(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background + + (1 | Rep)"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_CSS8_glmer_1byRep

saveRDS(CuSO4_CSS8_glmer_1byRep, file = "CuSO4_CSS8_glmer_1byRep.rds ")
```


```{r}
CuSO4_CSS1_glmer_1byRep <- readRDS("CuSO4_CSS1_glmer_1byRep.rds")
CuSO4_CSS8_glmer_1byRep <- readRDS("CuSO4_CSS8_glmer_1byRep.rds")
```

Merging all datasets instead

```{r}
CuSO4_CSS1_glmer_1byRep %>% filter(CHROM != "I") %>% mutate(CSS = "I") -> temp1
CuSO4_CSS8_glmer_1byRep %>% mutate(CSS = "VIII") %>%
  rbind(temp1) -> AllCuSO4_glmer_1byRep

rm(temp1)

AllCuSO4_glmer_1byRep %>% filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = zscore, color = CSS)) + 
  geom_line() +
  geom_hline(yintercept = 0)+
  facet_grid(label~CHROM, scales = "free", space = "free")
```

What genes are nearby?

```{r}
sgd_orfs <- read.csv("C:/Users/Cassandra/Documents/R_FileTransfer/SGD_ORFs.csv")

genes <- c("CRS5", "SOD1", "CUP1-1", "CUP1-2", "MET17", "MUP1", "PCL1", "OYE3", "FRE1", "FRE7", "CTR1", "IRC7")
sgd_orfs %>% filter(Gene.symbol %in% genes) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_copper

genes_eh <- c("VMA4", "MTL1", "PTR2", "LEU9", "IRC15")
sgd_orfs %>% filter(Gene.symbol %in% genes_eh) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_copper_eh
```

```{r}
AllCuSO4_glmer_1byRep %>% filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = zscore, linetype = CSS)) + 
  geom_line() +
  geom_hline(yintercept = 0)+
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
  geom_vline(data = sgd_copper_eh,
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             linetype = "dashed", size = 1, alpha = 0.3) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL) +
  theme(legend.position = "bottom")
```

## Permutations for significance

```{r, eval = FALSE}
# Shuffle the entire genome WITHIN replicates, of DILUTE bulks

rd_factor %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Background) %>%
  separate(NewPOS, into = c("CHROM", "POS", sep = "_"))-> rd_shuffled

rd_shuffled %>% ggplot(aes(x = POS, y = log(Oak/Wine), color = Dataset)) + geom_point(alpha = 0.1) + facet_grid(Pool ~ CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom")

rd_shuffled %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") 
```

Actually running a permutation on this

```{r, eval = FALSE}
rd_factor %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Background) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

rd_factor %>% filter(Bulk == "Dilute",
                     CHROM %in% c("I", "VIII", "M", "III", "V") == FALSE) %>% 
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Pool, Dataset) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Background = Background) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Dilute") %>%
  rbind(rd_shuffled_selected) -> Perm1

Perm1 %>% ungroup() %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") %>% unnest() %>%
  mutate(POS = as.numeric(POS)) %>%
  mutate_if(is.character, as.factor) -> Perm1contrasts

contrasts(Perm1contrasts$Background) <- matrix(c(-0.5, 0.5))
contrasts(Perm1contrasts$Bulk)

```

```{r, eval = FALSE}

Perm1contrasts %>%  filter(Pool %in% c("HNGLVDRXY", "HKTFTDRX2")) %>% distinct() %>% 
  
  group_by(CHROM, POS) %>% reframe(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background +  (1 | Rep)"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_CSS1_glmer_1byRep_PERM

saveRDS(CuSO4_CSS1_glmer_1byRep_PERM, "CuSO4_CSS1_glmer_1byRep_PERM.rds")

Perm1contrasts %>%
  filter(Pool %in% c("HVYTYDRX2","HTG3TDMXY")) %>% distinct() %>% 
  mutate(POS = as.numeric(POS)) %>%
  mutate_if(is.character, as.factor) %>%
  group_by(CHROM, POS) %>% reframe(zscore = glmer_cb2_short(Bulk = Bulk,  
                                                            Background = Background,
                                                            Rep = Dataset,
                                                            Allele = Allele,
                                                            W = SmoothCount, outputlength = 4,
                                                            formula = "Allele ~ Bulk*Background + (1 | Rep)"),
                                                                  label = c("cept","Bulk", "Background", "Interaction"))  -> CuSO4_CSS8_glmer_1byRep_PERM

saveRDS(CuSO4_CSS8_glmer_1byRep_PERM, "CuSO4_CSS8_glmer_1byRep_PERM.rds")
```

Cutoffs for this

```{r}

CuSO4_CSS8_glmer_1byRep_PERM <- readRDS("CuSO4_CSS8_glmer_1byRep_PERM.rds")
CuSO4_CSS1_glmer_1byRep_PERM <- readRDS("CuSO4_CSS1_glmer_1byRep_PERM.rds")

CuSO4_CSS8_glmer_1byRep_PERM %>%
  group_by(label) %>%
  summarize(q5 = quantile(abs(zscore), 0.95),
            q1 = quantile(abs(zscore), 0.99)) %>%
  mutate(CSS = "VIII") -> q5_CSS8

CuSO4_CSS1_glmer_1byRep_PERM %>%
  group_by(label) %>%
  summarize(q5 = quantile(abs(zscore), 0.95),
            q1 = quantile(abs(zscore), 0.99)) %>%
  mutate(CSS = "I")  -> q5_CSS1

rbind(q5_CSS1, q5_CSS8) -> q5_all

q5_all

```

Okay, so what's significant?

```{r}
AllCuSO4_glmer_1byRep %>% filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = zscore, linetype = CSS)) + 
  geom_line() +
  geom_hline(yintercept = 0, color = "gray")+
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
  geom_vline(data = sgd_copper_eh,
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             linetype = "dashed", size = 1, alpha = 0.3) +
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = -q5)) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL) +
  theme(legend.position = "bottom")
```


```{r}
AllCuSO4_glmer_1byRep %>% filter(label %in% c("Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) + 
  geom_point(aes(alpha = abs(zscore) > 1.01), color = "gray", size = 0.1) +
  geom_point(aes(alpha = abs(zscore) < 1.01), color = "black", size = 0.1) +
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
  geom_vline(data = sgd_copper_eh,
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             linetype = "dashed", size = 1, alpha = 0.3) +
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") + ggtitle("Interaction points") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))

AllCuSO4_glmer_1byRep %>% filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) + 
  geom_point(aes(alpha = abs(zscore) > 1.01), color = "gray", size = 0.1) +
  geom_point(aes(alpha = abs(zscore) < 1.01), color = "black", size = 0.1) +
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
  geom_vline(data = sgd_copper_eh,
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             linetype = "dashed", size = 1, alpha = 0.3) +
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") + ggtitle("Bulk points") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))


AllCuSO4_glmer_1byRep %>% filter(label %in% c("Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) + 
  geom_point(aes(color = abs(zscore) > 1.01, size = abs(zscore) > 1.01), alpha = 0.4) +
  geom_hline(yintercept = 0, color = "gray")+
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  labs(color = "FDR 5%", size = "FDR 5%") +
  theme(legend.position = "bottom") + ggtitle("Significant Interactions") +
  scale_color_manual(values = c("gray", "purple")) +
  scale_size_manual(values = c(0.5, 2))
```

What's going on with Chr I and VIII? Do they actually interact?

```{r}
AllCuSO4_glmer_1byRep %>% filter(label %in% c("Interaction"),
                                  CHROM %in% c("I", "VIII")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) + 
  geom_point(aes(color = abs(zscore) > 1.01, size = abs(zscore) > 1.01), alpha = 0.4) +
  geom_hline(yintercept = 0, color = "gray")+
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5), linetype = "dashed") +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  labs(color = "FDR 5%", size = "FDR 5%") +
  theme(legend.position = "bottom") + ggtitle("Fixed Chromosome Interactions") +
  scale_color_manual(values = c("gray", "purple")) +
  scale_size_manual(values = c(0.5, 2)) +
  geom_vline(aes(xintercept = c(212535))) +
  geom_vline(aes(xintercept = c(214718)))


```

# Checking original log ratios

```{r}
rd_factor <- readRDS("rd_factor.rds")

rd_factor %>% pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>% 
  mutate(logRatio = log(Wine/Oak)) -> rd_logratios

rd_logratios %>% filter(CHROM == "I") %>% ggplot(aes(x = POS, y = logRatio, color = Bulk)) + geom_point(alpha = 0.2) +
  facet_grid(Background~Pool)

rd_logratios %>% filter(CHROM == "I", Pool == "HVYTYDRX2") %>% group_by(Dataset) %>% summarize(unique(Dataset))

rd_factor %>% filter(Dataset %in% c("O1_aD", "W1_aD") == FALSE) %>% pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>% 
  mutate(logRatio = log(Wine/Oak)) -> rd_logratios

rd_logratios %>% filter(CHROM == "I") %>% ggplot(aes(x = POS, y = logRatio, color = Bulk)) + geom_point(alpha = 0.2) +
  facet_grid(Background~Pool)

```

# Checks from Mark Meeting

1. What do the residuals look like for different glms?
2. Do the outliers have more NAs? 
3. How exactly is the median happening?

```{r}
AllCuSO4_glmer_1byRep %>% group_by(CHROM, label, CSS) %>% arrange(POS) %>%
  reframe(POS = POS, 
          zscore = zscore,
          SmoothZ = (frollapply(zscore, n = 100, FUN = median, align = "center", na.rm = TRUE))) -> zscores_smoothed

zscores_smoothed %>% mutate(diff = abs(SmoothZ - zscore)) %>%
  ggplot(aes(x = POS, y = diff, color = label)) + geom_point(size = 0.2, alpha = 0.3) + facet_grid(CSS~CHROM, scales = "free", space = "free")

#Make a cutoff for how different the smoothed size has to temp2
zscores_smoothed %>% mutate(diff = abs(SmoothZ - zscore)) %>%
  mutate(outlier = diff > 2) -> outliers

#Plot if this gets the outliers
outliers %>% ggplot(aes(x = POS, y = abs(zscore), color = label, size = outlier, shape = outlier)) + 
  geom_point(alpha = 0.3)  + facet_grid(CSS~CHROM, scales = "free", space = "free") +
  scale_size_manual(values = c(0.2, 2, 0.2))

rd_logratios %>% select(Dataset, CHROM, POS, logRatio) %>% pivot_wider(names_from = Dataset, values_from = logRatio) %>% unnest() -> temp2
outliers %>% select(CHROM, POS, outlier, label) %>% pivot_wider(names_from = label, values_from = outlier) %>% unnest() %>% left_join(temp2) -> temp3

temp3 %>% 
  distinct() %>% count(POS) %>%
  merge(temp3) %>%
  filter(n > 1) %>% select(CHROM, POS) -> excl_duplicates


temp3 %>% filter(Background == TRUE)%>% select(-Bulk, -Interaction, -cept) %>% pivot_longer(-c(CHROM, POS)) %>%
  ggplot(aes(x = as.factor(POS), y = value, color = as.factor(POS))) + geom_boxplot(size = 2) + geom_jitter()

temp3 %>% filter(Bulk == TRUE)%>% select(-Background, -Interaction, -cept) %>% pivot_longer(-c(CHROM, POS)) %>%
  ggplot(aes(x = as.factor(POS), y = value, color = as.factor(POS))) + geom_boxplot() + 
  theme(legend.position = "none") + facet_grid(~CHROM,scales = "free", space = "free")

temp3 %>% filter(Interaction == TRUE)%>% select(-Background, -Bulk, -cept) %>% pivot_longer(-c(CHROM, POS)) %>%
  ggplot(aes(x = as.factor(POS), y = value, color = as.factor(POS))) + geom_boxplot() + 
  theme(legend.position = "none") + facet_grid(~CHROM,scales = "free", space = "free")

temp3 %>% filter(cept == TRUE)%>% select(-Background, -Bulk, -Interaction) %>% pivot_longer(-c(CHROM, POS)) %>%
  ggplot(aes(x = as.factor(POS), y = value, color = as.factor(POS))) + geom_boxplot() + 
  theme(legend.position = "none") + facet_grid(~CHROM,scales = "free", space = "free")

temp3 %>% filter(Background == TRUE | Bulk == TRUE | Interaction == TRUE | cept == TRUE) %>% pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>%
  ggplot(aes(y = name, color = value, x = factor(POS))) + geom_tile()

# temp3 %>% pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>%
#   pivot_longer(c(Background, Bulk, Interaction, cept), names_to = "label", values_to = "Outlier") %>%
#   ggplot(aes(x = Outlier, y = value, color = Outlier)) + geom_boxplot()

# temp3 %>% pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>%
#   pivot_longer(c(Background, Bulk, Interaction, cept), names_to = "label", values_to = "Outlier") %>%
#   filter(label == "Bulk") %>%
#   group_by(POS) %>% 
#   summarize(mean = mean(value, na.rm = TRUE),
#             sd = sd(value, na.rm = TRUE),
#             Outlier = Outlier) %>%
#   distinct() -> summaryofoutliers
# 
# summaryofoutliers %>% ggplot(aes(color = Outlier, x = sd)) + geom_density() +
#                      theme(axis.text.x=element_text())
# summaryofoutliers %>% ggplot(aes(color = Outlier, x = mean)) + geom_density()+
#                      theme(axis.text.x=element_text())

```

```{r, fig.width=10, fig.height=5}
temp3 %>% filter(CHROM == "VIII") %>%  pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>%
  ggplot(aes(x = POS, y = value, color = name)) + 
  geom_point(aes(alpha = 0.2, shape = Bulk, size = Bulk)) + geom_line(size = 0.3, alpha = 0.3) +
  xlim(c(1.6e5, 1.88e5))+ scale_size_manual(values = c(1, 2)) +
  theme(legend.position = "none") +
                  theme(axis.text.x=element_text())

temp3 %>% filter(CHROM == "VIII", POS > 1.6e5, POS < 1.88e5) %>%  pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>%
  ggplot(aes(x = factor(POS), y = name, fill = value)) + 
  geom_tile() + scale_color_manual(values = c("gray", "black"))
  
```

```{r}
temp3 %>% select(POS, Bulk) -> outlierinfo
temp3 %>% filter(CHROM == "VIII", POS > 1.6e5, POS < 1.88e5) %>% #distinct() %>%
  count(POS) %>% merge(outlierinfo) %>%
  ggplot(aes(x = Bulk, y = n, color = Bulk)) + geom_boxplot(outliers = FALSE) + geom_jitter(height = 0, alpha = 0.4) +
  ggtitle("All Counts")

temp3 %>% filter(CHROM == "VIII", POS > 1.6e5, POS < 1.88e5) %>% distinct() %>%
  count(POS) %>% merge(outlierinfo) %>%
  ggplot(aes(x = Bulk, y = n, color = Bulk)) + geom_boxplot(outliers = FALSE) + geom_jitter(height = 0, alpha = 0.4) +
  ggtitle("Distinct Only")

temp3 %>% filter(CHROM == "VIII") %>% 
  distinct() %>%
  count(POS) %>% merge(outlierinfo) %>%
  ggplot(aes(x = Bulk, y = n, color = Bulk)) + geom_boxplot(outliers = FALSE) + geom_jitter(height = 0, alpha = 0.4) +
  ggtitle("Bulk Outliers; Distinct Only")

temp3 %>% filter(CHROM == "VIII") %>% 
  distinct() %>% count(POS) %>%
  merge(temp3) %>%
  filter(n > 1) %>%
  filter(POS == 17019) %>%
  pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>% 
  filter(grepl("W8", name) == FALSE,
         grepl("O8", name) == FALSE) %>%
  ggplot(aes(y = name, x = value)) + geom_point(alpha = 0.3, size = 2)

# rawdata <- readRDS("rawdata.rds")
# 
# rawdata %>% filter(CHROM == "VIII", POS == 17019) %>% distinct() %>%
#   separate(Dataset, c("Pool", "null", "Dataset"), sep = "_", extra = "merge") %>% select(-null) %>%
#   mutate(Dataset = gsub(".fastq", "", Dataset))

temp3 %>% 
  distinct() %>% count(POS) %>%
  merge(temp3) %>%
  filter(n > 1) %>% select(CHROM, POS) -> excl_duplicates
```

```{r, fig.width=12, fig.height=8, eval = FALSE}

#Smoothed
rd_logratios %>% mutate(Coverage = Oak + Wine) %>% select(Dataset, CHROM, POS, Coverage) %>% pivot_wider(names_from = Dataset, values_from = Coverage) %>% unnest() -> temp4
outliers %>% select(CHROM, POS, outlier, label) %>% pivot_wider(names_from = label, values_from = outlier) %>% unnest() %>% left_join(temp4) -> temp_cov

#Not Smoothed
rawdata_called %>% mutate(Dataset = gsub("n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Pool", "Dataset"), sep = "_", extra = "merge") -> rd_notsmoothed

rd_notsmoothed %>% pivot_wider(names_from = Allele, values_from = Reads) %>% unnest() %>% 
  mutate(logRatio = log(Wine/Oak)) %>%
  mutate(Coverage = Oak + Wine) %>% select(Dataset, CHROM, POS, Coverage) %>% pivot_wider(names_from = Dataset, values_from = Coverage) %>% unnest() -> temp_unsm
outliers %>% select(CHROM, POS, outlier, label) %>% pivot_wider(names_from = label, values_from = outlier) %>% unnest() %>% left_join(temp_unsm) -> temp_unsmoothed

################################################################################
#Plots
temp_unsmoothed %>% filter(CHROM == "VIII", POS > 1.6e5, POS < 1.88e5) %>%  pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>%
  ggplot(aes(x = factor(POS), y = name, fill = value)) + 
  geom_tile() 

temp_cov %>% filter(CHROM == "VIII", POS > 1.6e5, POS < 1.88e5) %>%  pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>%
  ggplot(aes(x = factor(POS), y = name, fill = value)) + 
  geom_tile() 

temp_unsmoothed %>% filter(CHROM == "VIII", POS > 1.6e5, POS < 1.88e5) %>%  pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>%
  mutate(Smoothing = "Unsmoothed") -> anothertemp

temp_cov %>% filter(CHROM == "VIII", POS > 1.6e5, POS < 1.88e5) %>%  pivot_longer(-c(CHROM, POS, Background, Bulk, Interaction, cept)) %>%
  mutate(Smoothing = "Smoothed") %>%
  rbind(anothertemp) %>%
  ggplot(aes(x = factor(POS), y = name, fill = value)) + 
  geom_tile() + facet_grid(~Smoothing)
```

```{r, fig.height=8, fig.width=8, eval = FALSE}
testrawdata %>% filter(CHROM == "NC_001140.6", POS == 17019) %>%
  mutate_if(is.numeric, as.character) %>%
  pivot_longer(-c(CHROM, POS, REF, ALT)) %>%
  filter(grepl("AD", name)) %>%
  separate(value, c("REF_G","ALT_A", "ALT_C"), sep = ",") %>%
  pivot_longer(c(REF_G, ALT_A, ALT_C), values_to = "Reads", names_to = "Allele") %>%
  mutate(Reads = as.numeric(Reads)) %>%
  ggplot(aes(y = name, x = Reads, color = Allele)) + geom_point(alpha = 0.4)


testrawdata %>% filter(CHROM == "NC_001140.6", POS == 17019) %>%
  mutate_if(is.numeric, as.character) %>%
  pivot_longer(-c(CHROM, POS, REF, ALT)) %>%
  filter(grepl("AD", name)) %>%
  separate(value, c("REF_G","ALT_A", "ALT_C"), sep = ",") %>%
  filter(REF >0, ALT_A > 0, ALT_C > 0)
```

# Final Plotting

Ensure filters happen

```{r, warning = FALSE}
AllCuSO4_glmer_1byRep %>% group_by(CHROM, label, CSS) %>% arrange(POS) %>%
  reframe(POS = POS, 
          zscore = zscore,
          SmoothZ = (frollapply(zscore, n = 100, FUN = median, align = "center", na.rm = TRUE))) -> zscores_smoothed

#Make a cutoff for how different the smoothed size has to temp2
zscores_smoothed %>% mutate(diff = abs(SmoothZ - zscore)) %>%
  mutate(outlier = diff > 2) -> outliers

#Use rd_logratios
rd_factor <- readRDS("rd_factor.rds")
rd_factor %>% pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>% 
  mutate(logRatio = log(Wine/Oak)) -> rd_logratios

rd_logratios %>% select(Dataset, CHROM, POS, logRatio) %>% pivot_wider(names_from = Dataset, values_from = logRatio) %>% unnest() -> temp2

outliers %>% select(CHROM, POS, outlier, label) %>% pivot_wider(names_from = label, values_from = outlier) %>% unnest() %>% left_join(temp2) -> temp3

################################################################################
temp3 %>% 
  distinct() %>% count(POS) %>%
  merge(temp3) %>%
  filter(n > 1) %>% select(CHROM, POS) -> excl_duplicates
```

```{r}
excl_duplicates %>% mutate(excl = paste(CHROM, POS)) -> tempexcl
AllCuSO4_glmer_1byRep %>% mutate(excl = paste(CHROM, POS)) %>% filter(excl %in% tempexcl$excl == FALSE) -> AllCuSO4_glmer_1byRep_FILTERED

AllCuSO4_glmer_1byRep_FILTERED$CHROM <- factor(AllCuSO4_glmer_1byRep_FILTERED$CHROM, levels=as.character(as.roman(c(1:16,1000))))
sgd_copper$CHROM <- factor(sgd_copper$CHROM, levels=as.character(as.roman(c(1:16,1000))))
sgd_copper_eh$CHROM <- factor(sgd_copper_eh$CHROM, levels=as.character(as.roman(c(1:16,1000))))

AllCuSO4_glmer_1byRep_FILTERED %>% filter(label %in% c("Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) + 
  geom_point(aes(alpha = abs(zscore) > 1.01), color = "gray", size = 0.1) +
  geom_point(aes(alpha = abs(zscore) < 1.01), color = "black", size = 0.1) +
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
  geom_vline(data = sgd_copper_eh,
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             linetype = "dashed", size = 1, alpha = 0.3) +
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") + ggtitle("Interaction points") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))

AllCuSO4_glmer_1byRep_FILTERED %>% filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) + 
  geom_point(aes(alpha = abs(zscore) > 1.01), color = "gray", size = 0.1) +
  geom_point(aes(alpha = abs(zscore) < 1.01), color = "black", size = 0.1) +
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
  geom_vline(data = sgd_copper_eh,
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             linetype = "dashed", size = 1, alpha = 0.3) +
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") + ggtitle("Bulk points") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))
```
```{r, fig.height=6, fig.width=12}
AllCuSO4_glmer_1byRep_FILTERED %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), linetype = CSS)) + 
  geom_line() +
  geom_vline(data = sgd_copper, 
             aes(xintercept = Gene.chromosomeLocation.start, color = paste(CHROM,Gene.symbol)),
             size = 1, alpha = 0.3) +
    geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = "Gene", linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))
  
```

Looking at FRE1

```{r, fig.height=6, fig.width=12}
AllCuSO4_glmer_1byRep_FILTERED %>% filter(CHROM == "XII") %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), linetype = CSS)) + 
  geom_line() +
  geom_vline(aes(xintercept = 568567, color = "FRE1"),
             size = 1, alpha = 0.3) +
    geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))

AllCuSO4_glmer_1byRep_FILTERED %>% filter(CHROM == "VIII") %>%
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), linetype = CSS)) + 
  geom_line() +
  geom_vline(aes(xintercept = 212535, color = "CUP1"),
             size = 1, alpha = 0.3) +
    geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))

AllCuSO4_glmer_1byRep_FILTERED %>% 
  ungroup() %>%
  group_by(CHROM, CSS, label) %>%
  arrange(desc(abs(zscore))) %>%
  summarize(peakPOS = head(POS, 1),
            zscore = max(abs(zscore))) %>%
  merge(q5_all) %>%
  filter(abs(zscore) > q5) -> peaks

knowngenes <- data.frame(Gene = c("CUP1", "FRE1", "FRE1"),
                         POS = c(212535, 568567, 570627),
                         CHROM = c("VIII", "XII", "XII"))

AllCuSO4_glmer_1byRep_FILTERED %>% 
  filter(label %in% c("Bulk", "Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), linetype = CSS)) +
  #cutoffs and genes
  geom_vline(data = knowngenes, aes(xintercept = POS), color = "red", size = 1) +
  geom_vline(data = peaks[peaks$label %in% c("Bulk", "Interaction"),], 
             aes(xintercept = peakPOS, color = CHROM, linetype = CSS),
             size = 1, alpha = 0.8) +
    geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  #actual data
  geom_line() +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = "CSS", alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))

#peaks %>% filter(CHROM == "VIII")
#rd_logratios %>% filter(POS == 141525) %>% distinct()

AllCuSO4_glmer_1byRep_FILTERED %>% 
  filter(CHROM == "XII") %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), shape = CSS)) +
  #cutoffs and genes
  geom_rect(aes(xmin = 568567, xmax = 570627, ymin = -Inf, ymax = Inf), alpha = 0.2, fill = "lightblue") +
  geom_vline(data = peaks[peaks$label %in% c("Bulk") & peaks$CHROM == "XII",], 
             aes(xintercept = peakPOS, color = CHROM, linetype = CSS),
             size = 1, alpha = 0.8) +
    geom_hline(data = q5_all[q5_all$label %in% c("Bulk"),], aes(yintercept = q5)) +
  #actual data
  geom_point(alpha = 0.4, size = 1) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = "CSS", alpha = NULL) +
  theme(legend.position = "bottom") +
  xlim(520000, 600000) +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))+
  theme(axis.text.x=element_text())

AllCuSO4_glmer_1byRep_FILTERED %>% 
  filter(CHROM == "VIII") %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), shape = CSS)) +
  #cutoffs and genes
  geom_rect(aes(xmin = 212535, xmax = 212720, ymin = -Inf, ymax = Inf), alpha = 0.2, fill = "lightblue") +
  geom_rect(aes(xmin = 214533, xmax = 214718, ymin = -Inf, ymax = Inf), alpha = 0.2, fill = "lightblue") +
  geom_vline(data = peaks[peaks$label %in% c("Bulk") & peaks$CHROM == "VIII",], 
             aes(xintercept = peakPOS, color = CHROM, linetype = CSS),
             size = 1, alpha = 0.8) +
    geom_hline(data = q5_all[q5_all$label %in% c("Bulk"),], aes(yintercept = q5)) +
  #actual data
  geom_point(alpha = 0.4, size = 1) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = "CSS", alpha = NULL) +
  theme(legend.position = "bottom") +
  xlim(c(2e05, 2.25e05))+
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0))+
  theme(axis.text.x=element_text())

```

```{r}

peaks %>% filter(CHROM == "XII") %>%
  mutate(Dist = peakPOS - 570627) %>%
  arrange(abs(Dist)) -> FRE1_dist

FRE1_dist

peaks %>% filter(CHROM == "VIII") %>%
  mutate(Dist = peakPOS - 212535) %>%
  arrange(abs(Dist)) -> CUP1_dist

CUP1_dist


```

Zooming in on Chr XV

```{r}
AllCuSO4_glmer_1byRep_FILTERED %>% filter(CHROM == "XV") %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS/1000, y = abs(zscore), linetype = CSS)) + 
  geom_vline(data = peaks[peaks$label %in% c("Bulk") & peaks$CHROM == "XV",], 
             aes(xintercept = peakPOS/1000, color = CHROM, linetype = CSS),
             size = 1, alpha = 0.8) +
  geom_line() +
  geom_vline(xintercept = c(520, 560), color = "gray")+
    geom_hline(data = q5_all[q5_all$label %in% c("Bulk"),], aes(yintercept = q5)) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0)) +
  theme(axis.text.x=element_text())

AllCuSO4_glmer_1byRep_FILTERED %>% filter(CHROM == "XV") %>%
  filter(label %in% c("Interaction")) %>%
  ggplot(aes(x = POS/1000, y = abs(zscore), linetype = CSS)) + 
  geom_vline(data = peaks[peaks$label %in% c("Interaction") & peaks$CHROM == "XV",], 
             aes(xintercept = peakPOS/1000, color = CHROM, linetype = CSS),
             size = 1, alpha = 0.8) +
  geom_line() +
  geom_vline(xintercept = c(520, 560), color = "gray")+
    geom_hline(data = q5_all[q5_all$label %in% c("Interaction"),], aes(yintercept = q5)) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0)) +
  theme(axis.text.x=element_text())

AllCuSO4_glmer_1byRep_FILTERED %>% filter(CHROM == "XV") %>%
  filter(label %in% c("Interaction")) %>%
  ggplot(aes(x = POS/1000, y = abs(zscore), linetype = CSS)) +   
  geom_vline(data = peaks[peaks$label %in% c("Interaction") & peaks$CHROM == "XV",], 
             aes(xintercept = peakPOS/1000, color = CHROM, linetype = CSS),
             size = 1, alpha = 0.8) +
  geom_line() +
    geom_hline(data = q5_all[q5_all$label %in% c("Interaction"),], aes(yintercept = q5)) +
  facet_grid(label~CHROM, scales = "free", space = "free") +
  labs(color = NULL, linetype = NULL, alpha = NULL) +
  theme(legend.position = "bottom") +
    scale_size_manual(values = c(0.1, 2)) +
  scale_alpha_manual(values = c(0.4, 0)) +
  xlim(520, 560) +
  theme(axis.text.x=element_text())
```

# Peak finder based on actual change in direction

```{r}
AllCuSO4_glmer_1byRep_FILTERED %>% filter(label == "Bulk",
                                          CHROM == "XV",
                                          CSS == "VIII") -> testdata

testdata %>% ungroup() %>%
  reframe(POS = POS, 
          zscore = zscore,
          SmoothZ = frollapply(zscore, n = 300, FUN = mean, align = "center", na.rm = TRUE)) %>%
  ggplot(aes(x = POS, y = SmoothZ)) + geom_line()

testdata %>% ungroup() %>%
  reframe(POS = POS, 
          zscore = zscore,
          SmoothZ = frollapply(zscore, n = 300, FUN = mean, align = "center", na.rm = TRUE)) -> testder

findchange <- function(x){
  t <- length(x)
  diff <- x[t] - x[1]
  if(is.na(diff)){
    return(NA)
  }else if(diff == 0){
    return(0)
  }else{
    return(diff > 0)  
  }
}

testder %>% reframe(POS = POS,
                    zscore = zscore,
                    SmoothZ = SmoothZ,
                    change = frollapply(abs(SmoothZ), n = 4, findchange, align = "center")) %>%
  ggplot(aes(x = POS, y = change)) + geom_line() +
  geom_line(data = testdata, aes(x = POS, y = abs(zscore))) + 
  geom_line(data = testder, aes(x = POS, y = abs(SmoothZ)), color = "red", size = 2, alpha = 0.5)

findpeak <- function(x){
  t <- length(x)
  diff <- x[t] - x[1]
  return(diff)
}

testder %>% reframe(POS = POS,
                    zscore = zscore,
                    SmoothZ = SmoothZ,
                    change = frollapply(abs(SmoothZ), n = 4, findchange, align = "center")) %>%
  reframe(POS = POS,
                    zscore = zscore,
                    SmoothZ = SmoothZ,
          change = change,
                    peak = frollapply(abs(change), n = 2, findpeak, align = "center")) %>%
  filter(peak == -1) %>%
  ggplot(aes(x = POS)) + geom_vline(aes(xintercept = POS, color = factor(peak))) +
  geom_line(data = testdata, aes(x = POS, y = abs(zscore))) + 
  geom_line(data = testder, aes(x = POS, y = abs(SmoothZ)), color = "red", size = 2, alpha = 0.5)
  
```

Applying this to everythingg

```{r}
AllCuSO4_glmer_1byRep_FILTERED %>% 
  group_by(label, CSS, CHROM) %>%
  summarize(POS = POS,
            zscore = zscore,
            SmoothZ = frollapply(zscore, n = 600, FUN = mean, align = "center", na.rm = TRUE)) %>%
  summarize(POS = POS,
                    zscore = zscore,
                    SmoothZ = SmoothZ,
                    change = frollapply(abs(SmoothZ), n = 10, findchange, align = "center")) %>%
  summarize(POS = POS,
                    zscore = zscore,
                    SmoothZ = SmoothZ,
          change = change,
                    peak = frollapply(abs(change), n = 4, findpeak, align = "center")) %>%
  filter(peak == -1) %>% merge(q5_all) %>% filter(abs(zscore) > q5) -> TruePeaks

levels(TruePeaks$CHROM) <- as.character(as.roman(c(1:16, 1000)))

TruePeaks %>% 
  filter(label %in% c("Bulk", "Interaction")) %>% 
  ggplot(aes(x = POS, y = abs(SmoothZ), color = label)) + geom_vline(aes(xintercept = POS, color = label)) + geom_point() + facet_grid(CSS~CHROM, scales = "free", space = "free")

```

```{r, fig.width=12, fig.height=6}

# AllCuSO4_glmer_1byRep_FILTERED %>%
#   filter(label %in% c("Bulk", "Interaction")) %>%
#   ggplot(aes(x = POS, y = abs(zscore), color = CSS)) + 
#   geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk", "Interaction"),], 
#              aes(xintercept = POS, color = CSS), alpha = 0.2, linetype = "dashed")+
#   geom_line(alpha = 0.4, size = 1) +
#   geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
#   facet_grid(label~CHROM, scales = "free", space = "free") +
#   theme(legend.position = "bottom") +
#   scale_color_manual(values = c("navyblue", "darkorange")) +
#   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())  
# 
# AllCuSO4_glmer_1byRep_FILTERED %>%
#   filter(label %in% c("Bulk")) %>%
#   ggplot(aes(x = POS, y = abs(zscore), color = CSS)) + 
#   geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk"),], 
#              aes(xintercept = POS, color = CSS), alpha = 0.2, linetype = "dashed")+
#   geom_line(alpha = 0.4, size = 1) +
#   geom_hline(data = q5_all[q5_all$label %in% c("Bulk"),], aes(yintercept = q5)) +
#   facet_grid(CSS~CHROM, scales = "free", space = "free") +
#   theme(legend.position = "bottom") + ylim(0,12) +
#   scale_color_manual(values = c("navyblue", "darkorange")) +
#   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
#   ggtitle("Bulk Effect Peaks")

AllCuSO4_glmer_1byRep_FILTERED %>% 
  filter(label %in% c("Interaction")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CHROM %in% c("I", "III", "V", "VII", "IX", "XI", "XIII", "XV"))) + 
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Interaction"),], 
             aes(xintercept = POS, color = CHROM %in% c("I", "III", "V", "VII", "IX", "XI", "XIII", "XV")), alpha = 0.2, linetype = "dashed")+
  geom_line(alpha = 0.4, size = 1) +
  geom_hline(data = q5_all[q5_all$label %in% c("Interaction"),], aes(yintercept = q5)) +
  labs(color = NULL)+
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none") + ylim(0,12) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("Interaction Effect Peaks")

AllCuSO4_glmer_1byRep_FILTERED %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore), color = CHROM %in% c("I", "III", "V", "VII", "IX", "XI", "XIII", "XV"))) + 
  geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk"),], 
             aes(xintercept = POS, color = CHROM %in% c("I", "III", "V", "VII", "IX", "XI", "XIII", "XV")), alpha = 0.2, linetype = "dashed")+
  geom_line(alpha = 0.4, size = 1) +
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk"),], aes(yintercept = q5)) +
  labs(color = NULL)+
  facet_grid(CSS~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none") + ylim(0,12) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ggtitle("Bulk Effect Peaks")

```

