---
title: "MergedVCFs_ZivAIG_May24"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Load packages
require(ggplot2)
require(tidyr)
require(dplyr)
require(reshape2)
require(cowplot)
require(data.table)

require(doParallel)
require(RColorBrewer)
require(scales)
require(circlize) 
require(stringr)
require(ggrepel)

require(cybrBSA)
require(lme4)

ggplot2::theme_set(theme_minimal() + theme(text = element_text(size = 10)) +
                     theme(#axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()))

ChromosomeScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(Summary = NA, Label = NA) %>% select(-delete)

ChromosomeScale$Start <- 30000
ChromosomeScale$End <- ChromosomeScale$POS - 30000

ChromosomeScale2 <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                             levels = as.character(as.roman(1:16))),
                              start = rep(1, 16),
                              end = c(.23,.81,.32, 1.53, .58, .27, 1.09, .56, .44, .75, .67, 1.08, .92, .78, 1.09, .95)*1000000) %>% 
  pivot_longer(c(start, end), names_to = "delete", values_to = "POS") %>%
  mutate(summary = 0, label = "Bulk") %>% select(-delete)

# Loci to look at specifically

myloci <- data.frame(CHROM = c("V","V","IX", "III", "IV", "I", "III"),
                     POS = c(375215, 116167,389572, 198671, 46271, 203403, 260311),
                     name = c("FLO8", "URA3", "FLO11", "MAT", "HO", "FLO1", "TUP1"),
                     cat = c("Floc", "X", "Floc", "X", "X", "Floc", "Floc"))



## Functions for GLM all and Permutation all

glm_cb2_all <- function(..., W, formula, numgroups = FALSE, outputlength = 8) {
  data <- list(...)
  
  #Ensure that there is a formula and W parameter
  if (is.null(W) || is.null(formula)) {
    stop("Weights (W) and formula must be provided")
  }
  #Set formula
  glm_formula <- as.formula(formula)
  #Ensure that formula works for the data provided
  if (!all(names(data) %in% all.vars(glm_formula))) {
    stop("One or more variables in the formula are not provided as arguments")
  }
  
  #########################
  for(i in all.vars(glm_formula)){
    if(length(unique(as.data.frame(data)[,i])) < 2){
      output <- rep(NA, outputlength)
      #print("Not enough levels within groups")
      
      return(output)
    }
  }
  
  glm_fit <- glm(glm_formula, data = as.data.frame(data), weights = W, family = binomial)
  
  output <- summary(glm_fit)$coefficients[c(((length(summary(glm_fit)$coefficients)*0)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.25)),
                                            ((length(summary(glm_fit)$coefficients)*0.5)+1):
                                            ((length(summary(glm_fit)$coefficients)*0.75)))]

  
  if(length(output) == outputlength){
    return(output)
  }else{
    return(rep(NA, outputlength))
  }
  
} 


################################################################################
# No replicates, fix Parent
cybrPermute_cb2_all <- function(dataset, 
                                R = 0, perp = 1, outlength = 4, 
                                glmform = "Allele ~ Bulk * Parent", 
                                inputlabels = c("Intercept", "Bulk", "Parent", "Interaction")){
  start.time <- Sys.time()
    
  print("Make sure that dilute bulk is labeled aD")
  print(paste("Your labels are:", inputlabels))

    if(R > 0){ #INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Rep, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == 1) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest    
    }else{ #DO NOT INCLUDE REPLICATES
        dataset %>% 
        distinct() %>% ungroup() %>%
        group_by(CHROM, POS, Allele, Bulk, Parent) %>% 
        summarize(culprits = length((SmoothCount))) %>% 
        merge(dataset) %>% 
        filter(culprits == perp) %>% 
        ungroup() %>%
        distinct() %>% #THIS IS IMPORTANT
        pivot_wider(names_from = Allele, values_from = SmoothCount) -> newnewtest
    }
  
    #PERMUTE TWICE
  if(R > 0){
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Rep, 
             Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep, 
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, 
             Rep, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Rep = Rep,
                Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }else{
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "A") -> shuffled_DiluteA2
    
    newnewtest %>% filter(Bulk == "aD",
                          CHROM %in% c("I", "III", "V", "VIII", "M") == FALSE) %>% 
      select(CHROM, POS, Parent, Oak, Wine) %>% 
      group_by(Parent) %>% mutate(Loc = paste(CHROM, POS)) %>%
      summarize(Parent = Parent, Oak = Oak, Wine = Wine, Loc = sample(Loc)) %>%
      mutate(Bulk = "B") -> shuffled_DiluteB2
    
  }
    
    rbind(shuffled_DiluteA2, shuffled_DiluteB2) %>% pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") -> shuffletoglm2
    
    #RUN THE GLM
    if(R > 0) {
      shuffletoglm2 %>% na.omit() %>%
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 Rep = Rep,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }else{
      shuffletoglm2 %>% na.omit() %>%
      #Original Script
      group_by(Loc) %>%
      mutate_if(is.character, as.factor) %>%
        summarize(Summary = glm_cb2_all(Allele = Allele,
                                 Bulk = Bulk,
                                 Parent = Parent,
                                 W = SmoothCount,
                                 formula = glmform,
                                outputlength = length(inputlabels)*2),
                Factor = rep(inputlabels, 2),
                d = c(rep("Effect", length(inputlabels)),
                      rep("Z", length(inputlabels)))) -> glmresult
    }
    
    
    end.time = Sys.time()
    print(end.time - start.time)
    return(glmresult)
}

cybrInputGATKTable2 <- function(Ziv_rawData, yeast = TRUE){

  require(dplyr)
  require(doParallel)
  require(foreach)

  HNGLCDRXY <- read.table(Ziv_rawData, header = TRUE)

  #Identify the unique values besides AD/DP/GQ/PL
  gsub(".AD", "",
       gsub(".GQ", "",
            gsub(".DP","",
                 gsub(".PL","",
                      colnames(select(HNGLCDRXY, -CHROM, -POS, -REF, -ALT)))))) %>% unique() -> Samples
  #i <- Samples[1]

  resultscdf <- foreach(i=Samples,.combine=rbind) %dopar% {
    mydf <- HNGLCDRXY %>% select(CHROM, POS, REF, ALT) %>% mutate(Dataset = i)
    AD <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("AD"))
    GQ <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("GQ"))
    DP <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("DP"))
    PL <- select(HNGLCDRXY,matches(c(i), ignore.case = FALSE)) %>% select(., contains("PL"))
    cbind(mydf, AD , GQ , DP, PL) -> mydftotal
    colnames(mydftotal) <- c(colnames(mydf), "AD", "GQ", "DP", "PL")

    mydftotal %>% separate(AD, c('AD.REF','AD.ALT'), extra='drop') %>%
      separate(PL, c('PL.REF','PL.ALT'), extra='drop') %>%
      #Added 10/18/23:
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> mycdf

    mycdf %>% filter(grepl(",", ALT)) %>% 
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> doublecdf
    
    doublecdf %>% filter(grepl(",", ALT)) %>%
      separate(ALT, c("A1", "A2"), extra = 'merge') %>%
      separate(AD.ALT, c("AD1", "AD2"), extra = 'merge') %>%
      separate(PL.ALT, c("P1", "P2"), extra = 'merge') %>%
      
      pivot_longer(c(A1, A2), names_to = "NumAlt", values_to = "ALT") %>%
      pivot_longer(c(AD1, AD2), names_to = "NumADAlt", values_to = "AD.ALT") %>%
      pivot_longer(c(P1, P2), names_to = "NumPL", values_to = "PL.ALT") %>%
      mutate(NumAlt = gsub("A", "", NumAlt),
             NumADAlt = gsub("AD", "", NumADAlt),
             NumPL = gsub("P", "", NumPL)) %>%
      filter(NumAlt == NumPL,
             NumPL == NumADAlt) %>%
      select(CHROM, POS,REF,ALT,Dataset,AD.REF,AD.ALT,GQ,DP,PL.REF,PL.ALT) -> triplecdf
    
    rbind(mycdf, doublecdf, triplecdf) -> newcdf
    
    newcdf
  }

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    resultscdf %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> results
  }else{
    results <- resultscdf
  }
  return(results)

}

################################################################################

subtract <- function(POS){
  return(max(POS) - min(POS))
}

findchange <- function(x){
  t <- length(x)
  diff <- x[t] - x[1]
  if(is.na(diff)){
    return(NA)
  }else if(diff == 0){
    return(0)
  }else{
    return(diff > 0)  
  }
}

findpeak <- function(x){
  t <- length(x)
  diff <- x[t] - x[1]
  return(diff)
}

  
cybr_weightedgauss <- function(myx){
  myy <- dnorm(1:length(myx), mean = length(myx)/2, sd = 10)
  return(weighted.mean(x = myx, y = myy, na.rm = TRUE))
} 

################################################################################
sgd_orfs <- read.csv("C:/Users/Cassandra/Documents/R_FileTransfer/SGD_ORFs.csv")

genes <- c("CRS5", "SOD1", "CUP1-1", "CUP1-2", "MET17", "MUP1", "PCL1", "OYE3", "FRE1", "FRE7", "CTR1", "IRC7")
sgd_orfs %>% filter(Gene.symbol %in% genes) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_copper

genes_eh <- c("VMA4", "MTL1", "PTR2", "LEU9", "IRC15")
sgd_orfs %>% filter(Gene.symbol %in% genes_eh) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_copper_eh

#HUBS
sgd_orfs %>% filter(Gene.symbol %in% c("PTR3", "VPS70", "CUP1", "YLR257W")) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) %>% mutate(Hub = "Matsui") -> sgd_hubs_Matsui

sgd_orfs %>% filter(Gene.symbol %in% eQTL_hubs_Genes$GeneName) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) %>% mutate(Hub = "Albert") %>% arrange(CHROM) -> sgd_hubs_Albert

#Sporulation Genes
sgd_orfs %>% filter(Gene.symbol %in% c("RME1", "MKT1", "TAO3")) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) %>% mutate(Hub = "Sporulation") -> sgd_sporulation

```

## Load in table and match to oak or wine parents

```{r}
setwd("C:/Users/Cassandra/Documents/GitHub/Sequencing/Analysis/")

parentSNPids <- cybrConvertParentalAlleles(ParentFiles = c("Wine_VCF.txt", "Oak_VCF.txt"), Truncate = TRUE)
parentSNPids %>% group_by(CHROM, POS) %>% summarize(Type = Type, Unique = length(unique(Type))) -> pSNPs

```

```{r, message = FALSE, warning=FALSE}
#testfile <- "C:/Users/cassa/Downloads/AllZivAIG.REF_NC_001143.9.SortedCat.vcf.output.table"


myfile <- "C:/Users/Cassandra/Documents/Data_NYU/Sequencing/ZIV_AIGs.SortedCat.vcf.output.table"

read.table(myfile, header = TRUE) -> testZiv_rawdata

unique(testZiv_rawdata$CHROM)

cybrInputGATKTable2(myfile) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> Ziv_rawdata

```

## Actually determine which read is which

```{r}

#SMOOTHING ETC
Ziv_rawdata %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> Ziv_rawdata_called

# Ziv_rawdata_called %>% filter(CHROM == "II") %>%
#   ggplot(aes(x = POS, y = Dataset, color = Reads)) + geom_tile()

################################################################################
Ziv_rawdata_called %>% group_by(Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 300, FUN = median, align = "center", na.rm = TRUE))) -> Ziv_rawdata_smoothed #%>% na.omit()

# Ziv_rawdata_smoothed %>% filter(CHROM == "II") %>%
#   ggplot(aes(x = POS, y = Dataset, color = SmoothCount)) + geom_tile()
```


```{r, eval = FALSE}
#Checking to see how many of these are missing
Ziv_rawdata_smoothed %>% filter(CHROM == "II") %>%
  ggplot(aes(x = POS, y = Dataset, color = SmoothCount)) + geom_tile()

```

```{r, fig.width=12, fig.height=6}
Ziv_rawdata_smoothed %>% pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(logOdds = log(Wine/Oak)) -> Ziv_logodds

Ziv_logodds %>% na.omit() %>%
  ggplot(aes(x = POS, y = logOdds, color = Dataset)) + 
    geom_line(alpha = 0.8, size = 0.5) +
    facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "bottom")

```



# Loading in my AIG data

```{r}
myfile <- "C:/Users/Cassandra/Documents/Data_NYU/Sequencing/AllCuSO4.REF_.SortedCat.vcf.output.table"

#GET RID OF EXCLUDED ONES
read.table(myfile, header = TRUE) -> testrawdata
testrawdata %>% select(CHROM, POS, ALT) %>% separate(ALT, into = c("First", "Second"), sep = ",") %>% na.omit() -> exclude

ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                         "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                         CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                   "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                   "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                   "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

exclude %>% left_join(.,ChromKey) %>% select(-CHROM) %>% mutate(CHROM = chromosomes) %>% select(-chromosomes) -> exclude

exclude %>% mutate(locus = paste(CHROM, POS)) -> exclKEY

```


```{r}
CB_AIGs <- readRDS("AIG_CFZ_rawdata.rds")

#SMOOTHING ETC
CB_AIGs %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> CB_AIGs_called

CB_AIGs_called %>% mutate(locus = paste(CHROM, POS)) %>% filter(locus %in% exclKEY$locus == FALSE) %>% select(-locus) -> CB_AIGs_called

CB_AIGs_called %>% group_by(Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, SmoothCount = ceiling(frollapply(Reads, n = 300, FUN = median, align = "center", na.rm = TRUE))) -> CB_AIGs_smoothed 

CB_AIGs_smoothed %>%
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>% mutate(logOdds = log(Wine/Oak)) %>%
  mutate(Dataset = gsub("H5LHYDRX5_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Bulk", "Rep")) %>%
  filter(Bulk %in% c("C", "F", "Z", "U")) -> CB_AIGs_logOdds

CB_AIGs_called %>%
  pivot_wider(names_from = Allele, values_from = Reads) %>% unnest() %>% mutate(logOdds = log(Wine/Oak)) %>%
  mutate(Dataset = gsub("H5LHYDRX5_n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Bulk", "Rep")) %>%
  filter(Bulk %in% c("C", "F", "Z", "U")) -> CB_AIGs_logOdds_unsmoothed

```

```{r, fig.width=12, fig.height=6}
CB_AIGs_logOdds %>% na.omit() %>%
  ggplot(aes(x = POS, y = logOdds, color = Bulk)) + 
    geom_line(alpha = 0.4, size = 0.5) +
    facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none")
```

```{r, fig.width=12, fig.height=6}
# CB_AIGs_logOdds %>% mutate(Dataset = paste("CB_AIG ", Bulk, Rep)) %>%
#   select(Dataset, CHROM, POS, Oak, Wine, logOdds) %>%
#   rbind(Ziv_logodds) -> comparison

CB_AIGs_logOdds_unsmoothed %>% filter(abs(logOdds) != Inf) %>%
  mutate(Dataset = paste("CB_AIG ", Bulk, Rep)) %>%
  select(Dataset, CHROM, POS, Oak, Wine, logOdds) %>%
  rbind(Ziv_logodds) -> comparison

comparison %>% filter(CHROM %in% c("I", "M") == FALSE) %>%
  ggplot(aes(x = POS, y = logOdds, color = Dataset)) + 
    geom_line(alpha = 0.4, size = 0.5) +
    facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("orchid","orchid","orchid","orchid","orchid","orchid","orchid","orchid",
                                "black","black","black","black","black","black","black","black","black","black","black","black","black","black","black")) +
  theme(legend.position = "none") +
  ggtitle("logOdds of Ziv AIGs (black) vs CB AIGs (purple)")


```


## Look for sporulation genes

IME1, RME1, and RSF1

```{r, fig.width=12, fig.height=6}
sgd_orfs <- read.csv("C:/Users/Cassandra/Documents/R_FileTransfer/SGD_ORFs.csv")

genes <- c("IME1", "RME1", "RSF1")
sgd_orfs %>% filter(Gene.symbol %in% genes) %>%
  mutate(CHROM = gsub("chr", "", Gene.chromosome.primaryIdentifier)) -> sgd_spo

comparison %>% filter(CHROM %in% c("I", "M") == FALSE) %>%
  ggplot(aes(x = POS, y = logOdds, color = Dataset)) + 
    geom_line(alpha = 0.4, size = 0.5) +
  geom_hline(yintercept = 0) +
  geom_vline(data = sgd_spo, aes(xintercept = Gene.chromosomeLocation.start, linetype = Gene.symbol), 
             size = 1, alpha = 0.8, color = "lightblue") +
    facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("orchid","orchid","orchid","orchid","orchid","orchid","orchid","orchid",
                                "black","black","black","black","black","black","black","black","black","black","black","black","black","black","black")) +
  theme(legend.position = "bottom") +
  ggtitle("logOdds of Ziv AIGs (black) vs CB AIGs (purple)")


```




## What do the AIGs actually say?

```{r, fig.width=12, fig.height=6}
CB_AIGs_logOdds %>% filter(Bulk == "U") %>% group_by(CHROM, POS) %>% summarize(UMean = mean(logOdds)) -> Unselected_Avg

CB_AIGs_logOdds %>% merge(Unselected_Avg) %>% mutate(Diff = logOdds - UMean) -> logDiff

logDiff %>% filter(CHROM != "M") %>% ggplot(aes(x = POS, y = Diff, color = Bulk, linetype = Rep)) + 
    geom_line(alpha = 0.8, size = 0.5) +
  geom_hline(yintercept = 0) +
    facet_grid(~CHROM, scales = "free", space = "free")

logDiff %>% filter(CHROM == "VIII") %>% ggplot() + 
    geom_line(aes(x = POS, y = Diff, color = Bulk, linetype = Rep),alpha = 0.8, size = 0.5) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = c(120091,121683)) +
  geom_vline(xintercept = c(212535,212720,214533,214718)) +
    facet_grid(~CHROM, scales = "free", space = "free")

logDiff %>% filter(CHROM == "XII") %>% ggplot() + 
    geom_line(aes(x = POS, y = Diff, color = Bulk, linetype = Rep),alpha = 0.8, size = 0.5) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = c(568567,570627)) +
    facet_grid(~CHROM, scales = "free", space = "free")

CB_AIGs_logOdds %>% filter(CHROM == "VIII") %>% ggplot() + 
    geom_line(aes(x = POS, y = logOdds, color = Bulk, linetype = Rep),alpha = 0.8, size = 0.5) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = c(120091,121683)) +
  geom_vline(xintercept = c(212535,212720,214533,214718)) +
    facet_grid(~CHROM, scales = "free", space = "free")


```

```{r, fig.width=12, fig.height=6}
#CHROMOSOME VIII
CB_AIGs_logOdds_unsmoothed %>% group_by(Bulk, CHROM, POS) %>% summarize(avg = mean(logOdds)) %>% 
  filter(CHROM == "VIII", Bulk %in% c("U", "C"), POS < (562643 - 30000), POS > 30000) %>% 
  filter(abs(avg) != Inf) -> CHROM8_AIG_logOdds_unsmoothed

CB_AIGs_logOdds_unsmoothed %>% filter(CHROM == "VIII", Bulk %in% c("U", "C"), POS < (562643 - 30000), POS > 30000) %>% ggplot() + 
    geom_line(aes(x = POS, y = logOdds, color = Bulk, linetype = Rep),alpha = 0.8, size = 0.5) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = c(120091,121683), alpha = 0.3) +
  geom_vline(xintercept = c(212535,212720,214533,214718), alpha = 0.3) +
    facet_grid(~CHROM, scales = "free", space = "free")

CHROM8_AIG_logOdds_unsmoothed %>% 
  ggplot() + 
  geom_vline(xintercept = c(105586,105703), size = 2, color = "gray", alpha = 0.2) +
    geom_line(aes(x = POS, y = avg, color = Bulk),alpha = 0.8, size = 0.5) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = c(212535,212720,214533,214718), alpha = 0.3) +
    facet_grid(~CHROM, scales = "free", space = "free")

CHROM8_AIG_logOdds_unsmoothed %>% 
  ggplot() + 
  geom_vline(xintercept = c(105586,105703), size = 2, color = "gray", alpha = 0.2) + #CEN
    
  geom_point(aes(x = POS, y = avg, color = Bulk), size = 3, shape = 21) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = c(212535,212720,214533,214718), alpha = 0.3) + #CUP1
  xlim(105703+100000, 214718 + 105703 - 100000) + #zoom in
    facet_grid(~CHROM, scales = "free", space = "free")+
  theme(axis.text.x=element_text()) +
  scale_color_manual(values = c("darkturquoise", "black"))

CHROM8_AIG_logOdds_unsmoothed %>% 
  ggplot() + 
  geom_vline(xintercept = c(105586,105703), size = 2, color = "gray", alpha = 0.2) + #CEN
    
  geom_point(aes(x = POS, y = avg, color = Bulk), size = 3, shape = 21) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = c(212535,212720,214533,214718)) + #CUP1
  #geom_vline(xintercept = c(105703+80000, 214718 + 105703 - 80000), color = "red") + #zoom in
    facet_grid(~CHROM, scales = "free", space = "free") +
  scale_color_manual(values = c("darkturquoise", "black"))

CHROM8_AIG_logOdds_unsmoothed %>% filter(POS > (105703+80000), POS < 212535) %>% arrange(desc(avg))

```

```{r}
################################################################################
#Chromosome XII

CB_AIGs_logOdds_unsmoothed %>% filter(CHROM == "XII", Bulk %in% c("U", "C"), POS < (1078177 - 30000), POS > 30000) %>% ggplot() + 
    geom_line(aes(x = POS, y = logOdds, color = Bulk, linetype = Rep),alpha = 0.8, size = 0.5) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = c(568567,570627), alpha = 0.3) +
    facet_grid(~CHROM, scales = "free", space = "free")



CB_AIGs_logOdds_unsmoothed %>% group_by(Bulk, CHROM, POS) %>% summarize(avg = mean(logOdds)) %>% 
  filter(CHROM == "XII", Bulk %in% c("U", "C"), POS < (1078177 - 30000), POS > 30000) %>% 
  ggplot() + 
    geom_vline(xintercept = c(150828,150947), size = 2, color = "gray", alpha = 0.2) +
    geom_line(aes(x = POS, y = avg, color = Bulk),alpha = 0.8, size = 0.5) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = c(568567,570627), alpha = 0.3) +
    facet_grid(~CHROM, scales = "free", space = "free")

```

```{r}
CuSO4_CSS1_glmer_1byRep <- readRDS("CuSO4_CSS1_glmer_1byRepG.rds")
CuSO4_CSS8_glmer_1byRep <- readRDS("CuSO4_CSS8_glmer_1byRepG.rds")
q5_all <- readRDS("q5_CuSO4_CSS.rds")


CuSO4_CSS1_glmer_1byRep %>% filter(CHROM != "I") %>% mutate(CSS = "I") -> temp1
CuSO4_CSS8_glmer_1byRep %>% mutate(CSS = "VIII") %>%
  rbind(temp1) -> AllCuSO4_glmer_1byRep

```

```{r}
################################################################################
#Chromosome XVI
CB_AIGs_logOdds_unsmoothed %>% filter(CHROM == "XVI", Bulk %in% c("U", "C")) %>% ggplot() + 
    geom_line(aes(x = POS, y = abs(logOdds), color = Bulk, linetype = Rep),alpha = 0.8, size = 0.5) +
    geom_line(data = AllCuSO4_glmer_1byRep[AllCuSO4_glmer_1byRep$CHROM == "XVI" & AllCuSO4_glmer_1byRep$label == "Bulk",], aes(x = POS, y = abs(zscore), color = CSS)) +

  geom_hline(yintercept = 0) +
    facet_grid(~CHROM, scales = "free", space = "free")


CB_AIGs_logOdds_unsmoothed %>% group_by(Bulk, CHROM, POS) %>% select(-Coverage, -Oak, -Wine) %>% distinct() %>%
  filter(CHROM == "XVI", Bulk %in% c("U", "C"), POS < (1078177 - 30000), POS > 30000) %>% 
  pivot_wider(names_from = Bulk, values_from = logOdds) %>% unnest() %>%
  mutate(diff = C-U) %>%
  ggplot() + 
    geom_line(aes(x = POS, y = diff),alpha = 0.8, size = 0.5) +
  geom_line(data = AllCuSO4_glmer_1byRep[AllCuSO4_glmer_1byRep$CHROM == "XVI" & AllCuSO4_glmer_1byRep$label == "Bulk",], aes(x = POS, y = zscore, color = CSS)) +
  geom_hline(yintercept = 0) +
    facet_grid(~CHROM, scales = "free", space = "free")

```

# Kernel testing

```{r, eval = FALSE}
CHROM8_AIG_logOdds_unsmoothed %>% filter(Bulk == "C") %>% arrange(POS) -> testkernel

testkernelmat <- matrix(c(testkernel$POS, testkernel$avg), ncol = 2)
testpost <- smooth_ksmooth(testkernelmat, smoothness = 0.1, bandwidth = 100)

plot(testkernelmat[,1], testpost[,1])

dim(testkernelmat)
dim(testpost)

# plot(smooth_ksmooth(testkernelmat, smoothness = 0.1, bandwidth = 100), main = "0.1")
# plot(smooth_ksmooth(testkernelmat, smoothness = 1, bandwidth = 100), main = "1")
# plot(smooth_ksmooth(testkernelmat, smoothness = 10, bandwidth = 100), main = "10")
# plot(smooth_ksmooth(testkernelmat, smoothness = 100, bandwidth = 100), main = "100")

plot(smooth_ksmooth(testkernelmat, bandwidth = 100), main = "100")
plot(smooth_ksmooth(testkernelmat, bandwidth = 200), main = "200")
plot(smooth_ksmooth(testkernelmat, bandwidth = 500), main = "500")
plot(smooth_ksmooth(testkernelmat, bandwidth = 1000), main = "1000")
plot(smooth_ksmooth(testkernelmat, bandwidth = 5000), main = "5000")

smoothed_5k <- smooth_ksmooth(testkernelmat, bandwidth = 5000)

data.frame(smoothed_5k) %>% select(POS = X1, avg = X2) %>% mutate(TEST = "smoothed") -> test5k
testkernel %>% select(POS, avg) %>% mutate(TEST = "raw data") %>%
  rbind(test5k) %>%
  ggplot(aes(x = POS, y = avg, color = TEST)) + geom_line()

test5k %>% arrange(avg) %>% head()

#212354 - is this cup1?

212535 - 212354
```

# Smoothing after log odds

```{r}
Ziv_rawdata_called %>% pivot_wider(names_from = Allele, values_from = Reads) %>% unnest() %>%
  mutate(logOdds = log(Wine/Oak)) -> Ziv_logOdds_unsmoothed

CB_AIGs <- readRDS("AIG_CFZ_rawdata.rds")
CB_AIGs %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> CB_rawdata_called

CB_rawdata_called %>% pivot_wider(names_from = Allele, values_from = Reads) %>% unnest() %>%
  mutate(logOdds = log(Wine/Oak)) -> CBAIG_logOdds_unsmoothed

cybr_ksmooth <- function(z, p, b = 1000){
  result <- smoothr::smooth_ksmooth(matrix(c(z,p), ncol = 2), bandwidth = b)
  return(data.frame(result))
}
cybr_weightedgauss <- function(myx){
  myy <- dnorm(1:length(myx), mean = length(myx)/2, sd = 10)
  return(weighted.mean(x = myx, y = myy, na.rm = TRUE))
} 


#Using ksmooth
Ziv_logOdds_unsmoothed %>% ungroup() %>% 
  na.omit() %>%
  arrange(POS) %>%
  group_by(Dataset, CHROM) %>%
  summarize(POS = POS,
            LODsmooth = frollapply(logOdds, n = 200, FUN = cybr_weightedgauss, align = "center")) -> Ziv_logOdds_ksmooth

CBAIG_logOdds_unsmoothed %>% ungroup() %>% 
  na.omit() %>%
  arrange(POS) %>%
  group_by(Dataset, CHROM) %>%
  summarize(POS = POS,
            LODsmooth = frollapply(logOdds, n = 200, FUN = cybr_weightedgauss, align = "center")) -> CBAIG_logOdds_ksmooth


Ziv_logOdds_ksmooth %>% filter(abs(LODsmooth) != Inf) %>% ggplot(aes(x = POS, y = LODsmooth, color = Dataset)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")
CBAIG_logOdds_ksmooth %>% filter(abs(LODsmooth) != Inf) %>% ggplot(aes(x = POS, y = LODsmooth, color = Dataset)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")

```
Average these and find peaks

```{r, fig.width=10, fig.height=4}
CBAIG_logOdds_ksmooth %>% filter(abs(LODsmooth) != Inf) %>% ungroup() %>% group_by(CHROM, POS) %>% summarize(LODavg = mean(LODsmooth, na.rm = TRUE)) %>% mutate(ds = "CB") -> CBAIG_LODavg
Ziv_logOdds_ksmooth %>% filter(abs(LODsmooth) != Inf) %>% ungroup() %>% group_by(CHROM, POS) %>% summarize(LODavg = mean(LODsmooth, na.rm = TRUE)) %>% mutate(ds = "NZ") -> Ziv_LODavg

rbind(CBAIG_LODavg, Ziv_LODavg) %>%
  ggplot(aes(x = POS, y = LODavg, color = ds)) + geom_line() + facet_grid(~CHROM, scales = "free", space = "free")
  
rbind(CBAIG_LODavg, Ziv_LODavg) %>%
arrange(POS) %>%
  group_by(ds, CHROM) %>%
  summarize(cybr_ksmooth(z = LODavg, p = POS, b = 40000)) %>%
  summarize(POS = X2,
                    SmoothZ = X1,
                    change = frollapply(abs(X1), n = 100, findchange, align = "center")) %>%
  summarize(POS = POS,
                    SmoothZ = SmoothZ,
          change = change,
                    peak = frollapply(abs(change), n = 5, findpeak, align = "center")) -> smoothpeak_k_AIGs
  
rbind(CBAIG_LODavg, Ziv_LODavg) %>%
  arrange(POS) %>%
  group_by(ds, CHROM) %>%
  summarize(POS = POS,
                    SmoothZ = LODavg,
                    change = frollapply(abs(LODavg), n = 100, findchange, align = "center")) %>%
  summarize(POS = POS,
                    SmoothZ = SmoothZ,
          change = change,
                    peak = frollapply(abs(change), n = 5, findpeak, align = "center")) -> smoothpeak_AIGs

smoothpeak_AIGs %>% filter(peak == -1) -> AIGpeaks
smoothpeak_k_AIGs %>% filter(peak == -1) -> AIGpeaks

rbind(CBAIG_LODavg, Ziv_LODavg) %>%
  ggplot(aes(x = POS, y = LODavg, color = ds)) + geom_line() + 
  geom_vline(data = AIGpeaks, aes(xintercept = POS, color = ds)) +
  facet_grid(~CHROM, scales = "free", space = "free")
  
smoothpeak_k_AIGs %>%
  ggplot(aes(x = POS, y = SmoothZ, color = ds)) + geom_line(color = "black") +
    geom_vline(data = AIGpeaks, aes(xintercept = POS, color = ds)) +
  geom_hline(yintercept = 0) +
  facet_grid(ds~CHROM, scales = "free", space = "free")

# sgd_sporulation
# sgd_spo

smoothpeak_k_AIGs %>%
  ggplot(aes(x = POS, y = SmoothZ, color = ds)) +
    #geom_vline(data = AIGpeaks, aes(xintercept = POS, color = ds)) +
  geom_vline(data = sgd_sporulation, aes(xintercept = Gene.chromosomeLocation.start, linetype = Gene.symbol), 
             color = "skyblue3", size = 2, alpha = 0.4) +
  geom_vline(data = sgd_spo, aes(xintercept = Gene.chromosomeLocation.start, linetype = Gene.symbol), 
             color = "skyblue3", size = 2, alpha = 0.4) +
   geom_line() +
  ylab("Averaged Smoothed log(Wine/Oak)") +
  geom_hline(yintercept = 0) +
  facet_grid(~CHROM, scales = "free", space = "free")
```


# Using MultiPool

Python tool to be used in Ubuntu via WSL

```{python, eval = FALSE}
#Within Ubunt, navigate to environment
conda env list
activate MP2024_test2

#Check to make sure Python version is 2 and not 3
which conda

#Go to directory
cd /mnt/c/Users/Cassandra/Documents/GitHub/Sequencing/Analysis/Analysis_2024/

#Run either individual files or a loop of all of them
python mp_inference.py -n 10000 -r 2000 ../U_VII_b.txt ../C_VII_b.txt -m contrast -np -o bVII_out.txt

 for i in I II III IV V VI VII VIII IX X XI XII XIII XIV XV XVI; do python mp_inference.py -n 1000 C_B_$i.txt U_B_$i.txt -m contrast -np -o Bout_$i.txt; done

################################################################################
-n 10000 -r 2000 -m contrast -np #These are important
################################################################################
#CHANGE THE PARAMETERS
usage: mp_inference.py [-h] -n N [-m {replicates,contrast}] [-r RES] [-c CM]
                       [-t FILTER] [-np] [-o OUTFILE] [-v]
                       countfile [countfile ...]
    
Multipool: Efficient multi-locus genetic mapping with pooled sequencing,
version 0.10.2. See http://cgs.csail.mit.edu/multipool/ for more details.

positional arguments:
  countfile             Input file[s] of allele counts

optional arguments:
  -h, --help            show this help message and exit
  -n N, --individuals N
                        Individuals in each pool (required)
  -m {replicates,contrast}, --mode {replicates,contrast}
                        Mode for statistical testing. Default: replicates
  -r RES, --resolution RES
                        Bin size for discrete model. Default: 100 bp
  -c CM, --centimorgan CM
                        Length of a centimorgan, in base pairs. Default: 3300
                        (yeast average)
  -t FILTER, --truncate FILTER
                        Truncate possibly fixated (erroneous) markers.
                        Default: true
  -np, --noPlot         Turn off plotting output.. Default: false
  -o OUTFILE, --output OUTFILE
                        Output file for bin-level statistics
  -v, --version         show program's version number and exit
```


```{r}
#Produce text files for each chromosome

CB_AIGs <- readRDS("AIG_CFZ_rawdata.rds")

CB_AIGs %>% filter(CHROM == "VIII", Dataset %in% c("U_A", "U_B", "C_A", "C_B")) %>% ggplot(aes(x = POS, y = Coverage)) + geom_point() +geom_vline(xintercept = 212535, color = "red") 
CB_AIGs %>% #head(3000) %>%
  merge(parentSNPids) %>% mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Wine, Oak)  %>% filter(Dataset %in% c("H5LHYDRX5_n01_C_B.fastq",
                                            "H5LHYDRX5_n01_C_A.fastq",
                                            "H5LHYDRX5_n01_U_A.fastq",
                                            "H5LHYDRX5_n01_U_B.fastq")) -> mpsetup

mpsetup %>% mutate(Dataset = gsub("H5LHYDRX5_n01_", "", Dataset),
                   Dataset = gsub(".fastq", "", Dataset)) %>%
  na.omit() -> mpsetup

mpsetup %>% separate(Dataset, into = c("Selection", "Rep"), sep = "_") -> mpsetup_new

mpsetup_new %>% group_by(CHROM, POS, Selection, Rep) %>% mutate(MAF = min(Wine, Oak)/sum(Oak, Wine)) %>%
  filter(MAF > 0.1) -> mpsetup_maf

mpsetup_new %>% group_by(CHROM, POS, Selection, Rep) %>% mutate(MAF = min(Wine, Oak)/sum(Oak, Wine)) %>%
  filter(MAF < 0.1) -> mpsetup_maf_sub

mpsetup_maf %>% filter(Selection %in% c("C", "U"), CHROM == "VIII") %>% ggplot(aes(x = POS, y = Wine+Oak)) +
  geom_point(data = mpsetup_maf_sub[mpsetup_maf_sub$CHROM == "VIII",], color = "black") +
  geom_point(aes(color = Selection)) +geom_vline(xintercept = 212535, color = "red")

mpsetup_maf %>% filter(Selection %in% c("C", "U"), CHROM == "VIII") %>% ggplot(aes(x = POS, y = log(Wine/Oak))) +
  geom_point(data = mpsetup_maf_sub[mpsetup_maf_sub$CHROM == "VIII",], color = "black") +
  geom_point(aes(color = Selection)) +geom_vline(xintercept = 212535, color = "red")

# for(d in unique(mpsetup$Dataset)){
#   for(c in unique(mpsetup$CHROM)){
#    mpsetup %>% filter(CHROM == c, Dataset == d) %>%
#       arrange(POS) %>% ungroup() %>%
#       select(POS, Wine, Oak) -> temp
#     filename = paste(d,"_", c,".txt", sep = "")
#     write.table(temp, file = filename, col.names = FALSE, row.names = FALSE)
#   }
# }

for(d in unique(mpsetup_new$Selection)){
  for(c in unique(mpsetup_new$CHROM)){
   mpsetup_new %>% filter(CHROM == c, Selection == d) %>%
      arrange(POS) %>% ungroup() %>%
      select(POS, Wine, Oak) -> temp
    filename = paste("MultiPool_Inputs/", d,"_", c,"new.txt", sep = "")
    write.table(temp, file = filename, col.names = FALSE, row.names = FALSE)
  }
}

```

```{r}
rawdata_called_ex <- readRDS("rawdata_called_ex.rds")
rawdata_called_ex %>% pivot_wider(names_from = Allele, values_from = Reads) %>%
  select(CHROM, POS, Wine, Oak, Dataset) -> CuSO4_mpsetup

expnames <- read.csv("experiment_names_524.csv")

CuSO4_mpsetup %>% mutate(Dataset = gsub("n01_", "", Dataset)) %>%
  mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
  separate(Dataset, into = c("Pool", "Dataset"), sep = "_", extra = "merge") %>%
  merge(expnames) %>% unnest() %>% select(CHROM, POS, Bulk, Wine, Oak)-> CuSO4_mpsetup_final

CuSO4_mpsetup_final %>% na.omit() -> CuSO4_mpsetup_final

CuSO4_mpsetup_final %>% filter(Bulk != "Fluconazole") -> CuSO4_mpsetup_final

for(d in unique(CuSO4_mpsetup_final$Bulk)){
  for(c in unique(CuSO4_mpsetup_final$CHROM)){
   CuSO4_mpsetup_final %>% filter(CHROM == c, Bulk == d) %>%
      arrange(POS) %>% ungroup() %>%
      select(POS, Wine, Oak) -> temp
    filename = paste("MultiPool_Inputs/CUSO4all", d,"_", c,"new.txt", sep = "")
    write.table(temp, file = filename, col.names = FALSE, row.names = FALSE)
  }
}

```

```{r}
mpsetup_maf_sub %>%
  filter(CHROM == "VIII", Selection == "C") %>%
      arrange(POS) %>% ungroup() %>%
      select(POS, Wine, Oak) -> temp
  write.table(temp, file = "maf_C_VIII.txt", col.names = FALSE, row.names = FALSE)

  mpsetup_maf_sub %>%
  filter(CHROM == "VIII", Selection == "U") %>%
      arrange(POS) %>% ungroup() %>%
      select(POS, Wine, Oak) -> temp
  write.table(temp, file = "maf_U_VIII.txt", col.names = FALSE, row.names = FALSE)
  
  ###############################
  mpsetup_maf_sub %>%
  filter(CHROM == "XVI", Selection == "C") %>%
      arrange(POS) %>% ungroup() %>%
      select(POS, Wine, Oak) -> temp
  write.table(temp, file = "maf_C_XVI.txt", col.names = FALSE, row.names = FALSE)

  mpsetup_maf_sub %>%
  filter(CHROM == "XVI", Selection == "U") %>%
      arrange(POS) %>% ungroup() %>%
      select(POS, Wine, Oak) -> temp
  write.table(temp, file = "maf_U_XVI.txt", col.names = FALSE, row.names = FALSE)
  
```

Chr 7 Zooming in

```{r}
###############################
mpsetup_maf_sub %>%
filter(CHROM == "VII", Selection == "C") %>%
    arrange(POS) %>% ungroup() %>%
    select(POS, Wine, Oak) -> temp
write.table(temp, file = "maf_C_VII.txt", col.names = FALSE, row.names = FALSE)

mpsetup_maf_sub %>%
filter(CHROM == "VII", Selection == "U") %>%
    arrange(POS) %>% ungroup() %>%
    select(POS, Wine, Oak) -> temp
write.table(temp, file = "maf_U_VII.txt", col.names = FALSE, row.names = FALSE)

###############################
mpsetup_maf_sub %>% filter(POS < 661735, POS > 332360) %>%
filter(CHROM == "VII", Selection == "C") %>%
    arrange(POS) %>% ungroup() %>%
    select(POS, Wine, Oak) -> temp
write.table(temp, file = "maf_C_VII_a.txt", col.names = FALSE, row.names = FALSE)

mpsetup_maf_sub %>% filter(POS < 661735, POS > 332360) %>%
filter(CHROM == "VII", Selection == "U") %>%
    arrange(POS) %>% ungroup() %>%
    select(POS, Wine, Oak) -> temp
write.table(temp, file = "maf_U_VII_a.txt", col.names = FALSE, row.names = FALSE)

###############################
mpsetup_new %>% filter(POS < 661735, POS > 332360) %>%
filter(CHROM == "VII", Selection == "C") %>%
    arrange(POS) %>% ungroup() %>%
    select(POS, Wine, Oak) -> temp
write.table(temp, file = "maf_C_VII_b.txt", col.names = FALSE, row.names = FALSE)

mpsetup_new %>% filter(POS < 661735, POS > 332360) %>%
filter(CHROM == "VII", Selection == "U") %>%
    arrange(POS) %>% ungroup() %>%
    select(POS, Wine, Oak) -> temp
write.table(temp, file = "maf_U_VII_b.txt", col.names = FALSE, row.names = FALSE)

###############################
mpsetup_new %>% 
filter(CHROM == "VII", Selection == "C") %>%
    arrange(POS) %>% ungroup() %>%
    select(POS, Wine, Oak) -> temp
write.table(temp, file = "C_VII_b.txt", col.names = FALSE, row.names = FALSE)

mpsetup_new %>% 
filter(CHROM == "VII", Selection == "U") %>%
    arrange(POS) %>% ungroup() %>%
    select(POS, Wine, Oak) -> temp
write.table(temp, file = "U_VII_b.txt", col.names = FALSE, row.names = FALSE)

```

### How to run in python

python mp_inference.py -n 10000 -r 2000 ../U_VII_b.txt ../C_VII_b.txt -m contrast -np -o bVII_out.txt

Chr VII
```{r}
MAF_VII <- read.table("MultiPool_Inputs/VII_MAF_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "AIGmaf")

MAF_VII %>% ggplot(aes(x = Bin.start..bp., y = LOD.score)) + geom_line() +
  geom_vline(xintercept = c(171995, 172032, 172132, 187373, 187532, 188143, 205215, 205433, 205637), color = "red") + 
  geom_vline(xintercept = c(332360, 332484, 332582, 492468, 492815, 492864, 217776, 218225, 218227, 
                            484222, 484656, 484663, 661642, 661713, 661735)) +
  geom_vline(xintercept = c(332360, 332484, 332582, 661642, 661713, 661735), size = 2) +
  ggtitle("CHROM VII") -> MAF7_Plot

require(cowplot)  
cowplot::plot_grid(Chr7_Plot,MAF7_Plot, ncol = 1)

min(c(332360, 332484, 332582, 661642, 661713, 661735)) #332360
max(c(332360, 332484, 332582, 661642, 661713, 661735)) #661735


MAF_VIIa <- read.table("MultiPool_Inputs/aVII_MAF_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "AIGmaf")

MAF_VIIa %>% ggplot(aes(x = Bin.start..bp., y = LOD.score)) + geom_line() +
  geom_vline(xintercept = c(171995, 172032, 172132, 187373, 187532, 188143, 205215, 205433, 205637), color = "red") + 
  geom_vline(xintercept = c(332360, 332484, 332582, 492468, 492815, 492864, 217776, 218225, 218227, 
                            484222, 484656, 484663, 661642, 661713, 661735)) +
  geom_vline(xintercept = c(332360, 332484, 332582, 661642, 661713, 661735), size = 2) +
  ggtitle("CHROM VII")

MAF_VIIb <- read.table("MultiPool_Inputs/bVII_MAF_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "AIGmaf")

MAF_VIIb %>% ggplot(aes(x = Bin.start..bp., y = LOD.score)) + geom_line() +
  geom_vline(xintercept = c(171995, 172032, 172132, 187373, 187532, 188143, 205215, 205433, 205637), color = "red") + 
  geom_vline(xintercept = c(332360, 332484, 332582, 492468, 492815, 492864, 217776, 218225, 218227, 
                            484222, 484656, 484663, 661642, 661713, 661735)) +
  geom_vline(xintercept = c(332360, 332484, 332582, 661642, 661713, 661735), size = 2) +
  ggtitle("CHROM VII")

VIIb <- read.table("MultiPool_Inputs/bVII_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "AIGmaf")

VIIb %>% ggplot(aes(x = Bin.start..bp., y = LOD.score)) + geom_line() +
  geom_vline(xintercept = c(171995, 172032, 172132, 187373, 187532, 188143, 205215, 205433, 205637), color = "red") + 
  geom_vline(xintercept = c(332360, 332484, 332582, 492468, 492815, 492864, 217776, 218225, 218227, 
                            484222, 484656, 484663, 661642, 661713, 661735), color = "blue") +
  ggtitle("CHROM VII")
```

```{r, eval = FALSE}
MAF_VIII <- read.table("MultiPool_Inputs/VIII_MAF_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "AIGmaf")
A_VIII <- read.table("MultiPool_Inputs/New_out_VIII.txt", sep = "\t", header = TRUE) %>% mutate(exp = "AIG")

A_VIII %>% ggplot(aes(x = Bin.start..bp., y = LOD.score)) + geom_line() +
  geom_vline(xintercept = 212535, color = "red") + ggtitle("CHROM VIII")

read.table("MultiPool_Inputs/Aout_XII.txt", sep = "\t", header = TRUE) %>% ggplot(aes(x = Bin.start..bp., y = LOD.score)) + geom_line() +
  geom_vline(xintercept = 568567, color = "red") + ggtitle("CHROM XII")
```

Trying this with my own CuSO4 data actually

```{r, eval = FALSE}
rd_factor <- readRDS("rd_factorG.rds")

unique(rd_factor$Dataset)

testmultipool <- rd_factor %>% filter(Dataset %in% c("CuSO4_CSSI_SelectedA" ,  "CuSO4_CSSI_SelectedC" ,  "CuSO4_CSSI_UnselectedA", "CuSO4_CSSI_UnselectedC"))
################################################################################
testmultipool %>% filter(Background == "Oak", Bulk == "Copper", CHROM == "VIII") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) -> Oak_C_VIII
Oak_C_VIII %>% select(POS, Wine, Oak) %>% na.omit() -> temp
write.table(temp, file = "Oak_C_VIII.txt", col.names = FALSE, row.names = FALSE)

testmultipool %>% filter(Background == "Oak", Bulk == "Dilute", CHROM == "VIII") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) -> Oak_U_VIII
Oak_U_VIII %>% select(POS, Wine, Oak) %>% na.omit() -> temp
write.table(temp, file = "Oak_U_VIII.txt", col.names = FALSE, row.names = FALSE)

testmultipool %>% filter(Background == "Wine", Bulk == "Copper", CHROM == "VIII") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) -> Wine_C_VIII
Wine_C_VIII %>% select(POS, Wine, Oak) %>% na.omit() -> temp
write.table(temp, file = "Wine_C_VIII.txt", col.names = FALSE, row.names = FALSE)

testmultipool %>% filter(Background == "Wine", Bulk == "Dilute", CHROM == "VIII") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) -> Wine_U_VIII
Wine_U_VIII %>% select(POS, Wine, Oak) %>% na.omit() -> temp
write.table(temp, file = "Wine_U_VIII.txt", col.names = FALSE, row.names = FALSE)
################################################################################
OakVIII <- read.table("MultiPool_Inputs/VIII_Oak_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "CB_O")
WineVIII <- read.table("MultiPool_Inputs/VIII_Wine_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "CB_W")

rbind(OakVIII,WineVIII,MAF_VIII) %>% ggplot(aes(x = Bin.start..bp., y = LOD.score, color = exp)) + geom_line() +
  geom_vline(xintercept = 212535, color = "black") + ggtitle("CHROM VIII")

rbind(A_VIII, OakVIII,WineVIII, MAF_VIII) %>% ggplot(aes(x = Bin.start..bp., y = MLE.allele.freq., color = exp)) + geom_line() +
  geom_vline(xintercept = 212535, color = "black") + ggtitle("CHROM VIII")
```

Attempt with Chr XVI

```{r, eval = FALSE}
testmultipool <- rd_factor %>% filter(Dataset %in% c("CuSO4_CSSI_SelectedA" ,  "CuSO4_CSSI_SelectedC" ,  "CuSO4_CSSI_UnselectedA", "CuSO4_CSSI_UnselectedC"))
################################################################################
testmultipool %>% filter(Background == "Oak", Bulk == "Copper", CHROM == "XVI") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) -> Oak_C_XVI
Oak_C_XVI %>% select(POS, Wine, Oak) %>% na.omit() %>% unnest() -> temp
write.table(temp, file = "Oak_C_XVI.txt", col.names = FALSE, row.names = FALSE)

testmultipool %>% filter(Background == "Oak", Bulk == "Dilute", CHROM == "XVI") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) -> Oak_U_XVI
Oak_U_XVI %>% select(POS, Wine, Oak) %>% na.omit() %>% unnest() -> temp
write.table(temp, file = "Oak_U_XVI.txt", col.names = FALSE, row.names = FALSE)

testmultipool %>% filter(Background == "Wine", Bulk == "Copper", CHROM == "XVI") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) -> Wine_C_XVI
Wine_C_XVI %>% select(POS, Wine, Oak) %>% na.omit() %>% unnest() -> temp
write.table(temp, file = "Wine_C_XVI.txt", col.names = FALSE, row.names = FALSE)

testmultipool %>% filter(Background == "Wine", Bulk == "Dilute", CHROM == "XVI") %>% pivot_wider(names_from = Allele, values_from = SmoothCount) -> Wine_U_XVI
Wine_U_XVI %>% select(POS, Wine, Oak) %>% na.omit()  %>% unnest()-> temp
write.table(temp, file = "Wine_U_XVI.txt", col.names = FALSE, row.names = FALSE)
################################################################################
OakXVI <- read.table("MultiPool_Inputs/XVI_Oak_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "CB_O")
WineXVI <- read.table("MultiPool_Inputs/XVI_Wine_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "CB_W")
MAF_XVI <- read.table("MultiPool_Inputs/XVI_MAF_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "AIGmaf")


rbind(OakXVI,WineXVI,MAF_XVI) %>% ggplot(aes(x = Bin.start..bp., y = LOD.score, color = exp)) + geom_line() + ggtitle("CHROM XVI")

rbind(OakXVI,WineXVI, MAF_XVI) %>% ggplot(aes(x = Bin.start..bp., y = MLE.allele.freq., color = exp)) + geom_line() + ggtitle("CHROM XVI")
```

### Running everything with a larger window size and population

```{r, fig.width=12, fig.height=4}
test8 <- read.table("MultiPool_Inputs/2kr_out_VIII.txt", sep = "\t", header = TRUE) %>% mutate(CHROM = "VIII")
AllMP <- head(test8,1)
for(i in as.character(as.roman(1:16))){
  filename <- paste("MultiPool_Inputs/2kr_out_", i, ".txt", sep = "")
  tempfile <- read.table(filename, sep = "\t", header = TRUE) %>% mutate(CHROM = i)
  
  AllMP <- rbind(AllMP, tempfile)
}


AllMP %>% ggplot(aes(x = Bin.start..bp., y = LOD.score, color = CHROM)) + geom_line() + ggtitle("CHROM VIII") +
  facet_wrap("CHROM", scales = "free")

AllMP %>% ggplot(aes(x = Bin.start..bp., y = LOD.score)) + geom_line() + ggtitle("CHROM VIII") +
  facet_grid(~CHROM, scales = "free", space = "free") +
  ylim(0,100)

AllMP %>% distinct() %>%
  group_by(CHROM) %>%
  summarize(LOD = LOD.score, MLE = MLE.allele.freq.,
            POS = Bin.start..bp.,
            max = max(Bin.start..bp.)) %>%
  filter(POS < max - 30000, POS > 30000) %>%
  filter(LOD < 75) %>%
  ggplot(aes(x = POS, y = LOD, color = CHROM)) + geom_line() + ggtitle("AIG, n = 10k, r = 2k") +
  facet_grid(~CHROM, scales = "free", space = "free")  +
  theme(legend.position = "none")

```



Fewer individuals (n = 1000 instead of 10000)
```{r}
AllMP_few <- head(test8,1)
for(i in as.character(as.roman(1:16))){
  filename <- paste("MultiPool_Inputs/2kr_few_out_", i, ".txt", sep = "")
  tempfile <- read.table(filename, sep = "\t", header = TRUE) %>% mutate(CHROM = i)
  
  AllMP_few <- rbind(AllMP, tempfile)
}

AllMP_few4k <- head(test8, 1)
for(i in as.character(as.roman(1:16))){
  filename <- paste("MultiPool_Inputs/4kr_few_out_", i, ".txt", sep = "")
  tempfile <- read.table(filename, sep = "\t", header = TRUE) %>% mutate(CHROM = i)
  
  AllMP_few4k <- rbind(AllMP_few4k, tempfile)
}

AllMP_few8k <- head(test8, 1)
for(i in as.character(as.roman(1:16))){
  filename <- paste("MultiPool_Inputs/8kr_few_out_", i, ".txt", sep = "")
  tempfile <- read.table(filename, sep = "\t", header = TRUE) %>% mutate(CHROM = i)
  
  AllMP_few8k <- rbind(AllMP_few8k, tempfile)
}
```


```{r, fig.width=12, fig.height=4}
rbind(
      data.frame(AllMP_few, W = "2k"),
      data.frame(AllMP_few4k, W = "4k"),
      data.frame(AllMP_few8k, W = "8k")) %>%
  group_by(CHROM) %>%
  summarize(LOD = LOD.score, MLE = MLE.allele.freq., W = W,
            POS = Bin.start..bp.,
            max = max(Bin.start..bp.)) %>%
  filter(POS < max - 30000, POS > 30000) %>%
  ungroup() %>%
  group_by(W) %>%
  summarize(CHROM = CHROM, LOD = LOD, MLE = MLE, POS = POS, maxLOD = max(LOD)) %>%
  mutate(LOD_scaled = LOD/maxLOD) %>%
  ggplot() +
  geom_line(aes(x = POS, y = LOD_scaled, color = W)) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  ylim(0,0.5)

rbind(
      data.frame(AllMP_few, W = "2k"),
      data.frame(AllMP_few4k, W = "4k"),
      data.frame(AllMP_few8k, W = "8k")) %>%
  group_by(CHROM) %>%
  summarize(LOD = LOD.score, MLE = MLE.allele.freq., W = W,
            POS = Bin.start..bp.,
            max = max(Bin.start..bp.)) %>%
  filter(POS < max - 30000, POS > 30000) %>%
  ungroup() %>%
  group_by(W) %>%
  summarize(CHROM = CHROM, LOD = LOD, MLE = MLE, POS = POS, maxLOD = max(LOD)) %>%
  mutate(LOD_scaled = LOD/maxLOD) %>%
  filter(CHROM == "VII") %>%
  ggplot() +
  geom_point(aes(x = POS, y = LOD_scaled, color = W)) +
  facet_grid(~CHROM, scales = "free", space = "free") 

```

```{r, fig.width=12, fig.height=4}
AllMP_few %>% distinct() %>%
  group_by(CHROM) %>%
  summarize(LOD = LOD.score, MLE = MLE.allele.freq.,
            POS = Bin.start..bp.,
            max = max(Bin.start..bp.)) %>%
  filter(POS < max - 30000, POS > 30000) -> addtoplot


AllCuSO4_glmer_1byRep %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) +
  # geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk"),],
  #            aes(xintercept = POS), alpha = 0.2)+
  # geom_vline(data = TruePeaks[TruePeaks$label %in% c("Interaction"),],
             # aes(xintercept = POS), alpha = 0.2, color = "red", size = 2)+
  geom_line(aes(color = CSS)) +
  geom_point(data = addtoplot, aes(x = POS, y = LOD/5), size = 0.1, alpha = 0.3) +
  #geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_wrap("CHROM", scales = "free") +
  theme(legend.position = "none") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) 


AllCuSO4_glmer_1byRep %>%
  filter(label %in% c("Bulk")) %>%
  ggplot(aes(x = POS, y = abs(zscore))) +
  # geom_vline(data = TruePeaks[TruePeaks$label %in% c("Bulk"),],
  #            aes(xintercept = POS), alpha = 0.2)+
  # geom_vline(data = TruePeaks[TruePeaks$label %in% c("Interaction"),],
             # aes(xintercept = POS), alpha = 0.2, color = "red", size = 2)+
  geom_line(aes(color = CSS)) +
  geom_line(data = addtoplot, aes(x = POS, y = LOD/5)) +
  #geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_grid(~CHROM, scales = "free") +
  theme(legend.position = "none") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ylim(0, 20)

```

```{r, fig.width=6, fig.height=4}
AllMP_few %>% distinct() %>%
  group_by(CHROM) %>%
  summarize(LOD = LOD.score, MLE = MLE.allele.freq.,
            POS = Bin.start..bp.,
            max = max(Bin.start..bp.)) %>%
  filter(POS < max - 30000, POS > 30000, CHROM == "VII") -> addtoplot2

AllCuSO4_glmer_1byRep %>%
  filter(label %in% c("Bulk", "Interaction"), CHROM == "VII") %>%
  ggplot(aes(x = POS, y = abs(zscore))) +
  geom_line(aes(color = label, linetype = CSS)) +
  geom_line(data = addtoplot2, aes(x = POS, y = LOD/5)) +
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_color_manual(values = c("gray", "purple"))

AllMP_few %>% distinct() %>%
  group_by(CHROM) %>%
  summarize(LOD = LOD.score, MLE = MLE.allele.freq.,
            POS = Bin.start..bp.,
            max = max(Bin.start..bp.)) %>%
  filter(POS < max - 30000, POS > 30000, CHROM == "VIII") -> addtoplot2
AllCuSO4_glmer_1byRep %>%
  filter(label %in% c("Bulk"), CHROM == "VIII") %>%
  ggplot(aes(x = POS, y = abs(zscore))) +
  geom_line(aes(color = label, linetype = CSS)) +
  geom_line(data = addtoplot2, aes(x = POS, y = LOD/5)) +
  geom_hline(data = q5_all[q5_all$label %in% c("Bulk", "Interaction"),], aes(yintercept = q5)) +
  labs(linetype = NULL) +
  facet_grid(~CHROM, scales = "free", space = "free") +
  theme(legend.position = "none") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  scale_color_manual(values = c("gray", "purple"))

```

```{r}
OakVIII <- read.table("MultiPool_Inputs/VIII_Oak_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "CB_O")
WineVIII <- read.table("MultiPool_Inputs/VIII_Wine_out.txt", sep = "\t", header = TRUE) %>% mutate(exp = "CB_W")

AllMP_few %>% filter(CHROM == "VIII") %>%
  summarize(LOD.score = LOD.score, MLE.allele.freq. = MLE.allele.freq.,
            Bin.start..bp. = Bin.start..bp.,
            max = max(Bin.start..bp.)) %>%
    filter(Bin.start..bp. < max - 30000, Bin.start..bp. > 30000) %>%
  select(-max) %>%

  mutate(exp = "AIG") %>%
  rbind(OakVIII,WineVIII) %>% ggplot(aes(x = Bin.start..bp., y = LOD.score, color = exp)) + geom_line() + ggtitle("CHROM 8, n = 1k, r = 2k")

```

## My actual data run through multipool

```{r, fig.width=12, fig.height=4}
test8 <- read.table("MultiPool_Inputs/CuSO4_all_out_I.txt", sep = "\t", header = TRUE) %>% mutate(CHROM = "4")

Cu_MP <- head(test8,1)
for(i in as.character(as.roman(1:16))){
  filename <- paste("MultiPool_Inputs/CuSO4_all_out_", i, ".txt", sep = "")
  tempfile <- read.table(filename, sep = "\t", header = TRUE) %>% mutate(CHROM = i)
  
  Cu_MP <- rbind(Cu_MP, tempfile)
}

Cu_MP %>% ggplot(aes(x = Bin.start..bp., y = LOD.score, color = CHROM)) + geom_line() + ggtitle("CuSO4 Multipool, n = 10k, r = 2k") +
  facet_grid(~CHROM, scales = "free", space = "free")  +
    geom_line(data = addtoplot, aes(x = POS, y = LOD/5), color = "black") +

  theme(legend.position = "none") +
  ylim(0,20)


Cu_MP %>% ggplot(aes(x = Bin.start..bp., y = MLE.allele.freq., color = CHROM)) + geom_line() + ggtitle("CuSO4 Multipool, n = 10k, r = 2k") +
  facet_grid(~CHROM, scales = "free", space = "free")  +
    geom_line(data = addtoplot, aes(x = POS, y = MLE), color = "black") +
geom_hline(yintercept = 0.5, linetype = "dashed")+
  theme(legend.position = "none") 

Cu_MP %>% filter(CHROM == "VIII") %>% ggplot(aes(x = Bin.start..bp., y = LOD.score, color = CHROM)) + geom_line() + ggtitle("CuSO4 Multipool, n = 10k, r = 2k") +
  facet_grid(~CHROM, scales = "free", space = "free")  +
    geom_line(data = addtoplot[addtoplot$CHROM == "VIII",], aes(x = POS, y = LOD), color = "black") +
geom_hline(yintercept = 0.5, linetype = "dashed")+
  theme(legend.position = "none") 
```

